<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Calculate Clonal Diversity — clonalDiversity • scRepertoire</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate Clonal Diversity — clonalDiversity"><meta property="og:description" content="This function calculates a specified diversity metric for samples or groups
within a dataset. To control for variations in library size, the function
can perform bootstrapping with downsampling. It resamples each group to the
size of the smallest group and calculates the diversity metric across
multiple iterations, returning the mean value."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/dandelionR.html">Immune Repertoire Trajectory Analysis</a>
    </li>
    <li>
      <a href="../articles/immunarch.html">Using scRepertoire with immunarch</a>
    </li>
    <li>
      <a href="../articles/tcrpheno.html">TCRpheno applied to T cell fate</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/BorchLab/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate Clonal Diversity</h1>
    <small class="dont-index">Source: <a href="https://github.com/BorchLab/scRepertoire/blob/HEAD/R/clonalDiversity.R" class="external-link"><code>R/clonalDiversity.R</code></a></small>
    <div class="hidden name"><code>clonalDiversity.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function calculates a specified diversity metric for samples or groups
within a dataset. To control for variations in library size, the function
can perform bootstrapping with downsampling. It resamples each group to the
size of the smallest group and calculates the diversity metric across
multiple iterations, returning the mean value.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">clonalDiversity</span><span class="op">(</span></span>
<span>  <span class="va">input.data</span>,</span>
<span>  cloneCall <span class="op">=</span> <span class="st">"strict"</span>,</span>
<span>  metric <span class="op">=</span> <span class="st">"shannon"</span>,</span>
<span>  chain <span class="op">=</span> <span class="st">"both"</span>,</span>
<span>  group.by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  order.by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  x.axis <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  exportTable <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  palette <span class="op">=</span> <span class="st">"inferno"</span>,</span>
<span>  n.boots <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  return.boots <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  skip.boots <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-input-data">input.data<a class="anchor" aria-label="anchor" href="#arg-input-data"></a></dt>
<dd><p>The product of <code><a href="combineTCR.html">combineTCR()</a></code>,
<code><a href="combineBCR.html">combineBCR()</a></code>, or <code><a href="combineExpression.html">combineExpression()</a></code>.</p></dd>


<dt id="arg-clonecall">cloneCall<a class="anchor" aria-label="anchor" href="#arg-clonecall"></a></dt>
<dd><p>Defines the clonal sequence grouping. Accepted values
are: <code>gene</code> (VDJC genes), <code>nt</code> (CDR3 nucleotide sequence), <code>aa</code> (CDR3 amino
acid sequence), or <code>strict</code> (VDJC). A custom column header can also be used.</p></dd>


<dt id="arg-metric">metric<a class="anchor" aria-label="anchor" href="#arg-metric"></a></dt>
<dd><p>The diversity metric to calculate. Must be a single string from
the list of available metrics (see Details).</p></dd>


<dt id="arg-chain">chain<a class="anchor" aria-label="anchor" href="#arg-chain"></a></dt>
<dd><p>The TCR/BCR chain to use. Use <code>both</code> to include both chains
(e.g., TRA/TRB). Accepted values: <code>TRA</code>, <code>TRB</code>, <code>TRG</code>, <code>TRD</code>, <code>IGH</code>, <code>IGL</code>
(for both light chains), <code>both</code>.</p></dd>


<dt id="arg-group-by">group.by<a class="anchor" aria-label="anchor" href="#arg-group-by"></a></dt>
<dd><p>A column header in the metadata or lists to group the analysis
by (e.g., "sample", "treatment"). If <code>NULL</code>, data will be analyzed
by list element or active identity in the case of single-cell objects.</p></dd>


<dt id="arg-order-by">order.by<a class="anchor" aria-label="anchor" href="#arg-order-by"></a></dt>
<dd><p>A character vector defining the desired order of elements
of the <code>group.by</code> variable. Alternatively, use <code>alphanumeric</code> to sort groups
automatically.</p></dd>


<dt id="arg-x-axis">x.axis<a class="anchor" aria-label="anchor" href="#arg-x-axis"></a></dt>
<dd><p>An additional metadata variable to group samples along the x-axis.</p></dd>


<dt id="arg-exporttable">exportTable<a class="anchor" aria-label="anchor" href="#arg-exporttable"></a></dt>
<dd><p>If <code>TRUE</code>, returns a data frame or matrix of the results
instead of a plot.</p></dd>


<dt id="arg-palette">palette<a class="anchor" aria-label="anchor" href="#arg-palette"></a></dt>
<dd><p>Colors to use in visualization - input any
hcl.pals.</p></dd>


<dt id="arg-n-boots">n.boots<a class="anchor" aria-label="anchor" href="#arg-n-boots"></a></dt>
<dd><p>The number of bootstrap iterations to perform (default is 100).</p></dd>


<dt id="arg-return-boots">return.boots<a class="anchor" aria-label="anchor" href="#arg-return-boots"></a></dt>
<dd><p>If <code>TRUE</code>, returns all bootstrap values instead of the
mean. Automatically enables <code>exportTable</code>.</p></dd>


<dt id="arg-skip-boots">skip.boots<a class="anchor" aria-label="anchor" href="#arg-skip-boots"></a></dt>
<dd><p>If <code>TRUE</code>, disables downsampling and bootstrapping. The
metric will be calculated on the full dataset for each group. Defaults to <code>FALSE</code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments passed to the ggplot theme</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A ggplot object visualizing the diversity metric, or a data.frame if
<code>exportTable = TRUE</code>.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The function operates by first splitting the dataset by the specified <code>group.by</code>
variable.</p>
<p><strong>Downsampling and Bootstrapping:</strong>
To make a fair comparison between groups of different sizes, diversity metrics
often require normalization. This function implements this by downsampling.</p><ol><li><p>It determines the number of clones in the smallest group.</p></li>
<li><p>For each group, it performs <code>n.boots</code> iterations (default = 100).</p></li>
<li><p>In each iteration, it randomly samples the clones (with replacement) down to
the size of the smallest group.</p></li>
<li><p>It calculates the selected diversity metric on this downsampled set.</p></li>
<li><p>The final reported diversity value is the mean of the results from all
bootstrap iterations.</p></li>
</ol><p>This process can be disabled by setting <code>skip.boots = TRUE</code>.</p>
<p>Available Diversity Metrics (metric):
The function uses a registry of metrics imported from the immApex package.
You can select one of the following:</p><ul><li><p><code>"shannon"</code>: Shannon's Entropy. See <code><a href="https://rdrr.io/pkg/immApex/man/shannon_entropy.html" class="external-link">shannon_entropy</a></code>.</p></li>
<li><p><code>"inv.simpson"</code>: Inverse Simpson Index. See <code><a href="https://rdrr.io/pkg/immApex/man/inv_simpson.html" class="external-link">inv_simpson</a></code>.</p></li>
<li><p><code>"gini.simpson"</code>: Gini-Simpson Index. See <code><a href="https://rdrr.io/pkg/immApex/man/gini_simpson.html" class="external-link">gini_simpson</a></code>.</p></li>
<li><p><code>"norm.entropy"</code>: Normalized Shannon Entropy. See <code><a href="https://rdrr.io/pkg/immApex/man/norm_entropy.html" class="external-link">norm_entropy</a></code>.</p></li>
<li><p><code>"pielou"</code>: Pielou's Evenness (same as norm.entropy). See <code><a href="https://rdrr.io/pkg/immApex/man/pielou_evenness.html" class="external-link">pielou_evenness</a></code>.</p></li>
<li><p><code>"ace"</code>: Abundance-based Coverage Estimator. See <code><a href="https://rdrr.io/pkg/immApex/man/ace_richness.html" class="external-link">ace_richness</a></code>.</p></li>
<li><p><code>"chao1"</code>: Chao1 Richness Estimator. See <code><a href="https://rdrr.io/pkg/immApex/man/chao1_richness.html" class="external-link">chao1_richness</a></code>.</p></li>
<li><p><code>"gini"</code>: Gini Coefficient for inequality. See <code><a href="https://rdrr.io/pkg/immApex/man/gini_coef.html" class="external-link">gini_coef</a></code>.</p></li>
<li><p><code>"d50"</code>: The number of top clones making up 50% of the library. See <code><a href="https://rdrr.io/pkg/immApex/man/d50_dom.html" class="external-link">d50_dom</a></code>.</p></li>
<li><p><code>"hill0"</code>, <code>"hill1"</code>, <code>"hill2"</code>: Hill numbers of order 0, 1, and 2. See <code><a href="https://rdrr.io/pkg/immApex/man/hill_q.html" class="external-link">hill_q</a></code>.</p></li>
</ul></div>
    <div id="author">
    <h2>Author</h2>
    <p>Andrew Malone, Nick Borcherding, Nathan Vanderkraan</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Making combined contig data</span></span></span>
<span class="r-in"><span><span class="va">combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="combineTCR.html">combineTCR</a></span><span class="op">(</span><span class="va">contig_list</span>,</span></span>
<span class="r-in"><span>                       samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17B"</span>, <span class="st">"P17L"</span>, <span class="st">"P18B"</span>, <span class="st">"P18L"</span>,</span></span>
<span class="r-in"><span>                                   <span class="st">"P19B"</span>,<span class="st">"P19L"</span>, <span class="st">"P20B"</span>, <span class="st">"P20L"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Calculate Shannon diversity, grouped by sample</span></span></span>
<span class="r-in"><span><span class="fu">clonalDiversity</span><span class="op">(</span><span class="va">combined</span>, </span></span>
<span class="r-in"><span>                cloneCall <span class="op">=</span> <span class="st">"gene"</span>, </span></span>
<span class="r-in"><span>                metric <span class="op">=</span> <span class="st">"shannon"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="clonalDiversity-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Calculate Inverse Simpson without bootstrapping</span></span></span>
<span class="r-in"><span><span class="fu">clonalDiversity</span><span class="op">(</span><span class="va">combined</span>, </span></span>
<span class="r-in"><span>                cloneCall <span class="op">=</span> <span class="st">"aa"</span>, </span></span>
<span class="r-in"><span>                metric <span class="op">=</span> <span class="st">"inv.simpson"</span>, </span></span>
<span class="r-in"><span>                skip.boots <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="clonalDiversity-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

