<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Cluster clones by sequence similarity — clonalCluster • scRepertoire</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Cluster clones by sequence similarity — clonalCluster"><meta property="og:description" content="This function clusters TCRs or BCRs based on the edit distance of their CDR3
sequences. It can operate on either nucleotide (nt) or amino acid (aa)
sequences and can optionally enforce that clones share the same V and/or J
genes. The output can be the input object with an added metadata column for
cluster IDs, a sparse adjacency matrix, or an igraph graph object
representing the cluster network."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/dandelionR.html">VDJ Trajectory Analysis with dandelionR</a>
    </li>
    <li>
      <a href="../articles/immunarch.html">Using scRepertoire with immunarch</a>
    </li>
    <li>
      <a href="../articles/tcrpheno.html">TCRpheno applied to T cell fate</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/BorchLab/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Cluster clones by sequence similarity</h1>
    <small class="dont-index">Source: <a href="https://github.com/BorchLab/scRepertoire/blob/HEAD/R/clonalCluster.R" class="external-link"><code>R/clonalCluster.R</code></a></small>
    <div class="hidden name"><code>clonalCluster.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function clusters TCRs or BCRs based on the edit distance of their CDR3
sequences. It can operate on either nucleotide (<code>nt</code>) or amino acid (<code>aa</code>)
sequences and can optionally enforce that clones share the same V and/or J
genes. The output can be the input object with an added metadata column for
cluster IDs, a sparse adjacency matrix, or an <code>igraph</code> graph object
representing the cluster network.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">clonalCluster</span><span class="op">(</span></span>
<span>  <span class="va">input.data</span>,</span>
<span>  chain <span class="op">=</span> <span class="st">"TRB"</span>,</span>
<span>  sequence <span class="op">=</span> <span class="st">"aa"</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">0.85</span>,</span>
<span>  group.by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cluster.method <span class="op">=</span> <span class="st">"components"</span>,</span>
<span>  cluster.prefix <span class="op">=</span> <span class="st">"cluster."</span>,</span>
<span>  use.V <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  use.J <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  exportAdjMatrix <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  exportGraph <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-input-data">input.data<a class="anchor" aria-label="anchor" href="#arg-input-data"></a></dt>
<dd><p>The product of <code><a href="combineTCR.html">combineTCR()</a></code>,
<code><a href="combineBCR.html">combineBCR()</a></code> or <code><a href="combineExpression.html">combineExpression()</a></code>.</p></dd>


<dt id="arg-chain">chain<a class="anchor" aria-label="anchor" href="#arg-chain"></a></dt>
<dd><p>The TCR/BCR chain to use. Use <code>both</code> to include both chains
(e.g., TRA/TRB). Accepted values: <code>TRA</code>, <code>TRB</code>, <code>TRG</code>, <code>TRD</code>, <code>IGH</code>, <code>IGL</code>
(for both light chains), <code>both</code>.</p></dd>


<dt id="arg-sequence">sequence<a class="anchor" aria-label="anchor" href="#arg-sequence"></a></dt>
<dd><p>Clustering based on either <code>aa</code> or <code>nt</code> sequences.</p></dd>


<dt id="arg-threshold">threshold<a class="anchor" aria-label="anchor" href="#arg-threshold"></a></dt>
<dd><p>The similarity threshold. If &lt; 1, treated as normalized
similarity (higher is stricter). If &gt;= 1, treated as raw edit distance
(lower is stricter).</p></dd>


<dt id="arg-group-by">group.by<a class="anchor" aria-label="anchor" href="#arg-group-by"></a></dt>
<dd><p>A column header in the metadata or lists to group the analysis
by (e.g., "sample", "treatment"). If <code>NULL</code>, clusters will be calculated across
all sequences.</p></dd>


<dt id="arg-cluster-method">cluster.method<a class="anchor" aria-label="anchor" href="#arg-cluster-method"></a></dt>
<dd><p>The clustering algorithm to use. Defaults to <code>"components"</code>,
which finds connected subgraphs.</p></dd>


<dt id="arg-cluster-prefix">cluster.prefix<a class="anchor" aria-label="anchor" href="#arg-cluster-prefix"></a></dt>
<dd><p>A character prefix to add to the cluster names (e.g.,
"cluster.").</p></dd>


<dt id="arg-use-v">use.V<a class="anchor" aria-label="anchor" href="#arg-use-v"></a></dt>
<dd><p>If <code>TRUE</code>, sequences must share the same V gene to be
clustered together.</p></dd>


<dt id="arg-use-j">use.J<a class="anchor" aria-label="anchor" href="#arg-use-j"></a></dt>
<dd><p>If <code>TRUE</code>, sequences must share the same J gene to be
clustered together.</p></dd>


<dt id="arg-exportadjmatrix">exportAdjMatrix<a class="anchor" aria-label="anchor" href="#arg-exportadjmatrix"></a></dt>
<dd><p>If <code>TRUE</code>, the function returns a sparse
adjacency matrix (<code>dgCMatrix</code>) of the network.</p></dd>


<dt id="arg-exportgraph">exportGraph<a class="anchor" aria-label="anchor" href="#arg-exportgraph"></a></dt>
<dd><p>If <code>TRUE</code> returns an igraph object of connected
sequences  or the amended <code>input.data</code> with a new cluster-based variable</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>Depending on the export parameters, one of the following:</p><ul><li><p>An amended <code>input.data</code> object with a new metadata column containing cluster IDs (default).</p></li>
<li><p>An <code>igraph</code> object if <code>exportGraph = TRUE</code>.</p></li>
<li><p>A sparse <code>dgCMatrix</code> object if <code>exportAdjMatrix = TRUE</code>.</p></li>
</ul></div>
    <div id="details">
    <h2>Details</h2>
    <p>The clustering process is as follows:</p><ol><li><p>The function retrieves the relevant chain data from the input object.</p></li>
<li><p>It calculates the edit distance between all sequences within each group
(or across the entire dataset if <code>group.by</code> is <code>NULL</code>).</p></li>
<li><p>An edge list is constructed, connecting sequences that meet the similarity
<code>threshold</code>.</p></li>
<li><p>The <code>threshold</code> parameter behaves differently based on its value:</p><ul><li><p><strong><code>threshold</code> &lt; 1 (e.g., 0.85):</strong> Interpreted as a <em>normalized</em> edit
distance or sequence similarity. A higher value means greater
similarity is required. This is the default behavior.</p></li>
<li><p><strong><code>threshold</code> &gt;= 1 (e.g., 2):</strong> Interpreted as a maximum <em>raw</em> edit
distance. A lower value means greater similarity is required.</p></li>
</ul></li>
<li><p>An <code>igraph</code> graph is built from the edge list.</p></li>
<li><p>A clustering algorithm is run on the graph. The default
<code>cluster.method = "components"</code> simply identifies the connected
components (i.e., each cluster is a group of sequences connected by
edges). Other methods from <code>igraph</code> can be used.</p></li>
<li><p>The resulting cluster information is formatted and returned in the
specified format.</p></li>
</ol></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Getting the combined contigs</span></span></span>
<span class="r-in"><span><span class="va">combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="combineTCR.html">combineTCR</a></span><span class="op">(</span><span class="va">contig_list</span>,</span></span>
<span class="r-in"><span>                       samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17B"</span>, <span class="st">"P17L"</span>, <span class="st">"P18B"</span>, <span class="st">"P18L"</span>,</span></span>
<span class="r-in"><span>                                   <span class="st">"P19B"</span>,<span class="st">"P19L"</span>, <span class="st">"P20B"</span>, <span class="st">"P20L"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Add cluster information to the list</span></span></span>
<span class="r-in"><span><span class="va">sub_combined</span> <span class="op">&lt;-</span> <span class="fu">clonalCluster</span><span class="op">(</span><span class="va">combined</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                              chain <span class="op">=</span> <span class="st">"TRA"</span>,</span></span>
<span class="r-in"><span>                              sequence <span class="op">=</span> <span class="st">"aa"</span>,</span></span>
<span class="r-in"><span>                              threshold <span class="op">=</span> <span class="fl">0.85</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Export the graph object instead</span></span></span>
<span class="r-in"><span><span class="va">graph_obj</span> <span class="op">&lt;-</span> <span class="fu">clonalCluster</span><span class="op">(</span><span class="va">combined</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                           chain <span class="op">=</span> <span class="st">"TRA"</span>,</span></span>
<span class="r-in"><span>                           exportGraph <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

