<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Combine B Cell Receptor Contig Data — combineBCR • scRepertoire</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Combine B Cell Receptor Contig Data — combineBCR"><meta property="og:description" content="This function consolidates a list of BCR sequencing results to the level
of the individual cell barcodes. Using the samples and ID parameters,
the function will add the strings as prefixes to prevent issues with
repeated barcodes. The resulting new barcodes will need to match the
Seurat or SCE object in order to use, combineExpression().
Unlike combineTCR(), combineBCR produces a column
CTstrict of an index of nucleotide sequence and the
corresponding V and J genes using clonalCluster()."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.6.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Repertoire Analysis</li>
    <li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Working with Single-Cell Objects</li>
    <li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Deep Learning Integrations</li>
    <li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Other Workflow Integrations</li>
    <li>
      <a href="../articles/immcantation.html">Combining Immcantation and scRepertoire</a>
    </li>
    <li>
      <a href="../articles/tcrpheno.html">TCRpheno applied to T cell fate</a>
    </li>
    <li>
      <a href="../articles/immunarch.html">Using scRepertoire with immunarch</a>
    </li>
    <li>
      <a href="../articles/dandelionR.html">VDJ Trajectory Analysis with dandelionR</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/BorchLab/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Combine B Cell Receptor Contig Data</h1>
    <small class="dont-index">Source: <a href="https://github.com/BorchLab/scRepertoire/blob/HEAD/R/combineContigs.R" class="external-link"><code>R/combineContigs.R</code></a></small>
    <div class="hidden name"><code>combineBCR.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function consolidates a list of BCR sequencing results to the level
of the individual cell barcodes. Using the samples and ID parameters,
the function will add the strings as prefixes to prevent issues with
repeated barcodes. The resulting new barcodes will need to match the
Seurat or SCE object in order to use, <code><a href="combineExpression.html">combineExpression()</a></code>.
Unlike <code><a href="combineTCR.html">combineTCR()</a></code>, combineBCR produces a column
<code>CTstrict</code> of an index of nucleotide sequence and the
corresponding V and J genes using <code><a href="clonalCluster.html">clonalCluster()</a></code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">combineBCR</span><span class="op">(</span></span>
<span>  <span class="va">input.data</span>,</span>
<span>  samples <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ID <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chain <span class="op">=</span> <span class="st">"both"</span>,</span>
<span>  sequence <span class="op">=</span> <span class="st">"nt"</span>,</span>
<span>  call.related.clones <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  group.by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">0.85</span>,</span>
<span>  cluster.method <span class="op">=</span> <span class="st">"components"</span>,</span>
<span>  use.V <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  use.J <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  removeNA <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  removeMulti <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  filterMulti <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  filterNonproductive <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-input-data">input.data<a class="anchor" aria-label="anchor" href="#arg-input-data"></a></dt>
<dd><p>List of filtered contig annotations or outputs from
<code><a href="loadContigs.html">loadContigs()</a></code>.</p></dd>


<dt id="arg-samples">samples<a class="anchor" aria-label="anchor" href="#arg-samples"></a></dt>
<dd><p>A character vector of sample labels. Must be the same length
as the input list.</p></dd>


<dt id="arg-id">ID<a class="anchor" aria-label="anchor" href="#arg-id"></a></dt>
<dd><p>An optional character vector for additional sample identifiers.</p></dd>


<dt id="arg-chain">chain<a class="anchor" aria-label="anchor" href="#arg-chain"></a></dt>
<dd><p>The chain to use for clustering when <code>call.related.clones = TRUE</code>.
Passed to <code><a href="clonalCluster.html">clonalCluster()</a></code>. Default is <code>"both"</code>.</p></dd>


<dt id="arg-sequence">sequence<a class="anchor" aria-label="anchor" href="#arg-sequence"></a></dt>
<dd><p>The sequence type (<code>"nt"</code> or <code>"aa"</code>) to use for clustering.
Passed to <code><a href="clonalCluster.html">clonalCluster()</a></code>. Default is <code>"nt"</code>.</p></dd>


<dt id="arg-call-related-clones">call.related.clones<a class="anchor" aria-label="anchor" href="#arg-call-related-clones"></a></dt>
<dd><p>Logical. If <code>TRUE</code>, uses <code><a href="clonalCluster.html">clonalCluster()</a></code> to
identify related clones based on sequence similarity. If <code>FALSE</code>, defines
clones by the exact V-gene and CDR3 amino acid sequence.</p></dd>


<dt id="arg-group-by">group.by<a class="anchor" aria-label="anchor" href="#arg-group-by"></a></dt>
<dd><p>The column header used for to group clones.
If (`NULL“), clusters will be calculated across samples.</p></dd>


<dt id="arg-threshold">threshold<a class="anchor" aria-label="anchor" href="#arg-threshold"></a></dt>
<dd><p>The similarity threshold passed to <code><a href="clonalCluster.html">clonalCluster()</a></code> if
<code>call.related.clones = TRUE</code>. See <code><a href="clonalCluster.html">?clonalCluster</a></code> for details.</p></dd>


<dt id="arg-cluster-method">cluster.method<a class="anchor" aria-label="anchor" href="#arg-cluster-method"></a></dt>
<dd><p>The clustering algorithm to use. Defaults to <code>"components"</code>,
which finds connected subgraphs.</p></dd>


<dt id="arg-use-v">use.V<a class="anchor" aria-label="anchor" href="#arg-use-v"></a></dt>
<dd><p>Logical. If <code>TRUE</code>, sequences must share the same V gene to be
clustered together.</p></dd>


<dt id="arg-use-j">use.J<a class="anchor" aria-label="anchor" href="#arg-use-j"></a></dt>
<dd><p>Logical. If <code>TRUE</code>, sequences must share the same J gene to be
clustered together.</p></dd>


<dt id="arg-removena">removeNA<a class="anchor" aria-label="anchor" href="#arg-removena"></a></dt>
<dd><p>This will remove any chain without values.</p></dd>


<dt id="arg-removemulti">removeMulti<a class="anchor" aria-label="anchor" href="#arg-removemulti"></a></dt>
<dd><p>Logical. If <code>TRUE</code>, removes cells that have more than
one distinct heavy or light chain after processing.</p></dd>


<dt id="arg-filtermulti">filterMulti<a class="anchor" aria-label="anchor" href="#arg-filtermulti"></a></dt>
<dd><p>Logical. If <code>TRUE</code>, filters multi-chain cells to retain
only the most abundant IGH and IGL/IGK chains.</p></dd>


<dt id="arg-filternonproductive">filterNonproductive<a class="anchor" aria-label="anchor" href="#arg-filternonproductive"></a></dt>
<dd><p>Logical. If <code>TRUE</code>, removes non-productive contigs
from the analysis.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A list of data frames, where each data frame represents a sample.
Each row corresponds to a unique cell barcode, with columns detailing the
BCR chains and the assigned clone ID.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Data derived from the 10x Genomics intratumoral NSCLC B cells</span></span></span>
<span class="r-in"><span><span class="va">BCR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"https://www.borch.dev/uploads/contigs/b_contigs.csv"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">combined</span> <span class="op">&lt;-</span> <span class="fu">combineBCR</span><span class="op">(</span><span class="va">BCR</span>,</span></span>
<span class="r-in"><span>                       samples <span class="op">=</span> <span class="st">"Patient1"</span>,</span></span>
<span class="r-in"><span>                       threshold <span class="op">=</span> <span class="fl">0.85</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

