<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ scRepertoire</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/ncborcherding/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="loading-and-processing-contigs">Loading and Processing Contigs <a href="#loading-and-processing-contigs" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions that load, combine, and process the single-cell contig information.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="addVariable.html">addVariable()</a></code> </p>
        </td>
        <td><p>Adding variables after combineTCR() or combineBCR()</p></td>
      </tr><tr><td>
          <p><code><a href="combineBCR.html">combineBCR()</a></code> </p>
        </td>
        <td><p>Combining the list of B cell receptor contigs into clones</p></td>
      </tr><tr><td>
          <p><code><a href="combineTCR.html">combineTCR()</a></code> </p>
        </td>
        <td><p>Combining the list of T cell receptor contigs into clones</p></td>
      </tr><tr><td>
          <p><code><a href="createHTOContigList.html">createHTOContigList()</a></code> </p>
        </td>
        <td><p>Generate a contig list from a multiplexed experiment</p></td>
      </tr><tr><td>
          <p><code><a href="exportClones.html">exportClones()</a></code> </p>
        </td>
        <td><p>Exporting clones</p></td>
      </tr><tr><td>
          <p><code><a href="loadContigs.html">loadContigs()</a></code> </p>
        </td>
        <td><p>Loading the contigs derived from single-cell sequencing</p></td>
      </tr><tr><td>
          <p><code><a href="subsetClones.html">subsetClones()</a></code> </p>
        </td>
        <td><p>Subset the product of combineTCR() or combineBCR()</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="visualizing-clones">Visualizing Clones <a href="#visualizing-clones" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions plotting clonal data.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="clonalAbundance.html">clonalAbundance()</a></code> </p>
        </td>
        <td><p>Demonstrate the relative abundance of clones by group or sample</p></td>
      </tr><tr><td>
          <p><code><a href="clonalCluster.html">clonalCluster()</a></code> </p>
        </td>
        <td><p>Clustering adaptive receptor sequences by edit distance</p></td>
      </tr><tr><td>
          <p><code><a href="clonalCompare.html">clonalCompare()</a></code> </p>
        </td>
        <td><p>Demonstrate the difference in clonal proportion between clones</p></td>
      </tr><tr><td>
          <p><code><a href="clonalDiversity.html">clonalDiversity()</a></code> </p>
        </td>
        <td><p>Calculate the clonal diversity for samples or groupings</p></td>
      </tr><tr><td>
          <p><code><a href="clonalHomeostasis.html">clonalHomeostasis()</a></code> </p>
        </td>
        <td><p>Examining the clonal homeostasis of the repertoire</p></td>
      </tr><tr><td>
          <p><code><a href="clonalLength.html">clonalLength()</a></code> </p>
        </td>
        <td><p>Demonstrate the distribution of clonal length</p></td>
      </tr><tr><td>
          <p><code><a href="clonalOverlap.html">clonalOverlap()</a></code> </p>
        </td>
        <td><p>Examining the clonal overlap between groups or samples</p></td>
      </tr><tr><td>
          <p><code><a href="clonalProportion.html">clonalProportion()</a></code> </p>
        </td>
        <td><p>Examining the clonal space occupied by specific clones</p></td>
      </tr><tr><td>
          <p><code><a href="clonalQuant.html">clonalQuant()</a></code> </p>
        </td>
        <td><p>Quantify the unique clones by group or sample</p></td>
      </tr><tr><td>
          <p><code><a href="clonalRarefaction.html">clonalRarefaction()</a></code> </p>
        </td>
        <td><p>Calculate rarefaction based on the abundance of clones</p></td>
      </tr><tr><td>
          <p><code><a href="clonalScatter.html">clonalScatter()</a></code> </p>
        </td>
        <td><p>Scatter plot comparing the clonal expansion of two samples</p></td>
      </tr><tr><td>
          <p><code><a href="clonalSizeDistribution.html">clonalSizeDistribution()</a></code> </p>
        </td>
        <td><p>Hierarchical clustering of clones using Gamma-GPD spliced threshold model</p></td>
      </tr><tr><td>
          <p><code><a href="vizGenes.html">vizGenes()</a></code> </p>
        </td>
        <td><p>Visualizing the distribution of gene usage</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="summarizing-repertoire">Summarizing Repertoire <a href="#summarizing-repertoire" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to summarize clonal sequences across the repertoire.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="percentAA.html">percentAA()</a></code> </p>
        </td>
        <td><p>Examining the relative amino acid composition by position</p></td>
      </tr><tr><td>
          <p><code><a href="positionalEntropy.html">positionalEntropy()</a></code> </p>
        </td>
        <td><p>Examining the diversity of amino acids by position</p></td>
      </tr><tr><td>
          <p><code><a href="positionalProperty.html">positionalProperty()</a></code> </p>
        </td>
        <td><p>Examining the mean property of amino acids by position</p></td>
      </tr><tr><td>
          <p><code><a href="percentGenes.html">percentGenes()</a></code> </p>
        </td>
        <td><p>Examining the VDJ gene usage across clones</p></td>
      </tr><tr><td>
          <p><code><a href="percentKmer.html">percentKmer()</a></code> </p>
        </td>
        <td><p>Examining the relative composition of kmer motifs in clones.</p></td>
      </tr><tr><td>
          <p><code><a href="percentVJ.html">percentVJ()</a></code> </p>
        </td>
        <td><p>Quantifying the V and J gene usage across clones</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="single-cell-object-processing-and-visualizations">Single-Cell Object Processing and Visualizations <a href="#single-cell-object-processing-and-visualizations" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to add or visualize clonal information along a single-cell object.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="alluvialClones.html">alluvialClones()</a></code> </p>
        </td>
        <td><p>Alluvial plotting for single-cell object meta data</p></td>
      </tr><tr><td>
          <p><code><a href="clonalBias.html">clonalBias()</a></code> </p>
        </td>
        <td><p>Examine skew of clones towards a cluster or compartment</p></td>
      </tr><tr><td>
          <p><code><a href="clonalNetwork.html">clonalNetwork()</a></code> </p>
        </td>
        <td><p>Visualize clonal network along reduced dimensions</p></td>
      </tr><tr><td>
          <p><code><a href="clonalOccupy.html">clonalOccupy()</a></code> </p>
        </td>
        <td><p>Visualize the number of single cells with cloneSizes by cluster</p></td>
      </tr><tr><td>
          <p><code><a href="clonalOverlay.html">clonalOverlay()</a></code> </p>
        </td>
        <td><p>Visualize distribution of clonal frequency overlaid on dimensional reduction plots</p></td>
      </tr><tr><td>
          <p><code><a href="combineExpression.html">combineExpression()</a></code> </p>
        </td>
        <td><p>Adding clone information to a single-cell object</p></td>
      </tr><tr><td>
          <p><code><a href="getCirclize.html">getCirclize()</a></code> </p>
        </td>
        <td><p>Generate data frame to be used with circlize R package to visualize
clones as a chord diagram.</p></td>
      </tr><tr><td>
          <p><code><a href="highlightClones.html">highlightClones()</a></code> </p>
        </td>
        <td><p>Highlighting specific clones in Seurat</p></td>
      </tr><tr><td>
          <p><code><a href="StartracDiversity.html">StartracDiversity()</a></code> </p>
        </td>
        <td><p>Startrac-based diversity indices for single-cell RNA-seq</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="data">Data <a href="#data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Reference data for package functions.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="contig_list.html">contig_list</a></code> </p>
        </td>
        <td><p>A list of 8 single-cell T cell receptor sequences runs.</p></td>
      </tr><tr><td>
          <p><code><a href="mini_contig_list.html">mini_contig_list</a></code> </p>
        </td>
        <td><p>Processed subset of `contig_list`</p></td>
      </tr><tr><td>
          <p><code><a href="scRep_example.html">scRep_example</a></code> </p>
        </td>
        <td><p>A Seurat object of 500 single T cells,</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

