<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Calculate Startrac-based Diversity Indices — StartracDiversity • scRepertoire</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate Startrac-based Diversity Indices — StartracDiversity"><meta property="og:description" content="This function utilizes the STARTRAC approach to calculate T cell
diversity metrics based on the work of Zhang et al. (2018, Nature)
PMID: 30479382. It can compute
three distinct indices: clonal expansion (expa), cross-tissue migration
(migr), and state transition (tran)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Repertoire Analysis</li>
    <li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Working with Single-Cell Objects</li>
    <li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Deep Learning Integrations</li>
    <li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Other Workflow Integrations</li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
    <li>
      <a href="../articles/immcantation.html">Combining Immcantation and scRepertoire</a>
    </li>
    <li>
      <a href="../articles/tcrpheno.html">TCRpheno applied to T cell fate</a>
    </li>
    <li>
      <a href="../articles/immunarch.html">Using scRepertoire with immunarch</a>
    </li>
    <li>
      <a href="../articles/dandelionR.html">VDJ Trajectory Analysis with dandelionR</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/BorchLab/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate Startrac-based Diversity Indices</h1>
    <small class="dont-index">Source: <a href="https://github.com/BorchLab/scRepertoire/blob/HEAD/R/startracDiversity.R" class="external-link"><code>R/startracDiversity.R</code></a></small>
    <div class="hidden name"><code>StartracDiversity.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function utilizes the STARTRAC approach to calculate T cell
diversity metrics based on the work of Zhang et al. (2018, Nature)
<a href="https://pubmed.ncbi.nlm.nih.gov/30479382/" class="external-link">PMID: 30479382</a>. It can compute
three distinct indices: clonal expansion (<code>expa</code>), cross-tissue migration
(<code>migr</code>), and state transition (<code>tran</code>).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">StartracDiversity</span><span class="op">(</span></span>
<span>  <span class="va">sc.data</span>,</span>
<span>  cloneCall <span class="op">=</span> <span class="st">"strict"</span>,</span>
<span>  chain <span class="op">=</span> <span class="st">"both"</span>,</span>
<span>  index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"expa"</span>, <span class="st">"migr"</span>, <span class="st">"tran"</span><span class="op">)</span>,</span>
<span>  type <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  group.by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pairwise <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  exportTable <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  palette <span class="op">=</span> <span class="st">"inferno"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-sc-data">sc.data<a class="anchor" aria-label="anchor" href="#arg-sc-data"></a></dt>
<dd><p>The single-cell object after <code><a href="combineExpression.html">combineExpression()</a></code>.
For SCE objects, the cluster variable must be in the meta data under
"cluster".</p></dd>


<dt id="arg-clonecall">cloneCall<a class="anchor" aria-label="anchor" href="#arg-clonecall"></a></dt>
<dd><p>Defines the clonal sequence grouping. Accepted values
are: <code>gene</code> (VDJC genes), <code>nt</code> (CDR3 nucleotide sequence), <code>aa</code> (CDR3 amino
acid sequence), or <code>strict</code> (VDJC + nt). A custom column header can also be used.</p></dd>


<dt id="arg-chain">chain<a class="anchor" aria-label="anchor" href="#arg-chain"></a></dt>
<dd><p>The TCR/BCR chain to use. Use <code>both</code> to include both chains
(e.g., TRA/TRB). Accepted values: <code>TRA</code>, <code>TRB</code>, <code>TRG</code>, <code>TRD</code>, <code>IGH</code>, <code>IGL</code>
(for both light chains), <code>both</code>.</p></dd>


<dt id="arg-index">index<a class="anchor" aria-label="anchor" href="#arg-index"></a></dt>
<dd><p>A character vector specifying which indices to calculate.
Options: "expa", "migr", "tran". Default is all three.</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>The metadata variable that specifies tissue type for migration
analysis.</p></dd>


<dt id="arg-group-by">group.by<a class="anchor" aria-label="anchor" href="#arg-group-by"></a></dt>
<dd><p>A column header in the metadata or lists to group the analysis
by (e.g., "sample", "treatment"). If <code>NULL</code>, data will be analyzed as
by list element or active identity in the case of single-cell objects.</p></dd>


<dt id="arg-pairwise">pairwise<a class="anchor" aria-label="anchor" href="#arg-pairwise"></a></dt>
<dd><p>The metadata column to be used for pairwise comparisons.
Set to the <code>type</code> variable for pairwise migration or "cluster" for
pairwise transition.</p></dd>


<dt id="arg-exporttable">exportTable<a class="anchor" aria-label="anchor" href="#arg-exporttable"></a></dt>
<dd><p>If <code>TRUE</code>, returns a data frame or matrix of the results
instead of a plot.</p></dd>


<dt id="arg-palette">palette<a class="anchor" aria-label="anchor" href="#arg-palette"></a></dt>
<dd><p>Colors to use in visualization - input any
hcl.pals.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments passed to the ggplot theme</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A ggplot object visualizing STARTRAC diversity metrics or data.frame if
<code>exportTable = TRUE</code>.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The function requires a <code>type</code> variable in the metadata, which specifies the
tissue origin or any other categorical variable for migration analysis.</p>
<p><strong>Indices:</strong></p><ul><li><p><strong>expa (Clonal Expansion):</strong> Measures the extent of clonal
proliferation within a T cell cluster. It is calculated as
<code>1 - normalized Shannon entropy</code>. A higher value indicates greater
expansion of a few clones.</p></li>
<li><p><strong>migr (Cross-Tissue Migration):</strong> Quantifies the movement of
clonal T cells across different tissues (as defined by the <code>type</code>
parameter). It is based on the entropy of a clonotype's distribution
across tissues.</p></li>
<li><p><strong>tran (State Transition):</strong> Measures the developmental
transition of clonal T cells between different functional clusters.
It is based on the entropy  of a clonotype's distribution across
clusters.</p></li>
</ul><p><strong>Pairwise Analysis:</strong>
The <code>pairwise</code> parameter enables the calculation of migration or transition
between specific pairs of tissues or clusters, respectively.</p><ul><li><p>For migration (<code>index = "migr"</code>), set <code>pairwise</code> to the <code>type</code> column
(e.g., <code>pairwise = "Type"</code>).</p></li>
<li><p>For transition (<code>index = "tran"</code>), set <code>pairwise</code> to <code>"cluster"</code>.</p></li>
</ul></div>
    <div id="author">
    <h2>Author</h2>
    <p>Liangtao Zheng</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Getting the combined contigs</span></span></span>
<span class="r-in"><span><span class="va">combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="combineTCR.html">combineTCR</a></span><span class="op">(</span><span class="va">contig_list</span>, </span></span>
<span class="r-in"><span>                        samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17B"</span>, <span class="st">"P17L"</span>, <span class="st">"P18B"</span>, <span class="st">"P18L"</span>, </span></span>
<span class="r-in"><span>                                    <span class="st">"P19B"</span>,<span class="st">"P19L"</span>, <span class="st">"P20B"</span>, <span class="st">"P20L"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Getting a sample of a Seurat object</span></span></span>
<span class="r-in"><span><span class="va">scRep_example</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/get.html" class="external-link">get</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"scRep_example"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">scRep_example</span>  <span class="op">&lt;-</span> <span class="fu"><a href="combineExpression.html">combineExpression</a></span><span class="op">(</span><span class="va">combined</span>, <span class="va">scRep_example</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">scRep_example</span><span class="op">$</span><span class="va">Patient</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substring</a></span><span class="op">(</span><span class="va">scRep_example</span><span class="op">$</span><span class="va">orig.ident</span>,<span class="fl">1</span>,<span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">scRep_example</span><span class="op">$</span><span class="va">Type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substring</a></span><span class="op">(</span><span class="va">scRep_example</span><span class="op">$</span><span class="va">orig.ident</span>,<span class="fl">4</span>,<span class="fl">4</span><span class="op">)</span> </span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Calculate a single index (expansion)</span></span></span>
<span class="r-in"><span><span class="fu">StartracDiversity</span><span class="op">(</span><span class="va">scRep_example</span>, </span></span>
<span class="r-in"><span>                  type <span class="op">=</span> <span class="st">"Type"</span>, </span></span>
<span class="r-in"><span>                  group.by <span class="op">=</span> <span class="st">"Patient"</span>,</span></span>
<span class="r-in"><span>                  index <span class="op">=</span> <span class="st">"expa"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="StartracDiversity-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span>                  </span></span>
<span class="r-in"><span><span class="co"># Calculate pairwise transition </span></span></span>
<span class="r-in"><span><span class="fu">StartracDiversity</span><span class="op">(</span><span class="va">scRep_example</span>, </span></span>
<span class="r-in"><span>                  type <span class="op">=</span> <span class="st">"Type"</span>, </span></span>
<span class="r-in"><span>                  group.by <span class="op">=</span> <span class="st">"Patient"</span>,</span></span>
<span class="r-in"><span>                  index <span class="op">=</span> <span class="st">"tran"</span>,</span></span>
<span class="r-in"><span>                  pairwise <span class="op">=</span> <span class="st">"cluster"</span><span class="op">)</span> </span></span>
<span class="r-plt img"><img src="StartracDiversity-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

