<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Combining Clones and Single-Cell Objects • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Combining Clones and Single-Cell Objects">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.6.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Repertoire Analysis</li>
    <li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Working with Single-Cell Objects</li>
    <li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Deep Learning Integrations</li>
    <li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other Workflow Integrations</li>
    <li>
      <a href="../articles/immcantation.html">Combining Immcantation and scRepertoire</a>
    </li>
    <li>
      <a href="../articles/tcrpheno.html">TCRpheno applied to T cell fate</a>
    </li>
    <li>
      <a href="../articles/immunarch.html">Using scRepertoire with immunarch</a>
    </li>
    <li>
      <a href="../articles/dandelionR.html">VDJ Trajectory Analysis with dandelionR</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BorchLab/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Combining Clones and Single-Cell Objects</h1>
            
            <h4 data-toc-skip class="date">Compiled: October 31,
2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BorchLab/scRepertoire/blob/HEAD/vignettes/articles/Attaching_SC.Rmd" class="external-link"><code>vignettes/articles/Attaching_SC.Rmd</code></a></small>
      <div class="hidden name"><code>Attaching_SC.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="note-on-dimensional-reduction">Note on Dimensional Reduction<a class="anchor" aria-label="anchor" href="#note-on-dimensional-reduction"></a>
</h2>
<p>In single-cell RNA sequencing workflows, dimensional reduction is
typically performed by first identifying highly variable features. These
features are then used directly for UMAP/tSNE projection or as inputs
for principal component analysis. The same approach is commonly applied
to clustering as well.</p>
<p>However, in immune-focused datasets, VDJ genes from TCR and BCR are
often among the most variable genes. This variability arises naturally
due to clonal expansion and diversity within lymphocytes. As a result,
UMAP projections and clustering outcomes may be influenced by clonal
information rather than broader transcriptional differences across cell
types.</p>
<p>To mitigate this issue, a common strategy is to exclude VDJ genes
from the set of highly variable features before proceeding with
clustering and dimensional reduction. We introduce a set of functions
that facilitate this process by removing VDJ-related genes from either a
Seurat Object or a vector of gene names (useful for SCE-based
workflows).</p>
<div class="section level3">
<h3 id="functions-to-exclude-vdj-genes">Functions to Exclude VDJ Genes<a class="anchor" aria-label="anchor" href="#functions-to-exclude-vdj-genes"></a>
</h3>
<ul>
<li>
<code><a href="../reference/quietVDJgenes.html">quietVDJgenes()</a></code> – Removes both TCR and BCR VDJ
genes.</li>
<li>
<code><a href="../reference/quietVDJgenes.html">quietTCRgenes()</a></code> – Removes only TCR VDJ genes.</li>
<li>
<code><a href="../reference/quietVDJgenes.html">quietBCRgenes()</a></code> – Removes only BCR VDJ genes, but
retains BCR VDJ pseudogenes in the variable features.</li>
</ul>
<p>Let’s first check the top 10 variable features in the scRep_example
Seurat object before any removal:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check the first 10 variable features before removal</span></span>
<span><span class="fu">VariableFeatures</span><span class="op">(</span><span class="va">scRep_example</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "TRBV7-2"  "HSPA1B"   "HSPA1A"   "TRBV4-1"  "CCL4"     "TRBV5-1" </span></span>
<span><span class="co">##  [7] "TRBV10-3" "TRBV3-1"  "TRBV6-6"  "CD79A"</span></span></code></pre>
<p>Now, we’ll remove TCR VDJ genes from the scRep_example object:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Remove TCR VDJ genes</span></span>
<span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quietVDJgenes.html">quietTCRgenes</a></span><span class="op">(</span><span class="va">scRep_example</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check the first 10 variable features after removal</span></span>
<span><span class="fu">VariableFeatures</span><span class="op">(</span><span class="va">scRep_example</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "HSPA1B"  "HSPA1A"  "CCL4"    "CD79A"   "HLA-DRA" "CCL20"   "TUBA1B" </span></span>
<span><span class="co">##  [8] "HSPA6"   "TNF"     "MS4A1"</span></span></code></pre>
<p>By applying these functions, you can ensure that clustering and
dimensional reduction are driven by broader transcriptomic differences
across cell types rather than being skewed by the inherent variability
due to clonal expansion. This provides a more accurate representation of
cellular heterogeneity independent of clonal lineage.</p>
</div>
</div>
<div class="section level2">
<h2 id="preprocessed-single-cell-object">Preprocessed Single-Cell Object<a class="anchor" aria-label="anchor" href="#preprocessed-single-cell-object"></a>
</h2>
<p>The data in the <code>scRepertoire</code> package is derived from a
<a href="https://pubmed.ncbi.nlm.nih.gov/33622974/" class="external-link">study</a> of acute
respiratory stress disorder in the context of bacterial and COVID-19
infections. The internal single cell data (<code><a href="../reference/scRep_example.html">scRep_example()</a></code>)
built in to scRepertoire is randomly sampled 500 cells from the fully
integrated Seurat object to minimize the package size. However, for the
purpose of the vignette we will use the full single-cell object with
30,000 cells. We will use both Seurat and Single-Cell Experiment (SCE)
with <code>scater</code> to perform further visualizations in
tandem.</p>
<div class="section level3">
<h3 id="demonstrating-preprocessed-object-loading">Demonstrating Preprocessed Object Loading<a class="anchor" aria-label="anchor" href="#demonstrating-preprocessed-object-loading"></a>
</h3>
<p>Here’s how to load the full single-cell object and convert it to a
Single-Cell Experiment object:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"scRep_example_full.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Making a Single-Cell Experiment object</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/as.SingleCellExperiment.html" class="external-link">as.SingleCellExperiment</a></span><span class="op">(</span><span class="va">scRep_example</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="combineexpression">combineExpression<a class="anchor" aria-label="anchor" href="#combineexpression"></a>
</h2>
<p>After processing the contig data into clones via
<code><a href="../reference/combineBCR.html">combineBCR()</a></code> or <code><a href="../reference/combineTCR.html">combineTCR()</a></code>, we can add the
clonal information to the single-cell object using
<code><a href="../reference/combineExpression.html">combineExpression()</a></code>.</p>
<p><strong>Importantly</strong>, the major requirement for the
attachment is matching contig cell barcodes and barcodes in the row
names of the metadata of the Seurat or Single-Cell Experiment object. If
these do not match, the attachment will fail. We suggest making changes
to the single-cell object barcodes for ease of use.</p>
<div class="section level3">
<h3 id="calculating-clonesize">Calculating <code>cloneSize</code><a class="anchor" aria-label="anchor" href="#calculating-clonesize"></a>
</h3>
<p>Part of <code><a href="../reference/combineExpression.html">combineExpression()</a></code> is calculating the clonal
frequency and proportion, placing each clone into groups called
<code>cloneSize</code>. The default <code>cloneSize</code> argument uses
the following bins:
<code>c(Rare = 1e-4, Small = 0.001, Medium = 0.01, Large = 0.1, Hyperexpanded = 1)</code>,
which can be modified to include more/less bins or different names.</p>
<p>Clonal frequency and proportion are dependent on the repertoires
being compared. You can modify the calculation using the
<code>group.by</code> parameter, such as grouping by a Patient variable.
If <code>group.by</code> is not set, <code><a href="../reference/combineExpression.html">combineExpression()</a></code>
will calculate clonal frequency, proportion, and <code>cloneSize</code>
as a function of individual sequencing runs. Additionally,
<code>cloneSize</code> can use the frequency of clones when
<code>proportion = FALSE</code>.</p>
<p>Key Parameter(s) for <code><a href="../reference/combineExpression.html">combineExpression()</a></code></p>
<ul>
<li>
<code>input.data</code>: The product of <code><a href="../reference/combineTCR.html">combineTCR()</a></code>,
<code><a href="../reference/combineBCR.html">combineBCR()</a></code>, or a list containing both.</li>
<li>
<code>sc.data</code>: The Seurat or Single-Cell Experiment (SCE)
object to attach the clonal data to.</li>
<li>
<code>proportion</code>: If <code>TRUE</code> (default), calculates
the proportion of the clone; if <code>FALSE</code>, calculates total
frequency.</li>
<li>
<code>cloneSize</code>: Bins for grouping based on proportion or
frequency. If proportion is <code>FALSE</code> and cloneSize bins are
not set high enough, the upper limit will automatically adjust.</li>
<li>
<code>filterNA</code>: Method to subset the Seurat/SCE object of
barcodes without clone information.</li>
<li>
<code>addLabel</code>: Adds a label to the frequency header, useful
for trying multiple group.by variables or recalculating frequencies
after subsetting.</li>
</ul>
<p>We can look at the default cloneSize groupings using the Single-Cell
Experiment object we just created, with group.by set to the sample
variable used in combineTCR():</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineExpression.html">combineExpression</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                         <span class="va">sce</span>, </span>
<span>                         cloneCall<span class="op">=</span><span class="st">"gene"</span>, </span>
<span>                         group.by <span class="op">=</span> <span class="st">"sample"</span>, </span>
<span>                         proportion <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Define color palette </span></span>
<span><span class="va">colorblind_vector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.colors</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">7</span>, palette <span class="op">=</span> <span class="st">"inferno"</span>, fixup <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotUMAP</span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"cloneSize"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_color_manual</span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">colorblind_vector</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span>,<span class="fl">5</span>,<span class="fl">7</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Attaching_SC_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>Alternatively, if we want <code>cloneSize</code> to be based on the
frequency of the clone, we can set proportion = FALSE and will need to
change the <code>cloneSize</code> bins to integers. If we haven’t
inspected our clone data, setting the upper limit of the clonal
frequency might be difficult; <code><a href="../reference/combineExpression.html">combineExpression()</a></code> will
automatically adjust the upper limit to fit the distribution of the
frequencies.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineExpression.html">combineExpression</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                                   <span class="va">scRep_example</span>, </span>
<span>                                   cloneCall<span class="op">=</span><span class="st">"gene"</span>, </span>
<span>                                   group.by <span class="op">=</span> <span class="st">"sample"</span>, </span>
<span>                                   proportion <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                                   cloneSize<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Single<span class="op">=</span><span class="fl">1</span>, Small<span class="op">=</span><span class="fl">5</span>, Medium<span class="op">=</span><span class="fl">20</span>, Large<span class="op">=</span><span class="fl">100</span>, Hyperexpanded<span class="op">=</span><span class="fl">500</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">scRep_example</span>, group.by <span class="op">=</span> <span class="st">"cloneSize"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_color_manual</span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">colorblind_vector</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span>,<span class="fl">4</span>,<span class="fl">5</span>,<span class="fl">7</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Attaching_SC_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="combining-both-tcr-and-bcr">Combining both TCR and BCR<a class="anchor" aria-label="anchor" href="#combining-both-tcr-and-bcr"></a>
</h3>
<p>If you have both TCR and BCR enrichment data, or wish to include
information for both gamma-delta and alpha-beta T cells, you can create
a single list containing the outputs of <code><a href="../reference/combineTCR.html">combineTCR()</a></code> and
<code><a href="../reference/combineBCR.html">combineBCR()</a></code> and then use
<code><a href="../reference/combineExpression.html">combineExpression()</a></code>.</p>
<p><strong>Major Note</strong>: If there are duplicate barcodes (e.g.,
if a cell has both Ig and TCR information), the immune receptor
information will not be added for those cells. It might be worth
checking cluster identities and removing incongruent barcodes in the
products of <code><a href="../reference/combineTCR.html">combineTCR()</a></code> and <code><a href="../reference/combineBCR.html">combineBCR()</a></code>. As
an anecdote, the <a href="https://support.10xgenomics.com/single-cell-vdj/datasets/6.0.1/SC5v2_Melanoma_5Kcells_Connect_single_channel_SC5v2_Melanoma_5Kcells_Connect_single_channel" class="external-link">testing
data</a> used to improve this function showed 5-6% barcode overlap.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#This is an example of the process, which will not be evaluated during knit</span></span>
<span><span class="va">TCR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineTCR.html">combineTCR</a></span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span>
<span><span class="va">BCR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineBCR.html">combineBCR</a></span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span>
<span><span class="va">list.receptors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">TCR</span>, <span class="va">BCR</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineExpression.html">combineExpression</a></span><span class="op">(</span><span class="va">list.receptors</span>, </span>
<span>                            <span class="va">seurat</span>, </span>
<span>                            cloneCall<span class="op">=</span><span class="st">"gene"</span>, </span>
<span>                            proportion <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><code><a href="../reference/combineExpression.html">combineExpression()</a></code> is a core function in
<code>scRepertoire</code> that bridges the immune repertoire data with
single-cell gene expression data. It enriches your Seurat or SCE object
with crucial clonal information, including calculated frequencies and
proportions, allowing for integrated analysis of cellular identity and
clonal dynamics. Its flexibility in defining <code>cloneSize</code> and
handling various grouping scenarios makes it adaptable to diverse
experimental designs, even accommodating the integration of both TCR and
BCR data.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
