<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Combining Clones and Single-Cell Objects â€¢ scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Combining Clones and Single-Cell Objects">
<meta property="og:description" content="scRepertoire">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Combining Clones and Single-Cell Objects</h1>
            
            <h4 data-toc-skip class="date">Compiled: January 10,
2024</h4>
      
      
      <div class="hidden name"><code>Attaching_SC.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<p>The data in the scRepertoire package is derived from a <a href="https://pubmed.ncbi.nlm.nih.gov/33622974/" class="external-link">study</a> of acute
respiratory stress disorder in the context of bacterial and COVID-19
infections. The internal single cell data (<code><a href="../reference/scRep_example.html">scRep_example()</a></code>)
built in to scRepertoire is randomly sampled 500 cells from the fully
integrated Seurat object to minimize the package size. However, for the
purpose of the vignette we will use the full single-cell object with
30,000 cells. We will use both Seurat and Single-Cell Experiment (SCE)
with scater to perform further visualizations in tandem.</p>
<div class="section level2">
<h2 id="preprocessed-single-cell-object">Preprocessed Single-Cell Object<a class="anchor" aria-label="anchor" href="#preprocessed-single-cell-object"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"scRep_example_full.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Making a Single-Cell Experiment object</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/as.SingleCellExperiment.html" class="external-link">as.SingleCellExperiment</a></span><span class="op">(</span><span class="va">scRep_example</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="combineexpression">combineExpression<a class="anchor" aria-label="anchor" href="#combineexpression"></a>
</h2>
<p>After processing the contig data into clones via
<code><a href="../reference/combineBCR.html">combineBCR()</a></code> or <code><a href="../reference/combineTCR.html">combineTCR()</a></code>, we can add the
clonal information to the single-cell object using
<code><a href="../reference/combineExpression.html">combineExpression()</a></code>.</p>
<p><strong>Importantly</strong>, the major requirement for the
attachment is matching contig cell barcodes and barcodes in the row
names of the meta data of the Seurat or Single-Cell Experiment object.
If these do not match, the attachment will fail. Based on ease, we
suggest making changes to the single-cell object barcodes.</p>
<div class="section level3">
<h3 id="calculating-clonesize">Calculating cloneSize<a class="anchor" aria-label="anchor" href="#calculating-clonesize"></a>
</h3>
<p>Part of <code><a href="../reference/combineExpression.html">combineExpression()</a></code> is calculating the clonal
frequency and proportion, placing each clone into groups called
<strong>cloneSize</strong>. The default <strong>cloneSize</strong>
argument uses the following bins: c(Rare = 1e-4, Small = 0.001, Medium =
0.01, Large = 0.1, Hyperexpanded = 1), which can be modified to include
more/less bins or different names.</p>
<p>Clonal frequency and proportion is dependent on the repertoires being
compared, which we can modify the calculation using the
<strong>group.by</strong> parameter, such as grouping by the
<em>Patient</em> variable from above. If <strong>group.by</strong> is
not set, <code><a href="../reference/combineExpression.html">combineExpression()</a></code> will calculate clonal
frequency, proportion, and <strong>cloneSize</strong> as a function of
individual sequencing runs. In addition, <strong>cloneSize</strong> can
use the frequency of clones when <strong>proportion</strong> =
FALSE.</p>
<p>We can look at the default cloneSize groupings using the Single-Cell
Experiment object we just created above with using
<strong>group.by</strong> set to the <em>sample</em> variable used in
<code><a href="../reference/combineTCR.html">combineTCR()</a></code> :</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineExpression.html">combineExpression</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                         <span class="va">sce</span>, </span>
<span>                         cloneCall<span class="op">=</span><span class="st">"gene"</span>, </span>
<span>                         group.by <span class="op">=</span> <span class="st">"sample"</span>, </span>
<span>                         proportion <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Define color palette </span></span>
<span><span class="va">colorblind_vector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.colors</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">7</span>, palette <span class="op">=</span> <span class="st">"inferno"</span>, fixup <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotUMAP</span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"cloneSize"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_color_manual</span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">colorblind_vector</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span>,<span class="fl">5</span>,<span class="fl">7</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Attaching_SC_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<p>Alternatively, if we want <strong>cloneSize</strong> to be based on
the frequency of the clone, we can set <strong>proportion</strong> =
FALSE and we will need to change the <strong>cloneSize</strong> bins to
integers. If we have not inspected our clone data, setting the upper
limit of the clonal frequency might be difficult -
<code><a href="../reference/combineExpression.html">combineExpression()</a></code> will automatically adjust the upper
limit to fit the distribution of the frequencies. To demonstrate this,
check out the Seurat object below:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineExpression.html">combineExpression</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                                   <span class="va">scRep_example</span>, </span>
<span>                                   cloneCall<span class="op">=</span><span class="st">"gene"</span>, </span>
<span>                                   group.by <span class="op">=</span> <span class="st">"sample"</span>, </span>
<span>                                   proportion <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                                   cloneSize<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Single<span class="op">=</span><span class="fl">1</span>, Small<span class="op">=</span><span class="fl">5</span>, Medium<span class="op">=</span><span class="fl">20</span>, Large<span class="op">=</span><span class="fl">100</span>, Hyperexpanded<span class="op">=</span><span class="fl">500</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">scRep_example</span>, group.by <span class="op">=</span> <span class="st">"cloneSize"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_color_manual</span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">colorblind_vector</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span>,<span class="fl">4</span>,<span class="fl">5</span>,<span class="fl">7</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Attaching_SC_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="combining-both-tcr-and-bcr">Combining both TCR and BCR<a class="anchor" aria-label="anchor" href="#combining-both-tcr-and-bcr"></a>
</h3>
<p>If we have TCR/BCR enrichment or want to add info for gamma-delta and
alpha-beta T cells, we can make a single list and use
<code><a href="../reference/combineExpression.html">combineExpression()</a></code>.</p>
<p><strong>Major note</strong> if there are duplicate barcodes (if a
cell has both Ig and TCR), the immune receptor information will not be
added. It might be worth checking cluster identities and removing
incongruent barcodes in the products of <code><a href="../reference/combineTCR.html">combineTCR()</a></code> and
<code><a href="../reference/combineBCR.html">combineBCR()</a></code>.</p>
<p>As an anecdote, the <a href="https://support.10xgenomics.com/single-cell-vdj/datasets/6.0.1/SC5v2_Melanoma_5Kcells_Connect_single_channel_SC5v2_Melanoma_5Kcells_Connect_single_channel" class="external-link">testing
data</a> we used to improve this function had 5-6% of barcode
overlap.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#This is an example of the process, which will not be evaluated during knit</span></span>
<span><span class="va">TCR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineTCR.html">combineTCR</a></span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span>
<span><span class="va">BCR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineBCR.html">combineBCR</a></span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span>
<span><span class="va">list.receptors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">TCR</span>, <span class="va">BCR</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineExpression.html">combineExpression</a></span><span class="op">(</span><span class="va">list.receptors</span>, </span>
<span>                            <span class="va">seurat</span>, </span>
<span>                            cloneCall<span class="op">=</span><span class="st">"gene"</span>, </span>
<span>                            proportion <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
