<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Basic Clonal Visualizations • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Basic Clonal Visualizations">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.3.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BorchLab/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Basic Clonal Visualizations</h1>
            
            <h4 data-toc-skip class="date">Compiled: March 19, 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BorchLab/scRepertoire/blob/HEAD/vignettes/articles/Clonal_Visualizations.Rmd" class="external-link"><code>vignettes/articles/Clonal_Visualizations.Rmd</code></a></small>
      <div class="hidden name"><code>Clonal_Visualizations.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="clonalquant">clonalQuant<a class="anchor" aria-label="anchor" href="#clonalquant"></a>
</h2>
<p>The first function to explore the clones is
<code><a href="../reference/clonalQuant.html">clonalQuant()</a></code> to return the total or relative numbers of
unique clones.</p>
<p><strong>scale</strong></p>
<ul>
<li>TRUE - relative percent of unique clones scaled by the total size of
the clonal repertoire.</li>
<li>FALSE - Report the total number of unique clones
<strong>(default)</strong>.</li>
</ul>
<p><strong>chain</strong></p>
<ul>
<li>“both” for combined chain visualization<br>
</li>
<li>“TRA”, “TRB”, “TRD”, “TRG”, “IGH” or “IGL” to select single
chain</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalQuant.html">clonalQuant</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>            cloneCall<span class="op">=</span><span class="st">"strict"</span>, </span>
<span>            chain <span class="op">=</span> <span class="st">"both"</span>, </span>
<span>            scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Visualizations_files/figure-html/unnamed-chunk-1-1.png" width="700"></p>
<p>Another option here is to be able to define the visualization by data
classes. Here, we used the <code><a href="../reference/combineTCR.html">combineTCR()</a></code> list to define the
<strong>Type</strong> variable as part of the naming structure. We can
use the <strong>group.by</strong> to specifically use a column in the
data set to organize the visualization.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalQuant.html">clonalQuant</a></span><span class="op">(</span><span class="va">combined.TCR</span>, cloneCall <span class="op">=</span> <span class="st">"gene"</span>, group.by <span class="op">=</span> <span class="st">"Type"</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Visualizations_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="clonalabundance">clonalAbundance<a class="anchor" aria-label="anchor" href="#clonalabundance"></a>
</h2>
<p>We can also examine the relative distribution of clones by abundance.
Here <code><a href="../reference/clonalAbundance.html">clonalAbundance()</a></code> will produce a line graph with a
total number of clones by the number of instances within the sample or
run. Like above, we can also group.by this by vectors within the contig
object using the <strong>group.by</strong> variable in the function.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalAbundance.html">clonalAbundance</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                cloneCall <span class="op">=</span> <span class="st">"gene"</span>, </span>
<span>                scale <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Visualizations_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p><code><a href="../reference/clonalAbundance.html">clonalAbundance()</a></code> output can also be converted into a
density plot, which may allow for better comparisons between different
repertoire sizes, by setting <strong>scale</strong> = TRUE.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalAbundance.html">clonalAbundance</a></span><span class="op">(</span><span class="va">combined.TCR</span>, cloneCall <span class="op">=</span> <span class="st">"gene"</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Visualizations_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="clonallength">clonalLength<a class="anchor" aria-label="anchor" href="#clonallength"></a>
</h2>
<p>We can look at the length distribution of the CDR3 sequences by
calling the <code>lengtheContig()</code> function. Importantly, unlike
the other basic visualizations, the <strong>cloneCall</strong> can only
be <strong>“nt”</strong> or <strong>“aa”</strong>. Due to the method of
calling clones as outlined above, the length should reveal a multimodal
curve, this is a product of using the <strong>NA</strong> for the
unreturned chain sequence and multiple chains within a single
barcode.</p>
<p><strong>chain</strong></p>
<ul>
<li>“both” for combined chain visualization<br>
</li>
<li>“TRA”, “TRB”, “TRD”, “TRG”, “IGH” or “IGL” to select single
chain</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalLength.html">clonalLength</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>             cloneCall<span class="op">=</span><span class="st">"aa"</span>, </span>
<span>             chain <span class="op">=</span> <span class="st">"both"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="Clonal_Visualizations_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalLength.html">clonalLength</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>             cloneCall<span class="op">=</span><span class="st">"aa"</span>, </span>
<span>             chain <span class="op">=</span> <span class="st">"TRA"</span>, </span>
<span>             scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span></code></pre></div>
<p><img src="Clonal_Visualizations_files/figure-html/unnamed-chunk-5-2.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="clonalcompare">clonalCompare<a class="anchor" aria-label="anchor" href="#clonalcompare"></a>
</h2>
<p>We can also look at clones between samples and changes in dynamics by
using the <code><a href="../reference/clonalCompare.html">clonalCompare()</a></code> function.</p>
<p><strong>samples</strong></p>
<ul>
<li>Can be used to isolate specific samples based on the name of the
list element</li>
</ul>
<p><strong>graph</strong></p>
<ul>
<li>“alluvial” - graph imaged below<br>
</li>
<li>“area” - graph by area of the respective clone</li>
</ul>
<p><strong>top.clones</strong></p>
<ul>
<li>The top number of clones to graph, this will be calculated based on
the frequency of the individual sample. This can also be left
blank.</li>
</ul>
<p><strong>clones</strong></p>
<ul>
<li>Can be used to isolate specific clone sequences, ensure the call
matches the sequences you would like to visualize.</li>
</ul>
<p><strong>highlight.clones</strong></p>
<ul>
<li>Specifically color certain clones, other clones will be greyed
out.</li>
</ul>
<p><strong>relabel.clones</strong></p>
<ul>
<li>Simplify the isolated clones to numerical designations to allow for
a tidier visualization</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalCompare.html">clonalCompare</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                  top.clones <span class="op">=</span> <span class="fl">10</span>, </span>
<span>                  samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17B"</span>, <span class="st">"P17L"</span><span class="op">)</span>, </span>
<span>                  cloneCall<span class="op">=</span><span class="st">"aa"</span>, </span>
<span>                  graph <span class="op">=</span> <span class="st">"alluvial"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Visualizations_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>We can also choose to highlight specific clones, such as in the case
of <em>“CVVSDNTGGFKTIF_CASSVRRERANTGELFF”</em> and
<em>“NA_CASSVRRERANTGELFF”</em> using the
<strong>highlight.clones</strong> parameter. In addition, we can
simplify the plot to label the clones as clones 1:19.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalCompare.html">clonalCompare</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>              top.clones <span class="op">=</span> <span class="fl">10</span>,</span>
<span>              highlight.clones <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CVVSDNTGGFKTIF_CASSVRRERANTGELFF"</span>, <span class="st">"NA_CASSVRRERANTGELFF"</span><span class="op">)</span>,</span>
<span>              relabel.clones <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>              samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17B"</span>, <span class="st">"P17L"</span><span class="op">)</span>, </span>
<span>              cloneCall<span class="op">=</span><span class="st">"aa"</span>, </span>
<span>              graph <span class="op">=</span> <span class="st">"alluvial"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Visualizations_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>Alternatively, if we only want to show specific clones, we can use
the <strong>clones</strong> parameter.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalCompare.html">clonalCompare</a></span><span class="op">(</span><span class="va">combined.TCR</span>, clones <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CVVSDNTGGFKTIF_CASSVRRERANTGELFF"</span>, <span class="st">"NA_CASSVRRERANTGELFF"</span><span class="op">)</span>,</span>
<span>    relabel.clones <span class="op">=</span> <span class="cn">TRUE</span>, samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17B"</span>, <span class="st">"P17L"</span><span class="op">)</span>, cloneCall <span class="op">=</span> <span class="st">"aa"</span>, graph <span class="op">=</span> <span class="st">"alluvial"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Visualizations_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="clonalscatter">clonalScatter<a class="anchor" aria-label="anchor" href="#clonalscatter"></a>
</h2>
<p><code><a href="../reference/clonalScatter.html">clonalScatter()</a></code> will organize two repertoires, quantify
the relative clone sizes, and produce a scatter plot comparing the two
samples.</p>
<p><strong>x.axis</strong> and <strong>y.axis</strong></p>
<ul>
<li>Names of the list element to place on the x-axis and y-axis - so for
example <em>“P18B”</em> and <em>“P18L”</em>
</li>
</ul>
<p><strong>dot.size</strong></p>
<ul>
<li>“total” to display the total number of clones between the x- and
y-axis<br>
</li>
<li>Name of the list element to use for size calculation</li>
</ul>
<p><strong>graph</strong></p>
<ul>
<li>“proportion” for the relative proportion the clone represents across
all clones<br>
</li>
<li>“count” for the total count of clones by sample</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalScatter.html">clonalScatter</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>              cloneCall <span class="op">=</span><span class="st">"gene"</span>, </span>
<span>              x.axis <span class="op">=</span> <span class="st">"P18B"</span>, </span>
<span>              y.axis <span class="op">=</span> <span class="st">"P18L"</span>,</span>
<span>              dot.size <span class="op">=</span> <span class="st">"total"</span>,</span>
<span>              graph <span class="op">=</span> <span class="st">"proportion"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Visualizations_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
