<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Clustering by Edit Distance • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Clustering by Edit Distance">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.6.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Repertoire Analysis</li>
    <li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Working with Single-Cell Objects</li>
    <li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Deep Learning Integrations</li>
    <li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other Workflow Integrations</li>
    <li>
      <a href="../articles/immcantation.html">Combining Immcantation and scRepertoire</a>
    </li>
    <li>
      <a href="../articles/tcrpheno.html">TCRpheno applied to T cell fate</a>
    </li>
    <li>
      <a href="../articles/immunarch.html">Using scRepertoire with immunarch</a>
    </li>
    <li>
      <a href="../articles/dandelionR.html">VDJ Trajectory Analysis with dandelionR</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BorchLab/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Clustering by Edit Distance</h1>
            
            <h4 data-toc-skip class="date">Compiled: October 31,
2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BorchLab/scRepertoire/blob/HEAD/vignettes/articles/Clonal_Cluster.Rmd" class="external-link"><code>vignettes/articles/Clonal_Cluster.Rmd</code></a></small>
      <div class="hidden name"><code>Clonal_Cluster.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="clonalcluster-cluster-by-sequence-similarity">clonalCluster: Cluster by Sequence Similarity<a class="anchor" aria-label="anchor" href="#clonalcluster-cluster-by-sequence-similarity"></a>
</h2>
<p>The <code><a href="../reference/clonalCluster.html">clonalCluster()</a></code> function provides a powerful method
to group clonotypes based on sequence similarity. It calculates the edit
distance between CDR3 sequences and uses this information to build a
network, identifying closely related clusters of T or B cell receptors.
This functionality allows for a more nuanced definition of a “clone”
that extends beyond identical sequence matches.</p>
<div class="section level3">
<h3 id="core-concepts">Core Concepts<a class="anchor" aria-label="anchor" href="#core-concepts"></a>
</h3>
<p>The clustering process follows these key steps:</p>
<ul>
<li>
<strong>Sequence Selection</strong>: The function selects either the
nucleotide (<code>sequence = "nt"</code>) or amino acid
(<code>sequence = "aa"</code>) CDR3 sequences for a specified
chain.</li>
<li>
<strong>Distance Calculation</strong>: It calculates the edit
distance between all pairs of sequences. By default, it also requires
sequences to share the same V gene (<code>use.v = TRUE</code>).</li>
<li>
<strong>Network Construction</strong>: An edge is created between
any two sequences that meet the similarity threshold, forming a network
graph.</li>
<li>
<strong>Clustering</strong>: A graph-based clustering algorithm is
run to identify connected components or communities within the network.
By default, it identifies all directly or indirectly connected sequences
as a single cluster (<code>cluster.method = "components"</code>).</li>
<li>
<strong>Output</strong>: The function can either add the resulting
cluster IDs to the input object, return an <code>igraph</code> object
for network analysis, or export a sparse adjacency matrix.</li>
</ul>
</div>
<div class="section level3">
<h3 id="understanding-the-threshold-parameter">Understanding the <code>threshold</code> Parameter<a class="anchor" aria-label="anchor" href="#understanding-the-threshold-parameter"></a>
</h3>
<p>The behavior of the threshold parameter is critical for controlling
cluster granularity:</p>
<ul>
<li>
<strong>Normalized Similarity (threshold &lt; 1)</strong>: When the
threshold is a value between 0 and 1 (e.g., <code>0.85</code>), it
represents the normalized edit distance (Levenshtein distance / mean
sequence length). A higher value corresponds to a stricter similarity
requirement. This is useful for comparing sequences of varying
lengths.</li>
<li>
<strong>Raw Edit Distance (threshold &gt;= 1)</strong>: When the
threshold is a whole number (e.g., 2), it represents the maximum raw
edit distance allowed. A lower value corresponds to a stricter
similarity requirement. This is useful when you want to allow a specific
number of mutations.</li>
</ul>
<p>Key Parameter(s) for <code><a href="../reference/clonalCluster.html">clonalCluster()</a></code></p>
<ul>
<li>
<code>sequence</code>: Specifies whether to cluster based on
<code>aa</code> (amino acid) or <code>nt</code> (nucleotide)
sequences.</li>
<li>
<code>threshold</code>: The similarity threshold for clustering.
Values <code>&lt; 1</code> are normalized similarity, while values
<code>&gt;= 1</code> are raw edit distance.</li>
<li>
<code>group.by</code>: A column header in the metadata or lists to
group the analysis by (e.g., “sample”, “treatment”). If
<code>NULL</code>, clusters are calculated across all sequences.</li>
<li>
<code>use.V</code>: If <code>TRUE</code>, sequences must share the
same V gene to be clustered together.</li>
<li>
<code>use.J</code>: If <code>TRUE</code>, sequences must share the
same J gene to be clustered together.</li>
<li>
<code>cluster.method</code>: The clustering algorithm to use.
Defaults to <code>components</code>, which finds connected
subgraphs.</li>
<li>
<code>cluster.prefix</code>: A character prefix to add to the
cluster names (e.g., “cluster.”).</li>
<li>
<code>exportGraph</code>: If <code>TRUE</code>, returns an igraph
object of the sequence network.</li>
<li>
<code>exportAdjMatrix</code>: If <code>TRUE</code>, returns a sparse
adjacency matrix (dgCMatrix) of the network.</li>
</ul>
</div>
<div class="section level3">
<h3 id="demonstrating-basic-use">Demonstrating Basic Use<a class="anchor" aria-label="anchor" href="#demonstrating-basic-use"></a>
</h3>
<p>To run clustering on the first two samples for the TRA chain, using
amino acid sequences with a normalized similarity threshold of 0.85:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run clustering on the first two samples for the TRA chain</span></span>
<span><span class="va">sub_combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clonalCluster.html">clonalCluster</a></span><span class="op">(</span><span class="va">combined.TCR</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                              chain <span class="op">=</span> <span class="st">"TRA"</span>, </span>
<span>                              sequence <span class="op">=</span> <span class="st">"aa"</span>, </span>
<span>                              threshold <span class="op">=</span> <span class="fl">0.85</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View the new cluster column</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sub_combined</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"barcode"</span>, <span class="st">"TCR1"</span>, <span class="st">"TRA.Cluster"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                   barcode                     TCR1 TRA.Cluster</span></span>
<span><span class="co">## 1 P17B_AAACCTGAGTACGACG-1       TRAV25.TRAJ20.TRAC        &lt;NA&gt;</span></span>
<span><span class="co">## 2 P17B_AAACCTGCAACACGCC-1 TRAV38-2/DV8.TRAJ52.TRAC        &lt;NA&gt;</span></span>
<span><span class="co">## 3 P17B_AAACCTGCAGGCGATA-1      TRAV12-1.TRAJ9.TRAC  cluster.24</span></span>
<span><span class="co">## 4 P17B_AAACCTGCATGAGCGA-1      TRAV12-1.TRAJ9.TRAC  cluster.24</span></span>
<span><span class="co">## 5 P17B_AAACGGGAGAGCCCAA-1        TRAV20.TRAJ8.TRAC cluster.148</span></span>
<span><span class="co">## 6 P17B_AAACGGGAGCGTTTAC-1      TRAV12-1.TRAJ9.TRAC  cluster.24</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="demonstrating-clustering-with-a-single-cell-object">Demonstrating Clustering with a Single-Cell Object<a class="anchor" aria-label="anchor" href="#demonstrating-clustering-with-a-single-cell-object"></a>
</h3>
<p>You can calculate clusters based on specific metadata variables
within a single-cell object by using the <code>group.by</code>
parameter. This is useful for analyzing clusters on a per-sample or
per-patient basis without subsetting the data first.</p>
<p>First, add patient and type information to the
<code>scRep_example</code> Seurat object:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Adding patient information</span></span>
<span><span class="va">scRep_example</span><span class="op">$</span><span class="va">Patient</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="va">scRep_example</span><span class="op">$</span><span class="va">orig.ident</span>, <span class="fl">1</span>,<span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Adding type information</span></span>
<span><span class="va">scRep_example</span><span class="op">$</span><span class="va">Type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="va">scRep_example</span><span class="op">$</span><span class="va">orig.ident</span>, <span class="fl">4</span>,<span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p>Now, run clustering on the <code>scRep_example</code> Seurat object,
grouping calculations by “Patient”:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run clustering, but group calculations by "Patient"</span></span>
<span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clonalCluster.html">clonalCluster</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                               chain <span class="op">=</span> <span class="st">"TRA"</span>, </span>
<span>                               sequence <span class="op">=</span> <span class="st">"aa"</span>, </span>
<span>                               threshold <span class="op">=</span> <span class="fl">0.85</span>, </span>
<span>                               group.by <span class="op">=</span> <span class="st">"Patient"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Define color palette </span></span>
<span><span class="va">num_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="va">scRep_example</span><span class="op">$</span><span class="va">TRA.Cluster</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cluster_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.colors</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">num_clusters</span>, palette <span class="op">=</span> <span class="st">"inferno"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">scRep_example</span>, group.by <span class="op">=</span> <span class="st">"TRA.Cluster"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">cluster_colors</span>, na.value <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">NoLegend</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Cluster_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="returning-an-igraph-object">Returning an igraph Object:<a class="anchor" aria-label="anchor" href="#returning-an-igraph-object"></a>
</h3>
<p>Instead of modifying the input object, <code><a href="../reference/clonalCluster.html">clonalCluster()</a></code>
can export the underlying network structure for advanced analysis. Set
<code>exportGraph = TRUE</code> to get an igraph object consisting of
the networks of barcodes by the indicated clustering scheme.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="co">#Clustering Patient 19 samples</span></span>
<span><span class="va">igraph.object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clonalCluster.html">clonalCluster</a></span><span class="op">(</span><span class="va">combined.TCR</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>,<span class="fl">6</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                               chain <span class="op">=</span> <span class="st">"TRB"</span>,</span>
<span>                               sequence <span class="op">=</span> <span class="st">"aa"</span>,</span>
<span>                               group.by <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>                               threshold <span class="op">=</span> <span class="fl">0.85</span>, </span>
<span>                               exportGraph <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Setting color scheme</span></span>
<span><span class="va">col_legend</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">igraph.object</span><span class="op">)</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></span>
<span><span class="va">col_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.colors</a></span><span class="op">(</span><span class="fl">2</span>,<span class="st">"inferno"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">col_legend</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">color.legend</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">igraph.object</span><span class="op">)</span><span class="op">$</span><span class="va">group</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Sampling 1000 Barcodes</span></span>
<span><span class="va">sample.vertices</span> <span class="op">&lt;-</span> <span class="fu">V</span><span class="op">(</span><span class="va">igraph.object</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">igraph.object</span><span class="op">)</span>, <span class="fl">1000</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">subgraph.object</span> <span class="op">&lt;-</span> <span class="fu">induced_subgraph</span><span class="op">(</span><span class="va">igraph.object</span>, vids <span class="op">=</span> <span class="va">sample.vertices</span><span class="op">)</span></span>
<span><span class="fu">V</span><span class="op">(</span><span class="va">subgraph.object</span><span class="op">)</span><span class="op">$</span><span class="va">degrees</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/degree.html" class="external-link">degree</a></span><span class="op">(</span><span class="va">subgraph.object</span><span class="op">)</span></span>
<span><span class="va">edge_alpha_color</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/adjustcolor.html" class="external-link">adjustcolor</a></span><span class="op">(</span><span class="st">"gray"</span>, alpha.f <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Plotting</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">subgraph.object</span>,</span>
<span>     layout <span class="op">=</span> <span class="fu">layout_nicely</span><span class="op">(</span><span class="va">subgraph.object</span><span class="op">)</span>,</span>
<span>     vertex.label <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>     vertex.size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">subgraph.object</span><span class="op">)</span><span class="op">$</span><span class="va">degrees</span><span class="op">)</span>, </span>
<span>     vertex.color <span class="op">=</span> <span class="va">col_samples</span><span class="op">[</span><span class="va">sample.vertices</span><span class="op">]</span>,</span>
<span>     vertex.frame.color <span class="op">=</span> <span class="st">"white"</span>, </span>
<span>     edge.color <span class="op">=</span> <span class="va">edge_alpha_color</span>,</span>
<span>     edge.arrow.size <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>     edge.curved <span class="op">=</span> <span class="fl">0.05</span>, </span>
<span>     margin <span class="op">=</span> <span class="op">-</span><span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topleft"</span>, </span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">color.legend</span><span class="op">)</span>, </span>
<span>       pch <span class="op">=</span> <span class="fl">16</span>, </span>
<span>       col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">col_samples</span><span class="op">)</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Cluster_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="returning-a-sparse-adjacency-matrix">Returning a Sparse Adjacency Matrix<a class="anchor" aria-label="anchor" href="#returning-a-sparse-adjacency-matrix"></a>
</h3>
<p>For computational applications, you can export a sparse adjacency
matrix using <code>exportAdjMatrix = TRUE</code>. This matrix represents
the connections between all barcodes in the input data, with the edit
distance that meet the threshold in places of connection.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate the sparse matrix</span></span>
<span><span class="va">adj.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clonalCluster.html">clonalCluster</a></span><span class="op">(</span><span class="va">combined.TCR</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                            chain <span class="op">=</span> <span class="st">"TRB"</span>,</span>
<span>                            exportAdjMatrix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View the dimensions and a snippet of the matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">adj.matrix</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 5698 5698</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">adj.matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 10 x 10 sparse Matrix of class "dgCMatrix"</span></span></code></pre>
<pre><code><span><span class="co">##                                                                          </span></span>
<span><span class="co">## P17B_AAACCTGAGTACGACG-1 . . .      .      . .      .      .      . .     </span></span>
<span><span class="co">## P17B_AAACCTGCAACACGCC-1 . . .      .      . .      .      .      . .     </span></span>
<span><span class="co">## P17B_AAACCTGCAGGCGATA-1 . . .       1e-06 .  1e-06  1e-06  1e-06 .  1e-06</span></span>
<span><span class="co">## P17B_AAACCTGCATGAGCGA-1 . .  1e-06 .      .  1e-06  1e-06  1e-06 .  1e-06</span></span>
<span><span class="co">## P17B_AAACGGGAGAGCCCAA-1 . . .      .      . .      .      .      . .     </span></span>
<span><span class="co">## P17B_AAACGGGAGCGTTTAC-1 . .  1e-06  1e-06 . .       1e-06  1e-06 .  1e-06</span></span>
<span><span class="co">## P17B_AAACGGGAGGGCACTA-1 . .  1e-06  1e-06 .  1e-06 .       1e-06 .  1e-06</span></span>
<span><span class="co">## P17B_AAACGGGAGTGGTCCC-1 . .  1e-06  1e-06 .  1e-06  1e-06 .      .  1e-06</span></span>
<span><span class="co">## P17B_AAACGGGCAGTTAACC-1 . . .      .      . .      .      .      . .     </span></span>
<span><span class="co">## P17B_AAACGGGGTCGCCATG-1 . .  1e-06  1e-06 .  1e-06  1e-06  1e-06 . .</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="using-both-chains">Using Both Chains<a class="anchor" aria-label="anchor" href="#using-both-chains"></a>
</h3>
<p>You can analyze the combined network of both TRA/TRB or IGH/IGL
chains by setting <code>chain = "both"</code>. This will create a single
cluster column named <code>Multi.Cluster</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Cluster using both TRB and TRA chains simultaneously</span></span>
<span><span class="va">clustered_both</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clonalCluster.html">clonalCluster</a></span><span class="op">(</span><span class="va">combined.TCR</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                                chain <span class="op">=</span> <span class="st">"both"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View the new "Multi.Cluster" column</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">clustered_both</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"barcode"</span>, <span class="st">"TCR1"</span>, <span class="st">"TCR2"</span>, <span class="st">"Multi.Cluster"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                   barcode                     TCR1                        TCR2</span></span>
<span><span class="co">## 1 P17B_AAACCTGAGTACGACG-1       TRAV25.TRAJ20.TRAC  TRBV5-1.None.TRBJ2-7.TRBC2</span></span>
<span><span class="co">## 2 P17B_AAACCTGCAACACGCC-1 TRAV38-2/DV8.TRAJ52.TRAC TRBV10-3.None.TRBJ2-2.TRBC2</span></span>
<span><span class="co">## 3 P17B_AAACCTGCAGGCGATA-1      TRAV12-1.TRAJ9.TRAC    TRBV9.None.TRBJ2-2.TRBC2</span></span>
<span><span class="co">## 4 P17B_AAACCTGCATGAGCGA-1      TRAV12-1.TRAJ9.TRAC    TRBV9.None.TRBJ2-2.TRBC2</span></span>
<span><span class="co">## 5 P17B_AAACGGGAGAGCCCAA-1        TRAV20.TRAJ8.TRAC                        &lt;NA&gt;</span></span>
<span><span class="co">## 6 P17B_AAACGGGAGCGTTTAC-1      TRAV12-1.TRAJ9.TRAC    TRBV9.None.TRBJ2-2.TRBC2</span></span>
<span><span class="co">##   Multi.Cluster</span></span>
<span><span class="co">## 1          &lt;NA&gt;</span></span>
<span><span class="co">## 2          &lt;NA&gt;</span></span>
<span><span class="co">## 3    cluster.23</span></span>
<span><span class="co">## 4    cluster.23</span></span>
<span><span class="co">## 5   cluster.145</span></span>
<span><span class="co">## 6    cluster.23</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="using-different-clustering-algorithms">Using Different Clustering Algorithms<a class="anchor" aria-label="anchor" href="#using-different-clustering-algorithms"></a>
</h3>
<p>While the default <code>cluster.method = "components"</code> is
robust, you can use other algorithms from the igraph package, such as
<code>walktrap</code> or <code>louvain</code>, to potentially uncover
different community structures.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Cluster using the walktrap algorithm</span></span>
<span><span class="va">graph_walktrap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clonalCluster.html">clonalCluster</a></span><span class="op">(</span><span class="va">combined.TCR</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                                cluster.method <span class="op">=</span> <span class="st">"walktrap"</span>,</span>
<span>                                exportGraph <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compare the number of clusters found</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu">V</span><span class="op">(</span><span class="va">graph_walktrap</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 391</span></span></code></pre>
<p>Overall, <code><a href="../reference/clonalCluster.html">clonalCluster()</a></code> is a versatile function for
defining and analyzing clonal relationships based on sequence
similarity. It allows researchers to move beyond exact sequence matches,
providing a more comprehensive understanding of clonal families. The
ability to customize parameters like <code>threshold</code>,
<code>chain</code> selection, and <code>group.by</code> ensures
adaptability to diverse research questions. Furthermore, the option to
export <code>igraph</code> objects or sparse adjacency matrices provides
advanced users with the tools for in-depth network analysis.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
