<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Comparing Clonal Diversity and Overlap • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Comparing Clonal Diversity and Overlap">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.6.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Repertoire Analysis</li>
    <li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Working with Single-Cell Objects</li>
    <li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Deep Learning Integrations</li>
    <li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other Workflow Integrations</li>
    <li>
      <a href="../articles/immcantation.html">Combining Immcantation and scRepertoire</a>
    </li>
    <li>
      <a href="../articles/tcrpheno.html">TCRpheno applied to T cell fate</a>
    </li>
    <li>
      <a href="../articles/immunarch.html">Using scRepertoire with immunarch</a>
    </li>
    <li>
      <a href="../articles/dandelionR.html">VDJ Trajectory Analysis with dandelionR</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BorchLab/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Comparing Clonal Diversity and Overlap</h1>
            
            <h4 data-toc-skip class="date">Compiled: October 31,
2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BorchLab/scRepertoire/blob/HEAD/vignettes/articles/Clonal_Diversity.Rmd" class="external-link"><code>vignettes/articles/Clonal_Diversity.Rmd</code></a></small>
      <div class="hidden name"><code>Clonal_Diversity.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="clonaldiversity-clonal-diversity-quantification">clonalDiversity: Clonal Diversity Quantification<a class="anchor" aria-label="anchor" href="#clonaldiversity-clonal-diversity-quantification"></a>
</h2>
<p>Diversity can be measured for samples or by other variables.
<code><a href="../reference/clonalDiversity.html">clonalDiversity()</a></code> calculates a specified diversity metric
for groups within a dataset. To control for variations in library size,
the function can perform bootstrapping with downsampling. It resamples
each group to the size of the smallest group and calculates the
diversity metric across multiple iterations, returning the mean
values.</p>
<div class="section level3">
<h3 id="how-clonaldiversity-handles-downsampling-and-bootstrapping">How <code>clonalDiversity()</code> Handles Downsampling and
Bootstrapping<a class="anchor" aria-label="anchor" href="#how-clonaldiversity-handles-downsampling-and-bootstrapping"></a>
</h3>
<ul>
<li>It determines the number of clones in the smallest group.</li>
<li>For each group, it performs <code>n.boots</code> iterations (default
= 100).</li>
<li>In each iteration, it randomly samples the clones (with replacement)
down to the size of the smallest group.</li>
<li>It calculates the selected diversity metric on this downsampled
set.</li>
<li>The final reported diversity value is the mean of the results from
all bootstrap iterations.</li>
<li>This process can be disabled by setting
<code>skip.boots = TRUE</code>.</li>
</ul>
</div>
<div class="section level3">
<h3 id="available-diversity-metrics-metric">Available Diversity Metrics (<code>metric</code>)<a class="anchor" aria-label="anchor" href="#available-diversity-metrics-metric"></a>
</h3>
<table class="table">
<colgroup>
<col width="6%">
<col width="50%">
<col width="7%">
<col width="36%">
</colgroup>
<thead><tr class="header">
<th align="left"><strong>Metric</strong></th>
<th align="left"><strong>What it measures</strong></th>
<th align="left"><strong>Typical scale</strong></th>
<th align="left"><strong>Higher value indicates…</strong></th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left"><code>shannon</code></td>
<td align="left">Joint <em>richness + evenness</em> of clonotypes;
moderately sensitive to rare clones (entropy of the frequency
distribution).</td>
<td align="left">0 → ∞</td>
<td align="left">
<strong>More distinct clonotypes</strong> <em>and</em>
a more even clone-size distribution.</td>
</tr>
<tr class="even">
<td align="left"><code>inv.simpson</code></td>
<td align="left">Evenness-weighted diversity (inverse Simpson’s D);
heavily penalises dominance by a few large clones.</td>
<td align="left">1 → ∞</td>
<td align="left">
<strong>Few highly dominant clones</strong>; diversity
spread across many mid-abundance clones.</td>
</tr>
<tr class="odd">
<td align="left"><code>norm.entropy</code></td>
<td align="left">Shannon entropy divided by ln <em>S</em> (where
<em>S</em> = no. clonotypes); removes library-size dependence.</td>
<td align="left">0 → 1</td>
<td align="left">Same interpretation as <code>shannon</code>, but
<strong>comparable across samples of different depths</strong>.</td>
</tr>
<tr class="even">
<td align="left"><code>gini.simpson</code></td>
<td align="left">1 – Simpson’s D; probability that two reads drawn at
random come from <strong>different</strong> clonotypes.</td>
<td align="left">0 → 1</td>
<td align="left">
<strong>Greater heterogeneity</strong> (higher chance
the two reads represent distinct clones).</td>
</tr>
<tr class="odd">
<td align="left"><code>chao1</code></td>
<td align="left">Estimated <em>richness only</em> (number of
clonotypes), correcting for unseen singletons/doubletons.</td>
<td align="left">0 → ∞</td>
<td align="left">
<strong>More unique clonotypes</strong>, regardless of
their relative abundances.</td>
</tr>
<tr class="even">
<td align="left"><code>ACE</code></td>
<td align="left">Abundance-based Coverage Estimator—alternative richness
estimator robust when many clones are rare.</td>
<td align="left">0 → ∞</td>
<td align="left">
<strong>More unique clonotypes</strong>, especially
when the repertoire contains many low-frequency clones.</td>
</tr>
<tr class="odd">
<td align="left"><code>gini</code></td>
<td align="left">Inequality of clone size distribution (Gini
Coefficient); measures how unevenly clone sizes are distributed.</td>
<td align="left">0 → 1</td>
<td align="left">
<strong>Greater inequality</strong>; a few highly
abundant clones dominate the repertoire.</td>
</tr>
<tr class="even">
<td align="left"><code>d50</code></td>
<td align="left">Repertoire dominance; how many of the top clones are
needed to constitute 50% of the total library.</td>
<td align="left">1 → ∞</td>
<td align="left">
<strong>Less dominance</strong> by top clones; the
repertoire is more evenly distributed.</td>
</tr>
<tr class="odd">
<td align="left"><code>hill0</code></td>
<td align="left">Richness (Hill number q=0); the total number of unique
clonotypes observed in the sample.</td>
<td align="left">0 → ∞</td>
<td align="left">
<strong>More unique clonotypes</strong> (greater raw
richness).</td>
</tr>
<tr class="even">
<td align="left"><code>hill1</code></td>
<td align="left">Effective number of abundant clonotypes (Hill number
q=1); the exponential of Shannon entropy.</td>
<td align="left">1 → ∞</td>
<td align="left">
<strong>Higher diversity</strong> of common clonotypes,
balancing richness and evenness.</td>
</tr>
<tr class="odd">
<td align="left"><code>hill2</code></td>
<td align="left">Effective number of dominant clonotypes (Hill number
q=2); equivalent to the inverse Simpson index.</td>
<td align="left">1 → ∞</td>
<td align="left">
<strong>Higher diversity</strong> among the most
dominant clonotypes in the repertoire.</td>
</tr>
</tbody>
</table>
<p>Key Parameters for <code><a href="../reference/clonalDiversity.html">clonalDiversity()</a></code></p>
<ul>
<li>
<code>metric</code>: The diversity metric to calculate.</li>
<li>
<code>x.axis</code>: An additional metadata variable to group
samples along the x-axis.</li>
<li>
<code>return.boots</code>: If <code>TRUE</code>, returns all
bootstrap values instead of the mean.</li>
<li>
<code>skip.boots</code>: If <code>TRUE</code>, disables downsampling
and bootstrapping.</li>
<li>
<code>n.boots</code>: The number of bootstrap iterations to perform
(default is 100).</li>
</ul>
<p>To calculate Shannon diversity, grouped by sample:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalDiversity.html">clonalDiversity</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                cloneCall <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-1-1.png" width="700"></p>
<p>There are two options for grouping in <code><a href="../reference/clonalDiversity.html">clonalDiversity()</a></code>:
<code>group.by</code> and <code>x.axis</code>.</p>
<ul>
<li>
<code>group.by</code>: Reorganizes the clone information into new
groups that the calculation will be based on.</li>
<li>
<code>x.axis</code>: Keeps the organization of the clone information
the same, but plots along the x-axis for improved visibility or
grouping.</li>
</ul>
<p>First, add a “Patient” variable to combined.TCR</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">combined.TCR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addVariable.html">addVariable</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                            variable.name <span class="op">=</span> <span class="st">"Patient"</span>, </span>
<span>                             variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17"</span>, <span class="st">"P17"</span>, <span class="st">"P18"</span>, <span class="st">"P18"</span>, </span>
<span><span class="st">"P19"</span>,<span class="st">"P19"</span>, <span class="st">"P20"</span>, <span class="st">"P20"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Now, calculate Inverse Simpson diversity, grouped by “Patient”:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalDiversity.html">clonalDiversity</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                cloneCall <span class="op">=</span> <span class="st">"gene"</span>, </span>
<span>                group.by <span class="op">=</span> <span class="st">"Patient"</span>, </span>
<span>                metric <span class="op">=</span> <span class="st">"inv.simpson"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>Calculate Inverse Simpson diversity with “Patient” on the x-axis,
keeping the original grouping:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalDiversity.html">clonalDiversity</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                cloneCall <span class="op">=</span> <span class="st">"gene"</span>, </span>
<span>                x.axis <span class="op">=</span> <span class="st">"Patient"</span>, </span>
<span>                metric <span class="op">=</span> <span class="st">"inv.simpson"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p><code><a href="../reference/clonalDiversity.html">clonalDiversity()</a></code> functions in quantifying and comparing
the diversity of immune repertoires across different samples or
experimental conditions. By implementing bootstrapping and offering a
wide range of diversity metrics, it provides robust and comparable
measures of clonal richness and evenness, which are fundamental for
understanding immune responses and disease states</p>
</div>
</div>
<div class="section level2">
<h2 id="clonalrarefaction-sampling-based-extrapolation">clonalRarefaction: Sampling-based Extrapolation<a class="anchor" aria-label="anchor" href="#clonalrarefaction-sampling-based-extrapolation"></a>
</h2>
<p><code><a href="../reference/clonalRarefaction.html">clonalRarefaction()</a></code> uses Hill numbers to estimate
rarefaction, or estimating species richness, based on the abundance of
clones across groupings. The underlying rarefaction calculation uses the
observed receptor abundance to compute diversity. This function relies
on the <code>iNEXT</code> R package.</p>
<div class="section level3">
<h3 id="hill-numbers-and-their-interpretation">Hill Numbers and Their Interpretation<a class="anchor" aria-label="anchor" href="#hill-numbers-and-their-interpretation"></a>
</h3>
<ul>
<li>
<code>0</code>: Species richness</li>
<li>
<code>1</code>: Shannon Diversity</li>
<li>
<code>2</code>: Simpson Diversity</li>
</ul>
</div>
<div class="section level3">
<h3 id="plot-type-options">plot.type Options<a class="anchor" aria-label="anchor" href="#plot-type-options"></a>
</h3>
<ul>
<li>
<code>1</code>: Sample-size-based rarefaction/extrapolation
curve</li>
<li>
<code>2</code>: Sample completeness curve</li>
<li>
<code>3</code>: Coverage-based rarefaction/extrapolation curve</li>
</ul>
<p>Key Parameter(s) for <code><a href="../reference/clonalRarefaction.html">clonalRarefaction()</a></code></p>
<ul>
<li>
<code>plot.type</code>: The type of plot to generate.</li>
<li>
<code>hill.numbers</code>: The Hill numbers to be plotted (0, 1, or
2).</li>
<li>
<code>n.boots</code>: The number of bootstrap replicates used to
derive confidence intervals (default is 20).</li>
</ul>
<p>This function relies on the <a href="https://cran.r-project.org/web/packages/iNEXT/index.html" class="external-link">iNEXT</a>
with the accompanying <a href="http://chao.stat.nthu.edu.tw/wordpress/paper/120_pdf_appendix.pdf" class="external-link">manuscript</a>.
Like the other wrapping functions in scRepertoire, please cite the
original work. The sample completeness curve (<strong>plot.type</strong>
= 2), may not show full sample coverage due to the size/diversity of the
input data.</p>
</div>
<div class="section level3">
<h3 id="rarefaction-using-species-richness-q-0">Rarefaction using Species Richness (q = 0)<a class="anchor" aria-label="anchor" href="#rarefaction-using-species-richness-q-0"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalRarefaction.html">clonalRarefaction</a></span><span class="op">(</span><span class="va">combined.TCR</span>,</span>
<span>                  plot.type <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                  hill.numbers <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                  n.boots <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalRarefaction.html">clonalRarefaction</a></span><span class="op">(</span><span class="va">combined.TCR</span>,</span>
<span>                  plot.type <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                  hill.numbers <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                  n.boots <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-5-2.png" width="700"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalRarefaction.html">clonalRarefaction</a></span><span class="op">(</span><span class="va">combined.TCR</span>,</span>
<span>                  plot.type <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                  hill.numbers <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                  n.boots <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-5-3.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="rarefaction-using-shannon-diversity-q-1">Rarefaction using Shannon Diversity (q = 1)<a class="anchor" aria-label="anchor" href="#rarefaction-using-shannon-diversity-q-1"></a>
</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalRarefaction.html">clonalRarefaction</a></span><span class="op">(</span><span class="va">combined.TCR</span>,</span>
<span>                  plot.type <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                  hill.numbers <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                  n.boots <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalRarefaction.html">clonalRarefaction</a></span><span class="op">(</span><span class="va">combined.TCR</span>,</span>
<span>                  plot.type <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                  hill.numbers <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                  n.boots <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-6-2.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalRarefaction.html">clonalRarefaction</a></span><span class="op">(</span><span class="va">combined.TCR</span>,</span>
<span>                  plot.type <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                  hill.numbers <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                  n.boots <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-6-3.png" width="700"></p>
<p><code><a href="../reference/clonalRarefaction.html">clonalRarefaction()</a></code> provides robust estimates of immune
repertoire diversity, allowing for comparison of richness and effective
diversity across samples, even with varying sequencing depths. By
visualizing rarefaction and extrapolation curves, researchers can assess
the completeness of their sampling and make fair comparisons of
diversity, which is essential for understanding immune complexity.</p>
</div>
</div>
<div class="section level2">
<h2 id="clonalsizedistribution-advanced-modeling-of-clonal-composition">clonalSizeDistribution: Advanced Modeling Of Clonal Composition<a class="anchor" aria-label="anchor" href="#clonalsizedistribution-advanced-modeling-of-clonal-composition"></a>
</h2>
<p>Another method for modeling the repertoire distribution is a discrete
gamma-GPD spliced threshold model, proposed by <a href="https://pubmed.ncbi.nlm.nih.gov/30485278/" class="external-link">Koch et al.</a> The
spliced model models the repertoire and allows for the application of a
power law distribution for larger clonal-expanded sequences and a
Poisson distribution for smaller clones. After fitting the models,
repertoires can be compared using Euclidean distance. If using this
function, please read/cite <a href="https://pubmed.ncbi.nlm.nih.gov/30485278/" class="external-link">Koch et al.</a> and
check out the <a href="https://bioconductor.org/packages/release/bioc/html/powerTCR.html" class="external-link">powerTCR</a>
R package.</p>
<p>Key Parameter(s) for <code><a href="../reference/clonalSizeDistribution.html">clonalSizeDistribution()</a></code></p>
<ul>
<li>
<code>method</code>: The agglomeration method for hierarchical
clustering (e.g., “ward.D2”).</li>
</ul>
<p>To model the clonal size distribution using amino acid clone calls
and hierarchical clustering with the “ward.D2” method:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalSizeDistribution.html">clonalSizeDistribution</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                       cloneCall <span class="op">=</span> <span class="st">"aa"</span>, </span>
<span>                       method<span class="op">=</span> <span class="st">"ward.D2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p><code><a href="../reference/clonalSizeDistribution.html">clonalSizeDistribution()</a></code> offers a sophisticated approach
to modeling immune repertoire composition, distinguishing between the
distribution of rare and expanded clones. By applying a spliced
statistical model, it provides a more accurate representation of the
repertoire’s underlying clonal architecture, enabling robust comparisons
and a deeper understanding of immune system dynamics.</p>
</div>
<div class="section level2">
<h2 id="clonaloverlap-exploring-sequence-overlap">clonalOverlap: Exploring Sequence Overlap<a class="anchor" aria-label="anchor" href="#clonaloverlap-exploring-sequence-overlap"></a>
</h2>
<p>If you are interested in measures of similarity between the samples
loaded into scRepertoire, using <code><a href="../reference/clonalOverlap.html">clonalOverlap()</a></code> can assist
in the visualization. The underlying <code><a href="../reference/clonalOverlap.html">clonalOverlap()</a></code>
calculation varies by the <code>method</code> parameter.</p>
<div class="section level3">
<h3 id="method-parameters-for-clonaloverlap">
<code>method</code> Parameters for <code>clonalOverlap()</code><a class="anchor" aria-label="anchor" href="#method-parameters-for-clonaloverlap"></a>
</h3>
<ul>
<li>
<code>overlap</code> - Overlap coefficient</li>
<li>
<code>morisita</code> - Morisita’s overlap index</li>
<li>
<code>jaccard</code> - Jaccard index</li>
<li>
<code>cosine</code> - Cosine similarity</li>
<li>
<code>raw</code> - Exact number of overlapping clones</li>
</ul>
<p>To calculate and visualize the Morisita overlap index using strict
clone calls:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalOverlap.html">clonalOverlap</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>              cloneCall <span class="op">=</span> <span class="st">"strict"</span>, </span>
<span>              method <span class="op">=</span> <span class="st">"morisita"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>To calculate and visualize the raw number of overlapping clones using
strict clone calls:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalOverlap.html">clonalOverlap</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>              cloneCall <span class="op">=</span> <span class="st">"strict"</span>, </span>
<span>              method <span class="op">=</span> <span class="st">"raw"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p><code><a href="../reference/clonalOverlap.html">clonalOverlap()</a></code> is a tool for assessing the similarity
and shared clonotypes between different immune repertoires. By offering
various quantitative methods and a clear heatmap visualization, it
allows researchers to identify the degree of overlap, providing insights
into shared immune responses, cross-reactivity, or the impact of
different treatments on clonal composition.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
