<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Quantifying Clonal Bias • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Quantifying Clonal Bias">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/dandelionR.html">Immune Repertoire Trajectory Analysis</a>
    </li>
    <li>
      <a href="../articles/immunarch.html">Using scRepertoire with immunarch</a>
    </li>
    <li>
      <a href="../articles/tcrpheno.html">TCRpheno applied to T cell fate</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BorchLab/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Quantifying Clonal Bias</h1>
            
            <h4 data-toc-skip class="date">Compiled: July 14, 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BorchLab/scRepertoire/blob/HEAD/vignettes/articles/Clonal_Bias.Rmd" class="external-link"><code>vignettes/articles/Clonal_Bias.Rmd</code></a></small>
      <div class="hidden name"><code>Clonal_Bias.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="startracdiversity">StartracDiversity<a class="anchor" aria-label="anchor" href="#startracdiversity"></a>
</h2>
<p>From the excellent work by <a href="https://www.nature.com/articles/s41586-018-0694-x" class="external-link">Lei Zhang, et
al.</a>, the authors introduced new methods for looking at clones by
cellular origins and cluster identification. Their <a href="https://github.com/Japrin/STARTRAC" class="external-link">STARTRAC</a> software has been
adapted to work with <code>scRepertoire</code> and please read and cite
their excellent work.</p>
<p>To use the <code><a href="../reference/StartracDiversity.html">StartracDiversity()</a></code> function, you need to
include the product of the <code><a href="../reference/combineExpression.html">combineExpression()</a></code> function.
The second requirement is a column header in the metadata of the Seurat
object that has the tissue of origin. In the example data, type
corresponds to the column “Type”, which includes the “P” (peripheral
blood) and “L” (lung) classifiers. The indices can be subsetted for a
specific patient or examined overall using the group.by variable.
Importantly, the function uses only the strict definition of a clone,
which includes the VDJC genes and the CDR3 nucleotide sequence.</p>
<div class="section level3">
<h3 id="indices-output-from-startracdiversity">Indices Output from <code>StartracDiversity()</code><a class="anchor" aria-label="anchor" href="#indices-output-from-startracdiversity"></a>
</h3>
<ul>
<li>
<code>expa</code> - Clonal Expansion<br>
</li>
<li>
<code>migr</code> - Cross-tissue Migration<br>
</li>
<li>
<code>tran</code> - State Transition</li>
</ul>
<p>Key Parameters for <code><a href="../reference/StartracDiversity.html">StartracDiversity()</a></code></p>
<ul>
<li>
<code>type</code>: The variable in the metadata that provides tissue
type.</li>
<li>
<code>group.by</code>: A column header in the metadata to group the
analysis by (e.g., “sample”, “treatment”).</li>
</ul>
<p>Calculate and plot Startrac diversity metrics using “Type” for tissue
and “Patient” for grouping:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/StartracDiversity.html">StartracDiversity</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                  type <span class="op">=</span> <span class="st">"Type"</span>, </span>
<span>                  group.by <span class="op">=</span> <span class="st">"Patient"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Bias_files/figure-html/unnamed-chunk-1-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="clonalbias">clonalBias<a class="anchor" aria-label="anchor" href="#clonalbias"></a>
</h2>
<p>A new metric proposed by <a href="https://pubmed.ncbi.nlm.nih.gov/35829695/" class="external-link">Massimo et al</a>,
<code><a href="../reference/clonalBias.html">clonalBias()</a></code>, like STARTRAC, is a clonal metric that seeks
to quantify how individual clones are skewed towards a specific cellular
compartment or cluster. A clone bias of <code>1</code> indicates that a
clone is composed of cells from a single compartment or cluster, while a
clone bias of <code>0</code> matches the background subtype
distribution. Please read and cite the linked manuscript if using
<code><a href="../reference/clonalBias.html">clonalBias()</a></code></p>
<p>Key Parameter(s) for <code><a href="../reference/clonalBias.html">clonalBias()</a></code></p>
<ul>
<li>
<code>group.by</code>: A column header in the metadata that bias
will be based on.</li>
<li>
<code>split.by</code>: The variable to use for calculating the
baseline frequencies (e.g., “Type” for lung vs peripheral blood
comparison)</li>
<li>
<code>n.boots</code>: Number of bootstraps to downsample.</li>
<li>
<code>min.expand</code>: Clone frequency cut-off for the purpose of
comparison (default = 10).</li>
</ul>
<p>Here we calculate and plot clonal bias using <code>aa</code> clone
calls, splitting by “Patient” and grouping by “seurat_clusters”, with a
minimum expansion of 5 and 10 bootstraps:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalBias.html">clonalBias</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>           cloneCall <span class="op">=</span> <span class="st">"aa"</span>, </span>
<span>           split.by <span class="op">=</span> <span class="st">"Patient"</span>, </span>
<span>           group.by <span class="op">=</span> <span class="st">"seurat_clusters"</span>,</span>
<span>           n.boots <span class="op">=</span> <span class="fl">10</span>, </span>
<span>           min.expand <span class="op">=</span><span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Bias_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
