<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Basic Clonal Visualizations • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Basic Clonal Visualizations">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/dandelionR.html">VDJ Trajectory Analysis with dandelionR</a>
    </li>
    <li>
      <a href="../articles/immunarch.html">Using scRepertoire with immunarch</a>
    </li>
    <li>
      <a href="../articles/tcrpheno.html">TCRpheno applied to T cell fate</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BorchLab/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Basic Clonal Visualizations</h1>
            
            <h4 data-toc-skip class="date">Compiled: July 22, 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BorchLab/scRepertoire/blob/HEAD/vignettes/articles/Clonal_Visualizations.Rmd" class="external-link"><code>vignettes/articles/Clonal_Visualizations.Rmd</code></a></small>
      <div class="hidden name"><code>Clonal_Visualizations.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="clonalquant-quantifying-unique-clones">clonalQuant: Quantifying Unique Clones<a class="anchor" aria-label="anchor" href="#clonalquant-quantifying-unique-clones"></a>
</h2>
<p>The <code><a href="../reference/clonalQuant.html">clonalQuant()</a></code> function is used to explore the clones
by returning the total or relative numbers of unique clones.</p>
<p>Key Parameter(s) for <code><a href="../reference/clonalQuant.html">clonalQuant()</a></code></p>
<ul>
<li>
<code>scale</code>: If <code>TRUE</code>, converts the output to the
relative percentage of unique clones scaled by the total repertoire
size; if <code>FALSE</code> (default), reports the total number of
unique clones.</li>
</ul>
<p>To visualize the relative percent of unique clones across all chains
(<code>"both"</code>) using the <code>strict</code> clone
definition:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalQuant.html">clonalQuant</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>            cloneCall<span class="op">=</span><span class="st">"strict"</span>, </span>
<span>            chain <span class="op">=</span> <span class="st">"both"</span>, </span>
<span>            scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Visualizations_files/figure-html/unnamed-chunk-1-1.png" width="700"></p>
<p>Another option is to define the visualization by data classes using
the <code>group.by</code> parameter. Here, we’ll use the
<code>"Type"</code> variable, which was previously added to the
combined.TCR list.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalQuant.html">clonalQuant</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>            cloneCall<span class="op">=</span><span class="st">"gene"</span>, </span>
<span>            group.by <span class="op">=</span> <span class="st">"Type"</span>, </span>
<span>            scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Visualizations_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="clonalabundance-distribution-of-clones-by-size">clonalAbundance: Distribution of Clones by Size<a class="anchor" aria-label="anchor" href="#clonalabundance-distribution-of-clones-by-size"></a>
</h2>
<p><code><a href="../reference/clonalAbundance.html">clonalAbundance()</a></code> allows for the examination of the
relative distribution of clones by abundance. It produces a line graph
showing the total number of clones at specific frequencies within a
sample or group.</p>
<p>Key Parameter(s) for <code><a href="../reference/clonalAbundance.html">clonalAbundance()</a></code></p>
<ul>
<li>
<code>scale</code>: If <code>TRUE</code>, converts the graphs into
density plots to show relative distributions; if <code>FALSE</code>
(default), displays raw counts.</li>
</ul>
<p>To visualize the raw clonal abundance using the <code>gene</code>
clone definition:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalAbundance.html">clonalAbundance</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                cloneCall <span class="op">=</span> <span class="st">"gene"</span>, </span>
<span>                scale <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Visualizations_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p><code><a href="../reference/clonalAbundance.html">clonalAbundance()</a></code> output can also be converted into a
density plot, which may allow for better comparisons between different
repertoire sizes, by setting <code>scale = TRUE</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalAbundance.html">clonalAbundance</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                cloneCall <span class="op">=</span> <span class="st">"gene"</span>, </span>
<span>                scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Visualizations_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="clonallength-distribution-of-sequence-lengths">clonalLength: Distribution of Sequence Lengths<a class="anchor" aria-label="anchor" href="#clonallength-distribution-of-sequence-lengths"></a>
</h2>
<p><code><a href="../reference/clonalLength.html">clonalLength()</a></code> allows you to look at the length
distribution of the CDR3 sequences. Importantly, unlike the other basic
visualizations, the <code>cloneCall</code> can only be <code>nt</code>
(nucleotide) or <code>aa</code> (amino acid). Due to the method of
calling clones as outlined previously (e.g., using NA for unreturned
chain sequences or multiple chains within a single barcode), the length
distribution may reveal a multimodal curve.</p>
<p>To visualize the amino acid length distribution for both chains
(“both”):</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalLength.html">clonalLength</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>             cloneCall<span class="op">=</span><span class="st">"aa"</span>, </span>
<span>             chain <span class="op">=</span> <span class="st">"both"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="Clonal_Visualizations_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>To visualize the amino acid length distribution for the
<code>TRA</code> chain, scaled as a density plot:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalLength.html">clonalLength</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>             cloneCall<span class="op">=</span><span class="st">"aa"</span>, </span>
<span>             chain <span class="op">=</span> <span class="st">"TRA"</span>, </span>
<span>             scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span></code></pre></div>
<p><img src="Clonal_Visualizations_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="clonalcompare-clonal-dynamics-between-categorical-variables">clonalCompare: Clonal Dynamics Between Categorical Variables<a class="anchor" aria-label="anchor" href="#clonalcompare-clonal-dynamics-between-categorical-variables"></a>
</h2>
<p><code><a href="../reference/clonalCompare.html">clonalCompare()</a></code> allows you to look at clones between
samples and changes in dynamics. It is useful for tracking how the
proportions of top clones change between conditions.</p>
<p>Key Parameters for <code><a href="../reference/clonalCompare.html">clonalCompare()</a></code></p>
<ul>
<li>
<code>samples</code>: A character vector to isolate specific samples
by their list element name.</li>
<li>
<code>clones</code>: A character vector of specific clonal sequences
to visualize. If used, <code>top.clones</code> will be ignored.
*<code>top.clones</code>: The top n number of clones to graph,
calculated based on the frequency within individual samples.</li>
<li>
<code>highlight.clones</code>: A character vector of specific clonal
sequences to color; all other clones will be greyed out.</li>
<li>
<code>relabel.clones</code>: If <code>TRUE</code>, simplifies the
legend by labeling isolated clones numerically (e.g., “Clone: 1”).</li>
<li>
<code>graph</code>: The type of plot to generate;
<code>alluvial</code> (default) or <code>area</code>.</li>
<li>
<code>proportion</code>: If <code>TRUE</code> (default), the y-axis
represents proportional abundance; if <code>FALSE</code>, it represents
raw clone counts.</li>
</ul>
<p>To compare the top 10 clones between samples “P17B” and “P17L” using
amino acid sequences as an alluvial plot:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalCompare.html">clonalCompare</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                  top.clones <span class="op">=</span> <span class="fl">10</span>, </span>
<span>                  samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17B"</span>, <span class="st">"P17L"</span><span class="op">)</span>, </span>
<span>                  cloneCall<span class="op">=</span><span class="st">"aa"</span>, </span>
<span>                  graph <span class="op">=</span> <span class="st">"alluvial"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Visualizations_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>We can also choose to highlight specific clones, such as in the case
of <em>“CVVSDNTGGFKTIF_CASSVRRERANTGELFF”</em> and
<em>“NA_CASSVRRERANTGELFF”</em> using the <code>highlight.clones</code>
parameter. In addition, we can simplify the plot to label the clones as
“Clone: 1”, “Clone: 2”, etc., by setting
<code>relabel.clones = TRUE</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalCompare.html">clonalCompare</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>              top.clones <span class="op">=</span> <span class="fl">10</span>,</span>
<span>              highlight.clones <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CVVSDNTGGFKTIF_CASSVRRERANTGELFF"</span>, <span class="st">"NA_CASSVRRERANTGELFF"</span><span class="op">)</span>,</span>
<span>              relabel.clones <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>              samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17B"</span>, <span class="st">"P17L"</span><span class="op">)</span>, </span>
<span>              cloneCall<span class="op">=</span><span class="st">"aa"</span>, </span>
<span>              graph <span class="op">=</span> <span class="st">"alluvial"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Visualizations_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>Alternatively, if we only want to show specific clones, we can use
the <code>clones</code> parameter.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalCompare.html">clonalCompare</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>              clones <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CVVSDNTGGFKTIF_CASSVRRERANTGELFF"</span>, <span class="st">"NA_CASSVRRERANTGELFF"</span><span class="op">)</span>,</span>
<span>              relabel.clones <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>              samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17B"</span>, <span class="st">"P17L"</span><span class="op">)</span>, </span>
<span>              cloneCall<span class="op">=</span><span class="st">"aa"</span>, </span>
<span>              graph <span class="op">=</span> <span class="st">"alluvial"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Visualizations_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="clonalscatter-scatterplot-of-two-variables">clonalScatter: Scatterplot of Two Variables<a class="anchor" aria-label="anchor" href="#clonalscatter-scatterplot-of-two-variables"></a>
</h2>
<p>clonalScatter() organizes two repertoires, quantifies their relative
clone sizes, and produces a scatter plot comparing the two samples.
Clones are categorized by counts into singlets or expanded, either
exclusively present or shared between the selected samples.</p>
<p>Key Parameter(s) for <code><a href="../reference/clonalScatter.html">clonalScatter()</a></code></p>
<ul>
<li>
<code>x.axis</code>, <code>y.axis</code>: Names of the list elements
or meta data variable to place on the x-axis and y-axis.</li>
<li>
<code>dot.size</code>: Specifies how dot size is determined;
<code>total</code> (default) displays the total number of clones between
the x- and y-axis, or a specific list element name for size
calculation.</li>
<li>
<code>graph</code>: The type of graph to display;
<code>proportion</code> for the relative proportion of clones (default)
or <code>count</code> for the total count of clones by sample.</li>
</ul>
<p>To compare samples “P18B” and “P18L” based on <code>gene</code> clone
calls, with dot size representing the total number of clones, and
plotting clone proportions:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalScatter.html">clonalScatter</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>              cloneCall <span class="op">=</span><span class="st">"gene"</span>, </span>
<span>              x.axis <span class="op">=</span> <span class="st">"P18B"</span>, </span>
<span>              y.axis <span class="op">=</span> <span class="st">"P18L"</span>,</span>
<span>              dot.size <span class="op">=</span> <span class="st">"total"</span>,</span>
<span>              graph <span class="op">=</span> <span class="st">"proportion"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Visualizations_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
