<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Comparing Clonal Diversity and Overlap • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Comparing Clonal Diversity and Overlap">
<meta property="og:description" content="scRepertoire">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ncborcherding/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Comparing Clonal Diversity and Overlap</h1>
            
            <h4 data-toc-skip class="date">Compiled: February 24,
2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ncborcherding/scRepertoire/blob/HEAD/vignettes/articles/Clonal_Diversity.Rmd" class="external-link"><code>vignettes/articles/Clonal_Diversity.Rmd</code></a></small>
      <div class="hidden name"><code>Clonal_Diversity.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="clonaldiversity">clonalDiversity<a class="anchor" aria-label="anchor" href="#clonaldiversity"></a>
</h2>
<p>Diversity can also be measured for samples or by other variables.
Diversity <strong>metrics</strong> calculated, include:
<strong>“shannon”</strong>, <strong>“inv.simpson”</strong>,
<strong>“norm.entropy”</strong>, <strong>“gini.simpson”</strong>,
<strong>“chao1”</strong>, and <strong>“ACE”</strong>. Please see the
manual for more information on each metric and the underlying
calculations.</p>
<p>Inherent in diversity calculations is a bias for increasing diversity
with increasing repertoire size. <code><a href="../reference/clonalDiversity.html">clonalDiversity()</a></code> will
automatically downsample to the smallest repertoire size and perform
bootstrapping to return the mean diversity estimates. If the output of
diversity values are strange or minimally variable, it is likely due to
a sample with small repertoire size.</p>
<p><strong>n.boots</strong><br>
The number of calculations to perform <strong>(default =
100)</strong>.</p>
<p><strong>return.boots</strong></p>
<ul>
<li>TRUE: Return all the calculations.<br>
</li>
<li>FALSE: Return only the mean values <strong>(default)</strong>.</li>
</ul>
<p><strong>skip.boots</strong><br>
Skip the bootstrapping calculations.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalDiversity.html">clonalDiversity</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                cloneCall <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-1-1.png" width="700"></p>
<p>There are 2 options for grouping in <code><a href="../reference/clonalDiversity.html">clonalDiversity()</a></code>
and it is important to understand the difference.</p>
<p><strong>group.by</strong></p>
<ul>
<li>Reorganize the clone information into new groups that the
calculation will be based on.</li>
</ul>
<p><strong>x.axis</strong></p>
<ul>
<li>Keep the organization of the clone information the same, but will
plot along the x-axis for improved visibility or grouping.</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">combined.TCR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addVariable.html">addVariable</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                            variable.name <span class="op">=</span> <span class="st">"Patient"</span>, </span>
<span>                             variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17"</span>, <span class="st">"P17"</span>, <span class="st">"P18"</span>, <span class="st">"P18"</span>, </span>
<span><span class="st">"P19"</span>,<span class="st">"P19"</span>, <span class="st">"P20"</span>, <span class="st">"P20"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/clonalDiversity.html">clonalDiversity</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                cloneCall <span class="op">=</span> <span class="st">"gene"</span>, </span>
<span>                group.by <span class="op">=</span> <span class="st">"Patient"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalDiversity.html">clonalDiversity</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                cloneCall <span class="op">=</span> <span class="st">"gene"</span>, </span>
<span>                x.axis <span class="op">=</span> <span class="st">"Patient"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-2-2.png" width="700"></p>
<p>As a default, <code><a href="../reference/clonalDiversity.html">clonalDiversity()</a></code> will return all the
metrics calculated - <strong>“shannon”</strong>,
<strong>“inv.simpson”</strong>, <strong>“norm.entropy”</strong>,
<strong>“gini.simpson”</strong>, <strong>“chao1”</strong>, and
<strong>“ACE”</strong>. Selecting a single or a subset of these methods
using the <strong>metrics</strong> parameter.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Return only a subset of metrics</span></span>
<span><span class="fu"><a href="../reference/clonalDiversity.html">clonalDiversity</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                metrics <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"shannon"</span>, <span class="st">"ACE"</span><span class="op">)</span>,</span>
<span>                cloneCall <span class="op">=</span> <span class="st">"gene"</span>, </span>
<span>                n.boots <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="clonararefaction">clonaRarefaction<a class="anchor" aria-label="anchor" href="#clonararefaction"></a>
</h2>
<p>We can also use Hill numbers to estimate the rarefaction, or
estimating species richness, using the the abundance of clones across
groupings. Underlying the rarefaction calculation is the use of observed
receptor of <strong>abundance</strong> to compute diversity.</p>
<p><strong>hill.numbers</strong></p>
<ul>
<li>0 - species-richness<br>
</li>
<li>1 - Shannon Diversity</li>
<li>2 - Simpson Diversity</li>
</ul>
<p><strong>plot.type</strong></p>
<ul>
<li>1 - sample-size-based rarefaction/extrapolation<br>
</li>
<li>2 - sample completeness curve<br>
</li>
<li>3 - coverage-based rarefaction/extrapolation curve</li>
</ul>
<p>This function relies on the <a href="https://cran.r-project.org/web/packages/iNEXT/index.html" class="external-link">iNEXT</a>
with the accompanying <a href="http://chao.stat.nthu.edu.tw/wordpress/paper/120_pdf_appendix.pdf" class="external-link">manuscript</a>.
Like the other wrapping functions in scRepertoire, please cite the
original work. The sample completeness curve (<strong>plot.type</strong>
= 2), may not show full sample coverage due to the size/diversity of the
input data.</p>
<div class="section level3">
<h3 id="rarefaction-using-species-richness-q-0">Rarefaction using Species Richness (q = 0)<a class="anchor" aria-label="anchor" href="#rarefaction-using-species-richness-q-0"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalRarefaction.html">clonalRarefaction</a></span><span class="op">(</span><span class="va">combined.TCR</span>,</span>
<span>                  plot.type <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                  hill.numbers <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                  n.boots <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalRarefaction.html">clonalRarefaction</a></span><span class="op">(</span><span class="va">combined.TCR</span>,</span>
<span>                  plot.type <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                  hill.numbers <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                  n.boots <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-4-2.png" width="700"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalRarefaction.html">clonalRarefaction</a></span><span class="op">(</span><span class="va">combined.TCR</span>,</span>
<span>                  plot.type <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                  hill.numbers <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                  n.boots <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-4-3.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="rarefaction-using-shannon-diversity-q-1">Rarefaction using Shannon Diversity (q = 1)<a class="anchor" aria-label="anchor" href="#rarefaction-using-shannon-diversity-q-1"></a>
</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalRarefaction.html">clonalRarefaction</a></span><span class="op">(</span><span class="va">combined.TCR</span>,</span>
<span>                  plot.type <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                  hill.numbers <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                  n.boots <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalRarefaction.html">clonalRarefaction</a></span><span class="op">(</span><span class="va">combined.TCR</span>,</span>
<span>                  plot.type <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                  hill.numbers <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                  n.boots <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-5-2.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalRarefaction.html">clonalRarefaction</a></span><span class="op">(</span><span class="va">combined.TCR</span>,</span>
<span>                  plot.type <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                  hill.numbers <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                  n.boots <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-5-3.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="clonalsizedistribution">clonalSizeDistribution<a class="anchor" aria-label="anchor" href="#clonalsizedistribution"></a>
</h2>
<p>Another method for modeling the repertoire distribution is a discrete
gamma-GPD spliced threshold model, proposed by <a href="https://pubmed.ncbi.nlm.nih.gov/30485278/" class="external-link">Koch et al.</a> The
spliced model models the repertoire and allows for the application of a
power law distribution for larger clonal-expanded sequences and a
Poisson distribution for smaller clones. After fitting the models,
repertoires can be compared using Euclidean distance.</p>
<p>If using this function, please read/cite <a href="https://pubmed.ncbi.nlm.nih.gov/30485278/" class="external-link">Koch et al.</a> and
check out the <a href="https://bioconductor.org/packages/release/bioc/html/powerTCR.html" class="external-link">powerTCR</a>
R package.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalSizeDistribution.html">clonalSizeDistribution</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                       cloneCall <span class="op">=</span> <span class="st">"aa"</span>, </span>
<span>                       method<span class="op">=</span> <span class="st">"ward.D2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="clonaloverlap">clonalOverlap<a class="anchor" aria-label="anchor" href="#clonaloverlap"></a>
</h2>
<p>If you are interested in measures of similarity between the samples
loaded into scRepertoire, using <code><a href="../reference/clonalOverlap.html">clonalOverlap()</a></code> can assist
in the visualization.</p>
<p>The underlying <code><a href="../reference/clonalOverlap.html">clonalOverlap()</a></code> calculation varies by the
<strong>method</strong> parameter, more information on the exact
calculations are available in the manual.</p>
<p><strong>method</strong></p>
<ul>
<li>“overlap” - overlap coefficient</li>
<li>“morisita” - Morisita’s overlap index</li>
<li>“jaccard” - Jaccard index</li>
<li>“cosine” - cosine similarity</li>
<li>“raw” - exact number of overlapping clones</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalOverlap.html">clonalOverlap</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>              cloneCall <span class="op">=</span> <span class="st">"strict"</span>, </span>
<span>              method <span class="op">=</span> <span class="st">"morisita"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalOverlap.html">clonalOverlap</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>              cloneCall <span class="op">=</span> <span class="st">"strict"</span>, </span>
<span>              method <span class="op">=</span> <span class="st">"raw"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Clonal_Diversity_files/figure-html/unnamed-chunk-7-2.png" width="700"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
