<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Combining Contigs into Clones • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Combining Contigs into Clones">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Repertoire Analysis</li>
    <li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Working with Single-Cell Objects</li>
    <li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Deep Learning Integrations</li>
    <li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other Workflow Integrations</li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
    <li>
      <a href="../articles/immcantation.html">Combining Immcantation and scRepertoire</a>
    </li>
    <li>
      <a href="../articles/tcrpheno.html">TCRpheno applied to T cell fate</a>
    </li>
    <li>
      <a href="../articles/immunarch.html">Using scRepertoire with immunarch</a>
    </li>
    <li>
      <a href="../articles/dandelionR.html">VDJ Trajectory Analysis with dandelionR</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BorchLab/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Combining Contigs into Clones</h1>
            
            <h4 data-toc-skip class="date">Compiled: October 17,
2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BorchLab/scRepertoire/blob/HEAD/vignettes/articles/Combining_Contigs.Rmd" class="external-link"><code>vignettes/articles/Combining_Contigs.Rmd</code></a></small>
      <div class="hidden name"><code>Combining_Contigs.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<p>There are varying definitions of clones in the literature. For the
purposes of <code>scRepertoire</code>, we define a clone as cells with
shared/trackable complementarity-determining region 3 (CDR3) sequences.
Within this definition, one might use amino acid (<code>aa</code>)
sequences of one or both chains to define a clone. Alternatively, we
could use nucleotide (<code>nt</code>) or the V(D)JC genes
(<code>genes</code>) to define a clone. The latter, genes, would be a
more permissive definition of “clones,” as multiple amino acid or
nucleotide sequences can result from the same gene combination. Another
option to define a clone is the use of the V(D)JC and nucleotide
sequence (<code>strict</code>). <code>scRepertoire</code> allows for the
use of all these definitions of clones and enables users to select both
or individual chains to examine.</p>
<p>The first step in getting clones is to use the single-cell barcodes
to organize cells into paired sequences. This is accomplished using
<code><a href="../reference/combineTCR.html">combineTCR()</a></code> and <code><a href="../reference/combineBCR.html">combineBCR()</a></code>.</p>
<div class="section level2">
<h2 id="combinetcr">combineTCR<a class="anchor" aria-label="anchor" href="#combinetcr"></a>
</h2>
<p>The <code><a href="../reference/combineTCR.html">combineTCR()</a></code> function processes a list of TCR
sequencing results, consolidating them to the level of individual cell
barcodes. It handles potential issues with repeated barcodes by adding
prefixes from <code>samples</code> and <code>ID</code> parameters. The
output includes combined reads into clone calls by nucleotide sequence
(<code>CTnt</code>), amino acid sequence (<code>CTaa</code>), VDJC gene
sequence (<code>CTgene</code>), or a combination of nucleotide and gene
sequence (<code>CTstrict</code>).</p>
<p>Key Parameter(s) for <code><a href="../reference/combineTCR.html">combineTCR()</a></code></p>
<ul>
<li>
<code>input.data</code>: A list of filtered contig annotations
(e.g., filtered_contig_annotations.csv from 10x Cell Ranger) or outputs
from <code><a href="../reference/loadContigs.html">loadContigs()</a></code>.</li>
<li>
<code>samples</code>: Labels for your samples (recommended).</li>
<li>
<code>ID</code>: Additional sample labels (optional).</li>
<li>
<code>removeNA</code>: If <code>TRUE</code>, removes any cell
barcode with an NA value in at least one chain (default is
<code>FALSE</code>).</li>
<li>
<code>removeMulti</code>: If <code>TRUE</code>, removes any cell
barcode with more than two immune receptor chains (default is
<code>FALSE</code>).</li>
<li>
<code>filterMulti</code>: If <code>TRUE</code>, isolates the top two
expressed chains in cell barcodes with multiple chains (default is
<code>FALSE</code>).</li>
<li>
<code>filterNonproductive</code>: If <code>TRUE</code>, removes
non-productive chains if the variable exists in the contig data (default
is <code>TRUE</code>).</li>
</ul>
<p>To combine TCR contigs from <code>contig_list</code> and apply sample
prefixes:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">combined.TCR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineTCR.html">combineTCR</a></span><span class="op">(</span><span class="va">contig_list</span>, </span>
<span>                           samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17B"</span>, <span class="st">"P17L"</span>, <span class="st">"P18B"</span>, <span class="st">"P18L"</span>, </span>
<span>                                            <span class="st">"P19B"</span>,<span class="st">"P19L"</span>, <span class="st">"P20B"</span>, <span class="st">"P20L"</span><span class="op">)</span>,</span>
<span>                           removeNA <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                           removeMulti <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                           filterMulti <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">combined.TCR</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                    barcode sample                     TCR1           cdr3_aa1</span></span>
<span><span class="co">## 1  P17B_AAACCTGAGTACGACG-1   P17B       TRAV25.TRAJ20.TRAC        CGCSNDYKLSF</span></span>
<span><span class="co">## 3  P17B_AAACCTGCAACACGCC-1   P17B TRAV38-2/DV8.TRAJ52.TRAC CAYRSAQAGGTSYGKLTF</span></span>
<span><span class="co">## 5  P17B_AAACCTGCAGGCGATA-1   P17B      TRAV12-1.TRAJ9.TRAC     CVVSDNTGGFKTIF</span></span>
<span><span class="co">## 7  P17B_AAACCTGCATGAGCGA-1   P17B      TRAV12-1.TRAJ9.TRAC     CVVSDNTGGFKTIF</span></span>
<span><span class="co">## 9  P17B_AAACGGGAGAGCCCAA-1   P17B        TRAV20.TRAJ8.TRAC      CAVRGEGFQKLVF</span></span>
<span><span class="co">## 10 P17B_AAACGGGAGCGTTTAC-1   P17B      TRAV12-1.TRAJ9.TRAC     CVVSDNTGGFKTIF</span></span>
<span><span class="co">##                                                  cdr3_nt1</span></span>
<span><span class="co">## 1                       TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT</span></span>
<span><span class="co">## 3  TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT</span></span>
<span><span class="co">## 5              TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT</span></span>
<span><span class="co">## 7              TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT</span></span>
<span><span class="co">## 9                 TGTGCTGTGCGAGGAGAAGGCTTTCAGAAACTTGTATTT</span></span>
<span><span class="co">## 10             TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT</span></span>
<span><span class="co">##                           TCR2          cdr3_aa2</span></span>
<span><span class="co">## 1   TRBV5-1.None.TRBJ2-7.TRBC2    CASSLTDRTYEQYF</span></span>
<span><span class="co">## 3  TRBV10-3.None.TRBJ2-2.TRBC2     CAISEQGKGELFF</span></span>
<span><span class="co">## 5     TRBV9.None.TRBJ2-2.TRBC2 CASSVRRERANTGELFF</span></span>
<span><span class="co">## 7     TRBV9.None.TRBJ2-2.TRBC2 CASSVRRERANTGELFF</span></span>
<span><span class="co">## 9                         &lt;NA&gt;              &lt;NA&gt;</span></span>
<span><span class="co">## 10    TRBV9.None.TRBJ2-2.TRBC2 CASSVRRERANTGELFF</span></span>
<span><span class="co">##                                               cdr3_nt2</span></span>
<span><span class="co">## 1           TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC</span></span>
<span><span class="co">## 3              TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 5  TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 7  TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 9                                                 &lt;NA&gt;</span></span>
<span><span class="co">## 10 TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">##                                                  CTgene</span></span>
<span><span class="co">## 1         TRAV25.TRAJ20.TRAC_TRBV5-1.None.TRBJ2-7.TRBC2</span></span>
<span><span class="co">## 3  TRAV38-2/DV8.TRAJ52.TRAC_TRBV10-3.None.TRBJ2-2.TRBC2</span></span>
<span><span class="co">## 5          TRAV12-1.TRAJ9.TRAC_TRBV9.None.TRBJ2-2.TRBC2</span></span>
<span><span class="co">## 7          TRAV12-1.TRAJ9.TRAC_TRBV9.None.TRBJ2-2.TRBC2</span></span>
<span><span class="co">## 9                                  TRAV20.TRAJ8.TRAC_NA</span></span>
<span><span class="co">## 10         TRAV12-1.TRAJ9.TRAC_TRBV9.None.TRBJ2-2.TRBC2</span></span>
<span><span class="co">##                                                                                              CTnt</span></span>
<span><span class="co">## 1                    TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT_TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC</span></span>
<span><span class="co">## 3  TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT_TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 5  TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 7  TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 9                                                      TGTGCTGTGCGAGGAGAAGGCTTTCAGAAACTTGTATTT_NA</span></span>
<span><span class="co">## 10 TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">##                                CTaa</span></span>
<span><span class="co">## 1        CGCSNDYKLSF_CASSLTDRTYEQYF</span></span>
<span><span class="co">## 3  CAYRSAQAGGTSYGKLTF_CAISEQGKGELFF</span></span>
<span><span class="co">## 5  CVVSDNTGGFKTIF_CASSVRRERANTGELFF</span></span>
<span><span class="co">## 7  CVVSDNTGGFKTIF_CASSVRRERANTGELFF</span></span>
<span><span class="co">## 9                  CAVRGEGFQKLVF_NA</span></span>
<span><span class="co">## 10 CVVSDNTGGFKTIF_CASSVRRERANTGELFF</span></span>
<span><span class="co">##                                                                                                                                               CTstrict</span></span>
<span><span class="co">## 1                           TRAV25.TRAJ20.TRAC;TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT_TRBV5-1.None.TRBJ2-7.TRBC2;TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC</span></span>
<span><span class="co">## 3  TRAV38-2/DV8.TRAJ52.TRAC;TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT_TRBV10-3.None.TRBJ2-2.TRBC2;TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 5          TRAV12-1.TRAJ9.TRAC;TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TRBV9.None.TRBJ2-2.TRBC2;TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 7          TRAV12-1.TRAJ9.TRAC;TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TRBV9.None.TRBJ2-2.TRBC2;TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 9                                                                                      TRAV20.TRAJ8.TRAC;TGTGCTGTGCGAGGAGAAGGCTTTCAGAAACTTGTATTT_NA;NA</span></span>
<span><span class="co">## 10         TRAV12-1.TRAJ9.TRAC;TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TRBV9.None.TRBJ2-2.TRBC2;TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span></code></pre>
<p><code><a href="../reference/combineTCR.html">combineTCR()</a></code> is the essential first step for organizing
raw TCR sequencing data into a structured format for
<code>scRepertoire</code> analyses. It allows for flexible handling of
single and paired chains, barcode disambiguation, and initial filtering,
producing a list of data frames where each row represents a single cell
and its associated TCR clonotypes.</p>
</div>
<div class="section level2">
<h2 id="combinebcr">combineBCR<a class="anchor" aria-label="anchor" href="#combinebcr"></a>
</h2>
<p>The <code><a href="../reference/combineBCR.html">combineBCR()</a></code> function is the primary tool for
processing raw B cell receptor contig data into a format ready for
analysis. It is analogous to <code><a href="../reference/combineTCR.html">combineTCR()</a></code> but includes
specialized logic for handling the complexities of BCRs, such as somatic
hypermutation. The function consolidates contigs into a single data
frame per sample, with each row representing a unique cell.</p>
<p>By default <code>(call.related.clones = TRUE)</code>,
<code><a href="../reference/combineBCR.html">combineBCR()</a></code> groups B cells into clones based on the
similarity of their CDR3 sequences.</p>
<div class="section level3">
<h3 id="how-combinebcr-groups-related-clones">How <code>combineBCR()</code> Groups Related Clones<a class="anchor" aria-label="anchor" href="#how-combinebcr-groups-related-clones"></a>
</h3>
<ul>
<li>Internally calling <code><a href="../reference/clonalCluster.html">clonalCluster()</a></code> to build a network
of related sequences.</li>
<li>Using the <code>threshold</code> parameter to define connections.
The <code>threshold</code> is a normalized Levenshtein distance, where a
value closer to 1.0 requires higher sequence similarity. The default of
0.85 is a good starting point.</li>
<li>Assigning a cluster-based ID to the CTstrict column.</li>
</ul>
<p>Additionally, the <code>group.by</code> argument allows you to
constrain the clustering analysis to only occur within distinct
categories in your metadata. For example, using
<code>group.by = "sample"</code> ensures that sequences from different
samples are never compared or clustered together, even if they are
identical.</p>
<p>Key Parameter(s) for <code><a href="../reference/combineBCR.html">combineBCR()</a></code></p>
<ul>
<li>
<code>call.related.clones</code>: If <code>TRUE</code> (default),
uses <code><a href="../reference/clonalCluster.html">clonalCluster()</a></code> to identify related clones based on
sequence similarity.</li>
<li>
<code>group.by</code>: The column header used to group clones for
clustering (if <code>NULL</code>, clusters will be calculated across all
samples).</li>
<li>
<code>threshold</code>: The similarity threshold for
<code><a href="../reference/clonalCluster.html">clonalCluster()</a></code> (default: 0.85).</li>
<li>
<code>chain</code>: The chain to use for clustering when
call.related.clones = TRUE (default: <code>both</code>).</li>
<li>
<code>sequence</code>: The sequence type (<code>nt</code> or
<code>aa</code>) for clustering (default: <code>nt</code>).</li>
<li>use.V, use.J: If <code>TRUE</code>, sequences must share the same
V/J gene to be clustered (default: <code>TRUE</code>)</li>
<li>
<code>cluster.method</code>: The clustering algorithm to apply to
the edit-distance network (default: <code>components</code>).</li>
</ul>
<p>First, load the example BCR contig data:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load example BCR contig data</span></span>
<span><span class="va">BCR.contigs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"https://www.borch.dev/uploads/contigs/b_contigs.csv"</span><span class="op">)</span></span></code></pre></div>
<p>Then, combine BCR contigs using the default similarity
clustering:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Combine using the default similarity clustering</span></span>
<span><span class="va">combined.BCR.clustered</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineBCR.html">combineBCR</a></span><span class="op">(</span><span class="va">BCR.contigs</span>, </span>
<span>                                     samples <span class="op">=</span> <span class="st">"Patient1"</span>, </span>
<span>                                     threshold <span class="op">=</span> <span class="fl">0.85</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The CTstrict column contains cluster IDs (e.g., "cluster.1")</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">combined.BCR.clustered</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"barcode"</span>, <span class="st">"CTstrict"</span>, <span class="st">"IGH"</span>, <span class="st">"cdr3_aa1"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                       barcode   CTstrict                            IGH</span></span>
<span><span class="co">## 1 Patient1_AAACCTGAGGGCACTA-1 cluster.78                           &lt;NA&gt;</span></span>
<span><span class="co">## 2 Patient1_AAACCTGAGTACGTTC-1 cluster.16   IGHV3-64.IGHD6-6.IGHJ4.IGHG1</span></span>
<span><span class="co">## 3 Patient1_AAACCTGAGTCCGGTC-1       &lt;NA&gt;                           &lt;NA&gt;</span></span>
<span><span class="co">## 4 Patient1_AAACCTGCACCAGGTC-1       &lt;NA&gt; IGHV1-69-2.IGHD3-10.IGHJ6.IGHM</span></span>
<span><span class="co">## 5 Patient1_AAACCTGGTAAATGAC-1 cluster.16  IGHV1-46.IGHD6-13.IGHJ5.IGHA2</span></span>
<span><span class="co">## 6 Patient1_AAACGGGTCACCTCGT-1 cluster.44                           &lt;NA&gt;</span></span>
<span><span class="co">##                cdr3_aa1</span></span>
<span><span class="co">## 1                  &lt;NA&gt;</span></span>
<span><span class="co">## 2      CAKSYSRDLPRYFGSW</span></span>
<span><span class="co">## 3                  &lt;NA&gt;</span></span>
<span><span class="co">## 4 CVRDGAGHYGSGIGYYGMDVW</span></span>
<span><span class="co">## 5    CSRGRLPMPAAGSGLVSW</span></span>
<span><span class="co">## 6                  &lt;NA&gt;</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="filtering-and-cleaning-data">Filtering and Cleaning Data<a class="anchor" aria-label="anchor" href="#filtering-and-cleaning-data"></a>
</h3>
<p><code><a href="../reference/combineBCR.html">combineBCR()</a></code> includes several arguments to filter and
clean the contig data during processing. *
<code>filterNonproductive = TRUE</code> (Default): Removes any contigs
that are not classified as productive, ensuring that only functional
receptor chains are included in the analysis. *
<code>filterMulti = TRUE</code> (Default): For cells with more than one
heavy or light chain detected, this automatically selects the chain with
the highest UMI count (read abundance) and discards the others. This
helps resolve cellular multiplets or technical artifacts.</p>
<p>Here is an example applying these filters (though they are on by
default):</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cleaned.BCR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineBCR.html">combineBCR</a></span><span class="op">(</span><span class="va">BCR.contigs</span>,</span>
<span>                          samples <span class="op">=</span> <span class="st">"Patient1"</span>,</span>
<span>                          filterNonproductive <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                          filterMulti <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cleaned.BCR</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                       barcode   sample                            IGH</span></span>
<span><span class="co">## 1 Patient1_AAACCTGAGGGCACTA-1 Patient1                           &lt;NA&gt;</span></span>
<span><span class="co">## 2 Patient1_AAACCTGAGTACGTTC-1 Patient1   IGHV3-64.IGHD6-6.IGHJ4.IGHG1</span></span>
<span><span class="co">## 3 Patient1_AAACCTGAGTCCGGTC-1 Patient1                           &lt;NA&gt;</span></span>
<span><span class="co">## 4 Patient1_AAACCTGCACCAGGTC-1 Patient1 IGHV1-69-2.IGHD3-10.IGHJ6.IGHM</span></span>
<span><span class="co">## 5 Patient1_AAACCTGGTAAATGAC-1 Patient1  IGHV1-46.IGHD6-13.IGHJ5.IGHA2</span></span>
<span><span class="co">## 6 Patient1_AAACGGGTCACCTCGT-1 Patient1                           &lt;NA&gt;</span></span>
<span><span class="co">##                cdr3_aa1</span></span>
<span><span class="co">## 1                  &lt;NA&gt;</span></span>
<span><span class="co">## 2      CAKSYSRDLPRYFGSW</span></span>
<span><span class="co">## 3                  &lt;NA&gt;</span></span>
<span><span class="co">## 4 CVRDGAGHYGSGIGYYGMDVW</span></span>
<span><span class="co">## 5    CSRGRLPMPAAGSGLVSW</span></span>
<span><span class="co">## 6                  &lt;NA&gt;</span></span>
<span><span class="co">##                                                          cdr3_nt1</span></span>
<span><span class="co">## 1                                                            &lt;NA&gt;</span></span>
<span><span class="co">## 2                TGTGCGAAATCGTATAGCAGAGACCTGCCGCGGTACTTTGGCTCCTGG</span></span>
<span><span class="co">## 3                                                            &lt;NA&gt;</span></span>
<span><span class="co">## 4 TGTGTGAGAGATGGGGCGGGTCACTATGGTTCGGGGATAGGCTACTACGGTATGGACGTCTGG</span></span>
<span><span class="co">## 5          TGTTCGAGGGGTCGTCTCCCGATGCCAGCAGCTGGCAGTGGTTTGGTCTCCTGG</span></span>
<span><span class="co">## 6                                                            &lt;NA&gt;</span></span>
<span><span class="co">##                   IGLC       cdr3_aa2</span></span>
<span><span class="co">## 1 IGLV1-51.IGLJ3.IGLC2  CGTWDSSLSAGVF</span></span>
<span><span class="co">## 2  IGKV3-15.IGKJ4.IGKC    CQQYSNWPLTF</span></span>
<span><span class="co">## 3  IGKV3-20.IGKJ4.IGKC     CQQYGDSLPF</span></span>
<span><span class="co">## 4 IGLV2-14.IGLJ2.IGLC2 CSSYISTSTLEVLF</span></span>
<span><span class="co">## 5  IGKV3-15.IGKJ2.IGKC   CQHYHNWPPYTF</span></span>
<span><span class="co">## 6  IGKV3-11.IGKJ1.IGKC    CQQRSNWPLTF</span></span>
<span><span class="co">##                                     cdr3_nt2</span></span>
<span><span class="co">## 1    TGCGGAACATGGGATAGCAGCCTGAGTGCTGGCGTGTTC</span></span>
<span><span class="co">## 2          TGTCAGCAGTATAGTAACTGGCCGCTCACTTTC</span></span>
<span><span class="co">## 3             TGTCAACAGTATGGTGACTCTCTCCCTTTC</span></span>
<span><span class="co">## 4 TGCAGTTCATATATAAGTACCAGCACTCTCGAGGTCCTATTC</span></span>
<span><span class="co">## 5       TGTCAGCACTATCATAACTGGCCTCCGTACACTTTT</span></span>
<span><span class="co">## 6          TGTCAGCAGCGTAGCAACTGGCCTCTGACGTTC</span></span>
<span><span class="co">##                                                CTgene</span></span>
<span><span class="co">## 1                             NA_IGLV1-51.IGLJ3.IGLC2</span></span>
<span><span class="co">## 2    IGHV3-64.IGHD6-6.IGHJ4.IGHG1_IGKV3-15.IGKJ4.IGKC</span></span>
<span><span class="co">## 3                              NA_IGKV3-20.IGKJ4.IGKC</span></span>
<span><span class="co">## 4 IGHV1-69-2.IGHD3-10.IGHJ6.IGHM_IGLV2-14.IGLJ2.IGLC2</span></span>
<span><span class="co">## 5   IGHV1-46.IGHD6-13.IGHJ5.IGHA2_IGKV3-15.IGKJ2.IGKC</span></span>
<span><span class="co">## 6                              NA_IGKV3-11.IGKJ1.IGKC</span></span>
<span><span class="co">##                                                                                                         CTnt</span></span>
<span><span class="co">## 1                                                                 NA_TGCGGAACATGGGATAGCAGCCTGAGTGCTGGCGTGTTC</span></span>
<span><span class="co">## 2                         TGTGCGAAATCGTATAGCAGAGACCTGCCGCGGTACTTTGGCTCCTGG_TGTCAGCAGTATAGTAACTGGCCGCTCACTTTC</span></span>
<span><span class="co">## 3                                                                          NA_TGTCAACAGTATGGTGACTCTCTCCCTTTC</span></span>
<span><span class="co">## 4 TGTGTGAGAGATGGGGCGGGTCACTATGGTTCGGGGATAGGCTACTACGGTATGGACGTCTGG_TGCAGTTCATATATAAGTACCAGCACTCTCGAGGTCCTATTC</span></span>
<span><span class="co">## 5                TGTTCGAGGGGTCGTCTCCCGATGCCAGCAGCTGGCAGTGGTTTGGTCTCCTGG_TGTCAGCACTATCATAACTGGCCTCCGTACACTTTT</span></span>
<span><span class="co">## 6                                                                       NA_TGTCAGCAGCGTAGCAACTGGCCTCTGACGTTC</span></span>
<span><span class="co">##                                   CTaa   CTstrict</span></span>
<span><span class="co">## 1                     NA_CGTWDSSLSAGVF cluster.78</span></span>
<span><span class="co">## 2         CAKSYSRDLPRYFGSW_CQQYSNWPLTF cluster.16</span></span>
<span><span class="co">## 3                        NA_CQQYGDSLPF       &lt;NA&gt;</span></span>
<span><span class="co">## 4 CVRDGAGHYGSGIGYYGMDVW_CSSYISTSTLEVLF       &lt;NA&gt;</span></span>
<span><span class="co">## 5      CSRGRLPMPAAGSGLVSW_CQHYHNWPPYTF cluster.16</span></span>
<span><span class="co">## 6                       NA_CQQRSNWPLTF cluster.44</span></span></code></pre>
<p><code><a href="../reference/combineBCR.html">combineBCR()</a></code> is designed for processing B cell
repertoire data, going beyond simple contig aggregation to incorporate
advanced clustering based on CDR3 sequence similarity. This enables the
identification of clonally related B cells, crucial for studying B cell
development, affinity maturation, and humoral immune responses. Its
filtering options further ensure the quality and interpretability of the
processed data.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
