<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Combining Contigs into Clones • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Combining Contigs into Clones">
<meta property="og:description" content="scRepertoire">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ncborcherding/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Combining Contigs into Clones</h1>
            
            <h4 data-toc-skip class="date">Compiled: May 04, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ncborcherding/scRepertoire/blob/HEAD/vignettes/articles/Combining_Contigs.Rmd" class="external-link"><code>vignettes/articles/Combining_Contigs.Rmd</code></a></small>
      <div class="hidden name"><code>Combining_Contigs.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<p>There are varying definitions of clones or clones in the literature.
For the purposes of scRepertoire, we will use <strong>clone</strong> and
define this as the cells with shared/trackable
complementarity-determining region 3 (CDR3) sequences. Within this
definition, one might use amino acid (<strong>aa</strong>) sequences of
one or both chains to define a clone. Alternatively, we could use
nucleotide (<strong>nt</strong>) or the V(D)JC genes
(<strong>genes</strong>) to define a clone. The latter
<strong>genes</strong> would be a more permissive definition of
<em>“clones”</em>, as multiple amino acid or nucleotide sequences can
result from the same gene combination. Another option to define clone is
the use of the V(D)JC and nucleotide sequence (<strong>strict</strong>).
scRepertoire allows for the use of all these definitions of clones and
allows for users to select both or individual chains to examine.</p>
<p>The first step in getting clones is to use the single-cell barcodes
to organize cells into paired sequences. This is accomplished using
<code><a href="../reference/combineTCR.html">combineTCR()</a></code> and <code><a href="../reference/combineBCR.html">combineBCR()</a></code>.</p>
<div class="section level2">
<h2 id="combinetcr">combineTCR<a class="anchor" aria-label="anchor" href="#combinetcr"></a>
</h2>
<p><strong>input.data</strong></p>
<ul>
<li>List of <em>filtered_contig_annotations.csv</em> data frames from
the 10x Cell Ranger.<br>
</li>
<li>List of data processed using <code><a href="../reference/loadContigs.html">loadContigs()</a></code>.</li>
</ul>
<p><strong>samples</strong> and <strong>ID</strong></p>
<ul>
<li>Grouping variables for downstream analysis and will be added as
prefixes to prevent issues with duplicate barcodes
<strong>(optional)</strong>.</li>
</ul>
<p><strong>removeNA</strong></p>
<ul>
<li>TRUE - Filter to remove any cell barcode with an NA value in at
least one of the chains.<br>
</li>
<li>FALSE - Include and incorporate cells with 1 NA value
<strong>(default)</strong>.</li>
</ul>
<p><strong>removeMulti</strong></p>
<ul>
<li>TRUE - Filter to remove any cell barcode with more than 2 immune
receptor chains.<br>
</li>
<li>FALSE - Include and incorporate cells with &gt; 2 chains
<strong>(default)</strong>.</li>
</ul>
<p><strong>filterMulti</strong></p>
<ul>
<li>TRUE - Isolate the top 2 expressed chains in cell barcodes with
multiple chains.<br>
</li>
<li>FALSE - Include and incorporate cells with &gt; 2 chains
<strong>(default)</strong>.</li>
</ul>
<p>The output of <code><a href="../reference/combineTCR.html">combineTCR()</a></code> will be a list of contig data
frames that will be reduced to the reads associated with a single cell
barcode. It will also combine the multiple reads into clone calls by
either the nucleotide sequence (<strong>CTnt</strong>), amino acid
sequence (<strong>CTaa</strong>), the VDJC gene sequence
(<strong>CTgene</strong>), or the combination of the nucleotide and gene
sequence (<strong>CTstrict</strong>).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">combined.TCR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineTCR.html">combineTCR</a></span><span class="op">(</span><span class="va">contig_list</span>, </span>
<span>                           samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17B"</span>, <span class="st">"P17L"</span>, <span class="st">"P18B"</span>, <span class="st">"P18L"</span>, </span>
<span>                                            <span class="st">"P19B"</span>,<span class="st">"P19L"</span>, <span class="st">"P20B"</span>, <span class="st">"P20L"</span><span class="op">)</span>,</span>
<span>                           removeNA <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                           removeMulti <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                           filterMulti <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">combined.TCR</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                    barcode sample                     TCR1           cdr3_aa1</span></span>
<span><span class="co">## 1  P17B_AAACCTGAGTACGACG-1   P17B       TRAV25.TRAJ20.TRAC        CGCSNDYKLSF</span></span>
<span><span class="co">## 3  P17B_AAACCTGCAACACGCC-1   P17B TRAV38-2/DV8.TRAJ52.TRAC CAYRSAQAGGTSYGKLTF</span></span>
<span><span class="co">## 5  P17B_AAACCTGCAGGCGATA-1   P17B      TRAV12-1.TRAJ9.TRAC     CVVSDNTGGFKTIF</span></span>
<span><span class="co">## 7  P17B_AAACCTGCATGAGCGA-1   P17B      TRAV12-1.TRAJ9.TRAC     CVVSDNTGGFKTIF</span></span>
<span><span class="co">## 9  P17B_AAACGGGAGAGCCCAA-1   P17B        TRAV20.TRAJ8.TRAC      CAVRGEGFQKLVF</span></span>
<span><span class="co">## 10 P17B_AAACGGGAGCGTTTAC-1   P17B      TRAV12-1.TRAJ9.TRAC     CVVSDNTGGFKTIF</span></span>
<span><span class="co">##                                                  cdr3_nt1</span></span>
<span><span class="co">## 1                       TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT</span></span>
<span><span class="co">## 3  TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT</span></span>
<span><span class="co">## 5              TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT</span></span>
<span><span class="co">## 7              TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT</span></span>
<span><span class="co">## 9                 TGTGCTGTGCGAGGAGAAGGCTTTCAGAAACTTGTATTT</span></span>
<span><span class="co">## 10             TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT</span></span>
<span><span class="co">##                           TCR2          cdr3_aa2</span></span>
<span><span class="co">## 1   TRBV5-1.None.TRBJ2-7.TRBC2    CASSLTDRTYEQYF</span></span>
<span><span class="co">## 3  TRBV10-3.None.TRBJ2-2.TRBC2     CAISEQGKGELFF</span></span>
<span><span class="co">## 5     TRBV9.None.TRBJ2-2.TRBC2 CASSVRRERANTGELFF</span></span>
<span><span class="co">## 7     TRBV9.None.TRBJ2-2.TRBC2 CASSVRRERANTGELFF</span></span>
<span><span class="co">## 9                         &lt;NA&gt;              &lt;NA&gt;</span></span>
<span><span class="co">## 10    TRBV9.None.TRBJ2-2.TRBC2 CASSVRRERANTGELFF</span></span>
<span><span class="co">##                                               cdr3_nt2</span></span>
<span><span class="co">## 1           TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC</span></span>
<span><span class="co">## 3              TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 5  TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 7  TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 9                                                 &lt;NA&gt;</span></span>
<span><span class="co">## 10 TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">##                                                  CTgene</span></span>
<span><span class="co">## 1         TRAV25.TRAJ20.TRAC_TRBV5-1.None.TRBJ2-7.TRBC2</span></span>
<span><span class="co">## 3  TRAV38-2/DV8.TRAJ52.TRAC_TRBV10-3.None.TRBJ2-2.TRBC2</span></span>
<span><span class="co">## 5          TRAV12-1.TRAJ9.TRAC_TRBV9.None.TRBJ2-2.TRBC2</span></span>
<span><span class="co">## 7          TRAV12-1.TRAJ9.TRAC_TRBV9.None.TRBJ2-2.TRBC2</span></span>
<span><span class="co">## 9                                  TRAV20.TRAJ8.TRAC_NA</span></span>
<span><span class="co">## 10         TRAV12-1.TRAJ9.TRAC_TRBV9.None.TRBJ2-2.TRBC2</span></span>
<span><span class="co">##                                                                                              CTnt</span></span>
<span><span class="co">## 1                    TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT_TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC</span></span>
<span><span class="co">## 3  TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT_TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 5  TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 7  TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 9                                                      TGTGCTGTGCGAGGAGAAGGCTTTCAGAAACTTGTATTT_NA</span></span>
<span><span class="co">## 10 TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">##                                CTaa</span></span>
<span><span class="co">## 1        CGCSNDYKLSF_CASSLTDRTYEQYF</span></span>
<span><span class="co">## 3  CAYRSAQAGGTSYGKLTF_CAISEQGKGELFF</span></span>
<span><span class="co">## 5  CVVSDNTGGFKTIF_CASSVRRERANTGELFF</span></span>
<span><span class="co">## 7  CVVSDNTGGFKTIF_CASSVRRERANTGELFF</span></span>
<span><span class="co">## 9                  CAVRGEGFQKLVF_NA</span></span>
<span><span class="co">## 10 CVVSDNTGGFKTIF_CASSVRRERANTGELFF</span></span>
<span><span class="co">##                                                                                                                                               CTstrict</span></span>
<span><span class="co">## 1                           TRAV25.TRAJ20.TRAC;TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT_TRBV5-1.None.TRBJ2-7.TRBC2;TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC</span></span>
<span><span class="co">## 3  TRAV38-2/DV8.TRAJ52.TRAC;TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT_TRBV10-3.None.TRBJ2-2.TRBC2;TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 5          TRAV12-1.TRAJ9.TRAC;TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TRBV9.None.TRBJ2-2.TRBC2;TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 7          TRAV12-1.TRAJ9.TRAC;TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TRBV9.None.TRBJ2-2.TRBC2;TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 9                                                                                      TRAV20.TRAJ8.TRAC;TGTGCTGTGCGAGGAGAAGGCTTTCAGAAACTTGTATTT_NA;NA</span></span>
<span><span class="co">## 10         TRAV12-1.TRAJ9.TRAC;TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TRBV9.None.TRBJ2-2.TRBC2;TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="combinebcr">combineBCR<a class="anchor" aria-label="anchor" href="#combinebcr"></a>
</h2>
<p><code><a href="../reference/combineBCR.html">combineBCR()</a></code> is analogous to <code><a href="../reference/combineTCR.html">combineTCR()</a></code>
with 2 major changes: <strong>1)</strong> Each barcode can only have a
maximum of 2 sequences, if greater exists, the 2 with the highest reads
are selected; <strong>2)</strong> The <strong>strict</strong> definition
of a clone is based on the normalized Levenshtein edit distance of CDR3
nucleotide sequences and V-gene usage. For more information on this
approach, please see the respective <a href="https://pubmed.ncbi.nlm.nih.gov/34161770/" class="external-link">citation</a>. This
definition allows for the grouping of BCRs derived from the same
progenitor that have undergone mutation as part of somatic hypermutation
and affinity maturation.</p>
<p><strong>threshold</strong><br>
The level of similarity in sequences to group together.
<strong>Default</strong> is 0.85.</p>
<p><span class="math display">\[
\text{threshold}(s, t) = 1-\frac{\text{Levenshtein}(s,
t)}{\frac{\text{length}(s) + \text{length}(t)}{2}}
\]</span></p>
<p><strong>call.related.clones</strong><br>
Calculate the normalized edit distance (<strong>TRUE</strong>) or skip
the calculation (<strong>FALSE</strong>). Skipping the edit distance
calculation may save time, especially in the context of large data sets,
but is not recommended.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">BCR.contigs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"https://www.borch.dev/uploads/contigs/b_contigs.csv"</span><span class="op">)</span></span>
<span><span class="va">combined.BCR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineBCR.html">combineBCR</a></span><span class="op">(</span><span class="va">BCR.contigs</span>, </span>
<span>                           samples <span class="op">=</span> <span class="st">"P1"</span>, </span>
<span>                           threshold <span class="op">=</span> <span class="fl">0.85</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">combined.BCR</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                 barcode sample                             IGH</span></span>
<span><span class="co">## 1 P1_CGAACATTCCCTTGTG-1     P1                            &lt;NA&gt;</span></span>
<span><span class="co">## 2 P1_CCTTCGAAGACGCTTT-1     P1                            &lt;NA&gt;</span></span>
<span><span class="co">## 3 P1_CAGCTAACAGCTTAAC-1     P1                            &lt;NA&gt;</span></span>
<span><span class="co">## 4 P1_CCCAATCAGAGGGCTT-1     P1 IGHV1-69-2.IGHD5-12.IGHJ6.IGHG1</span></span>
<span><span class="co">## 5 P1_TACAGTGGTAAGGGCT-1     P1   IGHV1-18.IGHD3-10.IGHJ5.IGHA1</span></span>
<span><span class="co">## 6 P1_GACAGAGGTTGGTAAA-1     P1   IGHV1-18.IGHD3-10.IGHJ5.IGHA1</span></span>
<span><span class="co">##                cdr3_aa1</span></span>
<span><span class="co">## 1                  &lt;NA&gt;</span></span>
<span><span class="co">## 2                  &lt;NA&gt;</span></span>
<span><span class="co">## 3                  &lt;NA&gt;</span></span>
<span><span class="co">## 4  CASGGLVAIRHYYYYGMDVW</span></span>
<span><span class="co">## 5 CARVTYHHGSGSLVGGWFDPW</span></span>
<span><span class="co">## 6 CARVTYHHGSGSLVGGWFDPW</span></span>
<span><span class="co">##                                                          cdr3_nt1</span></span>
<span><span class="co">## 1                                                            &lt;NA&gt;</span></span>
<span><span class="co">## 2                                                            &lt;NA&gt;</span></span>
<span><span class="co">## 3                                                            &lt;NA&gt;</span></span>
<span><span class="co">## 4    TGTGCGAGCGGGGGATTAGTGGCTATTAGGCACTACTACTACTACGGTATGGACGTCTGG</span></span>
<span><span class="co">## 5 TGTGCGAGAGTGACATATCACCATGGTTCGGGGAGCCTTGTCGGGGGCTGGTTCGACCCCTGG</span></span>
<span><span class="co">## 6 TGTGCGAGAGTGACATATCACCATGGTTCGGGGAGCCTTGTCGGGGGCTGGTTCGACCCCTGG</span></span>
<span><span class="co">##                   IGLC      cdr3_aa2                                cdr3_nt2</span></span>
<span><span class="co">## 1 IGLV2-14.IGLJ6.IGLC2  CNSKGAGGTAVF    TGCAACTCAAAAGGAGCCGGAGGCACTGCGGTTTTC</span></span>
<span><span class="co">## 2 IGLV2-14.IGLJ6.IGLC2  CNSKGAGGTAVF    TGCAACTCAAAAGGAGCCGGAGGCACTGCGGTTTTC</span></span>
<span><span class="co">## 3 IGLV2-14.IGLJ2.IGLC2 CNSYTTSGTLWVF TGCAACTCATACACAACCAGCGGCACTCTCTGGGTATTC</span></span>
<span><span class="co">## 4 IGLV2-14.IGLJ2.IGLC2  CNSYTSSSTLVF    TGCAACTCATATACAAGCAGCAGCACTCTGGTCTTC</span></span>
<span><span class="co">## 5 IGLV2-14.IGLJ1.IGLC1  CNSYTSFGTSVF    TGCAACTCATATACAAGCTTCGGCACCTCGGTCTTC</span></span>
<span><span class="co">## 6 IGLV2-14.IGLJ1.IGLC1  CNSYTSFGTSVF    TGCAACTCATATACAAGCTTCGGCACCTCGGTCTTC</span></span>
<span><span class="co">##                                                 CTgene</span></span>
<span><span class="co">## 1                              NA_IGLV2-14.IGLJ6.IGLC2</span></span>
<span><span class="co">## 2                              NA_IGLV2-14.IGLJ6.IGLC2</span></span>
<span><span class="co">## 3                              NA_IGLV2-14.IGLJ2.IGLC2</span></span>
<span><span class="co">## 4 IGHV1-69-2.IGHD5-12.IGHJ6.IGHG1_IGLV2-14.IGLJ2.IGLC2</span></span>
<span><span class="co">## 5   IGHV1-18.IGHD3-10.IGHJ5.IGHA1_IGLV2-14.IGLJ1.IGLC1</span></span>
<span><span class="co">## 6   IGHV1-18.IGHD3-10.IGHJ5.IGHA1_IGLV2-14.IGLJ1.IGLC1</span></span>
<span><span class="co">##                                                                                                   CTnt</span></span>
<span><span class="co">## 1                                                              NA_TGCAACTCAAAAGGAGCCGGAGGCACTGCGGTTTTC</span></span>
<span><span class="co">## 2                                                              NA_TGCAACTCAAAAGGAGCCGGAGGCACTGCGGTTTTC</span></span>
<span><span class="co">## 3                                                           NA_TGCAACTCATACACAACCAGCGGCACTCTCTGGGTATTC</span></span>
<span><span class="co">## 4    TGTGCGAGCGGGGGATTAGTGGCTATTAGGCACTACTACTACTACGGTATGGACGTCTGG_TGCAACTCATATACAAGCAGCAGCACTCTGGTCTTC</span></span>
<span><span class="co">## 5 TGTGCGAGAGTGACATATCACCATGGTTCGGGGAGCCTTGTCGGGGGCTGGTTCGACCCCTGG_TGCAACTCATATACAAGCTTCGGCACCTCGGTCTTC</span></span>
<span><span class="co">## 6 TGTGCGAGAGTGACATATCACCATGGTTCGGGGAGCCTTGTCGGGGGCTGGTTCGACCCCTGG_TGCAACTCATATACAAGCTTCGGCACCTCGGTCTTC</span></span>
<span><span class="co">##                                 CTaa</span></span>
<span><span class="co">## 1                    NA_CNSKGAGGTAVF</span></span>
<span><span class="co">## 2                    NA_CNSKGAGGTAVF</span></span>
<span><span class="co">## 3                   NA_CNSYTTSGTLWVF</span></span>
<span><span class="co">## 4  CASGGLVAIRHYYYYGMDVW_CNSYTSSSTLVF</span></span>
<span><span class="co">## 5 CARVTYHHGSGSLVGGWFDPW_CNSYTSFGTSVF</span></span>
<span><span class="co">## 6 CARVTYHHGSGSLVGGWFDPW_CNSYTSFGTSVF</span></span>
<span><span class="co">##                                      CTstrict</span></span>
<span><span class="co">## 1                     NA.NA_IGLC.549.IGLV2-14</span></span>
<span><span class="co">## 2                     NA.NA_IGLC.549.IGLV2-14</span></span>
<span><span class="co">## 3                     NA.NA_IGLC.535.IGLV2-14</span></span>
<span><span class="co">## 4 IGH.481.IGHV1-69-2_IGLC:Cluster.56.IGLV2-14</span></span>
<span><span class="co">## 5   IGH.781.IGHV1-18_IGLC:Cluster.56.IGLV2-14</span></span>
<span><span class="co">## 6   IGH.781.IGHV1-18_IGLC:Cluster.56.IGLV2-14</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
