<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>VDJ Trajectory Analysis with dandelionR • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="VDJ Trajectory Analysis with dandelionR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/dandelionR.html">VDJ Trajectory Analysis with dandelionR</a>
    </li>
    <li>
      <a href="../articles/immunarch.html">Using scRepertoire with immunarch</a>
    </li>
    <li>
      <a href="../articles/tcrpheno.html">TCRpheno applied to T cell fate</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BorchLab/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>VDJ Trajectory Analysis with dandelionR</h1>
            
            <h4 data-toc-skip class="date">Compiled: July 31, 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BorchLab/scRepertoire/blob/HEAD/vignettes/articles/dandelionR.Rmd" class="external-link"><code>vignettes/articles/dandelionR.Rmd</code></a></small>
      <div class="hidden name"><code>dandelionR.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p><code>dandelionR</code> is an R package for performing single-cell
immune repertoire trajectory analysis, based on the original python
implementation in <a href="https://www.github.com/zktuong/dandelion" class="external-link">dandelion</a>.</p>
<p>It provides all the necessary tools to interface with <a href="https://github.com/ncborcherding/scRepertoire" class="external-link">scRepertoire</a>
and a custom implementation of absorbing markov chain for pseudotime
inference, inspired based on the <a href="https://github.com/dpeerlab/Palantir" class="external-link">palantir</a> python
package.</p>
<div class="section level3">
<h3 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h3>
<p>If using <em>dandelionR</em>, please cite the <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC12270724/" class="external-link">article</a>:
Yu, J., et al., dandelionR: Single-cell immune repertoire trajectory
analysis in R. <em>Computational and Structural Biotechnology
Journal.</em> 2025.</p>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"dandelionR"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"tuonglab/dandelionR"</span>, dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h3>
<div class="section level4">
<h4 id="load-the-required-libraries">Load the required libraries<a class="anchor" aria-label="anchor" href="#load-the-required-libraries"></a>
</h4>
</div>
<div class="section level4">
<h4 id="load-the-demo-data">Load the demo data<a class="anchor" aria-label="anchor" href="#load-the-demo-data"></a>
</h4>
<p>Due to size limitations of the package, we have provided a very
trimmed down version of the demo data to ~2000 cells. The full dataset
can be found here accordingly: GEX - <a href="https://developmental.cellatlas.io/fetal-immune" class="external-link uri">https://developmental.cellatlas.io/fetal-immune</a>
(Lymphoid Cells) and VDJ - <a href="https://github.com/zktuong/dandelion-demo-files/tree/master/dandelion_manuscript/data/dandelion-remap" class="external-link uri">https://github.com/zktuong/dandelion-demo-files/tree/master/dandelion_manuscript/data/dandelion-remap</a></p>
<p>Check out the other vignette for an example dataset that starts from
the original <code>dandelion</code> output associated with the original
<a href="https://www.nature.com/articles/s41587-023-01734-7" class="external-link">manuscript</a>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">demo_sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">demo_airr</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="use-screpertoire-to-load-the-vdj-data">Use <code>scRepertoire</code> to load the VDJ data<a class="anchor" aria-label="anchor" href="#use-screpertoire-to-load-the-vdj-data"></a>
</h4>
<p>For the trajectory analysis work here, we are focusing on the main
productive TCR chains. Therefore we will flag
<code>filterMulti = TRUE</code>, which will keep the selection of the 2
corresponding chains with the highest expression for a single barcode.
For more details, refer to <code>scRepertoire</code>’s <a href="https://www.borch.dev/uploads/screpertoire/reference/combinetcr" class="external-link">documentation</a>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">contig.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadContigs.html">loadContigs</a></span><span class="op">(</span>input <span class="op">=</span> <span class="va">demo_airr</span>, </span>
<span>                           format <span class="op">=</span> <span class="st">"AIRR"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Format to `scRepertoire`'s requirements and some light filtering</span></span>
<span><span class="va">combined.TCR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineTCR.html">combineTCR</a></span><span class="op">(</span><span class="va">contig.list</span>,</span>
<span>    removeNA <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    removeMulti <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    filterMulti <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Merge VDJ and GEX data</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineExpression.html">combineExpression</a></span><span class="op">(</span><span class="va">combined.TCR</span>, <span class="va">demo_sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Subsetting Complete TCR Gene Sequences</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"TRAC"</span>, <span class="va">sce</span><span class="op">$</span><span class="va">CTgene</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"TRBC"</span>, <span class="va">sce</span><span class="op">$</span><span class="va">CTgene</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="initiate-dandelionr-workflow">Initiate <code>dandelionR</code> workflow<a class="anchor" aria-label="anchor" href="#initiate-dandelionr-workflow"></a>
</h4>
<p>Here, the data is ready to be used for the pseudobulk and trajectory
analysis workflow in <code>dandelionR</code>.</p>
<p>Because this is a alpha-beta TCR data, we will set the
<code>mode_option</code> to “abT”. This will append <code>abT</code> to
the relevant columns holding the VDJ gene information. If you are going
to try other types of VDJ data e.g. BCR, you should set
<code>mode_option</code> to “B” instead. And this argument should be
consistently set with the <code>vdjPseudobulk</code> function later.</p>
<p>Since the TCR data is already filtered for productive chains in
<code>combineTCR</code>, we will set
<code>already.productive = TRUE</code> and can keep
<code>allowed_chain_status</code> as <code>NULL</code>.</p>
<p>We will also subset the data to only include the main T-cell types:
CD8+T, CD4+T, ABT(ENTRY), DP(P)_T, DP(Q)_T.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">setupVdjPseudobulk</span><span class="op">(</span><span class="va">sce</span>,</span>
<span>    mode_option <span class="op">=</span> <span class="st">"abT"</span>,</span>
<span>    already.productive <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    subsetby <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span>,</span>
<span>    groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD8+T"</span>, <span class="st">"CD4+T"</span>, <span class="st">"ABT(ENTRY)"</span>, <span class="st">"DP(P)_T"</span>, <span class="st">"DP(Q)_T"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The main output of this function is a
<code>SingleCellExperiment</code> object with the relevant VDJ
information appended to the <code>colData</code>, particularly the
columns with the <code>_main</code> suffix
e.g. <code>v_call_abT_VJ_main</code>, <code>j_call_abT_VJ_main</code>
etc.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 10 columns</span></span>
<span><span class="co">##                                  n_counts   n_genes           file      mito</span></span>
<span><span class="co">##                                 &lt;numeric&gt; &lt;integer&gt;       &lt;factor&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## FCAImmP7851891-CCTACCATCGGACAAG      2947      1275 FCAImmP7851891 0.0105192</span></span>
<span><span class="co">## FCAImmP7851892-ACGGGCTCAGCATGAG      4969      1971 FCAImmP7851892 0.0245522</span></span>
<span><span class="co">## FCAImmP7803035-CCAGCGATCCGAAGAG      7230      1733 FCAImmP7803035 0.0302905</span></span>
<span><span class="co">## FCAImmP7528296-ATAAGAGTCAAAGACA      2504       901 FCAImmP7528296 0.0207668</span></span>
<span><span class="co">## FCAImmP7555860-AACTTTCTCAACGGGA      8689      2037 FCAImmP7555860 0.0357924</span></span>
<span><span class="co">## FCAImmP7292034-CGTCACTGTGGTCTCG      3111      1254 FCAImmP7292034 0.0228222</span></span>
<span><span class="co">##                                 doublet_scores predicted_doublets</span></span>
<span><span class="co">##                                      &lt;numeric&gt;           &lt;factor&gt;</span></span>
<span><span class="co">## FCAImmP7851891-CCTACCATCGGACAAG      0.0439224              False</span></span>
<span><span class="co">## FCAImmP7851892-ACGGGCTCAGCATGAG      0.0610687              False</span></span>
<span><span class="co">## FCAImmP7803035-CCAGCGATCCGAAGAG      0.0383747              False</span></span>
<span><span class="co">## FCAImmP7528296-ATAAGAGTCAAAGACA      0.0236220              False</span></span>
<span><span class="co">## FCAImmP7555860-AACTTTCTCAACGGGA      0.0738255              False</span></span>
<span><span class="co">## FCAImmP7292034-CGTCACTGTGGTCTCG      0.0222841              False</span></span>
<span><span class="co">##                                 old_annotation_uniform    organ  Sort_id</span></span>
<span><span class="co">##                                               &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;</span></span>
<span><span class="co">## FCAImmP7851891-CCTACCATCGGACAAG              SP T CELL       TH    TOT  </span></span>
<span><span class="co">## FCAImmP7851892-ACGGGCTCAGCATGAG              DP T CELL       TH    TOT  </span></span>
<span><span class="co">## FCAImmP7803035-CCAGCGATCCGAAGAG              SP T CELL       SK    CD45P</span></span>
<span><span class="co">## FCAImmP7528296-ATAAGAGTCAAAGACA              SP T CELL       SK    CD45P</span></span>
<span><span class="co">## FCAImmP7555860-AACTTTCTCAACGGGA              SP T CELL       TH    CD45P</span></span>
<span><span class="co">## FCAImmP7292034-CGTCACTGTGGTCTCG              SP T CELL       TH    TOT  </span></span>
<span><span class="co">##                                       age</span></span>
<span><span class="co">##                                 &lt;integer&gt;</span></span>
<span><span class="co">## FCAImmP7851891-CCTACCATCGGACAAG        11</span></span>
<span><span class="co">## FCAImmP7851892-ACGGGCTCAGCATGAG        12</span></span>
<span><span class="co">## FCAImmP7803035-CCAGCGATCCGAAGAG        14</span></span>
<span><span class="co">## FCAImmP7528296-ATAAGAGTCAAAGACA        12</span></span>
<span><span class="co">## FCAImmP7555860-AACTTTCTCAACGGGA        16</span></span>
<span><span class="co">## FCAImmP7292034-CGTCACTGTGGTCTCG        14</span></span></code></pre>
<p>Visualize the UMAP of the filtered data.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotUMAP</span><span class="op">(</span><span class="va">sce</span>, </span>
<span>         color_by <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="milo-object-and-neighbourhood-graph-construction">Milo object and neighbourhood graph construction<a class="anchor" aria-label="anchor" href="#milo-object-and-neighbourhood-graph-construction"></a>
</h3>
<p>We will use miloR to create the pseudobulks based on the gene
expression data. The goal is to construct a neighbourhood graph with
many neighbors with which we can sample the representative neighbours to
form the objects.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://marionilab.github.io/miloR" class="external-link">miloR</a></span><span class="op">)</span></span>
<span><span class="va">milo_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://marionilab.github.io/miloR/reference/Milo.html" class="external-link">Milo</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">milo_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://marionilab.github.io/miloR/reference/buildGraph.html" class="external-link">buildGraph</a></span><span class="op">(</span><span class="va">milo_object</span>, </span>
<span>                          k <span class="op">=</span> <span class="fl">30</span>, </span>
<span>                          d <span class="op">=</span> <span class="fl">20</span>, </span>
<span>                          reduced.dim <span class="op">=</span> <span class="st">"X_scvi"</span><span class="op">)</span></span>
<span><span class="va">milo_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://marionilab.github.io/miloR/reference/makeNhoods.html" class="external-link">makeNhoods</a></span><span class="op">(</span><span class="va">milo_object</span>, </span>
<span>                          reduced_dims <span class="op">=</span> <span class="st">"X_scvi"</span>, </span>
<span>                          d <span class="op">=</span> <span class="fl">20</span>, </span>
<span>                          prop <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="construct-umap-on-milo-neighbor-graph">Construct UMAP on milo neighbor graph<a class="anchor" aria-label="anchor" href="#construct-umap-on-milo-neighbor-graph"></a>
</h4>
<p>We can visualize this milo object using UMAP.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">milo_object</span> <span class="op">&lt;-</span> <span class="fu">miloUmap</span><span class="op">(</span><span class="va">milo_object</span>, </span>
<span>                        n_neighbors <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu">plotUMAP</span><span class="op">(</span><span class="va">milo_object</span>, </span>
<span>         color_by <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span>, </span>
<span>         dimred <span class="op">=</span> <span class="st">"UMAP_knngraph"</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="construct-pseudobulked-vdj-feature-space">Construct pseudobulked VDJ feature space<a class="anchor" aria-label="anchor" href="#construct-pseudobulked-vdj-feature-space"></a>
</h3>
<p>Next, we will construct the pseudobulked VDJ feature space using the
neighbourhood graph constructed above. We will also run PCA on the
pseudobulked VDJ feature space.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pb.milo</span> <span class="op">&lt;-</span> <span class="fu">vdjPseudobulk</span><span class="op">(</span><span class="va">milo_object</span>, </span>
<span>                         mode_option <span class="op">=</span> <span class="st">"abT"</span>, </span>
<span>                         col_to_take <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span><span class="op">)</span></span></code></pre></div>
<p>We can compute and visualize the PCA of the pseudobulked VDJ feature
space.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pb.milo</span> <span class="op">&lt;-</span> <span class="fu">runPCA</span><span class="op">(</span><span class="va">pb.milo</span>, </span>
<span>                  assay.type <span class="op">=</span> <span class="st">"Feature_space"</span>, </span>
<span>                  ncomponents <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="fu">plotPCA</span><span class="op">(</span><span class="va">pb.milo</span>, </span>
<span>        color_by <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="tcr-trajectory-inference-using-absorbing-markov-chain">TCR trajectory inference using Absorbing Markov Chain<a class="anchor" aria-label="anchor" href="#tcr-trajectory-inference-using-absorbing-markov-chain"></a>
</h3>
<p>In the original <code>dandelion</code> python package, the trajectory
inference is done using the <code>palantir</code> package. Here, we
implement the absorbing markov chain approach in dandelionR to infer the
trajectory, leveraging on <code>destiny</code> for diffusion map
computation.</p>
<div class="section level4">
<h4 id="define-root-and-branch-tips">Define root and branch tips<a class="anchor" aria-label="anchor" href="#define-root-and-branch-tips"></a>
</h4>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu">reducedDim</span><span class="op">(</span><span class="va">pb.milo</span>, </span>
<span>                              type <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># define the CD8 terminal cell as the top-most cell and CD4 terminal cell as</span></span>
<span><span class="co"># the bottom-most cell</span></span>
<span><span class="va">branch.tips</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">pca</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.min</a></span><span class="op">(</span><span class="va">pca</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">branch.tips</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD8+T"</span>, <span class="st">"CD4+T"</span><span class="op">)</span></span>
<span><span class="co"># define the start of our trajectory as the right-most cell</span></span>
<span><span class="va">root</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">pca</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="construct-diffusion-map">Construct diffusion map<a class="anchor" aria-label="anchor" href="#construct-diffusion-map"></a>
</h4>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://theislab.github.io/destiny/" class="external-link">destiny</a></span><span class="op">)</span></span>
<span><span class="co"># Run diffusion map on the PCA</span></span>
<span><span class="va">feature_space</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">pb.milo</span>, <span class="st">"Feature_space"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/destiny/man/DiffusionMap-class.html" class="external-link">DiffusionMap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">feature_space</span><span class="op">)</span>, </span>
<span>                   n_pcs <span class="op">=</span> <span class="fl">10</span>, </span>
<span>                   n_eigs <span class="op">=</span> <span class="fl">10</span>, </span>
<span>                   sigma <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="compute-diffusion-pseudotime-on-diffusion-map">Compute diffusion pseudotime on diffusion map<a class="anchor" aria-label="anchor" href="#compute-diffusion-pseudotime-on-diffusion-map"></a>
</h4>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dif.pse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/destiny/man/DPT.html" class="external-link">DPT</a></span><span class="op">(</span><span class="va">dm</span>, tips <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">root</span>, <span class="va">branch.tips</span><span class="op">)</span>, w_width <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># the root is automatically called DPT + index of the root cell</span></span>
<span><span class="va">DPTroot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"DPT"</span>, <span class="va">root</span><span class="op">)</span></span>
<span><span class="co"># store pseudotime in milo object</span></span>
<span><span class="va">pb.milo</span><span class="op">$</span><span class="va">pseudotime</span> <span class="op">&lt;-</span> <span class="va">dif.pse</span><span class="op">[[</span><span class="va">DPTroot</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co"># set the colours for pseudotime</span></span>
<span><span class="va">pal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="op">(</span><span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">9</span>, <span class="st">"RdYlBu"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">255</span><span class="op">)</span></span>
<span><span class="fu">plotPCA</span><span class="op">(</span><span class="va">pb.milo</span>, color_by <span class="op">=</span> <span class="st">"pseudotime"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_colour_gradientn</span><span class="op">(</span>colours <span class="op">=</span> <span class="va">pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="markov-chain-construction-on-the-pseudobulk-vdj-feature-space">Markov chain construction on the pseudobulk VDJ feature space<a class="anchor" aria-label="anchor" href="#markov-chain-construction-on-the-pseudobulk-vdj-feature-space"></a>
</h4>
<p>This step will compute the Markov chain probabilities on the
pseudobulk VDJ feature space. It will return the branch probabilities in
the <code>colData</code> and the column name corresponds to the branch
tips defined earlier.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pb.milo</span> <span class="op">&lt;-</span> <span class="fu">markovProbability</span><span class="op">(</span></span>
<span>    milo <span class="op">=</span> <span class="va">pb.milo</span>,</span>
<span>    diffusionmap <span class="op">=</span> <span class="va">dm</span>,</span>
<span>    terminal_state <span class="op">=</span> <span class="va">branch.tips</span>,</span>
<span>    root_cell <span class="op">=</span> <span class="va">root</span>,</span>
<span>    n_eigs <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    pseudotime_key <span class="op">=</span> <span class="st">"pseudotime"</span>,</span>
<span>    knn <span class="op">=</span> <span class="fl">30</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="visualising-branch-probabilities">Visualising branch probabilities<a class="anchor" aria-label="anchor" href="#visualising-branch-probabilities"></a>
</h4>
<p>With the Markov chain probabilities computed, we can visualise the
branch probabilities towards CD4+ or CD8+ T-cell fate on the PCA
plot.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotPCA</span><span class="op">(</span><span class="va">pb.milo</span>, </span>
<span>        color_by <span class="op">=</span> <span class="st">"CD8+T"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">scale_color_gradientn</span><span class="op">(</span>colors <span class="op">=</span> <span class="va">pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotPCA</span><span class="op">(</span><span class="va">pb.milo</span>, </span>
<span>        color_by <span class="op">=</span> <span class="st">"CD4+T"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">scale_color_gradientn</span><span class="op">(</span>colors <span class="op">=</span> <span class="va">pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-17-2.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="transfer">Transfer<a class="anchor" aria-label="anchor" href="#transfer"></a>
</h3>
<p>The next step is to project the pseudotime and the branch probability
information from the pseudobulks back to each cell in the dataset. If
the cell do not belong to any of the pseudobulk, it will be removed. If
a cell belongs to multiple pseudobulk samples, its value should be
calculated as a weighted average of the corresponding values from each
pseudobulk, where each weight is inverse of the size of the
pseudobulk.</p>
<div class="section level4">
<h4 id="project-pseudobulk-data-to-each-cell">Project pseudobulk data to each cell<a class="anchor" aria-label="anchor" href="#project-pseudobulk-data-to-each-cell"></a>
</h4>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cdata</span> <span class="op">&lt;-</span> <span class="fu">projectPseudotimeToCell</span><span class="op">(</span><span class="va">milo_object</span>, </span>
<span>                                 <span class="va">pb.milo</span>, </span>
<span>                                 <span class="va">branch.tips</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="visualise-the-trajectory-data-on-a-per-cell-basis">Visualise the trajectory data on a per cell basis<a class="anchor" aria-label="anchor" href="#visualise-the-trajectory-data-on-a-per-cell-basis"></a>
</h4>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plotting Cell Anntotation</span></span>
<span><span class="fu">plotUMAP</span><span class="op">(</span><span class="va">cdata</span>, color_by <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span>, </span>
<span>         dimred <span class="op">=</span> <span class="st">"UMAP_knngraph"</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plotting pseudotime</span></span>
<span><span class="fu">plotUMAP</span><span class="op">(</span><span class="va">cdata</span>, color_by <span class="op">=</span> <span class="st">"pseudotime"</span>, </span>
<span>         dimred <span class="op">=</span> <span class="st">"UMAP_knngraph"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">scale_color_gradientn</span><span class="op">(</span>colors <span class="op">=</span> <span class="va">pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-19-2.png" width="700"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plotting CD4 trajectory</span></span>
<span><span class="fu">plotUMAP</span><span class="op">(</span><span class="va">cdata</span>, color_by <span class="op">=</span> <span class="st">"CD4+T"</span>, </span>
<span>         dimred <span class="op">=</span> <span class="st">"UMAP_knngraph"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">scale_color_gradientn</span><span class="op">(</span>colors <span class="op">=</span> <span class="va">pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-19-3.png" width="700"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plotting CD8 trajectory</span></span>
<span><span class="fu">plotUMAP</span><span class="op">(</span><span class="va">cdata</span>, color_by <span class="op">=</span> <span class="st">"CD8+T"</span>, </span>
<span>         dimred <span class="op">=</span> <span class="st">"UMAP_knngraph"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">scale_color_gradientn</span><span class="op">(</span>colors <span class="op">=</span> <span class="va">pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-19-4.png" width="700"></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
