<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Frequently Asked Questions • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Frequently Asked Questions">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/dandelionR.html">VDJ Trajectory Analysis with dandelionR</a>
    </li>
    <li>
      <a href="../articles/immunarch.html">Using scRepertoire with immunarch</a>
    </li>
    <li>
      <a href="../articles/tcrpheno.html">TCRpheno applied to T cell fate</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BorchLab/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Frequently Asked Questions</h1>
            
            <h4 data-toc-skip class="date">Compiled: July 22, 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BorchLab/scRepertoire/blob/HEAD/vignettes/articles/FAQ.Rmd" class="external-link"><code>vignettes/articles/FAQ.Rmd</code></a></small>
      <div class="hidden name"><code>FAQ.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="barcode-mismatch">Barcode Mismatch<a class="anchor" aria-label="anchor" href="#barcode-mismatch"></a>
</h2>
<p>Depending on the pipeline used to generate the single-cell object,
there may be inherent mismatches in the barcodes in the single-cell
object and the output of <code><a href="../reference/combineBCR.html">combineBCR()</a></code> or
<code><a href="../reference/combineTCR.html">combineTCR()</a></code>.</p>
<div class="section level3">
<h3 id="common-barcode-modification-examples">Common Barcode Modification Examples<a class="anchor" aria-label="anchor" href="#common-barcode-modification-examples"></a>
</h3>
<p><strong>Seurat-modified barcodes</strong>: By default, Seurat will
append _X to the suffix of the barcodes.</p>
<ul>
<li><code>original: ACGTACGTACGTACGT-1</code></li>
<li><code>seurat-modified: ACGTACGTACGTACGT-1_1</code></li>
</ul>
<p><strong>scRepertoire-modified barcodes</strong>:
<code>scRepertoire</code> uses the samples and/or ID parameters in
<code><a href="../reference/combineTCR.html">combineTCR()</a></code> or <code><a href="../reference/combineBCR.html">combineBCR()</a></code> to add a prefix
to the barcodes.</p>
<ul>
<li><code>original: ACGTACGTACGTACGT-1</code></li>
<li><code>scRepertoire-modified: Sample1_ACGTACGTACGTACGT-1</code></li>
</ul>
</div>
<div class="section level3">
<h3 id="solution-renaming-cell-barcodes-in-seurat">Solution: Renaming Cell Barcodes in Seurat<a class="anchor" aria-label="anchor" href="#solution-renaming-cell-barcodes-in-seurat"></a>
</h3>
<p>The easiest way to make these compatible is to rename the cell
barcodes in the Seurat object using <code>RenameCells()</code> from the
<code>SeuratObject</code> package</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Assuming 'seuratObj' is your Seurat object</span></span>
<span><span class="va">cell.barcodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">seuratObj</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co"># removing the _1 at the end of the barcodes (adjust regex if your suffix differs)</span></span>
<span><span class="va">cell.barcodes</span> <span class="op">&lt;-</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split</a></span><span class="op">(</span><span class="va">cell.barcodes</span>, <span class="st">"_"</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span></span>
<span><span class="co"># adding the prefix of the orig.ident to the barcodes, assuming that is the sample IDs</span></span>
<span><span class="va">cell.barcodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">seuratObj</span><span class="op">$</span><span class="va">orig.ident</span>, <span class="st">"_"</span>, <span class="va">cell.barcodes</span><span class="op">)</span></span>
<span><span class="va">seuratObj</span> <span class="op">&lt;-</span> <span class="fu">RenameCells</span><span class="op">(</span><span class="va">seuratObj</span>, new.names <span class="op">=</span> <span class="va">cell.barcodes</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="adjusting-color-palettes">Adjusting Color Palettes<a class="anchor" aria-label="anchor" href="#adjusting-color-palettes"></a>
</h2>
<p>For all visualizations in <code>scRepertoire</code>, you have two
primary ways to adjust the color scheme.</p>
<p>Methods for Color Adjustment</p>
<ul>
<li>
<strong>Internal Palette Selection</strong>: Change the
<code>palette</code> parameter within <code>scRepertoire</code>
functions to the desired color scheme. This approach uses the built-in
palettes of <code>grDevices</code>, and you can access the list of
available color schemes using <code><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.pals()</a></code>.</li>
<li>
<strong>Adding a ggplot Layer</strong>: Extend the
<code>scRepertoire</code> plot (which is a ggplot object) by adding a
<code>ggplot2</code> layer with a new color scheme using
<code>scale_fill_manual()</code> or similar functions.</li>
</ul>
<div class="section level3">
<h3 id="using-internal-palette-selection">Using Internal Palette Selection<a class="anchor" aria-label="anchor" href="#using-internal-palette-selection"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Internal Palette Selection</span></span>
<span><span class="fu"><a href="../reference/clonalQuant.html">clonalQuant</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>            cloneCall<span class="op">=</span><span class="st">"strict"</span>, </span>
<span>            chain <span class="op">=</span> <span class="st">"both"</span>, </span>
<span>            scale <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>            palette <span class="op">=</span> <span class="st">"Zissou 1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="FAQ_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="using-ggplot2-system">Using ggplot2 System:<a class="anchor" aria-label="anchor" href="#using-ggplot2-system"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using gg System</span></span>
<span><span class="fu"><a href="../reference/clonalQuant.html">clonalQuant</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>            cloneCall<span class="op">=</span><span class="st">"strict"</span>, </span>
<span>            chain <span class="op">=</span> <span class="st">"both"</span>, </span>
<span>            scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">scale_fill_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.colors</a></span><span class="op">(</span><span class="fl">8</span>,<span class="st">"geyser"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="FAQ_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="adjusting-plot-theme">Adjusting Plot Theme<a class="anchor" aria-label="anchor" href="#adjusting-plot-theme"></a>
</h3>
<p>Since <code>scRepertoire</code> functions return <code>ggplot</code>
objects, modifying the general appearance or theme of the plot is
straightforward, similar to adjusting color palettes—by adding a
<code>ggplot2</code> theme layer.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Original clonalQuant plot</span></span>
<span><span class="fu"><a href="../reference/clonalQuant.html">clonalQuant</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>            cloneCall<span class="op">=</span><span class="st">"strict"</span>, </span>
<span>            chain <span class="op">=</span> <span class="st">"both"</span>, </span>
<span>            scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="FAQ_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Modifying the theme of the clonalQuant plot</span></span>
<span><span class="fu"><a href="../reference/clonalQuant.html">clonalQuant</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>            cloneCall<span class="op">=</span><span class="st">"strict"</span>, </span>
<span>            chain <span class="op">=</span> <span class="st">"both"</span>, </span>
<span>            scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="FAQ_files/figure-html/unnamed-chunk-4-2.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="adjusting-order-of-plotting">Adjusting Order of Plotting<a class="anchor" aria-label="anchor" href="#adjusting-order-of-plotting"></a>
</h3>
<p>The order of grouping/group.by variables in <code>scRepertoire</code>
plots (whether along an axis or in color legends) can be precisely
controlled using the order.by parameter.</p>
<p>Key Parameter for Plot Order</p>
<ul>
<li>
<code>order.by</code>: A character vector defining the desired order
of elements for the <code>group.by</code> variable. It’s crucial that
the strings in this vector exactly match the <code>group.by</code>
strings. Alternatively, setting <code>order.by = "alphanumeric"</code>
will automatically sort groups alphanumerically.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalQuant.html">clonalQuant</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>            cloneCall<span class="op">=</span><span class="st">"strict"</span>, </span>
<span>            chain <span class="op">=</span> <span class="st">"both"</span>, </span>
<span>            scale <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>            order.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17B"</span>,<span class="st">"P18B"</span>,<span class="st">"P19B"</span>,<span class="st">"P20B"</span>,<span class="st">"P17L"</span>,<span class="st">"P18L"</span>,<span class="st">"P19L"</span>,<span class="st">"P20L"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="FAQ_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="getting-data-used-in-plots">Getting Data Used in Plots<a class="anchor" aria-label="anchor" href="#getting-data-used-in-plots"></a>
</h3>
<p>Within each of the general analysis functions in
<code>scRepertoire</code>, there’s an option to export the underlying
data frame used to create the visualization.</p>
<p>Key Parameter for Data Export</p>
<ul>
<li>
<code>exportTable</code>: Set this parameter to <code>TRUE</code> to
return the data frame used to generate the graph instead of the visual
output.</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clonalQuant_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clonalQuant.html">clonalQuant</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                                  cloneCall<span class="op">=</span><span class="st">"strict"</span>, </span>
<span>                                  scale <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                  exportTable <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">clonalQuant_output</span></span></code></pre></div>
<pre><code><span><span class="co">##   contigs values total   scaled</span></span>
<span><span class="co">## 1     745   P17B  2805 26.55971</span></span>
<span><span class="co">## 2    2117   P17L  2893 73.17663</span></span>
<span><span class="co">## 3    1254   P18B  1328 94.42771</span></span>
<span><span class="co">## 4    1202   P18L  1278 94.05321</span></span>
<span><span class="co">## 5    5544   P19B  6942 79.86171</span></span>
<span><span class="co">## 6    1619   P19L  2747 58.93702</span></span>
<span><span class="co">## 7    6087   P20B  8991 67.70103</span></span>
<span><span class="co">## 8     192   P20L   201 95.52239</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="citing-screpertoire">Citing scRepertoire<a class="anchor" aria-label="anchor" href="#citing-screpertoire"></a>
</h3>
<p>When using <code>scRepertoire</code> in your research, please cite
the appropriate version of the package.</p>
<p>Citation Details</p>
<ul>
<li>
<strong>Version 2</strong>: Yang, Q, &amp; Safina, K., Nguyen, K.,
Tuong, Z.K., &amp; Borcherding, N. (2025). “scRepertoire 2: Enhanced and
efficient toolkit for single-cell immune profiling.” <em>PLoS
Computational Biology</em> <a href="https://doi.org/10.1371/journal.pcbi.1012760" class="external-link uri">https://doi.org/10.1371/journal.pcbi.1012760</a>
</li>
<li>
<strong>Version 1</strong>: Borcherding, Nicholas, Nicholas L.
Bormann, and Gloria Kraus. “scRepertoire: An R-based toolkit for
single-cell immune receptor analysis.” <em>F1000Research</em> <a href="https://doi.org/10.12688/f1000research.22139.2" class="external-link uri">https://doi.org/10.12688/f1000research.22139.2</a>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="bug-reportsnew-features">Bug Reports/New Features<a class="anchor" aria-label="anchor" href="#bug-reportsnew-features"></a>
</h3>
<p>Your feedback is valuable for improving scRepertoire! If you
encounter a bug or have a suggestion for a new feature, please report
it.</p>
<p>Submit a <a href="https://github.com/BorchLab/scRepertoire/issues" class="external-link">GitHub issue</a>
- if possible please include a <a href="https://reprex.tidyverse.org/" class="external-link">reproducible example</a>.
Alternatively, an example with the internal
<strong>scRep_example</strong> and <strong>contig_list</strong> would be
extremely helpful.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
