<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Making Deep Learning Models with immApex • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Making Deep Learning Models with immApex">
<meta property="og:description" content="scRepertoire">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ncborcherding/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Making Deep Learning Models with immApex</h1>
            
            <h4 data-toc-skip class="date">Compiled: August 09,
2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ncborcherding/scRepertoire/blob/HEAD/vignettes/articles/immApex.Rmd" class="external-link"><code>vignettes/articles/immApex.Rmd</code></a></small>
      <div class="hidden name"><code>immApex.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="getting-started">Getting Started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p><strong>immApex</strong> is meant to serve as an API for
deep-learning models based on immune receptor sequencing. These
functions extract or generate amino acid or nucleotide sequences and
prepare them for deep learning tasks through <a href="https://tensorflow.rstudio.com/guides/keras/basics" class="external-link">Keras</a>.
<strong>immApex</strong> is the underlying structure for the BCR models
in <a href="https://github.com/ncborcherding/Ibex" class="external-link">Ibex</a> and TCR
models in <a href="https://github.com/ncborcherding/Trex" class="external-link">Trex</a>. It
should be noted, although the tools here are created for immune receptor
sequences, they will work more generally for nucleotide or amino acid
sequences.</p>
<p>More information is available at the <a href="https://github.com/ncborcherding/immApex" class="external-link">immApex GitHub
Repo</a>.</p>
</div>
<div class="section level2">
<h2 id="getting-and-manipulating-sequences">Getting and Manipulating Sequences<a class="anchor" aria-label="anchor" href="#getting-and-manipulating-sequences"></a>
</h2>
<div class="section level3">
<h3 id="generatesequences">generateSequences<a class="anchor" aria-label="anchor" href="#generatesequences"></a>
</h3>
<p>Generating synthetic sequences is a quick way to start testing the
model code. <code>generateSequences()</code> can also generate realistic
noise for generative adversarial networks.</p>
<p>Parameters for <code>generateSequences()</code></p>
<ul>
<li>
<strong>prefix.motif</strong> Add a defined sequence to the start of
the generated sequences.</li>
<li>
<strong>suffix.motif</strong> Add a defined sequence to the end of
the generated sequences number.of.sequences Number of sequences to
generate</li>
<li>
<strong>min.length</strong> Minimum length of the final sequence
(will be adjusted if incongruent with prefix.motif/suffix.motif)</li>
<li>
<strong>max.length</strong> Maximum length of the final
sequence</li>
<li>
<strong>sequence.dictionary</strong> The letters to use in sequence
generation (default are all amino acids)</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sequences</span> <span class="op">&lt;-</span> <span class="fu">generateSequences</span><span class="op">(</span>prefix.motif <span class="op">=</span> <span class="st">"CAS"</span>,</span>
<span>                               suffix.motif <span class="op">=</span> <span class="st">"YF"</span>,</span>
<span>                               number.of.sequences <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                               min.length <span class="op">=</span> <span class="fl">8</span>,</span>
<span>                               max.length <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sequences</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CASFKRRPLNFYF"    "CASDGLNSMSFTKYF"  "CASVVDNYVKHHFKYF" "CASMEHTRMYDRTYF" </span></span>
<span><span class="co">## [5] "CASPFWKERYRYF"    "CASMRIGYMRAGYF"</span></span></code></pre>
<p>If we want to generate nucleotide sequences instead of amino acids,
we just need to change the <strong>sequence.dictionary</strong>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nucleotide.sequences</span> <span class="op">&lt;-</span> <span class="fu">generateSequences</span><span class="op">(</span>number.of.sequences <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                                          min.length <span class="op">=</span> <span class="fl">8</span>,</span>
<span>                                          max.length <span class="op">=</span> <span class="fl">16</span>, </span>
<span>                                          sequence.dictionary <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"C"</span>, <span class="st">"T"</span>, <span class="st">"G"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">nucleotide.sequences</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "ACCGTTCGCCCGGCCC" "GATGGTACTGTATCC"  "CGACCGTGTGA"      "CTCACGGCTGGATAAT"</span></span>
<span><span class="co">## [5] "TCCTTTAGGTG"      "CGAGTTGACGCTCG"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="variationalsequences">variationalSequences<a class="anchor" aria-label="anchor" href="#variationalsequences"></a>
</h3>
<p>In addition to making random sequences with
<code>generateSequences()</code>, we can also use generative deep
learning to simulate similar, de novo sequences with
<code>variationalSequences()</code>. <code>variationalSequences()</code>
uses a variational autoencoder that allows for sampling and generation
of sequences similar to the input sequences.</p>
<p>It should be noted that the success of this approach is highly
dependent on the number of sequences used as input and the
hyperparameters of the model. As such,
<code>variationalSequences()</code> has a number of arguments to modify
to allow for optimization.</p>
<p>Parameters for `<code>variationalSequences()</code></p>
<ul>
<li>
<strong>input.sequences</strong> The amino acid or nucleotide
sequences to use</li>
<li>
<strong>encoder.function</strong> The method to prepare the
sequencing information - <strong>“onehotEncoder”</strong> or
<strong>“propertyEncoder”</strong>
</li>
<li>
<strong>aa.method.to.use</strong> The method or approach to use for
the conversion, see <code>propertyEncoder()</code>
</li>
<li>
<strong>number.of.sequences</strong> Number of sequences to
generate</li>
<li>
<strong>encoder.hidden.dim</strong> A vector of the neurons to use
in the hidden layers for the encoder portion of the model</li>
<li>
<strong>decoder.hidden.dim</strong> A vector of the neurons to use
in the hidden layers for the decoder portion of the model. If NULL
assumes symmetric autoencoder</li>
<li>
<strong>latent.dim</strong> The size of the latent dimensions</li>
<li>
<strong>batch.size</strong> The batch size to use for VAE
training</li>
<li>
<strong>epochs</strong> The number of epochs to use in VAE
training</li>
<li>
<strong>learning.rate</strong> The learning rate to use in VAE
training</li>
<li>
<strong>epsilon.std</strong> The epsilon to use in VAE training</li>
<li>
<strong>call.threshold</strong> The relative strictness of sequence
calling with higher values being more stringent</li>
<li>
<strong>activation.function</strong> The activation for the dense
connected layers</li>
<li>
<strong>optimizer</strong> The optimizer to use in VAE training</li>
<li>
<strong>disable.eager.execution</strong> Disable the eager execution
parameter for tensorflow.</li>
<li>
<strong>sequence.dictionary</strong> The letters to use in sequence
mutation (default are all amino acids)</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variational.sequences</span> <span class="op">&lt;-</span> <span class="fu">variationalSequences</span><span class="op">(</span><span class="va">sequences</span>, encoder <span class="op">=</span> <span class="st">"onehotEncoder"</span>, number.of.sequences <span class="op">=</span> <span class="fl">100</span>,</span>
<span>    encoder.hidden.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">256</span>, <span class="fl">128</span><span class="op">)</span>, latent.dim <span class="op">=</span> <span class="fl">16</span>, batch.size <span class="op">=</span> <span class="fl">16</span>, call.threshold <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Converting to matrix...."</span></span>
<span><span class="co">## [1] "Padding sequences..."</span></span>
<span><span class="co">## [1] "One Hot Encoding sequences..."</span></span>
<span><span class="co">## [1] "Preparing a matrix..."</span></span>
<span><span class="co">## [1] "Fitting Model...."</span></span>
<span><span class="co">## [1] "Generating New Sequences...."</span></span>
<span><span class="co">## 4/4 - 0s - 36ms/epoch - 9ms/step</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">variational.sequences</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CASNSHSGHRRQYF"   "CASCMMWQEEWYF"    "CASNRLTTKNRQYF"   "CASEDAKGDQ.MQYF" </span></span>
<span><span class="co">## [5] "CASAKVKGPVFM.YF"  "CASEKGDMPVGYF..F"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="mutatesequences">mutateSequences<a class="anchor" aria-label="anchor" href="#mutatesequences"></a>
</h3>
<p>A common approach is to mutate sequences randomly or at specific
intervals. This can be particularly helpful if we have fewer sequences
or want to test a model for accuracy given new, altered
sequences. mutateSequences() allows us to tune the type of mutation,
where along the sequences to introduce the mutation and the overall
number of mutations.</p>
<p>Parameters for <code>mutateSequences()</code></p>
<ul>
<li>
<strong>input.sequences</strong> The amino acid or nucleotide
sequences to use</li>
<li>
<strong>n.sequences</strong> The number of mutated sequences to
return per input.sequence</li>
<li>
<strong>mutation.rate</strong> The rate of mutations introduced into
sequences</li>
<li>
<strong>position.start</strong> The starting position to mutate
along the sequence. Default NULL will start the random mutations at
position 1.</li>
<li>
<strong>position.end</strong> The ending position to mutate along
the sequence. Default NULL will end the random mutations at the last
position.</li>
<li>
<strong>sequence.dictionary</strong> The letters to use in sequence
mutation (default are all amino acids)</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mutated.sequences</span> <span class="op">&lt;-</span> <span class="fu">mutateSequences</span><span class="op">(</span><span class="va">sequences</span>, </span>
<span>                                     n.sequence <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                                     position.start <span class="op">=</span> <span class="fl">3</span>,                                  </span>
<span>                                     position.end <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sequences</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CASFKRRPLNFYF"    "CASDGLNSMSFTKYF"  "CASVVDNYVKHHFKYF" "CASMEHTRMYDRTYF" </span></span>
<span><span class="co">## [5] "CASPFWKERYRYF"    "CASMRIGYMRAGYF"</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mutated.sequences</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CAKFKRRPLNFYF"    "CASDGLQSMSFTKYF"  "CASVVANYVKHHFKYF" "CASMEHTKMYDRTYF" </span></span>
<span><span class="co">## [5] "CASPPWKERYRYF"    "CASMRIAYMRAGYF"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="formatgenes">formatGenes<a class="anchor" aria-label="anchor" href="#formatgenes"></a>
</h3>
<p>Immune receptor nomenclature can be highly variable across sequencing
platforms. When preparing data for models, we can use
`<code>formatGenes()</code> to universalize the gene formats into IMGT
nomenclature.</p>
<p>Parameters for <code>formatGenes()</code></p>
<ul>
<li>
<strong>input.data</strong> Data frame of sequencing data or
scRepertoire outputs</li>
<li>
<strong>region</strong> Sequence gene loci to access - ‘v’, ‘d’,
‘j’, ‘c’ or a combination using c(‘v’, ‘d’, ‘j’)</li>
<li>
<strong>technology</strong> The sequencing technology employed -
‘TenX’, “Adaptive’, ‘AIRR’, or ‘Omniscope’.</li>
<li>
<strong>species</strong> One or two word designation of species.
Currently supporting: “human”, “mouse”, “rat”, “rabbit”, “rhesus
monkey”, “sheep”, “pig”, “platypus”, “alpaca”, “dog”, “chicken”, and
“ferret”</li>
<li>
<strong>simplify.format</strong> If applicable, remove the allelic
designation (TRUE) or retain all information (FALSE)</li>
</ul>
<p>Here, we will use the built-in example from Adaptive Biotechnologies
and reformat and simplify the <strong>v</strong> region.
<code>formatGenes()</code> will add 2 columns to the end of the data
frame per region selected - 1) <strong>v_IMGT</strong> will be the
formatted gene calls and 2) <strong>v_IMGT.check</strong> is a binary
for if the formatted region appears in the IMGT database. In the example
below, “TRBV2-1” is not recognized as a designation within IMGT.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Adaptive_example</span> <span class="op">&lt;-</span> <span class="fu">formatGenes</span><span class="op">(</span><span class="va">immapex_example.data</span><span class="op">[[</span><span class="st">"Adaptive"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                                region <span class="op">=</span> <span class="st">"v"</span>,</span>
<span>                                technology <span class="op">=</span> <span class="st">"Adaptive"</span>, </span>
<span>                                simplify.format <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Adaptive_example</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"aminoAcid"</span>,<span class="st">"vGeneName"</span>, <span class="st">"v_IMGT"</span>, <span class="st">"v_IMGT.check"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            aminoAcid  vGeneName   v_IMGT v_IMGT.check</span></span>
<span><span class="co">## 1     CASSAGEYGYEQYF TCRBV12-03 TRBV12-3            1</span></span>
<span><span class="co">## 2 CASRNPGLASTGSYGQYF TCRBV06-02  TRBV6-2            1</span></span>
<span><span class="co">## 3   CATSDLGDVGGETQYF TCRBV24-01 TRBV24-1            1</span></span>
<span><span class="co">## 5    CASSPSAGRINTIYF TCRBV06-05  TRBV6-5            1</span></span>
<span><span class="co">## 6     CASRDGGATGELFF TCRBV06-02  TRBV6-2            1</span></span>
<span><span class="co">## 7    CASPSPSGRRETQYF TCRBV02-01  TRBV2-1            0</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="getimgt">getIMGT<a class="anchor" aria-label="anchor" href="#getimgt"></a>
</h3>
<p>Depending on the sequencing technology and the version, we might want
to expand the length of our sequence embedding approach. The first step
in the process is pulling the reference sequences from the
ImMunoGeneTics (IMGT) system using <code>getIMGT()</code>. More
information for IMGT can be found at <a href="https://www.imgt.org/" class="external-link">imgt.org</a>.</p>
<p>Parameters for <code>getIMGT()</code></p>
<ul>
<li>
<strong>species</strong> One or two word designation of species.
Currently supporting: “human”, “mouse”, “rat”, “rabbit”, “rhesus
monkey”, “sheep”, “pig”, “platypus”, “alpaca”, “dog”, “chicken”, and
“ferret”</li>
<li>
<strong>chain</strong> Sequence chain to access</li>
<li>
<strong>frame</strong> Designation for “all”, “inframe” or
“inframe+gap”</li>
<li>
<strong>region</strong> Sequence gene loci to access</li>
<li>
<strong>sequence.type</strong> Type of sequence - “aa” for amino
acid or “nt” for nucleotide</li>
</ul>
<p>Here, we will use the <code>getIMGT()</code> function to get the
amino acid sequences for the TRBV region to get all the sequences by V
gene allele.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">TRBV_aa</span> <span class="op">&lt;-</span> <span class="fu">getIMGT</span><span class="op">(</span>species <span class="op">=</span> <span class="st">"human"</span>,</span>
<span>                   chain <span class="op">=</span> <span class="st">"TRB"</span>,</span>
<span>                   frame <span class="op">=</span> <span class="st">"inframe"</span>,</span>
<span>                   region <span class="op">=</span> <span class="st">"v"</span>,</span>
<span>                   sequence.type <span class="op">=</span> <span class="st">"aa"</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## [1] "Getting the sequences from IMGT..."</span></span>
<span><span class="co">## [1] "Formatting IMGT sequences..."</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">TRBV_aa</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## $`TRBV1*01`</span></span>
<span><span class="co">## [1] "HPVREGIONAADTGITQTPKYLVTAMGSKRTMKREHLGHDSMYWYRQKAKKSLEFMFYYNCKEFIENKTVPNHFTPECPDSSRLYLHVVALQQEDSAAYLCTSSQ"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="infercdr">inferCDR<a class="anchor" aria-label="anchor" href="#infercdr"></a>
</h3>
<p>We can now use <code>inferCDR()</code> to add additional sequence
elements to our example data using the outputs of
<code>formatGenes()</code> and <code>getIMGT()</code>. Here, we will use
the function to isolate the complementarity-determining regions (CDR) 1
and 2. If the gene nomenclature does not match the IMGT the result will
be NA for the given sequences. Likewise, if the IMGT nomenclature has
been simplified, the first allelic match will be used for sequence
extraction.</p>
<p>Parameters for <code>inferCDR</code></p>
<ul>
<li>
<strong>input.data</strong> Data frame of sequencing data or output
from formatGenes().</li>
<li>
<strong>reference</strong> IMGT sequences from
<code>getIMGT()</code>
</li>
<li>
<strong>technology</strong> The sequencing technology employed -
‘TenX’, “Adaptive’, ‘AIRR’, or ‘Omniscope’</li>
<li>
<strong>sequence.type</strong> Type of sequence - “aa” for amino
acid or “nt” for nucleotide</li>
<li>
<strong>sequences</strong> The specific regions of the CDR loop to
get from the data.</li>
</ul>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Adaptive_example</span> <span class="op">&lt;-</span> <span class="fu">inferCDR</span><span class="op">(</span><span class="va">Adaptive_example</span>,</span>
<span>                             chain <span class="op">=</span> <span class="st">"TRB"</span>, </span>
<span>                             reference <span class="op">=</span> <span class="va">TRBV_aa</span>,</span>
<span>                             technology <span class="op">=</span> <span class="st">"Adaptive"</span>, </span>
<span>                             sequence.type <span class="op">=</span> <span class="st">"aa"</span>,</span>
<span>                             sequences <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CDR1"</span>, <span class="st">"CDR2"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">Adaptive_example</span><span class="op">[</span><span class="fl">200</span><span class="op">:</span><span class="fl">210</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CDR1_IMGT"</span>, <span class="st">"CDR2_IMGT"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##        CDR1_IMGT  CDR2_IMGT</span></span>
<span><span class="co">## 200 GRQVTLMCHQTW LIYYSAAADI</span></span>
<span><span class="co">## 201 GRQVTLMCHQTW LIYYSAAADI</span></span>
<span><span class="co">## 202 GRQVTLMCHQTW LIYYSAAADI</span></span>
<span><span class="co">## 203 GRQVTLMCHQTW LIYYSAAADI</span></span>
<span><span class="co">## 204 GRQVTLMCHQTW LIYYSAAADI</span></span>
<span><span class="co">## 205 GRQVTLMCHQTW LIYYSAAADI</span></span>
<span><span class="co">## 206 GRQVTLMCHQTW LIYYSAAADI</span></span>
<span><span class="co">## 207 GRQVTLMCHQTW LIYYSAAADI</span></span>
<span><span class="co">## 208 GRQVTLMCHQTW LIYYSAAADI</span></span>
<span><span class="co">## 209 GRQVTLMCHQTW LIYYSAAADI</span></span>
<span><span class="co">## 210 GRQVTLMCHQTW LIYYSAAADI</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="encoders">Encoders<a class="anchor" aria-label="anchor" href="#encoders"></a>
</h2>
<div class="section level3">
<h3 id="onehotencoder">onehotEncoder<a class="anchor" aria-label="anchor" href="#onehotencoder"></a>
</h3>
<p>One hot encoding of amino acid or nucleotide sequences is a common
method for transforming sequences into numeric matrices compatible with
Keras (or other workflows).</p>
<p>Parameters for <code>onehotEncoder()</code></p>
<ul>
<li>
<strong>input.sequences</strong> The amino acid or nucleotide
sequences to use</li>
<li>
<strong>max.length</strong> Additional length to pad, NULL will pad
sequences to the max length of input.sequences</li>
<li>
<strong>convert.to.matrix</strong> Return a matrix
(<strong>TRUE</strong>) or a 3D array (<strong>FALSE</strong>)</li>
<li>
<strong>sequence.dictionary</strong> The letters to use in encoding
(default are all amino acids + NA value)</li>
</ul>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sequence.matrix</span> <span class="op">&lt;-</span> <span class="fu">onehotEncoder</span><span class="op">(</span>input.sequences <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">sequences</span>, <span class="va">mutated.sequences</span><span class="op">)</span>, </span>
<span>                                 convert.to.matrix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Padding sequences..."</span></span>
<span><span class="co">## [1] "One Hot Encoding sequences..."</span></span>
<span><span class="co">## [1] "Preparing a matrix..."</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sequence.matrix</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      Pos.1_A Pos.1_R Pos.1_N Pos.1_D Pos.1_C Pos.1_Q Pos.1_E Pos.1_G Pos.1_H</span></span>
<span><span class="co">## [1,]       0       0       0       0       1       0       0       0       0</span></span>
<span><span class="co">## [2,]       0       0       0       0       1       0       0       0       0</span></span>
<span><span class="co">## [3,]       0       0       0       0       1       0       0       0       0</span></span>
<span><span class="co">## [4,]       0       0       0       0       1       0       0       0       0</span></span>
<span><span class="co">## [5,]       0       0       0       0       1       0       0       0       0</span></span>
<span><span class="co">## [6,]       0       0       0       0       1       0       0       0       0</span></span>
<span><span class="co">##      Pos.1_I Pos.1_L Pos.1_K Pos.1_M Pos.1_F Pos.1_P Pos.1_S Pos.1_T Pos.1_W</span></span>
<span><span class="co">## [1,]       0       0       0       0       0       0       0       0       0</span></span>
<span><span class="co">## [2,]       0       0       0       0       0       0       0       0       0</span></span>
<span><span class="co">## [3,]       0       0       0       0       0       0       0       0       0</span></span>
<span><span class="co">## [4,]       0       0       0       0       0       0       0       0       0</span></span>
<span><span class="co">## [5,]       0       0       0       0       0       0       0       0       0</span></span>
<span><span class="co">## [6,]       0       0       0       0       0       0       0       0       0</span></span>
<span><span class="co">##      Pos.1_Y Pos.1_V</span></span>
<span><span class="co">## [1,]       0       0</span></span>
<span><span class="co">## [2,]       0       0</span></span>
<span><span class="co">## [3,]       0       0</span></span>
<span><span class="co">## [4,]       0       0</span></span>
<span><span class="co">## [5,]       0       0</span></span>
<span><span class="co">## [6,]       0       0</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="propertyencoder">propertyEncoder<a class="anchor" aria-label="anchor" href="#propertyencoder"></a>
</h3>
<p>An alternative to one hot encoding is transforming the sequences into
an array/matrix of numerical values using amino acid properties.</p>
<p>These properties are largely based on dimensional reduction
strategies, but it is essential to know the assumptions for each
approach (links to original work below). <strong>Important to note:
</strong> this encoding strategy is specific for amino acids.</p>
<p><strong>method.to.use</strong></p>
<ul>
<li>atchleyFactors - <a href="https://pubmed.ncbi.nlm.nih.gov/15851683/" class="external-link">citation</a>
</li>
<li>crucianiProperties - <a href="https://analyticalsciencejournals.onlinelibrary.wiley.com/doi/abs/10.1002/cem.856" class="external-link">citation</a>
</li>
<li>FASGAI - <a href="https://pubmed.ncbi.nlm.nih.gov/18318694/" class="external-link">citation</a>
</li>
<li>kideraFactors - <a href="https://link.springer.com/article/10.1007/BF01025492" class="external-link">citation</a>
</li>
<li>MSWHIM - <a href="https://pubs.acs.org/doi/10.1021/ci980211b" class="external-link">citation</a>
</li>
<li>ProtFP - <a href="https://pubmed.ncbi.nlm.nih.gov/24059694/" class="external-link">citation</a>
</li>
<li>stScales - <a href="https://pubmed.ncbi.nlm.nih.gov/19373543/" class="external-link">citation</a>
</li>
<li>tScales - <a href="https://www.sciencedirect.com/science/article/abs/pii/S0022286006006314" class="external-link">citation</a>
</li>
<li>VHSE - <a href="https://pubmed.ncbi.nlm.nih.gov/15895431/" class="external-link">citation</a>
</li>
<li>zScales - <a href="https://pubmed.ncbi.nlm.nih.gov/9651153/" class="external-link">citation</a>
</li>
</ul>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">property.matrix</span> <span class="op">&lt;-</span> <span class="fu">propertyEncoder</span><span class="op">(</span>input.sequences <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">sequences</span>, <span class="va">mutated.sequences</span><span class="op">)</span>, </span>
<span>                                   method.to.use <span class="op">=</span> <span class="st">"FASGAI"</span>,</span>
<span>                                   convert.to.matrix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Padding sequences..."</span></span>
<span><span class="co">## [1] "Property-based Encoding sequences..."</span></span>
<span><span class="co">## [1] "Preparing a matrix..."</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">property.matrix</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       Pos.1_F1  Pos.1_F2   Pos.1_F3 Pos.1_F4 Pos.1_F5  Pos.1_F6  Pos.2_F1</span></span>
<span><span class="co">## [1,] 0.8189626 0.6210147 0.06362416        0 0.063684 0.5042316 0.5475782</span></span>
<span><span class="co">## [2,] 0.8189626 0.6210147 0.06362416        0 0.063684 0.5042316 0.5475782</span></span>
<span><span class="co">## [3,] 0.8189626 0.6210147 0.06362416        0 0.063684 0.5042316 0.5475782</span></span>
<span><span class="co">## [4,] 0.8189626 0.6210147 0.06362416        0 0.063684 0.5042316 0.5475782</span></span>
<span><span class="co">## [5,] 0.8189626 0.6210147 0.06362416        0 0.063684 0.5042316 0.5475782</span></span>
<span><span class="co">## [6,] 0.8189626 0.6210147 0.06362416        0 0.063684 0.5042316 0.5475782</span></span>
<span><span class="co">##       Pos.2_F2  Pos.2_F3  Pos.2_F4  Pos.2_F5  Pos.2_F6  Pos.3_F1  Pos.3_F2</span></span>
<span><span class="co">## [1,] 0.8428057 0.1736913 0.8702309 0.2374043 0.2587973 0.3064239 0.3803715</span></span>
<span><span class="co">## [2,] 0.8428057 0.1736913 0.8702309 0.2374043 0.2587973 0.3064239 0.3803715</span></span>
<span><span class="co">## [3,] 0.8428057 0.1736913 0.8702309 0.2374043 0.2587973 0.3064239 0.3803715</span></span>
<span><span class="co">## [4,] 0.8428057 0.1736913 0.8702309 0.2374043 0.2587973 0.3064239 0.3803715</span></span>
<span><span class="co">## [5,] 0.8428057 0.1736913 0.8702309 0.2374043 0.2587973 0.3064239 0.3803715</span></span>
<span><span class="co">## [6,] 0.8428057 0.1736913 0.8702309 0.2374043 0.2587973 0.3064239 0.3803715</span></span>
<span><span class="co">##       Pos.3_F3  Pos.3_F4  Pos.3_F5  Pos.3_F6   Pos.4_F1   Pos.4_F2</span></span>
<span><span class="co">## [1,] 0.1548993 0.6681727 0.2317614 0.3772829 0.90484370 0.69642362</span></span>
<span><span class="co">## [2,] 0.1548993 0.6681727 0.2317614 0.3772829 0.03057369 0.49209870</span></span>
<span><span class="co">## [3,] 0.1548993 0.6681727 0.2317614 0.3772829 0.01889385 0.76628777</span></span>
<span><span class="co">## [4,] 0.1548993 0.6681727 0.2317614 0.3772829 0.78083133 0.98835597</span></span>
<span><span class="co">## [5,] 0.1548993 0.6681727 0.2317614 0.3772829 0.33665407 0.05018021</span></span>
<span><span class="co">## [6,] 0.1548993 0.6681727 0.2317614 0.3772829 0.78083133 0.98835597</span></span></code></pre>
<p><code>propertyEncoder()</code> also allows us to use multiple
approaches simultaneously by setting <strong>method.to.use</strong> as a
vector.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mulit.property.matrix</span> <span class="op">&lt;-</span> <span class="fu">propertyEncoder</span><span class="op">(</span>input.sequences <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">sequences</span>, <span class="va">mutated.sequences</span><span class="op">)</span>, </span>
<span>                                         method.to.use <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"atchleyFactors"</span>, <span class="st">"kideraFactors"</span><span class="op">)</span>,</span>
<span>                                         convert.to.matrix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Padding sequences..."</span></span>
<span><span class="co">## [1] "Property-based Encoding sequences..."</span></span>
<span><span class="co">## [1] "Preparing a matrix..."</span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mulit.property.matrix</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      Pos.1_AF1 Pos.1_AF2 Pos.1_AF3 Pos.1_AF4 Pos.1_AF5 Pos.1_KF1 Pos.1_KF2</span></span>
<span><span class="co">## [1,]         0  0.551818 0.4960678 0.2994783 0.4864792 0.4640884 0.2635468</span></span>
<span><span class="co">## [2,]         0  0.551818 0.4960678 0.2994783 0.4864792 0.4640884 0.2635468</span></span>
<span><span class="co">## [3,]         0  0.551818 0.4960678 0.2994783 0.4864792 0.4640884 0.2635468</span></span>
<span><span class="co">## [4,]         0  0.551818 0.4960678 0.2994783 0.4864792 0.4640884 0.2635468</span></span>
<span><span class="co">## [5,]         0  0.551818 0.4960678 0.2994783 0.4864792 0.4640884 0.2635468</span></span>
<span><span class="co">## [6,]         0  0.551818 0.4960678 0.2994783 0.4864792 0.4640884 0.2635468</span></span>
<span><span class="co">##      Pos.1_KF3 Pos.1_KF4 Pos.1_KF5 Pos.1_KF6 Pos.1_KF7 Pos.1_KF8 Pos.1_KF9</span></span>
<span><span class="co">## [1,] 0.5643836 0.1511628 0.2675676         1         1 0.3454936 0.8553616</span></span>
<span><span class="co">## [2,] 0.5643836 0.1511628 0.2675676         1         1 0.3454936 0.8553616</span></span>
<span><span class="co">## [3,] 0.5643836 0.1511628 0.2675676         1         1 0.3454936 0.8553616</span></span>
<span><span class="co">## [4,] 0.5643836 0.1511628 0.2675676         1         1 0.3454936 0.8553616</span></span>
<span><span class="co">## [5,] 0.5643836 0.1511628 0.2675676         1         1 0.3454936 0.8553616</span></span>
<span><span class="co">## [6,] 0.5643836 0.1511628 0.2675676         1         1 0.3454936 0.8553616</span></span>
<span><span class="co">##      Pos.1_KF10 Pos.2_AF1  Pos.2_AF2 Pos.2_AF3 Pos.2_AF4 Pos.2_AF5</span></span>
<span><span class="co">## [1,]  0.8661616 0.2366665 0.06143816 0.5124824         1 0.5043415</span></span>
<span><span class="co">## [2,]  0.8661616 0.2366665 0.06143816 0.5124824         1 0.5043415</span></span>
<span><span class="co">## [3,]  0.8661616 0.2366665 0.06143816 0.5124824         1 0.5043415</span></span>
<span><span class="co">## [4,]  0.8661616 0.2366665 0.06143816 0.5124824         1 0.5043415</span></span>
<span><span class="co">## [5,]  0.8661616 0.2366665 0.06143816 0.5124824         1 0.5043415</span></span>
<span><span class="co">## [6,]  0.8661616 0.2366665 0.06143816 0.5124824         1 0.5043415</span></span></code></pre>
<p>If, instead, we would like to get the set of summarized values across
all amino acid residues for a given <strong>method.to.use</strong>, we
can use <strong>summary.function</strong> and select “median”, “mean”,
“sum”, variance (“vars”), or Median Absolute Deviation (“mads”).</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">median.property.matrix</span> <span class="op">&lt;-</span> <span class="fu">propertyEncoder</span><span class="op">(</span>input.sequences <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">sequences</span>, <span class="va">mutated.sequences</span><span class="op">)</span>, </span>
<span>                                          method.to.use <span class="op">=</span> <span class="st">"crucianiProperties"</span>,</span>
<span>                                          summary.function <span class="op">=</span> <span class="st">"median"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Padding sequences..."</span></span>
<span><span class="co">## [1] "Property-based Encoding sequences..."</span></span>
<span><span class="co">## [1] "Summarising properties using median..."</span></span></code></pre>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">median.property.matrix</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         PP1    PP2    PP3</span></span>
<span><span class="co">## [1,] 0.0850 0.4075 0.4325</span></span>
<span><span class="co">## [2,] 0.4400 0.1975 0.5525</span></span>
<span><span class="co">## [3,] 0.6550 0.3650 0.4700</span></span>
<span><span class="co">## [4,] 0.6775 0.3550 0.5275</span></span>
<span><span class="co">## [5,] 0.3775 0.4250 0.4000</span></span>
<span><span class="co">## [6,] 0.0825 0.3700 0.5275</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="geometricencoder">geometricEncoder<a class="anchor" aria-label="anchor" href="#geometricencoder"></a>
</h3>
<p>One approach to encode amino acid sequences is geometric isometry,
such as <a href="https://pubmed.ncbi.nlm.nih.gov/34349111/" class="external-link">GIANA</a>.</p>
<p>Parameters for <code>geometricEncoder()</code></p>
<ul>
<li>
<strong>method.to.use</strong> Select the following substitution
matrices: “BLOSUM45”, “BLOSUM50”, “BLOSUM62”, “BLOSUM80”, “BLOSUM100”,
“PAM30”, “PAM40”, “PAM70”, “PAM120”, or “PAM250”</li>
<li>
<strong>theta</strong> The angle in which to create the rotation
matrix</li>
</ul>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">geometric.matrix</span> <span class="op">&lt;-</span> <span class="fu">geometricEncoder</span><span class="op">(</span><span class="va">sequences</span>, </span>
<span>                                     method.to.use <span class="op">=</span> <span class="st">"BLOSUM62"</span>,</span>
<span>                                     theta <span class="op">=</span> <span class="va">pi</span><span class="op">/</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Performing geometric transformation..."</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">geometric.matrix</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            [,1]        [,2]      [,3]        [,4]      [,5]      [,6]      [,7]</span></span>
<span><span class="co">## [1,] -0.9175541  0.35848108 -2.270512 -0.06735726 -1.548027 0.5274159 -2.529437</span></span>
<span><span class="co">## [2,] -1.4124356 -0.35358984 -1.248162 -0.10478645 -1.366025 0.3660254 -1.639230</span></span>
<span><span class="co">## [3,] -1.5970349  0.01614564 -1.682412  0.03902223 -1.522275 0.6366583 -2.317427</span></span>
<span><span class="co">## [4,] -0.7952135  0.31068360 -1.530299  0.11722202 -1.004145 0.8058972 -2.156453</span></span>
<span><span class="co">## [5,] -0.9560156  0.42509842 -2.434664  0.37080831 -1.358481 0.8144967 -2.298668</span></span>
<span><span class="co">## [6,] -0.8137361 -0.30485351 -2.631913  0.13003629 -1.392170 0.2684522 -1.723168</span></span>
<span><span class="co">##            [,8]       [,9]      [,10]     [,11]     [,12]      [,13]</span></span>
<span><span class="co">## [1,]  0.2272679 -1.6606506  0.1071004 -1.099556 0.5198716 -0.4127712</span></span>
<span><span class="co">## [2,]  0.4392305 -1.7547005  0.3725638 -1.374957 0.5148288 -0.8196152</span></span>
<span><span class="co">## [3,] -0.1110980 -0.7806488 -0.5228766 -1.328646 0.6762818 -0.6768829</span></span>
<span><span class="co">## [4,] -0.3982480 -1.3457689 -0.4690599 -1.292820 0.6392305 -0.8350853</span></span>
<span><span class="co">## [5,] -0.1724361 -1.9449700 -0.3235209 -1.302169 1.0246545 -0.6998520</span></span>
<span><span class="co">## [6,]  0.9846150 -1.4993129  0.4540290 -1.515888 0.1970236 -0.6020145</span></span>
<span><span class="co">##            [,14]      [,15]    [,16]     [,17]       [,18]      [,19]</span></span>
<span><span class="co">## [1,]  0.56109451 -1.4175541 1.224506 -1.965581  0.01987156 -1.3426525</span></span>
<span><span class="co">## [2,]  0.21961524 -0.7113249 1.898717 -1.783247 -0.64465820 -1.4636322</span></span>
<span><span class="co">## [3,]  0.54739564 -1.6578925 1.496554 -2.038421 -0.09435118 -0.4078925</span></span>
<span><span class="co">## [4,] -0.02025651 -1.2553418 1.507649 -1.518974 -0.56905989 -1.1058972</span></span>
<span><span class="co">## [5,]  0.75064079 -1.5123272 1.080966 -1.043244  0.73002931 -1.2732738</span></span>
<span><span class="co">## [6,] -0.10013741 -1.5401556 1.667628 -1.834752  0.17788436 -0.9470236</span></span>
<span><span class="co">##             [,20]</span></span>
<span><span class="co">## [1,] -0.597534553</span></span>
<span><span class="co">## [2,]  0.001751963</span></span>
<span><span class="co">## [3,] -0.668509526</span></span>
<span><span class="co">## [4,] -0.484529946</span></span>
<span><span class="co">## [5,] -1.179240540</span></span>
<span><span class="co">## [6,] -0.216849885</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="tokenizesequences">tokenizeSequences<a class="anchor" aria-label="anchor" href="#tokenizesequences"></a>
</h3>
<p>Another approach to transforming a sequence into numerical values is
tokenizing it into numbers. This is a common approach for recurrent
neural networks where one letter corresponds to a single integer. In
addition, we can add a start and stop tokens to our original sequences
to differentiate between the beginning and end of the sequences.</p>
<p>Parameters for <code>tokenizeSequences()</code></p>
<ul>
<li>
<strong>add.startstop</strong> Add start and stop tokens to the
sequence</li>
<li>
<strong>start.token</strong> The character to use for the start
token</li>
<li>
<strong>stop.token</strong> The character to use for the stop
token</li>
<li>
<strong>max.length</strong> Additional length to pad, NULL will pad
sequences to the max length of input.sequences</li>
<li>
<strong>convert.to.matrix</strong> Return a matrix
(<strong>TRUE</strong>) or a vector (<strong>FALSE</strong>)</li>
</ul>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">token.matrix</span> <span class="op">&lt;-</span> <span class="fu">tokenizeSequences</span><span class="op">(</span>input.sequences <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">sequences</span>, <span class="va">mutated.sequences</span><span class="op">)</span>, </span>
<span>                                  add.startstop <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                  start.token <span class="op">=</span> <span class="st">"!"</span>,</span>
<span>                                  stop.token <span class="op">=</span> <span class="st">"^"</span>, </span>
<span>                                  convert.to.matrix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Adding start and stop tokens..."</span></span>
<span><span class="co">## [1] "Tokenizing sequences..."</span></span>
<span><span class="co">## [1] "Padding sequences..."</span></span>
<span><span class="co">## [1] "Preparing a tokenized matrix..."</span></span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">token.matrix</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">18</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                  [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12]</span></span>
<span><span class="co">## CASFKRRPLNFYF       1    6    2   17   15   13    3    3   16    12     4    15</span></span>
<span><span class="co">## CASDGLNSMSFTKYF     1    6    2   17    5    9   12    4   17    14    17    15</span></span>
<span><span class="co">## CASVVDNYVKHHFKYF    1    6    2   17   21   21    5    4   20    21    13    10</span></span>
<span><span class="co">## CASMEHTRMYDRTYF     1    6    2   17   14    8   10   18    3    14    20     5</span></span>
<span><span class="co">## CASPFWKERYRYF       1    6    2   17   16   15   19   13    8     3    20     3</span></span>
<span><span class="co">## CASMRIGYMRAGYF      1    6    2   17   14    3   11    9   20    14     3     2</span></span>
<span><span class="co">##                  [,13] [,14] [,15] [,16] [,17] [,18]</span></span>
<span><span class="co">## CASFKRRPLNFYF       20    15    23    24    24    24</span></span>
<span><span class="co">## CASDGLNSMSFTKYF     18    13    20    15    23    24</span></span>
<span><span class="co">## CASVVDNYVKHHFKYF    10    15    13    20    15    23</span></span>
<span><span class="co">## CASMEHTRMYDRTYF      3    18    20    15    23    24</span></span>
<span><span class="co">## CASPFWKERYRYF       20    15    23    24    24    24</span></span>
<span><span class="co">## CASMRIGYMRAGYF       9    20    15    23    24    24</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="probabilitymatrix">probabilityMatrix<a class="anchor" aria-label="anchor" href="#probabilitymatrix"></a>
</h3>
<p>Another method for encoding a group of sequences is to calculate the
positional probability of sequences using
<code>probabilityMatrixs()</code>. This function could represent a
collection of antigen-specific sequences or even work on embedding a
total repertoire.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ppm.matrix</span> <span class="op">&lt;-</span> <span class="fu">probabilityMatrix</span><span class="op">(</span><span class="va">sequences</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Padding sequences..."</span></span>
<span><span class="co">## [1] "Calculating Positional Probabilites for sequences..."</span></span></code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ppm.matrix</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   Pos.1 Pos.2 Pos.3 Pos.4 Pos.5 Pos.6 Pos.7 Pos.8 Pos.9 Pos.10 Pos.11 Pos.12</span></span>
<span><span class="co">## A     0     1     0 0.055 0.048 0.044 0.040 0.041 0.042  0.043  0.048  0.029</span></span>
<span><span class="co">## R     0     0     0 0.041 0.046 0.062 0.044 0.052 0.054  0.056  0.042  0.042</span></span>
<span><span class="co">## N     0     0     0 0.055 0.037 0.042 0.041 0.045 0.036  0.063  0.056  0.027</span></span>
<span><span class="co">## D     0     0     0 0.057 0.047 0.042 0.042 0.051 0.051  0.048  0.047  0.037</span></span>
<span><span class="co">## C     1     0     0 0.054 0.050 0.053 0.060 0.054 0.039  0.044  0.048  0.037</span></span>
<span><span class="co">## Q     0     0     0 0.052 0.040 0.057 0.056 0.063 0.055  0.050  0.057  0.037</span></span>
<span><span class="co">##   Pos.13 Pos.14 Pos.15 Pos.16</span></span>
<span><span class="co">## A  0.020  0.011      0      0</span></span>
<span><span class="co">## R  0.027  0.009      0      0</span></span>
<span><span class="co">## N  0.020  0.016      0      0</span></span>
<span><span class="co">## D  0.027  0.009      0      0</span></span>
<span><span class="co">## C  0.026  0.013      0      0</span></span>
<span><span class="co">## Q  0.028  0.011      0      0</span></span></code></pre>
<p>In addition, <code>probabilityMatrix()</code> can convert the
positional probability matrix into a positional weight matrix using
log-likelihood using the argument <strong>convert.PWM</strong> = TRUE.
We can provide a set of background frequencies for the amino acids with
<strong>background.frequencies</strong> or leave this blank to assume a
uniform distribution for all amino acids. Here, we are going to use an
example background.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="va">back.freq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span>, <span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">back.freq</span> <span class="op">&lt;-</span> <span class="va">back.freq</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">back.freq</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pwm.matrix</span> <span class="op">&lt;-</span> <span class="fu">probabilityMatrix</span><span class="op">(</span><span class="va">sequences</span>,</span>
<span>                                max.length <span class="op">=</span> <span class="fl">20</span>,</span>
<span>                                convert.PWM <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                background.frequencies <span class="op">=</span> <span class="va">back.freq</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Padding sequences..."</span></span>
<span><span class="co">## [1] "Calculating Positional Probabilites for sequences..."</span></span>
<span><span class="co">## [1] "Converting to Liklihoods for a PWM..."</span></span></code></pre>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pwm.matrix</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       Pos.1     Pos.2     Pos.3      Pos.4      Pos.5       Pos.6       Pos.7</span></span>
<span><span class="co">## A -6.157338  3.809888 -6.157338 -0.3499835 -0.5426286 -0.66548534 -0.79978643</span></span>
<span><span class="co">## R -6.986931 -6.986931 -6.986931 -1.5946137 -1.4323423 -1.00965124 -1.49507807</span></span>
<span><span class="co">## N -5.351911 -5.351911 -5.351911  0.4554440 -0.1039834  0.07435379  0.04040646</span></span>
<span><span class="co">## D -4.282869 -4.282869 -4.282869  1.5751117  1.3020932  1.14339544  1.14339544</span></span>
<span><span class="co">## C  6.732291 -3.234935 -3.234935  2.5464249  2.4374905  2.51995266  2.69580250</span></span>
<span><span class="co">## Q -4.858371 -4.858371 -4.858371  0.8695490  0.4991805  0.99960951  0.97451853</span></span>
<span><span class="co">##       Pos.8      Pos.9     Pos.10     Pos.11     Pos.12       Pos.13     Pos.14</span></span>
<span><span class="co">## A -0.765021 -0.7310737 -0.6979068 -0.5426286 -1.2504478 -1.765021012 -2.5723759</span></span>
<span><span class="co">## R -1.259011 -1.2055715 -1.1540412 -1.5606664 -1.5606664 -2.179576246 -3.6650031</span></span>
<span><span class="co">## N  0.171651 -0.1424576  0.6480890  0.4809791 -0.5445560 -0.959593538 -1.2644481</span></span>
<span><span class="co">## D  1.417570  1.4175704  1.3318405  1.3020932  0.9650582  0.524485605 -0.9609412</span></span>
<span><span class="co">## C  2.546425  2.0869933  2.2569183  2.3797750  2.0129927  1.519952663  0.5724201</span></span>
<span><span class="co">## Q  1.141629  0.9489834  0.8140539  0.9996095  0.3895560 -0.000390493 -1.2734090</span></span>
<span><span class="co">##      Pos.15    Pos.16    Pos.17    Pos.18    Pos.19    Pos.20</span></span>
<span><span class="co">## A -6.157338 -6.157338 -6.157338 -6.157338 -6.157338 -6.157338</span></span>
<span><span class="co">## R -6.986931 -6.986931 -6.986931 -6.986931 -6.986931 -6.986931</span></span>
<span><span class="co">## N -5.351911 -5.351911 -5.351911 -5.351911 -5.351911 -5.351911</span></span>
<span><span class="co">## D -4.282869 -4.282869 -4.282869 -4.282869 -4.282869 -4.282869</span></span>
<span><span class="co">## C -3.234935 -3.234935 -3.234935 -3.234935 -3.234935 -3.234935</span></span>
<span><span class="co">## Q -4.858371 -4.858371 -4.858371 -4.858371 -4.858371 -4.858371</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="adjacencymatrix">adjacencyMatrix<a class="anchor" aria-label="anchor" href="#adjacencymatrix"></a>
</h3>
<p>Similar to the positional probability, we can also summarize a given
set of sequences by the frequency of adjacency for a given set of amino
acid or nucleotide residues using <code>adjacencyMatrix()</code>. For
this function, a matrix of n x n (defined by the length of
<strong>sequence.dictionary</strong>) is created and the number of times
a residue is adjacent to one another is calculated. We can
<strong>normalize</strong> the values using the total number of residues
evaluated.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">adj.matrix</span> <span class="op">&lt;-</span> <span class="fu">adjacencyMatrix</span><span class="op">(</span><span class="va">sequences</span>, </span>
<span>                              normalize <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">adj.matrix</span></span></code></pre></div>
<pre><code><span><span class="co">##      A  R  N  D    C   Q   E   G  H  I   L  K  M    F   P    S   T   W    Y  V</span></span>
<span><span class="co">## A   32 39 35 24 1029  49  38  36 31 40  43 40 36   35  31 1090  40  40   89 45</span></span>
<span><span class="co">## R   39 36 39 38   54  46  42  58 41 35  64 52 39   36  41   89  45  39   80 37</span></span>
<span><span class="co">## N   35 39 54 42   37  29  33  27 35 31  48 40 38   52  44   95  32  36   83 46</span></span>
<span><span class="co">## D   24 38 42 42   47  38  37  48 44 47  44 42 36   42  40   97  36  50   81 41</span></span>
<span><span class="co">## C 1029 54 37 47   38  58  43  45 38 45  25 38 47   56  42   88  41  48  101 36</span></span>
<span><span class="co">## Q   49 46 29 38   58  40  42  48 46 30  49 47 48   45  45   88  51  47  119 47</span></span>
<span><span class="co">## E   38 42 33 37   43  42  46  40 43 47  46 52 49   37  49  109  46  41   85 35</span></span>
<span><span class="co">## G   36 58 27 48   45  48  40  42 48 38  43 52 46   44  37   79  54  36  117 56</span></span>
<span><span class="co">## H   31 41 35 44   38  46  43  48 34 45  43 43 31   48  51   83  56  35   81 36</span></span>
<span><span class="co">## I   40 35 31 47   45  30  47  38 45 38  36 50 55   44  50   99  52  41   97 30</span></span>
<span><span class="co">## L   43 64 48 44   25  49  46  43 43 36  60 32 46   41  60  115  38  37   96 52</span></span>
<span><span class="co">## K   40 52 40 42   38  47  52  52 43 50  32 40 36   53  44   86  42  37   88 44</span></span>
<span><span class="co">## M   36 39 38 36   47  48  49  46 31 55  46 36 36   36  39   89  39  35   87 44</span></span>
<span><span class="co">## F   35 36 52 42   56  45  37  44 48 44  41 53 36   36  41  101  41  39 1079 42</span></span>
<span><span class="co">## P   31 41 44 40   42  45  49  37 51 50  60 44 39   41  36  109  48  44   94 45</span></span>
<span><span class="co">## S 1090 89 95 97   88  88 109  79 83 99 115 86 89  101 109  122  96 105  135 69</span></span>
<span><span class="co">## T   40 45 32 36   41  51  46  54 56 52  38 42 39   41  48   96  54  45  107 39</span></span>
<span><span class="co">## W   40 39 36 50   48  47  41  36 35 41  37 37 35   39  44  105  45  48   88 53</span></span>
<span><span class="co">## Y   89 80 83 81  101 119  85 117 81 97  96 88 87 1079  94  135 107  88  128 85</span></span>
<span><span class="co">## V   45 37 46 41   36  47  35  56 36 30  52 44 44   42  45   69  39  53   85 56</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="extracting-sequences">Extracting Sequences<a class="anchor" aria-label="anchor" href="#extracting-sequences"></a>
</h2>
<div class="section level3">
<h3 id="sequencedecoder">sequenceDecoder<a class="anchor" aria-label="anchor" href="#sequencedecoder"></a>
</h3>
<p>We have a function called <code>sequenceDecoder()</code> that
extracts sequences from one-hot or property-encoded matrices or arrays.
This function can be applied to any generative approach to sequence
generation.</p>
<p>Parameters for <code>sequenceDecoder()</code></p>
<ul>
<li>
<strong>sequence.matrix</strong> The encoded sequences to decode in
an array or matrix</li>
<li>
<strong>encoder</strong> The method to prepare the sequencing
information - “onehotEncoder” or “propertyEncoder”</li>
<li>
<strong>aa.method.to.use</strong> The method or approach to use for
the conversion corresponding to the input to
<code>propertyEncoder()</code>. This will be ignored if
<strong>encoder</strong> = “onehotEncoder”</li>
<li>
<strong>call.threshold</strong> The relative strictness of sequence
calling with higher values being more stringent</li>
</ul>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">property.matrix</span> <span class="op">&lt;-</span> <span class="fu">propertyEncoder</span><span class="op">(</span>input.sequences <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">sequences</span>, <span class="va">mutated.sequences</span><span class="op">)</span>, method.to.use <span class="op">=</span> <span class="st">"FASGAI"</span>,</span>
<span>    convert.to.matrix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Padding sequences..."</span></span>
<span><span class="co">## [1] "Property-based Encoding sequences..."</span></span>
<span><span class="co">## [1] "Preparing a matrix..."</span></span></code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">property.sequences</span> <span class="op">&lt;-</span> <span class="fu">sequenceDecoder</span><span class="op">(</span><span class="va">property.matrix</span>, encoder <span class="op">=</span> <span class="st">"propertyEncoder"</span>, aa.method.to.use <span class="op">=</span> <span class="st">"FASGAI"</span>,</span>
<span>    call.threshold <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sequences</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CASFKRRPLNFYF"    "CASDGLNSMSFTKYF"  "CASVVDNYVKHHFKYF" "CASMEHTRMYDRTYF" </span></span>
<span><span class="co">## [5] "CASPFWKERYRYF"    "CASMRIGYMRAGYF"</span></span></code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">property.sequences</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CASFKRRPLNFYF"    "CASDGLNSMSFTKYF"  "CASVVDNYVKHHFKYF" "CASMEHTRMYDRTYF" </span></span>
<span><span class="co">## [5] "CASPFWKERYRYF"    "CASMRIGYMRAGYF"</span></span></code></pre>
<p>A similar approach can be applied when using matrices or arrays
derived from one-hot encoding:</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sequence.matrix</span> <span class="op">&lt;-</span> <span class="fu">onehotEncoder</span><span class="op">(</span>input.sequences <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">sequences</span>, <span class="va">mutated.sequences</span><span class="op">)</span>, convert.to.matrix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Padding sequences..."</span></span>
<span><span class="co">## [1] "One Hot Encoding sequences..."</span></span>
<span><span class="co">## [1] "Preparing a matrix..."</span></span></code></pre>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">OHE.sequences</span> <span class="op">&lt;-</span> <span class="fu">sequenceDecoder</span><span class="op">(</span><span class="va">sequence.matrix</span>, encoder <span class="op">=</span> <span class="st">"onehotEncoder"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">OHE.sequences</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CASFKRRPLNFYF"    "CASDGLNSMSFTKYF"  "CASVVDNYVKHHFKYF" "CASMEHTRMYDRTYF" </span></span>
<span><span class="co">## [5] "CASPFWKERYRYF"    "CASMRIGYMRAGYF"</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="training-a-model">Training a Model<a class="anchor" aria-label="anchor" href="#training-a-model"></a>
</h2>
<div class="section level3">
<h3 id="autoencoder">Autoencoder<a class="anchor" aria-label="anchor" href="#autoencoder"></a>
</h3>
<p>For the vignette - we will use an autoencoder for sequence embedding.
The code below is based on the <a href="https://github.com/ncborcherding/Trex" class="external-link">Trex</a> R package. The
overall structure of the autoencoder is the same. However, some of the
parameters are modified for the sake of the vignette. We will use the
<strong>sequence.matrix</strong> we generated above from the
<code>onehotEncoder()</code>.</p>
<p>The steps to train the model include:</p>
<ol style="list-style-type: decimal">
<li>Subsetting sequences</li>
<li>Defining parameters for the model</li>
<li>Forming the autoencoder structure - encoder and decoder</li>
<li>Fitting the model</li>
</ol>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Sampling to make Training/Valid Data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="va">num_sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">sequence.matrix</span><span class="op">)</span></span>
<span><span class="va">indices</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="va">num_sequences</span></span>
<span><span class="va">train_indices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">indices</span>, size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="fl">0.8</span> <span class="op">*</span> <span class="va">num_sequences</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">val_indices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="va">indices</span>, <span class="va">train_indices</span><span class="op">)</span></span>
<span>    </span>
<span><span class="va">x_train</span> <span class="op">&lt;-</span> <span class="va">sequence.matrix</span><span class="op">[</span><span class="va">train_indices</span>,<span class="op">]</span></span>
<span><span class="va">x_val</span> <span class="op">&lt;-</span> <span class="va">sequence.matrix</span><span class="op">[</span><span class="va">val_indices</span>,<span class="op">]</span></span>
<span>   </span>
<span><span class="co"># Parameters</span></span>
<span><span class="va">input_shape</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">x_train</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">epochs</span> <span class="op">&lt;-</span> <span class="fl">20</span></span>
<span><span class="va">batch_size</span> <span class="op">&lt;-</span> <span class="fl">128</span></span>
<span><span class="va">encoding_dim</span> <span class="op">&lt;-</span> <span class="fl">40</span> </span>
<span><span class="va">hidden_dim1</span> <span class="op">&lt;-</span> <span class="fl">256</span> <span class="co"># Hidden layer 1 size</span></span>
<span><span class="va">hidden_dim2</span> <span class="op">&lt;-</span> <span class="fl">128</span>  <span class="co"># Hidden layer 2 size</span></span>
<span>    </span>
<span><span class="va">es</span> <span class="op">=</span> <span class="fu">callback_early_stopping</span><span class="op">(</span>monitor <span class="op">=</span> <span class="st">"val_loss"</span>,</span>
<span>                             min_delta <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                             patience <span class="op">=</span> <span class="fl">4</span>,</span>
<span>                             verbose <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                             mode <span class="op">=</span> <span class="st">"min"</span><span class="op">)</span></span>
<span>                    </span>
<span><span class="co"># Define the Model</span></span>
<span><span class="va">input_seq</span> <span class="op">&lt;-</span> <span class="fu">layer_input</span><span class="op">(</span>shape <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">input_shape</span><span class="op">)</span><span class="op">)</span></span>
<span>        </span>
<span><span class="co"># Encoder Layers</span></span>
<span><span class="va">encoded</span> <span class="op">&lt;-</span> <span class="va">input_seq</span> <span class="op">%&gt;%</span></span>
<span>          <span class="fu">layer_dense</span><span class="op">(</span>units <span class="op">=</span> <span class="va">hidden_dim1</span>, name <span class="op">=</span> <span class="st">"e.1"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>          <span class="fu">layer_batch_normalization</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"bn.1"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>          <span class="fu">layer_activation</span><span class="op">(</span><span class="st">'leaky_relu'</span>, name <span class="op">=</span> <span class="st">"act.1"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>          <span class="fu">layer_dense</span><span class="op">(</span>units <span class="op">=</span> <span class="va">hidden_dim2</span>, name <span class="op">=</span> <span class="st">"e.2"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>          <span class="fu">layer_batch_normalization</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"bn.2"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>          <span class="fu">layer_activation</span><span class="op">(</span><span class="st">'leaky_relu'</span>, name <span class="op">=</span> <span class="st">"act.2"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>          <span class="fu">layer_dense</span><span class="op">(</span>units <span class="op">=</span> <span class="va">encoding_dim</span>, activation <span class="op">=</span> <span class="st">'selu'</span>, name <span class="op">=</span> <span class="st">"latent"</span><span class="op">)</span></span>
<span>                </span>
<span><span class="co"># Decoder Layers</span></span>
<span><span class="va">decoded</span> <span class="op">&lt;-</span> <span class="va">encoded</span> <span class="op">%&gt;%</span></span>
<span>          <span class="fu">layer_dense</span><span class="op">(</span>units <span class="op">=</span> <span class="va">hidden_dim2</span>, name <span class="op">=</span> <span class="st">"d.2"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>          <span class="fu">layer_batch_normalization</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"bn.3"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>          <span class="fu">layer_activation</span><span class="op">(</span><span class="st">'leaky_relu'</span>, name <span class="op">=</span> <span class="st">"act.3"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>          <span class="fu">layer_dense</span><span class="op">(</span>units <span class="op">=</span> <span class="va">hidden_dim1</span>, name <span class="op">=</span> <span class="st">"d.1"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>          <span class="fu">layer_batch_normalization</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"bn.4"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>          <span class="fu">layer_activation</span><span class="op">(</span><span class="st">'leaky_relu'</span>, name <span class="op">=</span> <span class="st">"act.4"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>          <span class="fu">layer_dense</span><span class="op">(</span>units <span class="op">=</span> <span class="va">input_shape</span>, activation <span class="op">=</span> <span class="st">'sigmoid'</span><span class="op">)</span></span>
<span>      </span>
<span><span class="co"># Autoencoder Model</span></span>
<span><span class="va">autoencoder</span> <span class="op">&lt;-</span> <span class="fu">keras_model</span><span class="op">(</span><span class="va">input_seq</span>, <span class="va">decoded</span><span class="op">)</span></span>
<span><span class="va">autoencoder</span> <span class="op">%&gt;%</span> <span class="fu">keras</span><span class="fu">::</span><span class="fu"><a href="https://generics.r-lib.org/reference/compile.html" class="external-link">compile</a></span><span class="op">(</span>optimizer <span class="op">=</span> <span class="fu">optimizer_adam</span><span class="op">(</span>learning_rate <span class="op">=</span> <span class="fl">0.0001</span><span class="op">)</span>,</span>
<span>                                   loss <span class="op">=</span> <span class="st">"mse"</span><span class="op">)</span></span>
<span>      </span>
<span><span class="co"># Train the model</span></span>
<span><span class="va">history</span> <span class="op">&lt;-</span> <span class="va">autoencoder</span> <span class="op">%&gt;%</span> <span class="fu">fit</span><span class="op">(</span>x <span class="op">=</span> <span class="va">x_train</span>,</span>
<span>                               y <span class="op">=</span> <span class="va">x_train</span>,</span>
<span>                               validation_data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">x_val</span>, <span class="va">x_val</span><span class="op">)</span>,</span>
<span>                               epochs <span class="op">=</span> <span class="va">epochs</span>,</span>
<span>                               batch_size <span class="op">=</span> <span class="va">batch_size</span>,</span>
<span>                               shuffle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                               callbacks <span class="op">=</span> <span class="va">es</span>, </span>
<span>                               verbose <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">history</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">scale_color_viridis</span><span class="op">(</span>option <span class="op">=</span> <span class="st">"B"</span>, discrete <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">scale_fill_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="immApex_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="classifier">Classifier<a class="anchor" aria-label="anchor" href="#classifier"></a>
</h3>
<p>We can also build classifiers directly using deep or shallow neural
networks. Building deep classifiers requires more data than classical
machine learning methods, like random forests, so the vignette may not
be ideal.</p>
<p>The first step is to generate distinct types of sequences
using <code>generateSequences()</code> and <code>onehotEncoder()</code> to
prepare the data for the model.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">class1.sequences</span> <span class="op">&lt;-</span> <span class="fu">generateSequences</span><span class="op">(</span>prefix.motif <span class="op">=</span> <span class="st">"CAS"</span>,</span>
<span>                                      suffix.motif <span class="op">=</span> <span class="st">"YF"</span>,</span>
<span>                                      number.of.sequences <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>                                      min.length <span class="op">=</span> <span class="fl">8</span>,</span>
<span>                                      max.length <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span>
<span></span>
<span><span class="va">class2.sequences</span> <span class="op">&lt;-</span> <span class="fu">generateSequences</span><span class="op">(</span>prefix.motif <span class="op">=</span> <span class="st">"CASF"</span>,</span>
<span>                                      suffix.motif <span class="op">=</span> <span class="st">"YF"</span>,</span>
<span>                                      number.of.sequences <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>                                      min.length <span class="op">=</span> <span class="fl">8</span>,</span>
<span>                                      max.length <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span>
<span></span>
<span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">10000</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">10000</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">classifier.matrix</span> <span class="op">&lt;-</span> <span class="fu">onehotEncoder</span><span class="op">(</span>input.sequences <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">class1.sequences</span>, <span class="va">class2.sequences</span><span class="op">)</span>, </span>
<span>                                   convert.to.matrix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Padding sequences..."</span></span>
<span><span class="co">## [1] "One Hot Encoding sequences..."</span></span>
<span><span class="co">## [1] "Preparing a matrix..."</span></span></code></pre>
<p>Next, we will define and train the Keras classifier model using
artificial sequences. We will use a simple convolutional neural network
with 2 layers and then a single neuron that will classify the sequences
into class 1 or class 2 (here, the labels are 0 and 1).</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Input shape will be 1D as we are using a matrix</span></span>
<span><span class="va">input.shape</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">classifier.matrix</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#Simple model structure</span></span>
<span><span class="va">classifier.model</span> <span class="op">&lt;-</span> <span class="fu">keras_model_sequential</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>                        <span class="fu">layer_dense</span><span class="op">(</span>units <span class="op">=</span> <span class="fl">128</span>, activation <span class="op">=</span> <span class="st">"relu"</span>, </span>
<span>                                    input_shape <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">input.shape</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>                        <span class="fu">layer_dense</span><span class="op">(</span>units <span class="op">=</span> <span class="fl">32</span>, activation <span class="op">=</span> <span class="st">"relu"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>                        <span class="fu">layer_dense</span><span class="op">(</span>units <span class="op">=</span> <span class="fl">1</span>, activation <span class="op">=</span> <span class="st">"sigmoid"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">classifier.model</span> <span class="op">%&gt;%</span> <span class="fu">compile</span><span class="op">(</span></span>
<span>        optimizer <span class="op">=</span> <span class="fu">optimizer_adam</span><span class="op">(</span>learning_rate <span class="op">=</span> <span class="fl">0.00001</span><span class="op">)</span>,</span>
<span>        loss <span class="op">=</span> <span class="st">"binary_crossentropy"</span>,</span>
<span>        metrics <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"accuracy"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Seperating data and labels</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="va">val_indices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">classifier.matrix</span><span class="op">)</span>, <span class="fl">10000</span><span class="op">*</span><span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="va">x_val</span> <span class="op">&lt;-</span> <span class="va">classifier.matrix</span><span class="op">[</span><span class="va">val_indices</span>,<span class="op">]</span></span>
<span><span class="va">x_train</span> <span class="op">&lt;-</span> <span class="va">classifier.matrix</span><span class="op">[</span><span class="op">-</span><span class="va">val_indices</span>,<span class="op">]</span></span>
<span></span>
<span><span class="va">val_labels</span> <span class="op">&lt;-</span> <span class="va">labels</span><span class="op">[</span><span class="va">val_indices</span><span class="op">]</span></span>
<span><span class="va">train_labels</span> <span class="op">&lt;-</span> <span class="va">labels</span><span class="op">[</span><span class="op">-</span><span class="va">val_indices</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#Training the classifier.model</span></span>
<span><span class="va">history</span> <span class="op">&lt;-</span> <span class="va">classifier.model</span> <span class="op">%&gt;%</span> <span class="fu">fit</span><span class="op">(</span><span class="va">x_train</span>, </span>
<span>                                    <span class="va">train_labels</span>, </span>
<span>                                    epochs <span class="op">=</span> <span class="fl">20</span>, </span>
<span>                                    batch_size <span class="op">=</span> <span class="fl">32</span>, </span>
<span>                                    validation_data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">x_val</span>, <span class="va">val_labels</span><span class="op">)</span>,</span>
<span>                                    verbose <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">history</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">scale_color_viridis</span><span class="op">(</span>option <span class="op">=</span> <span class="st">"B"</span>, discrete <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">scale_fill_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="immApex_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<p>Here, we can achieve a validation accuracy of 98.25%, which is
impressive. But to contextualize, we used
<code>generateSequences()</code> and distinct motifs - “CAS” vs “CASF”
to create our 2 classes of sequences. Using sequences from experimental
data will likely result in lower accuracy or require greater model
complexity.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>This has been a general overview of the capabilities of immApex for
processing immune receptor sequences and making deep learning models. If
you have any questions, comments, or suggestions, feel free to visit the
<a href="https://github.com/ncborcherding/immApex" class="external-link">GitHub
repository</a>.</p>
<div class="section level3">
<h3 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h3>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">## Running under: macOS 14.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/Chicago</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] dplyr_1.1.4       viridis_0.6.5     viridisLite_0.4.2 ggplot2_3.5.1    </span></span>
<span><span class="co">## [5] keras_2.15.0      immApex_0.99.0   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] tidyselect_1.2.1            farver_2.1.2               </span></span>
<span><span class="co">##  [3] tensorflow_2.16.0           bitops_1.0-8               </span></span>
<span><span class="co">##  [5] fastmap_1.2.0               SingleCellExperiment_1.24.0</span></span>
<span><span class="co">##  [7] RCurl_1.98-1.16             digest_0.6.36              </span></span>
<span><span class="co">##  [9] lifecycle_1.0.4             magrittr_2.0.3             </span></span>
<span><span class="co">## [11] compiler_4.3.0              rlang_1.1.4                </span></span>
<span><span class="co">## [13] sass_0.4.9                  tools_4.3.0                </span></span>
<span><span class="co">## [15] utf8_1.2.4                  yaml_2.3.10                </span></span>
<span><span class="co">## [17] knitr_1.48                  labeling_0.4.3             </span></span>
<span><span class="co">## [19] S4Arrays_1.2.0              htmlwidgets_1.6.4          </span></span>
<span><span class="co">## [21] curl_5.2.1                  reticulate_1.37.0          </span></span>
<span><span class="co">## [23] here_1.0.1                  DelayedArray_0.28.0        </span></span>
<span><span class="co">## [25] xml2_1.3.6                  abind_1.4-5                </span></span>
<span><span class="co">## [27] withr_3.0.1                 purrr_1.0.2                </span></span>
<span><span class="co">## [29] BiocGenerics_0.48.1         desc_1.4.3                 </span></span>
<span><span class="co">## [31] grid_4.3.0                  hash_2.2.6.3               </span></span>
<span><span class="co">## [33] stats4_4.3.0                fansi_1.0.6                </span></span>
<span><span class="co">## [35] colorspace_2.1-1            scales_1.3.0               </span></span>
<span><span class="co">## [37] zeallot_0.1.0               SummarizedExperiment_1.32.0</span></span>
<span><span class="co">## [39] cli_3.6.3                   rmarkdown_2.27             </span></span>
<span><span class="co">## [41] crayon_1.5.3                ragg_1.3.2                 </span></span>
<span><span class="co">## [43] generics_0.1.3              rstudioapi_0.16.0          </span></span>
<span><span class="co">## [45] httr_1.4.7                  tfruns_1.5.3               </span></span>
<span><span class="co">## [47] cachem_1.1.0                stringr_1.5.1              </span></span>
<span><span class="co">## [49] splines_4.3.0               zlibbioc_1.48.0            </span></span>
<span><span class="co">## [51] rvest_1.0.4                 formatR_1.14               </span></span>
<span><span class="co">## [53] selectr_0.4-2               XVector_0.42.0             </span></span>
<span><span class="co">## [55] matrixStats_1.3.0           base64enc_0.1-3            </span></span>
<span><span class="co">## [57] vctrs_0.6.5                 Matrix_1.6-5               </span></span>
<span><span class="co">## [59] jsonlite_1.8.8              IRanges_2.36.0             </span></span>
<span><span class="co">## [61] S4Vectors_0.40.2            systemfonts_1.1.0          </span></span>
<span><span class="co">## [63] jquerylib_0.1.4             glue_1.7.0                 </span></span>
<span><span class="co">## [65] pkgdown_2.0.9               stringi_1.8.4              </span></span>
<span><span class="co">## [67] gtable_0.3.5                GenomeInfoDb_1.38.6        </span></span>
<span><span class="co">## [69] GenomicRanges_1.54.1        munsell_0.5.1              </span></span>
<span><span class="co">## [71] tibble_3.2.1                pillar_1.9.0               </span></span>
<span><span class="co">## [73] htmltools_0.5.8.1           GenomeInfoDbData_1.2.11    </span></span>
<span><span class="co">## [75] R6_2.5.1                    textshaping_0.4.0          </span></span>
<span><span class="co">## [77] rprojroot_2.0.4             evaluate_0.24.0            </span></span>
<span><span class="co">## [79] lattice_0.22-6              Biobase_2.62.0             </span></span>
<span><span class="co">## [81] highr_0.11                  png_0.1-8                  </span></span>
<span><span class="co">## [83] memoise_2.0.1               bslib_0.8.0                </span></span>
<span><span class="co">## [85] Rcpp_1.0.13                 nlme_3.1-165               </span></span>
<span><span class="co">## [87] gridExtra_2.3               SparseArray_1.2.4          </span></span>
<span><span class="co">## [89] mgcv_1.9-1                  whisker_0.4.1              </span></span>
<span><span class="co">## [91] xfun_0.46                   fs_1.6.4                   </span></span>
<span><span class="co">## [93] MatrixGenerics_1.14.0       pkgconfig_2.0.3</span></span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
