<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Making Deep Learning Models with immApex • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Making Deep Learning Models with immApex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/dandelionR.html">VDJ Trajectory Analysis with dandelionR</a>
    </li>
    <li>
      <a href="../articles/immunarch.html">Using scRepertoire with immunarch</a>
    </li>
    <li>
      <a href="../articles/tcrpheno.html">TCRpheno applied to T cell fate</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BorchLab/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Making Deep Learning Models with immApex</h1>
            
            <h4 data-toc-skip class="date">Compiled: July 17, 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BorchLab/scRepertoire/blob/HEAD/vignettes/articles/immApex.Rmd" class="external-link"><code>vignettes/articles/immApex.Rmd</code></a></small>
      <div class="hidden name"><code>immApex.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="getting-started">Getting Started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p><strong>immApex</strong> is meant to serve as an API for supervised
and unsupervised learning tasks based on immune receptor sequencing.
These functions extract or generate amino acid or nucleotide sequences
and prepare them for downstream machine learning.
<strong>immApex</strong> is the underlying structure for the BCR models
in <a href="https://github.com/BorchLab/Ibex" class="external-link">Ibex</a> and TCR models in
<a href="https://github.com/BorchLab/Trex" class="external-link">Trex</a>. It should be noted
that the tools here are created for immune receptor sequences; they will
work more generally for nucleotide or amino acid sequences. The package
itself supports AIRR, Adaptive, and 10x formats and interacts with the
<strong>scRepertoire</strong> R package.</p>
<p>More information is available at the <a href="https://github.com/BorchLab/immApex" class="external-link">immApex GitHub Repo</a>.</p>
<div class="section level3">
<h3 id="loading-libraries">Loading Libraries<a class="anchor" aria-label="anchor" href="#loading-libraries"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BorchLab/immApex/" class="external-link">immApex</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sjmgarnier.github.io/viridis/" class="external-link">viridis</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r.igraph.org/" class="external-link">igraph</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidygraph.data-imaginist.com" class="external-link">tidygraph</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggraph.data-imaginist.com" class="external-link">ggraph</a></span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="acquiring-and-preparing-repertoire-data">Acquiring and Preparing Repertoire Data<a class="anchor" aria-label="anchor" href="#acquiring-and-preparing-repertoire-data"></a>
</h2>
<div class="section level3">
<h3 id="getimgt">getIMGT<a class="anchor" aria-label="anchor" href="#getimgt"></a>
</h3>
<p>Depending on the sequencing technology and the version, we might want
to expand the length of our sequence embedding approach. The first step
in the process is pulling the reference sequences from the
ImMunoGeneTics (IMGT) system using <code><a href="https://rdrr.io/pkg/immApex/man/getIMGT.html" class="external-link">getIMGT()</a></code>. More
information for IMGT can be found at <a href="https://www.imgt.org/" class="external-link">imgt.org</a>. Data from IMGT is under a CC
BY-NC-ND 4.0 license. Please be aware that attribution is required for
usage and should not be used to create commercial or derivative
work.</p>
<p>Parameters for <code><a href="https://rdrr.io/pkg/immApex/man/getIMGT.html" class="external-link">getIMGT()</a></code></p>
<ul>
<li>
<strong>species</strong> One or two word designation of species.
Currently supporting: “human”, “mouse”, “rat”, “rabbit”, “rhesus
monkey”, “sheep”, “pig”, “platypus”, “alpaca”, “dog”, “chicken”, and
“ferret”</li>
<li>
<strong>chain</strong> Sequence chain to access</li>
<li>
<strong>frame</strong> Designation for “all”, “inframe” or
“inframe+gap”</li>
<li>
<strong>region</strong> Sequence gene loci to access</li>
<li>
<strong>sequence.type</strong> Type of sequence - “aa” for amino
acid or “nt” for nucleotide</li>
</ul>
<p>Here, we will use the <code><a href="https://rdrr.io/pkg/immApex/man/getIMGT.html" class="external-link">getIMGT()</a></code> function to get the
amino acid sequences for the TRBV region to get all the sequences by V
gene allele.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">TRBV_aa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/getIMGT.html" class="external-link">getIMGT</a></span><span class="op">(</span>species <span class="op">=</span> <span class="st">"human"</span>, chain <span class="op">=</span> <span class="st">"TRB"</span>, frame <span class="op">=</span> <span class="st">"inframe"</span>, region <span class="op">=</span> <span class="st">"v"</span>, sequence.type <span class="op">=</span> <span class="st">"aa"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">TRBV_aa</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## $`TRBV1*01`</span></span>
<span><span class="co">## [1] "TRBVHPVREGIONAADTGITQTPKYLVTAMGSKRTMKREHLGHDSMYWYRQKAKKSLEFMFYYNCKEFIENKTVPNHFTPECPDSSRLYLHVVALQQEDSAAYLCTSSQ"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="formatgenes">formatGenes<a class="anchor" aria-label="anchor" href="#formatgenes"></a>
</h3>
<p>Immune receptor nomenclature can be highly variable across sequencing
platforms. When preparing data for models, we can use
<code><a href="https://rdrr.io/pkg/immApex/man/formatGenes.html" class="external-link">formatGenes()</a></code> to universalize the gene formats into IMGT
nomenclature.</p>
<p>Parameters for <code><a href="https://rdrr.io/pkg/immApex/man/formatGenes.html" class="external-link">formatGenes()</a></code></p>
<ul>
<li>
<strong>input.data</strong> Data frame of sequencing data or
scRepertoire outputs</li>
<li>
<strong>region</strong> Sequence gene loci to access - ‘v’, ‘d’,
‘j’, ‘c’ or a combination using c(‘v’, ‘d’, ‘j’)</li>
<li>
<strong>technology</strong> The sequencing technology employed -
‘TenX’, “Adaptive’, or ‘AIRR’</li>
<li>
<strong>species</strong> One or two word designation of species.
Currently supporting: “human”, “mouse”, “rat”, “rabbit”, “rhesus
monkey”, “sheep”, “pig”, “platypus”, “alpaca”, “dog”, “chicken”, and
“ferret”</li>
<li>
<strong>simplify.format</strong> If applicable, remove the allelic
designation (TRUE) or retain all information (FALSE)</li>
</ul>
<p>Here, we will use the built-in example from Adaptive Biotechnologies
and reformat and simplify the <strong>v</strong> region.
<code><a href="https://rdrr.io/pkg/immApex/man/formatGenes.html" class="external-link">formatGenes()</a></code> will add 2 columns to the end of the data
frame per region selected - 1) <strong>v_IMGT</strong> will be the
formatted gene calls and 2) <strong>v_IMGT.check</strong> is a binary
for if the formatted region appears in the IMGT database. In the example
below, “TRBV2-1” is not recognized as a designation within IMGT.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"immapex_example.data"</span><span class="op">)</span></span>
<span><span class="va">Adaptive_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/formatGenes.html" class="external-link">formatGenes</a></span><span class="op">(</span><span class="va">immapex_example.data</span><span class="op">[[</span><span class="st">"Adaptive"</span><span class="op">]</span><span class="op">]</span>, region <span class="op">=</span> <span class="st">"v"</span>, technology <span class="op">=</span> <span class="st">"Adaptive"</span>,</span>
<span>    simplify.format <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Adaptive_example</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"aminoAcid"</span>, <span class="st">"vGeneName"</span>, <span class="st">"v_IMGT"</span>, <span class="st">"v_IMGT.check"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##              aminoAcid  vGeneName  v_IMGT v_IMGT.check</span></span>
<span><span class="co">## 4490  CASSQDGPSGIETQYF TCRBV04-02 TRBV4-2            1</span></span>
<span><span class="co">## 18266    CASSEGSNQPQHF TCRBV02-01 TRBV2-1            0</span></span>
<span><span class="co">## 22061   CSASAGDMVTEAFF       &lt;NA&gt;    &lt;NA&gt;            0</span></span>
<span><span class="co">## 22174  CASSQDPGETDTQYF       &lt;NA&gt;    &lt;NA&gt;            0</span></span>
<span><span class="co">## 19117     CATSAWTGELFF       &lt;NA&gt;    &lt;NA&gt;            0</span></span>
<span><span class="co">## 2659     CATSVPGQETQYF       &lt;NA&gt;    &lt;NA&gt;            0</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="infercdr">inferCDR<a class="anchor" aria-label="anchor" href="#infercdr"></a>
</h3>
<p>We can now use <code><a href="https://rdrr.io/pkg/immApex/man/inferCDR.html" class="external-link">inferCDR()</a></code> to add additional sequence
elements to our example data using the outputs of
<code><a href="https://rdrr.io/pkg/immApex/man/formatGenes.html" class="external-link">formatGenes()</a></code> and <code><a href="https://rdrr.io/pkg/immApex/man/getIMGT.html" class="external-link">getIMGT()</a></code>. Here, we will use
the function to isolate the complementarity-determining regions (CDR) 1
and 2. If the gene nomenclature does not match the IMGT the result will
be NA for the given sequences. Likewise, if the IMGT nomenclature has
been simplified, the first allelic match will be used for sequence
extraction.</p>
<p>Parameters for <code>inferCDR</code></p>
<ul>
<li>
<strong>input.data</strong> Data frame of sequencing data or output
from formatGenes().</li>
<li>
<strong>reference</strong> IMGT sequences from
<code><a href="https://rdrr.io/pkg/immApex/man/getIMGT.html" class="external-link">getIMGT()</a></code>
</li>
<li>
<strong>technology</strong> The sequencing technology employed -
‘TenX’, “Adaptive’, or ‘AIRR’,</li>
<li>
<strong>sequence.type</strong> Type of sequence - “aa” for amino
acid or “nt” for nucleotide</li>
<li>
<strong>sequences</strong> The specific regions of the CDR loop to
get from the data.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Adaptive_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/inferCDR.html" class="external-link">inferCDR</a></span><span class="op">(</span><span class="va">Adaptive_example</span>, chain <span class="op">=</span> <span class="st">"TRB"</span>, reference <span class="op">=</span> <span class="va">TRBV_aa</span>, technology <span class="op">=</span> <span class="st">"Adaptive"</span>,</span>
<span>    sequence.type <span class="op">=</span> <span class="st">"aa"</span>, sequences <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CDR1"</span>, <span class="st">"CDR2"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">Adaptive_example</span><span class="op">[</span><span class="fl">200</span><span class="op">:</span><span class="fl">210</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CDR1_IMGT"</span>, <span class="st">"CDR2_IMGT"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##          CDR1_IMGT  CDR2_IMGT</span></span>
<span><span class="co">## 27450         &lt;NA&gt;       &lt;NA&gt;</span></span>
<span><span class="co">## 14624         &lt;NA&gt;       &lt;NA&gt;</span></span>
<span><span class="co">## 10452         &lt;NA&gt;       &lt;NA&gt;</span></span>
<span><span class="co">## 31207 IIEKRQSVAFWC QGPKLLIQFQ</span></span>
<span><span class="co">## 23216 IKTRGQQVTLSC QGLQFLFEYF</span></span>
<span><span class="co">## 3356  VMGMTNKKSLKC KPPELMFVYS</span></span>
<span><span class="co">## 25030         &lt;NA&gt;       &lt;NA&gt;</span></span>
<span><span class="co">## 1889  VMGMTNKKSLKC KPLELMFVYN</span></span>
<span><span class="co">## 31150 VTEKGKDVELRC QGLEFLIYFQ</span></span>
<span><span class="co">## 20759         &lt;NA&gt;       &lt;NA&gt;</span></span>
<span><span class="co">## 13865 VMGMTNKKSLKC KPPELMFVYS</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="generating-and-augmenting-sequence-sets">Generating and Augmenting Sequence Sets<a class="anchor" aria-label="anchor" href="#generating-and-augmenting-sequence-sets"></a>
</h2>
<div class="section level3">
<h3 id="generatesequences">generateSequences<a class="anchor" aria-label="anchor" href="#generatesequences"></a>
</h3>
<p>Generating synthetic sequences is a quick way to start testing the
model code. <code><a href="https://rdrr.io/pkg/immApex/man/generateSequences.html" class="external-link">generateSequences()</a></code> can also generate realistic
noise for generative adversarial networks.</p>
<p>Parameters for <code><a href="https://rdrr.io/pkg/immApex/man/generateSequences.html" class="external-link">generateSequences()</a></code></p>
<ul>
<li>
<strong>prefix.motif</strong> Add a defined sequence to the start of
the generated sequences.</li>
<li>
<strong>suffix.motif</strong> Add a defined sequence to the end of
the generated sequences</li>
<li>
<strong>number.of.sequences</strong> Number of sequences to
generate</li>
<li>
<strong>min.length</strong> Minimum length of the final sequence
(will be adjusted if incongruent with prefix.motif/suffix.motif)</li>
<li>
<strong>max.length</strong> Maximum length of the final
sequence</li>
<li>
<strong>sequence.dictionary</strong> The letters to use in sequence
generation (default are all amino acids)</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/generateSequences.html" class="external-link">generateSequences</a></span><span class="op">(</span>prefix.motif <span class="op">=</span> <span class="st">"CAS"</span>, suffix.motif <span class="op">=</span> <span class="st">"YF"</span>, number.of.sequences <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>    min.length <span class="op">=</span> <span class="fl">8</span>, max.length <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span>
<span><span class="va">sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">sequences</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sequences</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CASHMCYF"         "CASKDINEQTYF"     "CASVIVYF"         "CASKVNWNPKPNVQYF"</span></span>
<span><span class="co">## [5] "CASIQLAINYF"      "CASMCISYF"</span></span></code></pre>
<p>If we want to generate nucleotide sequences instead of amino acids,
we must to change the <strong>sequence.dictionary</strong>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nucleotide.sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/generateSequences.html" class="external-link">generateSequences</a></span><span class="op">(</span>number.of.sequences <span class="op">=</span> <span class="fl">1000</span>, min.length <span class="op">=</span> <span class="fl">8</span>, max.length <span class="op">=</span> <span class="fl">16</span>,</span>
<span>    sequence.dictionary <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"C"</span>, <span class="st">"T"</span>, <span class="st">"G"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">nucleotide.sequences</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "TCGTGAAC"         "ATGCGAAC"         "CCGTGCCACGGGTCAA" "AAGAACGGATAGATT" </span></span>
<span><span class="co">## [5] "CTAGCCTGGCG"      "ACGCCGACAT"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="mutatesequences">mutateSequences<a class="anchor" aria-label="anchor" href="#mutatesequences"></a>
</h3>
<p>A common approach is to mutate sequences randomly or at specific
intervals. This can be particularly helpful if we have fewer sequences
or want to test a model for accuracy given new, altered sequences.
<code><a href="https://rdrr.io/pkg/immApex/man/mutateSequences.html" class="external-link">mutateSequences()</a></code> allows us to tune the type of mutation,
where along the sequences to introduce the mutation and the overall
number of mutations.</p>
<p>Parameters for <code><a href="https://rdrr.io/pkg/immApex/man/mutateSequences.html" class="external-link">mutateSequences()</a></code></p>
<ul>
<li>
<strong>input.sequences</strong> The amino acid or nucleotide
sequences to use</li>
<li>
<strong>number.of.sequences</strong> The number of mutated sequences
to return per input.sequence</li>
<li>
<strong>mutation.rate</strong> The rate of mutations introduced into
sequences</li>
<li>
<strong>position.start</strong> The starting position to mutate
along the sequence. Default NULL will start the random mutations at
position 1.</li>
<li>
<strong>position.end</strong> The ending position to mutate along
the sequence. Default NULL will end the random mutations at the last
position.</li>
<li>
<strong>sequence.dictionary</strong> The letters to use in sequence
mutation (default are all amino acids)</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mutated.sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/mutateSequences.html" class="external-link">mutateSequences</a></span><span class="op">(</span><span class="va">sequences</span>, number.of.sequences <span class="op">=</span> <span class="fl">1</span>, position.start <span class="op">=</span> <span class="fl">3</span>, position.end <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sequences</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CASHMCYF"         "CASKDINEQTYF"     "CASVIVYF"         "CASKVNWNPKPNVQYF"</span></span>
<span><span class="co">## [5] "CASIQLAINYF"      "CASMCISYF"</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mutated.sequences</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           CASHMCYF       CASKDINEQTYF           CASVIVYF   CASKVNWNPKPNVQYF </span></span>
<span><span class="co">##         "CASNMCYF"     "CASKDFNEQTYF"         "CASVIVYT" "CASSVNWNPKPNVQYF" </span></span>
<span><span class="co">##        CASIQLAINYF          CASMCISYF </span></span>
<span><span class="co">##      "CASIQMAINYF"        "CASACISYF"</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="feature-engineering-from-repertoires">Feature Engineering from Repertoires<a class="anchor" aria-label="anchor" href="#feature-engineering-from-repertoires"></a>
</h2>
<p>Beyond encoding single sequences, immApex provides functions to
calculate summary statistics and features across a collection of
sequences, such as a clonal repertoire.</p>
<div class="section level3">
<h3 id="calculatefrequency">calculateFrequency<a class="anchor" aria-label="anchor" href="#calculatefrequency"></a>
</h3>
<p>This function calculates the relative frequency of each residue at
each position across a set of sequences.</p>
<p>Parameters for <code><a href="https://rdrr.io/pkg/immApex/man/calculateFrequency.html" class="external-link">calculateFrequency()</a></code></p>
<ul>
<li>
<strong>input.sequences</strong>: A character vector of
sequences.</li>
<li>
<strong>max.length</strong>: The length to align sequences to by
padding/trimming.</li>
<li>
<strong>tidy</strong>: If TRUE, returns a long-format data frame
suitable for ggplot2.</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">freq.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/calculateFrequency.html" class="external-link">calculateFrequency</a></span><span class="op">(</span><span class="va">sequences</span>, max.length <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">freq.matrix</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   Pos.1 Pos.2 Pos.3      Pos.4      Pos.5      Pos.6      Pos.7      Pos.8</span></span>
<span><span class="co">## A     0     1     0 0.05210421 0.04308617 0.05210421 0.03807615 0.03106212</span></span>
<span><span class="co">## R     0     0     0 0.04208417 0.06412826 0.06012024 0.03707415 0.03306613</span></span>
<span><span class="co">## N     0     0     0 0.03907816 0.04208417 0.05911824 0.04809619 0.03306613</span></span>
<span><span class="co">## D     0     0     0 0.04809619 0.05210421 0.05611222 0.04208417 0.03306613</span></span>
<span><span class="co">## C     1     0     0 0.05010020 0.05410822 0.04609218 0.04008016 0.03807615</span></span>
<span><span class="co">## Q     0     0     0 0.04108216 0.05811623 0.05811623 0.04308617 0.02905812</span></span>
<span><span class="co">##        Pos.9     Pos.10</span></span>
<span><span class="co">## A 0.03306613 0.02605210</span></span>
<span><span class="co">## R 0.03006012 0.02805611</span></span>
<span><span class="co">## N 0.03206413 0.02404810</span></span>
<span><span class="co">## D 0.03807615 0.02204409</span></span>
<span><span class="co">## C 0.02905812 0.02805611</span></span>
<span><span class="co">## Q 0.02304609 0.01903808</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="calculateentropy">calculateEntropy<a class="anchor" aria-label="anchor" href="#calculateentropy"></a>
</h2>
<p>This function measures the diversity (or entropy) at each position in
a set of aligned sequences. It can use several common diversity
metrics.</p>
<p>Parameters for <code><a href="https://rdrr.io/pkg/immApex/man/calculateEntropy.html" class="external-link">calculateEntropy()</a></code></p>
<ul>
<li>
<strong>method</strong>: The diversity metric to use: “shannon”,
“inv.simpson”, “gini.simpson”, “norm.entropy”, “pielou”, or Hill numbers
(“hill0”, “hill1”, “hill2”).</li>
</ul>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">shannon.entropy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/calculateEntropy.html" class="external-link">calculateEntropy</a></span><span class="op">(</span><span class="va">sequences</span>, method <span class="op">=</span> <span class="st">"shannon"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">shannon.entropy</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     Pos1     Pos2     Pos3     Pos4     Pos5     Pos6 </span></span>
<span><span class="co">## 0.000000 0.000000 0.000000 2.986591 2.984538 2.985312</span></span></code></pre>
<div class="section level3">
<h3 id="calculateproperty">calculateProperty<a class="anchor" aria-label="anchor" href="#calculateproperty"></a>
</h3>
<p>This function computes position-wise summary statistics for amino
acid properties. For each position, it calculates a metric (like the
mean) of a specific physicochemical property for all residues found at
that position.</p>
<p>Parameters for <code><a href="https://rdrr.io/pkg/immApex/man/calculateProperty.html" class="external-link">calculateProperty()</a></code></p>
<ul>
<li>
<strong>property.set</strong>: The amino acid property scale to use
(e.g., “atchleyFactors”).</li>
<li>
<strong>summary.fun</strong>: The summary statistic to apply
(“mean”, “median”, “sum” etc.).</li>
</ul>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate the mean of Atchley factors at each position</span></span>
<span><span class="va">atchley.profile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/calculateProperty.html" class="external-link">calculateProperty</a></span><span class="op">(</span><span class="va">sequences</span>, property.set <span class="op">=</span> <span class="st">"atchleyFactors"</span>, summary.fun <span class="op">=</span> <span class="st">"mean"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">atchley.profile</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      position</span></span>
<span><span class="co">## scale  Pos.1  Pos.2  Pos.3       Pos.4       Pos.5       Pos.6</span></span>
<span><span class="co">##   AF1 -1.343 -0.591 -0.228 -0.04847395  0.06555010  0.03563226</span></span>
<span><span class="co">##   AF2  0.465 -1.302  1.399 -0.04388978 -0.05410321 -0.04387976</span></span>
<span><span class="co">##   AF3 -0.862 -0.733 -4.760 -0.02414930  0.06490982 -0.09043687</span></span>
<span><span class="co">##   AF4 -1.020  1.570  0.670  0.02724349 -0.02970441 -0.02113026</span></span>
<span><span class="co">##   AF5 -0.255 -0.146 -2.647 -0.02846994  0.06192184 -0.03162625</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="calculategeneusage">calculateGeneUsage<a class="anchor" aria-label="anchor" href="#calculategeneusage"></a>
</h3>
<p>This function quantifies the usage of gene loci (e.g., V and J genes)
within a repertoire.</p>
<p>Parameters for <code><a href="https://rdrr.io/pkg/immApex/man/calculateGeneUsage.html" class="external-link">calculateGeneUsage()</a></code> *
<strong>loci</strong>: A character vector of one or two column names
corresponding to the gene loci. * <strong>summary</strong>: The output
format: “proportion” (default), “count”, or “percent”.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using the built-in AIRR data</span></span>
<span><span class="va">data_airr</span> <span class="op">&lt;-</span> <span class="va">immapex_example.data</span><span class="op">[[</span><span class="st">"AIRR"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Calculate paired V-J gene usage as percentages</span></span>
<span><span class="va">vj_usage</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/calculateGeneUsage.html" class="external-link">calculateGeneUsage</a></span><span class="op">(</span><span class="va">data_airr</span>, loci <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"v_call"</span>, <span class="st">"j_call"</span><span class="op">)</span>, summary <span class="op">=</span> <span class="st">"percent"</span><span class="op">)</span></span>
<span><span class="va">vj_usage</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##             y</span></span>
<span><span class="co">## x             TRAJ10*01  TRAJ11*01  TRAJ12*01  TRAJ13*01  TRAJ13*02</span></span>
<span><span class="co">##   TRAV1-1*01 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000</span></span>
<span><span class="co">##   TRAV1-1*02 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000</span></span>
<span><span class="co">##   TRAV1-2*01 0.00000000 0.01666667 0.00000000 0.00000000 0.00000000</span></span>
<span><span class="co">##   TRAV1-2*03 0.01666667 0.00000000 0.03333333 0.00000000 0.01666667</span></span>
<span><span class="co">##   TRAV10*01  0.08333333 0.00000000 0.00000000 0.00000000 0.00000000</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="calculatemotif">calculateMotif<a class="anchor" aria-label="anchor" href="#calculatemotif"></a>
</h3>
<p>This function rapidly finds and counts contiguous (or gapped) amino
acid or nucleotide motifs of specified lengths across a set of
sequences.</p>
<p>Parameters for <code><a href="https://rdrr.io/pkg/immApex/man/calculateMotif.html" class="external-link">calculateMotif()</a></code></p>
<ul>
<li>
<strong>motif.lengths</strong>: An integer vector of motif sizes to
search for.</li>
<li>
<strong>min.depth</strong>: The minimum number of times a motif must
appear to be included in the output.</li>
<li>
<strong>discontinuous</strong>: If TRUE, also finds motifs with a
single gap (e.g., “C.S”).</li>
</ul>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">motif.counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/calculateMotif.html" class="external-link">calculateMotif</a></span><span class="op">(</span><span class="va">sequences</span>, motif.lengths <span class="op">=</span> <span class="fl">3</span>, min.depth <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">motif.counts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   motif frequency</span></span>
<span><span class="co">## 1   SYF        52</span></span>
<span><span class="co">## 2   LQY         5</span></span>
<span><span class="co">## 3   SAT         6</span></span>
<span><span class="co">## 4   STH         5</span></span>
<span><span class="co">## 5   SQI         5</span></span>
<span><span class="co">## 6   HNY         6</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="probabilitymatrix">probabilityMatrix<a class="anchor" aria-label="anchor" href="#probabilitymatrix"></a>
</h3>
<p>This function calculates a positional probability matrix (PPM) for a
group of sequences. This can be used to represent the sequence profile
of an antigen-specific repertoire or to generate a sequence logo.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ppm.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/probabilityMatrix.html" class="external-link">probabilityMatrix</a></span><span class="op">(</span><span class="va">sequences</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ppm.matrix</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   Pos.1 Pos.2 Pos.3      Pos.4      Pos.5      Pos.6      Pos.7      Pos.8</span></span>
<span><span class="co">## A     0     1     0 0.05210421 0.04308617 0.05210421 0.03807615 0.03106212</span></span>
<span><span class="co">## R     0     0     0 0.04208417 0.06412826 0.06012024 0.03707415 0.03306613</span></span>
<span><span class="co">## N     0     0     0 0.03907816 0.04208417 0.05911824 0.04809619 0.03306613</span></span>
<span><span class="co">## D     0     0     0 0.04809619 0.05210421 0.05611222 0.04208417 0.03306613</span></span>
<span><span class="co">## C     1     0     0 0.05010020 0.05410822 0.04609218 0.04008016 0.03807615</span></span>
<span><span class="co">## Q     0     0     0 0.04108216 0.05811623 0.05811623 0.04308617 0.02905812</span></span>
<span><span class="co">##        Pos.9     Pos.10    Pos.11     Pos.12     Pos.13     Pos.14 Pos.15</span></span>
<span><span class="co">## A 0.03780069 0.03439153 0.0312500 0.02592593 0.02347418 0.01242236      0</span></span>
<span><span class="co">## R 0.03436426 0.03703704 0.0312500 0.03333333 0.01643192 0.01552795      0</span></span>
<span><span class="co">## N 0.03665521 0.03174603 0.0359375 0.03148148 0.03521127 0.00931677      0</span></span>
<span><span class="co">## D 0.04352806 0.02910053 0.0375000 0.04259259 0.01877934 0.00621118      0</span></span>
<span><span class="co">## C 0.03321879 0.03703704 0.0328125 0.02407407 0.03755869 0.02795031      0</span></span>
<span><span class="co">## Q 0.02634593 0.02513228 0.0203125 0.03888889 0.02816901 0.01242236      0</span></span>
<span><span class="co">##   Pos.16</span></span>
<span><span class="co">## A      0</span></span>
<span><span class="co">## R      0</span></span>
<span><span class="co">## N      0</span></span>
<span><span class="co">## D      0</span></span>
<span><span class="co">## C      0</span></span>
<span><span class="co">## Q      0</span></span></code></pre>
<p>The PPM can also be converted to a positional weight matrix (PWM)
using a log-likelihood ratio.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="co"># Generate a sample background frequency</span></span>
<span><span class="va">back.freq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span>, <span class="fl">20</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">back.freq</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">amino.acids</span></span>
<span><span class="va">back.freq</span> <span class="op">&lt;-</span> <span class="va">back.freq</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">back.freq</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pwm.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/probabilityMatrix.html" class="external-link">probabilityMatrix</a></span><span class="op">(</span><span class="va">sequences</span>, max.length <span class="op">=</span> <span class="fl">20</span>, convert.PWM <span class="op">=</span> <span class="cn">TRUE</span>, background.frequencies <span class="op">=</span> <span class="va">back.freq</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pwm.matrix</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       Pos.1     Pos.2     Pos.3       Pos.4      Pos.5      Pos.6      Pos.7</span></span>
<span><span class="co">## A -6.153093  3.811248 -6.153093 -0.42517268 -0.6936615 -0.4251727 -0.8676909</span></span>
<span><span class="co">## R -6.982686 -6.982686 -6.982686 -1.55642111 -0.9603181 -1.0519485 -1.7347584</span></span>
<span><span class="co">## N -5.347666 -5.347666 -5.347666 -0.02573756  0.0785991  0.5592249  0.2670442</span></span>
<span><span class="co">## D -4.278624 -4.278624 -4.278624  1.33608583  1.4492964  1.5542660  1.1476407</span></span>
<span><span class="co">## C  6.733651 -3.230690 -3.230690  2.44173581  2.5506702  2.3238993  2.1268625</span></span>
<span><span class="co">## Q -4.854126 -4.854126 -4.854126  0.53819124  1.0285169  1.0285169  0.6053054</span></span>
<span><span class="co">##         Pos.8      Pos.9     Pos.10     Pos.11     Pos.12      Pos.13</span></span>
<span><span class="co">## A -1.15309313 -0.8766248 -1.0065966 -1.1355761 -1.3839637 -1.50303956</span></span>
<span><span class="co">## R -1.89522302 -1.8394841 -1.7330959 -1.9651688 -1.8725195 -2.79206392</span></span>
<span><span class="co">## N -0.26020282 -0.1142661 -0.3122005 -0.1375035 -0.3155018 -0.15704371</span></span>
<span><span class="co">## D  0.80883883  1.1957837  0.6365469  0.9904318  1.1685773  0.08192294</span></span>
<span><span class="co">## C  2.05471268  1.8652065  2.0189005  1.8539417  1.4389042  2.04739525</span></span>
<span><span class="co">## Q  0.05276441 -0.0801582 -0.1405891 -0.4215716  0.4675443  0.03693548</span></span>
<span><span class="co">##       Pos.14    Pos.15     Pos.16     Pos.17     Pos.18     Pos.19     Pos.20</span></span>
<span><span class="co">## A -2.2574957 -4.080435 -3.1615713 -0.4834994 -0.4834994 -0.4834994 -0.4834994</span></span>
<span><span class="co">## R -2.8240540 -4.910027 -3.9911640 -1.3130921 -1.3130921 -1.3130921 -1.3130921</span></span>
<span><span class="co">## N -1.7739963 -3.275007 -2.3561438  0.3219281  0.3219281  0.3219281  0.3219281</span></span>
<span><span class="co">## D -1.1199922 -2.205965 -1.2871022  1.3909697  1.3909697  1.3909697  1.3909697</span></span>
<span><span class="co">## C  1.6649079 -1.158031 -0.2391677  2.4389042  2.4389042  2.4389042  2.4389042</span></span>
<span><span class="co">## Q -0.9585288 -2.781468 -1.8626043  0.8154676  0.8154676  0.8154676  0.8154676</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="adjacencymatrix-and-buildnetwork">adjacencyMatrix and buildNetwork<a class="anchor" aria-label="anchor" href="#adjacencymatrix-and-buildnetwork"></a>
</h3>
<p>These functions help analyze relationships between residues or whole
sequences.</p>
<p><code><a href="https://rdrr.io/pkg/immApex/man/adjacencyMatrix.html" class="external-link">adjacencyMatrix()</a></code> summarizes transitions between
adjacent residues, creating a matrix of co-occurrence counts.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">adj.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/adjacencyMatrix.html" class="external-link">adjacencyMatrix</a></span><span class="op">(</span><span class="va">sequences</span>, normalize <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">adj.matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##      A  R  N  D    C  Q  E  G  H  I</span></span>
<span><span class="co">## A   20 28 38 26 1023 28 24 28 32 24</span></span>
<span><span class="co">## R   28 22 41 33   27 34 33 27 30 34</span></span>
<span><span class="co">## N   38 41 16 14   21 31 23 31 33 29</span></span>
<span><span class="co">## D   26 33 14 38   35 18 30 24 32 35</span></span>
<span><span class="co">## C 1023 27 21 35   24 28 30 30 27 34</span></span>
<span><span class="co">## Q   28 34 31 18   28 20 28 27 20 31</span></span>
<span><span class="co">## E   24 33 23 30   30 28 28 28 23 32</span></span>
<span><span class="co">## G   28 27 31 24   30 27 28 36 25 35</span></span>
<span><span class="co">## H   32 30 33 32   27 20 23 25 28 28</span></span>
<span><span class="co">## I   24 34 29 35   34 31 32 35 28 42</span></span></code></pre>
<p><code><a href="https://rdrr.io/pkg/immApex/man/buildNetwork.html" class="external-link">buildNetwork()</a></code> constructs a similarity network where
nodes are sequences and edges connect sequences with an edit distance
below a given threshold.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Building an edge list</span></span>
<span><span class="va">g1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/buildNetwork.html" class="external-link">buildNetwork</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>sequences <span class="op">=</span> <span class="va">sequences</span><span class="op">)</span>, seq_col <span class="op">=</span> <span class="st">"sequences"</span>, threshold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Forming network</span></span>
<span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r.igraph.org/reference/graph_from_edgelist.html" class="external-link">graph_from_edgelist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">g1</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/E.html" class="external-link">E</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">weight</span> <span class="op">&lt;-</span> <span class="va">g1</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Remove isolated nodes for clearer visualization</span></span>
<span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r.igraph.org/reference/delete_vertices.html" class="external-link">delete_vertices</a></span><span class="op">(</span><span class="va">graph</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://r.igraph.org/reference/degree.html" class="external-link">degree</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert to tidygraph for use with ggraph</span></span>
<span><span class="va">g_tidy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tidygraph.data-imaginist.com/reference/tbl_graph.html" class="external-link">as_tbl_graph</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the network</span></span>
<span><span class="fu"><a href="https://ggraph.data-imaginist.com/reference/ggraph.html" class="external-link">ggraph</a></span><span class="op">(</span><span class="va">g_tidy</span>, layout <span class="op">=</span> <span class="st">"fr"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_edge_link.html" class="external-link">geom_edge_link</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>width <span class="op">=</span> <span class="va">weight</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"black"</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_node_point.html" class="external-link">geom_node_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://r.igraph.org/reference/degree.html" class="external-link">degree</a></span><span class="op">(</span><span class="va">g_tidy</span>, mode <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span><span class="op">)</span>, fill <span class="op">=</span> <span class="st">"steelblue"</span>, color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>        shape <span class="op">=</span> <span class="fl">21</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/scale_edge_width.html" class="external-link">scale_edge_width</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>size <span class="op">=</span> <span class="st">"Degree"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span></span></code></pre></div>
<p><img src="immApex_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="encoding-sequences-for-model-input">Encoding Sequences for Model Input<a class="anchor" aria-label="anchor" href="#encoding-sequences-for-model-input"></a>
</h2>
<div class="section level3">
<h3 id="sequenceencoder">sequenceEncoder<a class="anchor" aria-label="anchor" href="#sequenceencoder"></a>
</h3>
<p>The primary tool for converting biological sequences into a numerical
format suitable for machine learning is the
<code><a href="https://rdrr.io/pkg/immApex/man/sequenceEncoder.html" class="external-link">sequenceEncoder()</a></code> function. It is a versatile wrapper that
can generate three different types of numerical representations,
controlled by the mode argument. Understanding this function is key to
using the package effectively.</p>
<p><strong>Key Arguments</strong>:</p>
<ul>
<li>
<strong>input.sequences</strong>: The character vector of your amino
acid or nucleotide sequences.</li>
<li>
<strong>mode</strong>: The most important argument. A string
specifying the encoding type you need: <code>"onehot"</code>,
<code>"property"</code>, or <code>"geometric"</code>.</li>
<li>
<strong>verbose</strong>: A logical flag (TRUE or FALSE) to control
whether progress messages are printed.</li>
</ul>
</div>
<div class="section level3">
<h3 id="onehotencoder">onehotEncoder<a class="anchor" aria-label="anchor" href="#onehotencoder"></a>
</h3>
<p>One-hot encoding is a common method that transforms each amino acid
at each position into a binary vector. This is useful when the specific
identity and position of each residue are important. This functionality
is accessed by setting <strong>mode</strong> = <code>"onehot"</code> in
<code><a href="https://rdrr.io/pkg/immApex/man/sequenceEncoder.html" class="external-link">sequenceEncoder()</a></code> or by using the
<code><a href="https://rdrr.io/pkg/immApex/man/sequenceEncoder.html" class="external-link">onehotEncoder()</a></code> alias.</p>
<p><strong>Mode-Specific Arguments</strong>:</p>
<ul>
<li>
<strong>max.length</strong>: Additional length to pad, NULL will pad
sequences to the max length of input.sequences</li>
<li>
<strong>padding.symbol</strong>: The single character used to pad
sequences shorter than max.length.</li>
<li>
<strong>sequence.dictionary</strong>: The letters to use in encoding
(default are all amino acids)</li>
</ul>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate one-hot encoding using the main function</span></span>
<span><span class="va">enc_onehot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/sequenceEncoder.html" class="external-link">sequenceEncoder</a></span><span class="op">(</span><span class="va">sequences</span>, mode <span class="op">=</span> <span class="st">"onehot"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># You can achieve the same result with the alias</span></span>
<span><span class="va">enc_onehot_alias</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/sequenceEncoder.html" class="external-link">onehotEncoder</a></span><span class="op">(</span><span class="va">sequences</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View the first few columns of the flattened matrix output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">enc_onehot</span><span class="op">$</span><span class="va">flattened</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      A_1 R_1 N_1 D_1 C_1 Q_1 E_1 G_1 H_1 I_1</span></span>
<span><span class="co">## [1,]   0   0   0   0   1   0   0   0   0   0</span></span>
<span><span class="co">## [2,]   0   0   0   0   1   0   0   0   0   0</span></span>
<span><span class="co">## [3,]   0   0   0   0   1   0   0   0   0   0</span></span>
<span><span class="co">## [4,]   0   0   0   0   1   0   0   0   0   0</span></span>
<span><span class="co">## [5,]   0   0   0   0   1   0   0   0   0   0</span></span>
<span><span class="co">## [6,]   0   0   0   0   1   0   0   0   0   0</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="propertyencoder">propertyEncoder<a class="anchor" aria-label="anchor" href="#propertyencoder"></a>
</h3>
<p>Instead of a binary vector, this method represents each amino acid
using a set of continuous numerical values based on physicochemical
properties (e.g., kideraFactors, FASGAI, etc). This can capture the
biochemical similarity between amino acids. Access this method by
setting <strong>mode</strong> = <code>"property"</code> or using the
<code><a href="https://rdrr.io/pkg/immApex/man/sequenceEncoder.html" class="external-link">propertyEncoder()</a></code> alias.</p>
<p><strong>Mode-Specific Arguments</strong>:</p>
<ul>
<li>
<strong>property.set</strong>: A character string specifying which
set of pre-defined amino acid properties to use.</li>
<li>
<strong>property.matrix</strong>: A custom numeric matrix of
property values you provide yourself.</li>
</ul>
<p>Available <strong>property.set</strong> options include:</p>
<ul>
<li>atchleyFactors - <a href="https://pubmed.ncbi.nlm.nih.gov/15851683/" class="external-link">citation</a>
</li>
<li>crucianiProperties - <a href="https://analyticalsciencejournals.onlinelibrary.wiley.com/doi/abs/10.1002/cem.856" class="external-link">citation</a>
</li>
<li>FASGAI - <a href="https://pubmed.ncbi.nlm.nih.gov/18318694/" class="external-link">citation</a>
</li>
<li>kideraFactors - <a href="https://link.springer.com/article/10.1007/BF01025492" class="external-link">citation</a>
</li>
<li>MSWHIM - <a href="https://pubs.acs.org/doi/10.1021/ci980211b" class="external-link">citation</a>
</li>
<li>ProtFP - <a href="https://pubmed.ncbi.nlm.nih.gov/24059694/" class="external-link">citation</a>
</li>
<li>stScales - <a href="https://pubmed.ncbi.nlm.nih.gov/19373543/" class="external-link">citation</a>
</li>
<li>tScales - <a href="https://www.sciencedirect.com/science/article/abs/pii/S0022286006006314" class="external-link">citation</a>
</li>
<li>VHSE - <a href="https://pubmed.ncbi.nlm.nih.gov/15895431/" class="external-link">citation</a>
</li>
<li>zScales - <a href="https://pubmed.ncbi.nlm.nih.gov/9651153/" class="external-link">citation</a>
</li>
</ul>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate property-based encoding using FASGAI factors</span></span>
<span><span class="va">enc_prop</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/sequenceEncoder.html" class="external-link">sequenceEncoder</a></span><span class="op">(</span><span class="va">sequences</span>, mode <span class="op">=</span> <span class="st">"property"</span>, property.set <span class="op">=</span> <span class="st">"FASGAI"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The propertyEncoder() alias is a convenient shortcut</span></span>
<span><span class="va">enc_prop_alias</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/sequenceEncoder.html" class="external-link">propertyEncoder</a></span><span class="op">(</span><span class="va">sequences</span>, property.set <span class="op">=</span> <span class="st">"FASGAI"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View the first few columns of the flattened property matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">enc_prop</span><span class="op">$</span><span class="va">flattened</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       F1_1  F2_1   F3_1  F4_1   F5_1  F6_1</span></span>
<span><span class="co">## [1,] 0.997 0.021 -1.419 -2.08 -0.799 0.502</span></span>
<span><span class="co">## [2,] 0.997 0.021 -1.419 -2.08 -0.799 0.502</span></span>
<span><span class="co">## [3,] 0.997 0.021 -1.419 -2.08 -0.799 0.502</span></span>
<span><span class="co">## [4,] 0.997 0.021 -1.419 -2.08 -0.799 0.502</span></span>
<span><span class="co">## [5,] 0.997 0.021 -1.419 -2.08 -0.799 0.502</span></span>
<span><span class="co">## [6,] 0.997 0.021 -1.419 -2.08 -0.799 0.502</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="geometricencoder">geometricEncoder<a class="anchor" aria-label="anchor" href="#geometricencoder"></a>
</h3>
<p>This approach creates a single, fixed-length vector (an “embedding”)
for an entire sequence. It works by averaging the vectors for each amino
acid from a substitution matrix (like BLOSUM62) and then applying a
geometric rotation. This is useful for tasks where a summary of the
entire sequence is needed, similar to approaches like <a href="https://pubmed.ncbi.nlm.nih.gov/34349111/" class="external-link">GIANA</a>. Use this
with <strong>mode</strong> = <code>"geometric"</code> or the
<code><a href="https://rdrr.io/pkg/immApex/man/sequenceEncoder.html" class="external-link">geometricEncoder()</a></code> alias.</p>
<p>Parameters for <code><a href="https://rdrr.io/pkg/immApex/man/sequenceEncoder.html" class="external-link">geometricEncoder()</a></code></p>
<ul>
<li>
<strong>method</strong>: Select the following substitution matrices:
“BLOSUM45”, “BLOSUM50”, “BLOSUM62”, “BLOSUM80”, “BLOSUM100”, “PAM30”,
“PAM40”, “PAM70”, “PAM120”, or “PAM250”</li>
<li>
<strong>theta</strong>: The angle in which to create the rotation
matrix</li>
</ul>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate a geometric embedding for each sequence</span></span>
<span><span class="va">enc_geo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/sequenceEncoder.html" class="external-link">sequenceEncoder</a></span><span class="op">(</span><span class="va">sequences</span>, mode <span class="op">=</span> <span class="st">"geometric"</span>, method <span class="op">=</span> <span class="st">"BLOSUM62"</span>, theta <span class="op">=</span> <span class="va">pi</span><span class="op">/</span><span class="fl">3</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The alias provides a direct path to this functionality</span></span>
<span><span class="va">enc_geo_alias</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/sequenceEncoder.html" class="external-link">geometricEncoder</a></span><span class="op">(</span><span class="va">sequences</span>, method <span class="op">=</span> <span class="st">"BLOSUM62"</span>, theta <span class="op">=</span> <span class="va">pi</span><span class="op">/</span><span class="fl">3</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The output is a single summary matrix where each row is a sequence</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">enc_geo</span><span class="op">$</span><span class="va">summary</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        [,1]        [,2]       [,3]        [,4]       [,5]       [,6]      [,7]</span></span>
<span><span class="co">## 1 -1.640544 -0.65849365 -2.7610572  0.28229128 -0.6282849 -1.6617786 -2.840304</span></span>
<span><span class="co">## 2 -1.229861 -0.03648518 -0.7135148  0.06917725 -0.8720085  0.8436963 -1.754380</span></span>
<span><span class="co">## 3 -2.056810 -1.18750000 -3.2900635  0.69855716 -1.5612976 -1.0457532 -3.119310</span></span>
<span><span class="co">## 4 -1.380529  0.14114564 -1.3470349 -0.41686706 -1.2371393  1.0177881 -1.930168</span></span>
<span><span class="co">## 5 -1.462587 -0.73945224 -2.3684144  0.10221415 -1.1720277 -0.1518066 -2.628962</span></span>
<span><span class="co">## 6 -1.732051 -1.00000000 -2.8133898  0.42848961 -0.7915951 -1.7400282 -2.883831</span></span>
<span><span class="co">##          [,8]       [,9]       [,10]     [,11]      [,12]       [,13]</span></span>
<span><span class="co">## 1  0.66955081 -0.9910254 -0.28349365 -1.953044 -0.1172278 -0.12500000</span></span>
<span><span class="co">## 2 -0.62799153 -1.6852027  0.08552329 -1.255181  1.0073714 -1.51036297</span></span>
<span><span class="co">## 3  0.65280398 -0.3337341  1.82804446 -1.794551 -0.8917468  0.04575318</span></span>
<span><span class="co">## 4 -0.03185118 -1.9072913  0.05352540 -1.191386  1.1885413 -1.76778811</span></span>
<span><span class="co">## 5  0.37167793 -0.8392773  1.09003464 -1.474767 -0.5365385 -0.53001155</span></span>
<span><span class="co">## 6  0.77272030 -1.0664529  1.18048396 -1.761823 -0.5039887 -0.09622504</span></span>
<span><span class="co">##          [,14]      [,15]     [,16]     [,17]      [,18]      [,19]       [,20]</span></span>
<span><span class="co">## 1  0.216506351 -1.5122595 1.8693103 -1.582532  0.2410254 -0.6785254 -0.82475953</span></span>
<span><span class="co">## 2  0.282692070 -0.9166667 1.5877132 -2.043055 -0.7946582 -1.3853630  0.06618572</span></span>
<span><span class="co">## 3  0.170753175 -1.8370191 1.6818103 -1.828044 -0.3337341  0.7120191  0.26674682</span></span>
<span><span class="co">## 4  0.061898816 -0.7935095 0.6243988 -1.859294 -0.2796075 -1.4346552  0.10989564</span></span>
<span><span class="co">## 5  0.008916019 -1.4512820 1.7864214 -2.095687 -0.3701633 -0.4180069  0.17855469</span></span>
<span><span class="co">## 6 -0.055555556 -1.1666667 2.0207259 -1.680484 -0.2004275 -0.4960113 -0.02977213</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="tokenizesequences">tokenizeSequences<a class="anchor" aria-label="anchor" href="#tokenizesequences"></a>
</h3>
<p>Another approach to transforming a sequence into numerical values is
tokenizing it into numbers. This is a common approach for recurrent
neural networks where one letter corresponds to a single integer. In
addition, we can add start and stop tokens to our original sequences to
differentiate between the beginning and end of the sequences.</p>
<p>Parameters for <code><a href="https://rdrr.io/pkg/immApex/man/tokenizeSequences.html" class="external-link">tokenizeSequences()</a></code></p>
<ul>
<li>
<strong>add.startstop</strong> Add start and stop tokens to the
sequence</li>
<li>
<strong>start.token</strong> The character to use for the start
token</li>
<li>
<strong>stop.token</strong> The character to use for the stop
token</li>
<li>
<strong>max.length</strong> Additional length to pad, NULL will pad
sequences to the max length of input.sequences</li>
<li>
<strong>convert.to.matrix</strong> Return a matrix
(<strong>TRUE</strong>) or a vector (<strong>FALSE</strong>)</li>
</ul>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">token.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/tokenizeSequences.html" class="external-link">tokenizeSequences</a></span><span class="op">(</span>input.sequences <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">sequences</span>, <span class="va">mutated.sequences</span><span class="op">)</span>, add.startstop <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    start.token <span class="op">=</span> <span class="st">"!"</span>, stop.token <span class="op">=</span> <span class="st">"^"</span>, convert.to.matrix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">token.matrix</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">18</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13] [,14]</span></span>
<span><span class="co">## [1,]    1    6    2   17   10   14    6   20   15    22    23    23    23    23</span></span>
<span><span class="co">## [2,]    1    6    2   17   13    5   11    4    8     7    18    20    15    22</span></span>
<span><span class="co">## [3,]    1    6    2   17   21   11   21   20   15    22    23    23    23    23</span></span>
<span><span class="co">## [4,]    1    6    2   17   13   21    4   19    4    16    13    16     4    21</span></span>
<span><span class="co">## [5,]    1    6    2   17   11    7   12    2   11     4    20    15    22    23</span></span>
<span><span class="co">## [6,]    1    6    2   17   14    6   11   17   20    15    22    23    23    23</span></span>
<span><span class="co">##      [,15] [,16] [,17] [,18]</span></span>
<span><span class="co">## [1,]    23    23    23    23</span></span>
<span><span class="co">## [2,]    23    23    23    23</span></span>
<span><span class="co">## [3,]    23    23    23    23</span></span>
<span><span class="co">## [4,]     7    20    15    22</span></span>
<span><span class="co">## [5,]    23    23    23    23</span></span>
<span><span class="co">## [6,]    23    23    23    23</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="sequencedecoder">sequenceDecoder<a class="anchor" aria-label="anchor" href="#sequencedecoder"></a>
</h3>
<p>We have a function called <code><a href="https://rdrr.io/pkg/immApex/man/sequenceDecoder.html" class="external-link">sequenceDecoder()</a></code> that
extracts sequences from one-hot or property-encoded matrices or arrays.
This function can be applied to any generative approach to sequence
generation.</p>
<p>Parameters for <code><a href="https://rdrr.io/pkg/immApex/man/sequenceDecoder.html" class="external-link">sequenceDecoder()</a></code></p>
<ul>
<li>
<strong>sequence.matrix</strong> The encoded sequences to decode in
an array or matrix</li>
<li>
<strong>mode</strong> The encoding mode used for decoding:
<code>"onehot"</code> or <code>"property"</code>
</li>
<li>
<strong>property.set</strong> For <code>mode = "property"</code>, a
character vector of property names (e.g., <code>"Atchley"</code>) that
were used for the original encoding.</li>
<li>
<strong>call.threshold</strong> The relative strictness of sequence
calling with higher values being more stringent</li>
</ul>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">property.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/sequenceEncoder.html" class="external-link">propertyEncoder</a></span><span class="op">(</span>input.sequences <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">sequences</span>, <span class="va">mutated.sequences</span><span class="op">)</span>, property.set <span class="op">=</span> <span class="st">"FASGAI"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">property.sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/sequenceDecoder.html" class="external-link">sequenceDecoder</a></span><span class="op">(</span><span class="va">property.matrix</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, mode <span class="op">=</span> <span class="st">"property"</span>, property.set <span class="op">=</span> <span class="st">"FASGAI"</span>,</span>
<span>    call.threshold <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sequences</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CASHMCYF"         "CASKDINEQTYF"     "CASVIVYF"         "CASKVNWNPKPNVQYF"</span></span>
<span><span class="co">## [5] "CASIQLAINYF"      "CASMCISYF"</span></span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">property.sequences</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                                                             </span></span>
<span><span class="co">##         "CASHMCYF"     "CASKDINEQTYF"         "CASVIVYF" "CASKVNWNPKPNVQYF" </span></span>
<span><span class="co">##                                       </span></span>
<span><span class="co">##      "CASIQLAINYF"        "CASMCISYF"</span></span></code></pre>
<p>A similar approach can be applied when using matrices or arrays
derived from one-hot encoding:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sequence.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/sequenceEncoder.html" class="external-link">onehotEncoder</a></span><span class="op">(</span>input.sequences <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">sequences</span>, <span class="va">mutated.sequences</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">OHE.sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/sequenceDecoder.html" class="external-link">sequenceDecoder</a></span><span class="op">(</span><span class="va">sequence.matrix</span>,</span>
<span>                                 mode<span class="op">=</span> <span class="st">"onehot"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">OHE.sequences</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CASHMCYF"         "CASKDINEQTYF"     "CASVIVYF"         "CASKVNWNPKPNVQYF"</span></span>
<span><span class="co">## [5] "CASIQLAINYF"      "CASMCISYF"</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="training-a-model">Training a Model<a class="anchor" aria-label="anchor" href="#training-a-model"></a>
</h2>
<div class="section level3">
<h3 id="example-1-classifying-sequences-with-random-forest">Example 1: Classifying Sequences with Random Forest<a class="anchor" aria-label="anchor" href="#example-1-classifying-sequences-with-random-forest"></a>
</h3>
<p>A common task is to classify sequences into groups, such as
predicting whether an immune receptor binds to a specific antigen. Here,
we’ll train a Random Forest model to distinguish between two classes of
sequences based on their physicochemical properties</p>
<p><strong>Step 1: Simulate Data and Engineer Features</strong></p>
<p>First, we’ll simulate two distinct classes of sequences using
<code><a href="https://rdrr.io/pkg/immApex/man/generateSequences.html" class="external-link">generateSequences()</a></code>. We’ll then use
<code><a href="https://rdrr.io/pkg/immApex/man/sequenceEncoder.html" class="external-link">propertyEncoder()</a></code> with the “atchleyFactors” to convert each
sequence into a flattened numerical vector. Each element in the vector
represents a specific physicochemical property at a specific position,
transforming our variable-length sequences into a fixed-size feature
matrix suitable for machine learning.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Step 1a: Generate two distinct classes of sequences</span></span>
<span><span class="va">class1.sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/generateSequences.html" class="external-link">generateSequences</a></span><span class="op">(</span>prefix.motif <span class="op">=</span> <span class="st">"CAS"</span>, min.length <span class="op">=</span> <span class="fl">3</span>, number.of.sequences <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span></span>
<span><span class="va">class2.sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/generateSequences.html" class="external-link">generateSequences</a></span><span class="op">(</span>prefix.motif <span class="op">=</span> <span class="st">"CSG"</span>, min.length <span class="op">=</span> <span class="fl">3</span>, number.of.sequences <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combine sequences and create labels</span></span>
<span><span class="va">all.sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">class1.sequences</span>, <span class="va">class2.sequences</span><span class="op">)</span></span>
<span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Class1"</span>, <span class="fl">500</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Class2"</span>, <span class="fl">500</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Step 1b: Use propertyEncoder to create a feature matrix from Atchley factors</span></span>
<span><span class="va">feature.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/sequenceEncoder.html" class="external-link">propertyEncoder</a></span><span class="op">(</span><span class="va">all.sequences</span>, property.set <span class="op">=</span> <span class="st">"atchleyFactors"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combine the flattened feature matrix and labels into the final data frame</span></span>
<span><span class="va">training.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">feature.matrix</span><span class="op">$</span><span class="va">flattened</span>, <span class="va">labels</span><span class="op">)</span></span></code></pre></div>
<p><strong>Step 2: Train and Evaluate the Model</strong></p>
<p>Now we can train the Random Forest classifier. This model is robust,
requires minimal tuning, and can provide insights into which features
(in this case, which property at which position) are most important for
distinguishing between the classes.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.stat.berkeley.edu/~breiman/RandomForests/" class="external-link">randomForest</a></span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Train the Random Forest model</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span>  <span class="co"># for reproducibility</span></span>
<span><span class="va">rf.model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/randomForest/man/randomForest.html" class="external-link">randomForest</a></span><span class="op">(</span><span class="va">labels</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">training.data</span>, ntree <span class="op">=</span> <span class="fl">100</span>, importance <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>  <span class="co"># Set importance=TRUE to calculate scores</span></span>
<span></span>
<span><span class="co"># Print the confusion matrix to see model performance</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">rf.model</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Call:</span></span>
<span><span class="co">##  randomForest(formula = labels ~ ., data = training.data, ntree = 100,      importance = TRUE) </span></span>
<span><span class="co">##                Type of random forest: classification</span></span>
<span><span class="co">##                      Number of trees: 100</span></span>
<span><span class="co">## No. of variables tried at each split: 7</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##         OOB estimate of  error rate: 0%</span></span>
<span><span class="co">## Confusion matrix:</span></span>
<span><span class="co">##        Class1 Class2 class.error</span></span>
<span><span class="co">## Class1    500      0           0</span></span>
<span><span class="co">## Class2      0    500           0</span></span></code></pre>
<p>A key advantage of Random Forest is the ability to easily extract
feature importance. We can create a plot to see which positional Atchley
factors were most influential in the model’s predictions.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract feature importance data from the model</span></span>
<span><span class="va">importance.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/randomForest/man/importance.html" class="external-link">importance</a></span><span class="op">(</span><span class="va">rf.model</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">importance.data</span><span class="op">$</span><span class="va">Feature</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">importance.data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get the top 15 most important features</span></span>
<span><span class="va">top_features</span> <span class="op">&lt;-</span> <span class="va">importance.data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">importance.data</span><span class="op">$</span><span class="va">MeanDecreaseGini</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">15</span>,</span>
<span>    <span class="op">]</span></span>
<span></span>
<span><span class="co"># Plot using ggplot2</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">top_features</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">MeanDecreaseGini</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html" class="external-link">reorder</a></span><span class="op">(</span><span class="va">Feature</span>, <span class="va">MeanDecreaseGini</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_col</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">MeanDecreaseGini</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Top 15 Most Important Features"</span>, x <span class="op">=</span> <span class="st">"Mean Decrease in Gini Impurity"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Feature (Property and Position)"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="immApex_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<p>The plot shows the features that the model found most useful for
classification. The feature names correspond to specific Atchley factors
at specific positions within the sequence - here residues 2 and 3 where
the difference was encoded - demonstrating how
<code><a href="https://rdrr.io/pkg/immApex/man/sequenceEncoder.html" class="external-link">propertyEncoder()</a></code> allows the model to learn from the
underlying biochemistry.</p>
</div>
<div class="section level3">
<h3 id="example-2-unsupervised-clustering-with-pca-and-geometric-encoding">Example 2: Unsupervised Clustering with PCA and Geometric
Encoding<a class="anchor" aria-label="anchor" href="#example-2-unsupervised-clustering-with-pca-and-geometric-encoding"></a>
</h3>
<p>Sometimes we don’t have labels, but we want to see if our sequences
form natural clusters. We can use dimensionality reduction techniques
like Principal Component Analysis (PCA) to visualize the relationships
between sequences. The <code><a href="https://rdrr.io/pkg/immApex/man/sequenceEncoder.html" class="external-link">geometricEncoder()</a></code> function is
perfect for this, as it creates a rich, fixed-length numerical embedding
for each sequence.</p>
<p><strong>Step 1: Simulate and Encode a Mixed Population of
Sequences</strong></p>
<p>We’ll generate three distinct families of sequences. Then, we will
use <code><a href="https://rdrr.io/pkg/immApex/man/sequenceEncoder.html" class="external-link">geometricEncoder()</a></code> to transform them into a
20-dimensional numerical matrix.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate three distinct groups of sequences</span></span>
<span><span class="va">groupA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/generateSequences.html" class="external-link">generateSequences</a></span><span class="op">(</span>prefix.motif <span class="op">=</span> <span class="st">"CA"</span>, number.of.sequences <span class="op">=</span> <span class="fl">100</span>, min.length <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">groupB</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/generateSequences.html" class="external-link">generateSequences</a></span><span class="op">(</span>prefix.motif <span class="op">=</span> <span class="st">"QR"</span>, number.of.sequences <span class="op">=</span> <span class="fl">100</span>, min.length <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">groupC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/generateSequences.html" class="external-link">generateSequences</a></span><span class="op">(</span>prefix.motif <span class="op">=</span> <span class="st">"YY"</span>, number.of.sequences <span class="op">=</span> <span class="fl">100</span>, min.length <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combine them and create a vector of original group IDs for later visualization</span></span>
<span><span class="va">mixed.sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">groupA</span>, <span class="va">groupB</span>, <span class="va">groupC</span><span class="op">)</span></span>
<span><span class="va">original.groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Group A"</span>, <span class="st">"Group B"</span>, <span class="st">"Group C"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">100</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Use geometricEncoder to create embeddings</span></span>
<span><span class="va">geometric.embeddings</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/sequenceEncoder.html" class="external-link">geometricEncoder</a></span><span class="op">(</span><span class="va">mixed.sequences</span>, method <span class="op">=</span> <span class="st">"BLOSUM62"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><strong>Step 2: Perform PCA and Visualize Clusters</strong></p>
<p>With our geometric.embeddings matrix, we can now perform PCA and plot
the results. If the geometric encoding captures the underlying
differences in our sequence families, we should see distinct clusters in
the plot of the first two principal components.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform PCA on the embedding matrix</span></span>
<span><span class="va">pca.result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="va">geometric.embeddings</span><span class="op">$</span><span class="va">summary</span>, center <span class="op">=</span> <span class="cn">TRUE</span>, scale. <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Prepare data for plotting with ggplot2</span></span>
<span><span class="va">pca.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>PC1 <span class="op">=</span> <span class="va">pca.result</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, PC2 <span class="op">=</span> <span class="va">pca.result</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>, Group <span class="op">=</span> <span class="va">original.groups</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot PC1 vs PC2 using ggplot2 and viridis</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">pca.data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span>, color <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span>, alpha <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_color_viridis</a></span><span class="op">(</span>discrete <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"PCA of Geometric Sequence Embeddings"</span>, x <span class="op">=</span> <span class="st">"Principal Component 1"</span>, y <span class="op">=</span> <span class="st">"Principal Component 2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="immApex_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
<p>The resulting plot clearly shows three distinct clusters,
demonstrating that the <code><a href="https://rdrr.io/pkg/immApex/man/sequenceEncoder.html" class="external-link">geometricEncoder()</a></code> successfully
captured the structural differences between our sequence families,
allowing for their separation in an unsupervised manner.</p>
</div>
<div class="section level3">
<h3 id="example-3-identifying-sequence-communities-with-network-analysis">Example 3: Identifying Sequence Communities with Network
Analysis<a class="anchor" aria-label="anchor" href="#example-3-identifying-sequence-communities-with-network-analysis"></a>
</h3>
<p>Beyond analyzing individual sequences, we can explore the
relationships between sequences by building a similarity network. A
powerful machine learning application for these networks is community
detection, an unsupervised clustering method that finds groups of
densely connected nodes. In immunology, this can be used to identify
clonal families or groups of sequences with shared features.</p>
<p><strong>Step 1: Simulate Data with Inherent Structure</strong></p>
<p>First, we will simulate a dataset containing three distinct
“families” of sequences. Our goal is to see if the network analysis can
blindly recover these groups without being given the labels.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="co"># Generate three distinct groups of sequences to simulate clonal families</span></span>
<span><span class="va">family1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/immApex/man/generateSequences.html" class="external-link">generateSequences</a></span><span class="op">(</span>prefix.motif <span class="op">=</span> <span class="st">"CASS"</span>, suffix.motif <span class="op">=</span> <span class="st">"YF"</span>, min.length <span class="op">=</span> <span class="fl">6</span>,</span>
<span>    number.of.sequences <span class="op">=</span> <span class="fl">60</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">family2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/immApex/man/generateSequences.html" class="external-link">generateSequences</a></span><span class="op">(</span>prefix.motif <span class="op">=</span> <span class="st">"CARS"</span>, suffix.motif <span class="op">=</span> <span class="st">"GF"</span>, min.length <span class="op">=</span> <span class="fl">6</span>,</span>
<span>    number.of.sequences <span class="op">=</span> <span class="fl">60</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">family3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/immApex/man/generateSequences.html" class="external-link">generateSequences</a></span><span class="op">(</span>prefix.motif <span class="op">=</span> <span class="st">"CSVA"</span>, suffix.motif <span class="op">=</span> <span class="st">"HF"</span>, min.length <span class="op">=</span> <span class="fl">6</span>,</span>
<span>    number.of.sequences <span class="op">=</span> <span class="fl">60</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combine into a single data frame</span></span>
<span><span class="va">all_sequences_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>sequence <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">family1</span>, <span class="va">family2</span>, <span class="va">family3</span><span class="op">)</span>, original_family <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Family 1"</span>,</span>
<span>    <span class="fl">42</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Family 2"</span>, <span class="fl">40</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Family 3"</span>, <span class="fl">46</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><strong>Step 2: Build the Sequence Network</strong></p>
<p>Next, we use <code><a href="https://rdrr.io/pkg/immApex/man/buildNetwork.html" class="external-link">buildNetwork()</a></code> to create an edge list based
on sequence similarity (edit distance). We then convert this list into a
formal graph object using the igraph package, which is the standard for
network analysis in R.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Build the edge list: connect sequences with an edit distance of 3 or less</span></span>
<span><span class="va">edge_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/immApex/man/buildNetwork.html" class="external-link">buildNetwork</a></span><span class="op">(</span><span class="va">all_sequences_df</span>, seq_col <span class="op">=</span> <span class="st">"sequence"</span>, threshold <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Replace numerical edge list with the sequences</span></span>
<span><span class="va">edge_list_sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">all_sequences_df</span><span class="op">$</span><span class="va">sequence</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">edge_list</span><span class="op">$</span><span class="va">from</span><span class="op">)</span><span class="op">]</span>,</span>
<span>    to <span class="op">=</span> <span class="va">all_sequences_df</span><span class="op">$</span><span class="va">sequence</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">edge_list</span><span class="op">$</span><span class="va">to</span><span class="op">)</span><span class="op">]</span>, dist <span class="op">=</span> <span class="va">edge_list</span><span class="op">$</span><span class="va">dist</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a graph object from the edge list and node data This graph now contains all our</span></span>
<span><span class="co"># sequences as nodes</span></span>
<span><span class="va">sequence_graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r.igraph.org/reference/graph_from_data_frame.html" class="external-link">graph_from_data_frame</a></span><span class="op">(</span>d <span class="op">=</span> <span class="va">edge_list_sequences</span>, vertices <span class="op">=</span> <span class="va">all_sequences_df</span>, directed <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/E.html" class="external-link">E</a></span><span class="op">(</span><span class="va">sequence_graph</span><span class="op">)</span><span class="op">$</span><span class="va">weight</span> <span class="op">&lt;-</span> <span class="va">edge_list_sequences</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<p><strong>Step 3: Perform Community Detection</strong></p>
<p>Now we apply a machine learning algorithm to find communities. The
<code>igraph</code> package offers many algorithms; we will use
“Walktrap,” a method that finds communities through a series of short
random walks. The idea is that walks are more likely to get “trapped”
within densely connected parts of the network (i.e., communities).</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform community detection using the Walktrap algorithm</span></span>
<span><span class="va">communities</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r.igraph.org/reference/cluster_walktrap.html" class="external-link">cluster_walktrap</a></span><span class="op">(</span><span class="va">sequence_graph</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add the community membership as an attribute to the graph nodes</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">sequence_graph</span><span class="op">)</span><span class="op">$</span><span class="va">community</span> <span class="op">&lt;-</span> <span class="va">communities</span><span class="op">$</span><span class="va">membership</span></span></code></pre></div>
<p><strong>Step 4: Visualize and Interpret the Communities</strong></p>
<p>Finally, we use <code>ggraph</code> to visualize the network,
coloring each node by the community it was assigned to by the algorithm.
If the analysis was successful, the colors should align with the
original sequence families we simulated.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert to tidygraph for use with ggraph</span></span>
<span><span class="va">g_tidy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tidygraph.data-imaginist.com/reference/tbl_graph.html" class="external-link">as_tbl_graph</a></span><span class="op">(</span><span class="va">sequence_graph</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the network, coloring nodes by their detected community</span></span>
<span><span class="fu"><a href="https://ggraph.data-imaginist.com/reference/ggraph.html" class="external-link">ggraph</a></span><span class="op">(</span><span class="va">g_tidy</span>, layout <span class="op">=</span> <span class="st">"fr"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_edge_link.html" class="external-link">geom_edge_link</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="va">weight</span><span class="op">)</span>, show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_node_point.html" class="external-link">geom_node_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">community</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_color_viridis</a></span><span class="op">(</span>discrete <span class="op">=</span> <span class="cn">TRUE</span>, option <span class="op">=</span> <span class="st">"D"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Sequence Network with Detected Communities"</span>,</span>
<span>    color <span class="op">=</span> <span class="st">"Detected\nCommunity"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="immApex_files/figure-html/unnamed-chunk-31-1.png" width="700"></p>
<p>The plot clearly visualizes the network structure, and the distinct
color groups demonstrate that the community detection algorithm
successfully identified and separated the three original sequence
families without any prior information. This unsupervised approach is a
powerful tool for exploring the hidden structure within a complex
repertoire.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
