<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Combining Immcantation and scRepertoire • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Combining Immcantation and scRepertoire">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.6.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Repertoire Analysis</li>
    <li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Working with Single-Cell Objects</li>
    <li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Deep Learning Integrations</li>
    <li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other Workflow Integrations</li>
    <li>
      <a href="../articles/immcantation.html">Combining Immcantation and scRepertoire</a>
    </li>
    <li>
      <a href="../articles/tcrpheno.html">TCRpheno applied to T cell fate</a>
    </li>
    <li>
      <a href="../articles/immunarch.html">Using scRepertoire with immunarch</a>
    </li>
    <li>
      <a href="../articles/dandelionR.html">VDJ Trajectory Analysis with dandelionR</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BorchLab/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Combining Immcantation and scRepertoire</h1>
            
            <h4 data-toc-skip class="date">Compiled: October 31,
2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BorchLab/scRepertoire/blob/HEAD/vignettes/articles/immcantation.Rmd" class="external-link"><code>vignettes/articles/immcantation.Rmd</code></a></small>
      <div class="hidden name"><code>immcantation.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>The <code>Immcantation</code> workflow is a suite of R
packages/python scripts designed for the analysis of B cell receptor
(BCR) sequencing data. This guide will walk you through the installation
and loading of the core packages required for analyzing single-cell BCR
data with accompanying interaction with <code>scRepertoire</code>. More
information on the <code>Immcantation</code> workflow is available at
their <a href="https://immcantation.readthedocs.io/en/stable/" class="external-link">website</a>.</p>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>Here is a brief overview of the role each package plays in the
<code>Immcantation</code> workflow:</p>
<ul>
<li>
<code>alakazam</code>: This package provides general tools for
working with AIRR (Adaptive Immune Receptor Repertoire) data. It
includes functions for data import/export, quality control, and basic
frequency analysis of V(D)J genes and clonotypes.</li>
<li>
<code>dowser</code>: After identifying clones with scoper, dowser is
used to perform phylogenetic analysis on the sequences within each
clone. This helps in reconstructing the lineage trees that trace the
affinity maturation process from the unmutated common ancestor.</li>
<li>
<code>scoper</code>: Used for inferring B cell clones from BCR
sequencing data. It implements several methods for partitioning
sequences into clonal groups based on shared VJ genes and junction
region similarity.</li>
<li>
<code>shazam</code>: This package focuses on the analysis of somatic
hypermutation (SHM) in immunoglobulin (Ig) sequences. Its functions
allow you to quantify mutation frequencies, analyze mutational spectra,
and measure the strength and nature of selection pressures.</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">packages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"alakazam"</span>, <span class="st">"shazam"</span>, <span class="st">"scoper"</span>, <span class="st">"dowser"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="va">packages</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h3>
<p>The <code>Immcantation</code> workflow has involved extensive
development across a range of methods and packages. If using the
vignette and packages, please read and cite the following:</p>
<div class="section level4">
<h4 id="alakazam">alakazam<a class="anchor" aria-label="anchor" href="#alakazam"></a>
</h4>
<pre><code>@article{gupta2015change,
  title={Change-O: a toolkit for analyzing large-scale B cell immunoglobulin repertoire sequencing data},
  author={Gupta, Namita T and Vander Heiden, Jason A and Uduman, Mohamed and Gadala-Maria, Daniel and Yaari, Gur and Kleinstein, Steven H},
  journal={Bioinformatics},
  volume={31},
  number={20},
  pages={3356--3358},
  year={2015},
  publisher={Oxford University Press}
}</code></pre>
</div>
<div class="section level4">
<h4 id="dowser">dowser<a class="anchor" aria-label="anchor" href="#dowser"></a>
</h4>
<pre><code># Original Publication
@article{hoehn2022phylogenetic,
  title={Phylogenetic analysis of migration, differentiation, and class switching in B cells},
  author={Hoehn, Kenneth B and Pybus, Oliver G and Kleinstein, Steven H},
  journal={PLoS computational biology},
  volume={18},
  number={4},
  pages={e1009885},
  year={2022},
  publisher={Public Library of Science San Francisco, CA USA}
}

# Methodology for paired BCR sequences
@article{jensen2024inferring,
  title={Inferring B cell phylogenies from paired H and L chain BCR sequences with Dowser},
  author={Jensen, Cole G and Sumner, Jacob A and Kleinstein, Steven H and Hoehn, Kenneth B},
  journal={The Journal of Immunology},
  volume={212},
  number={10},
  pages={1579--1588},
  year={2024},
  publisher={American Association of Immunologists}
}</code></pre>
</div>
<div class="section level4">
<h4 id="scoper">scoper<a class="anchor" aria-label="anchor" href="#scoper"></a>
</h4>
<pre><code>@article{nouri2018spectral,
  title={A spectral clustering-based method for identifying clones from high-throughput B cell repertoire sequencing data},
  author={Nouri, Nima and Kleinstein, Steven H},
  journal={Bioinformatics},
  volume={34},
  number={13},
  pages={i341--i349},
  year={2018},
  publisher={Oxford University Press}
}</code></pre>
</div>
<div class="section level4">
<h4 id="shazam">shazam<a class="anchor" aria-label="anchor" href="#shazam"></a>
</h4>
<pre><code># Selection analysis methods
@article{yaari2012quantifying,
  title={Quantifying selection in high-throughput Immunoglobulin sequencing data sets},
  author={Yaari, Gur and Uduman, Mohamed and Kleinstein, Steven H},
  journal={Nucleic acids research},
  volume={40},
  number={17},
  pages={e134--e134},
  year={2012},
  publisher={Oxford University Press}
}

# HH_S5F model and the targeting model generation methods
@article{yaari2013models,
  title={Models of somatic hypermutation targeting and substitution based on synonymous mutations from high-throughput immunoglobulin sequencing data},
  author={Yaari, Gur and Vander Heiden, Jason A and Uduman, Mohamed and Gadala-Maria, Daniel and Gupta, Namita and Stern, Joel NH and O’Connor, Kevin C and Hafler, David A and Laserson, Uri and Vigneault, Francois and others},
  journal={Frontiers in immunology},
  volume={4},
  pages={358},
  year={2013},
  publisher={Frontiers Media SA}
}</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="loading-packages-for-the-tutorial">Loading Packages for the Tutorial<a class="anchor" aria-label="anchor" href="#loading-packages-for-the-tutorial"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://alakazam.readthedocs.io/" class="external-link">alakazam</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dowser.readthedocs.io" class="external-link">dowser</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.amazon.com/Integration-Manipulation-Visualization-Phylogenetic-Computational-ebook/dp/B0B5NLZR1Z/" class="external-link">ggtree</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scoper.readthedocs.io" class="external-link">scoper</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.borch.dev/uploads/scRepertoire/">scRepertoire</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://shazam.readthedocs.io" class="external-link">shazam</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sjmgarnier.github.io/viridis/" class="external-link">viridis</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Using standard ggplot theme</span></span>
<span><span class="va">theme_vignette</span> <span class="op">&lt;-</span> <span class="fu">scRepertoire</span><span class="fu">:::</span><span class="va">.themeRepertoire</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="loading-bcr-data">Loading BCR Data<a class="anchor" aria-label="anchor" href="#loading-bcr-data"></a>
</h2>
<p>Data from this vignette is derived from <a href="https://pubmed.ncbi.nlm.nih.gov/33891889/" class="external-link">PMID: 33891889</a>,
which examines the immune dysregulation of multisystem inflammatory
syndrome in children following COVID-19 infection. Raw data from SRA was
downloaded and re-aligned using the standard Cell Ranger v9.0
workflow.</p>
<p>After processing the raw FASTA files with IgBLAST using the provided
bash script <a href="https://www.borch.dev/uploads/scripts/Immcantation_Bash.sh" class="external-link">Immcantation_Bash.sh</a>,
you will have a set of tab-separated value (<code>.tsv</code>) files.
The bash script automates the use of Change-O’s
<code>AssignGenes.py</code> and <code>MakeDb.py</code> tools, which
perform V(D)J gene assignment and construct a database file for each
sample. The output file we are interested in is
<code>all_contig_igblast_db-pass.tsv</code>.</p>
<p>Our first task in R is to locate all of these database files, load
them, and combine them into a single, clean data frame. The following
code block handles this process, performing several essential quality
control steps along the way to ensure the data is suitable for
downstream analysis.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># -------- Building BCR Data Base -------------------</span></span>
<span></span>
<span><span class="co"># Recursively find  files that match the igBlast output</span></span>
<span><span class="va">file.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span></span>
<span>  path <span class="op">=</span> <span class="st">"./Alignments"</span>,</span>
<span>  pattern <span class="op">=</span> <span class="st">"all_contig_igblast_db-pass.tsv"</span>,</span>
<span>  recursive <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generating Categorical Variables to use Downstream</span></span>
<span><span class="va">sample_id</span> <span class="op">&lt;-</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split</a></span><span class="op">(</span><span class="va">file.list</span>, <span class="st">"/"</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">subject_id</span> <span class="op">&lt;-</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="va">sample_id</span>, <span class="st">"PBMC_"</span><span class="op">)</span></span>
<span><span class="va">subject_id</span> <span class="op">&lt;-</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="va">subject_id</span>, <span class="st">"_1"</span><span class="op">)</span></span>
<span><span class="va">subject_id</span> <span class="op">&lt;-</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="va">subject_id</span>, <span class="st">"_2"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Iterative Processing and Filtering of Each Sample ---</span></span>
<span><span class="va">lib_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">file.list</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"./Alignments/"</span>, <span class="va">file.list</span><span class="op">[</span><span class="va">x</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">tmp</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">&lt;-</span> <span class="va">sample_id</span><span class="op">[</span><span class="va">x</span><span class="op">]</span></span>
<span>  <span class="va">tmp</span><span class="op">$</span><span class="va">subject_id</span> <span class="op">&lt;-</span> <span class="va">subject_id</span><span class="op">[</span><span class="va">x</span><span class="op">]</span></span>
<span>  <span class="va">tmp</span><span class="op">$</span><span class="va">cell_id</span> <span class="op">&lt;-</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split</a></span><span class="op">(</span><span class="va">tmp</span><span class="op">$</span><span class="va">sequence_id</span>, <span class="st">"_"</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="va">tmp</span><span class="op">$</span><span class="va">cell_id_unique</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">sample_id</span><span class="op">[</span><span class="va">x</span><span class="op">]</span>, <span class="st">"_"</span>, <span class="va">tmp</span><span class="op">$</span><span class="va">cell_id</span><span class="op">)</span></span>
<span>  </span>
<span>  </span>
<span>  <span class="co"># Data Cleaning and Filtering Steps </span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="va">tmp</span><span class="op">$</span><span class="va">productive</span> <span class="op">==</span> <span class="st">"TRUE"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">tmp</span> <span class="op">&lt;-</span> <span class="va">tmp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">productive</span> <span class="op">==</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="va">tmp</span><span class="op">$</span><span class="va">productive</span> <span class="op">==</span> <span class="st">"T"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">tmp</span> <span class="op">&lt;-</span> <span class="va">tmp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">productive</span> <span class="op">==</span> <span class="st">"T"</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="co"># Remove cells that have more than one heavy chain contig assigned.</span></span>
<span>  <span class="va">multi_heavy</span> <span class="op">&lt;-</span> <span class="va">tmp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">locus</span> <span class="op">==</span> <span class="st">"IGH"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co"># Isolate heavy chain records</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">cell_id</span><span class="op">)</span> <span class="co"># Count heavy chains per cell</span></span>
<span>  </span>
<span>  <span class="va">multi_heavy_cells</span> <span class="op">&lt;-</span> <span class="va">multi_heavy</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">n</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">cell_id</span><span class="op">)</span> <span class="co"># Extract the cell IDs as a vector</span></span>
<span>  </span>
<span>  <span class="co"># Filter out all records (both heavy and light) from these problematic cells.</span></span>
<span>  <span class="va">tmp</span> <span class="op">&lt;-</span> <span class="va">tmp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">cell_id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">multi_heavy_cells</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># Standardize isotype calls ('c_call').</span></span>
<span>  <span class="va">tmp</span><span class="op">$</span><span class="va">c_call</span> <span class="op">&lt;-</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split</a></span><span class="op">(</span><span class="va">tmp</span><span class="op">$</span><span class="va">c_call</span>, <span class="st">","</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span></span>
<span>  </span>
<span>  <span class="co"># Remove any records where an isotype could not be determined.</span></span>
<span>  <span class="va">tmp</span> <span class="op">&lt;-</span> <span class="va">tmp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">c_call</span> <span class="op">!=</span> <span class="st">""</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># Ensure all remaining cells have a paired heavy chain.</span></span>
<span>  <span class="va">heavy_cells</span> <span class="op">&lt;-</span> <span class="va">tmp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">locus</span> <span class="op">==</span> <span class="st">"IGH"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">cell_id</span><span class="op">)</span></span>
<span>  <span class="va">light_cells</span> <span class="op">&lt;-</span> <span class="va">tmp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">locus</span> <span class="op">==</span> <span class="st">"IGK"</span> <span class="op">|</span> <span class="va">locus</span> <span class="op">==</span> <span class="st">"IGL"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">cell_id</span><span class="op">)</span></span>
<span>  <span class="va">no_heavy_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="va">light_cells</span>, <span class="va">heavy_cells</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># Remove the orphan light chain records.</span></span>
<span>  <span class="va">tmp</span> <span class="op">&lt;-</span> <span class="va">tmp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">cell_id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">no_heavy_cells</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">tmp</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combining all the data together</span></span>
<span><span class="va">bcr_db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">lib_data</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">bcr_db</span>, <span class="st">"Immcantation_bcr_db.rds"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="determining-a-clonal-assignment-threshold">Determining a Clonal Assignment Threshold<a class="anchor" aria-label="anchor" href="#determining-a-clonal-assignment-threshold"></a>
</h2>
<p>The core task in B cell repertoire analysis is to group sequences
into “clones”—sets of cells that originated from the same initial B cell
and underwent affinity maturation. The <code>Immcantation</code>
workflow defines clones as sequences that use the same V and J genes and
have a high degree of similarity in their junction regions.</p>
<p>But how do we define “high similarity”? We need to determine a
specific, objective distance threshold. Sequences with a junction
similarity distance below this threshold will be considered part of the
same clone. This step uses an automated, data-driven approach to find
that threshold.</p>
<div class="section level4">
<h4 id="calculate-nearest-neighbor-distances-disttonearest">Calculate Nearest Neighbor Distances
(<code>distToNearest</code>)<a class="anchor" aria-label="anchor" href="#calculate-nearest-neighbor-distances-disttonearest"></a>
</h4>
<p>The first step is to calculate the Hamming distance (the number of
positions at which the characters are different) between the junction
sequence of each B cell and its “nearest neighbor.” The nearest neighbor
is the sequence from a different cell that has the most similar
junction.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Loading Data from Above Step</span></span>
<span><span class="va">bcr_db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://www.borch.dev/uploads/data/Immcantation_bcr_db.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Split by Subject Id</span></span>
<span><span class="va">dist_heavy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/shazam/man/distToNearest.html" class="external-link">distToNearest</a></span><span class="op">(</span><span class="va">bcr_db</span>, </span>
<span>                            cellIdColumn <span class="op">=</span> <span class="st">"cell_id_unique"</span>,</span>
<span>                            first <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                            onlyHeavy <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                            fields <span class="op">=</span> <span class="st">"subject_id"</span>,</span>
<span>                            nproc <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                            progress <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="find-the-threshold-automatically-findthreshold">Find the Threshold Automatically (<code>findThreshold</code>)<a class="anchor" aria-label="anchor" href="#find-the-threshold-automatically-findthreshold"></a>
</h3>
<p>The distribution of nearest-neighbor distances is typically bimodal.
There is a sharp peak of very small distances (sequences within the same
clone) and a broader peak of larger distances (sequences from different
clones). The <code>findThreshold</code> function from the
<code>scoper</code> package automatically finds the valley between these
two peaks.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Find threshold for cloning automatically</span></span>
<span><span class="va">threshold_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/shazam/man/findThreshold.html" class="external-link">findThreshold</a></span><span class="op">(</span><span class="va">dist_heavy</span><span class="op">$</span><span class="va">dist_nearest</span>, </span>
<span>                                  progress <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                  method <span class="op">=</span> <span class="st">"gmm"</span>, </span>
<span>                                  model <span class="op">=</span> <span class="st">"gamma-norm"</span>,</span>
<span>                                  cutoff <span class="op">=</span> <span class="st">"user"</span>, </span>
<span>                                  spc <span class="op">=</span> <span class="fl">0.995</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       STEP&gt; Parameter initialization</span></span>
<span><span class="co">##     VALUES&gt; 6104</span></span>
<span><span class="co">## ITERATIONS&gt; 4</span></span></code></pre>
<pre><code><span><span class="co">##       STEP&gt; Fitting gamma-norm</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">threshold_auto</span> <span class="op">&lt;-</span> <span class="va">threshold_output</span><span class="op">@</span><span class="va">threshold</span></span></code></pre></div>
<p>This function fits a Gaussian Mixture Model
(<code>method = "gmm"</code>) to the distance distribution. It assumes
the distribution is a mix of a gamma distribution (for the
clonally-related distances) and a normal distribution (for the unrelated
distances).</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">dist_heavy</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">dist_nearest</span><span class="op">)</span><span class="op">)</span>,</span>
<span>             <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">dist_nearest</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_vignette</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Hamming distance"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Count"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"white"</span>, binwidth <span class="op">=</span> <span class="fl">0.02</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="va">threshold_auto</span>, lty<span class="op">=</span><span class="fl">2</span>, color<span class="op">=</span><span class="st">'red'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="va">subject_id</span> <span class="op">~</span> <span class="va">.</span>, scales<span class="op">=</span><span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>strip.text.y.right <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="immcantation_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="defining-and-quantifying-clonal-groups">Defining and Quantifying Clonal Groups<a class="anchor" aria-label="anchor" href="#defining-and-quantifying-clonal-groups"></a>
</h2>
<p>With the distance threshold established, we can now formally
partition our sequences into clonal lineages. This step uses the
<code>scoper</code> package to apply the threshold and assign a unique
clone ID to each sequence. We will then visualize the resulting clone
size distributions, which is a fundamental way to characterize an immune
repertoire.</p>
<div class="section level3">
<h3 id="assigning-sequences-to-clones-hierarchicalclones">Assigning Sequences to Clones (<code>hierarchicalClones</code>)<a class="anchor" aria-label="anchor" href="#assigning-sequences-to-clones-hierarchicalclones"></a>
</h3>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># call clones based on heavy chain</span></span>
<span><span class="co"># keep light chain information for subclone calling later</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scoper/man/hierarchicalClones.html" class="external-link">hierarchicalClones</a></span><span class="op">(</span><span class="va">dist_heavy</span>, </span>
<span>                              cell_id <span class="op">=</span> <span class="st">'cell_id_unique'</span>,</span>
<span>                              threshold <span class="op">=</span> <span class="va">threshold_auto</span>,</span>
<span>                              only_heavy <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                              split_light <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                              summarize_clones <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                              nproc <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                              verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      MAX_N_FILTER&gt;  0 invalid junction(s) ( # of N &gt; 0 ) in the junction column removed.</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results_db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span></span>
<span><span class="va">results_db</span><span class="op">$</span><span class="va">sequence_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">results_db</span><span class="op">$</span><span class="va">sample_id</span>, <span class="st">"_"</span>, <span class="va">results_db</span><span class="op">$</span><span class="va">sequence_id</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="counting-clone-sizes-countclones">Counting Clone Sizes (<code>countClones</code>)<a class="anchor" aria-label="anchor" href="#counting-clone-sizes-countclones"></a>
</h3>
<p>Now that clones are defined, a common first analysis is to quantify
their sizes. How many cells belong to each clone?</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get clone sizes per subject</span></span>
<span><span class="va">clone_sizes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/alakazam/man/countClones.html" class="external-link">countClones</a></span><span class="op">(</span><span class="va">results_db</span>,</span>
<span>                           groups<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"locus"</span>, <span class="st">"subject_id"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">clone_sizes</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">locus</span><span class="op">==</span><span class="st">"IGH"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">seq_count</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"black"</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label<span class="op">=</span><span class="va">seq_count</span>, y<span class="op">=</span><span class="fl">1</span><span class="op">)</span>, vjust<span class="op">=</span><span class="fl">1.5</span>, size<span class="op">=</span><span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">subject_id</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_sqrt</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_vignette</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"Cells per clone"</span>, </span>
<span>         y<span class="op">=</span><span class="st">"# Clonal lineages"</span><span class="op">)</span></span></code></pre></div>
<p><img src="immcantation_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="germline-reconstruction-and-subclone-definition">Germline Reconstruction and Subclone Definition<a class="anchor" aria-label="anchor" href="#germline-reconstruction-and-subclone-definition"></a>
</h2>
<p>Now that we have defined our clonal groups based on heavy chain
similarity, the next steps are to prepare for mutation analysis and to
refine our clone definitions using the paired light chain information
available in single-cell data.</p>
<p>This involves two main goals:</p>
<ul>
<li>Reconstruct the unmutated germline sequence for each observed BCR.
This is the baseline against which we will measure somatic
hypermutation.</li>
<li>Define subclones by identifying groups of cells within a heavy-chain
clone that share the same light chain.</li>
</ul>
<div class="section level3">
<h3 id="reconstructing-germline-sequences-creategermlines">Reconstructing Germline Sequences
(<code>createGermlines</code>)<a class="anchor" aria-label="anchor" href="#reconstructing-germline-sequences-creategermlines"></a>
</h3>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in IMGT-gapped sequences</span></span>
<span><span class="va">references</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/dowser/man/readIMGT.html" class="external-link">readIMGT</a></span><span class="op">(</span>dir <span class="op">=</span> <span class="st">"~/share/germlines/imgt/human/vdj"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Read in 1213 from 17 fasta files"</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results_db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/dowser/man/createGermlines.html" class="external-link">createGermlines</a></span><span class="op">(</span><span class="va">results_db</span>, </span>
<span>                              reference <span class="op">=</span> <span class="va">references</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Data cleaning: Some downstream functions do not tolerate NA values in columns</span></span>
<span><span class="va">results_db</span><span class="op">$</span><span class="va">c_call</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">results_db</span><span class="op">$</span><span class="va">c_call</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"NA"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="defining-subclones-by-light-chain-resolvelightchains">Defining Subclones by Light Chain
(<code>resolveLightChains</code>)<a class="anchor" aria-label="anchor" href="#defining-subclones-by-light-chain-resolvelightchains"></a>
</h3>
<p>A key advantage of single-cell data is having paired heavy and light
chain information. While we defined our clones using only the heavy
chain, cells within that clone may use different light chains. These
represent distinct “subclones.” The <code>resolveLightChains</code>
function from <code>scoper</code> identifies these groupings.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Split clones into subclones by light chain</span></span>
<span><span class="va">subclones_db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/dowser/man/resolveLightChains.html" class="external-link">resolveLightChains</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">results_db</span>,</span>
<span>                                   cell <span class="op">=</span> <span class="st">"cell_id_unique"</span>, </span>
<span>                                   id <span class="op">=</span> <span class="st">"sequence_id"</span>, </span>
<span>                                   nproc <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># General Clean Up</span></span>
<span><span class="va">subclones_db</span><span class="op">$</span><span class="va">v_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split</a></span><span class="op">(</span><span class="va">subclones_db</span><span class="op">$</span><span class="va">v_call</span>, <span class="st">"[*]"</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">subclones_db</span><span class="op">$</span><span class="va">j_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split</a></span><span class="op">(</span><span class="va">subclones_db</span><span class="op">$</span><span class="va">j_call</span>, <span class="st">"[*]"</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">subclones_db</span><span class="op">$</span><span class="va">c_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split</a></span><span class="op">(</span><span class="va">subclones_db</span><span class="op">$</span><span class="va">c_call</span>, <span class="st">"[*]"</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">subclones_db</span> <span class="op">&lt;-</span> <span class="va">subclones_db</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>chain <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">locus</span> <span class="op">==</span> <span class="st">"IGH"</span>, <span class="st">"heavy"</span>, <span class="st">"light"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">subclones_db</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">subject_id</span>, <span class="va">chain</span>, <span class="va">c_gene</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">tally</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">subject_id</span>, y<span class="op">=</span><span class="va">n</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">c_gene</span><span class="op">)</span><span class="op">)</span>, </span>
<span>           stat<span class="op">=</span><span class="st">"identity"</span>, position<span class="op">=</span><span class="st">"stack"</span>, color <span class="op">=</span> <span class="st">"black"</span>, lwd <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">chain</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_fill_viridis</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"B"</span>, discrete <span class="op">=</span> <span class="cn">TRUE</span>, name<span class="op">=</span><span class="st">"Isotype"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"Subject"</span>, y<span class="op">=</span><span class="st">"# BCRs"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_vignette</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="immcantation_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="preparing-data-for-lineage-tree-analysis">Preparing Data for Lineage Tree Analysis<a class="anchor" aria-label="anchor" href="#preparing-data-for-lineage-tree-analysis"></a>
</h2>
<p>The ultimate goal of many BCR repertoire analyses is to reconstruct
the evolutionary history of each B cell clone. This is done by building
phylogenetic trees that trace the path of somatic hypermutation from the
unmutated germline ancestor to the observed, mature B cells.</p>
<p>The <code>dowser</code> package is used for this phylogenetic
analysis, but it requires the data to be in a specific format: a
<code>ScoperClones</code> object. This step uses the
<code>formatClones</code> function to perform some final data cleaning
and to create this specialized object.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Remove any rows where junction_length is NA, as this is a critical field.</span></span>
<span><span class="va">subclones_db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">subclones_db</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">junction_length</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Replace NA values in key annotation columns with the string "NA".</span></span>
<span><span class="va">subclones_db</span><span class="op">$</span><span class="va">c_gene</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">subclones_db</span><span class="op">$</span><span class="va">c_gene</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"NA"</span></span>
<span><span class="va">subclones_db</span><span class="op">$</span><span class="va">v_gene</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">subclones_db</span><span class="op">$</span><span class="va">v_gene</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"NA"</span></span>
<span><span class="va">subclones_db</span><span class="op">$</span><span class="va">j_gene</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">subclones_db</span><span class="op">$</span><span class="va">j_gene</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"NA"</span></span>
<span><span class="va">subclones_db</span><span class="op">$</span><span class="va">vj_gene</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">subclones_db</span><span class="op">$</span><span class="va">vj_gene</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"NA"</span></span>
<span></span>
<span><span class="va">vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split</a></span><span class="op">(</span><span class="va">subclones_db</span><span class="op">$</span><span class="va">cell_id_unique</span>, <span class="st">"_"</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span></span>
<span><span class="va">subclones_db</span><span class="op">$</span><span class="va">timepoint</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">vec</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1"</span>, <span class="st">"2"</span><span class="op">)</span>, <span class="va">vec</span>, <span class="st">"1"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">clones_hl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/dowser/man/formatClones.html" class="external-link">formatClones</a></span><span class="op">(</span><span class="va">subclones_db</span>,</span>
<span>                          cell <span class="op">=</span> <span class="st">'cell_id_unique'</span>,</span>
<span>                          chain <span class="op">=</span> <span class="st">"HL"</span>, </span>
<span>                          split_light <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          traits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"subject_id"</span>, <span class="st">"c_gene"</span><span class="op">)</span>,</span>
<span>                          columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"c_gene"</span>, <span class="st">"v_gene"</span>, <span class="st">"j_gene"</span>, <span class="st">"vj_gene"</span>, <span class="st">"subject_id"</span>, <span class="st">"timepoint"</span><span class="op">)</span>,</span>
<span>                          minseq <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                          verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="building-and-visualizing-clonal-trees">Building and Visualizing Clonal Trees<a class="anchor" aria-label="anchor" href="#building-and-visualizing-clonal-trees"></a>
</h2>
<p>The culmination of the <code>Immcantation</code> workflow is the
reconstruction and visualization of clonal lineage trees. These trees
map the evolutionary relationships between sequences in a clone,
starting from the unmutated germline ancestor and tracing the branches
of somatic hypermutation and class-switching that occur during affinity
maturation.</p>
<p>First, we use <code><a href="https://rdrr.io/pkg/dowser/man/getTrees.html" class="external-link">getTrees()</a></code> function to build a maximum
parsimony tree for each clone, and then we scale the branches to
represent the number of mutations. We will then use
<code><a href="https://rdrr.io/pkg/dowser/man/collapseNodes.html" class="external-link">collapseNodes()</a></code> to cleanup the trees and
<code><a href="https://rdrr.io/pkg/dowser/man/scaleBranches.html" class="external-link">scaleBranches()</a></code> to enhanced the visual comparison of the
tree nodes.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clones_hl</span><span class="op">$</span><span class="va">num_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">clones_hl</span><span class="op">$</span><span class="va">data</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">@</span><span class="va">data</span><span class="op">$</span><span class="va">collapse_count</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># --- Build trees for each clone ---</span></span>
<span><span class="va">trees</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/dowser/man/getTrees.html" class="external-link">getTrees</a></span><span class="op">(</span><span class="va">clones_hl</span><span class="op">)</span></span>
<span><span class="va">trees</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/dowser/man/collapseNodes.html" class="external-link">collapseNodes</a></span><span class="op">(</span><span class="va">trees</span><span class="op">)</span></span>
<span><span class="va">trees_m</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/dowser/man/scaleBranches.html" class="external-link">scaleBranches</a></span><span class="op">(</span><span class="va">trees</span>, edge_type<span class="op">=</span><span class="st">"mutations"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="visualizing-clonal-trees">Visualizing Clonal Trees<a class="anchor" aria-label="anchor" href="#visualizing-clonal-trees"></a>
</h3>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># --- Generate and Customize Tree Plots ---</span></span>
<span><span class="co"># Define a color palette for isotypes for consistent plotting</span></span>
<span><span class="va">isotypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"IGHA1"</span>, <span class="st">"IGHA2"</span>, <span class="st">"IGHD"</span>, <span class="st">"IGHG1"</span>, <span class="st">"IGHG2"</span>, <span class="st">"IGHG3"</span>, <span class="st">"IGHG4"</span>, <span class="st">"IGHM"</span>, <span class="st">"Germline"</span><span class="op">)</span></span>
<span><span class="va">col_pal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/viridis_pal.html" class="external-link">viridis_pal</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"B"</span><span class="op">)</span><span class="op">(</span><span class="fl">8</span><span class="op">)</span>, <span class="st">"grey"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">col_pal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">isotypes</span></span>
<span></span>
<span><span class="co"># plotTrees generates a list of ggplot objects, one for each tree.</span></span>
<span><span class="va">plots_m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/dowser/man/plotTrees.html" class="external-link">plotTrees</a></span><span class="op">(</span><span class="va">trees_m</span>, scale<span class="op">=</span><span class="fl">1</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">plots_m</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">trees_m</span><span class="op">$</span><span class="va">clone_id</span></span>
<span></span>
<span><span class="co"># --- Example: Plotting a Single Interesting Tree ---</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="va">trees_m</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_max</a></span><span class="op">(</span>order_by <span class="op">=</span> <span class="va">seqs</span>, n <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">clone_id</span><span class="op">)</span></span>
<span><span class="va">example_tree_plot</span> <span class="op">&lt;-</span> <span class="va">plots_m</span><span class="op">[[</span><span class="va">idx</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">max_x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">example_tree_plot</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">x</span><span class="op">)</span> <span class="co"># Find x-axis limit</span></span>
<span></span>
<span><span class="va">example_tree_plot</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Lineage Tree for Clone:"</span>, <span class="va">idx</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tippoint.html" class="external-link">geom_tippoint</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">c_gene</span>, </span>
<span>                      size <span class="op">=</span> <span class="va">collapse_count</span><span class="op">)</span>, </span>
<span>                  alpha <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">col_pal</span>, name <span class="op">=</span> <span class="st">"Isotype"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html" class="external-link">scale_size</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"# Cells"</span>, range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">max_x</span> <span class="op">+</span> <span class="op">(</span><span class="va">max_x</span> <span class="op">*</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_vignette</span><span class="op">(</span>grid_lines <span class="op">=</span> <span class="st">"no"</span><span class="op">)</span></span></code></pre></div>
<p><img src="immcantation_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="quantifying-and-analyzing-somatic-hypermutation">Quantifying and Analyzing Somatic Hypermutation<a class="anchor" aria-label="anchor" href="#quantifying-and-analyzing-somatic-hypermutation"></a>
</h2>
<p>A hallmark of the adaptive immune response is somatic hypermutation
(SHM), the process by which B cells introduce mutations into their
antibody genes to improve antigen binding. Quantifying the frequency,
location, and type of these mutations is a primary goal of BCR
repertoire analysis. This step uses the shazam package to calculate SHM
and explore its patterns.</p>
<div class="section level3">
<h3 id="calculate-somatic-hypermutation-observedmutations">Calculate Somatic Hypermutation
(<code>observedMutations</code>)<a class="anchor" aria-label="anchor" href="#calculate-somatic-hypermutation-observedmutations"></a>
</h3>
<p>The <code><a href="https://rdrr.io/pkg/shazam/man/observedMutations.html" class="external-link">observedMutations()</a></code> function is the workhorse for
all SHM analysis. It compares each observed sequence to its
reconstructed germline sequence (which we created in Step 4) and tallies
the mutations. The function is called multiple times to calculate
different metrics. It compares the <code>sequence_alignment</code>
(observed sequence) to the <code>germline_alignment_d_mask</code>
(unmutated ancestor).</p>
<p>Key Parameter(s) for <code><a href="https://rdrr.io/pkg/shazam/man/observedMutations.html" class="external-link">observedMutations()</a></code></p>
<ul>
<li>
<code>regionDefinition = IMGT_VDJ</code>: This tells the function to
use the standard IMGT definitions to identify Framework Regions (FWRs)
and Complementarity-Determining Regions (CDRs). CDRs are the
hypervariable loops that directly contact the antigen, while FWRs
provide the structural scaffold.</li>
<li>
<code>frequency = TRUE</code>: Calculates the mutation rate
(mutations per 100 base pairs). If FALSE, it calculates the raw count of
mutations.</li>
<li>
<code>combine = TRUE</code>: This calculates a single mutation value
across the entire VDJ segment. If FALSE, it calculates mutations
separately for each sub-region (e.g., <code>mu_freq_cdr_s</code> for
silent mutations in the CDR, <code>mu_freq_fwr_r</code> for replacement
mutations in the FWR, etc.).</li>
</ul>
<p>After running these commands, your shm_db data frame will be
populated with many new columns (e.g., <code>mu_freq</code>,
<code>mu_count</code>, <code>mu_freq_cdr_s</code>,
<code>mu_freq_cdr_r</code>, etc) that quantify SHM in great detail.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Detailed Mutation Counts by Region</span></span>
<span><span class="va">shm_db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/shazam/man/observedMutations.html" class="external-link">observedMutations</a></span><span class="op">(</span><span class="va">subclones_db</span>,</span>
<span>                            sequenceColumn <span class="op">=</span> <span class="st">"sequence_alignment"</span>,</span>
<span>                            germlineColumn <span class="op">=</span> <span class="st">"germline_alignment_d_mask"</span>,</span>
<span>                            regionDefinition <span class="op">=</span> <span class="va">IMGT_VDJ</span>,</span>
<span>                            frequency <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                            combine <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                            nproc <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Total Mutation Count (Combined VDJ regions)</span></span>
<span><span class="va">shm_db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/shazam/man/observedMutations.html" class="external-link">observedMutations</a></span><span class="op">(</span><span class="va">shm_db</span>,</span>
<span>                            sequenceColumn <span class="op">=</span> <span class="st">"sequence_alignment"</span>,</span>
<span>                            germlineColumn <span class="op">=</span> <span class="st">"germline_alignment_d_mask"</span>,</span>
<span>                            regionDefinition <span class="op">=</span> <span class="va">IMGT_VDJ</span>,</span>
<span>                            frequency <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                            combine <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                            nproc <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Detailed Mutation Frequency by Region</span></span>
<span><span class="va">shm_db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/shazam/man/observedMutations.html" class="external-link">observedMutations</a></span><span class="op">(</span><span class="va">shm_db</span>,</span>
<span>                            sequenceColumn <span class="op">=</span> <span class="st">"sequence_alignment"</span>,</span>
<span>                            germlineColumn <span class="op">=</span> <span class="st">"germline_alignment_d_mask"</span>,</span>
<span>                            regionDefinition <span class="op">=</span> <span class="va">IMGT_VDJ</span>,</span>
<span>                            frequency <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                            combine <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                            nproc <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Total Mutation Frequency (Combined VDJ regions</span></span>
<span><span class="va">shm_db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/shazam/man/observedMutations.html" class="external-link">observedMutations</a></span><span class="op">(</span><span class="va">shm_db</span>,</span>
<span>                            sequenceColumn <span class="op">=</span> <span class="st">"sequence_alignment"</span>,</span>
<span>                            germlineColumn <span class="op">=</span> <span class="st">"germline_alignment_d_mask"</span>,</span>
<span>                            regionDefinition <span class="op">=</span> <span class="va">IMGT_VDJ</span>,</span>
<span>                            frequency <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                            combine <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                            nproc <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="visualizing-replacement-and-silent-mutations-in-cdrs">Visualizing Replacement and Silent Mutations in CDRs<a class="anchor" aria-label="anchor" href="#visualizing-replacement-and-silent-mutations-in-cdrs"></a>
</h4>
<p>A key way to infer positive or negative selection is to compare the
rate of replacement (<strong>R</strong>) mutations (which change the
amino acid sequence) to silent (<strong>S</strong>) mutations (which do
not). High levels of <strong>R</strong> mutations in the antigen-binding
CDRs are a classic sign of positive selection for improved binding.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Clean up isotype calls for plotting</span></span>
<span><span class="va">shm_db</span><span class="op">$</span><span class="va">c_call</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split</a></span><span class="op">(</span><span class="va">shm_db</span><span class="op">$</span><span class="va">c_call</span>, <span class="st">"[*]"</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Plot silent mutation frequency in the CDRs</span></span>
<span><span class="va">shm_db</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">locus</span> <span class="op">==</span> <span class="st">"IGH"</span> <span class="op">&amp;</span> <span class="va">c_call</span> <span class="op">!=</span> <span class="st">"NA"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">subject_id</span>, y<span class="op">=</span><span class="va">mu_freq_cdr_s</span>, fill<span class="op">=</span><span class="va">c_call</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>          <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>outlier.alpha <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span> </span>
<span>          <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"CDR silent mutations"</span>, </span>
<span>               x <span class="op">=</span> <span class="st">"Isotype"</span>, y <span class="op">=</span> <span class="st">"Mutation frequency"</span><span class="op">)</span> <span class="op">+</span></span>
<span>          <span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_fill_viridis</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"B"</span>, discrete <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>          <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="va">.</span><span class="op">~</span><span class="va">c_call</span>, scales<span class="op">=</span><span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>     <span class="fu">theme_vignette</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, vjust <span class="op">=</span> <span class="fl">0.5</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span>  <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="immcantation_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot replacement mutation frequency in the CDRs</span></span>
<span><span class="va">shm_db</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">locus</span> <span class="op">==</span> <span class="st">"IGH"</span> <span class="op">&amp;</span> <span class="va">c_call</span> <span class="op">!=</span> <span class="st">"NA"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">subject_id</span>, y<span class="op">=</span><span class="va">mu_freq_cdr_r</span>, fill<span class="op">=</span><span class="va">c_call</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>          <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>outlier.alpha <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span> </span>
<span>          <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"CDR replacement mutations"</span>, </span>
<span>               x <span class="op">=</span> <span class="st">"Isotype"</span>, y <span class="op">=</span> <span class="st">"Mutation frequency"</span><span class="op">)</span> <span class="op">+</span></span>
<span>          <span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_fill_viridis</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"B"</span>, discrete <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>          <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="va">.</span><span class="op">~</span><span class="va">c_call</span>, scales<span class="op">=</span><span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>     <span class="fu">theme_vignette</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, vjust <span class="op">=</span> <span class="fl">0.5</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span>  <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="immcantation_files/figure-html/unnamed-chunk-16-2.png" width="700"></p>
<p>This boxplot displays the frequency of replacement mutations
specifically within the CDRs of heavy chains. It is faceted by isotype
(<code>c_call</code>), allowing you to compare SHM levels in naive
(IgM/IgD) versus mature (IgG/IgA) B cells. You would typically expect to
see higher R mutation frequencies in class-switched isotypes, indicating
that these clones have undergone selection.</p>
</div>
<div class="section level4">
<h4 id="visualizing-overall-shm-distribution">Visualizing Overall SHM Distribution<a class="anchor" aria-label="anchor" href="#visualizing-overall-shm-distribution"></a>
</h4>
<p>Finally, we can look at the overall distribution of SHM across all
clones to get a global picture of the repertoire’s maturity.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First, calculate the mean mutation frequency for each clone</span></span>
<span><span class="va">mut_freq_clone</span> <span class="op">&lt;-</span> <span class="va">shm_db</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">locus</span> <span class="op">==</span> <span class="st">"IGH"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">clone_id</span>, <span class="va">subject_id</span>, <span class="va">c_call</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>mean_mut_freq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">mu_freq</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Order isotypes for consistent plotting</span></span>
<span><span class="va">isotypes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"IGHA1"</span>, <span class="st">"IGHA2"</span>, <span class="st">"IGHD"</span>, <span class="st">"IGHE"</span>, <span class="st">"IGHG1"</span>, <span class="st">"IGHG2"</span>, <span class="st">"IGHG3"</span>, <span class="st">"IGHG4"</span>, <span class="st">"IGHM"</span><span class="op">)</span></span>
<span><span class="va">mut_freq_clone</span><span class="op">$</span><span class="va">c_call</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">mut_freq_clone</span><span class="op">$</span><span class="va">c_call</span>, levels<span class="op">=</span><span class="va">isotypes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a density plot of mutation frequencies</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">mut_freq_clone</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">c_call</span> <span class="op">!=</span> <span class="st">"NA"</span><span class="op">)</span>, </span>
<span>       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">mean_mut_freq</span>, fill <span class="op">=</span> <span class="va">c_call</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="va">c_call</span><span class="op">~</span><span class="va">subject_id</span>, scales<span class="op">=</span><span class="st">"free"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_fill_viridis</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"B"</span>, discrete <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"Mean mutation frequency per clone"</span>, y<span class="op">=</span><span class="st">"Density"</span>,</span>
<span>       title<span class="op">=</span><span class="st">"Somatic Hypermutation Rate by Isotype and Subject"</span>, </span>
<span>       fill <span class="op">=</span> <span class="st">"Isotype"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_vignette</span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="immcantation_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<p>This series of density plots shows the distribution of SHM
frequencies for each isotype within each subject. The x-axis represents
the mutation frequency. A peak near zero indicates a large population of
naive or unmutated cells (such as expected for IGHM and IGHD). In a
mature immune response, you would expect to see the peaks for IgG and
IgA shifted to the right, indicating higher overall mutation levels.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="interaction-with-single-cell-object">Interaction with Single-Cell Object<a class="anchor" aria-label="anchor" href="#interaction-with-single-cell-object"></a>
</h2>
<p>In this final step, we will add the clonal and SHM information we
derived from the <code>Immcantation</code> workflow into a Seurat object
and then use the powerful tools in the <code>scRepertoire</code> package
to perform integrated analyses.</p>
<p>More information on the upstream processing to create the Seurat
Object can be obtained from this <a href="https://www.borch.dev/uploads/scripts/Immcantation_Seurat.R" class="external-link">R
script</a>.</p>
<div class="section level3">
<h3 id="merging-immcantation-data-into-the-seurat-object">Merging Immcantation Data into the Seurat Object<a class="anchor" aria-label="anchor" href="#merging-immcantation-data-into-the-seurat-object"></a>
</h3>
<p>Our first task is to add the Immcantation-derived metadata to the
Seurat object. The key is the unique cell identifier
(<code>cell_id_unique</code>) we created earlier, which matches the cell
barcodes in the Seurat object.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Loading the Seurat Object</span></span>
<span><span class="va">SeuratObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://www.borch.dev/uploads/data/Immcantation_SeuratObject.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Preparing Our Immcantation Clonal Info</span></span>
<span><span class="va">Immc.df</span> <span class="op">&lt;-</span> <span class="va">shm_db</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">chain</span> <span class="op">==</span> <span class="st">"heavy"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">cell_id_unique</span>,</span>
<span>           <span class="va">clone_id</span>,</span>
<span>           <span class="va">subject_id</span>,</span>
<span>           new_mu_count <span class="op">=</span> <span class="va">mu_count</span>,</span>
<span>           new_mu_freq <span class="op">=</span> <span class="va">mu_freq</span>,</span>
<span>           <span class="va">mu_count_cdr_r</span>,</span>
<span>           <span class="va">mu_count_cdr_s</span>,</span>
<span>           <span class="va">mu_count_fwr_r</span>,</span>
<span>           <span class="va">mu_count_fwr_s</span>,</span>
<span>           <span class="va">mu_freq_cdr_r</span>,</span>
<span>           <span class="va">mu_freq_cdr_s</span>,</span>
<span>           <span class="va">mu_freq_fwr_r</span>,</span>
<span>           <span class="va">mu_freq_fwr_s</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>clone_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">subject_id</span>, <span class="st">"_"</span>, <span class="va">clone_id</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span><span class="st">"cell_id_unique"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add Metadata to Seurat Object </span></span>
<span><span class="va">SeuratObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/AddMetaData.html" class="external-link">AddMetaData</a></span><span class="op">(</span><span class="va">SeuratObject</span>, <span class="va">Immc.df</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualizing-clonal-distribution-on-the-umap">Visualizing Clonal Distribution on the UMAP<a class="anchor" aria-label="anchor" href="#visualizing-clonal-distribution-on-the-umap"></a>
</h3>
<p>We can now immediately visualize where our B cell clones and the
mutational frequency along the UMAP projection, which is organized by
transcriptional similarity.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">SeuratObject</span>, </span>
<span>                 label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>            <span class="fu">theme_vignette</span><span class="op">(</span>grid_lines <span class="op">=</span> <span class="st">"no"</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">SeuratObject</span>, </span>
<span>                 group.by <span class="op">=</span> <span class="st">"clone_id"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>            <span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_color_viridis</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"B"</span>, discrete <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>            <span class="fu">theme_vignette</span><span class="op">(</span>grid_lines <span class="op">=</span> <span class="st">"no"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">plot3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">SeuratObject</span>, </span>
<span>                     features <span class="op">=</span> <span class="st">"new_mu_freq"</span><span class="op">)</span><span class="op">+</span> </span>
<span>            <span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_color_viridis</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"B"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>            <span class="fu">theme_vignette</span><span class="op">(</span>grid_lines <span class="op">=</span> <span class="st">"no"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">plot1</span> <span class="op">+</span> <span class="va">plot2</span> <span class="op">+</span> <span class="va">plot3</span></span></code></pre></div>
<p><img src="immcantation_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="quantitative-analysis-with-screpertoire">Quantitative Analysis with scRepertoire<a class="anchor" aria-label="anchor" href="#quantitative-analysis-with-screpertoire"></a>
</h3>
<p><code>scRepertoire</code> provides a suite of functions designed for
quantitative analysis of single-cell immune repertoires and functions
with custom clonal definitions, such as <code>clone_id</code>.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalQuant.html">clonalQuant</a></span><span class="op">(</span><span class="va">SeuratObject</span>, </span>
<span>            group.by <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>            cloneCall<span class="op">=</span><span class="st">"clone_id"</span><span class="op">)</span></span></code></pre></div>
<p><img src="immcantation_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalDiversity.html">clonalDiversity</a></span><span class="op">(</span><span class="va">SeuratObject</span>, </span>
<span>                group.by <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>                cloneCall<span class="op">=</span><span class="st">"clone_id"</span>, </span>
<span>                metric <span class="op">=</span> <span class="st">"shannon"</span><span class="op">)</span></span></code></pre></div>
<p><img src="immcantation_files/figure-html/unnamed-chunk-20-2.png" width="700"></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/percentAA.html">percentAA</a></span><span class="op">(</span><span class="va">SeuratObject</span>, </span>
<span>          group.by <span class="op">=</span> <span class="st">"subject_id"</span>,</span>
<span>          chain <span class="op">=</span> <span class="st">"IGH"</span>, </span>
<span>          aa.length <span class="op">=</span> <span class="fl">20</span>, </span>
<span>          base_size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span></code></pre></div>
<p><img src="immcantation_files/figure-html/unnamed-chunk-20-3.png" width="700"></p>
<p>By integrating the detailed clonal definitions from Immcantation with
the powerful single-cell analysis tools in Seurat and scRepertoire, you
can now perform a truly comprehensive analysis, linking the evolution of
the antibody sequence directly to the dynamic transcriptional state of
the B cell.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
