<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Loading Data into scRepertoire â€¢ scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Loading Data into scRepertoire">
<meta property="og:description" content="scRepertoire">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Loading Data into scRepertoire</h1>
            
            <h4 data-toc-skip class="date">Compiled: January 10,
2024</h4>
      
      
      <div class="hidden name"><code>Loading.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="loading-and-processing-contig-data">Loading and Processing Contig Data<a class="anchor" aria-label="anchor" href="#loading-and-processing-contig-data"></a>
</h2>
<div class="section level3">
<h3 id="what-data-to-load-into-screpertoire">What data to load into scRepertoire?<a class="anchor" aria-label="anchor" href="#what-data-to-load-into-screpertoire"></a>
</h3>
<p>scRepertoire functions using the
<strong>filtered_contig_annotations.csv</strong> output from the 10x
Genomics Cell Ranger. This file is located in the ./outs/ directory of
the VDJ alignment folder. To generate a list of contigs to use for
scRepertoire:</p>
<ul>
<li>load the <strong>filtered_contig_annotations.csv</strong> for each
of the samples.<br>
</li>
<li>make a list in the R environment.</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">S1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">".../Sample1/outs/filtered_contig_annotations.csv"</span><span class="op">)</span></span>
<span><span class="va">S2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">".../Sample2/outs/filtered_contig_annotations.csv"</span><span class="op">)</span></span>
<span><span class="va">S3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">".../Sample3/outs/filtered_contig_annotations.csv"</span><span class="op">)</span></span>
<span><span class="va">S4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">".../Sample4/outs/filtered_contig_annotations.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">contig_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">S1</span>, <span class="va">S2</span>, <span class="va">S3</span>, <span class="va">S4</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="other-alignment-workflows">Other alignment workflows<a class="anchor" aria-label="anchor" href="#other-alignment-workflows"></a>
</h2>
<p>Beyond the default 10x Genomic Cell Ranger pipeline outputs,
scRepertoire supports the following <strong>single-cell</strong>
formats:</p>
<ul>
<li>AIRR<br>
</li>
<li>BD Rhapsody Multiomic Immune Profiling<br>
</li>
<li>
<a href="https://immcantation.readthedocs.io/en/stable/index.html" class="external-link">Immcantation</a><br>
</li>
<li>JSON-formatted contig data<br>
</li>
<li>
<a href="https://pubmed.ncbi.nlm.nih.gov/25924071/" class="external-link">MiXCR</a><br>
</li>
<li>
<a href="https://www.omniscope.ai/" class="external-link">Omniscope</a><br>
</li>
<li>
<a href="https://pubmed.ncbi.nlm.nih.gov/33986545/" class="external-link">TRUST4</a><br>
</li>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/35674381/" class="external-link">WAT3R</a></li>
</ul>
<p><code><a href="../reference/loadContigs.html">loadContigs()</a></code> can be given a directory where the
sequencing experiments are located and it will recursively load and
process the contig data based on the file names. Alternatively,
<code><a href="../reference/loadContigs.html">loadContigs()</a></code> can be given a list of data frames and
process the contig data</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Directory example</span></span>
<span><span class="va">contig.output</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"~/Documents/MyExperiment"</span><span class="op">)</span></span>
<span><span class="va">contig.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadContigs.html">loadContigs</a></span><span class="op">(</span>input <span class="op">=</span> <span class="va">contig.output</span>, </span>
<span>                           format <span class="op">=</span> <span class="st">"TRUST4"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#List of data frames example</span></span>
<span><span class="va">S1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"~/Documents/MyExperiment/Sample1/outs/barcode_results.csv"</span><span class="op">)</span></span>
<span><span class="va">S2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"~/Documents/MyExperiment/Sample2/outs/barcode_results.csv"</span><span class="op">)</span></span>
<span><span class="va">S3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"~/Documents/MyExperiment/Sample3/outs/barcode_results.csv"</span><span class="op">)</span></span>
<span><span class="va">S4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"~/Documents/MyExperiment/Sample4/outs/barcode_results.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">contig_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">S1</span>, <span class="va">S2</span>, <span class="va">S3</span>, <span class="va">S4</span><span class="op">)</span></span>
<span><span class="va">contig.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadContigs.html">loadContigs</a></span><span class="op">(</span>input <span class="op">=</span> <span class="va">contig.output</span>, </span>
<span>                           format <span class="op">=</span> <span class="st">"WAT3R"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="multiplexed-experiment">Multiplexed Experiment<a class="anchor" aria-label="anchor" href="#multiplexed-experiment"></a>
</h2>
<p>It is now easy to create the contig list from a multiplexed
experiment by first generating a single-cell RNA object (either Seurat
or Single Cell Experiment), loading the filtered contig file and then
using <code><a href="../reference/createHTOContigList.html">createHTOContigList()</a></code>. This function will return a
list separated by the <strong>group.by</strong> variable(s).</p>
<p>This function depends on the match of barcodes between the
single-cell object and contigs. If there is a prefix or different suffix
added to the barcode, this will result in no contigs recovered.
Currently, it is recommended you do this step before the integration, as
integration workflows commonly alter the barcodes. There is a
<strong>multi.run</strong> variable that can be used on the integrated
object. However, it assumes you have modified the barcodes with the
Seurat pipeline (automatic addition of _# to end), and your contig list
is in the same order.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">contigs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">".../outs/filtered_contig_annotations.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">contig.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createHTOContigList.html">createHTOContigList</a></span><span class="op">(</span><span class="va">contigs</span>, </span>
<span>                                   <span class="va">Seurat.Obj</span>, </span>
<span>                                   group.by <span class="op">=</span> <span class="st">"HTO_maxID"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="example-data-in-screpertoire">Example Data in scRepertoire<a class="anchor" aria-label="anchor" href="#example-data-in-screpertoire"></a>
</h2>
<p>scRepertoire comes with a data set from T cells derived from four
patients with acute respiratory distress to demonstrate the
functionality of the R package. More information on the data set can be
found in the corresponding <a href="https://pubmed.ncbi.nlm.nih.gov/33622974/" class="external-link">manuscript</a>. The
samples consist of paired peripheral-blood (B) and bronchoalveolar
lavage (L), effectively creating 8 distinct runs for T cell receptor
(TCR) enrichment. We can preview the elements in the list by using the
head function and looking at the first contig annotation.</p>
<p>The built-in example data is derived from the 10x Cell Ranger
pipeline, so it is ready to go for downstream processing and
analysis.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"contig_list"</span><span class="op">)</span> <span class="co">#the data built into scRepertoire</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">contig_list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##              barcode is_cell                   contig_id high_confidence length</span></span>
<span><span class="co">## 1 AAACCTGAGTACGACG-1    True AAACCTGAGTACGACG-1_contig_1            True    500</span></span>
<span><span class="co">## 2 AAACCTGAGTACGACG-1    True AAACCTGAGTACGACG-1_contig_2            True    478</span></span>
<span><span class="co">## 4 AAACCTGCAACACGCC-1    True AAACCTGCAACACGCC-1_contig_1            True    506</span></span>
<span><span class="co">## 5 AAACCTGCAACACGCC-1    True AAACCTGCAACACGCC-1_contig_2            True    470</span></span>
<span><span class="co">## 6 AAACCTGCAGGCGATA-1    True AAACCTGCAGGCGATA-1_contig_1            True    558</span></span>
<span><span class="co">## 7 AAACCTGCAGGCGATA-1    True AAACCTGCAGGCGATA-1_contig_2            True    505</span></span>
<span><span class="co">##   chain       v_gene d_gene  j_gene c_gene full_length productive</span></span>
<span><span class="co">## 1   TRA       TRAV25   None  TRAJ20   TRAC        True       True</span></span>
<span><span class="co">## 2   TRB      TRBV5-1   None TRBJ2-7  TRBC2        True       True</span></span>
<span><span class="co">## 4   TRA TRAV38-2/DV8   None  TRAJ52   TRAC        True       True</span></span>
<span><span class="co">## 5   TRB     TRBV10-3   None TRBJ2-2  TRBC2        True       True</span></span>
<span><span class="co">## 6   TRA     TRAV12-1   None   TRAJ9   TRAC        True       True</span></span>
<span><span class="co">## 7   TRB        TRBV9   None TRBJ2-2  TRBC2        True       True</span></span>
<span><span class="co">##                 cdr3                                                cdr3_nt</span></span>
<span><span class="co">## 1        CGCSNDYKLSF                      TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT</span></span>
<span><span class="co">## 2     CASSLTDRTYEQYF             TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC</span></span>
<span><span class="co">## 4 CAYRSAQAGGTSYGKLTF TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT</span></span>
<span><span class="co">## 5      CAISEQGKGELFF                TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 6     CVVSDNTGGFKTIF             TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT</span></span>
<span><span class="co">## 7  CASSVRRERANTGELFF    TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">##   reads umis raw_clonotype_id         raw_consensus_id</span></span>
<span><span class="co">## 1  8344    4     clonotype123 clonotype123_consensus_2</span></span>
<span><span class="co">## 2 65390   38     clonotype123 clonotype123_consensus_1</span></span>
<span><span class="co">## 4 18372    8     clonotype124 clonotype124_consensus_1</span></span>
<span><span class="co">## 5 34054    9     clonotype124 clonotype124_consensus_2</span></span>
<span><span class="co">## 6  5018    2       clonotype1   clonotype1_consensus_2</span></span>
<span><span class="co">## 7 25110   11       clonotype1   clonotype1_consensus_1</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
