<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Additional Processing Steps • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Additional Processing Steps">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Repertoire Analysis</li>
    <li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Working with Single-Cell Objects</li>
    <li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Deep Learning Integrations</li>
    <li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other Workflow Integrations</li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
    <li>
      <a href="../articles/immcantation.html">Combining Immcantation and scRepertoire</a>
    </li>
    <li>
      <a href="../articles/tcrpheno.html">TCRpheno applied to T cell fate</a>
    </li>
    <li>
      <a href="../articles/immunarch.html">Using scRepertoire with immunarch</a>
    </li>
    <li>
      <a href="../articles/dandelionR.html">VDJ Trajectory Analysis with dandelionR</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BorchLab/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Additional Processing Steps</h1>
            
            <h4 data-toc-skip class="date">Compiled: October 17,
2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BorchLab/scRepertoire/blob/HEAD/vignettes/articles/Processing.Rmd" class="external-link"><code>vignettes/articles/Processing.Rmd</code></a></small>
      <div class="hidden name"><code>Processing.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="addvariable-adding-variables-for-plotting">addVariable: Adding Variables for Plotting<a class="anchor" aria-label="anchor" href="#addvariable-adding-variables-for-plotting"></a>
</h2>
<p>What if there are more variables to add than just sample and ID? We
can add them by using the <code><a href="../reference/addVariable.html">addVariable()</a></code> function. For each
element, the function will add a column (labeled by
<code>variable.name</code>) with the <code>variable</code>. The length
of the <code>variables</code> parameter needs to match the length of the
combined object.</p>
<p>Key Parameter(s) for <code><a href="../reference/addVariable.html">addVariable()</a></code></p>
<ul>
<li>
<code>variable.name</code>: A character string that defines the new
variable to add (e.g., “Type”, “Treatment”).</li>
<li>
<code>variables</code>: A character vector defining the desired
column value for each list element. Its length must match the number of
elements in the input.data list.</li>
</ul>
<p>As an example, here we add the Type in which the samples were
processed and sequenced to the <code>combined.TCR</code> object:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">combined.TCR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addVariable.html">addVariable</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                            variable.name <span class="op">=</span> <span class="st">"Type"</span>, </span>
<span>                            variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B"</span>, <span class="st">"L"</span><span class="op">)</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">combined.TCR</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                    barcode sample                     TCR1           cdr3_aa1</span></span>
<span><span class="co">## 1  P17B_AAACCTGAGTACGACG-1   P17B       TRAV25.TRAJ20.TRAC        CGCSNDYKLSF</span></span>
<span><span class="co">## 3  P17B_AAACCTGCAACACGCC-1   P17B TRAV38-2/DV8.TRAJ52.TRAC CAYRSAQAGGTSYGKLTF</span></span>
<span><span class="co">## 5  P17B_AAACCTGCAGGCGATA-1   P17B      TRAV12-1.TRAJ9.TRAC     CVVSDNTGGFKTIF</span></span>
<span><span class="co">## 7  P17B_AAACCTGCATGAGCGA-1   P17B      TRAV12-1.TRAJ9.TRAC     CVVSDNTGGFKTIF</span></span>
<span><span class="co">## 9  P17B_AAACGGGAGAGCCCAA-1   P17B        TRAV20.TRAJ8.TRAC      CAVRGEGFQKLVF</span></span>
<span><span class="co">## 10 P17B_AAACGGGAGCGTTTAC-1   P17B      TRAV12-1.TRAJ9.TRAC     CVVSDNTGGFKTIF</span></span>
<span><span class="co">##                                                  cdr3_nt1</span></span>
<span><span class="co">## 1                       TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT</span></span>
<span><span class="co">## 3  TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT</span></span>
<span><span class="co">## 5              TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT</span></span>
<span><span class="co">## 7              TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT</span></span>
<span><span class="co">## 9                 TGTGCTGTGCGAGGAGAAGGCTTTCAGAAACTTGTATTT</span></span>
<span><span class="co">## 10             TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT</span></span>
<span><span class="co">##                           TCR2          cdr3_aa2</span></span>
<span><span class="co">## 1   TRBV5-1.None.TRBJ2-7.TRBC2    CASSLTDRTYEQYF</span></span>
<span><span class="co">## 3  TRBV10-3.None.TRBJ2-2.TRBC2     CAISEQGKGELFF</span></span>
<span><span class="co">## 5     TRBV9.None.TRBJ2-2.TRBC2 CASSVRRERANTGELFF</span></span>
<span><span class="co">## 7     TRBV9.None.TRBJ2-2.TRBC2 CASSVRRERANTGELFF</span></span>
<span><span class="co">## 9                         &lt;NA&gt;              &lt;NA&gt;</span></span>
<span><span class="co">## 10    TRBV9.None.TRBJ2-2.TRBC2 CASSVRRERANTGELFF</span></span>
<span><span class="co">##                                               cdr3_nt2</span></span>
<span><span class="co">## 1           TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC</span></span>
<span><span class="co">## 3              TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 5  TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 7  TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 9                                                 &lt;NA&gt;</span></span>
<span><span class="co">## 10 TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">##                                                  CTgene</span></span>
<span><span class="co">## 1         TRAV25.TRAJ20.TRAC_TRBV5-1.None.TRBJ2-7.TRBC2</span></span>
<span><span class="co">## 3  TRAV38-2/DV8.TRAJ52.TRAC_TRBV10-3.None.TRBJ2-2.TRBC2</span></span>
<span><span class="co">## 5          TRAV12-1.TRAJ9.TRAC_TRBV9.None.TRBJ2-2.TRBC2</span></span>
<span><span class="co">## 7          TRAV12-1.TRAJ9.TRAC_TRBV9.None.TRBJ2-2.TRBC2</span></span>
<span><span class="co">## 9                                  TRAV20.TRAJ8.TRAC_NA</span></span>
<span><span class="co">## 10         TRAV12-1.TRAJ9.TRAC_TRBV9.None.TRBJ2-2.TRBC2</span></span>
<span><span class="co">##                                                                                              CTnt</span></span>
<span><span class="co">## 1                    TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT_TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC</span></span>
<span><span class="co">## 3  TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT_TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 5  TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 7  TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 9                                                      TGTGCTGTGCGAGGAGAAGGCTTTCAGAAACTTGTATTT_NA</span></span>
<span><span class="co">## 10 TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">##                                CTaa</span></span>
<span><span class="co">## 1        CGCSNDYKLSF_CASSLTDRTYEQYF</span></span>
<span><span class="co">## 3  CAYRSAQAGGTSYGKLTF_CAISEQGKGELFF</span></span>
<span><span class="co">## 5  CVVSDNTGGFKTIF_CASSVRRERANTGELFF</span></span>
<span><span class="co">## 7  CVVSDNTGGFKTIF_CASSVRRERANTGELFF</span></span>
<span><span class="co">## 9                  CAVRGEGFQKLVF_NA</span></span>
<span><span class="co">## 10 CVVSDNTGGFKTIF_CASSVRRERANTGELFF</span></span>
<span><span class="co">##                                                                                                                                               CTstrict</span></span>
<span><span class="co">## 1                           TRAV25.TRAJ20.TRAC;TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT_TRBV5-1.None.TRBJ2-7.TRBC2;TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC</span></span>
<span><span class="co">## 3  TRAV38-2/DV8.TRAJ52.TRAC;TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT_TRBV10-3.None.TRBJ2-2.TRBC2;TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 5          TRAV12-1.TRAJ9.TRAC;TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TRBV9.None.TRBJ2-2.TRBC2;TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 7          TRAV12-1.TRAJ9.TRAC;TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TRBV9.None.TRBJ2-2.TRBC2;TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">## 9                                                                                      TRAV20.TRAJ8.TRAC;TGTGCTGTGCGAGGAGAAGGCTTTCAGAAACTTGTATTT_NA;NA</span></span>
<span><span class="co">## 10         TRAV12-1.TRAJ9.TRAC;TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TRBV9.None.TRBJ2-2.TRBC2;TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</span></span>
<span><span class="co">##    Type</span></span>
<span><span class="co">## 1     B</span></span>
<span><span class="co">## 3     B</span></span>
<span><span class="co">## 5     B</span></span>
<span><span class="co">## 7     B</span></span>
<span><span class="co">## 9     B</span></span>
<span><span class="co">## 10    B</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="subsetclones-filter-out-clonal-information">subsetClones: Filter Out Clonal Information<a class="anchor" aria-label="anchor" href="#subsetclones-filter-out-clonal-information"></a>
</h2>
<p>Likewise, we can remove specific list elements after
<code><a href="../reference/combineTCR.html">combineTCR()</a></code> or <code><a href="../reference/combineBCR.html">combineBCR()</a></code> using the
<code><a href="../reference/subsetClones.html">subsetClones()</a></code> function. In order to subset, we need to
identify the column header we would like to use for subsetting
(<code>name</code>) and the specific values to include
(<code>variables</code>).</p>
<p>Key Parameter(s) for <code><a href="../reference/subsetClones.html">subsetClones()</a></code></p>
<ul>
<li>
<code>name</code>: The column header/name in the metadata of
input.data to use for subsetting (e.g., “sample”, “Type”).</li>
<li>
<code>variables</code>: A character vector of the specific values
within the chosen name column to retain in the subsetted data.</li>
</ul>
<p>Below, we isolate just the two sequencing results from “P18L” and
“P18B” samples:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">subset1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subsetClones.html">subsetClones</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                        name <span class="op">=</span> <span class="st">"sample"</span>, </span>
<span>                        variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P18L"</span>, <span class="st">"P18B"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">subset1</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                    barcode sample                 TCR1         cdr3_aa1</span></span>
<span><span class="co">## 1  P18B_AAACCTGAGGCTCAGA-1   P18B TRAV26-1.TRAJ37.TRAC  CIVRGGSSNTGKLIF</span></span>
<span><span class="co">## 3  P18B_AAACCTGCATGACATC-1   P18B    TRAV3.TRAJ20.TRAC    CAVQRSNDYKLSF</span></span>
<span><span class="co">## 5  P18B_AAACCTGGTATGCTTG-1   P18B TRAV26-1.TRAJ53.TRAC   CIGSSGGSNYKLTF</span></span>
<span><span class="co">## 8  P18B_AAACGGGCAGATGGGT-1   P18B                 &lt;NA&gt;             &lt;NA&gt;</span></span>
<span><span class="co">## 9  P18B_AAACGGGTCTTACCGC-1   P18B    TRAV20.TRAJ9.TRAC CAVQAKRYTGGFKTIF</span></span>
<span><span class="co">## 12 P18B_AAAGATGAGTTACGGG-1   P18B   TRAV8-3.TRAJ8.TRAC   CAVGGDTGFQKLVF</span></span></code></pre>
<p>Alternatively, we can also just select the list elements after
<code><a href="../reference/combineTCR.html">combineTCR()</a></code> or <code><a href="../reference/combineBCR.html">combineBCR()</a></code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">subset2</span> <span class="op">&lt;-</span> <span class="va">combined.TCR</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">4</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">subset2</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                    barcode sample                 TCR1         cdr3_aa1</span></span>
<span><span class="co">## 1  P18B_AAACCTGAGGCTCAGA-1   P18B TRAV26-1.TRAJ37.TRAC  CIVRGGSSNTGKLIF</span></span>
<span><span class="co">## 3  P18B_AAACCTGCATGACATC-1   P18B    TRAV3.TRAJ20.TRAC    CAVQRSNDYKLSF</span></span>
<span><span class="co">## 5  P18B_AAACCTGGTATGCTTG-1   P18B TRAV26-1.TRAJ53.TRAC   CIGSSGGSNYKLTF</span></span>
<span><span class="co">## 8  P18B_AAACGGGCAGATGGGT-1   P18B                 &lt;NA&gt;             &lt;NA&gt;</span></span>
<span><span class="co">## 9  P18B_AAACGGGTCTTACCGC-1   P18B    TRAV20.TRAJ9.TRAC CAVQAKRYTGGFKTIF</span></span>
<span><span class="co">## 12 P18B_AAAGATGAGTTACGGG-1   P18B   TRAV8-3.TRAJ8.TRAC   CAVGGDTGFQKLVF</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="exportclones-save-clonal-data">exportClones: Save Clonal Data<a class="anchor" aria-label="anchor" href="#exportclones-save-clonal-data"></a>
</h2>
<p>After assigning the clone by barcode, we can export the clonal
information using <code><a href="../reference/exportClones.html">exportClones()</a></code> to save for later use or
to integrate with other bioinformatics pipelines. This function supports
various output formats tailored for different analytical needs.</p>
<p>Key Parameter(s) for <code><a href="../reference/exportClones.html">exportClones()</a></code> *
<code>format</code>: The desired output format for the clonal data. *
<code>airr</code>: Exports data in an Adaptive Immune Receptor
Repertoire (AIRR) Community-compliant format, with each row representing
a single receptor chain. * <code>immunarch</code>: Exports a list
containing a data frame and metadata formatted for use with the
<code>immunarch</code> package. * <code>paired</code>: Exports a data
frame with paired chain information (amino acid, nucleotide, genes) per
barcode. This is the default. * <code>TCRMatch</code>: Exports a data
frame specifically for the TCRMatch algorithm, containing TRB chain
amino acid sequence and clonal frequency. * <code>tcrpheno</code>:
Exports a data frame compatible with the <code>tcrpheno</code> pipeline,
with TRA and TRB chains in separate columns. * <code>write.file</code>:
If <code>TRUE</code> (default), saves the output to a CSV file. If
<code>FALSE</code>, returns the data frame or list to the R environment.
* <code>dir</code>: The directory where the output file will be saved.
Defaults to the current working directory. * <code>file.name</code>: The
name of the CSV file to be saved.</p>
<p>To export the combined clonotypes as a <code>paired</code> data frame
and save it to a specified directory:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">exportClones</span>(combined, </span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>             <span class="at">write.file =</span> <span class="cn">TRUE</span>,</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>             <span class="at">dir =</span> <span class="st">"~/Documents/MyExperiment/Sample1/"</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>             <span class="at">file.name =</span> <span class="st">"clones.csv"</span>)</span></code></pre></div>
<p>To return an <code>immunarch</code>-formatted data frame directly to
your R environment without saving a file:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">immunarch</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exportClones.html">exportClones</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                          format <span class="op">=</span> <span class="st">"immunarch"</span>, </span>
<span>                          write.file <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">immunarch</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   Clones   Proportion                                             CDR3.nt</span></span>
<span><span class="co">## 1      1 0.0003565062    NA;TGCGCCAGCAGTCGGGGACTAGCGGGATACAATGAGCAGTTCTTC</span></span>
<span><span class="co">## 2      1 0.0003565062       NA;TGTGCCATCAGCGCGGACCCCCGCTACAATGAGCAGTTCTTC</span></span>
<span><span class="co">## 3      1 0.0003565062 NA;TGTGCCAGCAGCTTGAGGGACAGCTATCGGTACTATGGCTACACCTTC</span></span>
<span><span class="co">## 4      2 0.0007130125          NA;TGTGCCAGCAGCCGGCAGGGCGCAGATACGCAGTATTTT</span></span>
<span><span class="co">## 5      1 0.0003565062       NA;TGTGCCAGCAGTCCCTTTACAGGGTTCTATGGCTACACCTTC</span></span>
<span><span class="co">## 6      1 0.0003565062          NA;TGTGCCAGCTCATCCGGGATCAATCAGCCCCAGCATTTT</span></span>
<span><span class="co">##               CDR3.aa      V.name  D.name     J.name   C.name</span></span>
<span><span class="co">## 1  NA;CASSRGLAGYNEQFF NA;TRBV10-2 NA;None NA;TRBJ2-1 NA;TRBC2</span></span>
<span><span class="co">## 2   NA;CAISADPRYNEQFF NA;TRBV10-3 NA;None NA;TRBJ2-1 NA;TRBC2</span></span>
<span><span class="co">## 3 NA;CASSLRDSYRYYGYTF NA;TRBV11-3 NA;None NA;TRBJ1-2 NA;TRBC1</span></span>
<span><span class="co">## 4    NA;CASSRQGADTQYF NA;TRBV11-3 NA;None NA;TRBJ2-3 NA;TRBC2</span></span>
<span><span class="co">## 5   NA;CASSPFTGFYGYTF NA;TRBV12-4 NA;None NA;TRBJ1-2 NA;TRBC1</span></span>
<span><span class="co">## 6    NA;CASSSGINQPQHF   NA;TRBV18 NA;None NA;TRBJ1-5 NA;TRBC1</span></span>
<span><span class="co">##                                           Barcode</span></span>
<span><span class="co">## 1                         P17B_AGCGGTCCAAAGGAAG-1</span></span>
<span><span class="co">## 2                         P17B_GGCTCGAGTCGCGGTT-1</span></span>
<span><span class="co">## 3                         P17B_CGCGTTTTCGGCTACG-1</span></span>
<span><span class="co">## 4 P17B_CACCAGGGTTCCTCCA-1;P17B_TCTATTGCAGGTGCCT-1</span></span>
<span><span class="co">## 5                         P17B_GCTGGGTGTACGAAAT-1</span></span>
<span><span class="co">## 6                         P17B_AGGGTGACATTGGTAC-1</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="annotateinvariant">annotateInvariant<a class="anchor" aria-label="anchor" href="#annotateinvariant"></a>
</h2>
<p>The <code><a href="../reference/annotateInvariant.html">annotateInvariant()</a></code> function enables the
identification of mucosal-associated invariant T (<code>MAIT</code>)
cells and invariant natural killer T (<code>iNKT</code>) cells in
single-cell sequencing datasets. These specialized T-cell subsets are
defined by their characteristic TCR usage, making them distinguishable
within single-cell immune profiling data. The function extracts TCR
chain information from the provided single-cell dataset and evaluates it
against known invariant TCR criteria for either MAIT or iNKT cells. Each
cell is assigned a score indicating the presence (1) or absence (0) of
the specified invariant T-cell population.</p>
<p>Key Parameter(s) for <code>annotateInavriant()</code></p>
<ul>
<li>
<code>type</code>: Character string specifying the type of invariant
T cell to annotate (<code>MAIT</code> or <code>iNKT</code>).</li>
<li>
<code>species</code>: Character string specifying the species
(<code>mouse</code> or <code>human</code>).</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotateInvariant.html">annotateInvariant</a></span><span class="op">(</span><span class="va">combined</span>, </span>
<span>                              type <span class="op">=</span> <span class="st">"MAIT"</span>, </span>
<span>                              species <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotateInvariant.html">annotateInvariant</a></span><span class="op">(</span><span class="va">combined</span>, </span>
<span>                              type <span class="op">=</span> <span class="st">"iNKT"</span>, </span>
<span>                              species <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
