<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Summarizing Repertoires • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Summarizing Repertoires">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Repertoire Analysis</li>
    <li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Working with Single-Cell Objects</li>
    <li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Deep Learning Integrations</li>
    <li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other Workflow Integrations</li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
    <li>
      <a href="../articles/immcantation.html">Combining Immcantation and scRepertoire</a>
    </li>
    <li>
      <a href="../articles/tcrpheno.html">TCRpheno applied to T cell fate</a>
    </li>
    <li>
      <a href="../articles/immunarch.html">Using scRepertoire with immunarch</a>
    </li>
    <li>
      <a href="../articles/dandelionR.html">VDJ Trajectory Analysis with dandelionR</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BorchLab/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Summarizing Repertoires</h1>
            
            <h4 data-toc-skip class="date">Compiled: October 17,
2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BorchLab/scRepertoire/blob/HEAD/vignettes/articles/Repertoire_Summary.Rmd" class="external-link"><code>vignettes/articles/Repertoire_Summary.Rmd</code></a></small>
      <div class="hidden name"><code>Repertoire_Summary.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="percentaa">percentAA<a class="anchor" aria-label="anchor" href="#percentaa"></a>
</h2>
<p>Quantify the proportion of amino acids along the CDR3 sequence with
<code><a href="../reference/percentAA.html">percentAA()</a></code>. By default, the function will pad the
sequences with NAs up to the maximum of <code>aa.length</code>.
Sequences longer than aa.length will be removed before visualization
(default <code>aa.length</code> = 20).</p>
<p>Key Parameter(s) for <code><a href="../reference/percentAA.html">percentAA()</a></code></p>
<ul>
<li>
<code>aa.length</code>: The maximum length of the CDR3 amino acid
sequence to consider.</li>
</ul>
<p>To visualize the relative percentage of amino acids at each position
of the CDR3 sequences for the TRB chain, up to a length of 20 amino
acids:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/percentAA.html">percentAA</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>          chain <span class="op">=</span> <span class="st">"TRB"</span>, </span>
<span>          aa.length <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-1-1.png" width="700"></p>
<p>This plot displays the relative proportion of each amino acid at
different positions across the CDR3 sequence for the specified chain. It
provides insights into the amino acid composition and variability at
each residue, which can be indicative of functional constraints or
selection pressures on the CDR3</p>
</div>
<div class="section level2">
<h2 id="positionalentropy">positionalEntropy<a class="anchor" aria-label="anchor" href="#positionalentropy"></a>
</h2>
<p>We can also quantify the level of entropy/diversity across amino acid
residues along the CDR3 sequence. <code><a href="../reference/positionalEntropy.html">positionalEntropy()</a></code>
combines the quantification by residue of <code><a href="../reference/percentAA.html">percentAA()</a></code> with
diversity calculations. Positions without variance will have a value
reported as 0 for the purposes of comparison.</p>
<p>Key Parameter(s) for <code><a href="../reference/positionalEntropy.html">positionalEntropy()</a></code></p>
<ul>
<li>
<code>method</code>
<ul>
<li>
<code>shannon</code> - Shannon Index</li>
<li>
<code>inv.simpson</code> - Inverse Simpson Index</li>
<li>
<code>gini.simpson</code> - Gini-Simpson Index</li>
<li>
<code>norm.entropy</code> - Normalized Entropy</li>
<li>
<code>pielou</code> - Pielou’s Evenness</li>
<li>
<code>hill1</code>, <code>hill2</code>, <code>hill3</code> - Hill
Numbers</li>
</ul>
</li>
</ul>
<p>To visualize the normalized entropy across amino acid residues for
the TRB chain, up to a length of 20 amino acids:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/positionalEntropy.html">positionalEntropy</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                  chain <span class="op">=</span> <span class="st">"TRB"</span>, </span>
<span>                  aa.length <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<p>The plot generated by <code><a href="../reference/positionalEntropy.html">positionalEntropy()</a></code> illustrates
the diversity or entropy at each amino acid position within the CDR3
sequence. Higher entropy values indicate greater variability in amino
acid usage at that position, suggesting less selective pressure or more
promiscuous binding, while lower values suggest conserved positions
critical for structural integrity or antigen recognition.</p>
</div>
<div class="section level2">
<h2 id="positionalproperty">positionalProperty<a class="anchor" aria-label="anchor" href="#positionalproperty"></a>
</h2>
<p>Like <code><a href="../reference/positionalEntropy.html">positionalEntropy()</a></code>, we can also examine a series
of amino acid properties along the CDR3 sequences using
<code><a href="../reference/positionalProperty.html">positionalProperty()</a></code>. Important differences for
<code><a href="../reference/positionalProperty.html">positionalProperty()</a></code> is dropping NA values as they would
void the mean calculation and displaying a ribbon with the 95%
confidence interval surrounding the mean value for the selected
properties.</p>
<p>Key Parameter(s) for <code><a href="../reference/positionalEntropy.html">positionalEntropy()</a></code></p>
<ul>
<li>
<code>method</code>
<ul>
<li>
<code>atchleyFactors</code>: <a href="https://pubmed.ncbi.nlm.nih.gov/15851683/" class="external-link">citation</a>
</li>
<li>
<code>crucianiProperties</code>: <a href="https://analyticalsciencejournals.onlinelibrary.wiley.com/doi/abs/10.1002/cem.856" class="external-link">citation</a>
</li>
<li>
<code>FASGAI</code>: <a href="https://pubmed.ncbi.nlm.nih.gov/18318694/" class="external-link">citation</a>
</li>
<li>
<code>kideraFactors</code>: <a href="https://link.springer.com/article/10.1007/BF01025492" class="external-link">citation</a>
</li>
<li>
<code>MSWHIM</code>: <a href="https://pubs.acs.org/doi/10.1021/ci980211b" class="external-link">citation</a>
</li>
<li>
<code>ProtFP</code>: <a href="https://pubmed.ncbi.nlm.nih.gov/24059694/" class="external-link">citation</a>
</li>
<li>
<code>stScales</code>: <a href="https://pubmed.ncbi.nlm.nih.gov/19373543/" class="external-link">citation</a>
</li>
<li>
<code>tScales</code>: <a href="https://www.sciencedirect.com/science/article/abs/pii/S0022286006006314" class="external-link">citation</a>
</li>
<li>
<code>VHSE</code>: <a href="https://pubmed.ncbi.nlm.nih.gov/15895431/" class="external-link">citation</a>
</li>
<li>
<code>zScales</code>: <a href="https://pubmed.ncbi.nlm.nih.gov/9651153/" class="external-link">citation</a>
</li>
</ul>
</li>
</ul>
<p>To examine the Atchley factors of amino acids across the CDR3
sequence for the first two samples:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/positionalProperty.html">positionalProperty</a></span><span class="op">(</span><span class="va">combined.TCR</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                  chain <span class="op">=</span> <span class="st">"TRB"</span>, </span>
<span>                  aa.length <span class="op">=</span> <span class="fl">20</span>, </span>
<span>                  method <span class="op">=</span> <span class="st">"atchleyFactors"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.colors</a></span><span class="op">(</span><span class="fl">5</span>, <span class="st">"inferno"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>This function provides a detailed view of how physicochemical
properties of amino acids change along the CDR3 sequence. By visualizing
the mean property value and its confidence interval, one can identify
positions with distinct characteristics or significant variations
between groups, offering insights into structural and functional aspects
of the CDR3.</p>
</div>
<div class="section level2">
<h2 id="percentgeneusage">percentGeneUsage<a class="anchor" aria-label="anchor" href="#percentgeneusage"></a>
</h2>
<p>Gene quantification and visualization has been redesigned to offer a
more robust and translatable function under
<code><a href="../reference/percentGeneUsage.html">percentGeneUsage()</a></code>. We have maintained the functionality of
the previous functions for gene-level quantification under the aliases
<code><a href="../reference/percentGeneUsage.html">vizGenes()</a></code>, <code><a href="../reference/percentGeneUsage.html">percentGenes()</a></code>, and
<code><a href="../reference/percentGeneUsage.html">percentVJ()</a></code>.</p>
<div class="section level3">
<h3 id="vizgenes-flexible-gene-usage-visualization">vizGenes: Flexible Gene Usage Visualization<a class="anchor" aria-label="anchor" href="#vizgenes-flexible-gene-usage-visualization"></a>
</h3>
<p>The <code><a href="../reference/percentGeneUsage.html">vizGenes()</a></code> function offers a highly adaptable
approach to visualizing the relative usage of TCR or BCR genes. It acts
as a versatile alias for <code><a href="../reference/percentGeneUsage.html">percentGeneUsage()</a></code>, allowing for
comparisons across different chains, scaling of values, and selection
between bar charts and heatmaps.</p>
<ul>
<li>
<code>x.axis</code>: Specifies the gene segment to display along the
x-axis (e.g., “TRBV”, “TRBD”, “IGKJ”).</li>
<li>
<code>y.axis</code>
<ul>
<li>Another gene segment (e.g., “TRAV”, “TRBJ”) for paired gene
analysis. When x.axis and y.axis are both gene segments, vizGenes()
internally calls percentGeneUsage() with genes = c(x.axis, y.axis),
resulting in a heatmap.</li>
<li>A categorical variable (e.g., “sample”, “orig.ident”) to visualize
gene usage across different groups. When y.axis is a categorical
variable, vizGenes() maps it to the group.by parameter of
percentGeneUsage(), creating facets for each category.</li>
</ul>
</li>
<li>
<code>plot</code>: Determines the visualization type:
<ul>
<li>“barplot”: Ideal for visualizing the distribution of a single gene
segment.</li>
<li>“heatmap”: Suitable for single gene usage (with a group.by or y.axis
categorical variable) or for paired gene analysis.</li>
</ul>
</li>
<li>
<code>summary.fun</code>: (Inherited from
<code>percentGeneUsage</code>) Defines the statistic to display:
“percent” (default), “proportion”, or “count”. This implicitly handles
the scaling of values.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/percentGeneUsage.html">vizGenes</a></span><span class="op">(</span><span class="va">combined.TCR</span>,</span>
<span>         x.axis <span class="op">=</span> <span class="st">"TRBV"</span>,</span>
<span>         y.axis <span class="op">=</span> <span class="cn">NULL</span>, <span class="co"># No specific y-axis variable, will group all samples</span></span>
<span>         plot <span class="op">=</span> <span class="st">"barplot"</span>,</span>
<span>         summary.fun <span class="op">=</span> <span class="st">"proportion"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>This plot shows the proportion of each <em>TRBV</em> gene segment
observed across the entire combined.TCR dataset. Since
<code>y.axis</code> is NULL, samples are grouped by the list element of
the combined.TCR data.</p>
<p><code><a href="../reference/percentGeneUsage.html">vizGenes()</a></code> is particularly useful for examining gene
pairings. Let’s look at the differences in <em>TRBV</em> and
<em>TRBJ</em> usage between peripheral blood and lung samples from your
dataset. We’ll subset combined.TCR for this analysis.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Peripheral Blood Samples</span></span>
<span><span class="fu"><a href="../reference/percentGeneUsage.html">vizGenes</a></span><span class="op">(</span><span class="va">combined.TCR</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17B"</span>, <span class="st">"P18B"</span>, <span class="st">"P19B"</span>, <span class="st">"P20B"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>         x.axis <span class="op">=</span> <span class="st">"TRBV"</span>,</span>
<span>         y.axis <span class="op">=</span> <span class="st">"TRBJ"</span>,</span>
<span>         plot <span class="op">=</span> <span class="st">"heatmap"</span>,</span>
<span>         summary.fun <span class="op">=</span> <span class="st">"percent"</span><span class="op">)</span> <span class="co"># Display percentages</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Lung Samples</span></span>
<span><span class="fu"><a href="../reference/percentGeneUsage.html">vizGenes</a></span><span class="op">(</span><span class="va">combined.TCR</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17L"</span>, <span class="st">"P18L"</span>, <span class="st">"P19L"</span>, <span class="st">"P20L"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>         x.axis <span class="op">=</span> <span class="st">"TRBV"</span>,</span>
<span>         y.axis <span class="op">=</span> <span class="st">"TRBJ"</span>,</span>
<span>         plot <span class="op">=</span> <span class="st">"heatmap"</span>,</span>
<span>         summary.fun <span class="op">=</span> <span class="st">"percent"</span><span class="op">)</span> <span class="co"># Display percentages</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-5-2.png" width="700"></p>
<p>In these examples, by providing both <code>x.axis</code> and
<code>y.axis</code> as gene segments (“TRBV” and “TRBJ”),
<code><a href="../reference/percentGeneUsage.html">vizGenes()</a></code> automatically performs a paired gene analysis,
generating a heatmap where the intensity reflects the percentage of each
V-J pairing.</p>
<p>Beyond V-J pairings within a single chain, <code><a href="../reference/percentGeneUsage.html">vizGenes()</a></code>
can also visualize gene usage across different chains. For instance, to
examine <em>TRBV</em> and <em>TRAV</em> pairings for patient P17’s
samples:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/percentGeneUsage.html">vizGenes</a></span><span class="op">(</span><span class="va">combined.TCR</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17B"</span>, <span class="st">"P17L"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>         x.axis <span class="op">=</span> <span class="st">"TRBV"</span>,</span>
<span>         y.axis <span class="op">=</span> <span class="st">"TRAV"</span>,</span>
<span>         plot <span class="op">=</span> <span class="st">"heatmap"</span>,</span>
<span>         summary.fun <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="percentgenes-quantifying-single-gene-usage">percentGenes: Quantifying Single Gene Usage<a class="anchor" aria-label="anchor" href="#percentgenes-quantifying-single-gene-usage"></a>
</h3>
<p>The <code><a href="../reference/percentGeneUsage.html">percentGenes()</a></code> function is a specialized alias for
<code><a href="../reference/percentGeneUsage.html">percentGeneUsage()</a></code> designed to quantify the usage of a
single V, D, or J gene locus for a specified immune receptor chain. By
default, it returns a heatmap visualization.</p>
<p>Key Parameters for <code><a href="../reference/percentGeneUsage.html">percentGenes()</a></code>:</p>
<ul>
<li>
<code>chain</code>: Specifies the immune receptor chain (e.g.,
“TRB”, “TRA”, “IGH”, “IGL”).</li>
<li>
<code>gene</code>: Indicates the gene locus to quantify: “Vgene”,
“Dgene”, or “Jgene”.</li>
<li>
<code>group.by</code>, <code>order.by</code>,
<code>summary.fun</code>, <code>exportTable</code>,
<code>palette</code>: These parameters are directly passed to
<code><a href="../reference/percentGeneUsage.html">percentGeneUsage()</a></code>.</li>
</ul>
<p>To quantify and visualize the percentage of TRBV gene usage across
your samples:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/percentGeneUsage.html">percentGenes</a></span><span class="op">(</span><span class="va">combined.TCR</span>,</span>
<span>             chain <span class="op">=</span> <span class="st">"TRB"</span>,</span>
<span>             gene <span class="op">=</span> <span class="st">"Vgene"</span>,</span>
<span>             summary.fun <span class="op">=</span> <span class="st">"percent"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>This generates a heatmap showing the percentage of each TRBV gene
segment within each sample, allowing for easy visual comparison of gene
usage profiles across your samples.</p>
<p>The raw data returned by <code><a href="../reference/percentGeneUsage.html">percentGenes()</a></code> (when
<code>exportTable</code> = TRUE) can be a powerful input for further
downstream analysis, such as dimensionality reduction techniques like
Principal Component Analysis. This allows you to summarize the complex
gene usage patterns and identify samples with similar or distinct
repertoires.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/percentGeneUsage.html">percentGenes</a></span><span class="op">(</span><span class="va">combined.TCR</span>,</span>
<span>                         chain <span class="op">=</span> <span class="st">"TRB"</span>,</span>
<span>                         gene <span class="op">=</span> <span class="st">"Vgene"</span>,</span>
<span>                         exportTable <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                         summary.fun <span class="op">=</span> <span class="st">"proportion"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Performing PCA on the gene usage matrix</span></span>
<span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">df.genes</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Getting data frame to plot from</span></span>
<span><span class="va">df_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">df.genes</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">df_plot</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"Sample"</span><span class="op">)</span></span>
<span><span class="va">df_plot</span><span class="op">$</span><span class="va">PC1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">df_plot</span><span class="op">$</span><span class="va">PC1</span><span class="op">)</span></span>
<span><span class="va">df_plot</span><span class="op">$</span><span class="va">PC2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">df_plot</span><span class="op">$</span><span class="va">PC2</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df_plot</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">Sample</span><span class="op">)</span>, shape <span class="op">=</span> <span class="fl">21</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>fill<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>title<span class="op">=</span><span class="st">"Samples"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.colors</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">df_plot</span><span class="op">)</span>, <span class="st">"inferno"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"PCA of TRBV Gene Usage"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>This PCA plot visually clusters samples based on their <em>TRBV</em>
gene usage profiles, helping to identify underlying patterns or
relationships between samples.</p>
</div>
<div class="section level3">
<h3 id="percentvj-quantifying-v-j-gene-pairings">percentVJ: Quantifying V-J Gene Pairings<a class="anchor" aria-label="anchor" href="#percentvj-quantifying-v-j-gene-pairings"></a>
</h3>
<p>The <code><a href="../reference/percentGeneUsage.html">percentVJ()</a></code> function is another specialized alias
for <code><a href="../reference/percentGeneUsage.html">percentGeneUsage()</a></code>, specifically designed to quantify
the proportion or percentage of V and J gene segments paired together
within individual repertoires. It always produces a heatmap
visualization.</p>
<p>Key Parameters for <code>percenVJ()</code>:</p>
<ul>
<li>
<code>chain</code>: Specifies the immune receptor chain (e.g.,
<strong>“TRB”</strong>, <strong>“TRA”</strong>, <strong>“IGH”</strong>,
<strong>“IGL”</strong>). This dictates which V and J gene segments are
analyzed (e.g., <strong>TRBV</strong> and <strong>TRBJ</strong> for
<code>chain</code> = “TRB”)</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/percentGeneUsage.html">percentVJ</a></span><span class="op">(</span><span class="va">combined.TCR</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>          chain <span class="op">=</span> <span class="st">"TRB"</span>,</span>
<span>          summary.fun <span class="op">=</span> <span class="st">"percent"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>This heatmap displays the percentage of each <em>TRBV-TRBJ</em>
pairing for each sample, providing a detailed view of the V-J
recombination landscape.</p>
<p>Similar to <code><a href="../reference/percentGeneUsage.html">percentGenes()</a></code>, the quantitative output of
<code><a href="../reference/percentGeneUsage.html">percentVJ()</a></code> can be used for dimensionality reduction to
summarize V-J pairing patterns across samples.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df.vj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/percentGeneUsage.html">percentVJ</a></span><span class="op">(</span><span class="va">combined.TCR</span>,</span>
<span>                   chain <span class="op">=</span> <span class="st">"TRB"</span>,</span>
<span>                   exportTable <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                   summary.fun <span class="op">=</span> <span class="st">"proportion"</span><span class="op">)</span> <span class="co"># Export proportions for PCA</span></span>
<span></span>
<span><span class="co"># Performing PCA on the V-J pairing matrix</span></span>
<span><span class="va">pc.vj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">df.vj</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Getting data frame to plot from</span></span>
<span><span class="va">df_plot_vj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">pc.vj</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">df.vj</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">df_plot_vj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"Sample"</span><span class="op">)</span></span>
<span><span class="va">df_plot_vj</span><span class="op">$</span><span class="va">PC1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">df_plot_vj</span><span class="op">$</span><span class="va">PC1</span><span class="op">)</span></span>
<span><span class="va">df_plot_vj</span><span class="op">$</span><span class="va">PC2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">df_plot_vj</span><span class="op">$</span><span class="va">PC2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plotting the PCA results</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df_plot_vj</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">Sample</span><span class="op">)</span>, shape <span class="op">=</span> <span class="fl">21</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>fill<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>title<span class="op">=</span><span class="st">"Samples"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.colors</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">df_plot_vj</span><span class="op">)</span>, <span class="st">"inferno"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"PCA of TRBV-TRBJ Gene Pairings"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="percentkmer">percentKmer<a class="anchor" aria-label="anchor" href="#percentkmer"></a>
</h2>
<p>Another quantification of the composition of the CDR3 sequence is to
define motifs by sliding across the amino acid or nucleotide sequences
at set intervals, resulting in substrings or kmers.</p>
<p>Key Parameter(s) for <code><a href="../reference/percentKmer.html">percentKmer()</a></code>:</p>
<ul>
<li>
<code>cloneCall</code>: Defines the clonal sequence grouping; only
accepts <code>aa</code> (amino acids) or <code>nt</code>
(nucleotides)</li>
<li>
<code>motif.length</code>: The length of the kmer to analyze.</li>
<li>
<code>top.motifs</code>: Displays the n most variable motifs as a
function of median absolute deviation.</li>
</ul>
<p>To visualize the percentage of the top 25 most variable 3-mer amino
acid motifs for the TRB chain:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/percentKmer.html">percentKmer</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>            cloneCall <span class="op">=</span> <span class="st">"aa"</span>,</span>
<span>            chain <span class="op">=</span> <span class="st">"TRB"</span>, </span>
<span>            motif.length <span class="op">=</span> <span class="fl">3</span>, </span>
<span>            top.motifs <span class="op">=</span> <span class="fl">25</span><span class="op">)</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>To perform the same analysis but for nucleotide motifs:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/percentKmer.html">percentKmer</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>            cloneCall <span class="op">=</span> <span class="st">"nt"</span>,</span>
<span>            chain <span class="op">=</span> <span class="st">"TRB"</span>, </span>
<span>            motif.length <span class="op">=</span> <span class="fl">3</span>, </span>
<span>            top.motifs <span class="op">=</span> <span class="fl">25</span><span class="op">)</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>The heatmaps generated by <code><a href="../reference/percentKmer.html">percentKmer()</a></code> illustrate the
relative composition of frequently occurring k-mer motifs across samples
or groups. This analysis can highlight recurrent sequence patterns in
the CDR3, which may be associated with specific antigen recognition,
disease states, or processing mechanisms. By examining the most variable
motifs, researchers can identify key sequence features that
differentiate immune repertoires.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
