<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Summarizing Repertoires • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Summarizing Repertoires">
<meta property="og:description" content="scRepertoire">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ncborcherding/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Summarizing Repertoires</h1>
            
            <h4 data-toc-skip class="date">Compiled: May 20, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ncborcherding/scRepertoire/blob/HEAD/vignettes/articles/Repertoire_Summary.Rmd" class="external-link"><code>vignettes/articles/Repertoire_Summary.Rmd</code></a></small>
      <div class="hidden name"><code>Repertoire_Summary.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="percentaa">percentAA<a class="anchor" aria-label="anchor" href="#percentaa"></a>
</h2>
<p>Quantify the proportion of amino acids along the cdr3 sequence with
<code><a href="../reference/percentAA.html">percentAA()</a></code>. By default, the function will pad the
sequences with NAs up to the maximum of <strong>aa.length</strong>.
Sequences longer than <strong>aa.length</strong> will be removed before
visualization <strong>(default aa.length = 20)</strong>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/percentAA.html">percentAA</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>          chain <span class="op">=</span> <span class="st">"TRB"</span>, </span>
<span>          aa.length <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-1-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="positionalentropy">positionalEntropy<a class="anchor" aria-label="anchor" href="#positionalentropy"></a>
</h2>
<p>We can also quantify the level of entropy/diversity across amino acid
residues along the cdr3 sequence. <code><a href="../reference/positionalEntropy.html">positionalEntropy()</a></code>
combines the quantification by residue of <code><a href="../reference/percentAA.html">percentAA()</a></code> with
the diversity calls in <code><a href="../reference/clonalDiversity.html">clonalDiversity()</a></code>.</p>
<p><strong>method</strong></p>
<ul>
<li>“shannon” - Shannon Diversity<br>
</li>
<li>“inv.simpson” - Inverse Simpson Diversity</li>
<li>“norm.entropy” - Normalized Entropy</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/positionalEntropy.html">positionalEntropy</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                  chain <span class="op">=</span> <span class="st">"TRB"</span>, </span>
<span>                  aa.length <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="positionalproperty">positionalProperty<a class="anchor" aria-label="anchor" href="#positionalproperty"></a>
</h2>
<p>Like <code><a href="../reference/positionalEntropy.html">positionalEntropy()</a></code>, we can also examine a series
of amino acid properties along the cdr3 sequences using
<code><a href="../reference/positionalProperty.html">positionalProperty()</a></code>. Important differences from the above
function for <code><a href="../reference/positionalProperty.html">positionalProperty()</a></code> is dropping NA values as
they would void the mean calculation. <code><a href="../reference/positionalProperty.html">positionalProperty()</a></code>
also display a ribbon with the 95% confidence interval surrounding the
mean value for the selected properties.</p>
<p><strong>method</strong></p>
<ul>
<li>“Atchley” - Atchley Factors</li>
<li>“Kidera” - Kidera Factors</li>
<li>“stScales” - stScales Vectors</li>
<li>“tScales” - tScales Vectors</li>
<li>“VHSE” - Vectors of Hydrophobic, Steric, and Electronic
properties</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/positionalProperty.html">positionalProperty</a></span><span class="op">(</span><span class="va">combined.TCR</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                  chain <span class="op">=</span> <span class="st">"TRB"</span>, </span>
<span>                  aa.length <span class="op">=</span> <span class="fl">20</span>, </span>
<span>                  method <span class="op">=</span> <span class="st">"Atchley"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.colors</a></span><span class="op">(</span><span class="fl">5</span>, <span class="st">"inferno"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="vizgenes">vizGenes<a class="anchor" aria-label="anchor" href="#vizgenes"></a>
</h2>
<p>A visualization of the relative usage of genes of the TCR or BCR,
using <code><a href="../reference/vizGenes.html">vizGenes()</a></code>. There is some functional crossover
between <code><a href="../reference/vizGenes.html">vizGenes()</a></code> and two functions below called
<code><a href="../reference/percentGenes.html">percentGenes()</a></code> and <code><a href="../reference/percentVJ.html">percentVJ()</a></code>. But
<code><a href="../reference/vizGenes.html">vizGenes()</a></code> is more adaptable to allow for comparisons
across chains, scaling, etc.</p>
<p><strong>x.axis</strong></p>
<ul>
<li>Gene segment to separate the x-axis, such as “TRAV”, “TRBD”,
“IGKJ”</li>
</ul>
<p><strong>y.axis</strong></p>
<ul>
<li>Gene segment or categorical variable to visualize along the
y-axis</li>
</ul>
<p><strong>plot</strong></p>
<ul>
<li>“barplot” for a bar chart<br>
</li>
<li>“heatmap” for a heatmap</li>
</ul>
<p><strong>scale</strong></p>
<ul>
<li>TRUE to scale the graph by number of genes per sample<br>
</li>
<li>FALSE to report raw numbers</li>
</ul>
<p><strong>order</strong></p>
<ul>
<li>“gene” to order by gene name<br>
</li>
<li>“variance” to order by variance between the separate variable
categories</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/vizGenes.html">vizGenes</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>         x.axis <span class="op">=</span> <span class="st">"TRBV"</span>,</span>
<span>         y.axis <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>         plot <span class="op">=</span> <span class="st">"barplot"</span>,  </span>
<span>         scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p><code><a href="../reference/vizGenes.html">vizGenes()</a></code> can also be used to look at the usage of
genes in a single chain. For example, say we are interested in the
difference in TRB V and J usage between lung and peripheral blood
samples - we can easily take a look at this using the following
code:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Peripheral Blood</span></span>
<span><span class="fu"><a href="../reference/vizGenes.html">vizGenes</a></span><span class="op">(</span><span class="va">combined.TCR</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span>,<span class="fl">5</span>,<span class="fl">7</span><span class="op">)</span><span class="op">]</span>, </span>
<span>         x.axis <span class="op">=</span> <span class="st">"TRBV"</span>,</span>
<span>         y.axis <span class="op">=</span> <span class="st">"TRBJ"</span>,</span>
<span>         plot <span class="op">=</span> <span class="st">"heatmap"</span>,  </span>
<span>         scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Lung</span></span>
<span><span class="fu"><a href="../reference/vizGenes.html">vizGenes</a></span><span class="op">(</span><span class="va">combined.TCR</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">6</span>,<span class="fl">8</span><span class="op">)</span><span class="op">]</span>, </span>
<span>         x.axis <span class="op">=</span> <span class="st">"TRBV"</span>,</span>
<span>         y.axis <span class="op">=</span> <span class="st">"TRBJ"</span>,</span>
<span>         plot <span class="op">=</span> <span class="st">"heatmap"</span>,  </span>
<span>         scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-5-2.png" width="700"></p>
<p>For the P17 patient samples, what if we are interested in chain
pairings, we can look at TRBV and TRAV at the same time using them as
inputs to <strong>x.axis</strong> and <strong>y.axis</strong>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/vizGenes.html">vizGenes</a></span><span class="op">(</span><span class="va">combined.TCR</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">]</span>, </span>
<span>         x.axis <span class="op">=</span> <span class="st">"TRBV"</span>,</span>
<span>         y.axis <span class="op">=</span> <span class="st">"TRAV"</span>,</span>
<span>         plot <span class="op">=</span> <span class="st">"heatmap"</span>, </span>
<span>         scale <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="percentgenes">percentGenes<a class="anchor" aria-label="anchor" href="#percentgenes"></a>
</h2>
<p>Quantify the proportion of V or J gene usage with
<code><a href="../reference/percentGenes.html">percentGenes()</a></code>. Like <code><a href="../reference/percentAA.html">percentAA()</a></code>, we select
the chain of interest and then indicate the gene of interest with the
<strong>gene</strong> parameter. Two major limitations of
<code><a href="../reference/percentGenes.html">percentGenes()</a></code> are, 1) the function quantifies only V or J
genes, and 2) the quantification of the genes are limited to all the V
or J genes seen across the samples, not all possible V or J genes.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/percentGenes.html">percentGenes</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>             chain <span class="op">=</span> <span class="st">"TRB"</span>, </span>
<span>             gene <span class="op">=</span> <span class="st">"Vgene"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>We can also use the output <code><a href="../reference/percentGenes.html">percentGenes()</a></code> for
dimensional reduction to summarise the gene usage by sample. This can be
done with a simple principal component analysis (below) or even more
complex reductions.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/percentGenes.html">percentGenes</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                         chain <span class="op">=</span> <span class="st">"TRB"</span>, </span>
<span>                         gene <span class="op">=</span> <span class="st">"Vgene"</span>, </span>
<span>                         exportTable <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Performing PCA</span></span>
<span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="va">df.genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Getting data frame to plot from</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">df.genes</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">df</span><span class="op">$</span><span class="va">PC1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">PC1</span><span class="op">)</span></span>
<span><span class="va">df</span><span class="op">$</span><span class="va">PC2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">PC2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Plotting</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">df</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>fill <span class="op">=</span><span class="va">df</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span><span class="op">)</span>, shape <span class="op">=</span> <span class="fl">21</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">guides</span><span class="op">(</span>fill<span class="op">=</span><span class="fu">guide_legend</span><span class="op">(</span>title<span class="op">=</span><span class="st">"Samples"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">scale_fill_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.colors</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span>, <span class="st">"inferno"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="percentvj">percentVJ<a class="anchor" aria-label="anchor" href="#percentvj"></a>
</h2>
<p>Quantify the proportion of V and J gene usage with
<code><a href="../reference/percentVJ.html">percentVJ()</a></code>. Like <code><a href="../reference/percentGenes.html">percentGenes()</a></code>, this
function will quantify the percentage of V and J paired together across
individual repertoires. The output can be visualized using a heatmap or
as input for further dimensional reduction.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/percentVJ.html">percentVJ</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>          chain <span class="op">=</span> <span class="st">"TRB"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/percentVJ.html">percentVJ</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                      chain <span class="op">=</span> <span class="st">"TRB"</span>, </span>
<span>                      exportTable <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Performing PCA</span></span>
<span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="va">df.genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Getting data frame to plot from</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">df.genes</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">df</span><span class="op">$</span><span class="va">PC1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">PC1</span><span class="op">)</span></span>
<span><span class="va">df</span><span class="op">$</span><span class="va">PC2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">PC2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Plotting</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">df</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>fill <span class="op">=</span><span class="va">df</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span><span class="op">)</span>, shape <span class="op">=</span> <span class="fl">21</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">guides</span><span class="op">(</span>fill<span class="op">=</span><span class="fu">guide_legend</span><span class="op">(</span>title<span class="op">=</span><span class="st">"Samples"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">scale_fill_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.colors</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span>, <span class="st">"inferno"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-9-2.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="percentkmer">percentKmer<a class="anchor" aria-label="anchor" href="#percentkmer"></a>
</h2>
<p>Another quantification of the composition of the CDR3 sequence is to
define motifs by sliding across the amino acid or nucleotide sequences
at set intervals resulting in substrings or kmers.</p>
<p><strong>motif.length</strong></p>
<ul>
<li>Numerical value for the length of the kmer.</li>
</ul>
<p><strong>top.motifs</strong></p>
<ul>
<li>Display the most variable genes determined by mean absolute
deviation (MAD).</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/percentKmer.html">percentKmer</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>            cloneCall <span class="op">=</span> <span class="st">"aa"</span>,</span>
<span>            chain <span class="op">=</span> <span class="st">"TRB"</span>, </span>
<span>            motif.length <span class="op">=</span> <span class="fl">3</span>, </span>
<span>            top.motifs <span class="op">=</span> <span class="fl">25</span><span class="op">)</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/percentKmer.html">percentKmer</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>            cloneCall <span class="op">=</span> <span class="st">"nt"</span>,</span>
<span>            chain <span class="op">=</span> <span class="st">"TRB"</span>, </span>
<span>            motif.length <span class="op">=</span> <span class="fl">3</span>, </span>
<span>            top.motifs <span class="op">=</span> <span class="fl">25</span><span class="op">)</span></span></code></pre></div>
<p><img src="Repertoire_Summary_files/figure-html/unnamed-chunk-10-2.png" width="700"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
