<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Single-cell Gene Set Enrichment Analysis • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Single-cell Gene Set Enrichment Analysis">
<meta property="og:description" content="scRepertoire">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ncborcherding/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Single-cell Gene Set Enrichment Analysis</h1>
            
            <h4 data-toc-skip class="date">Compiled: May 04, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ncborcherding/scRepertoire/blob/HEAD/vignettes/articles/Running_Escape.Rmd" class="external-link"><code>vignettes/articles/Running_Escape.Rmd</code></a></small>
      <div class="hidden name"><code>Running_Escape.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="what-is-escape">What is escape?<a class="anchor" aria-label="anchor" href="#what-is-escape"></a>
</h2>
<p><em>escape</em> is a R package designed for <strong>E</strong>asy
<strong>s</strong>ingle-<strong>c</strong>ell <strong>a</strong>nalysis
<strong>p</strong>latform for <strong>e</strong>nrichment. As the
tortuous acronym implies, <em>escape</em> was designed as a
user-friendly package for gene set enrichment analysis that leverages
the heterogeneity of single-cell data.</p>
<p>More information is available at the <a href="https://github.com/ncborcherding/escape" class="external-link">GitHub Repo</a>.</p>
<div class="section level3">
<h3 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h3>
<p>If using <em>escape</em>, please cite the <a href="https://www.nature.com/articles/s42003-020-01625-6" class="external-link">article</a>:
Borcherding, N., Vishwakarma, A., Voigt, A.P. et al. Mapping the immune
environment in clear cell renal carcinoma by single-cell genomics.
Commun Biol 4, 122 (2021). <a href="https://doi.org/10.1038/s42003-020-01625-6" class="external-link uri">https://doi.org/10.1038/s42003-020-01625-6</a>.</p>
</div>
<div class="section level3">
<h3 id="installing-escape">Installing escape<a class="anchor" aria-label="anchor" href="#installing-escape"></a>
</h3>
<p><em>escape</em> is available via github using:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"ncborcherding/escape"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"ncborcherding/escape"</span><span class="op">)</span></span></code></pre></div>
<p>For now, the newest version of <em>escape</em> is available in the
Bioconductor dev version (3.19).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span>version<span class="op">=</span><span class="st">'devel'</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"escape"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="loading-data">Loading Data<a class="anchor" aria-label="anchor" href="#loading-data"></a>
</h3>
<p>For the purposes of the vignette, we will use the scRepertoire full
example data. More information is available <a href="https://www.borch.dev/uploads/screpertoire/articles/loading#example-data-in-screpertoire" class="external-link">here</a>.
We will subset the data for patient 17 and 18 from the data set in order
to speed up the calculations.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">scRep_example</span>, <span class="va">Patient</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17"</span>, <span class="st">"P18"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="getting-gene-sets">Getting Gene Sets<a class="anchor" aria-label="anchor" href="#getting-gene-sets"></a>
</h2>
<div class="section level3">
<h3 id="option-1-molecular-signature-database">Option 1: Molecular Signature Database<a class="anchor" aria-label="anchor" href="#option-1-molecular-signature-database"></a>
</h3>
<p>The first step in the process of performing gene set enrichment
analysis is identifying the gene sets we would like to use. The function
<code>getGeneSets()</code> allows users to isolate a whole or multiple
libraries from a list of GSEABase GeneSetCollection objects.</p>
<p>We can do this for gene set collections from the built-in <a href="https://www.gsea-msigdb.org/gsea/msigdb/search.jsp" class="external-link">Molecular
Signature Database</a> by setting the parameter <strong>library</strong>
equal to library/libraries of interest. For multiple libraries, just set
<strong>library = c(“Library1”, “Library2”, etc)</strong>.</p>
<p>Additional parameters include:</p>
<ul>
<li>
<strong>subcategory</strong> Can be used to select subcategories in
the larger database.<br>
</li>
<li>
<strong>gene.sets</strong> Can select individual pathways/gene sets
to be isolated.</li>
</ul>
<p>If the sequencing of the single-cell data is performed on a species
other than “Homo sapiens”, make sure to use the <strong>species</strong>
parameter in <code>getGeneSets()</code> in order to get the correct gene
nomenclature.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GS.hallmark</span> <span class="op">&lt;-</span> <span class="fu">getGeneSets</span><span class="op">(</span>library <span class="op">=</span> <span class="st">"H"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="option-2-built-in-gene-sets">Option 2: Built-In gene sets<a class="anchor" aria-label="anchor" href="#option-2-built-in-gene-sets"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"escape.gene.sets"</span>, package<span class="op">=</span><span class="st">"escape"</span><span class="op">)</span></span>
<span><span class="va">gene.sets</span> <span class="op">&lt;-</span> <span class="va">escape.gene.sets</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="option-3-define-personal-gene-sets">Option 3: Define personal gene sets<a class="anchor" aria-label="anchor" href="#option-3-define-personal-gene-sets"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>gene.sets <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">Bcells =</span> <span class="fu">c</span>(<span class="st">"MS4A1"</span>,<span class="st">"CD79B"</span>,<span class="st">"CD79A"</span>,<span class="st">"IGH1"</span>,<span class="st">"IGH2"</span>)</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>                        <span class="at">Myeloid =</span> <span class="fu">c</span>(<span class="st">"SPI1"</span>,<span class="st">"FCER1G"</span>,<span class="st">"CSF1R"</span>),</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>                        <span class="at">Tcells =</span> <span class="fu">c</span>(<span class="st">"CD3E"</span>, <span class="st">"CD3D"</span>, <span class="st">"CD3G"</span>, <span class="st">"CD7"</span>,<span class="st">"CD8A"</span>))</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="performing-enrichment-calculation">Performing Enrichment Calculation<a class="anchor" aria-label="anchor" href="#performing-enrichment-calculation"></a>
</h2>
<p>Several popular methods exist for Gene Set Enrichment Analysis
(GSEA). These methods can vary in the underlying assumptions.
<em>escape</em> incorporates several methods that are particularly
advantageous for single-cell RNA values:</p>
<div class="section level4">
<h4 id="ssgsea">
<strong>ssGSEA</strong><a class="anchor" aria-label="anchor" href="#ssgsea"></a>
</h4>
<p>This method calculates the enrichment score using a rank-normalized
approach and generating an empirical cumulative distribution function
for each individual cell. The enrichment score is defined for a gene set
(<em>G</em>) using the number of genes in the gene set (<em>NG</em>) and
total number of genes (<em>N</em>).</p>
<p><span class="math display">\[
ES(G,S) \sum_{i = 1}^{n} [P_W^G(G,S,i)-P_{NG}(G,S,i)]
\]</span> Please see the following <a href="https://pubmed.ncbi.nlm.nih.gov/19847166/" class="external-link">citation</a> for more
information.</p>
</div>
<div class="section level4">
<h4 id="gsva">
<strong>GSVA</strong><a class="anchor" aria-label="anchor" href="#gsva"></a>
</h4>
<p>GSVA varies slightly by estimating a Poisson-based kernel cumulative
density function. But like ssGSEA, the ultimate enrichment score
reported is based on the maximum value of the random walk statistic.
GSVA appears to have better overall consistency and runs faster than
ssGSEA.</p>
<p><span class="math display">\[
ES_{jk}^{max} = V_{jk} [max_{l=1,...,p}|v_{jk}(l)]
\]</span> Please see the following <a href="https://pubmed.ncbi.nlm.nih.gov/23323831/" class="external-link">citation</a> for more
information.</p>
</div>
<div class="section level4">
<h4 id="aucell">
<strong>AUCell</strong><a class="anchor" aria-label="anchor" href="#aucell"></a>
</h4>
<p>In contrast to ssGSEA and GSVA, AUCell takes the gene rankings for
each cell and step-wise plots the position of each gene in the gene set
along the y-axis. The output score is the area under the curve for this
approach.</p>
<p>Please see the following <a href="https://pubmed.ncbi.nlm.nih.gov/28991892/" class="external-link">citation</a> for more
information.</p>
</div>
<div class="section level4">
<h4 id="ucell">UCell<a class="anchor" aria-label="anchor" href="#ucell"></a>
</h4>
<p>UCell calculates a Mann-Whitney U statistic based on the gene rank
list. <strong>Importantly</strong>, UCell has a cut-off for ranked genes
(<span class="math inline">\(r_{max}\)</span>) at 1500 - this is per
design as drop-out in single-cell can alter enrichment results. This
also substantially speeds the calculations up.</p>
<p>The enrichment score output is then calculated using the complement
of the U statistic scaled by the gene set size and cut-off.</p>
<p><span class="math display">\[
U_j^` = 1-\frac{U_j}{n \bullet r_{max}}
\]</span></p>
<p>Please see the following <a href="https://pubmed.ncbi.nlm.nih.gov/34285779/" class="external-link">citation</a> for more
information.</p>
</div>
<div class="section level3">
<h3 id="escape-matrix">escape.matrix<a class="anchor" aria-label="anchor" href="#escape-matrix"></a>
</h3>
<p>escape has 2 major functions - the first being
<code>escape.matrix()</code>, which serves as the backbone of enrichment
calculations. Using count-level data supplied from a single-cell object
or matrix, <code>escape.matrix()</code> will produce an enrichment score
for the individual cells with the gene sets selected and output the
values as a matrix.</p>
<p><strong>method</strong></p>
<ul>
<li>AUCell</li>
<li>GSVA</li>
<li>ssGSEA</li>
<li>UCell</li>
</ul>
<p><strong>groups</strong></p>
<ul>
<li>The number of cells to calculate at once.</li>
</ul>
<p><strong>min.size</strong></p>
<ul>
<li>The minimum size of detectable genes in a gene set. Gene sets less
than the <strong>min.size</strong> will be removed before the
calculation.</li>
</ul>
<p><strong>normalize</strong></p>
<ul>
<li>Use the number of genes from the gene sets in each cell to normalize
the enrichment scores. The default value is <strong>FALSE</strong>.</li>
</ul>
<p><strong>make.positive</strong></p>
<ul>
<li>During normalization, whether to shift the enrichment values to a
positive range (<strong>TRUE</strong>) or not (<strong>FALSE</strong>).
The default value is <strong>FALSE</strong>.</li>
</ul>
<p><em>Cautionary note:</em> <strong>make.positive</strong> was added to
allow for differential analysis downstream of enrichment as some methods
may produce negative values. It preserves log-fold change, but
ultimately modifies the enrichment values and should be used with
caution.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enrichment.scores</span> <span class="op">&lt;-</span> <span class="fu">escape.matrix</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                                   gene.sets <span class="op">=</span> <span class="va">GS.hallmark</span>, </span>
<span>                                   groups <span class="op">=</span> <span class="fl">5000</span>, </span>
<span>                                   min.size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>We can visualize the raw results with a ggplot:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">enrichment.scores</span><span class="op">)</span>, </span>
<span>      mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">`HALLMARK-DNA-REPAIR`</span>, <span class="va">`HALLMARK-G2M-CHECKPOINT`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="runescape">runEscape<a class="anchor" aria-label="anchor" href="#runescape"></a>
</h3>
<p>Alternatively, we can use <code>runEscape()</code> to calculate the
enrichment score and directly attach the output to a single-cell object.
The additional parameter for ``<code>runEscape</code> is
<strong>new.assay.name</strong>, in order to save the enrichment scores
as a custom assay in the single-cell object.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu">runEscape</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                           method <span class="op">=</span> <span class="st">"ssGSEA"</span>,</span>
<span>                           gene.sets <span class="op">=</span> <span class="va">GS.hallmark</span>, </span>
<span>                           groups <span class="op">=</span> <span class="fl">5000</span>, </span>
<span>                           min.size <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                           new.assay.name <span class="op">=</span> <span class="st">"escape.ssGSEA"</span><span class="op">)</span></span></code></pre></div>
<p>We can quickly examine the attached enrichment scores using the
visualization/workflow we prefer - here we will use just
<code>FeaturePlot()</code> from the Seurat R package.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Define color palette </span></span>
<span><span class="va">colorblind_vector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.colors</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">7</span>, palette <span class="op">=</span> <span class="st">"inferno"</span>, fixup <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">FeaturePlot</span><span class="op">(</span><span class="va">scRep_example</span>, <span class="st">"HALLMARK-APOPTOSIS"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">scale_color_gradientn</span><span class="op">(</span>colors <span class="op">=</span> <span class="va">colorblind_vector</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="performnormalization">performNormalization<a class="anchor" aria-label="anchor" href="#performnormalization"></a>
</h3>
<p>Although we glossed over the normalization that can be used in
<code>escape.matrix()</code> and <code>runEscape()</code>, it is worth
mentioning here as normalization can affect all downstream analyses.</p>
<p>There can be inherent bias in enrichment values due to drop out in
single-cell expression data. Cells with larger numbers of features and
counts will likely have higher enrichment values.
<code>performNormalization()</code> will normalize the enrichment values
by calculating the number of genes expressed in each gene set and cell.
This is similar to the normalization in classic GSEA and it will be
stored in a new assay.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu">performNormalization</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                                      assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>, </span>
<span>                                      gene.sets <span class="op">=</span> <span class="va">GS.hallmark</span><span class="op">)</span></span></code></pre></div>
<p>An alternative for scaling by expressed gene sets would be to use a
scaling factor previously calculated during normal single-cell data
processing and quality control. This can be done using the
<strong>scale.factor</strong> argument and providing a vector. This
approach may be particularly advantageous for large data sets.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu">performNormalization</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                                      assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>, </span>
<span>                                      gene.sets <span class="op">=</span> <span class="va">GS.hallmark</span>, </span>
<span>                                      scale.factor <span class="op">=</span> <span class="va">scRep_example</span><span class="op">$</span><span class="va">nFeature_RNA</span><span class="op">)</span></span></code></pre></div>
<p><code>performNormalization()</code> has an additional parameter
<strong>make.positive</strong>. Across the individual gene sets, if
negative normalized enrichment scores are seen, the minimum value is
added to all values. For example if the normalized enrichment scores
(after the above accounting for drop out) ranges from -50 to 50,
<strong>make.positive</strong> will adjust the range to 0 to 100 (by
adding 50). This allows for compatible log2-fold change downstream, but
can alter the enrichment score interpretation.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="visualizations">Visualizations<a class="anchor" aria-label="anchor" href="#visualizations"></a>
</h2>
<p>There are a number of ways to look at the enrichment values
downstream of <code>runEscape()</code> with the myriad plotting and
visualizations functions/packages for single-cell data. <em>escape</em>
include several additional plotting functions to assist in the
analysis.</p>
<div class="section level3">
<h3 id="heatmapenrichment">heatmapEnrichment<a class="anchor" aria-label="anchor" href="#heatmapenrichment"></a>
</h3>
<p>We can examine the enrichment values across our gene sets by using
<code>heatmapEnrichment()</code>. This visualization will return the
mean of the <strong>group.by</strong> variable. As a default - all
visualizations of single-cell objects will use the cluster assignment or
active identity as a default for visualizations. We will use a subset of
the Hallmark Gene Sets to display for the vignette format using the
<strong>gene.set.use</strong> set to first 12 gene sets. Alternatively,
we can plot all gene sets using <strong>gene.set.use</strong> =
“all”.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">heatmapEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                  group.by <span class="op">=</span> <span class="st">"ident"</span>,</span>
<span>                  gene.set.use <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scRep_example</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">escape.ssGSEA</span><span class="op">@</span><span class="va">data</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">12</span><span class="op">]</span>,</span>
<span>                  assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-13-1.png" width="700">
Most of the visualizations in <em>escape</em> have a defined set of
parameters.</p>
<p><strong>group.by</strong></p>
<ul>
<li>The grouping variable for the comparison.</li>
</ul>
<p><strong>facet.by</strong></p>
<ul>
<li>Using a variable to facet the graph into separate
visualizations.</li>
</ul>
<p><strong>scale</strong></p>
<ul>
<li>
<strong>TRUE</strong> - z-transform the enrichment values.</li>
<li>
<strong>FALSE</strong> - leave raw values
(<strong>default</strong>).</li>
</ul>
<p>In addition, <code>heatmapEnrichment()</code> allows for the
reclustering of rows and columns using Euclidean distance of the
enrichment scores and the Ward2 methods for clustering using
<strong>cluster.rows</strong> and <strong>cluster.columns</strong>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">heatmapEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                  group.by <span class="op">=</span> <span class="st">"ident"</span>,</span>
<span>                  assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>,</span>
<span>                  gene.set.use <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scRep_example</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">escape.ssGSEA</span><span class="op">@</span><span class="va">data</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">12</span><span class="op">]</span>,</span>
<span>                  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                  cluster.rows <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                  cluster.columns <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>Each visualization has an additional argument called
<strong>palette</strong> that supplies the coloring scheme to be used -
available color palettes can be viewed with <code><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.pals()</a></code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.pals</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   [1] "Pastel 1"      "Dark 2"        "Dark 3"        "Set 2"        </span></span>
<span><span class="co">##   [5] "Set 3"         "Warm"          "Cold"          "Harmonic"     </span></span>
<span><span class="co">##   [9] "Dynamic"       "Grays"         "Light Grays"   "Blues 2"      </span></span>
<span><span class="co">##  [13] "Blues 3"       "Purples 2"     "Purples 3"     "Reds 2"       </span></span>
<span><span class="co">##  [17] "Reds 3"        "Greens 2"      "Greens 3"      "Oslo"         </span></span>
<span><span class="co">##  [21] "Purple-Blue"   "Red-Purple"    "Red-Blue"      "Purple-Orange"</span></span>
<span><span class="co">##  [25] "Purple-Yellow" "Blue-Yellow"   "Green-Yellow"  "Red-Yellow"   </span></span>
<span><span class="co">##  [29] "Heat"          "Heat 2"        "Terrain"       "Terrain 2"    </span></span>
<span><span class="co">##  [33] "Viridis"       "Plasma"        "Inferno"       "Rocket"       </span></span>
<span><span class="co">##  [37] "Mako"          "Dark Mint"     "Mint"          "BluGrn"       </span></span>
<span><span class="co">##  [41] "Teal"          "TealGrn"       "Emrld"         "BluYl"        </span></span>
<span><span class="co">##  [45] "ag_GrnYl"      "Peach"         "PinkYl"        "Burg"         </span></span>
<span><span class="co">##  [49] "BurgYl"        "RedOr"         "OrYel"         "Purp"         </span></span>
<span><span class="co">##  [53] "PurpOr"        "Sunset"        "Magenta"       "SunsetDark"   </span></span>
<span><span class="co">##  [57] "ag_Sunset"     "BrwnYl"        "YlOrRd"        "YlOrBr"       </span></span>
<span><span class="co">##  [61] "OrRd"          "Oranges"       "YlGn"          "YlGnBu"       </span></span>
<span><span class="co">##  [65] "Reds"          "RdPu"          "PuRd"          "Purples"      </span></span>
<span><span class="co">##  [69] "PuBuGn"        "PuBu"          "Greens"        "BuGn"         </span></span>
<span><span class="co">##  [73] "GnBu"          "BuPu"          "Blues"         "Lajolla"      </span></span>
<span><span class="co">##  [77] "Turku"         "Hawaii"        "Batlow"        "Blue-Red"     </span></span>
<span><span class="co">##  [81] "Blue-Red 2"    "Blue-Red 3"    "Red-Green"     "Purple-Green" </span></span>
<span><span class="co">##  [85] "Purple-Brown"  "Green-Brown"   "Blue-Yellow 2" "Blue-Yellow 3"</span></span>
<span><span class="co">##  [89] "Green-Orange"  "Cyan-Magenta"  "Tropic"        "Broc"         </span></span>
<span><span class="co">##  [93] "Cork"          "Vik"           "Berlin"        "Lisbon"       </span></span>
<span><span class="co">##  [97] "Tofino"        "ArmyRose"      "Earth"         "Fall"         </span></span>
<span><span class="co">## [101] "Geyser"        "TealRose"      "Temps"         "PuOr"         </span></span>
<span><span class="co">## [105] "RdBu"          "RdGy"          "PiYG"          "PRGn"         </span></span>
<span><span class="co">## [109] "BrBG"          "RdYlBu"        "RdYlGn"        "Spectral"     </span></span>
<span><span class="co">## [113] "Zissou 1"      "Cividis"       "Roma"</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">heatmapEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                  group.by <span class="op">=</span> <span class="st">"ident"</span>,</span>
<span>                  gene.set.use <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scRep_example</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">escape.ssGSEA</span><span class="op">@</span><span class="va">data</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">12</span><span class="op">]</span>,</span>
<span>                  assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>,</span>
<span>                  palette <span class="op">=</span> <span class="st">"Spectral"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>Alternatively, we can add an additional layer to the ggplot object
that is returned by the visualizations using something like
<code>scale_fill_gradientn()</code> for continuous values or
<code>scale_fill_manual()</code> for the categorical variables.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">heatmapEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                  group.by <span class="op">=</span> <span class="st">"ident"</span>,</span>
<span>                  gene.set.use <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scRep_example</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">escape.ssGSEA</span><span class="op">@</span><span class="va">data</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">12</span><span class="op">]</span>,</span>
<span>                  assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">scale_fill_gradientn</span><span class="op">(</span>colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">11</span>, <span class="st">"RdYlBu"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="geyserenrichment">geyserEnrichment<a class="anchor" aria-label="anchor" href="#geyserenrichment"></a>
</h3>
<p>We can also focus on individual gene sets - one approach is to use
<code>geyserEnrichment()</code>. Here individual cells are plotted along
the Y-axis with graphical summary where the central dot refers to the
median enrichment value and the thicker/thinner lines demonstrate the
interval summaries referring to the 66% and 95%.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">geyserEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                 assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>,</span>
<span>                 gene.set <span class="op">=</span> <span class="st">"HALLMARK-INTERFERON-GAMMA-RESPONSE"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<p>To show the additional parameters that appear in visualizations of
individual enrichment gene sets - we can reorder the groups by the mean
of the gene set using <strong>order.by</strong> = “mean”.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">geyserEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                 assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>,</span>
<span>                 gene.set <span class="op">=</span> <span class="st">"HALLMARK-INTERFERON-GAMMA-RESPONSE"</span>, </span>
<span>                 order.by <span class="op">=</span> <span class="st">"mean"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<p>What if we had 2 separate samples or groups within the data? Another
parameter we can use is <strong>facet.by</strong> to allow for direct
visualization of an additional variable.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">geyserEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                 assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>,</span>
<span>                 gene.set <span class="op">=</span> <span class="st">"HALLMARK-INTERFERON-GAMMA-RESPONSE"</span>, </span>
<span>                 facet.by <span class="op">=</span> <span class="st">"Type"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-20-1.png" width="700">
Lastly, we can select the way the color is applied to the plot using the
<strong>color.by</strong> parameter. Here we can set it to the gene set
of interest <em>“HALLMARK-INTERFERON-GAMMA-RESPONSE”</em>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">geyserEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                 assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>,</span>
<span>                 gene.set <span class="op">=</span> <span class="st">"HALLMARK-INTERFERON-GAMMA-RESPONSE"</span>, </span>
<span>                 color.by  <span class="op">=</span> <span class="st">"HALLMARK-INTERFERON-GAMMA-RESPONSE"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="ridgeenrichment">ridgeEnrichment<a class="anchor" aria-label="anchor" href="#ridgeenrichment"></a>
</h3>
<p>Similar to the <code>geyserEnrichment()</code> the
<code>ridgeEnrichment()</code> can display the distribution of
enrichment values across the selected gene set. The central line is at
the median value for the respective grouping.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ridgeEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>,</span>
<span>                gene.set <span class="op">=</span> <span class="st">"HALLMARK-IL2-STAT5-SIGNALING"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-22-1.png" width="700">
We can get the relative position of individual cells along the x-axis
using the <strong>add.rug</strong> parameter.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ridgeEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>,</span>
<span>                gene.set <span class="op">=</span> <span class="st">"HALLMARK-IL2-STAT5-SIGNALING"</span>,</span>
<span>                add.rug <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="splitenrichment">splitEnrichment<a class="anchor" aria-label="anchor" href="#splitenrichment"></a>
</h3>
<p>Another distribution visualization is a violin plot, which we
separate and directly compare using a binary classification. Like
<code>ridgeEnrichment()</code>, this allows for greater use of
categorical variables. For <code>splitEnrichment()</code>, the output
will be two halves of a violin plot based on the
<strong>split.by</strong> parameter with a central boxplot with the
relative distribution across all samples. Here, we split the samples by
<em>Type</em> where <strong>B</strong> refers to the peripheral blood
and <strong>L</strong> refers to lung.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">splitEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>,</span>
<span>                gene.set <span class="op">=</span> <span class="st">"HALLMARK-IL2-STAT5-SIGNALING"</span>, </span>
<span>                split.by <span class="op">=</span> <span class="st">"Type"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="densityenrichment">densityEnrichment<a class="anchor" aria-label="anchor" href="#densityenrichment"></a>
</h3>
<p><code>densityEnrichment()</code> is a method to visualize the mean
rank position of the gene set features along the total feature space by
group. This is similar to traditional GSEA analysis, but is not
calculating the walk-based enrichment score. As this function calculates
mean ranks on the fly, it may take some time depending on the size of
the single-cell object.</p>
<p><strong>gene.set.use</strong></p>
<ul>
<li>The selected gene set to visualize</li>
</ul>
<p><strong>gene.sets</strong></p>
<ul>
<li>The gene set library from either of the 3 options in the first
section of the vignette.</li>
</ul>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">densityEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                  gene.set.use <span class="op">=</span> <span class="st">"HALLMARK-IL6-JAK-STAT3-SIGNALING"</span>, </span>
<span>                  group.by <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>                  gene.sets <span class="op">=</span> <span class="va">GS.hallmark</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="scatterenrichment">scatterEnrichment<a class="anchor" aria-label="anchor" href="#scatterenrichment"></a>
</h3>
<p>It may be advantageous to look at the distribution of multiple gene
sets - here we can use <code>scatterEnrichment()</code> for a 2 gene set
comparison. The color values are based on the density of points
determined by the number of neighbors, similar to the <a href="https://www.bioconductor.org/packages/release/bioc/html/Nebulosa.html" class="external-link">Nebulosa
R package</a>. We just need to define which gene set to plot on the
<strong>x.axis</strong> and which to plot on the
<strong>y.axis</strong>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scatterEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                  assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>,</span>
<span>                  x.axis <span class="op">=</span> <span class="st">"HALLMARK-INTERFERON-GAMMA-RESPONSE"</span>,</span>
<span>                  y.axis <span class="op">=</span> <span class="st">"HALLMARK-IL6-JAK-STAT3-SIGNALING"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
<p>The scatter plot can also be converted into a hexbin, another method
for summarizing the individual cell distributions along the x and y
axis, by setting <strong>style</strong> = “hex”.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scatterEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                  assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>,</span>
<span>                  x.axis <span class="op">=</span> <span class="st">"HALLMARK-INTERFERON-GAMMA-RESPONSE"</span>,</span>
<span>                  y.axis <span class="op">=</span> <span class="st">"HALLMARK-IL6-JAK-STAT3-SIGNALING"</span>, </span>
<span>                  style <span class="op">=</span> <span class="st">"hex"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="statistical-analysis">Statistical Analysis<a class="anchor" aria-label="anchor" href="#statistical-analysis"></a>
</h2>
<div class="section level3">
<h3 id="principal-component-analysis-pca">Principal Component Analysis (PCA)<a class="anchor" aria-label="anchor" href="#principal-component-analysis-pca"></a>
</h3>
<p>escape has its own PCA function <code>performPCA()</code> which will
work on a single-cell object or a matrix of enrichment values. This is
specifically useful for downstream visualizations as it stores the
eigenvalues and rotations. If we want to look at the relative
contribution to overall variance of each component or a Biplot-like
overlay of the individual features, use <code>performPCA()</code>.</p>
<p>Alternatively, other PCA-based functions like Seurat’s
<code>RunPCA()</code> or scater’s ``<code>runPCA()</code> can be used.
These functions are likely faster and would be ideal if we have a larger
number of cells and/or gene sets.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu">performPCA</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                            assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>,</span>
<span>                            n.dim <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><em>escape</em> has a built in method for plotting PCA
<code>pcaEnrichment()</code> that functions similarly to the
<code>scatterEnrichment()</code> function where <strong>x.axis</strong>
and <strong>y.axis</strong> are the components to plot.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pcaEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>              dimRed <span class="op">=</span> <span class="st">"escape.PCA"</span>,</span>
<span>              x.axis <span class="op">=</span> <span class="st">"PC1"</span>,</span>
<span>              y.axis <span class="op">=</span> <span class="st">"PC2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
<p><code>pcaEnrichment()</code> can plot additional information on the
principal component analysis.</p>
<p><strong>add.percent.contribution</strong> will add the relative
percent contribution of the x and y.axis to total variability observed
in the PCA.</p>
<p><strong>display.factors</strong> will overlay the magnitude and
direction that the features/gene sets contribute to the selected
components. The number of gene sets is determined by
<strong>number.of.factors</strong>. This can assist in understanding the
underlying differences in enrichment across different cells.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pcaEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>              dimRed <span class="op">=</span> <span class="st">"escape.PCA"</span>,</span>
<span>              x.axis <span class="op">=</span> <span class="st">"PC1"</span>,</span>
<span>              y.axis <span class="op">=</span> <span class="st">"PC2"</span>,</span>
<span>              add.percent.contribution <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>              display.factors <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>              number.of.factors <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-30-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="differential-enrichment">Differential Enrichment<a class="anchor" aria-label="anchor" href="#differential-enrichment"></a>
</h3>
<p>Differential enrichment analysis can be performed similar to
differential gene expression analysis. For the purposes of finding the
differential enrichment values, we can first normalize the enrichment
values for the ssGSEA calculations. Here we are using the
<strong>scale.factor</strong> parameter of <em>nFeature_RNA</em>, which
is the total feature space for each cell.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu">performNormalization</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                                      assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>, </span>
<span>                                      gene.sets <span class="op">=</span> <span class="va">GS.hallmark</span>, </span>
<span>                                      scale.factor <span class="op">=</span> <span class="va">scRep_example</span><span class="op">$</span><span class="va">nFeature_RNA</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Normalizing enrichment scores per cell..."</span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all.markers</span> <span class="op">&lt;-</span> <span class="fu">FindAllMarkers</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                              assay <span class="op">=</span> <span class="st">"escape.ssGSEA_normalized"</span>, </span>
<span>                              min.pct <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                              logfc.threshold <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">all.markers</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                                 p_val avg_log2FC pct.1 pct.2</span></span>
<span><span class="co">## HALLMARK-REACTIVE-OXYGEN-SPECIES-PATHWAY 1.840790e-50  0.4954612 0.994 0.927</span></span>
<span><span class="co">## HALLMARK-OXIDATIVE-PHOSPHORYLATION       1.730400e-42  0.2908800 1.000 0.998</span></span>
<span><span class="co">## HALLMARK-ADIPOGENESIS                    5.911400e-36  0.1871677 1.000 1.000</span></span>
<span><span class="co">## HALLMARK-APICAL-JUNCTION                 7.642810e-19  0.2126941 0.983 0.921</span></span>
<span><span class="co">## HALLMARK-DNA-REPAIR                      1.817561e-18  0.2708074 0.985 0.945</span></span>
<span><span class="co">## HALLMARK-UV-RESPONSE-DN                  2.792584e-17 -4.3160697 1.000 1.000</span></span>
<span><span class="co">##                                             p_val_adj cluster</span></span>
<span><span class="co">## HALLMARK-REACTIVE-OXYGEN-SPECIES-PATHWAY 9.203948e-49       1</span></span>
<span><span class="co">## HALLMARK-OXIDATIVE-PHOSPHORYLATION       8.651999e-41       1</span></span>
<span><span class="co">## HALLMARK-ADIPOGENESIS                    2.955700e-34       1</span></span>
<span><span class="co">## HALLMARK-APICAL-JUNCTION                 3.821405e-17       1</span></span>
<span><span class="co">## HALLMARK-DNA-REPAIR                      9.087807e-17       1</span></span>
<span><span class="co">## HALLMARK-UV-RESPONSE-DN                  1.396292e-15       1</span></span>
<span><span class="co">##                                                                              gene</span></span>
<span><span class="co">## HALLMARK-REACTIVE-OXYGEN-SPECIES-PATHWAY HALLMARK-REACTIVE-OXYGEN-SPECIES-PATHWAY</span></span>
<span><span class="co">## HALLMARK-OXIDATIVE-PHOSPHORYLATION             HALLMARK-OXIDATIVE-PHOSPHORYLATION</span></span>
<span><span class="co">## HALLMARK-ADIPOGENESIS                                       HALLMARK-ADIPOGENESIS</span></span>
<span><span class="co">## HALLMARK-APICAL-JUNCTION                                 HALLMARK-APICAL-JUNCTION</span></span>
<span><span class="co">## HALLMARK-DNA-REPAIR                                           HALLMARK-DNA-REPAIR</span></span>
<span><span class="co">## HALLMARK-UV-RESPONSE-DN                                   HALLMARK-UV-RESPONSE-DN</span></span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
