<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Single-cell Gene Set Enrichment Analysis • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Single-cell Gene Set Enrichment Analysis">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/dandelionR.html">VDJ Trajectory Analysis with dandelionR</a>
    </li>
    <li>
      <a href="../articles/immunarch.html">Using scRepertoire with immunarch</a>
    </li>
    <li>
      <a href="../articles/tcrpheno.html">TCRpheno applied to T cell fate</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BorchLab/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Single-cell Gene Set Enrichment Analysis</h1>
            
            <h4 data-toc-skip class="date">Compiled: July 31, 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BorchLab/scRepertoire/blob/HEAD/vignettes/articles/Running_Escape.Rmd" class="external-link"><code>vignettes/articles/Running_Escape.Rmd</code></a></small>
      <div class="hidden name"><code>Running_Escape.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p><em>escape</em> is a R package designed for <strong>E</strong>asy
<strong>s</strong>ingle-<strong>c</strong>ell <strong>a</strong>nalysis
<strong>p</strong>latform for <strong>e</strong>nrichment. As the
tortuous acronym implies, <em>escape</em> was designed as a
user-friendly package for gene set enrichment analysis that leverages
the heterogeneity of single-cell data. escape turns raw single-cell
counts into intuitive, per-cell gene-set scores with a single command
and then provides plotting helpers to interrogate them.</p>
<p>The core workflow is:</p>
<ol style="list-style-type: decimal">
<li>Choose gene-set library (<code>getGeneSets()</code> or your own
list)</li>
<li>Score cells (<code>runEscape()</code>)</li>
<li>(Optional) Normalize for drop-out
(<code>performNormalization()</code>)</li>
<li>Explore with the built-in visualization gallery</li>
</ol>
<p>More information is available at the <a href="https://github.com/BorchLab/escape" class="external-link">GitHub Repo</a>.</p>
<div class="section level3">
<h3 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h3>
<p>If using <em>escape</em>, please cite the <a href="https://www.nature.com/articles/s42003-020-01625-6" class="external-link">article</a>:
Borcherding, N., Vishwakarma, A., Voigt, A.P. et al. Mapping the immune
environment in clear cell renal carcinoma by single-cell genomics.
Commun Biol 4, 122 (2021).</p>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"BorchLab/escape"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"escape"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="loading-data">Loading Data<a class="anchor" aria-label="anchor" href="#loading-data"></a>
</h2>
<p>For the purposes of the vignette, we will use the scRepertoire full
example data. More information is available <a href="https://www.borch.dev/uploads/screpertoire/articles/loading#example-data-in-screpertoire" class="external-link">here</a>.
We will subset the data for patient 17 and 18 from the data set in order
to speed up the calculations.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">scRep_example</span>, <span class="va">Patient</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P17"</span>, <span class="st">"P18"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="getting-gene-sets">Getting Gene Sets<a class="anchor" aria-label="anchor" href="#getting-gene-sets"></a>
</h2>
<div class="section level3">
<h3 id="option-1-built-in-gene-sets">Option 1: Built-In gene sets<a class="anchor" aria-label="anchor" href="#option-1-built-in-gene-sets"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"escape.gene.sets"</span>, package<span class="op">=</span><span class="st">"escape"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="option-2-msigdb-via-getgenesets">Option 2: MSigDB via <code>getGeneSets()</code><a class="anchor" aria-label="anchor" href="#option-2-msigdb-via-getgenesets"></a>
</h3>
<p>Gene set enrichment analysis begins by identifying the appropriate
gene sets for your study. The <code>getGeneSets()</code> function
simplifies this process by extracting one or more gene set libraries
from the Molecular Signature Database (MSigDB) and returning them as a
GSEABase GeneSetCollection object. Note that the first time you run
<code>getGeneSets()</code>, it downloads a complete local copy of the
gene sets, which may take a little while. Future calls will use the
cached version, greatly improving performance.</p>
<p>To retrieve gene set collections from MSigDB, specify the library or
libraries of interest using the library parameter. For example, to
select multiple libraries, use:</p>
<ul>
<li>
<strong>subcategory</strong>: Narrow down your selection by
specifying subcategories within a library. Examples include “CGN”,
“CGP”, “CP:BIOCARTA”, “CP:KEGG”, “<a href="GO:BP" class="uri">GO:BP</a>”, “IMMUNESIGDB”, etc.</li>
<li>
<strong>gene.sets:</strong> Isolate individual pathways or gene sets
by providing their specific names.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GS.hallmark</span> <span class="op">&lt;-</span> <span class="fu">getGeneSets</span><span class="op">(</span>library <span class="op">=</span> <span class="st">"H"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="option-3-define-personal-gene-sets">Option 3: Define personal gene sets<a class="anchor" aria-label="anchor" href="#option-3-define-personal-gene-sets"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene.sets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Bcells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MS4A1"</span>,<span class="st">"CD79B"</span>,<span class="st">"CD79A"</span>,<span class="st">"IGH1"</span>,<span class="st">"IGH2"</span><span class="op">)</span>,</span>
<span>                        Myeloid <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SPI1"</span>,<span class="st">"FCER1G"</span>,<span class="st">"CSF1R"</span><span class="op">)</span>,</span>
<span>                        Tcells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3E"</span>, <span class="st">"CD3D"</span>, <span class="st">"CD3G"</span>, <span class="st">"CD7"</span>,<span class="st">"CD8A"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="option-4-using-msigdbr">Option 4: Using msigdbr<a class="anchor" aria-label="anchor" href="#option-4-using-msigdbr"></a>
</h3>
<p><a href="https://cran.r-project.org/web/packages/msigdbr/index.html" class="external-link">msigdbr</a>
is an alternative R package to access the Molecular Signature Database
in R. There is expanded support for species in the package as well as a
mix of accessible versus downloadable gene sets, so it can be faster
than caching a copy locally.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GS.hallmark</span> <span class="op">&lt;-</span> <span class="fu">msigdbr</span><span class="op">(</span>species  <span class="op">=</span> <span class="st">"Homo sapiens"</span>,  </span>
<span>                       category <span class="op">=</span> <span class="st">"H"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="performing-enrichment-calculation">Performing Enrichment Calculation<a class="anchor" aria-label="anchor" href="#performing-enrichment-calculation"></a>
</h2>
<p>Several popular methods exist for Gene Set Enrichment Analysis
(GSEA). These methods can vary in the underlying assumptions.
<em>escape</em> incorporates several methods that are particularly
advantageous for single-cell RNA values:</p>
<div class="section level4">
<h4 id="ssgsea">
<strong>ssGSEA</strong><a class="anchor" aria-label="anchor" href="#ssgsea"></a>
</h4>
<p>This method calculates the enrichment score using a rank-normalized
approach and generating an empirical cumulative distribution function
for each individual cell. The enrichment score is defined for a gene set
(<em>G</em>) using the number of genes in the gene set (<em>NG</em>) and
total number of genes (<em>N</em>).</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>E</mi><mi>S</mi><mo stretchy="false" form="prefix">(</mo><mi>G</mi><mo>,</mo><mi>S</mi><mo stretchy="false" form="postfix">)</mo><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><mo stretchy="false" form="prefix">[</mo><msubsup><mi>P</mi><mi>W</mi><mi>G</mi></msubsup><mo stretchy="false" form="prefix">(</mo><mi>G</mi><mo>,</mo><mi>S</mi><mo>,</mo><mi>i</mi><mo stretchy="false" form="postfix">)</mo><mo>−</mo><msub><mi>P</mi><mrow><mi>N</mi><mi>G</mi></mrow></msub><mo stretchy="false" form="prefix">(</mo><mi>G</mi><mo>,</mo><mi>S</mi><mo>,</mo><mi>i</mi><mo stretchy="false" form="postfix">)</mo><mo stretchy="false" form="postfix">]</mo></mrow><annotation encoding="application/x-tex">
ES(G,S) \sum_{i = 1}^{n} [P_W^G(G,S,i)-P_{NG}(G,S,i)]
</annotation></semantics></math></p>
<p>Please see the following <a href="https://pubmed.ncbi.nlm.nih.gov/19847166/" class="external-link">citation</a> for more
information.</p>
</div>
<div class="section level4">
<h4 id="gsva">
<strong>GSVA</strong><a class="anchor" aria-label="anchor" href="#gsva"></a>
</h4>
<p>GSVA varies slightly by estimating a Poisson-based kernel cumulative
density function. But like ssGSEA, the ultimate enrichment score
reported is based on the maximum value of the random walk statistic.
GSVA appears to have better overall consistency and runs faster than
ssGSEA.</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>E</mi><msubsup><mi>S</mi><mrow><mi>j</mi><mi>k</mi></mrow><mrow><mi>m</mi><mi>a</mi><mi>x</mi></mrow></msubsup><mo>=</mo><msub><mi>V</mi><mrow><mi>j</mi><mi>k</mi></mrow></msub><mo stretchy="false" form="prefix">[</mo><mi>m</mi><mi>a</mi><msub><mi>x</mi><mrow><mi>l</mi><mo>=</mo><mn>1</mn><mo>,</mo><mi>.</mi><mi>.</mi><mi>.</mi><mo>,</mo><mi>p</mi></mrow></msub><mo stretchy="false" form="prefix">|</mo><msub><mi>v</mi><mrow><mi>j</mi><mi>k</mi></mrow></msub><mo stretchy="false" form="prefix">(</mo><mi>l</mi><mo stretchy="false" form="postfix">)</mo><mo stretchy="false" form="postfix">]</mo></mrow><annotation encoding="application/x-tex">
ES_{jk}^{max} = V_{jk} [max_{l=1,...,p}|v_{jk}(l)]
</annotation></semantics></math></p>
<p>Please see the following <a href="https://pubmed.ncbi.nlm.nih.gov/23323831/" class="external-link">citation</a> for more
information.</p>
</div>
<div class="section level4">
<h4 id="aucell">
<strong>AUCell</strong><a class="anchor" aria-label="anchor" href="#aucell"></a>
</h4>
<p>In contrast to ssGSEA and GSVA, AUCell takes the gene rankings for
each cell and step-wise plots the position of each gene in the gene set
along the y-axis. The output score is the area under the curve for this
approach.</p>
<p>Please see the following <a href="https://pubmed.ncbi.nlm.nih.gov/28991892/" class="external-link">citation</a> for more
information.</p>
</div>
<div class="section level4">
<h4 id="ucell">UCell<a class="anchor" aria-label="anchor" href="#ucell"></a>
</h4>
<p>UCell calculates a Mann-Whitney U statistic based on the gene rank
list. <strong>Importantly</strong>, UCell has a cut-off for ranked genes
(<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>r</mi><mrow><mi>m</mi><mi>a</mi><mi>x</mi></mrow></msub><annotation encoding="application/x-tex">r_{max}</annotation></semantics></math>)
at 1500 - this is per design as drop-out in single-cell can alter
enrichment results. This also substantially speeds the calculations
up.</p>
<p>The enrichment score output is then calculated using the complement
of the U statistic scaled by the gene set size and cut-off.</p>
<p><span class="math display">$$
U_j^` = 1-\frac{U_j}{n \bullet r_{max}}
$$</span></p>
<p>Please see the following <a href="https://pubmed.ncbi.nlm.nih.gov/34285779/" class="external-link">citation</a> for more
information.</p>
</div>
<div class="section level3">
<h3 id="escape-matrix">escape.matrix<a class="anchor" aria-label="anchor" href="#escape-matrix"></a>
</h3>
<p>escape has 2 major functions - the first being
<code>escape.matrix()</code>, which serves as the backbone of enrichment
calculations. Using count-level data supplied from a single-cell object
or matrix, <code>escape.matrix()</code> will produce an enrichment score
for the individual cells with the gene sets selected and output the
values as a matrix.</p>
<p><strong>method</strong></p>
<ul>
<li>AUCell</li>
<li>GSVA</li>
<li>ssGSEA</li>
<li>UCell</li>
</ul>
<p><strong>groups</strong></p>
<ul>
<li>The number of cells to calculate at once.</li>
</ul>
<p><strong>min.size</strong></p>
<ul>
<li>The minimum size of detectable genes in a gene set. Gene sets less
than the <strong>min.size</strong> will be removed before the
calculation.</li>
</ul>
<p><strong>normalize</strong></p>
<ul>
<li>Use the number of genes from the gene sets in each cell to normalize
the enrichment scores. The default value is <strong>FALSE</strong>.</li>
</ul>
<p><strong>make.positive</strong></p>
<ul>
<li>During normalization, whether to shift the enrichment values to a
positive range (<strong>TRUE</strong>) or not (<strong>FALSE</strong>).
The default value is <strong>FALSE</strong>.</li>
</ul>
<p><em>Cautionary note:</em> <strong>make.positive</strong> was added to
allow for differential analysis downstream of enrichment as some methods
may produce negative values. It preserves log-fold change, but
ultimately modifies the enrichment values and should be used with
caution.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enrichment.scores</span> <span class="op">&lt;-</span> <span class="fu">escape.matrix</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                                   gene.sets <span class="op">=</span> <span class="va">escape.gene.sets</span>, </span>
<span>                                   groups <span class="op">=</span> <span class="fl">1000</span>, </span>
<span>                                   min.size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">enrichment.scores</span><span class="op">)</span>, </span>
<span>      mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">enrichment.scores</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">enrichment.scores</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>Multi-core support is for all methods is available through <a href="https://bioconductor.org/packages/release/bioc/html/BiocParallel.html" class="external-link">BiocParallel</a>.
To add more cores, use the argument <strong>BPPARAM</strong> to
<code>escape.matrix()</code>. Here we will use the
<code>SnowParam()</code> for it’s support across platforms and
explicitly call 2 workers (or cores).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enrichment.scores</span> <span class="op">&lt;-</span> <span class="fu">escape.matrix</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                                   gene.sets <span class="op">=</span> <span class="va">escape.gene.sets</span>, </span>
<span>                                   groups <span class="op">=</span> <span class="fl">1000</span>, </span>
<span>                                   min.size <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                                   BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="runescape">runEscape<a class="anchor" aria-label="anchor" href="#runescape"></a>
</h3>
<p>Alternatively, we can use <code>runEscape()</code> to calculate the
enrichment score and directly attach the output to a single-cell object.
The additional parameter for ``<code>runEscape</code> is
<strong>new.assay.name</strong>, in order to save the enrichment scores
as a custom assay in the single-cell object.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu">runEscape</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                           method <span class="op">=</span> <span class="st">"ssGSEA"</span>,</span>
<span>                           gene.sets <span class="op">=</span> <span class="va">escape.gene.sets</span>, </span>
<span>                           groups <span class="op">=</span> <span class="fl">1000</span>, </span>
<span>                           min.size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                           new.assay.name <span class="op">=</span> <span class="st">"escape.ssGSEA"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu">runEscape</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                           method <span class="op">=</span> <span class="st">"UCell"</span>,</span>
<span>                           gene.sets <span class="op">=</span> <span class="va">escape.gene.sets</span>, </span>
<span>                           groups <span class="op">=</span> <span class="fl">1000</span>, </span>
<span>                           min.size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                           new.assay.name <span class="op">=</span> <span class="st">"escape.UCell"</span><span class="op">)</span></span></code></pre></div>
<p>We can quickly examine the attached enrichment scores using the
visualization/workflow we prefer - here we will use just
<code>FeaturePlot()</code> from the Seurat R package.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Define color palette </span></span>
<span><span class="va">colorblind_vector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.colors</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">7</span>, palette <span class="op">=</span> <span class="st">"inferno"</span>, fixup <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">DefaultAssay</span><span class="op">(</span><span class="va">scRep_example</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"escape.ssGSEA"</span></span>
<span><span class="fu">FeaturePlot</span><span class="op">(</span><span class="va">scRep_example</span>, <span class="st">"Proinflammatory"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="va">colorblind_vector</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="performnormalization">performNormalization<a class="anchor" aria-label="anchor" href="#performnormalization"></a>
</h3>
<p>Although we glossed over the normalization that can be used in
<code>escape.matrix()</code> and <code>runEscape()</code>, it is worth
mentioning here as normalization can affect all downstream analyses.</p>
<p>There can be inherent bias in enrichment values due to drop out in
single-cell expression data. Cells with larger numbers of features and
counts will likely have higher enrichment values.
<code>performNormalization()</code> will normalize the enrichment values
by calculating the number of genes expressed in each gene set and cell.
This is similar to the normalization in classic GSEA and it will be
stored in a new assay.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu">performNormalization</span><span class="op">(</span>input.data <span class="op">=</span> <span class="va">scRep_example</span>, </span>
<span>                                   assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>, </span>
<span>                                   gene.sets <span class="op">=</span> <span class="va">escape.gene.sets</span><span class="op">)</span></span></code></pre></div>
<p>An alternative for scaling by expressed gene sets would be to use a
scaling factor previously calculated during normal single-cell data
processing and quality control. This can be done using the
<strong>scale.factor</strong> argument and providing a vector.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu">performNormalization</span><span class="op">(</span>input.data <span class="op">=</span> <span class="va">scRep_example</span>, </span>
<span>                                   assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>, </span>
<span>                                   gene.sets <span class="op">=</span> <span class="va">escape.gene.sets</span>, </span>
<span>                                   scale.factor <span class="op">=</span> <span class="va">scRep_example</span><span class="op">$</span><span class="va">nFeature_RNA</span><span class="op">)</span></span></code></pre></div>
<p><code>performNormalization()</code> has an additional parameter
<strong>make.positive</strong>. Across the individual gene sets, if
negative normalized enrichment scores are seen, the minimum value is
added to all values. For example if the normalized enrichment scores
(after the above accounting for drop out) ranges from -50 to 50,
<strong>make.positive</strong> will adjust the range to 0 to 100 (by
adding 50). This allows for compatible log2-fold change downstream, but
can alter the enrichment score interpretation.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="visualizations">Visualizations<a class="anchor" aria-label="anchor" href="#visualizations"></a>
</h2>
<p>There are a number of ways to look at the enrichment values
downstream of <code>runEscape()</code> with the myriad plotting and
visualizations functions/packages for single-cell data. <em>escape</em>
include several additional plotting functions to assist in the
analysis.</p>
<div class="section level3">
<h3 id="heatmapenrichment">heatmapEnrichment<a class="anchor" aria-label="anchor" href="#heatmapenrichment"></a>
</h3>
<p>We can examine the enrichment values across our gene sets by using
<code>heatmapEnrichment()</code>. This visualization will return the
mean of the <strong>group.by</strong> variable. As a default - all
visualizations of single-cell objects will use the cluster assignment or
active identity as a default for visualizations.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">heatmapEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                  group.by <span class="op">=</span> <span class="st">"ident"</span>,</span>
<span>                  gene.set.use <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>                  assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>Most of the visualizations in <em>escape</em> have a defined set of
parameters.</p>
<p><strong>group.by</strong></p>
<ul>
<li>The grouping variable for the comparison.</li>
</ul>
<p><strong>facet.by</strong></p>
<ul>
<li>Using a variable to facet the graph into separate
visualizations.</li>
</ul>
<p><strong>scale</strong></p>
<ul>
<li>
<strong>TRUE</strong> - z-transform the enrichment values.</li>
<li>
<strong>FALSE</strong> - leave raw values
(<strong>DEFAULT</strong>).</li>
</ul>
<p>In addition, <code>heatmapEnrichment()</code> allows for the
reclustering of rows and columns using Euclidean distance of the
enrichment scores and the Ward2 methods for clustering using
<strong>cluster.rows</strong> and <strong>cluster.columns</strong>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">heatmapEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                  group.by <span class="op">=</span> <span class="st">"ident"</span>,</span>
<span>                  assay <span class="op">=</span> <span class="st">"escape.UCell"</span>,</span>
<span>                  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                  cluster.rows <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                  cluster.columns <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>Each visualization has an additional argument called **palette that
supplies the coloring scheme to be used - available color palettes can
be viewed with <code><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.pals()</a></code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.pals</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   [1] "Pastel 1"      "Dark 2"        "Dark 3"        "Set 2"        </span></span>
<span><span class="co">##   [5] "Set 3"         "Warm"          "Cold"          "Harmonic"     </span></span>
<span><span class="co">##   [9] "Dynamic"       "Grays"         "Light Grays"   "Blues 2"      </span></span>
<span><span class="co">##  [13] "Blues 3"       "Purples 2"     "Purples 3"     "Reds 2"       </span></span>
<span><span class="co">##  [17] "Reds 3"        "Greens 2"      "Greens 3"      "Oslo"         </span></span>
<span><span class="co">##  [21] "Purple-Blue"   "Red-Purple"    "Red-Blue"      "Purple-Orange"</span></span>
<span><span class="co">##  [25] "Purple-Yellow" "Blue-Yellow"   "Green-Yellow"  "Red-Yellow"   </span></span>
<span><span class="co">##  [29] "Heat"          "Heat 2"        "Terrain"       "Terrain 2"    </span></span>
<span><span class="co">##  [33] "Viridis"       "Plasma"        "Inferno"       "Rocket"       </span></span>
<span><span class="co">##  [37] "Mako"          "Dark Mint"     "Mint"          "BluGrn"       </span></span>
<span><span class="co">##  [41] "Teal"          "TealGrn"       "Emrld"         "BluYl"        </span></span>
<span><span class="co">##  [45] "ag_GrnYl"      "Peach"         "PinkYl"        "Burg"         </span></span>
<span><span class="co">##  [49] "BurgYl"        "RedOr"         "OrYel"         "Purp"         </span></span>
<span><span class="co">##  [53] "PurpOr"        "Sunset"        "Magenta"       "SunsetDark"   </span></span>
<span><span class="co">##  [57] "ag_Sunset"     "BrwnYl"        "YlOrRd"        "YlOrBr"       </span></span>
<span><span class="co">##  [61] "OrRd"          "Oranges"       "YlGn"          "YlGnBu"       </span></span>
<span><span class="co">##  [65] "Reds"          "RdPu"          "PuRd"          "Purples"      </span></span>
<span><span class="co">##  [69] "PuBuGn"        "PuBu"          "Greens"        "BuGn"         </span></span>
<span><span class="co">##  [73] "GnBu"          "BuPu"          "Blues"         "Lajolla"      </span></span>
<span><span class="co">##  [77] "Turku"         "Hawaii"        "Batlow"        "Blue-Red"     </span></span>
<span><span class="co">##  [81] "Blue-Red 2"    "Blue-Red 3"    "Red-Green"     "Purple-Green" </span></span>
<span><span class="co">##  [85] "Purple-Brown"  "Green-Brown"   "Blue-Yellow 2" "Blue-Yellow 3"</span></span>
<span><span class="co">##  [89] "Green-Orange"  "Cyan-Magenta"  "Tropic"        "Broc"         </span></span>
<span><span class="co">##  [93] "Cork"          "Vik"           "Berlin"        "Lisbon"       </span></span>
<span><span class="co">##  [97] "Tofino"        "ArmyRose"      "Earth"         "Fall"         </span></span>
<span><span class="co">## [101] "Geyser"        "TealRose"      "Temps"         "PuOr"         </span></span>
<span><span class="co">## [105] "RdBu"          "RdGy"          "PiYG"          "PRGn"         </span></span>
<span><span class="co">## [109] "BrBG"          "RdYlBu"        "RdYlGn"        "Spectral"     </span></span>
<span><span class="co">## [113] "Zissou 1"      "Cividis"       "Roma"</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">heatmapEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                  assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>,</span>
<span>                  palette <span class="op">=</span> <span class="st">"Spectral"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<p>Alternatively, we can add an additional layer to the ggplot object
that is returned by the visualizations using something like
<code><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradientn()</a></code> for continuous values or
<code><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual()</a></code> for the categorical variables.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">heatmapEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                  group.by <span class="op">=</span> <span class="st">"ident"</span>,</span>
<span>                  assay <span class="op">=</span> <span class="st">"escape.UCell"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">11</span>, <span class="st">"RdYlBu"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="geyserenrichment">geyserEnrichment<a class="anchor" aria-label="anchor" href="#geyserenrichment"></a>
</h3>
<p>We can also focus on individual gene sets - one approach is to use
<code>geyserEnrichment()</code>. Here individual cells are plotted along
the Y-axis with graphical summary where the central dot refers to the
median enrichment value and the thicker/thinner lines demonstrate the
interval summaries referring to the 66% and 95%.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">geyserEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                 assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>,</span>
<span>                 gene.set <span class="op">=</span> <span class="st">"T1-Interferon"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<p>To show the additional parameters that appear in visualizations of
individual enrichment gene sets - we can reorder the groups by the mean
of the gene set using <strong>order.by</strong> = “mean”.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">geyserEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                 assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>,</span>
<span>                 gene.set <span class="op">=</span> <span class="st">"T1-Interferon"</span>, </span>
<span>                 order.by <span class="op">=</span> <span class="st">"mean"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<p>What if we had 2 separate samples or groups within the data? Another
parameter we can use is <strong>facet.by</strong> to allow for direct
visualization of an additional variable.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">geyserEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                 assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>,</span>
<span>                 gene.set <span class="op">=</span> <span class="st">"T1-Interferon"</span>, </span>
<span>                 facet.by <span class="op">=</span> <span class="st">"Type"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<p>Lastly, we can select the way the color is applied to the plot using
the <strong>color.by</strong> parameter. Here we can set it to the gene
set of interest <em>“HALLMARK-INTERFERON-GAMMA-RESPONSE”</em>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">geyserEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                 assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>,</span>
<span>                 gene.set <span class="op">=</span> <span class="st">"T1-Interferon"</span>, </span>
<span>                 color.by  <span class="op">=</span> <span class="st">"T1-Interferon"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="ridgeenrichment">ridgeEnrichment<a class="anchor" aria-label="anchor" href="#ridgeenrichment"></a>
</h3>
<p>Similar to the <code>geyserEnrichment()</code> the
<code>ridgeEnrichment()</code> can display the distribution of
enrichment values across the selected gene set. The central line is at
the median value for the respective grouping.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ridgeEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                assay <span class="op">=</span> <span class="st">"escape.UCell"</span>,</span>
<span>                gene.set <span class="op">=</span> <span class="st">"T2-Interferon"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
<p>We can get the relative position of individual cells along the x-axis
using the <strong>add.rug</strong> parameter.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ridgeEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                assay <span class="op">=</span> <span class="st">"escape.UCell"</span>,</span>
<span>                gene.set <span class="op">=</span> <span class="st">"T2-Interferon"</span>,</span>
<span>                add.rug <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="splitenrichment">splitEnrichment<a class="anchor" aria-label="anchor" href="#splitenrichment"></a>
</h3>
<p>Another distribution visualization is a violin plot, which we
separate and directly compare using a binary classification. Like
<code>ridgeEnrichment()</code>, this allows for greater use of
categorical variables. For <code>splitEnrichment()</code>, the output
will be two halves of a violin plot based on the
<strong>split.by</strong> parameter with a central boxplot with the
relative distribution across all samples.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">splitEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>,</span>
<span>                gene.set <span class="op">=</span> <span class="st">"Lipid-mediators"</span>, </span>
<span>                split.by <span class="op">=</span> <span class="st">"Type"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<p>If selecting a <strong>split.by</strong> variable with more than 2
levels, <code>splitEnrichment()</code> will convert the violin plots to
dodge.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">splitEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>,</span>
<span>                gene.set <span class="op">=</span> <span class="st">"Lipid-mediators"</span>, </span>
<span>                split.by <span class="op">=</span> <span class="st">"ident"</span>, </span>
<span>                group.by <span class="op">=</span> <span class="st">"Patient"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="gseaenrichment">gseaEnrichment<a class="anchor" aria-label="anchor" href="#gseaenrichment"></a>
</h3>
<p><code>gseaEnrichment()</code> reproduces the two-panel GSEA graphic
from Subramanian et al. (2005): * Panel A – the running enrichment score
(RES) as you “walk” down the ranked list. * Panel B – a rug showing
exact positions of each pathway gene.</p>
<p>It works on escape’s per-cell ranks, but collapses them across cells
with a summary statistic (summary.fun = “median” by default).</p>
<p><strong>How it works:</strong></p>
<ol style="list-style-type: decimal">
<li>Rank all genes in each group by summary.fun of
expression/statistic.</li>
<li>Perform the weighted Kolmogorov–Smirnov walk: +w when the next gene
is in the set, −1/(N − NG) otherwise.</li>
<li>ES = maximum signed deviation; permutation on gene labels (or
phenotypes) to derive NES and p.</li>
</ol>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">gseaEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>,</span>
<span>               gene.set.use <span class="op">=</span> <span class="st">"T2_Interferon"</span>,</span>
<span>               gene.sets    <span class="op">=</span> <span class="va">escape.gene.sets</span>,</span>
<span>               group.by     <span class="op">=</span> <span class="st">"ident"</span>,</span>
<span>               summary.fun  <span class="op">=</span> <span class="st">"mean"</span>,</span>
<span>               nperm        <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="densityenrichment">densityEnrichment<a class="anchor" aria-label="anchor" href="#densityenrichment"></a>
</h3>
<p><code>densityEnrichment()</code> is a method to visualize the mean
rank position of the gene set features along the total feature space by
group. Instead of the classic GSEA running-score, it overlays
<strong>kernel-density traces</strong> of the <em>gene ranks</em> (1 =
most highly expressed/ranked gene) for every group or cluster. High
densities at the <em>left-hand</em> side mean the pathway is
collectively <strong>up-regulated</strong>; peaks on the <em>right</em>
imply down-regulation.</p>
<p><strong>Anatomy of the plot</strong></p>
<ol style="list-style-type: decimal">
<li>
<strong>X-axis</strong> – gene rank (1 … <em>N</em>). Left =
top-ranked genes.<br>
</li>
<li>
<strong>Y-axis</strong> – density estimate (area under each curve =
1).<br>
</li>
<li>
<strong>One coloured line per level of
<code>group.by</code></strong> – default is Seurat/SCE cluster.</li>
</ol>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">densityEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                  gene.set.use <span class="op">=</span> <span class="st">"T2_Interferon"</span>, </span>
<span>                  gene.sets <span class="op">=</span> <span class="va">escape.gene.sets</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="scatterenrichment">scatterEnrichment<a class="anchor" aria-label="anchor" href="#scatterenrichment"></a>
</h3>
<p>It may be advantageous to look at the distribution of multiple gene
sets - here we can use <code>scatterEnrichment()</code> for a 2 gene set
comparison. The color values are based on the density of points
determined by the number of neighbors, similar to the <a href="https://www.bioconductor.org/packages/release/bioc/html/Nebulosa.html" class="external-link">Nebulosa
R package</a>. We just need to define which gene set to plot on the
<strong>x.axis</strong> and which to plot on the
<strong>y.axis</strong>.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scatterEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                  assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>,</span>
<span>                  x.axis <span class="op">=</span> <span class="st">"T2-Interferon"</span>,</span>
<span>                  y.axis <span class="op">=</span> <span class="st">"Lipid-mediators"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
<p>The scatter plot can also be converted into a hexbin, another method
for summarizing the individual cell distributions along the x and y
axis, by setting <strong>style</strong> = “hex”.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scatterEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                  assay <span class="op">=</span> <span class="st">"escape.UCell"</span>,</span>
<span>                  x.axis <span class="op">=</span> <span class="st">"T2-Interferon"</span>,</span>
<span>                  y.axis <span class="op">=</span> <span class="st">"Lipid-mediators"</span>,</span>
<span>                  style <span class="op">=</span> <span class="st">"hex"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-30-1.png" width="700"></p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="statistical-analysis">Statistical Analysis<a class="anchor" aria-label="anchor" href="#statistical-analysis"></a>
</h2>
<div class="section level3">
<h3 id="principal-component-analysis-pca">Principal Component Analysis (PCA)<a class="anchor" aria-label="anchor" href="#principal-component-analysis-pca"></a>
</h3>
<p>escape has its own PCA function <code>performPCA()</code> which will
work on a single-cell object or a matrix of enrichment values. This is
specifically useful for downstream visualizations as it stores the
eigenvalues and rotations. If we want to look at the relative
contribution to overall variance of each component or a Biplot-like
overlay of the individual features, use <code>performPCA()</code>.</p>
<p>Alternatively, other PCA-based functions like Seurat’s
<code>RunPCA()</code> or scater’s ``<code>runPCA()</code> can be used.
These functions are likely faster and would be ideal if we have a larger
number of cells and/or gene sets.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu">performPCA</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                         assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>,</span>
<span>                         n.dim <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><em>escape</em> has a built in method for plotting PCA
<code>pcaEnrichment()</code> that functions similarly to the
<code>scatterEnrichment()</code> function where <strong>x.axis</strong>
and <strong>y.axis</strong> are the components to plot.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pcaEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>              dimRed <span class="op">=</span> <span class="st">"escape.PCA"</span>,</span>
<span>              x.axis <span class="op">=</span> <span class="st">"PC1"</span>,</span>
<span>              y.axis <span class="op">=</span> <span class="st">"PC2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-32-1.png" width="700"></p>
<p><code>pcaEnrichment()</code> can plot additional information on the
principal component analysis.</p>
<p><strong>add.percent.contribution</strong> will add the relative
percent contribution of the x and y.axis to total variability observed
in the PCA.</p>
<p><strong>display.factors</strong> will overlay the magnitude and
direction that the features/gene sets contribute to the selected
components. The number of gene sets is determined by
<strong>number.of.factors</strong>. This can assist in understanding the
underlying differences in enrichment across different cells.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pcaEnrichment</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>              dimRed <span class="op">=</span> <span class="st">"escape.PCA"</span>,</span>
<span>              x.axis <span class="op">=</span> <span class="st">"PC1"</span>,</span>
<span>              y.axis <span class="op">=</span> <span class="st">"PC2"</span>,</span>
<span>              add.percent.contribution <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>              display.factors <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>              number.of.factors <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-33-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="precomputed-rank-lists">Precomputed Rank Lists<a class="anchor" aria-label="anchor" href="#precomputed-rank-lists"></a>
</h3>
<p>Functional enrichment is not limited to per-cell scores. Many
workflows start with <strong>differential-expression (DE)
statistics</strong> (e.g. Seurat’s <code>FindMarkers()</code>, DESeq2’s
<code>results()</code>, edgeR’s <code>topTags()</code>). Those produce a
<em>ranked gene list</em> that can be fed into a classical
<strong>Gene-Set Enrichment Analysis (GSEA)</strong>.</p>
<div class="section level4">
<h4 id="why-do-this">Why do this?<a class="anchor" aria-label="anchor" href="#why-do-this"></a>
</h4>
<ul>
<li>
<strong>Aggregates signal across genes</strong>: a borderline but
<em>consistent</em> trend across 30 pathway genes is often more
informative than a single high-logFC gene.<br>
</li>
<li>
<strong>Directionality</strong>: by combining log-fold-change
(<em>effect size</em>) and an adjusted <em>p</em>-value
(<em>confidence</em>)</li>
<li>
<strong>Speed</strong>: you avoid re-scoring every cell; only one
numeric vector is needed.</li>
</ul>
<p><code>enrichIt()</code> accepts either</p>
<ol style="list-style-type: decimal">
<li>a <strong>named numeric vector</strong> (<em>already ranked</em>),
or<br>
</li>
<li>a <strong>data frame</strong> containing logFC + <em>p</em> (or
<em>adj.p</em>).</li>
</ol>
<p>The helper <strong>automatically chooses</strong> the best
<em>p</em>-value column in this order:</p>
<ol style="list-style-type: decimal">
<li>
<code>p_val_adj</code><br>
</li>
<li>
<code>padj</code> (DESeq2)<br>
</li>
<li>
<code>FDR</code> (edgeR)<br>
</li>
<li>plain <code>p_val</code>
</li>
</ol>
</div>
<div class="section level4">
<h4 id="example-enrichit-workflow">Example <code>enrichIt()</code> workflow<a class="anchor" aria-label="anchor" href="#example-enrichit-workflow"></a>
</h4>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DefaultAssay</span><span class="op">(</span><span class="va">scRep_example</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"RNA"</span></span>
<span><span class="va">DEG.markers</span> <span class="op">&lt;-</span> <span class="fu">FindMarkers</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                           ident.1 <span class="op">=</span> <span class="st">"1"</span>, </span>
<span>                           ident.2 <span class="op">=</span> <span class="st">"2"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">GSEA.results</span> <span class="op">&lt;-</span> <span class="fu">enrichIt</span><span class="op">(</span>input.data <span class="op">=</span> <span class="va">DEG.markers</span>, </span>
<span>                         gene.sets <span class="op">=</span> <span class="va">escape.gene.sets</span>, </span>
<span>                         ranking_fun <span class="op">=</span> <span class="st">"signed_log10_p"</span><span class="op">)</span>               </span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">GSEA.results</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                            pathway        pval       padj   log2err        ES</span></span>
<span><span class="co">##                             &lt;char&gt;       &lt;num&gt;      &lt;num&gt;     &lt;num&gt;     &lt;num&gt;</span></span>
<span><span class="co">## 1:                             G1S 0.004459379 0.08918759 0.4070179 0.9495747</span></span>
<span><span class="co">## 2:                   T1_Interferon 0.045590848 0.43715847 0.3217759 0.9916719</span></span>
<span><span class="co">## 3: T_Cell_Terminal_Differentiation 0.088135593 0.43715847 0.2820134 0.9905942</span></span>
<span><span class="co">## 4:             Glucose_Deprivation 0.108474576 0.43715847 0.2529611 0.9869766</span></span>
<span><span class="co">## 5:                       Cytolytic 0.121951220 0.43715847 0.2249661 0.9827274</span></span>
<span><span class="co">## 6:                 Antinflammatory 0.131147541 0.43715847 0.2529611 0.9883311</span></span>
<span><span class="co">##         NES  size                                        leadingEdge  geneRatio</span></span>
<span><span class="co">##       &lt;num&gt; &lt;int&gt;                                             &lt;char&gt;      &lt;num&gt;</span></span>
<span><span class="co">## 1: 1.552116    63 HIST1H4C;CDKN1A;HIST1H1D;CCNG2;H2AFZ;HIST1H1E;H1FX 0.11111111</span></span>
<span><span class="co">## 2: 1.541180    17                                               OASL 0.05882353</span></span>
<span><span class="co">## 3: 1.933083     7                                        PDCD1;CTLA4 0.28571429</span></span>
<span><span class="co">## 4: 1.926023     7                              FAM129A;HSPA1A;HSPA1B 0.42857143</span></span>
<span><span class="co">## 5: 1.996093     6                                     GZMB;GZMA;GZMH 0.50000000</span></span>
<span><span class="co">## 6: 1.824227     9                    LGALS3;PDCD1;FOXP3;CTLA4;ENTPD1 0.55555556</span></span></code></pre>
<p>What does the result look like?</p>
<ul>
<li>
<strong>ES / NES</strong> – raw and normalised enrichment scores
from fgsea</li>
<li>
<strong>pval / padj</strong> – nominal and
multiple-testing-corrected p</li>
<li>
<strong>size</strong> – total number of genes in the set</li>
<li>
<strong>geneRatio</strong> – (core hits)/(size), useful for dot
plots</li>
<li>
<strong>leadingEdge</strong> – semi-colon-separated genes driving
the signal</li>
</ul>
</div>
<div class="section level4">
<h4 id="visualizing-the-enrichment-table">Visualizing the enrichment table<a class="anchor" aria-label="anchor" href="#visualizing-the-enrichment-table"></a>
</h4>
<p>The companion <code>enrichItPlot()</code> gives three quick chart
types.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## (1) Bar plot –20 most significant per database</span></span>
<span><span class="fu">enrichItPlot</span><span class="op">(</span><span class="va">GSEA.results</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>strip.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        strip.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>              </span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-35-1.png" width="700"></p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## (2) Dot plot – colored by –log10 padj, sized by core-hits</span></span>
<span><span class="fu">enrichItPlot</span><span class="op">(</span><span class="va">GSEA.results</span>, <span class="st">"dot"</span>, top <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>strip.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        strip.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>   </span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-35-2.png" width="700"></p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## (3) C-net plot – network of pathways ↔ leading-edge genes</span></span>
<span><span class="fu">enrichItPlot</span><span class="op">(</span><span class="va">GSEA.results</span>, <span class="st">"cnet"</span>, top <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> </span></code></pre></div>
<p><img src="Running_Escape_files/figure-html/unnamed-chunk-35-3.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="differential-enrichment">Differential Enrichment<a class="anchor" aria-label="anchor" href="#differential-enrichment"></a>
</h3>
<p>Differential enrichment analysis can be performed similar to
differential gene expression analysis. For the purposes of finding the
differential enrichment values, we can first normalize the enrichment
values for the ssGSEA calculations. Notice here, we are using
<strong>make.positive</strong> = TRUE in order to adjust any negative
values. This is a particular issue when it comes to ssGSEA and GSVA
enrichment scores.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu">performNormalization</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                                   assay <span class="op">=</span> <span class="st">"escape.ssGSEA"</span>, </span>
<span>                                   gene.sets <span class="op">=</span> <span class="va">escape.gene.sets</span>,</span>
<span>                                   make.positive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">all.markers</span> <span class="op">&lt;-</span> <span class="fu">FindAllMarkers</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                              assay <span class="op">=</span> <span class="st">"escape.ssGSEA_normalized"</span>, </span>
<span>                              min.pct <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                              logfc.threshold <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">all.markers</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                        p_val  avg_log2FC pct.1 pct.2    p_val_adj cluster</span></span>
<span><span class="co">## Cytolytic       8.835726e-33 -0.22290562     1     1 1.767145e-31       1</span></span>
<span><span class="co">## TCA-cycle       3.828829e-15 -0.18921087     1     1 7.657658e-14       1</span></span>
<span><span class="co">## Antinflammatory 7.518488e-11 -0.09552935     1     1 1.503698e-09       1</span></span>
<span><span class="co">## Hypoxia         8.299977e-10  0.02923536     1     1 1.659995e-08       1</span></span>
<span><span class="co">## G1S             2.501304e-09 -0.31214094     1     1 5.002607e-08       1</span></span>
<span><span class="co">## PPS             3.565329e-09 -0.10738891     1     1 7.130657e-08       1</span></span>
<span><span class="co">##                            gene</span></span>
<span><span class="co">## Cytolytic             Cytolytic</span></span>
<span><span class="co">## TCA-cycle             TCA-cycle</span></span>
<span><span class="co">## Antinflammatory Antinflammatory</span></span>
<span><span class="co">## Hypoxia                 Hypoxia</span></span>
<span><span class="co">## G1S                         G1S</span></span>
<span><span class="co">## PPS                         PPS</span></span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
