<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Visualizations for Single-Cell Objects • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Visualizations for Single-Cell Objects">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.6.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Repertoire Analysis</li>
    <li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Working with Single-Cell Objects</li>
    <li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Deep Learning Integrations</li>
    <li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other Workflow Integrations</li>
    <li>
      <a href="../articles/immcantation.html">Combining Immcantation and scRepertoire</a>
    </li>
    <li>
      <a href="../articles/tcrpheno.html">TCRpheno applied to T cell fate</a>
    </li>
    <li>
      <a href="../articles/immunarch.html">Using scRepertoire with immunarch</a>
    </li>
    <li>
      <a href="../articles/dandelionR.html">VDJ Trajectory Analysis with dandelionR</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BorchLab/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Visualizations for Single-Cell Objects</h1>
            
            <h4 data-toc-skip class="date">Compiled: October 31,
2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BorchLab/scRepertoire/blob/HEAD/vignettes/articles/SC_Visualizations.Rmd" class="external-link"><code>vignettes/articles/SC_Visualizations.Rmd</code></a></small>
      <div class="hidden name"><code>SC_Visualizations.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="clonaloverlay">clonalOverlay<a class="anchor" aria-label="anchor" href="#clonaloverlay"></a>
</h2>
<p>Using the dimensional reduction graphs as a reference,
<code><a href="../reference/clonalOverlay.html">clonalOverlay()</a></code> generates an overlay of the positions of
clonally-expanded cells. It highlights areas of high clonal frequency or
proportion on your UMAP or tSNE plots.</p>
<p>Key Parameters for clonalOverlay()</p>
<ul>
<li>
<code>sc.data</code>: The single-cell object
after<code><a href="../reference/combineExpression.html">combineExpression()</a></code>.</li>
<li>
<code>reduction</code>: The dimensional reduction to visualize
(e.g., “umap”, “pca”). Default is “pca”.</li>
<li>
<code>cut.category</code>: The metadata variable to use for
filtering the overlay (e.g., “clonalFrequency” or
“clonalProportion”).</li>
<li>
<code>cutpoint</code>: The lowest clonal frequency or proportion to
include in the contour plot.</li>
<li>
<code>bins</code>: The number of contours to draw.</li>
</ul>
<p><code><a href="../reference/clonalOverlay.html">clonalOverlay()</a></code> can be used to look across all cells or
faceted by a metadata variable using <code>facet.by</code>. The overall
dimensional reduction will be maintained as we facet, while the contour
plots will adjust based on the <code>facet.by</code> variable. The
coloring of the dot plot is based on the active identity of the
single-cell object.</p>
<p>This visualization was authored by Dr. Francesco Mazziotta and
inspired by Drs. Carmona and Andreatta and their work with <a href="https://github.com/carmonalab/ProjecTILs" class="external-link">ProjectTIL</a>, a
pipeline for annotating T cell subtypes.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalOverlay.html">clonalOverlay</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>              reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>              cutpoint <span class="op">=</span> <span class="fl">1</span>, </span>
<span>              bins <span class="op">=</span> <span class="fl">10</span>, </span>
<span>              facet.by <span class="op">=</span> <span class="st">"Patient"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>              <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SC_Visualizations_files/figure-html/unnamed-chunk-1-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="clonalnetwork">clonalNetwork<a class="anchor" aria-label="anchor" href="#clonalnetwork"></a>
</h2>
<p>Similar to <code><a href="../reference/clonalOverlay.html">clonalOverlay()</a></code>, <code><a href="../reference/clonalNetwork.html">clonalNetwork()</a></code>
visualizes the network interaction of clones shared between clusters
along the single-cell dimensional reduction. This function shows the
relative proportion of clones flowing from a starting node, with the
ending node indicated by an arrow</p>
<div class="section level3">
<h3 id="filtering-options-for-clonalnetwork">Filtering Options for clonalNetwork()<a class="anchor" aria-label="anchor" href="#filtering-options-for-clonalnetwork"></a>
</h3>
<ul>
<li>
<code>filter.clones</code>:
<ul>
<li>Select a number to isolate the clones comprising the top n number of
cells (e.g., <code>filter.clones = 2000</code>).</li>
<li>Select <code>min</code> to scale all groups to the size of the
minimum group.</li>
</ul>
</li>
<li>
<code>filter.identity</code>: For the identity chosen to visualize,
show the “to” and “from” network connections for a single group.</li>
<li>
<code>filter.proportion</code>: Remove clones from the network that
comprise less than a certain proportion of clones in groups.</li>
<li>
<code>filter.graph</code>: Remove reciprocal edges from one half of
the graph, allowing for cleaner visualization.</li>
</ul>
<p>Now, visualize the clonal network with no specific identity
filter:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#ggraph needs to be loaded due to issues with ggplot</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggraph.data-imaginist.com" class="external-link">ggraph</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">#No Identity filter</span></span>
<span><span class="fu"><a href="../reference/clonalNetwork.html">clonalNetwork</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>              reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>              group.by <span class="op">=</span> <span class="st">"seurat_clusters"</span>,</span>
<span>              filter.clones <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>              filter.identity <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>              cloneCall <span class="op">=</span> <span class="st">"aa"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SC_Visualizations_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<p>We can look at the clonal relationships relative to a single cluster
using the <code>filter.identity</code> parameter. For example, focusing
on Cluster 3:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Examining Cluster 3 only</span></span>
<span><span class="fu"><a href="../reference/clonalNetwork.html">clonalNetwork</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>              reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>              group.by <span class="op">=</span> <span class="st">"seurat_clusters"</span>,</span>
<span>              filter.identity <span class="op">=</span> <span class="fl">3</span>,</span>
<span>              cloneCall <span class="op">=</span> <span class="st">"aa"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SC_Visualizations_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>You can also use the <code>exportClones</code> parameter to quickly
get clones that are shared across multiple identity groups, along with
the total number of clone copies in the dataset.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">shared.clones</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clonalNetwork.html">clonalNetwork</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                               reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>                               group.by <span class="op">=</span> <span class="st">"seurat_clusters"</span>,</span>
<span>                               cloneCall <span class="op">=</span> <span class="st">"aa"</span>, </span>
<span>                               exportClones <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">shared.clones</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 2</span></span></span>
<span><span class="co">##   clone                                sum</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>                              <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> CVVSDNTGGFKTIF_CASSVRRERANTGELFF     906</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> CAERGSGGSYIPTF_CASSDPSGRQGPRWDTQYF   140</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> CAVTFHYNTDKLIF_CASSQDRTGLDYEQYF      122</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> CAVRDDGNTGFQKLVF_CASSQDFNDGGLNIQYF   119</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> CARKVRDSSYKLIF_CASSDSGYNEQFF         106</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> CAVGAQQGGKLIF_CASSLSLSDGRHGYTF       101</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="highlightclones">highlightClones<a class="anchor" aria-label="anchor" href="#highlightclones"></a>
</h2>
<p>The <code><a href="../reference/highlightClones.html">highlightClones()</a></code> function allows you to
specifically visualize the distribution of particular clonal sequences
on your single-cell dimensional reduction plots. This helps in tracking
the location and expansion of clones of interest.</p>
<p>Key Parameters for <code><a href="../reference/highlightClones.html">highlightClones()</a></code></p>
<ul>
<li>
<code>cloneCall</code>: The type of sequence to use for highlighting
(e.g., “aa”, “nt”, “strict”).</li>
<li>
<code>sequence</code>: A character vector of the specific clonal
sequences to highlight.</li>
</ul>
<p>To highlight the most prominent amino acid sequences:
<em>CAERGSGGSYIPTF_CASSDPSGRQGPRWDTQYF</em> and
<em>CARKVRDSSYKLIF_CASSDSGYNEQFF</em>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/highlightClones.html">highlightClones</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                    cloneCall<span class="op">=</span> <span class="st">"aa"</span>, </span>
<span>                    sequence <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CAERGSGGSYIPTF_CASSDPSGRQGPRWDTQYF"</span>, </span>
<span>                                 <span class="st">"CARKVRDSSYKLIF_CASSDSGYNEQFF"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">scRep_example</span>, group.by <span class="op">=</span> <span class="st">"highlight"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>nrow<span class="op">=</span><span class="fl">3</span>,byrow<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, </span>
<span>                 legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SC_Visualizations_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="clonaloccupy">clonalOccupy<a class="anchor" aria-label="anchor" href="#clonaloccupy"></a>
</h2>
<p><code><a href="../reference/clonalOccupy.html">clonalOccupy()</a></code> visualizes the count of cells by cluster,
categorized into specific clonal frequency ranges. It uses the cloneSize
metadata variable (generated by <code><a href="../reference/combineExpression.html">combineExpression()</a></code>) to
plot the number of cells within each clonal expansion designation, using
a secondary variable like cluster. Credit for the idea goes to Drs.
Carmona and Andreatta.</p>
<p>Key Parameters for <code><a href="../reference/clonalOccupy.html">clonalOccupy()</a></code> * <code>x.axis</code>:
The variable in the metadata to graph along the x-axis (e.g.,
“seurat_clusters”, “ident”). * <code>label</code>: If TRUE, includes the
number of clones in each category by x.axis variable. *
<code>proportion</code>: If TRUE, converts the stacked bars into
relative proportions. * <code>na.include</code>: If TRUE, visualizes NA
values.</p>
<p>To visualize the count of cells by <code>seurat_clusters</code> based
on <code>cloneSize</code> groupings:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalOccupy.html">clonalOccupy</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>              x.axis <span class="op">=</span> <span class="st">"seurat_clusters"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SC_Visualizations_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>To visualize the proportion of cells by <code>ident</code> (active
identity), without labels:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalOccupy.html">clonalOccupy</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>             x.axis <span class="op">=</span> <span class="st">"ident"</span>, </span>
<span>             proportion <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>             label <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="SC_Visualizations_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="alluvialclones">alluvialClones<a class="anchor" aria-label="anchor" href="#alluvialclones"></a>
</h2>
<p>After the metadata has been modified with clonal information,
<code><a href="../reference/alluvialClones.html">alluvialClones()</a></code> allows you to look at clones across
multiple categorical variables, enabling the examination of the
interchange between these variables. Because this function produces a
graph with each clone arranged by called stratification, it may take
some time depending on the size of the repertoire.</p>
<p>Key Parameters for <code><a href="../reference/alluvialClones.html">alluvialClones()</a></code></p>
<ul>
<li>
<code>y.axes</code>: The columns that will separate the proportional
visualizations.</li>
<li>
<code>color</code>: The column header or clone(s) to be
highlighted.</li>
<li>
<code>facet</code>: The column label to separate facets.</li>
<li>
<code>alpha</code>: The column header to have gradated opacity.</li>
</ul>
<p>To visualize clonal flow across “Patient”, “ident”, and “Type”,
highlighting specific amino acid clones:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/alluvialClones.html">alluvialClones</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>               cloneCall <span class="op">=</span> <span class="st">"aa"</span>, </span>
<span>               y.axes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Patient"</span>, <span class="st">"ident"</span>, <span class="st">"Type"</span><span class="op">)</span>, </span>
<span>               color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CVVSDNTGGFKTIF_CASSVRRERANTGELFF"</span>, <span class="st">"NA_CASSVRRERANTGELFF"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grey"</span>, <span class="va">colorblind_vector</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SC_Visualizations_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>To visualize clonal flow across “Patient”, “ident”, and “Type”,
coloring by “ident”:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/alluvialClones.html">alluvialClones</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                   cloneCall <span class="op">=</span> <span class="st">"gene"</span>, </span>
<span>                   y.axes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Patient"</span>, <span class="st">"ident"</span>, <span class="st">"Type"</span><span class="op">)</span>, </span>
<span>                   color <span class="op">=</span> <span class="st">"ident"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="SC_Visualizations_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p><code><a href="../reference/alluvialClones.html">alluvialClones()</a></code> provides a visual representation of
clonal distribution and movement across multiple categorical
annotations. It is particularly effective for tracking how specific
clones or clonal groups transition between different states, tissues, or
cell types, offering a dynamic perspective on immune repertoire
evolution and function.</p>
</div>
<div class="section level2">
<h2 id="getcirclize">getCirclize<a class="anchor" aria-label="anchor" href="#getcirclize"></a>
</h2>
<p>Like alluvial graphs, we can also visualize the interconnection of
clusters using chord diagrams from the <a href="https://jokergoo.github.io/circlize_book/book/" class="external-link">circlize R
package</a>. The first step is to get the data frame output to feed into
the<code><a href="https://rdrr.io/pkg/circlize/man/chordDiagram.html" class="external-link">chordDiagram()</a></code> function in <code>circlize</code>, which
can be done using <code><a href="../reference/getCirclize.html">getCirclize()</a></code>.</p>
<p>This function calculates the relative number of unique and shared
clones based on the <code>group.by</code> variable using the product of
<code><a href="../reference/combineExpression.html">combineExpression()</a></code>. It creates a matrix the size of the
<code>group.by</code> variable and then simplifies it into instructions
readable by the <code>circlize</code> R package. The output represents
the total number of unique and shared clones by the group.by variable.
If using the downstream <code>circlize</code> R package, please read and
cite the following <a href="https://pubmed.ncbi.nlm.nih.gov/24930139/" class="external-link">manuscript</a>.</p>
<p>Key Parameters for getCirclize()</p>
<ul>
<li>
<code>proportion</code>: If <code>TRUE</code>, normalizes the
relationship by proportion; if <code>FALSE</code> (default), uses unique
clone counts.</li>
<li>
<code>include.self</code>: If <code>TRUE</code>, includes counting
clones within a single group.by comparison.</li>
</ul>
<p>To get data for a chord diagram showing shared clones between
<code>seurat_clusters</code>:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/circlize" class="external-link">circlize</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org" class="external-link">scales</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">circles</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCirclize.html">getCirclize</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                       group.by <span class="op">=</span> <span class="st">"seurat_clusters"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Just assigning the normal colors to each cluster</span></span>
<span><span class="va">grid.cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://scales.r-lib.org/reference/pal_hue.html" class="external-link">hue_pal</a></span><span class="op">(</span><span class="op">)</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">scRep_example</span><span class="op">$</span><span class="va">seurat_clusters</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">grid.cols</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">scRep_example</span><span class="op">$</span><span class="va">seurat_clusters</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Graphing the chord diagram</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/circlize/man/chordDiagram.html" class="external-link">chordDiagram</a></span><span class="op">(</span><span class="va">circles</span>, self.link <span class="op">=</span> <span class="fl">1</span>, grid.col <span class="op">=</span> <span class="va">grid.cols</span><span class="op">)</span></span></code></pre></div>
<p><img src="SC_Visualizations_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>This can also be used if we want to explore just the lung-specific T
cells by subsetting the single-cell object. For the sake of this
vignette, we can also look at setting <code>proportion = TRUE</code> to
get a scaled output.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">scRep_example</span>, <span class="va">Type</span> <span class="op">==</span> <span class="st">"L"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">circles</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCirclize.html">getCirclize</a></span><span class="op">(</span><span class="va">subset</span>, group.by <span class="op">=</span> <span class="st">"ident"</span>, proportion <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">grid.cols</span> <span class="op">&lt;-</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/pal_hue.html" class="external-link">hue_pal</a></span><span class="op">(</span><span class="op">)</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">subset</span><span class="op">@</span><span class="va">active.ident</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">grid.cols</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">subset</span><span class="op">@</span><span class="va">active.ident</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/circlize/man/chordDiagram.html" class="external-link">chordDiagram</a></span><span class="op">(</span><span class="va">circles</span>, </span>
<span>             self.link <span class="op">=</span> <span class="fl">1</span>, </span>
<span>             grid.col <span class="op">=</span> <span class="va">grid.cols</span>, </span>
<span>             directional <span class="op">=</span> <span class="fl">1</span>, </span>
<span>             direction.type <span class="op">=</span>  <span class="st">"arrows"</span>,</span>
<span>             link.arr.type <span class="op">=</span> <span class="st">"big.arrow"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SC_Visualizations_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p><code><a href="../reference/getCirclize.html">getCirclize()</a></code> facilitates the creation of visually
striking and informative chord diagrams to represent shared clonal
relationships between distinct groups within your single-cell data. By
providing a flexible way to quantify and format clonal overlap, it
enables researchers to effectively illustrate complex clonal
connectivity patterns, which are crucial for understanding immune
communication and migration.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
