<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Visualizations for Single-Cell Objects • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Visualizations for Single-Cell Objects">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ncborcherding/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Visualizations for Single-Cell Objects</h1>
            
            <h4 data-toc-skip class="date">Compiled: November 04,
2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ncborcherding/scRepertoire/blob/HEAD/vignettes/articles/SC_Visualizations.Rmd" class="external-link"><code>vignettes/articles/SC_Visualizations.Rmd</code></a></small>
      <div class="hidden name"><code>SC_Visualizations.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="clonaloverlay">clonalOverlay<a class="anchor" aria-label="anchor" href="#clonaloverlay"></a>
</h2>
<p>Using the dimensional reduction graphs as a reference, we can also
generate an overlay of the position of clonally-expanded cells using
<code><a href="../reference/clonalOverlay.html">clonalOverlay()</a></code>.</p>
<p><strong>reduction</strong></p>
<ul>
<li>The dimensional reduction for the visualization, <strong>(default =
“pca”)</strong>
</li>
</ul>
<p><strong>cut.category</strong></p>
<ul>
<li>Use of “clonalFrequency” or “clonalProportion” to filter the
overlay</li>
</ul>
<p><strong>cutpoint</strong></p>
<ul>
<li>lowest clonal frequency or proportion to generate the contour
plot</li>
</ul>
<p><strong>bins</strong></p>
<ul>
<li>the number of contours to draw</li>
</ul>
<p><code><a href="../reference/clonalOverlay.html">clonalOverlay()</a></code> can be used to look across all cells or
faceted by a meta data variable using <strong>facet.by</strong>. The
overall dimensional reduction will be maintained as we facet, while the
contour plots will adjust based on the <strong>facet.by</strong>
variable. The coloring of the dot plot is based on the active identity
of the single-cell object.</p>
<p>This visualization was authored by Dr. Francesco Mazziotta and
inspired by Drs. Carmona and Andreatta and their work with <a href="https://github.com/carmonalab/ProjecTILs" class="external-link">ProjectTIL</a>, which is
a great pipeline to annotated T cell subtypes.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalOverlay.html">clonalOverlay</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>              reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>              cutpoint <span class="op">=</span> <span class="fl">1</span>, </span>
<span>              bins <span class="op">=</span> <span class="fl">10</span>, </span>
<span>              facet.by <span class="op">=</span> <span class="st">"Patient"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>              <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SC_Visualizations_files/figure-html/unnamed-chunk-1-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="clonalnetwork">clonalNetwork<a class="anchor" aria-label="anchor" href="#clonalnetwork"></a>
</h2>
<p>Similar to <code><a href="../reference/clonalOverlay.html">clonalOverlay()</a></code>, we can look at the network
interaction of clones shared between clusters along the single-cell
dimensional reduction using <code><a href="../reference/clonalNetwork.html">clonalNetwork()</a></code>. This function
shows the relative proportion of clones from the starting node, with the
ending node indicated by the arrow.</p>
<p>Filtering for clones can be accomplished using 3 methods:</p>
<p><strong>filter.clones</strong></p>
<ul>
<li>Select a number to isolate the clones comprising the top n number of
cells (e.g., <strong>filter.clones</strong> = 2000)<br>
</li>
<li>Select “min” to make sure all groups are scaled to the size of the
minimum group</li>
</ul>
<p><strong>filter.identity</strong></p>
<ul>
<li>For the identity chosen to visualize, show the to and from network
connections for a single group</li>
</ul>
<p><strong>filter.proportion</strong></p>
<ul>
<li>Remove clones that comprise less than a certain proportion of clones
in groups.</li>
</ul>
<p><strong>filter.graph</strong></p>
<ul>
<li>Remove the reciprocal edges from the half of the graph, allowing for
cleaner visualization.</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#ggraph needs to be loaded due to issues with ggplot</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggraph.data-imaginist.com" class="external-link">ggraph</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">#No Identity filter</span></span>
<span><span class="fu"><a href="../reference/clonalNetwork.html">clonalNetwork</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>              reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>              group.by <span class="op">=</span> <span class="st">"seurat_clusters"</span>,</span>
<span>              filter.clones <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>              filter.identity <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>              cloneCall <span class="op">=</span> <span class="st">"aa"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SC_Visualizations_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<p>We can look at the clonal relationships relative to a single cluster
using the <strong>filter.identity</strong> parameter.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Examining Cluster 3 only</span></span>
<span><span class="fu"><a href="../reference/clonalNetwork.html">clonalNetwork</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>              reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>              group.by <span class="op">=</span> <span class="st">"seurat_clusters"</span>,</span>
<span>              filter.identity <span class="op">=</span> <span class="fl">3</span>,</span>
<span>              cloneCall <span class="op">=</span> <span class="st">"aa"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SC_Visualizations_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>We can also use <strong>exportClones</strong> parameter to quickly
get clones that are shared across multiple identity groups, along with
the total number of clones in the data set.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">shared.clones</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clonalNetwork.html">clonalNetwork</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                               reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>                               group.by <span class="op">=</span> <span class="st">"seurat_clusters"</span>,</span>
<span>                               cloneCall <span class="op">=</span> <span class="st">"aa"</span>, </span>
<span>                               exportClones <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">shared.clones</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 2</span></span></span>
<span><span class="co">##   clone                                sum</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                              <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> CVVSDNTGGFKTIF_CASSVRRERANTGELFF     906</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> CAERGSGGSYIPTF_CASSDPSGRQGPRWDTQYF   140</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> CAVTFHYNTDKLIF_CASSQDRTGLDYEQYF      122</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> CAVRDDGNTGFQKLVF_CASSQDFNDGGLNIQYF   119</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> CARKVRDSSYKLIF_CASSDSGYNEQFF         106</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> CAVGAQQGGKLIF_CASSLSLSDGRHGYTF       101</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="highlightclones">highlightClones<a class="anchor" aria-label="anchor" href="#highlightclones"></a>
</h2>
<p>We can also look at the clones by calling specific sequences in the
<code>highlightclones()</code> below. In order to highlight the clones,
we first need to use the <strong>cloneCall</strong>, the type of
sequence we will be using, and then the specific sequences themselves
using <strong>sequence</strong>. Below, we can see the steps to
highlight the most prominent sequences
<em>CAERGSGGSYIPTF_CASSDPSGRQGPRWDTQYF</em>, and
<em>CARKVRDSSYKLIF_CASSDSGYNEQFF</em>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/highlightClones.html">highlightClones</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                    cloneCall<span class="op">=</span> <span class="st">"aa"</span>, </span>
<span>                    sequence <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CAERGSGGSYIPTF_CASSDPSGRQGPRWDTQYF"</span>, </span>
<span>                                 <span class="st">"CARKVRDSSYKLIF_CASSDSGYNEQFF"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">scRep_example</span>, group.by <span class="op">=</span> <span class="st">"highlight"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SC_Visualizations_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="clonaloccupy">clonalOccupy<a class="anchor" aria-label="anchor" href="#clonaloccupy"></a>
</h2>
<p>We can also look at the count of cells by cluster assigned into
specific frequency ranges by using the <code><a href="../reference/clonalOccupy.html">clonalOccupy()</a></code>
function and selecting the <strong>x.axis</strong> to display cluster or
other variables in the meta data of the single cell object.</p>
<p><strong>proportion</strong></p>
<ul>
<li>can be used to look at relative level groupings</li>
</ul>
<p><strong>label</strong></p>
<ul>
<li>will still return the absolute number of clones</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalOccupy.html">clonalOccupy</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>              x.axis <span class="op">=</span> <span class="st">"seurat_clusters"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SC_Visualizations_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clonalOccupy.html">clonalOccupy</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>             x.axis <span class="op">=</span> <span class="st">"ident"</span>, </span>
<span>             proportion <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>             label <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="SC_Visualizations_files/figure-html/unnamed-chunk-6-2.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="alluvialclones">alluvialClones<a class="anchor" aria-label="anchor" href="#alluvialclones"></a>
</h2>
<p>After the metadata has been modified, we can look at clones across
multiple categories using the <code><a href="../reference/alluvialClones.html">alluvialClones()</a></code> function. We
are able to use the plots to examine the interchange of categorical
variables. Because this function will produce a graph with each clone
arranged by called stratification, this will take some time depending on
the size of the repertoire.</p>
<p>To understand the basic concepts of this graphing method and the
ggalluvial R package, we recommend reading <a href="https://cran.r-project.org/web/packages/ggalluvial/vignettes/ggalluvial.html" class="external-link">this
post</a>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/alluvialClones.html">alluvialClones</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>               cloneCall <span class="op">=</span> <span class="st">"aa"</span>, </span>
<span>               y.axes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Patient"</span>, <span class="st">"ident"</span>, <span class="st">"Type"</span><span class="op">)</span>, </span>
<span>               color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CVVSDNTGGFKTIF_CASSVRRERANTGELFF"</span>, <span class="st">"NA_CASSVRRERANTGELFF"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grey"</span>, <span class="va">colorblind_vector</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SC_Visualizations_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/alluvialClones.html">alluvialClones</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                   cloneCall <span class="op">=</span> <span class="st">"gene"</span>, </span>
<span>                   y.axes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Patient"</span>, <span class="st">"ident"</span>, <span class="st">"Type"</span><span class="op">)</span>, </span>
<span>                   color <span class="op">=</span> <span class="st">"ident"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="SC_Visualizations_files/figure-html/unnamed-chunk-7-2.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="getcirclize">getCirclize<a class="anchor" aria-label="anchor" href="#getcirclize"></a>
</h2>
<p>Like alluvial graphs, we can also visualize the interconnection of
clusters using the chord diagrams from the <a href="https://jokergoo.github.io/circlize_book/book/" class="external-link">circlize R
package</a>.</p>
<p>The first step is getting the data frame output to feed into the
<code><a href="https://rdrr.io/pkg/circlize/man/chordDiagram.html" class="external-link">chordDiagram()</a></code> function in circlize, which can be done
using <code><a href="../reference/getCirclize.html">getCirclize()</a></code>.</p>
<p>This will calculate the relative number of clones shared based on the
<strong>group.by</strong> variable using the product of
<code><a href="../reference/combineExpression.html">combineExpression()</a></code>. It is important to note
<code><a href="../reference/getCirclize.html">getCirclize()</a></code> will create a matrix the size of the
<strong>group.by</strong> variable and then simplify into instructions
to be read by the circlize R package. The output is the total number of
unique and shared clones by the <strong>group.by</strong> variable.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/circlize" class="external-link">circlize</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org" class="external-link">scales</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">circles</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCirclize.html">getCirclize</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                       group.by <span class="op">=</span> <span class="st">"seurat_clusters"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Just assigning the normal colors to each cluster</span></span>
<span><span class="va">grid.cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://scales.r-lib.org/reference/pal_hue.html" class="external-link">hue_pal</a></span><span class="op">(</span><span class="op">)</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">scRep_example</span><span class="op">$</span><span class="va">seurat_clusters</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">grid.cols</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">scRep_example</span><span class="op">$</span><span class="va">seurat_clusters</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Graphing the chord diagram</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/circlize/man/chordDiagram.html" class="external-link">chordDiagram</a></span><span class="op">(</span><span class="va">circles</span>, self.link <span class="op">=</span> <span class="fl">1</span>, grid.col <span class="op">=</span> <span class="va">grid.cols</span><span class="op">)</span></span></code></pre></div>
<p><img src="SC_Visualizations_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>This can also be used if we want to explore just the lung-specific T
cells by just subsetting the single-cell object. For the sake of this
vignette, we can also look at setting <strong>proportion</strong> = TRUE
to get a scaled output.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">scRep_example</span>, <span class="va">Type</span> <span class="op">==</span> <span class="st">"L"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">circles</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCirclize.html">getCirclize</a></span><span class="op">(</span><span class="va">subset</span>, group.by <span class="op">=</span> <span class="st">"ident"</span>, proportion <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">grid.cols</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/pal_hue.html" class="external-link">hue_pal</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">subset</span><span class="op">@</span><span class="va">active.ident</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">grid.cols</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">subset</span><span class="op">@</span><span class="va">active.ident</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/circlize/man/chordDiagram.html" class="external-link">chordDiagram</a></span><span class="op">(</span><span class="va">circles</span>, self.link <span class="op">=</span> <span class="fl">1</span>, grid.col <span class="op">=</span> <span class="va">grid.cols</span>, directional <span class="op">=</span> <span class="fl">1</span>, direction.type <span class="op">=</span> <span class="st">"arrows"</span>,</span>
<span>    link.arr.type <span class="op">=</span> <span class="st">"big.arrow"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SC_Visualizations_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
