<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>TCRpheno applied to T cell fate • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="TCRpheno applied to T cell fate">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/dandelionR.html">VDJ Trajectory Analysis with dandelionR</a>
    </li>
    <li>
      <a href="../articles/immunarch.html">Using scRepertoire with immunarch</a>
    </li>
    <li>
      <a href="../articles/tcrpheno.html">TCRpheno applied to T cell fate</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BorchLab/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>TCRpheno applied to T cell fate</h1>
            
            <h4 data-toc-skip class="date">Compiled: July 22, 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BorchLab/scRepertoire/blob/HEAD/vignettes/articles/tcrpheno.Rmd" class="external-link"><code>vignettes/articles/tcrpheno.Rmd</code></a></small>
      <div class="hidden name"><code>tcrpheno.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p><code>tcrpheno</code> is an R package that applies a logistic
regression model to the amino acid sequences of T-cell receptor
complementarity-determining regions (CDRs) 1, 2, and 3. This model
produces phenotype scores associated with specific T cell fates,
providing insights into the potential functional trajectory of T cells
based on their TCR sequences.</p>
<div class="section level3">
<h3 id="more-information-on-individual-phenotypes">More information on individual phenotypes:<a class="anchor" aria-label="anchor" href="#more-information-on-individual-phenotypes"></a>
</h3>
<p>The tcrpheno package calculates four distinct scores, each linked to
a potential T cell phenotype</p>
<ul>
<li>
<strong>TCRinnate</strong>: Higher scores suggest a greater
likelihood of the T cell adopting an innate-like, <em>PLZF</em>-high
phenotype, characteristic of mucosal-associated invariant T (MAIT) cells
or invariant natural killer T (iNKT). This score is strongly influenced
by features in CDR2α and specific TRAV gene usage.<br>
</li>
<li>
<strong>TCR.8</strong>: Higher scores indicate a predisposition
towards a CD8+ T cell fate over a CD4+ fate. TCRs with high TCR.8 scores
tend to have a depletion of positive charge in the mid-region of their
CDR3 loops.<br>
</li>
<li>
<strong>TCRreg</strong>: Higher scores point to an increased
probability of the T cell becoming a regulatory T cell (Treg),
encompassing both CD4+ and CD8+ Treg populations. This is associated
with increased hydrophobicity in CDR3β and CDR3α residues.<br>
</li>
<li>
<strong>TCRmem</strong>: Higher scores suggest a T cell is more
likely to differentiate into a memory cell rather than remaining naive.
This score reflects a general propensity for T-cell activation and is
influenced by features in both CDR3α and CDR3β. Notably, higher TCRmem
scores correlate with increased T-cell activation even among T cells
recognizing the same antigen and correspond to the strength of positive
selection in the thymus.</li>
</ul>
</div>
<div class="section level3">
<h3 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h3>
<p>If using <em>tcrpheno</em>, please cite the <a href="https://pubmed.ncbi.nlm.nih.gov/39731734/" class="external-link">article</a>: Lagattuta,
K. et al. The T cell receptor sequence influences the likelihood of T
cell memory formation. Cell Reports. 2025 Jan 28;44(1)</p>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Ensure 'remotes' is installed: install.packages("remotes")</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"kalaga27/tcrpheno"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="loading-data">Loading Data<a class="anchor" aria-label="anchor" href="#loading-data"></a>
</h2>
<p>his vignette uses example data provided by the scRepertoire package
to demonstrate the workflow. For more details on scRepertoire’s example
data and loading mechanisms, refer <a href="https://www.borch.dev/uploads/screpertoire/articles/loading#example-data-in-screpertoire" class="external-link">here</a>.
For the sake of comparison, we will also filter out any cell without
clonal information.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load and normalize RNA</span></span>
<span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"scRep_example_full.rds"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                    <span class="fu">NormalizeData</span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Adding Clonal Information</span></span>
<span><span class="va">scRep_example</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineExpression.html">combineExpression</a></span><span class="op">(</span><span class="va">combined.TCR</span>, </span>
<span>                                    <span class="va">scRep_example</span> , </span>
<span>                                    cloneCall<span class="op">=</span><span class="st">"aa"</span>, </span>
<span>                                    group.by <span class="op">=</span> <span class="st">"sample"</span>, </span>
<span>                                    proportion <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Filtering for single-cells with TCRs</span></span>
<span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                        cells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">scRep_example</span><span class="op">)</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">scRep_example</span><span class="op">$</span><span class="va">CTaa</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="exporting-clonal-information-for-tcrpheno">Exporting Clonal Information for tcrpheno<a class="anchor" aria-label="anchor" href="#exporting-clonal-information-for-tcrpheno"></a>
</h2>
<p>The <code>tcrpheno</code> package requires TCR data in a specific
format. The <code><a href="../reference/exportClones.html">exportClones()</a></code> function from scRepertoire can
now output data directly in this <code>tcrpheno</code> format. This
format includes separate columns for TRA V gene, TRA J gene, TRA CDR3
sequence, TRB V gene, TRB J gene, and TRB CDR3 sequence, along with a
cell identifier.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exported_clones</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exportClones.html">exportClones</a></span><span class="op">(</span><span class="va">scRep_example</span>,</span>
<span>                                write.file <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                format <span class="op">=</span> <span class="st">"tcrpheno"</span><span class="op">)</span> </span>
<span><span class="va">exported_clones</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="va">exported_clones</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">exported_clones</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                      cell     TCRA_cdr3aa TCRA_vgene TCRA_jgene</span></span>
<span><span class="co">## 1 P17B_AAAGCAACAGACAAAT-1  CALFTSGNTGKLIF     TRAV16     TRAJ37</span></span>
<span><span class="co">## 2 P17B_AACCATGAGGCATGTG-1   CAVEDPRDYKLSF      TRAV2     TRAJ20</span></span>
<span><span class="co">## 4 P17B_AAGCCGCCAATCTACG-1 CALSEARETGNQFYF     TRAV19     TRAJ49</span></span>
<span><span class="co">## 5 P17B_AAGCCGCCATCCCACT-1   CAASINNNARLMF   TRAV13-1     TRAJ31</span></span>
<span><span class="co">## 6 P17B_AAGCCGCGTCACCTAA-1  CAVQAGDSWGKLQF     TRAV20     TRAJ24</span></span>
<span><span class="co">## 7 P17B_AAGGAGCGTTCTCATT-1  CATAPRDSWGKLQF     TRAV17     TRAJ24</span></span>
<span><span class="co">##                                     TCRA_cdr3nt      TCRB_cdr3aa TCRB_vgene</span></span>
<span><span class="co">## 1    TGTGCTCTCTTTACCTCTGGCAACACAGGCAAACTAATCTTT    CAIKGTGNGEQYF   TRBV10-3</span></span>
<span><span class="co">## 2       TGTGCTGTGGAGGATCCTCGGGACTACAAGCTCAGCTTT CASSLGGAGGGYEQYF   TRBV11-2</span></span>
<span><span class="co">## 4 TGTGCTCTGAGTGAGGCGAGGGAAACCGGTAACCAGTTCTATTTT  CASSQDADSFYEQYF    TRBV4-1</span></span>
<span><span class="co">## 5       TGTGCAGCAAGTATAAATAACAATGCCAGACTCATGTTT    CASSPVRTDTQYF    TRBV5-4</span></span>
<span><span class="co">## 6    TGTGCTGTGCAGGCCGGTGACAGCTGGGGGAAATTGCAGTTT   CASSLDGGSDTQYF   TRBV11-3</span></span>
<span><span class="co">## 7    TGTGCTACGGCCCCCAGGGACAGCTGGGGGAAATTGCAGTTT  CASSLYDTNTGELFF    TRBV5-1</span></span>
<span><span class="co">##   TCRB_jgene                                      TCRB_cdr3nt</span></span>
<span><span class="co">## 1    TRBJ2-7          TGTGCCATCAAGGGGACAGGGAATGGTGAGCAGTACTTC</span></span>
<span><span class="co">## 2    TRBJ2-7 TGTGCCAGCAGCTTGGGGGGCGCGGGTGGGGGCTACGAGCAGTACTTC</span></span>
<span><span class="co">## 4    TRBJ2-7    TGCGCCAGCAGCCAAGATGCGGACAGCTTCTACGAGCAGTACTTC</span></span>
<span><span class="co">## 5    TRBJ2-3          TGTGCCAGCAGCCCCGTTCGAACAGATACGCAGTATTTT</span></span>
<span><span class="co">## 6    TRBJ2-3       TGTGCCAGCAGCTTAGACGGGGGCTCAGATACGCAGTATTTT</span></span>
<span><span class="co">## 7    TRBJ2-2    TGCGCCAGCAGCTTGTACGACACAAACACCGGGGAGCTGTTTTTT</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="generating-phenotype-scores">Generating Phenotype Scores<a class="anchor" aria-label="anchor" href="#generating-phenotype-scores"></a>
</h2>
<p>With the TCR data correctly formatted, we can use the
<code>score_tcrs()</code> function from the <code>tcrpheno</code>
package to calculate the phenotype scores.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tcrpheno.results</span> <span class="op">&lt;-</span> <span class="fu">score_tcrs</span><span class="op">(</span><span class="va">exported_clones</span>, <span class="st">"ab"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "adding CDR1 and CDR2 based on V gene..."</span></span>
<span><span class="co">## [1] "identifying amino acids at each position..."</span></span>
<span><span class="co">## [1] "converting amino acids into Atchley factors..."</span></span>
<span><span class="co">## [1] 18670</span></span>
<span><span class="co">## [1] 18670</span></span>
<span><span class="co">## [1] "adding interactions between adjacent residues..."</span></span>
<span><span class="co">## [1] "TCRs featurized!"</span></span>
<span><span class="co">## [1] "scoring TCRs..."</span></span>
<span><span class="co">## [1] "all done!"</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tcrpheno.results</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                         TCR.innate     TCR.CD8    TCR.reg    TCR.mem</span></span>
<span><span class="co">## P17B_AAAGCAACAGACAAAT-1 -0.0675532 -1.19523036  0.4049552 -0.1555845</span></span>
<span><span class="co">## P17B_AACCATGAGGCATGTG-1  0.6501125  0.02883556 -1.1217330 -0.1431962</span></span>
<span><span class="co">## P17B_AAGCCGCCAATCTACG-1  0.2681321  3.05374879 -0.8430967 -0.1919169</span></span>
<span><span class="co">## P17B_AAGCCGCCATCCCACT-1 -0.1228840 -0.97613780  0.4996984 -1.2836692</span></span>
<span><span class="co">## P17B_AAGCCGCGTCACCTAA-1 -0.6160124  0.27602943 -1.0244453  0.5860345</span></span>
<span><span class="co">## P17B_AAGGAGCGTTCTCATT-1 -0.1305800 -0.06879752  0.3131880 -0.2055134</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="adding-to-single-cell-object">Adding to Single Cell Object<a class="anchor" aria-label="anchor" href="#adding-to-single-cell-object"></a>
</h2>
<p>To visualize and analyze these TCR-derived phenotype scores in
conjunction with gene expression data, we add them to the metadata of
our Seurat object.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scRep_example</span> <span class="op">&lt;-</span> <span class="fu">AddMetaData</span><span class="op">(</span><span class="va">scRep_example</span>, <span class="va">tcrpheno.results</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="visualizing-predictions">Visualizing Predictions<a class="anchor" aria-label="anchor" href="#visualizing-predictions"></a>
</h3>
<p>Now that the phenotype scores are part of the Seurat object, we can
visualize them on dimensionality reduction plots, such as UMAPs. This
helps to see if cells with particular TCR-derived phenotype scores
cluster together or co-localize with known cell populations.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tcrpheno.plots</span> <span class="op">&lt;-</span> <span class="fu">FeaturePlot</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                              features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TCR.mem"</span>, </span>
<span>                                           <span class="st">"TCR.reg"</span>, </span>
<span>                                           <span class="st">"TCR.CD8"</span>, </span>
<span>                                           <span class="st">"TCR.innate"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">tcrpheno.plots</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">x</span> <span class="op">+</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_color_viridis</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"B"</span><span class="op">)</span> </span>
<span><span class="op">}</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">tcrpheno.plots</span></span>
<span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">tcrpheno.plots</span><span class="op">)</span></span></code></pre></div>
<p><img src="tcrpheno_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="comparing-with-gene-expression">Comparing With Gene Expression<a class="anchor" aria-label="anchor" href="#comparing-with-gene-expression"></a>
</h3>
<p>A key aspect of integrating TCR phenotype scores is to compare them
with the expression of known marker genes associated with different T
cell states. This can help validate or provide biological context to the
<code>tcrpheno</code> predictions.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">RNA.plots</span> <span class="op">&lt;-</span> <span class="fu">FeaturePlot</span><span class="op">(</span><span class="va">scRep_example</span>, </span>
<span>                          features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>,<span class="st">"CD8A"</span>, <span class="st">"FOXP3"</span>, <span class="st">"KLRB1"</span><span class="op">)</span>, </span>
<span>                          combine <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">RNA.plots</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">x</span> <span class="op">+</span> <span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_color_viridis</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"B"</span><span class="op">)</span> </span>
<span><span class="op">}</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">RNA.plots</span></span>
<span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">RNA.plots</span><span class="op">)</span></span></code></pre></div>
<p><img src="tcrpheno_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>This concludes the vignette on applying <code>tcrpheno</code> to
predict T cell fate from TCR sequences and integrating these predictions
with single-cell RNA sequencing data.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
