<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Combining Deep Learning and TCRs with Trex • scRepertoire</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Combining Deep Learning and TCRs with Trex">
<meta property="og:description" content="scRepertoire">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    </li>
<li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ncborcherding/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Combining Deep Learning and TCRs with Trex</h1>
            
            <h4 data-toc-skip class="date">Compiled: April 05, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ncborcherding/scRepertoire/blob/HEAD/vignettes/articles/Trex.Rmd" class="external-link"><code>vignettes/articles/Trex.Rmd</code></a></small>
      <div class="hidden name"><code>Trex.Rmd</code></div>

    </div>

    
    
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="getting-started">Getting Started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>The idea behind Trex is to combine TCR CDR3 amino acid information
with phenotypic RNA/protein data to direct the use of single-cell
sequencing towards antigen-specific discoveries. This is a growing field
- specifically <a href="https://github.com/jcao89757/TESSA" class="external-link">TESSA</a>
uses amino acid characteristics and autoencoder as a means to get a
dimensional reduction. Another option is <a href="https://github.com/phbradley/conga" class="external-link">CoNGA</a>, which produces an
embedding using TCR and RNA with python. Trex was designed to make a
customizable approach to this combined approach using R.</p>
<p>More information is available at the <a href="https://github.com/ncborcherding/Trex" class="external-link">Trex GitHub Repo</a>.</p>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"ncborcherding/Trex"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="the-data-set">The Data Set<a class="anchor" aria-label="anchor" href="#the-data-set"></a>
</h3>
<p>To show the multiple options of Trex, the example data is derived
from <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE167118" class="external-link">GSE167118</a>,
a cohort of CITE-seq data derived from severe COVID-19 patients. More
information is available in the <a href="https://pubmed.ncbi.nlm.nih.gov/33622974/" class="external-link">corresponding
manuscript</a>.</p>
<div class="section level4">
<h4 id="formation-of-the-single-cell-object">Formation of the Single-cell Object<a class="anchor" aria-label="anchor" href="#formation-of-the-single-cell-object"></a>
</h4>
<p>Here is the basic workflow that was used to make the single-cell
object to use in the vignette. Notice there is a removal of TCR-related
CITE-seq data and RNA features (using the <strong>Trex</strong> function
<code>quietTCRgenes()</code>). As we are going to combine multimodal
data, both the GEX and CITE probes may cause bias in the weighted
output.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##################################</span></span>
<span><span class="co">#scRNA/ADT loading and processing</span></span>
<span><span class="co">#################################</span></span>
<span><span class="va">tmp</span> <span class="op">&lt;-</span>  <span class="fu">Read10X</span><span class="op">(</span><span class="st">"~/Patient17/filtered_feature_bc_matrix"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">Pt17</span> <span class="op">&lt;-</span> <span class="fu">CreateSeuratObject</span><span class="op">(</span>counts <span class="op">=</span> <span class="va">tmp</span><span class="op">$</span><span class="va">`Gene Expression`</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Removing TCR-specific antibody</span></span>
<span><span class="va">adt_assay</span> <span class="op">&lt;-</span> <span class="fu">CreateAssayObject</span><span class="op">(</span>counts <span class="op">=</span> <span class="va">tmp</span><span class="op">$</span><span class="va">`Antibody Capture`</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">37</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="va">Pt17</span><span class="op">[[</span><span class="st">"ADT"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">adt_assay</span></span>
<span><span class="va">Pt17</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">Pt17</span>, subset <span class="op">=</span> <span class="va">nFeature_RNA</span> <span class="op">&gt;</span> <span class="fl">100</span><span class="op">)</span> </span>
<span><span class="va">Pt17</span>  <span class="op">&lt;-</span> <span class="fu">RenameCells</span><span class="op">(</span>object <span class="op">=</span> <span class="va">Pt17</span> , new.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Pt17_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">Pt17</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">Pt17</span><span class="op">[[</span><span class="st">"mito.genes"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">PercentageFeatureSet</span><span class="op">(</span><span class="va">Pt17</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span></span>
<span>    </span>
<span><span class="co">#Filtering step</span></span>
<span><span class="va">standev</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html" class="external-link">sd</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">Pt17</span><span class="op">$</span><span class="va">nFeature_RNA</span><span class="op">)</span><span class="op">)</span><span class="op">*</span><span class="fl">2.5</span> <span class="co">#cutting off above standard deviation of 2.5</span></span>
<span><span class="va">mean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">Pt17</span><span class="op">$</span><span class="va">nFeature_RNA</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cut</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">standev</span><span class="op">+</span><span class="va">mean</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">Pt17</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">Pt17</span>, subset <span class="op">=</span> <span class="va">mito.genes</span> <span class="op">&lt;</span> <span class="fl">10</span> <span class="op">&amp;</span> <span class="va">nFeature_RNA</span> <span class="op">&lt;</span> <span class="va">cut</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Processing RNA</span></span>
<span><span class="fu">DefaultAssay</span><span class="op">(</span><span class="va">Pt17</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'RNA'</span></span>
<span><span class="va">Pt17</span> <span class="op">&lt;-</span> <span class="fu">NormalizeData</span><span class="op">(</span><span class="va">Pt17</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">FindVariableFeatures</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">quietTCRgenes</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">ScaleData</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">RunPCA</span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Processing ADT</span></span>
<span><span class="fu">DefaultAssay</span><span class="op">(</span><span class="va">Pt17</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'ADT'</span></span>
<span><span class="fu">VariableFeatures</span><span class="op">(</span><span class="va">Pt17</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">Pt17</span><span class="op">[[</span><span class="st">"ADT"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">Pt17</span> <span class="op">&lt;-</span> <span class="fu">NormalizeData</span><span class="op">(</span><span class="va">Pt17</span>, normalization.method <span class="op">=</span> <span class="st">'CLR'</span>, margin <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">ScaleData</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">RunPCA</span><span class="op">(</span>reduction.name <span class="op">=</span> <span class="st">'apca'</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">##################################</span></span>
<span><span class="co">#Processing and Adding Contig Info</span></span>
<span><span class="co">##################################</span></span>
<span></span>
<span><span class="va">contigs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"~/Patient17/filtered_contig_annotations.csv"</span><span class="op">)</span></span>
<span><span class="va">clones</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineTCR.html">combineTCR</a></span><span class="op">(</span><span class="va">contigs</span>, samples <span class="op">=</span> <span class="st">"Pt17"</span>, cells <span class="op">=</span> <span class="st">"T-AB"</span>, filterMulti <span class="op">=</span> <span class="cn">TRUE</span>, removeNA <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">Pt17</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineExpression.html">combineExpression</a></span><span class="op">(</span><span class="va">clones</span>, <span class="va">Pt17</span>, cloneCall<span class="op">=</span><span class="st">"aa"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">Pt17</span>, file <span class="op">=</span> <span class="st">"Trex_FullExample.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">###################################</span></span>
<span><span class="co">#Making Example Data Set for Trex</span></span>
<span><span class="co">#################################</span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="va">Pt17</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="va">meta</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span><span class="op">*</span><span class="fl">0.1</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">trex_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">Pt17</span>, cells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">trex_example</span>, file <span class="op">=</span> <span class="st">"trex_example.rda"</span>, compress <span class="op">=</span> <span class="st">"xz"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="loading-the-data-object">Loading the Data Object<a class="anchor" aria-label="anchor" href="#loading-the-data-object"></a>
</h4>
<p>For the purpose of the vignette, we will load the full object. The
data example built into the package (<strong>trex_example</strong>) is
derived from randomly sampling T cells from patient 17 (see above
workflow).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SeuratObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://www.borch.dev/uploads/data/Trex_FullExample.rds"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="running-trex">Running Trex<a class="anchor" aria-label="anchor" href="#running-trex"></a>
</h2>
<div class="section level3">
<h3 id="matrex-function">maTrex Function<a class="anchor" aria-label="anchor" href="#matrex-function"></a>
</h3>
<p>Trex has 2 major functions - the first being <code>maTrex()</code>,
which is the backbone of the algorithm and returns the encoded values
based on the selection of variables. Unlike <code>runTrex()</code>
below, <code>maTrex()</code> does not filter the input for only T cells
with attached TCR data. In addition, <code>maTrex()</code> is compatible
with the list output from the <code><a href="../reference/combineTCR.html">combineTCR()</a></code> function from
the <a href="https://github.com/ncborcherding/scRepertoire" class="external-link">scRepertoire</a> R
package, while <code>runTrex()</code> must be performed on a single-cell
object.</p>
<p><strong>chains</strong></p>
<ul>
<li>“TRA” for TCR Alpha Chain</li>
<li>“TRB” for TCR Beta Chain</li>
</ul>
<p><strong>method</strong></p>
<ul>
<li>“encoder” for a convolution neural network (CNN) based
encoding.</li>
<li>“geometric” for a geometric transformation</li>
</ul>
<p><strong>encoder.model</strong></p>
<ul>
<li>“VAE” for a variational autoencoder<br>
</li>
<li>“AE” for a traditional autoencoder</li>
</ul>
<p><strong>encoder.input</strong></p>
<ul>
<li>“AF” to use Atchley factors<br>
</li>
<li>“KF” to use Kidera factors<br>
</li>
<li>“both” to use both Atchley and Kidera factors</li>
<li>“OHE” for a One Hot Autoencoder</li>
</ul>
<p><strong>theta</strong><br>
If choosing the geometric transformation, what value of theta to use
(default is pi)</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Trex_vectors</span> <span class="op">&lt;-</span> <span class="fu">maTrex</span><span class="op">(</span><span class="va">SeuratObj</span>, </span>
<span>                       chains <span class="op">=</span> <span class="st">"TRA"</span>,</span>
<span>                       encoder.model <span class="op">=</span> <span class="st">"VAE"</span>, </span>
<span>                       encoder.input <span class="op">=</span> <span class="st">"AF"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Calculating the encoding values..."</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">Trex_vectors</span><span class="op">)</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">Trex_1</span>, <span class="va">Trex_2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="Trex_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Trex_vectors2</span> <span class="op">&lt;-</span> <span class="fu">maTrex</span><span class="op">(</span><span class="va">SeuratObj</span>, </span>
<span>                       chains <span class="op">=</span> <span class="st">"TRB"</span>,</span>
<span>                       method <span class="op">=</span> <span class="st">"geometric"</span>,</span>
<span>                       theta <span class="op">=</span> <span class="va">pi</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Performing geometric transformation..."</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">Trex_vectors2</span><span class="op">)</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Trex_1</span>, y <span class="op">=</span> <span class="va">Trex_2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="Trex_files/figure-html/unnamed-chunk-4-2.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="runtrex">runTrex<a class="anchor" aria-label="anchor" href="#runtrex"></a>
</h3>
<p>Additionally, <code>runTrex()</code> can be used to append the Seurat
or Single-cell Experiment object with the Trex vectors and allow for
further analysis. Importantly, <code>runTrex()</code> will remove single
cells that do not have recovered TCR data in the metadata of the
object.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SeuratObj</span> <span class="op">&lt;-</span> <span class="fu">runTrex</span><span class="op">(</span><span class="va">SeuratObj</span>, </span>
<span>                    chains <span class="op">=</span> <span class="st">"TRB"</span>,</span>
<span>                    encoder.model <span class="op">=</span> <span class="st">"VAE"</span>, </span>
<span>                    encoder.input <span class="op">=</span> <span class="st">"KF"</span>,</span>
<span>                    reduction.name <span class="op">=</span> <span class="st">"Trex.KF"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Calculating the encoding values..."</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="using-trex-vectors">Using Trex Vectors<a class="anchor" aria-label="anchor" href="#using-trex-vectors"></a>
</h2>
<p>After <code>runTrex()</code> we have the encoded values stored under
<strong>“Trex…”</strong>. Using the Trex reduction stored in Seurat, we
can calculate the nearest neighbor and shared nearest neighbor indexes
and generate a UMAP.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Generating UMAP from Trex Neighbors</span></span>
<span><span class="va">SeuratObj</span> <span class="op">&lt;-</span> <span class="fu">RunUMAP</span><span class="op">(</span><span class="va">SeuratObj</span>, </span>
<span>                     reduction <span class="op">=</span> <span class="st">"Trex.KF"</span>,</span>
<span>                     dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>,</span>
<span>                     reduction.name <span class="op">=</span> <span class="st">'Trex.umap'</span>, </span>
<span>                     reduction.key <span class="op">=</span> <span class="st">'trexUMAP_'</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Trex UMAP</span></span>
<span><span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">SeuratObj</span>, </span>
<span>                 reduction <span class="op">=</span> <span class="st">"Trex.umap"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>         <span class="fu">NoLegend</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">SeuratObj</span>, </span>
<span>                 group.by <span class="op">=</span> <span class="st">"CTaa"</span>, </span>
<span>                 reduction <span class="op">=</span> <span class="st">"Trex.umap"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>          <span class="fu">scale_color_viridis</span><span class="op">(</span>discrete <span class="op">=</span> <span class="cn">TRUE</span>, option <span class="op">=</span> <span class="st">"B"</span><span class="op">)</span>  <span class="op">+</span> </span>
<span>          <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>          <span class="fu">NoLegend</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot1</span> <span class="op">+</span> <span class="va">plot2</span></span></code></pre></div>
<p><img src="Trex_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>We now can use this in a similar way as other single-cell modalities
and calculate weighted nearest neighbor (WNN). To check out more on WNN,
please read the Satija’s group <a href="https://pubmed.ncbi.nlm.nih.gov/34062119/" class="external-link">paper</a>. We will use
the RNA, ADT protein levels, and Trex vectors for the WNN
calculations.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SeuratObj</span> <span class="op">&lt;-</span> <span class="fu">FindMultiModalNeighbors</span><span class="op">(</span></span>
<span>                    <span class="va">SeuratObj</span>, </span>
<span>                    reduction.list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"pca"</span>, <span class="st">"apca"</span>, <span class="st">"Trex.KF"</span><span class="op">)</span>, </span>
<span>                    dims.list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span>, </span>
<span>                    modality.weight.name <span class="op">=</span> <span class="st">"RNA.weight"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">SeuratObj</span> <span class="op">&lt;-</span> <span class="fu">RunUMAP</span><span class="op">(</span><span class="va">SeuratObj</span>, </span>
<span>                     nn.name <span class="op">=</span> <span class="st">"weighted.nn"</span>, </span>
<span>                     reduction.name <span class="op">=</span> <span class="st">"wnn.umap"</span>, </span>
<span>                     reduction.key <span class="op">=</span> <span class="st">"wnnUMAP_"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">SeuratObj</span> <span class="op">&lt;-</span> <span class="fu">FindClusters</span><span class="op">(</span><span class="va">SeuratObj</span>, </span>
<span>                          graph.name <span class="op">=</span> <span class="st">"wsnn"</span>, </span>
<span>                          resolution <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>                          algorithm <span class="op">=</span> <span class="fl">3</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#WNN UMAP</span></span>
<span><span class="va">plot3</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">SeuratObj</span>, </span>
<span>                 reduction <span class="op">=</span> <span class="st">"wnn.umap"</span><span class="op">)</span></span>
<span><span class="va">plot4</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">SeuratObj</span>, </span>
<span>                 reduction <span class="op">=</span> <span class="st">"wnn.umap"</span>, </span>
<span>                 group.by <span class="op">=</span> <span class="st">"CTaa"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>          <span class="fu">scale_color_viridis</span><span class="op">(</span>discrete <span class="op">=</span> <span class="cn">TRUE</span>, option <span class="op">=</span> <span class="st">"B"</span><span class="op">)</span>  <span class="op">+</span> </span>
<span>          <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>          <span class="fu">NoLegend</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot3</span> <span class="op">+</span> <span class="va">plot4</span></span></code></pre></div>
<p><img src="Trex_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<div class="section level3">
<h3 id="comparing-the-outcome-to-just-one-modality">Comparing the outcome to just one modality<a class="anchor" aria-label="anchor" href="#comparing-the-outcome-to-just-one-modality"></a>
</h3>
<p>We can also look at the differences in the UMAP generated from RNA,
ADT, or Trex as individual components. Remember, the clusters that we
are displaying in UMAP are based on clusters defined by the weighted
nearest neighbors calculated above.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SeuratObj</span> <span class="op">&lt;-</span> <span class="fu">RunUMAP</span><span class="op">(</span><span class="va">SeuratObj</span>, </span>
<span>                     reduction <span class="op">=</span> <span class="st">'pca'</span>, </span>
<span>                     dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, </span>
<span>                     assay <span class="op">=</span> <span class="st">'RNA'</span>, </span>
<span>                     reduction.name <span class="op">=</span> <span class="st">'rna.umap'</span>, </span>
<span>                     reduction.key <span class="op">=</span> <span class="st">'rnaUMAP_'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">SeuratObj</span> <span class="op">&lt;-</span> <span class="fu">RunUMAP</span><span class="op">(</span><span class="va">SeuratObj</span>, </span>
<span>                     reduction <span class="op">=</span> <span class="st">'apca'</span>, </span>
<span>                     dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, </span>
<span>                     assay <span class="op">=</span> <span class="st">'ADT'</span>, </span>
<span>                     reduction.name <span class="op">=</span> <span class="st">'adt.umap'</span>, </span>
<span>                     reduction.key <span class="op">=</span> <span class="st">'adtUMAP_'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot5</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">SeuratObj</span>, reduction <span class="op">=</span> <span class="st">"rna.umap"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">NoLegend</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">plot6</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">SeuratObj</span>, reduction <span class="op">=</span> <span class="st">"adt.umap"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">NoLegend</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">plot7</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">SeuratObj</span>, reduction <span class="op">=</span> <span class="st">"Trex.umap"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">NoLegend</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot5</span> <span class="op">+</span> <span class="va">plot6</span> <span class="op">+</span> <span class="va">plot7</span></span></code></pre></div>
<p><img src="Trex_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="conga-reduction">CoNGA Reduction<a class="anchor" aria-label="anchor" href="#conga-reduction"></a>
</h2>
<p>Recent <a href="https://pubmed.ncbi.nlm.nih.gov/34426704/" class="external-link">work</a>
has proposed using representative cells for the characterization of
clone and gene expression relationships. In order to generate these
representative cells, either a mean expression across a clone or using
the PCA dimensional space to identify a single cell that has the minimum
euclidean distance across a clone.</p>
<p>In order to generate a single-cell object based on the CoNGA
approach, Trex offers the function <code>CoNGAfy()</code>. For
<strong>method</strong>, select either “mean” or “dist” as described
above. After performing <code>CoNGAfy()</code>, the user can use any of
the above reduction strategies.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span>, include.only <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"%&gt;%"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">CoNGA.seurat</span> <span class="op">&lt;-</span> <span class="fu">CoNGAfy</span><span class="op">(</span><span class="va">SeuratObj</span>, </span>
<span>                         method <span class="op">=</span> <span class="st">"dist"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">CoNGA.seurat</span> <span class="op">&lt;-</span> <span class="fu">runTrex</span><span class="op">(</span><span class="va">CoNGA.seurat</span>, </span>
<span>                        chains <span class="op">=</span> <span class="st">"TRB"</span>,</span>
<span>                        encoder.model <span class="op">=</span> <span class="st">"VAE"</span>, </span>
<span>                        encoder.input <span class="op">=</span> <span class="st">"KF"</span>,</span>
<span>                        reduction.name <span class="op">=</span> <span class="st">"Trex.KF"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Calculating the encoding values..."</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CoNGA.seurat</span> <span class="op">&lt;-</span> <span class="va">CoNGA.seurat</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                  <span class="fu">FindNeighbors</span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">"Trex.KF"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                  <span class="fu">FindClusters</span><span class="op">(</span>algorithm <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of nodes: 805</span></span>
<span><span class="co">## Number of edges: 27729</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Running smart local moving algorithm...</span></span>
<span><span class="co">## Maximum modularity in 10 random starts: 0.6491</span></span>
<span><span class="co">## Number of communities: 7</span></span>
<span><span class="co">## Elapsed time: 0 seconds</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CoNGA.seurat</span> <span class="op">&lt;-</span> <span class="fu">RunUMAP</span><span class="op">(</span><span class="va">CoNGA.seurat</span>, </span>
<span>                     reduction <span class="op">=</span> <span class="st">"Trex.KF"</span>, </span>
<span>                     dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, </span>
<span>                     reduction.name <span class="op">=</span> <span class="st">'Trex.umap'</span>, </span>
<span>                     reduction.key <span class="op">=</span> <span class="st">'trexUMAP_'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">CoNGA.seurat</span>, reduction <span class="op">=</span> <span class="st">"Trex.umap"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">NoLegend</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="Trex_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="annotating-possible-epitopes">Annotating possible epitopes<a class="anchor" aria-label="anchor" href="#annotating-possible-epitopes"></a>
</h2>
<p>Towards find epitope relationships, Trex has a built in data base of
TCRA and TCRB sequences associated with epitopes. To append the database
to the single-cell object (either before or after
<code>CoNGAfy()</code>), you can use <code>annotateDB()</code>.</p>
<p>The database is collated from multiple sources and represents a great
deal of effort from these sources:</p>
<ol style="list-style-type: decimal">
<li>
<a href="https://vdjdb.CDR3.net/" class="external-link">VDJdb</a> - <a href="https://pubmed.ncbi.nlm.nih.gov/28977646/" class="external-link">citation</a>
</li>
<li>
<a href="http://friedmanlab.weizmann.ac.il/McPAS-TCR/" class="external-link">McPAS-TCR</a>
- <a href="https://pubmed.ncbi.nlm.nih.gov/28481982/" class="external-link">citation</a>
</li>
<li>
<a href="https://www.iedb.org/" class="external-link">iedb</a> - <a href="https://pubmed.ncbi.nlm.nih.gov/30357391/" class="external-link">citation</a>
</li>
<li>
<a href="https://db.cngb.org/pird/home/" class="external-link">TBA</a> (now called PIRD) -
<a href="https://pubmed.ncbi.nlm.nih.gov/31373607/" class="external-link">citation</a>
</li>
</ol>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CoNGA.seurat</span> <span class="op">&lt;-</span> <span class="fu">annotateDB</span><span class="op">(</span><span class="va">CoNGA.seurat</span>, </span>
<span>                           chains <span class="op">=</span> <span class="st">"TRB"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">CoNGA.seurat</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"Trex.umap"</span>, </span>
<span>        group.by <span class="op">=</span> <span class="st">"TRB_Epitope.species"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Trex_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">CoNGA.seurat</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"Trex.umap"</span>, </span>
<span>        group.by <span class="op">=</span> <span class="st">"TRB_Epitope.sequence"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Trex_files/figure-html/unnamed-chunk-10-2.png" width="700"></p>
<p>Or <code>annotateDB()</code> can be used on the full single-cell
object and examine the sequence information along the RNA-based UMAP. An
added feature to the function allows the annotations to be extended to
CDR3 sequences that are within defined edit distances from the reference
using <strong>edit.distance</strong>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SeuratObj</span> <span class="op">&lt;-</span> <span class="fu">annotateDB</span><span class="op">(</span><span class="va">SeuratObj</span>,</span>
<span>                        chains <span class="op">=</span> <span class="st">"TRB"</span>, </span>
<span>                        edit.distance <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">SeuratObj</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"rna.umap"</span>, </span>
<span>        group.by <span class="op">=</span> <span class="st">"TRB_Epitope.species"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">theme</span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Trex_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
