<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Starting work with scRepertoire. • scRepertoire</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Starting work with scRepertoire."><meta property="og:description" content="scRepertoire"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/ncborcherding/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>





<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Starting work with scRepertoire.</h1>
                        <h4 data-toc-skip class="author">Nick
Borcherding</h4>
            <address class="author_afil">
      Washington University in St. Louis, School of Medicine, St. Louis,
MO, USA<br>            <a class="author_email" href="mailto:#"><a
href="mailto:ncborch@gmail.com" class="email">ncborch@gmail.com</a></a>
      </address>
                  
            <h4 data-toc-skip class="date">November 17th, 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ncborcherding/scRepertoire/blob/HEAD/vignettes/vignette.Rmd" class="external-link"><code>vignettes/vignette.Rmd</code></a></small>
      <div class="hidden name"><code>vignette.Rmd</code></div>

    </div>

    
    
<pre><code>## Warning: package &#39;GenomeInfoDb&#39; was built under R version 4.3.3</code></pre>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>scRepertoire is designed to take filter contig outputs from the 10x
Genomics Cell Ranger pipeline, process that data to assign clonotype
based on two TCR or Ig chains and analyze the clonotype dynamics. The
latter can be separated into 1) clonotype-only analysis functions, such
as unique clonotypes or clonal space quantification, and 2) interaction
with mRNA expression data using Seurat or SingleCellExperiment
packages.</p>
<div id="loading-libraries" class="section level2">
<h2>Loading Libraries</h2>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(scRepertoire))</span></code></pre></div>
</div>
<div id="loading-data" class="section level2">
<h2>Loading Data</h2>
<div id="what-data-to-load-into-screpertoire" class="section level3">
<h3>What data to load into scRepertoire?</h3>
<p>scRepertoire functions using the
<strong>filtered_contig_annotations.csv</strong> output from the 10x
Genomics Cell Ranger. This file is located in the ./outs/ directory of
the VDJ alignment folder. To generate a list of contigs to use for
scRepertoire:</p>
<ul>
<li>load the <strong>filtered_contig_annotations.csv</strong> for each
of the samples.<br />
</li>
<li>make a list in the R environment.</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>S1 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;.../Sample1/outs/filtered_contig_annotations.csv&quot;</span>)</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>S2 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;.../Sample2/outs/filtered_contig_annotations.csv&quot;</span>)</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>S3 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;.../Sample3/outs/filtered_contig_annotations.csv&quot;</span>)</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>S4 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;.../Sample4/outs/filtered_contig_annotations.csv&quot;</span>)</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>contig_list <span class="ot">&lt;-</span> <span class="fu">list</span>(S1, S2, S3, S4)</span></code></pre></div>
</div>
<div id="other-alignment-workflows" class="section level3">
<h3>Other alignment workflows</h3>
<p>Beyond the default 10x Genomic Cell Ranger pipeline outputs,
scRepertoire supports the following <strong>single-cell</strong>
formats:</p>
<ul>
<li>AIRR<br />
</li>
<li>BD Rhapsody Multiomic Immune Profiling<br />
</li>
<li><a
href="https://immcantation.readthedocs.io/en/stable/index.html">Immcantation</a><br />
</li>
<li>JSON-formatted contig data<br />
</li>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/25924071/">MiXCR</a><br />
</li>
<li><a href="https://www.omniscope.ai/">Omniscope</a><br />
</li>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/33986545/">TRUST4</a><br />
</li>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/35674381/">WAT3R</a></li>
</ul>
<p><code>loadContigs()</code> can be given a directory where the
sequencing experiments are located and it will recursively load and
process the contig data based on the file names. Alternatively,
<code>loadContigs()</code> can be given a list of data frames and
process the contig data</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co">#Directory example</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>contig.output <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;~/Documents/MyExperiment&quot;</span>)</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>contig.list <span class="ot">&lt;-</span> <span class="fu">loadContigs</span>(<span class="at">input =</span> contig.output, </span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>                           <span class="at">format =</span> <span class="st">&quot;TRUST4&quot;</span>)</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co">#List of data frames example</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>S1 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;~/Documents/MyExperiment/Sample1/outs/barcode_results.csv&quot;</span>)</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>S2 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;~/Documents/MyExperiment/Sample2/outs/barcode_results.csv&quot;</span>)</span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>S3 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;~/Documents/MyExperiment/Sample3/outs/barcode_results.csv&quot;</span>)</span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>S4 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;~/Documents/MyExperiment/Sample4/outs/barcode_results.csv&quot;</span>)</span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a>contig_list <span class="ot">&lt;-</span> <span class="fu">list</span>(S1, S2, S3, S4)</span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a>contig.list <span class="ot">&lt;-</span> <span class="fu">loadContigs</span>(<span class="at">input =</span> contig.output, </span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a>                           <span class="at">format =</span> <span class="st">&quot;WAT3R&quot;</span>)</span></code></pre></div>
</div>
<div id="multiplexed-experiment" class="section level3">
<h3>Multiplexed Experiment</h3>
<p>It is now easy to create the contig list from a multiplexed
experiment by first generating a single-cell RNA object (either Seurat
or Single Cell Experiment), loading the filtered contig file and then
using <code>createHTOContigList()</code>. This function will return a
list separated by the <strong>group.by</strong> variable(s).</p>
<p>This function depends on the match of barcodes between the
single-cell object and contigs. If there is a prefix or different suffix
added to the barcode, this will result in no contigs recovered.
Currently, it is recommended you do this step before the integration, as
integration workflows commonly alter the barcodes. There is a
<strong>multi.run</strong> variable that can be used on the integrated
object. However, it assumes you have modified the barcodes with the
Seurat pipeline (automatic addition of _# to end), and your contig list
is in the same order.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>contigs <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;.../outs/filtered_contig_annotations.csv&quot;</span>)</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>contig.list <span class="ot">&lt;-</span> <span class="fu">createHTOContigList</span>(contigs, </span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>                                   Seurat.Obj, </span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>                                   <span class="at">group.by =</span> <span class="st">&quot;HTO_maxID&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="example-data-in-screpertoire" class="section level2">
<h2>Example Data in scRepertoire</h2>
<p>scRepertoire comes with a data set from T cells derived from four
patients with acute respiratory distress to demonstrate the
functionality of the R package. More information on the data set can be
found in the corresponding <a
href="https://pubmed.ncbi.nlm.nih.gov/33622974/">manuscript</a>. The
samples consist of paired peripheral-blood (B) and bronchoalveolar
lavage (L), effectively creating 8 distinct runs for T cell receptor
(TCR) enrichment. We can preview the elements in the list by using the
head function and looking at the first contig annotation.</p>
<p>The built-in example data is derived from the 10x Cell Ranger
pipeline, so it is ready to go for downstream processing and
analysis.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;contig_list&quot;</span>) <span class="co">#the data built into scRepertoire</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="fu">head</span>(contig_list[[<span class="dv">1</span>]])</span></code></pre></div>
<pre><code>##              barcode is_cell                   contig_id high_confidence length
## 1 AAACCTGAGTACGACG-1    True AAACCTGAGTACGACG-1_contig_1            True    500
## 2 AAACCTGAGTACGACG-1    True AAACCTGAGTACGACG-1_contig_2            True    478
## 4 AAACCTGCAACACGCC-1    True AAACCTGCAACACGCC-1_contig_1            True    506
## 5 AAACCTGCAACACGCC-1    True AAACCTGCAACACGCC-1_contig_2            True    470
## 6 AAACCTGCAGGCGATA-1    True AAACCTGCAGGCGATA-1_contig_1            True    558
## 7 AAACCTGCAGGCGATA-1    True AAACCTGCAGGCGATA-1_contig_2            True    505
##   chain       v_gene d_gene  j_gene c_gene full_length productive
## 1   TRA       TRAV25   None  TRAJ20   TRAC        True       True
## 2   TRB      TRBV5-1   None TRBJ2-7  TRBC2        True       True
## 4   TRA TRAV38-2/DV8   None  TRAJ52   TRAC        True       True
## 5   TRB     TRBV10-3   None TRBJ2-2  TRBC2        True       True
## 6   TRA     TRAV12-1   None   TRAJ9   TRAC        True       True
## 7   TRB        TRBV9   None TRBJ2-2  TRBC2        True       True
##                 cdr3                                                cdr3_nt
## 1        CGCSNDYKLSF                      TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT
## 2     CASSLTDRTYEQYF             TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC
## 4 CAYRSAQAGGTSYGKLTF TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT
## 5      CAISEQGKGELFF                TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT
## 6     CVVSDNTGGFKTIF             TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT
## 7  CASSVRRERANTGELFF    TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
##   reads umis raw_clonotype_id         raw_consensus_id
## 1  8344    4     clonotype123 clonotype123_consensus_2
## 2 65390   38     clonotype123 clonotype123_consensus_1
## 4 18372    8     clonotype124 clonotype124_consensus_1
## 5 34054    9     clonotype124 clonotype124_consensus_2
## 6  5018    2       clonotype1   clonotype1_consensus_2
## 7 25110   11       clonotype1   clonotype1_consensus_1</code></pre>
<hr />
</div>
</div>
<div id="combining-contigs-into-clones" class="section level1">
<h1>Combining Contigs into Clones</h1>
<div id="combinetcr" class="section level2">
<h2>combineTCR</h2>
<p><strong>input.data</strong></p>
<ul>
<li>List of <em>filtered_contig_annotations.csv</em> data frames from
the 10x Cell Ranger.<br />
</li>
<li>List of data processed using <code>loadContigs()</code>.</li>
</ul>
<p><strong>samples</strong> and <strong>ID</strong></p>
<ul>
<li>Grouping variables for downstream analysis and will be added as
prefixes to prevent issues with duplicate barcodes
<strong>(optional)</strong>.</li>
</ul>
<p><strong>removeNA</strong></p>
<ul>
<li>TRUE - Filter to remove any cell barcode with an NA value in at
least one of the chains.<br />
</li>
<li>FALSE - Include and incorporate cells with 1 NA value
<strong>(default)</strong>.</li>
</ul>
<p><strong>removeMulti</strong></p>
<ul>
<li>TRUE - Filter to remove any cell barcode with more than 2 immune
receptor chains.<br />
</li>
<li>FALSE - Include and incorporate cells with &gt; 2 chains
<strong>(default)</strong>.</li>
</ul>
<p><strong>filterMulti</strong></p>
<ul>
<li>TRUE - Isolate the top 2 expressed chains in cell barcodes with
multiple chains.<br />
</li>
<li>FALSE - Include and incorporate cells with &gt; 2 chains
<strong>(default)</strong>.</li>
</ul>
<p>The output of <code>combineTCR()</code> will be a list of contig data
frames that will be reduced to the reads associated with a single cell
barcode. It will also combine the multiple reads into clone calls by
either the nucleotide sequence (<strong>CTnt</strong>), amino acid
sequence (<strong>CTaa</strong>), the VDJC gene sequence
(<strong>CTgene</strong>), or the combination of the nucleotide and gene
sequence (<strong>CTstrict</strong>).</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>combined.TCR <span class="ot">&lt;-</span> <span class="fu">combineTCR</span>(contig_list, </span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>                           <span class="at">samples =</span> <span class="fu">c</span>(<span class="st">&quot;P17B&quot;</span>, <span class="st">&quot;P17L&quot;</span>, <span class="st">&quot;P18B&quot;</span>, <span class="st">&quot;P18L&quot;</span>, </span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>                                            <span class="st">&quot;P19B&quot;</span>,<span class="st">&quot;P19L&quot;</span>, <span class="st">&quot;P20B&quot;</span>, <span class="st">&quot;P20L&quot;</span>),</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>                           <span class="at">removeNA =</span> <span class="cn">FALSE</span>, </span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>                           <span class="at">removeMulti =</span> <span class="cn">FALSE</span>, </span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>                           <span class="at">filterMulti =</span> <span class="cn">FALSE</span>)</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="fu">head</span>(combined.TCR[[<span class="dv">1</span>]])</span></code></pre></div>
<pre><code>##                    barcode sample                     TCR1           cdr3_aa1
## 1  P17B_AAACCTGAGTACGACG-1   P17B       TRAV25.TRAJ20.TRAC        CGCSNDYKLSF
## 3  P17B_AAACCTGCAACACGCC-1   P17B TRAV38-2/DV8.TRAJ52.TRAC CAYRSAQAGGTSYGKLTF
## 5  P17B_AAACCTGCAGGCGATA-1   P17B      TRAV12-1.TRAJ9.TRAC     CVVSDNTGGFKTIF
## 7  P17B_AAACCTGCATGAGCGA-1   P17B      TRAV12-1.TRAJ9.TRAC     CVVSDNTGGFKTIF
## 9  P17B_AAACGGGAGAGCCCAA-1   P17B        TRAV20.TRAJ8.TRAC      CAVRGEGFQKLVF
## 10 P17B_AAACGGGAGCGTTTAC-1   P17B      TRAV12-1.TRAJ9.TRAC     CVVSDNTGGFKTIF
##                                                  cdr3_nt1
## 1                       TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT
## 3  TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT
## 5              TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT
## 7              TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT
## 9                 TGTGCTGTGCGAGGAGAAGGCTTTCAGAAACTTGTATTT
## 10             TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT
##                           TCR2          cdr3_aa2
## 1   TRBV5-1.None.TRBJ2-7.TRBC2    CASSLTDRTYEQYF
## 3  TRBV10-3.None.TRBJ2-2.TRBC2     CAISEQGKGELFF
## 5     TRBV9.None.TRBJ2-2.TRBC2 CASSVRRERANTGELFF
## 7     TRBV9.None.TRBJ2-2.TRBC2 CASSVRRERANTGELFF
## 9                         &lt;NA&gt;              &lt;NA&gt;
## 10    TRBV9.None.TRBJ2-2.TRBC2 CASSVRRERANTGELFF
##                                               cdr3_nt2
## 1           TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC
## 3              TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT
## 5  TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 7  TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 9                                                 &lt;NA&gt;
## 10 TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
##                                                  CTgene
## 1         TRAV25.TRAJ20.TRAC_TRBV5-1.None.TRBJ2-7.TRBC2
## 3  TRAV38-2/DV8.TRAJ52.TRAC_TRBV10-3.None.TRBJ2-2.TRBC2
## 5          TRAV12-1.TRAJ9.TRAC_TRBV9.None.TRBJ2-2.TRBC2
## 7          TRAV12-1.TRAJ9.TRAC_TRBV9.None.TRBJ2-2.TRBC2
## 9                                  TRAV20.TRAJ8.TRAC_NA
## 10         TRAV12-1.TRAJ9.TRAC_TRBV9.None.TRBJ2-2.TRBC2
##                                                                                              CTnt
## 1                    TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT_TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC
## 3  TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT_TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT
## 5  TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 7  TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 9                                                      TGTGCTGTGCGAGGAGAAGGCTTTCAGAAACTTGTATTT_NA
## 10 TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
##                                CTaa
## 1        CGCSNDYKLSF_CASSLTDRTYEQYF
## 3  CAYRSAQAGGTSYGKLTF_CAISEQGKGELFF
## 5  CVVSDNTGGFKTIF_CASSVRRERANTGELFF
## 7  CVVSDNTGGFKTIF_CASSVRRERANTGELFF
## 9                  CAVRGEGFQKLVF_NA
## 10 CVVSDNTGGFKTIF_CASSVRRERANTGELFF
##                                                                                                                                               CTstrict
## 1                           TRAV25.TRAJ20.TRAC;TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT_TRBV5-1.None.TRBJ2-7.TRBC2;TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC
## 3  TRAV38-2/DV8.TRAJ52.TRAC;TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT_TRBV10-3.None.TRBJ2-2.TRBC2;TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT
## 5          TRAV12-1.TRAJ9.TRAC;TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TRBV9.None.TRBJ2-2.TRBC2;TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 7          TRAV12-1.TRAJ9.TRAC;TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TRBV9.None.TRBJ2-2.TRBC2;TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 9                                                                                      TRAV20.TRAJ8.TRAC;TGTGCTGTGCGAGGAGAAGGCTTTCAGAAACTTGTATTT_NA;NA
## 10         TRAV12-1.TRAJ9.TRAC;TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TRBV9.None.TRBJ2-2.TRBC2;TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</code></pre>
</div>
<div id="combinebcr" class="section level2">
<h2>combineBCR</h2>
<p><code>combineBCR()</code> is analogous to <code>combineTCR()</code>
with 2 major changes: <strong>1)</strong> Each barcode can only have a
maximum of 2 sequences, if greater exists, the 2 with the highest reads
are selected; <strong>2)</strong> The <strong>strict</strong> definition
of a clone is based on the normalized Levenshtein edit distance of CDR3
nucleotide sequences and V-gene usage. For more information on this
approach, please see the respective <a
href="https://pubmed.ncbi.nlm.nih.gov/34161770/">citation</a>. This
definition allows for the grouping of BCRs derived from the same
progenitor that have undergone mutation as part of somatic hypermutation
and affinity maturation.</p>
<p><strong>threshold</strong><br />
The level of similarity in sequences to group together.
<strong>Default</strong> is 0.85.</p>
<p><span class="math display">\[
\text{threshold}(s, t) = 1-\frac{\text{Levenshtein}(s,
t)}{\frac{\text{length}(s) + \text{length}(t)}{2}}
\]</span></p>
<p><strong>call.related.clones</strong><br />
Calculate the normalized edit distance (<strong>TRUE</strong>) or skip
the calculation (<strong>FALSE</strong>). Skipping the edit distance
calculation may save time, especially in the context of large data sets,
but is not recommended.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>BCR.contigs <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;https://www.borch.dev/uploads/contigs/b_contigs.csv&quot;</span>)</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>combined.BCR <span class="ot">&lt;-</span> <span class="fu">combineBCR</span>(BCR.contigs, </span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>                           <span class="at">samples =</span> <span class="st">&quot;P1&quot;</span>, </span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>                           <span class="at">threshold =</span> <span class="fl">0.85</span>)</span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a><span class="fu">head</span>(combined.BCR[[<span class="dv">1</span>]])</span></code></pre></div>
<pre><code>##                 barcode sample                             IGH
## 1 P1_CGAACATTCCCTTGTG-1     P1                            &lt;NA&gt;
## 2 P1_CCTTCGAAGACGCTTT-1     P1                            &lt;NA&gt;
## 3 P1_CAGCTAACAGCTTAAC-1     P1                            &lt;NA&gt;
## 4 P1_CCCAATCAGAGGGCTT-1     P1 IGHV1-69-2.IGHD5-12.IGHJ6.IGHG1
## 5 P1_TACAGTGGTAAGGGCT-1     P1   IGHV1-18.IGHD3-10.IGHJ5.IGHA1
## 6 P1_GACAGAGGTTGGTAAA-1     P1   IGHV1-18.IGHD3-10.IGHJ5.IGHA1
##                cdr3_aa1
## 1                  &lt;NA&gt;
## 2                  &lt;NA&gt;
## 3                  &lt;NA&gt;
## 4  CASGGLVAIRHYYYYGMDVW
## 5 CARVTYHHGSGSLVGGWFDPW
## 6 CARVTYHHGSGSLVGGWFDPW
##                                                          cdr3_nt1
## 1                                                            &lt;NA&gt;
## 2                                                            &lt;NA&gt;
## 3                                                            &lt;NA&gt;
## 4    TGTGCGAGCGGGGGATTAGTGGCTATTAGGCACTACTACTACTACGGTATGGACGTCTGG
## 5 TGTGCGAGAGTGACATATCACCATGGTTCGGGGAGCCTTGTCGGGGGCTGGTTCGACCCCTGG
## 6 TGTGCGAGAGTGACATATCACCATGGTTCGGGGAGCCTTGTCGGGGGCTGGTTCGACCCCTGG
##                   IGLC      cdr3_aa2                                cdr3_nt2
## 1 IGLV2-14.IGLJ6.IGLC2  CNSKGAGGTAVF    TGCAACTCAAAAGGAGCCGGAGGCACTGCGGTTTTC
## 2 IGLV2-14.IGLJ6.IGLC2  CNSKGAGGTAVF    TGCAACTCAAAAGGAGCCGGAGGCACTGCGGTTTTC
## 3 IGLV2-14.IGLJ2.IGLC2 CNSYTTSGTLWVF TGCAACTCATACACAACCAGCGGCACTCTCTGGGTATTC
## 4 IGLV2-14.IGLJ2.IGLC2  CNSYTSSSTLVF    TGCAACTCATATACAAGCAGCAGCACTCTGGTCTTC
## 5 IGLV2-14.IGLJ1.IGLC1  CNSYTSFGTSVF    TGCAACTCATATACAAGCTTCGGCACCTCGGTCTTC
## 6 IGLV2-14.IGLJ1.IGLC1  CNSYTSFGTSVF    TGCAACTCATATACAAGCTTCGGCACCTCGGTCTTC
##                                                 CTgene
## 1                              NA_IGLV2-14.IGLJ6.IGLC2
## 2                              NA_IGLV2-14.IGLJ6.IGLC2
## 3                              NA_IGLV2-14.IGLJ2.IGLC2
## 4 IGHV1-69-2.IGHD5-12.IGHJ6.IGHG1_IGLV2-14.IGLJ2.IGLC2
## 5   IGHV1-18.IGHD3-10.IGHJ5.IGHA1_IGLV2-14.IGLJ1.IGLC1
## 6   IGHV1-18.IGHD3-10.IGHJ5.IGHA1_IGLV2-14.IGLJ1.IGLC1
##                                                                                                   CTnt
## 1                                                              NA_TGCAACTCAAAAGGAGCCGGAGGCACTGCGGTTTTC
## 2                                                              NA_TGCAACTCAAAAGGAGCCGGAGGCACTGCGGTTTTC
## 3                                                           NA_TGCAACTCATACACAACCAGCGGCACTCTCTGGGTATTC
## 4    TGTGCGAGCGGGGGATTAGTGGCTATTAGGCACTACTACTACTACGGTATGGACGTCTGG_TGCAACTCATATACAAGCAGCAGCACTCTGGTCTTC
## 5 TGTGCGAGAGTGACATATCACCATGGTTCGGGGAGCCTTGTCGGGGGCTGGTTCGACCCCTGG_TGCAACTCATATACAAGCTTCGGCACCTCGGTCTTC
## 6 TGTGCGAGAGTGACATATCACCATGGTTCGGGGAGCCTTGTCGGGGGCTGGTTCGACCCCTGG_TGCAACTCATATACAAGCTTCGGCACCTCGGTCTTC
##                                 CTaa
## 1                    NA_CNSKGAGGTAVF
## 2                    NA_CNSKGAGGTAVF
## 3                   NA_CNSYTTSGTLWVF
## 4  CASGGLVAIRHYYYYGMDVW_CNSYTSSSTLVF
## 5 CARVTYHHGSGSLVGGWFDPW_CNSYTSFGTSVF
## 6 CARVTYHHGSGSLVGGWFDPW_CNSYTSFGTSVF
##                                      CTstrict
## 1                     NA.NA_IGLC.549.IGLV2-14
## 2                     NA.NA_IGLC.549.IGLV2-14
## 3                     NA.NA_IGLC.535.IGLV2-14
## 4 IGH.481.IGHV1-69-2_IGLC:Cluster.56.IGLV2-14
## 5   IGH.781.IGHV1-18_IGLC:Cluster.56.IGLV2-14
## 6   IGH.781.IGHV1-18_IGLC:Cluster.56.IGLV2-14</code></pre>
<hr />
</div>
</div>
<div id="additional-processing" class="section level1">
<h1>Additional Processing</h1>
<div id="addvariable" class="section level2">
<h2>addVariable</h2>
<p>What if there are more variables to add than just sample and ID? We
can add them by using the <code>addVariable()</code> function. All we
need is the <strong>variable.name</strong> of the variable you’d like to
add and the specific character or numeric values
(<strong>variables</strong>). As an example, here we add the
<strong>Type</strong> in which the samples were processed and
sequenced.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a>combined.TCR <span class="ot">&lt;-</span> <span class="fu">addVariable</span>(combined.TCR, </span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>                            <span class="at">variable.name =</span> <span class="st">&quot;Type&quot;</span>, </span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>                            <span class="at">variables =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">&quot;B&quot;</span>, <span class="st">&quot;L&quot;</span>), <span class="dv">4</span>))</span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="fu">head</span>(combined.TCR[[<span class="dv">1</span>]])</span></code></pre></div>
<pre><code>##                    barcode sample                     TCR1           cdr3_aa1
## 1  P17B_AAACCTGAGTACGACG-1   P17B       TRAV25.TRAJ20.TRAC        CGCSNDYKLSF
## 3  P17B_AAACCTGCAACACGCC-1   P17B TRAV38-2/DV8.TRAJ52.TRAC CAYRSAQAGGTSYGKLTF
## 5  P17B_AAACCTGCAGGCGATA-1   P17B      TRAV12-1.TRAJ9.TRAC     CVVSDNTGGFKTIF
## 7  P17B_AAACCTGCATGAGCGA-1   P17B      TRAV12-1.TRAJ9.TRAC     CVVSDNTGGFKTIF
## 9  P17B_AAACGGGAGAGCCCAA-1   P17B        TRAV20.TRAJ8.TRAC      CAVRGEGFQKLVF
## 10 P17B_AAACGGGAGCGTTTAC-1   P17B      TRAV12-1.TRAJ9.TRAC     CVVSDNTGGFKTIF
##                                                  cdr3_nt1
## 1                       TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT
## 3  TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT
## 5              TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT
## 7              TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT
## 9                 TGTGCTGTGCGAGGAGAAGGCTTTCAGAAACTTGTATTT
## 10             TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT
##                           TCR2          cdr3_aa2
## 1   TRBV5-1.None.TRBJ2-7.TRBC2    CASSLTDRTYEQYF
## 3  TRBV10-3.None.TRBJ2-2.TRBC2     CAISEQGKGELFF
## 5     TRBV9.None.TRBJ2-2.TRBC2 CASSVRRERANTGELFF
## 7     TRBV9.None.TRBJ2-2.TRBC2 CASSVRRERANTGELFF
## 9                         &lt;NA&gt;              &lt;NA&gt;
## 10    TRBV9.None.TRBJ2-2.TRBC2 CASSVRRERANTGELFF
##                                               cdr3_nt2
## 1           TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC
## 3              TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT
## 5  TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 7  TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 9                                                 &lt;NA&gt;
## 10 TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
##                                                  CTgene
## 1         TRAV25.TRAJ20.TRAC_TRBV5-1.None.TRBJ2-7.TRBC2
## 3  TRAV38-2/DV8.TRAJ52.TRAC_TRBV10-3.None.TRBJ2-2.TRBC2
## 5          TRAV12-1.TRAJ9.TRAC_TRBV9.None.TRBJ2-2.TRBC2
## 7          TRAV12-1.TRAJ9.TRAC_TRBV9.None.TRBJ2-2.TRBC2
## 9                                  TRAV20.TRAJ8.TRAC_NA
## 10         TRAV12-1.TRAJ9.TRAC_TRBV9.None.TRBJ2-2.TRBC2
##                                                                                              CTnt
## 1                    TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT_TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC
## 3  TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT_TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT
## 5  TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 7  TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 9                                                      TGTGCTGTGCGAGGAGAAGGCTTTCAGAAACTTGTATTT_NA
## 10 TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
##                                CTaa
## 1        CGCSNDYKLSF_CASSLTDRTYEQYF
## 3  CAYRSAQAGGTSYGKLTF_CAISEQGKGELFF
## 5  CVVSDNTGGFKTIF_CASSVRRERANTGELFF
## 7  CVVSDNTGGFKTIF_CASSVRRERANTGELFF
## 9                  CAVRGEGFQKLVF_NA
## 10 CVVSDNTGGFKTIF_CASSVRRERANTGELFF
##                                                                                                                                               CTstrict
## 1                           TRAV25.TRAJ20.TRAC;TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT_TRBV5-1.None.TRBJ2-7.TRBC2;TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC
## 3  TRAV38-2/DV8.TRAJ52.TRAC;TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT_TRBV10-3.None.TRBJ2-2.TRBC2;TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT
## 5          TRAV12-1.TRAJ9.TRAC;TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TRBV9.None.TRBJ2-2.TRBC2;TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 7          TRAV12-1.TRAJ9.TRAC;TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TRBV9.None.TRBJ2-2.TRBC2;TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 9                                                                                      TRAV20.TRAJ8.TRAC;TGTGCTGTGCGAGGAGAAGGCTTTCAGAAACTTGTATTT_NA;NA
## 10         TRAV12-1.TRAJ9.TRAC;TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TRBV9.None.TRBJ2-2.TRBC2;TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
##    Type
## 1     B
## 3     B
## 5     B
## 7     B
## 9     B
## 10    B</code></pre>
</div>
<div id="subsetclones" class="section level2">
<h2>subsetClones</h2>
<p>Likewise, we can remove specific list elements after
<code>combineTCR()</code> using the <code>subsetClones()</code>
function. In order to subset, we need to identify the vector we would
like to use for subsetting (<strong>name</strong>) and the variable
values to subset (<strong>variables</strong>). Below, we isolate just
the 2 sequencing results from P18L and P18B.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a>subset1 <span class="ot">&lt;-</span> <span class="fu">subsetClones</span>(combined.TCR, </span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>                        <span class="at">name =</span> <span class="st">&quot;sample&quot;</span>, </span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>                        <span class="at">variables =</span> <span class="fu">c</span>(<span class="st">&quot;P18L&quot;</span>, <span class="st">&quot;P18B&quot;</span>))</span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="fu">head</span>(subset1[[<span class="dv">1</span>]])</span></code></pre></div>
<pre><code>##                    barcode sample                 TCR1         cdr3_aa1
## 1  P18B_AAACCTGAGGCTCAGA-1   P18B TRAV26-1.TRAJ37.TRAC  CIVRGGSSNTGKLIF
## 3  P18B_AAACCTGCATGACATC-1   P18B    TRAV3.TRAJ20.TRAC    CAVQRSNDYKLSF
## 5  P18B_AAACCTGGTATGCTTG-1   P18B TRAV26-1.TRAJ53.TRAC   CIGSSGGSNYKLTF
## 8  P18B_AAACGGGCAGATGGGT-1   P18B                 &lt;NA&gt;             &lt;NA&gt;
## 9  P18B_AAACGGGTCTTACCGC-1   P18B    TRAV20.TRAJ9.TRAC CAVQAKRYTGGFKTIF
## 12 P18B_AAAGATGAGTTACGGG-1   P18B   TRAV8-3.TRAJ8.TRAC   CAVGGDTGFQKLVF
##                                            cdr3_nt1
## 1     TGCATCGTCAGGGGCGGCTCTAGCAACACAGGCAAACTAATCTTT
## 3           TGTGCTGTGCAACGTTCTAACGACTACAAGCTCAGCTTT
## 5        TGCATCGGCTCAAGTGGAGGTAGCAACTATAAACTGACATTT
## 8                                              &lt;NA&gt;
## 9  TGTGCTGTGCAGGCCAAGCGGTATACTGGAGGCTTCAAAACTATCTTT
## 12       TGTGCTGTGGGTGGTGACACAGGCTTTCAGAAACTTGTATTT
##                                                     TCR2
## 1                             TRBV6-1.None.TRBJ2-3.TRBC2
## 3                             TRBV3-1.None.TRBJ2-3.TRBC2
## 5   TRBV4-1.None.TRBJ2-2.TRBC2;TRBV19.None.TRBJ1-5.TRBC1
## 8                             TRBV5-1.None.TRBJ1-2.TRBC1
## 9  TRBV5-1.None.TRBJ1-1.TRBC1;TRBV7-9.None.TRBJ2-2.TRBC2
## 12                           TRBV12-4.None.TRBJ1-1.TRBC1
##                          cdr3_aa2
## 1                 CASIGRSFGRDTQYF
## 3                CASSPPRGGFTDTQYF
## 5  CASSQGGQGGRELFF;CASSYAVGRQPQHF
## 8                  CASSLRETNYGYTF
## 9  CASSLGTGTGVEAFF;CAIDPGLLTGELFF
## 12                  CASRNSQATEAFF
##                                                                                    cdr3_nt2
## 1                                             TGTGCCAGTATCGGGAGGTCCTTTGGCCGAGATACGCAGTATTTT
## 3                                          TGTGCCAGCAGCCCCCCCCGCGGCGGATTCACAGATACGCAGTATTTT
## 5  TGCGCCAGCAGCCAAGGTGGACAGGGCGGAAGGGAGCTGTTTTTT;TGTGCCAGTAGCTACGCGGTGGGGAGGCAGCCCCAGCATTTT
## 8                                                TGCGCCAGCAGCTTGAGGGAAACCAACTATGGCTACACCTTC
## 9  TGCGCCAGCAGCTTGGGAACGGGGACAGGGGTTGAAGCTTTCTTT;TGTGCCATCGATCCGGGACTACTCACCGGGGAGCTGTTTTTT
## 12                                                  TGTGCCAGCAGAAACTCCCAAGCCACTGAAGCTTTCTTT
##                                                                       CTgene
## 1                            TRAV26-1.TRAJ37.TRAC_TRBV6-1.None.TRBJ2-3.TRBC2
## 3                               TRAV3.TRAJ20.TRAC_TRBV3-1.None.TRBJ2-3.TRBC2
## 5  TRAV26-1.TRAJ53.TRAC_TRBV4-1.None.TRBJ2-2.TRBC2;TRBV19.None.TRBJ1-5.TRBC1
## 8                                              NA_TRBV5-1.None.TRBJ1-2.TRBC1
## 9    TRAV20.TRAJ9.TRAC_TRBV5-1.None.TRBJ1-1.TRBC1;TRBV7-9.None.TRBJ2-2.TRBC2
## 12                            TRAV8-3.TRAJ8.TRAC_TRBV12-4.None.TRBJ1-1.TRBC1
##                                                                                                                                         CTnt
## 1                                                TGCATCGTCAGGGGCGGCTCTAGCAACACAGGCAAACTAATCTTT_TGTGCCAGTATCGGGAGGTCCTTTGGCCGAGATACGCAGTATTTT
## 3                                                   TGTGCTGTGCAACGTTCTAACGACTACAAGCTCAGCTTT_TGTGCCAGCAGCCCCCCCCGCGGCGGATTCACAGATACGCAGTATTTT
## 5        TGCATCGGCTCAAGTGGAGGTAGCAACTATAAACTGACATTT_TGCGCCAGCAGCCAAGGTGGACAGGGCGGAAGGGAGCTGTTTTTT;TGTGCCAGTAGCTACGCGGTGGGGAGGCAGCCCCAGCATTTT
## 8                                                                                              NA_TGCGCCAGCAGCTTGAGGGAAACCAACTATGGCTACACCTTC
## 9  TGTGCTGTGCAGGCCAAGCGGTATACTGGAGGCTTCAAAACTATCTTT_TGCGCCAGCAGCTTGGGAACGGGGACAGGGGTTGAAGCTTTCTTT;TGTGCCATCGATCCGGGACTACTCACCGGGGAGCTGTTTTTT
## 12                                                        TGTGCTGTGGGTGGTGACACAGGCTTTCAGAAACTTGTATTT_TGTGCCAGCAGAAACTCCCAAGCCACTGAAGCTTTCTTT
##                                               CTaa
## 1                  CIVRGGSSNTGKLIF_CASIGRSFGRDTQYF
## 3                   CAVQRSNDYKLSF_CASSPPRGGFTDTQYF
## 5    CIGSSGGSNYKLTF_CASSQGGQGGRELFF;CASSYAVGRQPQHF
## 8                                NA_CASSLRETNYGYTF
## 9  CAVQAKRYTGGFKTIF_CASSLGTGTGVEAFF;CAIDPGLLTGELFF
## 12                    CAVGGDTGFQKLVF_CASRNSQATEAFF
##                                                                                                                                                                                                             CTstrict
## 1                                                                        TRAV26-1.TRAJ37.TRAC;TGCATCGTCAGGGGCGGCTCTAGCAACACAGGCAAACTAATCTTT_TRBV6-1.None.TRBJ2-3.TRBC2;TGTGCCAGTATCGGGAGGTCCTTTGGCCGAGATACGCAGTATTTT
## 3                                                                              TRAV3.TRAJ20.TRAC;TGTGCTGTGCAACGTTCTAACGACTACAAGCTCAGCTTT_TRBV3-1.None.TRBJ2-3.TRBC2;TGTGCCAGCAGCCCCCCCCGCGGCGGATTCACAGATACGCAGTATTTT
## 5      TRAV26-1.TRAJ53.TRAC;TGCATCGGCTCAAGTGGAGGTAGCAACTATAAACTGACATTT_TRBV4-1.None.TRBJ2-2.TRBC2;TRBV19.None.TRBJ1-5.TRBC1;TGCGCCAGCAGCCAAGGTGGACAGGGCGGAAGGGAGCTGTTTTTT;TGTGCCAGTAGCTACGCGGTGGGGAGGCAGCCCCAGCATTTT
## 8                                                                                                                                        NA;NA_TRBV5-1.None.TRBJ1-2.TRBC1;TGCGCCAGCAGCTTGAGGGAAACCAACTATGGCTACACCTTC
## 9  TRAV20.TRAJ9.TRAC;TGTGCTGTGCAGGCCAAGCGGTATACTGGAGGCTTCAAAACTATCTTT_TRBV5-1.None.TRBJ1-1.TRBC1;TRBV7-9.None.TRBJ2-2.TRBC2;TGCGCCAGCAGCTTGGGAACGGGGACAGGGGTTGAAGCTTTCTTT;TGTGCCATCGATCCGGGACTACTCACCGGGGAGCTGTTTTTT
## 12                                                                                 TRAV8-3.TRAJ8.TRAC;TGTGCTGTGGGTGGTGACACAGGCTTTCAGAAACTTGTATTT_TRBV12-4.None.TRBJ1-1.TRBC1;TGTGCCAGCAGAAACTCCCAAGCCACTGAAGCTTTCTTT
##    Type
## 1     B
## 3     B
## 5     B
## 8     B
## 9     B
## 12    B</code></pre>
<p>Alternatively, we can also just select the list elements after
<code>combineTCR()</code> or <code>combineBCR()</code>.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a>subset2 <span class="ot">&lt;-</span> combined.TCR[<span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">4</span>)]</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a><span class="fu">head</span>(subset2[[<span class="dv">1</span>]])</span></code></pre></div>
<pre><code>##                    barcode sample                 TCR1         cdr3_aa1
## 1  P18B_AAACCTGAGGCTCAGA-1   P18B TRAV26-1.TRAJ37.TRAC  CIVRGGSSNTGKLIF
## 3  P18B_AAACCTGCATGACATC-1   P18B    TRAV3.TRAJ20.TRAC    CAVQRSNDYKLSF
## 5  P18B_AAACCTGGTATGCTTG-1   P18B TRAV26-1.TRAJ53.TRAC   CIGSSGGSNYKLTF
## 8  P18B_AAACGGGCAGATGGGT-1   P18B                 &lt;NA&gt;             &lt;NA&gt;
## 9  P18B_AAACGGGTCTTACCGC-1   P18B    TRAV20.TRAJ9.TRAC CAVQAKRYTGGFKTIF
## 12 P18B_AAAGATGAGTTACGGG-1   P18B   TRAV8-3.TRAJ8.TRAC   CAVGGDTGFQKLVF
##                                            cdr3_nt1
## 1     TGCATCGTCAGGGGCGGCTCTAGCAACACAGGCAAACTAATCTTT
## 3           TGTGCTGTGCAACGTTCTAACGACTACAAGCTCAGCTTT
## 5        TGCATCGGCTCAAGTGGAGGTAGCAACTATAAACTGACATTT
## 8                                              &lt;NA&gt;
## 9  TGTGCTGTGCAGGCCAAGCGGTATACTGGAGGCTTCAAAACTATCTTT
## 12       TGTGCTGTGGGTGGTGACACAGGCTTTCAGAAACTTGTATTT
##                                                     TCR2
## 1                             TRBV6-1.None.TRBJ2-3.TRBC2
## 3                             TRBV3-1.None.TRBJ2-3.TRBC2
## 5   TRBV4-1.None.TRBJ2-2.TRBC2;TRBV19.None.TRBJ1-5.TRBC1
## 8                             TRBV5-1.None.TRBJ1-2.TRBC1
## 9  TRBV5-1.None.TRBJ1-1.TRBC1;TRBV7-9.None.TRBJ2-2.TRBC2
## 12                           TRBV12-4.None.TRBJ1-1.TRBC1
##                          cdr3_aa2
## 1                 CASIGRSFGRDTQYF
## 3                CASSPPRGGFTDTQYF
## 5  CASSQGGQGGRELFF;CASSYAVGRQPQHF
## 8                  CASSLRETNYGYTF
## 9  CASSLGTGTGVEAFF;CAIDPGLLTGELFF
## 12                  CASRNSQATEAFF
##                                                                                    cdr3_nt2
## 1                                             TGTGCCAGTATCGGGAGGTCCTTTGGCCGAGATACGCAGTATTTT
## 3                                          TGTGCCAGCAGCCCCCCCCGCGGCGGATTCACAGATACGCAGTATTTT
## 5  TGCGCCAGCAGCCAAGGTGGACAGGGCGGAAGGGAGCTGTTTTTT;TGTGCCAGTAGCTACGCGGTGGGGAGGCAGCCCCAGCATTTT
## 8                                                TGCGCCAGCAGCTTGAGGGAAACCAACTATGGCTACACCTTC
## 9  TGCGCCAGCAGCTTGGGAACGGGGACAGGGGTTGAAGCTTTCTTT;TGTGCCATCGATCCGGGACTACTCACCGGGGAGCTGTTTTTT
## 12                                                  TGTGCCAGCAGAAACTCCCAAGCCACTGAAGCTTTCTTT
##                                                                       CTgene
## 1                            TRAV26-1.TRAJ37.TRAC_TRBV6-1.None.TRBJ2-3.TRBC2
## 3                               TRAV3.TRAJ20.TRAC_TRBV3-1.None.TRBJ2-3.TRBC2
## 5  TRAV26-1.TRAJ53.TRAC_TRBV4-1.None.TRBJ2-2.TRBC2;TRBV19.None.TRBJ1-5.TRBC1
## 8                                              NA_TRBV5-1.None.TRBJ1-2.TRBC1
## 9    TRAV20.TRAJ9.TRAC_TRBV5-1.None.TRBJ1-1.TRBC1;TRBV7-9.None.TRBJ2-2.TRBC2
## 12                            TRAV8-3.TRAJ8.TRAC_TRBV12-4.None.TRBJ1-1.TRBC1
##                                                                                                                                         CTnt
## 1                                                TGCATCGTCAGGGGCGGCTCTAGCAACACAGGCAAACTAATCTTT_TGTGCCAGTATCGGGAGGTCCTTTGGCCGAGATACGCAGTATTTT
## 3                                                   TGTGCTGTGCAACGTTCTAACGACTACAAGCTCAGCTTT_TGTGCCAGCAGCCCCCCCCGCGGCGGATTCACAGATACGCAGTATTTT
## 5        TGCATCGGCTCAAGTGGAGGTAGCAACTATAAACTGACATTT_TGCGCCAGCAGCCAAGGTGGACAGGGCGGAAGGGAGCTGTTTTTT;TGTGCCAGTAGCTACGCGGTGGGGAGGCAGCCCCAGCATTTT
## 8                                                                                              NA_TGCGCCAGCAGCTTGAGGGAAACCAACTATGGCTACACCTTC
## 9  TGTGCTGTGCAGGCCAAGCGGTATACTGGAGGCTTCAAAACTATCTTT_TGCGCCAGCAGCTTGGGAACGGGGACAGGGGTTGAAGCTTTCTTT;TGTGCCATCGATCCGGGACTACTCACCGGGGAGCTGTTTTTT
## 12                                                        TGTGCTGTGGGTGGTGACACAGGCTTTCAGAAACTTGTATTT_TGTGCCAGCAGAAACTCCCAAGCCACTGAAGCTTTCTTT
##                                               CTaa
## 1                  CIVRGGSSNTGKLIF_CASIGRSFGRDTQYF
## 3                   CAVQRSNDYKLSF_CASSPPRGGFTDTQYF
## 5    CIGSSGGSNYKLTF_CASSQGGQGGRELFF;CASSYAVGRQPQHF
## 8                                NA_CASSLRETNYGYTF
## 9  CAVQAKRYTGGFKTIF_CASSLGTGTGVEAFF;CAIDPGLLTGELFF
## 12                    CAVGGDTGFQKLVF_CASRNSQATEAFF
##                                                                                                                                                                                                             CTstrict
## 1                                                                        TRAV26-1.TRAJ37.TRAC;TGCATCGTCAGGGGCGGCTCTAGCAACACAGGCAAACTAATCTTT_TRBV6-1.None.TRBJ2-3.TRBC2;TGTGCCAGTATCGGGAGGTCCTTTGGCCGAGATACGCAGTATTTT
## 3                                                                              TRAV3.TRAJ20.TRAC;TGTGCTGTGCAACGTTCTAACGACTACAAGCTCAGCTTT_TRBV3-1.None.TRBJ2-3.TRBC2;TGTGCCAGCAGCCCCCCCCGCGGCGGATTCACAGATACGCAGTATTTT
## 5      TRAV26-1.TRAJ53.TRAC;TGCATCGGCTCAAGTGGAGGTAGCAACTATAAACTGACATTT_TRBV4-1.None.TRBJ2-2.TRBC2;TRBV19.None.TRBJ1-5.TRBC1;TGCGCCAGCAGCCAAGGTGGACAGGGCGGAAGGGAGCTGTTTTTT;TGTGCCAGTAGCTACGCGGTGGGGAGGCAGCCCCAGCATTTT
## 8                                                                                                                                        NA;NA_TRBV5-1.None.TRBJ1-2.TRBC1;TGCGCCAGCAGCTTGAGGGAAACCAACTATGGCTACACCTTC
## 9  TRAV20.TRAJ9.TRAC;TGTGCTGTGCAGGCCAAGCGGTATACTGGAGGCTTCAAAACTATCTTT_TRBV5-1.None.TRBJ1-1.TRBC1;TRBV7-9.None.TRBJ2-2.TRBC2;TGCGCCAGCAGCTTGGGAACGGGGACAGGGGTTGAAGCTTTCTTT;TGTGCCATCGATCCGGGACTACTCACCGGGGAGCTGTTTTTT
## 12                                                                                 TRAV8-3.TRAJ8.TRAC;TGTGCTGTGGGTGGTGACACAGGCTTTCAGAAACTTGTATTT_TRBV12-4.None.TRBJ1-1.TRBC1;TGTGCCAGCAGAAACTCCCAAGCCACTGAAGCTTTCTTT
##    Type
## 1     B
## 3     B
## 5     B
## 8     B
## 9     B
## 12    B</code></pre>
</div>
<div id="exportclones" class="section level2">
<h2>exportClones</h2>
<p>After assigning the clone by barcode, we can export the paired
clonotypes using <code>exportClones()</code> to save for later use or to
use in other pipelines.</p>
<p><strong>format</strong></p>
<ul>
<li>“paired” - Export the paired sequences
<strong>(default)</strong>.<br />
</li>
<li>“airr” - Export data in an AIRR-compliant format.<br />
</li>
<li>“TCRMatch” - Export TCRB chain information.</li>
</ul>
<p><strong>write.file</strong></p>
<ul>
<li>TRUE, save the file.<br />
</li>
<li>FALSE, return a data.frame.</li>
</ul>
<p><strong>dir</strong><br />
directory location to save the csv</p>
<p><strong>file.name</strong><br />
the csv file name</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="fu">exportClones</span>(combined, </span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a>             <span class="at">write.file =</span> <span class="cn">TRUE</span>,</span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a>             <span class="at">dir =</span> <span class="st">&quot;~/Documents/MyExperiment/Sample1/&quot;</span></span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a>             <span class="at">file.name =</span> <span class="st">&quot;clones.csv&quot;</span></span></code></pre></div>
<hr />
</div>
</div>
<div id="basic-clonal-visualizations" class="section level1">
<h1>Basic Clonal Visualizations</h1>
<p><strong>cloneCall</strong><br />
* “gene” - use the VDJC genes comprising the TCR/Ig<br />
* “nt” - use the nucleotide sequence of the CDR3 region<br />
* “aa” - use the amino acid sequence of the CDR3 region<br />
* “strict” - use the VDJC genes comprising the TCR + the nucleotide
sequence of the CDR3 region. This is the <a
href="https://www.ncbi.nlm.nih.gov/pubmed/19568742">proper definition of
clonotype</a>. For <code>combineBCR()</code> strict refers to the edit
distance clusters + Vgene of the Ig.</p>
<p>It is important to note that the clonotype is called using
essentially the combination of genes or nt/aa CDR3 sequences for both
loci. As of this implementation of scRepertoire, clonotype calling is
not incorporating small variations within the CDR3 sequences. As such
the <strong>gene</strong> approach will be the most sensitive, while the
use of <strong>nt</strong> or <strong>aa</strong> is moderately so, and
the most specific for clonotypes being <strong>strict</strong>.
Additionally, the clonotype call is trying to incorporate both loci,
<em>i.e.</em>, both <strong>TCRA</strong> and <strong>TCRB</strong>
chains and if a single cell barcode has multiple sequences identified
(<em>i.e.</em>, 2 TCRA chains expressed in one cell). Using the 10x
approach, there is a subset of barcodes that only return one of the
immune receptor chains. The unreturned chain is assigned an
<strong>NA</strong> value.</p>
<div id="clonalquant" class="section level2">
<h2>clonalQuant</h2>
<p>The first function to explore the clones is
<code>clonalQuant()</code> to return the total or relative numbers of
unique clones.</p>
<p><strong>scale</strong></p>
<ul>
<li>TRUE - relative percent of unique clones scaled by the total size of
the clonal repertoire.</li>
<li>FALSE - Report the total number of unique clones
<strong>(default)</strong>.</li>
</ul>
<p><strong>chain</strong></p>
<ul>
<li>“both” for combined chain visualization<br />
</li>
<li>“TRA”, “TRB”, “TRD”, “TRG”, “IGH” or “IGL” to select single
chain</li>
</ul>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="fu">clonalQuant</span>(combined.TCR, </span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>            <span class="at">cloneCall=</span><span class="st">&quot;strict&quot;</span>, </span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>            <span class="at">chain =</span> <span class="st">&quot;both&quot;</span>, </span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>            <span class="at">scale =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-13-1.png" width="700" /></p>
<p>Another option here is to be able to define the visualization by data
classes. Here, we used the <code>combineTCR()</code> list to define the
<strong>Type</strong> variable as part of the naming structure. We can
use the <strong>group.by</strong> to specifically use a column in the
data set to organize the visualization.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="fu">clonalQuant</span>(combined.TCR, </span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a>            <span class="at">cloneCall=</span><span class="st">&quot;gene&quot;</span>, </span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a>            <span class="at">group.by =</span> <span class="st">&quot;Type&quot;</span>, </span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a>            <span class="at">scale =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-14-1.png" width="700" /></p>
</div>
<div id="clonalabundance" class="section level2">
<h2>clonalAbundance</h2>
<p>We can also examine the relative distribution of clones by abundance.
Here <code>clonalAbundance()</code> will produce a line graph with a
total number of clones by the number of instances within the sample or
run. Like above, we can also group.by this by vectors within the contig
object using the <strong>group.by</strong> variable in the function.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="fu">clonalAbundance</span>(combined.TCR, </span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a>                <span class="at">cloneCall =</span> <span class="st">&quot;gene&quot;</span>, </span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a>                <span class="at">scale =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-15-1.png" width="700" /></p>
<p><code>clonalAbundance()</code> output can also be converted into a
density plot, which may allow for better comparisons between different
repertoire sizes, by setting <strong>scale</strong> = TRUE.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="fu">clonalAbundance</span>(combined.TCR, </span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>                <span class="at">cloneCall =</span> <span class="st">&quot;gene&quot;</span>, </span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a>                <span class="at">scale =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-16-1.png" width="700" /></p>
</div>
<div id="clonallength" class="section level2">
<h2>clonalLength</h2>
<p>We can look at the length distribution of the CDR3 sequences by
calling the <code>lengtheContig()</code> function. Importantly, unlike
the other basic visualizations, the <strong>cloneCall</strong> can only
be <strong>“nt”</strong> or <strong>“aa”</strong>. Due to the method of
calling clones as outlined above, the length should reveal a multimodal
curve, this is a product of using the <strong>NA</strong> for the
unreturned chain sequence and multiple chains within a single
barcode.</p>
<p><strong>chain</strong></p>
<ul>
<li>“both” for combined chain visualization<br />
</li>
<li>“TRA”, “TRB”, “TRD”, “TRG”, “IGH” or “IGL” to select single
chain</li>
</ul>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="fu">clonalLength</span>(combined.TCR, </span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a>             <span class="at">cloneCall=</span><span class="st">&quot;aa&quot;</span>, </span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a>             <span class="at">chain =</span> <span class="st">&quot;both&quot;</span>) </span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-17-1.png" width="700" /></p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="fu">clonalLength</span>(combined.TCR, </span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a>             <span class="at">cloneCall=</span><span class="st">&quot;aa&quot;</span>, </span>
<span id="cb24-3"><a href="#cb24-3" tabindex="-1"></a>             <span class="at">chain =</span> <span class="st">&quot;TRA&quot;</span>, </span>
<span id="cb24-4"><a href="#cb24-4" tabindex="-1"></a>             <span class="at">scale =</span> <span class="cn">TRUE</span>) </span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-17-2.png" width="700" /></p>
</div>
<div id="clonalcompare" class="section level2">
<h2>clonalCompare</h2>
<p>We can also look at clones between samples and changes in dynamics by
using the <code>clonalCompare()</code> function.</p>
<p><strong>samples</strong></p>
<ul>
<li>Can be used to isolate specific samples based on the name of the
list element</li>
</ul>
<p><strong>graph</strong></p>
<ul>
<li>“alluvial” - graph imaged below<br />
</li>
<li>“area” - graph by area of the respective clone</li>
</ul>
<p><strong>top.clones</strong></p>
<ul>
<li>The top number of clones to graph, this will be calculated based on
the frequency of the individual sample. This can also be left
blank.</li>
</ul>
<p><strong>clones</strong></p>
<ul>
<li>Can be used to isolate specific clone sequences, ensure the call
matches the sequences you would like to visualize.</li>
</ul>
<p><strong>highlight.clones</strong></p>
<ul>
<li>Specifically color certain clones, other clones will be greyed
out.</li>
</ul>
<p><strong>relabel.clones</strong></p>
<ul>
<li>Simplify the isolated clones to numerical designations to allow for
a tidier visualization</li>
</ul>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="fu">clonalCompare</span>(combined.TCR, </span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a>                  <span class="at">top.clones =</span> <span class="dv">10</span>, </span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a>                  <span class="at">samples =</span> <span class="fu">c</span>(<span class="st">&quot;P17B&quot;</span>, <span class="st">&quot;P17L&quot;</span>), </span>
<span id="cb25-4"><a href="#cb25-4" tabindex="-1"></a>                  <span class="at">cloneCall=</span><span class="st">&quot;aa&quot;</span>, </span>
<span id="cb25-5"><a href="#cb25-5" tabindex="-1"></a>                  <span class="at">graph =</span> <span class="st">&quot;alluvial&quot;</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-18-1.png" width="700" /></p>
<p>We can also choose to highlight specific clones, such as in the case
of <em>“CVVSDNTGGFKTIF_CASSVRRERANTGELFF”</em> and
<em>“NA_CASSVRRERANTGELFF”</em> using the
<strong>highlight.clones</strong> parameter. In addition, we can
simplify the plot to label the clones as clones 1:19.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="fu">clonalCompare</span>(combined.TCR, </span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a>              <span class="at">top.clones =</span> <span class="dv">10</span>,</span>
<span id="cb26-3"><a href="#cb26-3" tabindex="-1"></a>              <span class="at">highlight.clones =</span> <span class="fu">c</span>(<span class="st">&quot;CVVSDNTGGFKTIF_CASSVRRERANTGELFF&quot;</span>, <span class="st">&quot;NA_CASSVRRERANTGELFF&quot;</span>),</span>
<span id="cb26-4"><a href="#cb26-4" tabindex="-1"></a>              <span class="at">relabel.clones =</span> <span class="cn">TRUE</span>,</span>
<span id="cb26-5"><a href="#cb26-5" tabindex="-1"></a>              <span class="at">samples =</span> <span class="fu">c</span>(<span class="st">&quot;P17B&quot;</span>, <span class="st">&quot;P17L&quot;</span>), </span>
<span id="cb26-6"><a href="#cb26-6" tabindex="-1"></a>              <span class="at">cloneCall=</span><span class="st">&quot;aa&quot;</span>, </span>
<span id="cb26-7"><a href="#cb26-7" tabindex="-1"></a>              <span class="at">graph =</span> <span class="st">&quot;alluvial&quot;</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-19-1.png" width="700" /></p>
<p>Alternatively, if we only want to show specific clones, we can use
the <strong>clones</strong> parameter.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="fu">clonalCompare</span>(combined.TCR, </span>
<span id="cb27-2"><a href="#cb27-2" tabindex="-1"></a>              <span class="at">clones =</span> <span class="fu">c</span>(<span class="st">&quot;CVVSDNTGGFKTIF_CASSVRRERANTGELFF&quot;</span>, <span class="st">&quot;NA_CASSVRRERANTGELFF&quot;</span>),</span>
<span id="cb27-3"><a href="#cb27-3" tabindex="-1"></a>              <span class="at">relabel.clones =</span> <span class="cn">TRUE</span>,</span>
<span id="cb27-4"><a href="#cb27-4" tabindex="-1"></a>              <span class="at">samples =</span> <span class="fu">c</span>(<span class="st">&quot;P17B&quot;</span>, <span class="st">&quot;P17L&quot;</span>), </span>
<span id="cb27-5"><a href="#cb27-5" tabindex="-1"></a>              <span class="at">cloneCall=</span><span class="st">&quot;aa&quot;</span>, </span>
<span id="cb27-6"><a href="#cb27-6" tabindex="-1"></a>              <span class="at">graph =</span> <span class="st">&quot;alluvial&quot;</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-20-1.png" width="700" /></p>
</div>
<div id="clonalscatter" class="section level2">
<h2>clonalScatter</h2>
<p><code>clonalScatter()</code> will organize two repertoires, quantify
the relative clone sizes, and produce a scatter plot comparing the two
samples.</p>
<p><strong>x.axis</strong> and <strong>y.axis</strong></p>
<ul>
<li>Names of the list element to place on the x-axis and y-axis - so for
example <em>“P18B”</em> and <em>“P18L”</em></li>
</ul>
<p><strong>dot.size</strong></p>
<ul>
<li>“total” to display the total number of clones between the x- and
y-axis<br />
</li>
<li>Name of the list element to use for size calculation</li>
</ul>
<p><strong>graph</strong></p>
<ul>
<li>“proportion” for the relative proportion the clone represents across
all clones<br />
</li>
<li>“count” for the total count of clones by sample</li>
</ul>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a><span class="fu">clonalScatter</span>(combined.TCR, </span>
<span id="cb28-2"><a href="#cb28-2" tabindex="-1"></a>              <span class="at">cloneCall =</span><span class="st">&quot;gene&quot;</span>, </span>
<span id="cb28-3"><a href="#cb28-3" tabindex="-1"></a>              <span class="at">x.axis =</span> <span class="st">&quot;P18B&quot;</span>, </span>
<span id="cb28-4"><a href="#cb28-4" tabindex="-1"></a>              <span class="at">y.axis =</span> <span class="st">&quot;P18L&quot;</span>,</span>
<span id="cb28-5"><a href="#cb28-5" tabindex="-1"></a>              <span class="at">dot.size =</span> <span class="st">&quot;total&quot;</span>,</span>
<span id="cb28-6"><a href="#cb28-6" tabindex="-1"></a>              <span class="at">graph =</span> <span class="st">&quot;proportion&quot;</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-21-1.png" width="700" /></p>
<hr />
</div>
</div>
<div id="visualizing-clonal-dynamics" class="section level1">
<h1>Visualizing Clonal Dynamics</h1>
<div id="clonalhomeostasis" class="section level2">
<h2>clonalHomeostasis</h2>
<p>By examining the clonal space, we effectively look at the relative
space occupied by clones at specific proportions. Another way to think
about this would be to think of the total immune receptor sequencing run
as a measuring cup. In this cup, we will fill liquids of different
viscosity - or different numbers of clonal proportions. Clonal space
homeostasis asks what percentage of the cup is filled by clones in
distinct proportions (or liquids of different viscosity, to extend the
analogy). The proportional cut points are set under the
<strong>cloneSize</strong> variable in the function and can be
adjusted.</p>
<p><strong>cloneSize</strong></p>
<ul>
<li>Rare = 0.0001<br />
</li>
<li>Small = 0.001<br />
</li>
<li>Medium = 0.01<br />
</li>
<li>Large = 0.1<br />
</li>
<li>Hyperexpanded = 1</li>
</ul>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a><span class="fu">clonalHomeostasis</span>(combined.TCR, </span>
<span id="cb29-2"><a href="#cb29-2" tabindex="-1"></a>                  <span class="at">cloneCall =</span> <span class="st">&quot;gene&quot;</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-22-1.png" width="700" /></p>
<p>We can reassign the proportional cut points for cloneSize, which can
drastically alter the visualization and analysis.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="fu">clonalHomeostasis</span>(combined.TCR, </span>
<span id="cb30-2"><a href="#cb30-2" tabindex="-1"></a>                  <span class="at">cloneCall =</span> <span class="st">&quot;gene&quot;</span>,</span>
<span id="cb30-3"><a href="#cb30-3" tabindex="-1"></a>                  <span class="at">cloneSize =</span> <span class="fu">c</span>(<span class="at">Rare =</span> <span class="fl">0.001</span>, <span class="at">Small =</span> <span class="fl">0.01</span>, <span class="at">Medium =</span> <span class="fl">0.1</span>, <span class="at">Large =</span> <span class="fl">0.3</span>, <span class="at">Hyperexpanded =</span></span>
<span id="cb30-4"><a href="#cb30-4" tabindex="-1"></a>    <span class="dv">1</span>))</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-23-1.png" width="700" /></p>
<p>In addition, we can use the <strong>group.by</strong> parameter to
look at the relative proportion of clones between groups - such as
tissue.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a>combined.TCR <span class="ot">&lt;-</span> <span class="fu">addVariable</span>(combined.TCR, </span>
<span id="cb31-2"><a href="#cb31-2" tabindex="-1"></a>                            <span class="at">variable.name =</span> <span class="st">&quot;Type&quot;</span>, </span>
<span id="cb31-3"><a href="#cb31-3" tabindex="-1"></a>                            <span class="at">variables =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">&quot;B&quot;</span>, <span class="st">&quot;L&quot;</span>), <span class="dv">4</span>))</span>
<span id="cb31-4"><a href="#cb31-4" tabindex="-1"></a></span>
<span id="cb31-5"><a href="#cb31-5" tabindex="-1"></a><span class="fu">clonalHomeostasis</span>(combined.TCR, </span>
<span id="cb31-6"><a href="#cb31-6" tabindex="-1"></a>                  <span class="at">group.by =</span> <span class="st">&quot;Type&quot;</span>,</span>
<span id="cb31-7"><a href="#cb31-7" tabindex="-1"></a>                  <span class="at">cloneCall =</span> <span class="st">&quot;gene&quot;</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-24-1.png" width="700" /></p>
</div>
<div id="clonalproportion" class="section level2">
<h2>clonalProportion</h2>
<p>Like clonal space homeostasis above, clonal proportion places clones
into separate bins. The key difference is that instead of looking at the
relative proportion of the clone to the total, the
<code>clonalProportion()</code> function will rank the clones by total
number and place them into bins.</p>
<p>The <strong>clonalSplit</strong> represents the ranking of clonotypes
by copy or frequency of occurrence, meaning 1:10 are the top 10
clonotypes in each sample. The default bins are under the
<strong>clonalSplit</strong> variable in the function and can be
adjusted, but they are as follows at baseline.</p>
<p><strong>clonalSplit</strong></p>
<ul>
<li>10<br />
</li>
<li>100<br />
</li>
<li>1000<br />
</li>
<li>10000</li>
<li>30000</li>
<li>100000</li>
</ul>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="fu">clonalProportion</span>(combined.TCR, </span>
<span id="cb32-2"><a href="#cb32-2" tabindex="-1"></a>                 <span class="at">cloneCall =</span> <span class="st">&quot;gene&quot;</span>) </span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-25-1.png" width="700" /></p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" tabindex="-1"></a><span class="fu">clonalProportion</span>(combined.TCR, </span>
<span id="cb33-2"><a href="#cb33-2" tabindex="-1"></a>                 <span class="at">cloneCall =</span> <span class="st">&quot;nt&quot;</span>,</span>
<span id="cb33-3"><a href="#cb33-3" tabindex="-1"></a>                 <span class="at">clonalSplit =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">10</span>, <span class="dv">100</span>, <span class="dv">1000</span>, <span class="dv">10000</span>)) </span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-25-2.png" width="700" /></p>
<hr />
</div>
</div>
<div id="summarizing-repertoires" class="section level1">
<h1>Summarizing Repertoires</h1>
<div id="vizgenes" class="section level2">
<h2>vizGenes</h2>
<p>A visualization of the relative usage of genes of the TCR or BCR,
using <code>vizGenes()</code>. There is some functional crossover
between <code>vizGenes()</code> and two functions below called
<code>percentGenes()</code> and <code>percentVJ()</code>. But
<code>vizGenes()</code> is more adaptable to allow for comparisons
across chains, scaling, etc.</p>
<p><strong>x.axis</strong></p>
<ul>
<li>Gene segment to separate the x-axis, such as “TRAV”, “TRBD”,
“IGKJ”</li>
</ul>
<p><strong>y.axis</strong></p>
<ul>
<li>Gene segment or categorical variable to visualize along the
y-axis</li>
</ul>
<p><strong>plot</strong></p>
<ul>
<li>“barplot” for a bar chart<br />
</li>
<li>“heatmap” for a heatmap</li>
</ul>
<p><strong>scale</strong></p>
<ul>
<li>TRUE to scale the graph by number of genes per sample<br />
</li>
<li>FALSE to report raw numbers</li>
</ul>
<p><strong>order</strong></p>
<ul>
<li>“gene” to order by gene name<br />
</li>
<li>“variance” to order by variance between the separate variable
categories</li>
</ul>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" tabindex="-1"></a><span class="fu">vizGenes</span>(combined.TCR, </span>
<span id="cb34-2"><a href="#cb34-2" tabindex="-1"></a>         <span class="at">x.axis =</span> <span class="st">&quot;TRBV&quot;</span>,</span>
<span id="cb34-3"><a href="#cb34-3" tabindex="-1"></a>         <span class="at">y.axis =</span> <span class="cn">NULL</span>,</span>
<span id="cb34-4"><a href="#cb34-4" tabindex="-1"></a>         <span class="at">plot =</span> <span class="st">&quot;barplot&quot;</span>,  </span>
<span id="cb34-5"><a href="#cb34-5" tabindex="-1"></a>         <span class="at">scale =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-26-1.png" width="700" /></p>
<p><code>vizGenes()</code> can also be used to look at the usage of
genes in a single chain. For example, say we are interested in the
difference in TRB V and J usage between lung and peripheral blood
samples - we can easily take a look at this using the following
code:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" tabindex="-1"></a><span class="co">#Peripheral Blood</span></span>
<span id="cb35-2"><a href="#cb35-2" tabindex="-1"></a><span class="fu">vizGenes</span>(combined.TCR[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">5</span>,<span class="dv">7</span>)], </span>
<span id="cb35-3"><a href="#cb35-3" tabindex="-1"></a>         <span class="at">x.axis =</span> <span class="st">&quot;TRBV&quot;</span>,</span>
<span id="cb35-4"><a href="#cb35-4" tabindex="-1"></a>         <span class="at">y.axis =</span> <span class="st">&quot;TRBJ&quot;</span>,</span>
<span id="cb35-5"><a href="#cb35-5" tabindex="-1"></a>         <span class="at">plot =</span> <span class="st">&quot;heatmap&quot;</span>,  </span>
<span id="cb35-6"><a href="#cb35-6" tabindex="-1"></a>         <span class="at">scale =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-27-1.png" width="700" /></p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" tabindex="-1"></a><span class="co">#Lung</span></span>
<span id="cb36-2"><a href="#cb36-2" tabindex="-1"></a><span class="fu">vizGenes</span>(combined.TCR[<span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">6</span>,<span class="dv">8</span>)], </span>
<span id="cb36-3"><a href="#cb36-3" tabindex="-1"></a>         <span class="at">x.axis =</span> <span class="st">&quot;TRBV&quot;</span>,</span>
<span id="cb36-4"><a href="#cb36-4" tabindex="-1"></a>         <span class="at">y.axis =</span> <span class="st">&quot;TRBJ&quot;</span>,</span>
<span id="cb36-5"><a href="#cb36-5" tabindex="-1"></a>         <span class="at">plot =</span> <span class="st">&quot;heatmap&quot;</span>,  </span>
<span id="cb36-6"><a href="#cb36-6" tabindex="-1"></a>         <span class="at">scale =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-27-2.png" width="700" /></p>
<p>For the P17 patient samples, what if we are interested in chain
pairings, we can look at TRBV and TRAV at the same time using them as
inputs to <strong>x.axis</strong> and <strong>y.axis</strong>.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a><span class="fu">vizGenes</span>(combined.TCR[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)], </span>
<span id="cb37-2"><a href="#cb37-2" tabindex="-1"></a>         <span class="at">x.axis =</span> <span class="st">&quot;TRBV&quot;</span>,</span>
<span id="cb37-3"><a href="#cb37-3" tabindex="-1"></a>         <span class="at">y.axis =</span> <span class="st">&quot;TRAV&quot;</span>,</span>
<span id="cb37-4"><a href="#cb37-4" tabindex="-1"></a>         <span class="at">plot =</span> <span class="st">&quot;heatmap&quot;</span>, </span>
<span id="cb37-5"><a href="#cb37-5" tabindex="-1"></a>         <span class="at">scale =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-28-1.png" width="700" /></p>
</div>
<div id="percentaa" class="section level2">
<h2>percentAA</h2>
<p>Quantify the proportion of amino acids along the cdr3 sequence with
<code>percentAA()</code>. By default, the function will pad the
sequences with NAs up to the maximum of <strong>aa.length</strong>.
Sequences longer than <strong>aa.length</strong> will be removed before
visualization <strong>(default aa.length = 20)</strong>.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" tabindex="-1"></a><span class="fu">percentAA</span>(combined.TCR, </span>
<span id="cb38-2"><a href="#cb38-2" tabindex="-1"></a>          <span class="at">chain =</span> <span class="st">&quot;TRB&quot;</span>, </span>
<span id="cb38-3"><a href="#cb38-3" tabindex="-1"></a>          <span class="at">aa.length =</span> <span class="dv">20</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-29-1.png" width="700" /></p>
</div>
<div id="positionalentropy" class="section level2">
<h2>positionalEntropy</h2>
<p>We can also quantify the level of entropy/diversity across amino acid
residues along the cdr3 sequence. <code>positionalEntropy()</code>
combines the quantification by residue of <code>percentAA()</code> with
the diversity calls in <code>clonalDiversity()</code>. Importantly,
because diversity can be affected by size of sample, similar to
<code>clonalDiversity()</code>, <code>positionalEntropy()</code> will
also downsample/bootstrap the calculation. Positions without variance
will have a value reported as 0 for the purposes of comparison.</p>
<p><strong>method</strong></p>
<ul>
<li>“shannon” - Shannon Diversity<br />
</li>
<li>“inv.simpson” - Inverse Simpson Diversity</li>
<li>“norm.entropy” - Normalized Entropy</li>
</ul>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" tabindex="-1"></a><span class="fu">positionalEntropy</span>(combined.TCR, </span>
<span id="cb39-2"><a href="#cb39-2" tabindex="-1"></a>                  <span class="at">chain =</span> <span class="st">&quot;TRB&quot;</span>, </span>
<span id="cb39-3"><a href="#cb39-3" tabindex="-1"></a>                  <span class="at">aa.length =</span> <span class="dv">20</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-30-1.png" width="700" /></p>
</div>
<div id="percentgenes" class="section level2">
<h2>percentGenes</h2>
<p>Quantify the proportion of V or J gene usage with
<code>percentGenes()</code>. Like <code>percentAA()</code>, we select
the chain of interest and then indicate the gene of interest with the
<strong>gene</strong> parameter. Two major limitations of
<code>percentGenes()</code> are, 1) the function quantifies only V or J
genes, and 2) the quantification of the genes are limited to all the V
or J genes seen across the samples, not all possible V or J genes.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" tabindex="-1"></a><span class="fu">percentGenes</span>(combined.TCR, </span>
<span id="cb40-2"><a href="#cb40-2" tabindex="-1"></a>             <span class="at">chain =</span> <span class="st">&quot;TRB&quot;</span>, </span>
<span id="cb40-3"><a href="#cb40-3" tabindex="-1"></a>             <span class="at">gene =</span> <span class="st">&quot;Vgene&quot;</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-31-1.png" width="700" /></p>
<p>We can also use the output <code>percentGenes()</code> for
dimensional reduction to summarise the gene usage by sample. This can be
done with a simple principal component analysis (below) or even more
complex reductions.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" tabindex="-1"></a>df.genes <span class="ot">&lt;-</span> <span class="fu">percentGenes</span>(combined.TCR, </span>
<span id="cb41-2"><a href="#cb41-2" tabindex="-1"></a>                         <span class="at">chain =</span> <span class="st">&quot;TRB&quot;</span>, </span>
<span id="cb41-3"><a href="#cb41-3" tabindex="-1"></a>                         <span class="at">gene =</span> <span class="st">&quot;Vgene&quot;</span>, </span>
<span id="cb41-4"><a href="#cb41-4" tabindex="-1"></a>                         <span class="at">exportTable =</span> <span class="cn">TRUE</span>)</span>
<span id="cb41-5"><a href="#cb41-5" tabindex="-1"></a></span>
<span id="cb41-6"><a href="#cb41-6" tabindex="-1"></a><span class="co">#Performing PCA</span></span>
<span id="cb41-7"><a href="#cb41-7" tabindex="-1"></a>pc <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(df.genes)</span>
<span id="cb41-8"><a href="#cb41-8" tabindex="-1"></a></span>
<span id="cb41-9"><a href="#cb41-9" tabindex="-1"></a><span class="co">#Getting data frame to plot from</span></span>
<span id="cb41-10"><a href="#cb41-10" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">cbind</span>(pc<span class="sc">$</span>x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="fu">rownames</span>(df.genes)))</span>
<span id="cb41-11"><a href="#cb41-11" tabindex="-1"></a>df<span class="sc">$</span>PC1 <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(df<span class="sc">$</span>PC1)</span>
<span id="cb41-12"><a href="#cb41-12" tabindex="-1"></a>df<span class="sc">$</span>PC2 <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(df<span class="sc">$</span>PC2)</span>
<span id="cb41-13"><a href="#cb41-13" tabindex="-1"></a></span>
<span id="cb41-14"><a href="#cb41-14" tabindex="-1"></a><span class="co">#Plotting</span></span>
<span id="cb41-15"><a href="#cb41-15" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2)) <span class="sc">+</span> </span>
<span id="cb41-16"><a href="#cb41-16" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">fill =</span>df[,<span class="dv">3</span>]), <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">size =</span> <span class="dv">5</span>) <span class="sc">+</span> </span>
<span id="cb41-17"><a href="#cb41-17" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">fill=</span><span class="fu">guide_legend</span>(<span class="at">title=</span><span class="st">&quot;Samples&quot;</span>)) <span class="sc">+</span> </span>
<span id="cb41-18"><a href="#cb41-18" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">hcl.colors</span>(<span class="fu">nrow</span>(df), <span class="st">&quot;inferno&quot;</span>)) <span class="sc">+</span> </span>
<span id="cb41-19"><a href="#cb41-19" tabindex="-1"></a>  <span class="fu">theme_classic</span>() </span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-32-1.png" width="700" /></p>
</div>
<div id="percentvj" class="section level2">
<h2>percentVJ</h2>
<p>Quantify the proportion of V and J gene usage with
<code>percentVJ()</code>. Like <code>percentGenes()</code>, this
function will quantify the percentage of V and J paired together across
individual repertoires. The output can be visualized using a heatmap or
as input for further dimensional reduction.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" tabindex="-1"></a><span class="fu">percentVJ</span>(combined.TCR, </span>
<span id="cb42-2"><a href="#cb42-2" tabindex="-1"></a>          <span class="at">chain =</span> <span class="st">&quot;TRB&quot;</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-33-1.png" width="700" /></p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" tabindex="-1"></a>df.genes <span class="ot">&lt;-</span> <span class="fu">percentVJ</span>(combined.TCR, </span>
<span id="cb43-2"><a href="#cb43-2" tabindex="-1"></a>                      <span class="at">chain =</span> <span class="st">&quot;TRB&quot;</span>, </span>
<span id="cb43-3"><a href="#cb43-3" tabindex="-1"></a>                      <span class="at">exportTable =</span> <span class="cn">TRUE</span>)</span>
<span id="cb43-4"><a href="#cb43-4" tabindex="-1"></a></span>
<span id="cb43-5"><a href="#cb43-5" tabindex="-1"></a><span class="co">#Performing PCA</span></span>
<span id="cb43-6"><a href="#cb43-6" tabindex="-1"></a>pc <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(df.genes)</span>
<span id="cb43-7"><a href="#cb43-7" tabindex="-1"></a></span>
<span id="cb43-8"><a href="#cb43-8" tabindex="-1"></a><span class="co">#Getting data frame to plot from</span></span>
<span id="cb43-9"><a href="#cb43-9" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">cbind</span>(pc<span class="sc">$</span>x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="fu">rownames</span>(df.genes))) </span>
<span id="cb43-10"><a href="#cb43-10" tabindex="-1"></a>df<span class="sc">$</span>PC1 <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(df<span class="sc">$</span>PC1)</span>
<span id="cb43-11"><a href="#cb43-11" tabindex="-1"></a>df<span class="sc">$</span>PC2 <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(df<span class="sc">$</span>PC2)</span>
<span id="cb43-12"><a href="#cb43-12" tabindex="-1"></a></span>
<span id="cb43-13"><a href="#cb43-13" tabindex="-1"></a><span class="co">#Plotting</span></span>
<span id="cb43-14"><a href="#cb43-14" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2)) <span class="sc">+</span> </span>
<span id="cb43-15"><a href="#cb43-15" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">fill =</span>df[,<span class="dv">3</span>]), <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">size =</span> <span class="dv">5</span>) <span class="sc">+</span> </span>
<span id="cb43-16"><a href="#cb43-16" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">fill=</span><span class="fu">guide_legend</span>(<span class="at">title=</span><span class="st">&quot;Samples&quot;</span>)) <span class="sc">+</span> </span>
<span id="cb43-17"><a href="#cb43-17" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">hcl.colors</span>(<span class="fu">nrow</span>(df), <span class="st">&quot;inferno&quot;</span>)) <span class="sc">+</span> </span>
<span id="cb43-18"><a href="#cb43-18" tabindex="-1"></a>  <span class="fu">theme_classic</span>() </span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-33-2.png" width="700" /></p>
</div>
<div id="percentkmer" class="section level2">
<h2>percentKmer</h2>
<p>Another quantification of the composition of the CDR3 sequence is to
define motifs by sliding across the amino acid or nucleotide sequences
at set intervals resulting in substrings or kmers.</p>
<p><strong>motif.length</strong></p>
<ul>
<li>Numerical value for the length of the kmer.</li>
</ul>
<p><strong>top.motifs</strong></p>
<ul>
<li>Display the most variable genes determined by mean absolute
deviation (MAD).</li>
</ul>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" tabindex="-1"></a><span class="fu">percentKmer</span>(combined.TCR, </span>
<span id="cb44-2"><a href="#cb44-2" tabindex="-1"></a>            <span class="at">cloneCall =</span> <span class="st">&quot;aa&quot;</span>,</span>
<span id="cb44-3"><a href="#cb44-3" tabindex="-1"></a>            <span class="at">chain =</span> <span class="st">&quot;TRB&quot;</span>, </span>
<span id="cb44-4"><a href="#cb44-4" tabindex="-1"></a>            <span class="at">motif.length =</span> <span class="dv">3</span>, </span>
<span id="cb44-5"><a href="#cb44-5" tabindex="-1"></a>            <span class="at">top.motifs =</span> <span class="dv">25</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-34-1.png" width="700" /></p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" tabindex="-1"></a><span class="fu">percentKmer</span>(combined.TCR, </span>
<span id="cb45-2"><a href="#cb45-2" tabindex="-1"></a>            <span class="at">cloneCall =</span> <span class="st">&quot;nt&quot;</span>,</span>
<span id="cb45-3"><a href="#cb45-3" tabindex="-1"></a>            <span class="at">chain =</span> <span class="st">&quot;TRB&quot;</span>, </span>
<span id="cb45-4"><a href="#cb45-4" tabindex="-1"></a>            <span class="at">motif.length =</span> <span class="dv">3</span>, </span>
<span id="cb45-5"><a href="#cb45-5" tabindex="-1"></a>            <span class="at">top.motifs =</span> <span class="dv">25</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-34-2.png" width="700" /></p>
<hr />
</div>
</div>
<div id="comparing-clonal-diversity-and-overlap" class="section level1">
<h1>Comparing Clonal Diversity and Overlap</h1>
<div id="clonaldiversity" class="section level2">
<h2>clonalDiversity</h2>
<p>Diversity can also be measured for samples or by other variables.
Diversity <strong>metrics</strong> calculated, include:
<strong>“shannon”</strong>, <strong>“inv.simpson”</strong>,
<strong>“norm.entropy”</strong>, <strong>“gini.simpson”</strong>,
<strong>“chao1”</strong>, and <strong>“ACE”</strong>. Please see the
manual for more information on each metric and the underlying
calculations.</p>
<p>Inherent in diversity calculations is a bias for increasing diversity
with increasing repertoire size. <code>clonalDiversity()</code> will
automatically downsample to the smallest repertoire size and perform
bootstrapping to return the mean diversity estimates. If the output of
diversity values are strange or minimally variable, it is likely due to
a sample with small repertoire size.</p>
<p><strong>n.boots</strong><br />
The number of calculations to perform <strong>(default =
100)</strong>.</p>
<p><strong>return.boots</strong></p>
<ul>
<li>TRUE: Return all the calculations.<br />
</li>
<li>FALSE: Return only the mean values <strong>(default)</strong>.</li>
</ul>
<p><strong>skip.boots</strong><br />
Skip the bootstrapping calculations.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" tabindex="-1"></a><span class="fu">clonalDiversity</span>(combined.TCR, </span>
<span id="cb46-2"><a href="#cb46-2" tabindex="-1"></a>                <span class="at">cloneCall =</span> <span class="st">&quot;gene&quot;</span>, </span>
<span id="cb46-3"><a href="#cb46-3" tabindex="-1"></a>                <span class="at">n.boots =</span> <span class="dv">20</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-35-1.png" width="700" /></p>
<p>As a default, <code>clonalDiversity()</code> will return all the
metrics calculated - <strong>“shannon”</strong>,
<strong>“inv.simpson”</strong>, <strong>“norm.entropy”</strong>,
<strong>“gini.simpson”</strong>, <strong>“chao1”</strong>, and
<strong>“ACE”</strong>. Selecting a single or a subset of these methods
using the <strong>metrics</strong> parameter.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" tabindex="-1"></a><span class="co">#Return only a subset of metrics</span></span>
<span id="cb47-2"><a href="#cb47-2" tabindex="-1"></a><span class="fu">clonalDiversity</span>(combined.TCR, </span>
<span id="cb47-3"><a href="#cb47-3" tabindex="-1"></a>                <span class="at">metrics =</span> <span class="fu">c</span>(<span class="st">&quot;shannon&quot;</span>, <span class="st">&quot;ACE&quot;</span>),</span>
<span id="cb47-4"><a href="#cb47-4" tabindex="-1"></a>                <span class="at">cloneCall =</span> <span class="st">&quot;gene&quot;</span>, </span>
<span id="cb47-5"><a href="#cb47-5" tabindex="-1"></a>                <span class="at">n.boots =</span> <span class="dv">20</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-36-1.png" width="700" /></p>
</div>
<div id="clonararefaction" class="section level2">
<h2>clonaRarefaction</h2>
<p>We can also use Hill numbers to estimate the rarefaction, or
estimating species richness, using the the abundance of clones across
groupings. Underlying the rarefaction calculation is the use of observed
receptor of abundance to compute diversity.</p>
<p><strong>hill.numbers</strong></p>
<ul>
<li>0 - species-richness<br />
</li>
<li>1 - Shannon<br />
</li>
<li>2 - Simpson</li>
</ul>
<p><strong>plot.type</strong></p>
<ul>
<li>1 - sample-size-based rarefaction/extrapolation<br />
</li>
<li>2 - sample completeness curve<br />
</li>
<li>3 - coverage-based rarefaction/extrapolation curve</li>
</ul>
<p>This function relies on the <a
href="https://cran.r-project.org/web/packages/iNEXT/index.html">iNEXT</a>
with the accompanying <a
href="http://chao.stat.nthu.edu.tw/wordpress/paper/120_pdf_appendix.pdf">manuscript</a>.
Like the other wrapping functions in scRepertoire, please cite the
original work. The sample completeness curve (<strong>plot.type</strong>
= 2), may not show full sample coverage due to the size/diversity of the
input data.</p>
<div id="rarefaction-using-species-richness-q-0" class="section level3">
<h3>Rarefaction using Species Richness (q = 0)</h3>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" tabindex="-1"></a><span class="fu">clonalRarefaction</span>(combined.TCR,</span>
<span id="cb48-2"><a href="#cb48-2" tabindex="-1"></a>                  <span class="at">plot.type =</span> <span class="dv">1</span>,</span>
<span id="cb48-3"><a href="#cb48-3" tabindex="-1"></a>                  <span class="at">hill.numbers =</span> <span class="dv">0</span>,</span>
<span id="cb48-4"><a href="#cb48-4" tabindex="-1"></a>                  <span class="at">n.boots =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-37-1.png" width="700" /></p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" tabindex="-1"></a><span class="fu">clonalRarefaction</span>(combined.TCR,</span>
<span id="cb49-2"><a href="#cb49-2" tabindex="-1"></a>                  <span class="at">plot.type =</span> <span class="dv">2</span>,</span>
<span id="cb49-3"><a href="#cb49-3" tabindex="-1"></a>                  <span class="at">hill.numbers =</span> <span class="dv">0</span>,</span>
<span id="cb49-4"><a href="#cb49-4" tabindex="-1"></a>                  <span class="at">n.boots =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-37-2.png" width="700" /></p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" tabindex="-1"></a><span class="fu">clonalRarefaction</span>(combined.TCR,</span>
<span id="cb50-2"><a href="#cb50-2" tabindex="-1"></a>                  <span class="at">plot.type =</span> <span class="dv">3</span>,</span>
<span id="cb50-3"><a href="#cb50-3" tabindex="-1"></a>                  <span class="at">hill.numbers =</span> <span class="dv">0</span>,</span>
<span id="cb50-4"><a href="#cb50-4" tabindex="-1"></a>                  <span class="at">n.boots =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-37-3.png" width="700" /></p>
</div>
<div id="rarefaction-using-shannon-diversity-q-1"
class="section level3">
<h3>Rarefaction using Shannon Diversity (q = 1)</h3>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" tabindex="-1"></a><span class="fu">clonalRarefaction</span>(combined.TCR,</span>
<span id="cb51-2"><a href="#cb51-2" tabindex="-1"></a>                  <span class="at">plot.type =</span> <span class="dv">1</span>,</span>
<span id="cb51-3"><a href="#cb51-3" tabindex="-1"></a>                  <span class="at">hill.numbers =</span> <span class="dv">1</span>,</span>
<span id="cb51-4"><a href="#cb51-4" tabindex="-1"></a>                  <span class="at">n.boots =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-38-1.png" width="700" /></p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" tabindex="-1"></a><span class="fu">clonalRarefaction</span>(combined.TCR,</span>
<span id="cb52-2"><a href="#cb52-2" tabindex="-1"></a>                  <span class="at">plot.type =</span> <span class="dv">2</span>,</span>
<span id="cb52-3"><a href="#cb52-3" tabindex="-1"></a>                  <span class="at">hill.numbers =</span> <span class="dv">1</span>,</span>
<span id="cb52-4"><a href="#cb52-4" tabindex="-1"></a>                  <span class="at">n.boots =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-38-2.png" width="700" /></p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" tabindex="-1"></a><span class="fu">clonalRarefaction</span>(combined.TCR,</span>
<span id="cb53-2"><a href="#cb53-2" tabindex="-1"></a>                  <span class="at">plot.type =</span> <span class="dv">3</span>,</span>
<span id="cb53-3"><a href="#cb53-3" tabindex="-1"></a>                  <span class="at">hill.numbers =</span> <span class="dv">1</span>,</span>
<span id="cb53-4"><a href="#cb53-4" tabindex="-1"></a>                  <span class="at">n.boots =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-38-3.png" width="700" /></p>
</div>
</div>
<div id="clonalsizedistribution" class="section level2">
<h2>clonalSizeDistribution</h2>
<p>Another method for modeling the repertoire distribution is a discrete
gamma-GPD spliced threshold model, proposed by <a
href="https://pubmed.ncbi.nlm.nih.gov/30485278/">Koch et al.</a> The
spliced model models the repertoire and allows for the application of a
power law distribution for larger clonal-expanded sequences and a
Poisson distribution for smaller clones. After fitting the models,
repertoires can be compared using Euclidean distance.</p>
<p>If using this function, please read/cite <a
href="https://pubmed.ncbi.nlm.nih.gov/30485278/">Koch et al.</a> and
check out the <a
href="https://bioconductor.org/packages/release/bioc/html/powerTCR.html">powerTCR</a>
R package.</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" tabindex="-1"></a><span class="fu">clonalSizeDistribution</span>(combined.TCR, </span>
<span id="cb54-2"><a href="#cb54-2" tabindex="-1"></a>                       <span class="at">cloneCall =</span> <span class="st">&quot;aa&quot;</span>, </span>
<span id="cb54-3"><a href="#cb54-3" tabindex="-1"></a>                       <span class="at">method=</span> <span class="st">&quot;ward.D2&quot;</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-39-1.png" width="700" /></p>
</div>
<div id="clonaloverlap" class="section level2">
<h2>clonalOverlap</h2>
<p>If you are interested in measures of similarity between the samples
loaded into scRepertoire, using <code>clonalOverlap()</code> can assist
in the visualization.</p>
<p>The underlying <code>clonalOverlap()</code> calculation varies by the
<strong>method</strong> parameter, more information on the exact
calculations are available in the manual.</p>
<p><strong>method</strong></p>
<ul>
<li>“overlap” - overlap coefficient</li>
<li>“morisita” - Morisita’s overlap index</li>
<li>“jaccard” - Jaccard index</li>
<li>“cosine” - cosine similarity</li>
<li>“raw” - exact number of overlapping clones</li>
</ul>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" tabindex="-1"></a><span class="fu">clonalOverlap</span>(combined.TCR, </span>
<span id="cb55-2"><a href="#cb55-2" tabindex="-1"></a>              <span class="at">cloneCall =</span> <span class="st">&quot;strict&quot;</span>, </span>
<span id="cb55-3"><a href="#cb55-3" tabindex="-1"></a>              <span class="at">method =</span> <span class="st">&quot;morisita&quot;</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-40-1.png" width="700" /></p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" tabindex="-1"></a><span class="fu">clonalOverlap</span>(combined.TCR, </span>
<span id="cb56-2"><a href="#cb56-2" tabindex="-1"></a>              <span class="at">cloneCall =</span> <span class="st">&quot;strict&quot;</span>, </span>
<span id="cb56-3"><a href="#cb56-3" tabindex="-1"></a>              <span class="at">method =</span> <span class="st">&quot;raw&quot;</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-40-2.png" width="700" /></p>
<hr />
</div>
</div>
<div id="combining-clones-and-single-cell-objects"
class="section level1">
<h1>Combining Clones and Single-Cell Objects</h1>
<p>The data in the scRepertoire package is derived from a <a
href="https://pubmed.ncbi.nlm.nih.gov/33622974/">study</a> of acute
respiratory stress disorder in the context of bacterial and COVID-19
infections. The internal single cell data (<code>scRep_example()</code>)
built in to scRepertoire is randomly sampled 500 cells from the fully
integrated Seurat object to minimize the package size. We will use both
Seurat and Single-Cell Experiment (SCE) with scater to perform further
visualizations in tandem.</p>
<div id="preprocessed-single-cell-object" class="section level2">
<h2>Preprocessed Single-Cell Object</h2>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" tabindex="-1"></a>scRep_example <span class="ot">&lt;-</span> <span class="fu">get</span>(<span class="fu">data</span>(<span class="st">&quot;scRep_example&quot;</span>))</span>
<span id="cb57-2"><a href="#cb57-2" tabindex="-1"></a></span>
<span id="cb57-3"><a href="#cb57-3" tabindex="-1"></a><span class="co">#Making a Single-Cell Experiment object</span></span>
<span id="cb57-4"><a href="#cb57-4" tabindex="-1"></a>sce <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">as.SingleCellExperiment</span>(scRep_example)</span></code></pre></div>
</div>
<div id="combineexpression" class="section level2">
<h2>combineExpression</h2>
<p>After processing the contig data into clones via
<code>combineBCR()</code> or <code>combineTCR()</code>, we can add the
clonal information to the single-cell object using
<code>combineExpression()</code>.</p>
<p><strong>Importantly</strong>, the major requirement for the
attachment is matching contig cell barcodes and barcodes in the row
names of the meta data of the Seurat or Single-Cell Experiment object.
If these do not match, the attachment will fail. Based on ease, we
suggest making changes to the single-cell object barcodes.</p>
<div id="calculating-clonesize" class="section level3">
<h3>Calculating cloneSize</h3>
<p>Part of <code>combineExpression()</code> is calculating the clonal
frequency and proportion, placing each clone into groups called
<strong>cloneSize</strong>. The default <strong>cloneSize</strong>
argument uses the following bins: c(Rare = 1e-4, Small = 0.001, Medium =
0.01, Large = 0.1, Hyperexpanded = 1), which can be modified to include
more/less bins or different names.</p>
<p>Clonal frequency and proportion is dependent on the repertoires being
compared, which we can modify the calculation using the
<strong>group.by</strong> parameter, such as grouping by the
<em>Patient</em> variable from above. If <strong>group.by</strong> is
not set, <code>combineExpression()</code> will calculate clonal
frequency, proportion, and <strong>cloneSize</strong> as a function of
individual sequencing runs. In addition, <strong>cloneSize</strong> can
use the frequency of clones when <strong>proportion</strong> =
FALSE.</p>
<p>We can look at the default cloneSize groupings using the Single-Cell
Experiment object we just created above with using
<strong>group.by</strong> set to the <em>sample</em> variable used in
<code>combineTCR()</code> :</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">combineExpression</span>(combined.TCR, </span>
<span id="cb58-2"><a href="#cb58-2" tabindex="-1"></a>                         sce, </span>
<span id="cb58-3"><a href="#cb58-3" tabindex="-1"></a>                         <span class="at">cloneCall=</span><span class="st">&quot;gene&quot;</span>, </span>
<span id="cb58-4"><a href="#cb58-4" tabindex="-1"></a>                         <span class="at">group.by =</span> <span class="st">&quot;sample&quot;</span>, </span>
<span id="cb58-5"><a href="#cb58-5" tabindex="-1"></a>                         <span class="at">proportion =</span> <span class="cn">TRUE</span>)</span>
<span id="cb58-6"><a href="#cb58-6" tabindex="-1"></a></span>
<span id="cb58-7"><a href="#cb58-7" tabindex="-1"></a><span class="co">#Define color palette </span></span>
<span id="cb58-8"><a href="#cb58-8" tabindex="-1"></a>colorblind_vector <span class="ot">&lt;-</span> <span class="fu">hcl.colors</span>(<span class="at">n=</span><span class="dv">7</span>, <span class="at">palette =</span> <span class="st">&quot;inferno&quot;</span>, <span class="at">fixup =</span> <span class="cn">TRUE</span>)</span>
<span id="cb58-9"><a href="#cb58-9" tabindex="-1"></a></span>
<span id="cb58-10"><a href="#cb58-10" tabindex="-1"></a><span class="fu">plotUMAP</span>(sce, <span class="at">colour_by =</span> <span class="st">&quot;cloneSize&quot;</span>) <span class="sc">+</span></span>
<span id="cb58-11"><a href="#cb58-11" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">rev</span>(colorblind_vector[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">5</span>,<span class="dv">7</span>)]))</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-42-1.png" width="700" /></p>
<p>Alternatively, if we want <strong>cloneSize</strong> to be based on
the frequency of the clone, we can set <strong>proportion</strong> =
FALSE and we will need to change the <strong>cloneSize</strong> bins to
integers. If we have not inspected our clone data, setting the upper
limit of the clonal frequency might be difficult -
<code>combineExpression()</code> will automatically adjust the upper
limit to fit the distribution of the frequencies. To demonstrate this,
check out the Seurat object below:</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" tabindex="-1"></a>scRep_example <span class="ot">&lt;-</span> <span class="fu">combineExpression</span>(combined.TCR, </span>
<span id="cb59-2"><a href="#cb59-2" tabindex="-1"></a>                                   scRep_example, </span>
<span id="cb59-3"><a href="#cb59-3" tabindex="-1"></a>                                   <span class="at">cloneCall=</span><span class="st">&quot;gene&quot;</span>, </span>
<span id="cb59-4"><a href="#cb59-4" tabindex="-1"></a>                                   <span class="at">group.by =</span> <span class="st">&quot;sample&quot;</span>, </span>
<span id="cb59-5"><a href="#cb59-5" tabindex="-1"></a>                                   <span class="at">proportion =</span> <span class="cn">FALSE</span>, </span>
<span id="cb59-6"><a href="#cb59-6" tabindex="-1"></a>                                   <span class="at">cloneSize=</span><span class="fu">c</span>(<span class="at">Single=</span><span class="dv">1</span>, <span class="at">Small=</span><span class="dv">5</span>, <span class="at">Medium=</span><span class="dv">20</span>, <span class="at">Large=</span><span class="dv">100</span>, <span class="at">Hyperexpanded=</span><span class="dv">500</span>))</span>
<span id="cb59-7"><a href="#cb59-7" tabindex="-1"></a></span>
<span id="cb59-8"><a href="#cb59-8" tabindex="-1"></a>Seurat<span class="sc">::</span><span class="fu">DimPlot</span>(scRep_example, <span class="at">group.by =</span> <span class="st">&quot;cloneSize&quot;</span>) <span class="sc">+</span></span>
<span id="cb59-9"><a href="#cb59-9" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">rev</span>(colorblind_vector[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">7</span>)]))</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-43-1.png" width="700" /></p>
</div>
<div id="combining-both-tcr-and-bcr" class="section level3">
<h3>Combining both TCR and BCR</h3>
<p>If we have TCR/BCR enrichment or want to add info for gamma-delta and
alpha-beta T cells, we can make a single list and use
<code>combineExpression()</code>.</p>
<p><strong>Major note</strong> if there are duplicate barcodes (if a
cell has both Ig and TCR), the immune receptor information will not be
added. It might be worth checking cluster identities and removing
incongruent barcodes in the products of <code>combineTCR()</code> and
<code>combineBCR()</code>.</p>
<p>As an anecdote, the <a
href="https://support.10xgenomics.com/single-cell-vdj/datasets/6.0.1/SC5v2_Melanoma_5Kcells_Connect_single_channel_SC5v2_Melanoma_5Kcells_Connect_single_channel">testing
data</a> we used to improve this function had 5-6% of barcode
overlap.</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" tabindex="-1"></a><span class="co">#This is an example of the process, which will not be evaluated during knit</span></span>
<span id="cb60-2"><a href="#cb60-2" tabindex="-1"></a>TCR <span class="ot">&lt;-</span> <span class="fu">combineTCR</span>(...)</span>
<span id="cb60-3"><a href="#cb60-3" tabindex="-1"></a>BCR <span class="ot">&lt;-</span> <span class="fu">combineBCR</span>(...)</span>
<span id="cb60-4"><a href="#cb60-4" tabindex="-1"></a>list.receptors <span class="ot">&lt;-</span> <span class="fu">c</span>(TCR, BCR)</span>
<span id="cb60-5"><a href="#cb60-5" tabindex="-1"></a></span>
<span id="cb60-6"><a href="#cb60-6" tabindex="-1"></a></span>
<span id="cb60-7"><a href="#cb60-7" tabindex="-1"></a>seurat <span class="ot">&lt;-</span> <span class="fu">combineExpression</span>(list.receptors, </span>
<span id="cb60-8"><a href="#cb60-8" tabindex="-1"></a>                            seurat, </span>
<span id="cb60-9"><a href="#cb60-9" tabindex="-1"></a>                            <span class="at">cloneCall=</span><span class="st">&quot;gene&quot;</span>, </span>
<span id="cb60-10"><a href="#cb60-10" tabindex="-1"></a>                            <span class="at">proportion =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<hr />
</div>
</div>
</div>
<div id="visualizations-for-single-cell-objects" class="section level1">
<h1>Visualizations for Single-Cell Objects</h1>
<div id="clonaloverlay" class="section level2">
<h2>clonalOverlay</h2>
<p>Using the dimensional reduction graphs as a reference, we can also
generate an overlay of the position of clonally-expanded cells using
<code>clonalOverlay()</code>.</p>
<p><strong>reduction</strong></p>
<ul>
<li>The dimensional reduction for the visualization, <strong>(default =
“pca”)</strong></li>
</ul>
<p><strong>cut.category</strong></p>
<ul>
<li>Use of “clonalFrequency” or “clonalProportion” to filter the
overlay</li>
</ul>
<p><strong>cutpoint</strong></p>
<ul>
<li>lowest clonal frequency or proportion to generate the contour
plot</li>
</ul>
<p><strong>bins</strong></p>
<ul>
<li>the number of contours to draw</li>
</ul>
<p><code>clonalOverlay()</code> can be used to look across all cells or
faceted by a meta data variable using <strong>facet.by</strong>. The
overall dimensional reduction will be maintained as we facet, while the
contour plots will adjust based on the <strong>facet.by</strong>
variable. The coloring of the dot plot is based on the active identity
of the single-cell object.</p>
<p>This visualization was authored by Dr. Francesco Mazziotta and
inspired by Drs. Carmona and Andreatta and their work with <a
href="https://github.com/carmonalab/ProjecTILs">ProjectTIL</a>, which is
a great pipeline to annotated T cell subtypes.</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" tabindex="-1"></a><span class="co">#Adding patient information</span></span>
<span id="cb61-2"><a href="#cb61-2" tabindex="-1"></a>scRep_example<span class="sc">$</span>Patient <span class="ot">&lt;-</span> <span class="fu">substr</span>(scRep_example<span class="sc">$</span>orig.ident, <span class="dv">1</span>,<span class="dv">3</span>)</span>
<span id="cb61-3"><a href="#cb61-3" tabindex="-1"></a></span>
<span id="cb61-4"><a href="#cb61-4" tabindex="-1"></a><span class="fu">clonalOverlay</span>(scRep_example, </span>
<span id="cb61-5"><a href="#cb61-5" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, </span>
<span id="cb61-6"><a href="#cb61-6" tabindex="-1"></a>              <span class="at">cutpoint =</span> <span class="dv">1</span>, </span>
<span id="cb61-7"><a href="#cb61-7" tabindex="-1"></a>              <span class="at">bins =</span> <span class="dv">10</span>, </span>
<span id="cb61-8"><a href="#cb61-8" tabindex="-1"></a>              <span class="at">facet.by =</span> <span class="st">&quot;Patient&quot;</span>) <span class="sc">+</span> </span>
<span id="cb61-9"><a href="#cb61-9" tabindex="-1"></a>              <span class="fu">guides</span>(<span class="at">color =</span> <span class="st">&quot;none&quot;</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-45-1.png" width="700" /></p>
</div>
<div id="clonalnetwork" class="section level2">
<h2>clonalNetwork</h2>
<p>Similar to <code>clonalOverlay()</code>, we can look at the network
interaction of clones shared between clusters along the single-cell
dimensional reduction using <code>clonalNetwork()</code>. This function
shows the relative proportion of clones from the starting node, with the
ending node indicated by the arrow.</p>
<p>Filtering for clones can be accomplished using 3 methods:</p>
<p><strong>filter.clones</strong></p>
<ul>
<li>Select a number to isolate the clones comprising the top n number of
cells (e.g., <strong>filter.clones</strong> = 2000)<br />
</li>
<li>Select “min” to make sure all groups are scaled to the size of the
minimum group</li>
</ul>
<p><strong>filter.identity</strong></p>
<ul>
<li>For the identity chosen to visualize, show the to and from network
connections for a single group</li>
</ul>
<p><strong>filter.proportion</strong></p>
<ul>
<li>Remove clones that comprise less than a certain proportion of clones
in groups.</li>
</ul>
<p><strong>filter.graph</strong></p>
<ul>
<li>Remove the reciprocal edges from the half of the graph, allowing for
cleaner visualization.</li>
</ul>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" tabindex="-1"></a><span class="co">#ggraph needs to be loaded due to issues with ggplot</span></span>
<span id="cb62-2"><a href="#cb62-2" tabindex="-1"></a><span class="fu">library</span>(ggraph)</span>
<span id="cb62-3"><a href="#cb62-3" tabindex="-1"></a></span>
<span id="cb62-4"><a href="#cb62-4" tabindex="-1"></a><span class="co">#No Identity filter</span></span>
<span id="cb62-5"><a href="#cb62-5" tabindex="-1"></a><span class="fu">clonalNetwork</span>(scRep_example, </span>
<span id="cb62-6"><a href="#cb62-6" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, </span>
<span id="cb62-7"><a href="#cb62-7" tabindex="-1"></a>              <span class="at">group.by =</span> <span class="st">&quot;seurat_clusters&quot;</span>,</span>
<span id="cb62-8"><a href="#cb62-8" tabindex="-1"></a>              <span class="at">filter.clones =</span> <span class="cn">NULL</span>,</span>
<span id="cb62-9"><a href="#cb62-9" tabindex="-1"></a>              <span class="at">filter.identity =</span> <span class="cn">NULL</span>,</span>
<span id="cb62-10"><a href="#cb62-10" tabindex="-1"></a>              <span class="at">cloneCall =</span> <span class="st">&quot;aa&quot;</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-46-1.png" width="700" /></p>
<p>We can look at the clonal relationships relative to a single cluster
using the <strong>filter.identity</strong> parameter.</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" tabindex="-1"></a><span class="co">#Examining Cluster 3 only</span></span>
<span id="cb63-2"><a href="#cb63-2" tabindex="-1"></a><span class="fu">clonalNetwork</span>(scRep_example, </span>
<span id="cb63-3"><a href="#cb63-3" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, </span>
<span id="cb63-4"><a href="#cb63-4" tabindex="-1"></a>              <span class="at">group.by =</span> <span class="st">&quot;seurat_clusters&quot;</span>,</span>
<span id="cb63-5"><a href="#cb63-5" tabindex="-1"></a>              <span class="at">filter.identity =</span> <span class="dv">3</span>,</span>
<span id="cb63-6"><a href="#cb63-6" tabindex="-1"></a>              <span class="at">cloneCall =</span> <span class="st">&quot;aa&quot;</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-47-1.png" width="700" /></p>
<p>We can also use <strong>exportClones</strong> parameter to quickly
get clones that are shared across multiple identity groups, along with
the total number of clones in the data set.</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" tabindex="-1"></a>shared.clones <span class="ot">&lt;-</span> <span class="fu">clonalNetwork</span>(scRep_example, </span>
<span id="cb64-2"><a href="#cb64-2" tabindex="-1"></a>                               <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, </span>
<span id="cb64-3"><a href="#cb64-3" tabindex="-1"></a>                               <span class="at">group.by =</span> <span class="st">&quot;seurat_clusters&quot;</span>,</span>
<span id="cb64-4"><a href="#cb64-4" tabindex="-1"></a>                               <span class="at">cloneCall =</span> <span class="st">&quot;aa&quot;</span>, </span>
<span id="cb64-5"><a href="#cb64-5" tabindex="-1"></a>                               <span class="at">exportClones =</span> <span class="cn">TRUE</span>)</span>
<span id="cb64-6"><a href="#cb64-6" tabindex="-1"></a><span class="fu">head</span>(shared.clones)</span></code></pre></div>
<pre><code>## [38;5;246m# A tibble: 6 × 2[39m
##   clone                              sum
##   [3m[38;5;246m&lt;chr&gt;[39m[23m                            [3m[38;5;246m&lt;int&gt;[39m[23m
## [38;5;250m1[39m CVVSDNTGGFKTIF_CASSVRRERANTGELFF    11
## [38;5;250m2[39m CAELNQAGTALIF_CASSQAPFSTSGELFF       3
## [38;5;250m3[39m CASLSGSARQLTF_CASSSTVAGEQYF          3
## [38;5;250m4[39m CALSGSRDDKIIF_NA                     2
## [38;5;250m5[39m CARKVRDSSYKLIF_CASSDSGYNEQFF         2
## [38;5;250m6[39m CASLSGSARQLTF_CASSPTVAGEQFF          2</code></pre>
</div>
<div id="highlightclones" class="section level2">
<h2>highlightClones</h2>
<p>We can also look at the clones by calling specific sequences in the
<code>highlightclones()</code> below. In order to highlight the clones,
we first need to use the <strong>cloneCall</strong>, the type of
sequence we will be using, and then the specific sequences themselves
using <strong>sequence</strong>. Below, we can see the steps to
highlight the most prominent sequences
<em>CAERGSGGSYIPTF_CASSDPSGRQGPRWDTQYF</em>, and
<em>CARKVRDSSYKLIF_CASSDSGYNEQFF</em>.</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" tabindex="-1"></a>scRep_example <span class="ot">&lt;-</span> <span class="fu">highlightClones</span>(scRep_example, </span>
<span id="cb66-2"><a href="#cb66-2" tabindex="-1"></a>                    <span class="at">cloneCall=</span> <span class="st">&quot;aa&quot;</span>, </span>
<span id="cb66-3"><a href="#cb66-3" tabindex="-1"></a>                    <span class="at">sequence =</span> <span class="fu">c</span>(<span class="st">&quot;CAERGSGGSYIPTF_CASSDPSGRQGPRWDTQYF&quot;</span>, </span>
<span id="cb66-4"><a href="#cb66-4" tabindex="-1"></a>                                 <span class="st">&quot;CARKVRDSSYKLIF_CASSDSGYNEQFF&quot;</span>))</span>
<span id="cb66-5"><a href="#cb66-5" tabindex="-1"></a></span>
<span id="cb66-6"><a href="#cb66-6" tabindex="-1"></a>Seurat<span class="sc">::</span><span class="fu">DimPlot</span>(scRep_example, <span class="at">group.by =</span> <span class="st">&quot;highlight&quot;</span>) <span class="sc">+</span> </span>
<span id="cb66-7"><a href="#cb66-7" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_blank</span>())</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-49-1.png" width="700" /></p>
</div>
<div id="clonaloccupy" class="section level2">
<h2>clonalOccupy</h2>
<p>We can also look at the count of cells by cluster assigned into
specific frequency ranges by using the <code>clonalOccupy()</code>
function and selecting the <strong>x.axis</strong> to display cluster or
other variables in the meta data of the single cell object.</p>
<p><strong>proportion</strong></p>
<ul>
<li>can be used to look at relative level groupings</li>
</ul>
<p><strong>label</strong></p>
<ul>
<li>will still return the absolute number of clones</li>
</ul>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" tabindex="-1"></a><span class="fu">clonalOccupy</span>(scRep_example, </span>
<span id="cb67-2"><a href="#cb67-2" tabindex="-1"></a>              <span class="at">x.axis =</span> <span class="st">&quot;seurat_clusters&quot;</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-50-1.png" width="700" /></p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" tabindex="-1"></a><span class="fu">clonalOccupy</span>(scRep_example, </span>
<span id="cb68-2"><a href="#cb68-2" tabindex="-1"></a>             <span class="at">x.axis =</span> <span class="st">&quot;ident&quot;</span>, </span>
<span id="cb68-3"><a href="#cb68-3" tabindex="-1"></a>             <span class="at">proportion =</span> <span class="cn">TRUE</span>, </span>
<span id="cb68-4"><a href="#cb68-4" tabindex="-1"></a>             <span class="at">label =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-50-2.png" width="700" /></p>
</div>
<div id="alluvialclones" class="section level2">
<h2>alluvialClones</h2>
<p>After the metadata has been modified, we can look at clones across
multiple categories using the <code>alluvialClones()</code> function. We
are able to use the plots to examine the interchange of categorical
variables. Because this function will produce a graph with each clone
arranged by called stratification, this will take some time depending on
the size of the repertoire.</p>
<p>To understand the basic concepts of this graphing method and the
ggalluvial R package, we recommend reading <a
href="https://cran.r-project.org/web/packages/ggalluvial/vignettes/ggalluvial.html">this
post</a>.</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" tabindex="-1"></a><span class="co">#Adding type information</span></span>
<span id="cb69-2"><a href="#cb69-2" tabindex="-1"></a>scRep_example<span class="sc">$</span>Type <span class="ot">&lt;-</span> <span class="fu">substr</span>(scRep_example<span class="sc">$</span>orig.ident, <span class="dv">4</span>,<span class="dv">4</span>)</span>
<span id="cb69-3"><a href="#cb69-3" tabindex="-1"></a></span>
<span id="cb69-4"><a href="#cb69-4" tabindex="-1"></a><span class="fu">alluvialClones</span>(scRep_example, </span>
<span id="cb69-5"><a href="#cb69-5" tabindex="-1"></a>               <span class="at">cloneCall =</span> <span class="st">&quot;aa&quot;</span>, </span>
<span id="cb69-6"><a href="#cb69-6" tabindex="-1"></a>               <span class="at">y.axes =</span> <span class="fu">c</span>(<span class="st">&quot;Patient&quot;</span>, <span class="st">&quot;ident&quot;</span>, <span class="st">&quot;Type&quot;</span>), </span>
<span id="cb69-7"><a href="#cb69-7" tabindex="-1"></a>               <span class="at">color =</span> <span class="fu">c</span>(<span class="st">&quot;CVVSDNTGGFKTIF_CASSVRRERANTGELFF&quot;</span>, <span class="st">&quot;NA_CASSVRRERANTGELFF&quot;</span>)) <span class="sc">+</span> </span>
<span id="cb69-8"><a href="#cb69-8" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;grey&quot;</span>, colorblind_vector[<span class="dv">3</span>]))</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-51-1.png" width="700" /></p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" tabindex="-1"></a><span class="fu">alluvialClones</span>(scRep_example, </span>
<span id="cb70-2"><a href="#cb70-2" tabindex="-1"></a>                   <span class="at">cloneCall =</span> <span class="st">&quot;gene&quot;</span>, </span>
<span id="cb70-3"><a href="#cb70-3" tabindex="-1"></a>                   <span class="at">y.axes =</span> <span class="fu">c</span>(<span class="st">&quot;Patient&quot;</span>, <span class="st">&quot;ident&quot;</span>, <span class="st">&quot;Type&quot;</span>), </span>
<span id="cb70-4"><a href="#cb70-4" tabindex="-1"></a>                   <span class="at">color =</span> <span class="st">&quot;ident&quot;</span>) </span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-51-2.png" width="700" /></p>
</div>
<div id="getcirclize" class="section level2">
<h2>getCirclize</h2>
<p>Like alluvial graphs, we can also visualize the interconnection of
clusters using the chord diagrams from the <a
href="https://jokergoo.github.io/circlize_book/book/">circlize R
package</a>.</p>
<p>The first step is getting the data frame output to feed into the
<code>chordDiagram()</code> function in circlize, which can be done
using <code>getCirclize()</code>.</p>
<p>This will calculate the relative number of clones shared based on the
<strong>group.by</strong> variable using the product of
<code>combineExpression()</code>. It is important to note
<code>getCirclize()</code> will create a matrix the size of the
<strong>group.by</strong> variable and then simplify into instructions
to be read by the circlize R package. The output is the total number of
unique and shared clones by the <strong>group.by</strong> variable.</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" tabindex="-1"></a><span class="fu">library</span>(circlize)</span>
<span id="cb71-2"><a href="#cb71-2" tabindex="-1"></a><span class="fu">library</span>(scales)</span>
<span id="cb71-3"><a href="#cb71-3" tabindex="-1"></a></span>
<span id="cb71-4"><a href="#cb71-4" tabindex="-1"></a>circles <span class="ot">&lt;-</span> <span class="fu">getCirclize</span>(scRep_example, </span>
<span id="cb71-5"><a href="#cb71-5" tabindex="-1"></a>                       <span class="at">group.by =</span> <span class="st">&quot;seurat_clusters&quot;</span>)</span>
<span id="cb71-6"><a href="#cb71-6" tabindex="-1"></a></span>
<span id="cb71-7"><a href="#cb71-7" tabindex="-1"></a><span class="co">#Just assigning the normal colors to each cluster</span></span>
<span id="cb71-8"><a href="#cb71-8" tabindex="-1"></a>grid.cols <span class="ot">&lt;-</span> <span class="fu">hue_pal</span>()(<span class="fu">length</span>(<span class="fu">unique</span>(scRep_example<span class="sc">$</span>seurat_clusters)))</span>
<span id="cb71-9"><a href="#cb71-9" tabindex="-1"></a><span class="fu">names</span>(grid.cols) <span class="ot">&lt;-</span> <span class="fu">unique</span>(scRep_example<span class="sc">$</span>seurat_clusters)</span>
<span id="cb71-10"><a href="#cb71-10" tabindex="-1"></a></span>
<span id="cb71-11"><a href="#cb71-11" tabindex="-1"></a><span class="co">#Graphing the chord diagram</span></span>
<span id="cb71-12"><a href="#cb71-12" tabindex="-1"></a><span class="fu">chordDiagram</span>(circles, <span class="at">self.link =</span> <span class="dv">1</span>, <span class="at">grid.col =</span> grid.cols)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-52-1.png" width="700" /></p>
<p>This can also be used if we want to explore just the lung-specific T
cells by just subsetting the single-cell object. For the sake of this
vignette, we can also look at setting <strong>proportion</strong> = TRUE
to get a scaled output.</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" tabindex="-1"></a>subset <span class="ot">&lt;-</span> <span class="fu">subset</span>(scRep_example, Type <span class="sc">==</span> <span class="st">&quot;L&quot;</span>)</span>
<span id="cb72-2"><a href="#cb72-2" tabindex="-1"></a></span>
<span id="cb72-3"><a href="#cb72-3" tabindex="-1"></a>circles <span class="ot">&lt;-</span> <span class="fu">getCirclize</span>(subset, <span class="at">group.by =</span> <span class="st">&quot;ident&quot;</span>, <span class="at">proportion =</span> <span class="cn">TRUE</span>)</span>
<span id="cb72-4"><a href="#cb72-4" tabindex="-1"></a></span>
<span id="cb72-5"><a href="#cb72-5" tabindex="-1"></a>grid.cols <span class="ot">&lt;-</span> scales<span class="sc">::</span><span class="fu">hue_pal</span>()(<span class="fu">length</span>(<span class="fu">unique</span>(subset<span class="sc">@</span>active.ident)))</span>
<span id="cb72-6"><a href="#cb72-6" tabindex="-1"></a><span class="fu">names</span>(grid.cols) <span class="ot">&lt;-</span> <span class="fu">levels</span>(subset<span class="sc">@</span>active.ident)</span>
<span id="cb72-7"><a href="#cb72-7" tabindex="-1"></a></span>
<span id="cb72-8"><a href="#cb72-8" tabindex="-1"></a><span class="fu">chordDiagram</span>(circles, </span>
<span id="cb72-9"><a href="#cb72-9" tabindex="-1"></a>             <span class="at">self.link =</span> <span class="dv">1</span>, </span>
<span id="cb72-10"><a href="#cb72-10" tabindex="-1"></a>             <span class="at">grid.col =</span> grid.cols, </span>
<span id="cb72-11"><a href="#cb72-11" tabindex="-1"></a>             <span class="at">directional =</span> <span class="dv">1</span>, </span>
<span id="cb72-12"><a href="#cb72-12" tabindex="-1"></a>             <span class="at">direction.type =</span>  <span class="st">&quot;arrows&quot;</span>,</span>
<span id="cb72-13"><a href="#cb72-13" tabindex="-1"></a>             <span class="at">link.arr.type =</span> <span class="st">&quot;big.arrow&quot;</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-53-1.png" width="700" /></p>
<hr />
</div>
</div>
<div id="quantifying-clonal-bias" class="section level1">
<h1>Quantifying Clonal Bias</h1>
<div id="startracdiversity" class="section level2">
<h2>StartracDiversity</h2>
<p>From the excellent work by <a
href="https://www.nature.com/articles/s41586-018-0694-x">Lei Zhang, et
al.</a>, the authors introduce new methods for looking at clones by
cellular origins and cluster identification. Their <a
href="https://github.com/Japrin/STARTRAC">STARTRAC</a> software has been
adapted to work with scRepertoire and please read and cite their
excellent work.</p>
<p>In order to use the <code>StartracDiversity()</code> function, you
will need to include the product of the
<code>combinedExpression()</code> function. The second requirement is a
column header in the meta data of the Seurat object that has tissue of
origin. In the example data, <strong>type</strong> corresponds to the
column “Type”, which includes the “P” and “T” classifiers. The indices
can be subsetted for a specific patient or examined overall using the
<strong>by</strong> variable. Importantly, the function uses only the
strict definition of a clone of the VDJC genes and the CDR3 nucleotide
sequence.</p>
<p><strong>The indices output includes:</strong></p>
<ul>
<li>expa - Clonal Expansion<br />
</li>
<li>migr - Cross-tissue Migration<br />
</li>
<li>tran - State Transition</li>
</ul>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" tabindex="-1"></a><span class="fu">StartracDiversity</span>(scRep_example, </span>
<span id="cb73-2"><a href="#cb73-2" tabindex="-1"></a>                  <span class="at">type =</span> <span class="st">&quot;Type&quot;</span>, </span>
<span id="cb73-3"><a href="#cb73-3" tabindex="-1"></a>                  <span class="at">group.by =</span> <span class="st">&quot;Patient&quot;</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-54-1.png" width="700" /></p>
</div>
<div id="clonalbias" class="section level2">
<h2>clonalBias</h2>
<p>A new metric proposed by <a
href="https://pubmed.ncbi.nlm.nih.gov/35829695/">Massimo et al</a>,
<code>clonalBias()</code>, like STARTRAC is a clonal metric that seeks
to quantify how individual clones are skewed towards a specific cellular
compartment or cluster.</p>
<p><strong>split.by</strong></p>
<ul>
<li>Variable used to calculate the baseline frequencies</li>
</ul>
<p><strong>group.by</strong></p>
<ul>
<li>The compartment/variable used for the purpose of the comparison</li>
</ul>
<p><strong>min.expand</strong></p>
<ul>
<li>Cut point for frequency <strong>(default = 10)</strong></li>
</ul>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" tabindex="-1"></a><span class="fu">clonalBias</span>(scRep_example, </span>
<span id="cb74-2"><a href="#cb74-2" tabindex="-1"></a>           <span class="at">cloneCall =</span> <span class="st">&quot;aa&quot;</span>, </span>
<span id="cb74-3"><a href="#cb74-3" tabindex="-1"></a>           <span class="at">split.by =</span> <span class="st">&quot;Patient&quot;</span>, </span>
<span id="cb74-4"><a href="#cb74-4" tabindex="-1"></a>           <span class="at">group.by =</span> <span class="st">&quot;seurat_clusters&quot;</span>,</span>
<span id="cb74-5"><a href="#cb74-5" tabindex="-1"></a>           <span class="at">n.boots =</span> <span class="dv">10</span>, </span>
<span id="cb74-6"><a href="#cb74-6" tabindex="-1"></a>           <span class="at">min.expand =</span><span class="dv">0</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-55-1.png" width="700" /></p>
<hr />
</div>
</div>
<div id="clustering-by-edit-distance" class="section level1">
<h1>Clustering by Edit Distance</h1>
<div id="clonalcluster" class="section level2">
<h2>clonalCluster</h2>
<p>The nucleotide or amino acid sequences of the chains can be used to
cluster clonotypes by examining the edit distance of the sequences. This
approach is underlying the <code>combineBCR()</code> function but now
can be applied to the B or T cell receptors at the level of nucleotides
(<strong>sequence = “nt”</strong>) or amino acids (<strong>sequence =
“aa”</strong>). It will add a cluster to the end of each list element by
generating a network connected by the similarity in sequence. This
network is directed by the <strong>threshold</strong> variable, where
0.85 is the normalized mean edit distance.</p>
<p>Edit-distance based clusters will have the following format:</p>
<ul>
<li>[chain:] + :Cluster + [number] e.g., TRA:Cluster.1</li>
</ul>
<p><strong>Cluster</strong> denotes if the cluster was called using the
normalized Levenshtein distance, which takes the edit distance
calculated between 2 sequences and divides that by the mean of the
sequence lengths. Unconnected sequences will have <em>NA</em>
values.</p>
<div id="basic-use" class="section level3">
<h3>Basic use</h3>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" tabindex="-1"></a>sub_combined <span class="ot">&lt;-</span> <span class="fu">clonalCluster</span>(combined.TCR[[<span class="dv">2</span>]], </span>
<span id="cb75-2"><a href="#cb75-2" tabindex="-1"></a>                              <span class="at">chain =</span> <span class="st">&quot;TRA&quot;</span>, </span>
<span id="cb75-3"><a href="#cb75-3" tabindex="-1"></a>                              <span class="at">sequence =</span> <span class="st">&quot;aa&quot;</span>, </span>
<span id="cb75-4"><a href="#cb75-4" tabindex="-1"></a>                              <span class="at">threshold =</span> <span class="fl">0.85</span>, </span>
<span id="cb75-5"><a href="#cb75-5" tabindex="-1"></a>                              <span class="at">group.by =</span> <span class="cn">NULL</span>)</span>
<span id="cb75-6"><a href="#cb75-6" tabindex="-1"></a></span>
<span id="cb75-7"><a href="#cb75-7" tabindex="-1"></a><span class="fu">head</span>(sub_combined[,<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">13</span>)])</span></code></pre></div>
<pre><code>##                   barcode sample Type
## 1 P17L_AAACCTGAGGTGTTAA-1   P17L    L
## 2 P17L_AAACCTGCACGTTGGC-1   P17L    L
## 3 P17L_AAACCTGGTACGACCC-1   P17L    L
## 4 P17L_AAACCTGGTTCGCTAA-1   P17L    L
## 5 P17L_AAACCTGTCAGCACAT-1   P17L    L
## 6 P17L_AAACCTGTCCGGCACA-1   P17L    L</code></pre>
</div>
<div id="clustering-with-a-single-cell-object" class="section level3">
<h3>Clustering with a single-cell object</h3>
<p>If performed over the number of samples, such as the list elements,
<strong>group.by</strong> can used to calculate only the clusters on the
setting of patient sample (<strong>group.by</strong> = “Patient”) or
tissue type (<strong>group.by</strong> = “Type”). This will add the
selected group to the beginning of the cluster designation. We can also
call <code>clonalCluster()</code> directly on a Single-Cell Object.</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" tabindex="-1"></a><span class="co">#Define color palette </span></span>
<span id="cb77-2"><a href="#cb77-2" tabindex="-1"></a>colorblind_vector <span class="ot">&lt;-</span> <span class="fu">hcl.colors</span>(<span class="at">n=</span><span class="dv">7</span>, <span class="at">palette =</span> <span class="st">&quot;inferno&quot;</span>, <span class="at">fixup =</span> <span class="cn">TRUE</span>)</span>
<span id="cb77-3"><a href="#cb77-3" tabindex="-1"></a></span>
<span id="cb77-4"><a href="#cb77-4" tabindex="-1"></a>scRep_example <span class="ot">&lt;-</span> <span class="fu">clonalCluster</span>(scRep_example, </span>
<span id="cb77-5"><a href="#cb77-5" tabindex="-1"></a>                               <span class="at">chain =</span> <span class="st">&quot;TRA&quot;</span>, </span>
<span id="cb77-6"><a href="#cb77-6" tabindex="-1"></a>                               <span class="at">sequence =</span> <span class="st">&quot;aa&quot;</span>, </span>
<span id="cb77-7"><a href="#cb77-7" tabindex="-1"></a>                               <span class="at">threshold =</span> <span class="fl">0.85</span>, </span>
<span id="cb77-8"><a href="#cb77-8" tabindex="-1"></a>                               <span class="at">group.by =</span> <span class="st">&quot;Patient&quot;</span>)</span>
<span id="cb77-9"><a href="#cb77-9" tabindex="-1"></a></span>
<span id="cb77-10"><a href="#cb77-10" tabindex="-1"></a>Seurat<span class="sc">::</span><span class="fu">DimPlot</span>(scRep_example, <span class="at">group.by =</span> <span class="st">&quot;TRA_cluster&quot;</span>) <span class="sc">+</span></span>
<span id="cb77-11"><a href="#cb77-11" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values =</span>  <span class="fu">hcl.colors</span>(<span class="at">n=</span><span class="fu">length</span>(<span class="fu">unique</span>(scRep_example<span class="sc">@</span>meta.data[,<span class="st">&quot;TRA_cluster&quot;</span>])), <span class="st">&quot;inferno&quot;</span>)) <span class="sc">+</span> </span>
<span id="cb77-12"><a href="#cb77-12" tabindex="-1"></a>  Seurat<span class="sc">::</span><span class="fu">NoLegend</span>() <span class="sc">+</span> </span>
<span id="cb77-13"><a href="#cb77-13" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_blank</span>())</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-57-1.png" width="700" /></p>
<p>Using <code>clonalCluster()</code>, we can also return an igraph
object of all the related sequences using <strong>exportGraph</strong> =
TRUE. The returned igraph object contains only the sequences that have
at least one connection with another sequence. The igraph can then be
directly visualized (below) or used for downstream analysis (see the
igraph <a href="https://r.igraph.org/">website</a>).</p>
</div>
<div id="returning-an-igraph-object" class="section level3">
<h3>Returning an igraph object:</h3>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" tabindex="-1"></a><span class="co">#Clustering Patient 19 samples</span></span>
<span id="cb78-2"><a href="#cb78-2" tabindex="-1"></a>igraph.object <span class="ot">&lt;-</span> <span class="fu">clonalCluster</span>(combined.TCR[<span class="fu">c</span>(<span class="dv">5</span>,<span class="dv">6</span>)],</span>
<span id="cb78-3"><a href="#cb78-3" tabindex="-1"></a>                               <span class="at">chain =</span> <span class="st">&quot;TRB&quot;</span>,</span>
<span id="cb78-4"><a href="#cb78-4" tabindex="-1"></a>                               <span class="at">sequence =</span> <span class="st">&quot;aa&quot;</span>,</span>
<span id="cb78-5"><a href="#cb78-5" tabindex="-1"></a>                               <span class="at">group.by =</span> <span class="st">&quot;sample&quot;</span>,</span>
<span id="cb78-6"><a href="#cb78-6" tabindex="-1"></a>                               <span class="at">threshold =</span> <span class="fl">0.85</span>, </span>
<span id="cb78-7"><a href="#cb78-7" tabindex="-1"></a>                               <span class="at">exportGraph =</span> <span class="cn">TRUE</span>)</span>
<span id="cb78-8"><a href="#cb78-8" tabindex="-1"></a></span>
<span id="cb78-9"><a href="#cb78-9" tabindex="-1"></a><span class="co">#Setting color scheme</span></span>
<span id="cb78-10"><a href="#cb78-10" tabindex="-1"></a>col_legend <span class="ot">&lt;-</span> <span class="fu">factor</span>(igraph<span class="sc">::</span><span class="fu">V</span>(igraph.object)<span class="sc">$</span>group)</span>
<span id="cb78-11"><a href="#cb78-11" tabindex="-1"></a>col_samples <span class="ot">&lt;-</span> <span class="fu">hcl.colors</span>(<span class="dv">3</span>,<span class="st">&quot;inferno&quot;</span>)[<span class="fu">as.numeric</span>(col_legend)]</span>
<span id="cb78-12"><a href="#cb78-12" tabindex="-1"></a>color.legend <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">unique</span>(igraph<span class="sc">::</span><span class="fu">V</span>(igraph.object)<span class="sc">$</span>group))</span>
<span id="cb78-13"><a href="#cb78-13" tabindex="-1"></a></span>
<span id="cb78-14"><a href="#cb78-14" tabindex="-1"></a><span class="co">#Plotting</span></span>
<span id="cb78-15"><a href="#cb78-15" tabindex="-1"></a><span class="fu">plot</span>(</span>
<span id="cb78-16"><a href="#cb78-16" tabindex="-1"></a>  igraph.object,</span>
<span id="cb78-17"><a href="#cb78-17" tabindex="-1"></a>  <span class="at">vertex.size     =</span> <span class="fu">sqrt</span>(igraph<span class="sc">::</span><span class="fu">V</span>(igraph.object)<span class="sc">$</span>size),</span>
<span id="cb78-18"><a href="#cb78-18" tabindex="-1"></a>  <span class="at">vertex.label    =</span> <span class="cn">NA</span>,</span>
<span id="cb78-19"><a href="#cb78-19" tabindex="-1"></a>  <span class="at">edge.arrow.size =</span> .<span class="dv">25</span>,</span>
<span id="cb78-20"><a href="#cb78-20" tabindex="-1"></a>  <span class="at">vertex.color    =</span> col_samples</span>
<span id="cb78-21"><a href="#cb78-21" tabindex="-1"></a>)</span>
<span id="cb78-22"><a href="#cb78-22" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;topleft&quot;</span>, <span class="at">legend =</span> <span class="fu">levels</span>(color.legend), <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">col =</span> <span class="fu">unique</span>(col_samples), <span class="at">bty =</span> <span class="st">&quot;n&quot;</span>)</span></code></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-58-1.png" width="700" /></p>
<hr />
</div>
</div>
</div>
<div id="conclusion" class="section level1">
<h1>Conclusion</h1>
<p>This has been a general overview of the capabilities of scRepertoire
from the initial processing and visualization to attach to the mRNA
expression values in a single-cell object. If you have any questions,
comments, or suggestions, please visit the <a
href="https://github.com/ncborcherding/scRepertoire">GitHub
repository</a>.</p>
<div id="session-info" class="section level3">
<h3>Session Info</h3>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.3.2 (2023-10-31)
## Platform: aarch64-apple-darwin20 (64-bit)
## Running under: macOS Ventura 13.4
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
## LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## time zone: America/Chicago
## tzcode source: internal
## 
## attached base packages:
## [1] stats4    stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] scales_1.3.0                circlize_0.4.16            
##  [3] ggraph_2.2.1                scRepertoire_2.0.4         
##  [5] scater_1.30.1               ggplot2_3.5.0              
##  [7] scuttle_1.12.0              SingleCellExperiment_1.24.0
##  [9] SummarizedExperiment_1.32.0 Biobase_2.62.0             
## [11] GenomicRanges_1.54.1        GenomeInfoDb_1.38.8        
## [13] IRanges_2.36.0              S4Vectors_0.40.2           
## [15] BiocGenerics_0.48.1         MatrixGenerics_1.14.0      
## [17] matrixStats_1.2.0           BiocStyle_2.30.0           
## 
## loaded via a namespace (and not attached):
##   [1] cubature_2.1.0            RcppAnnoy_0.0.22         
##   [3] later_1.3.2               splines_4.3.2            
##   [5] bitops_1.0-7              tibble_3.2.1             
##   [7] polyclip_1.10-6           fastDummies_1.7.3        
##   [9] lifecycle_1.0.4           globals_0.16.3           
##  [11] lattice_0.22-6            MASS_7.3-60.0.1          
##  [13] magrittr_2.0.3            plotly_4.10.4            
##  [15] sass_0.4.9                rmarkdown_2.26           
##  [17] jquerylib_0.1.4           yaml_2.3.8               
##  [19] httpuv_1.6.15             Seurat_5.0.3             
##  [21] sctransform_0.4.1         spam_2.10-0              
##  [23] spatstat.sparse_3.0-3     sp_2.1-3                 
##  [25] reticulate_1.35.0         pbapply_1.7-2            
##  [27] cowplot_1.1.3             RColorBrewer_1.1-3       
##  [29] abind_1.4-5               zlibbioc_1.48.2          
##  [31] Rtsne_0.17                purrr_1.0.2              
##  [33] RCurl_1.98-1.14           hash_2.2.6.3             
##  [35] tweenr_2.0.3              evmix_2.12               
##  [37] GenomeInfoDbData_1.2.11   ggrepel_0.9.5            
##  [39] irlba_2.3.5.1             spatstat.utils_3.0-4     
##  [41] listenv_0.9.1             iNEXT_3.0.1              
##  [43] goftest_1.2-3             MatrixModels_0.5-3       
##  [45] RSpectra_0.16-1           spatstat.random_3.2-3    
##  [47] fitdistrplus_1.1-11       parallelly_1.37.1        
##  [49] pkgdown_2.0.7             DelayedMatrixStats_1.24.0
##  [51] leiden_0.4.3.1            codetools_0.2-20         
##  [53] DelayedArray_0.28.0       ggforce_0.4.2            
##  [55] shape_1.4.6.1             tidyselect_1.2.1         
##  [57] farver_2.1.1              ScaledMatrix_1.10.0      
##  [59] viridis_0.6.5             spatstat.explore_3.2-7   
##  [61] jsonlite_1.8.8            BiocNeighbors_1.20.2     
##  [63] tidygraph_1.3.1           progressr_0.14.0         
##  [65] ggridges_0.5.6            ggalluvial_0.12.5        
##  [67] survival_3.5-8            systemfonts_1.0.6        
##  [69] tools_4.3.2               ragg_1.3.0               
##  [71] stringdist_0.9.12         ica_1.0-3                
##  [73] Rcpp_1.0.12               glue_1.7.0               
##  [75] gridExtra_2.3             SparseArray_1.2.4        
##  [77] xfun_0.43                 dplyr_1.1.4              
##  [79] withr_3.0.0               BiocManager_1.30.22      
##  [81] fastmap_1.1.1             fansi_1.0.6              
##  [83] SparseM_1.81              digest_0.6.35            
##  [85] rsvd_1.0.5                mime_0.12                
##  [87] R6_2.5.1                  textshaping_0.3.7        
##  [89] colorspace_2.1-0          scattermore_1.2          
##  [91] tensor_1.5                spatstat.data_3.0-4      
##  [93] utf8_1.2.4                tidyr_1.3.1              
##  [95] generics_0.1.3            data.table_1.15.4        
##  [97] htmlwidgets_1.6.4         graphlayouts_1.1.1       
##  [99] httr_1.4.7                S4Arrays_1.2.1           
## [101] uwot_0.1.16               pkgconfig_2.0.3          
## [103] gtable_0.3.4              lmtest_0.9-40            
## [105] XVector_0.42.0            htmltools_0.5.8.1        
## [107] dotCall64_1.1-1           bookdown_0.38            
## [109] SeuratObject_5.0.1        png_0.1-8                
## [111] ggdendro_0.2.0            knitr_1.45               
## [113] rstudioapi_0.16.0         reshape2_1.4.4           
## [115] rjson_0.2.21              nlme_3.1-164             
## [117] GlobalOptions_0.1.2       zoo_1.8-12               
## [119] cachem_1.0.8              stringr_1.5.1            
## [121] KernSmooth_2.23-22        miniUI_0.1.1.1           
## [123] parallel_4.3.2            vipor_0.4.7              
## [125] desc_1.4.3                pillar_1.9.0             
## [127] grid_4.3.2                vctrs_0.6.5              
## [129] RANN_2.6.1                promises_1.2.1           
## [131] VGAM_1.1-10               BiocSingular_1.18.0      
## [133] beachmat_2.18.1           xtable_1.8-4             
## [135] cluster_2.1.6             beeswarm_0.4.0           
## [137] evaluate_0.23             isoband_0.2.7            
## [139] truncdist_1.0-2           cli_3.6.2                
## [141] compiler_4.3.2            rlang_1.1.3              
## [143] crayon_1.5.2              future.apply_1.11.2      
## [145] labeling_0.4.3            plyr_1.8.9               
## [147] fs_1.6.3                  ggbeeswarm_0.7.2         
## [149] stringi_1.8.3             deldir_2.0-4             
## [151] viridisLite_0.4.2         BiocParallel_1.36.0      
## [153] munsell_0.5.1             gsl_2.1-8                
## [155] lazyeval_0.2.2            spatstat.geom_3.2-9      
## [157] quantreg_5.97             Matrix_1.6-5             
## [159] RcppHNSW_0.6.0            patchwork_1.2.0          
## [161] sparseMatrixStats_1.14.0  future_1.33.2            
## [163] shiny_1.8.1.1             highr_0.10               
## [165] evd_2.3-6.1               ROCR_1.0-11              
## [167] igraph_2.0.3              memoise_2.0.1            
## [169] bslib_0.7.0</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
      </div>

</div>



      <footer><div class="copyright">
  <p></p><p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>
