<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Starting work with scRepertoire. • scRepertoire</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Starting work with scRepertoire."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.6.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Repertoire Analysis</li>
    <li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Working with Single-Cell Objects</li>
    <li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Quantifying Clonal Bias</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Deep Learning Integrations</li>
    <li>
      <a href="../articles/immApex.html">Making Deep Learning Models with immApex</a>
    </li>
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
    <li>
      <a href="../articles/Running_Escape.html">Single-cell Gene Set Enrichment Analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Other Workflow Integrations</li>
    <li>
      <a href="../articles/immcantation.html">Combining Immcantation and scRepertoire</a>
    </li>
    <li>
      <a href="../articles/tcrpheno.html">TCRpheno applied to T cell fate</a>
    </li>
    <li>
      <a href="../articles/immunarch.html">Using scRepertoire with immunarch</a>
    </li>
    <li>
      <a href="../articles/dandelionR.html">VDJ Trajectory Analysis with dandelionR</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/BorchLab/scRepertoire/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header>





<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Starting work with scRepertoire.</h1>
                        <h4 data-toc-skip class="author">Nick
Borcherding</h4>
            <address class="author_afil">
      Washington University in St. Louis, School of Medicine, St. Louis,
MO, USA<br>            <a class="author_email" href="mailto:#"><a
href="mailto:ncborch@gmail.com" class="email">ncborch@gmail.com</a></a>
      </address>
                  
            <h4 data-toc-skip class="date">Compiled: October 31,
2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BorchLab/scRepertoire/blob/HEAD/vignettes/vignette.Rmd" class="external-link"><code>vignettes/vignette.Rmd</code></a></small>
      <div class="hidden name"><code>vignette.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>scRepertoire is designed to take filter contig outputs from the 10x
Genomics Cell Ranger pipeline, process that data to assign clonotype
based on two TCR or Ig chains and analyze the clonotype dynamics. The
latter can be separated into 1) clonotype-only analysis functions, such
as unique clonotypes or clonal space quantification, and 2) interaction
with mRNA expression data using Seurat or SingleCellExperiment
packages.</p>
<div id="loading-libraries" class="section level2">
<h2>Loading Libraries</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(scRepertoire))</span></code></pre></div>
</div>
<div id="loading-data" class="section level2">
<h2>Loading Data</h2>
<div id="what-data-to-load-into-screpertoire" class="section level3">
<h3>What data to load into scRepertoire?</h3>
<p><code>scRepertoire</code> primarily functions using the
<code>filtered_contig_annotations.csv</code> output generated by 10x
Genomics Cell Ranger. This file is typically found in the ./outs/
directory of your VDJ alignment folder.</p>
</div>
<div id="demonstrating-manual-data-loading-10x-genomics"
class="section level3">
<h3>Demonstrating Manual Data Loading (10x Genomics)</h3>
<p>To prepare data for scRepertoire from 10x Genomics outputs, you
would:</p>
<ul>
<li>Load the <code>filtered_contig_annotations.csv</code> file for each
of your samples.</li>
<li>Combine these loaded data frames into a single list in your R
environment.</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>S1 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;.../Sample1/outs/filtered_contig_annotations.csv&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>S2 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;.../Sample2/outs/filtered_contig_annotations.csv&quot;</span>)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>S3 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;.../Sample3/outs/filtered_contig_annotations.csv&quot;</span>)</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>S4 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;.../Sample4/outs/filtered_contig_annotations.csv&quot;</span>)</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>contig_list <span class="ot">&lt;-</span> <span class="fu">list</span>(S1, S2, S3, S4)</span></code></pre></div>
</div>
</div>
<div id="other-alignment-workflows" class="section level2">
<h2>Other alignment workflows</h2>
<p>Beyond the default 10x Genomics Cell Ranger pipeline outputs,
<code>scRepertoire</code> supports various other single-cell immune
receptor sequencing formats through the <code>loadContigs()</code>
function.</p>
<div id="supported-formats-and-expected-file-names"
class="section level3">
<h3>Supported Formats and Expected File Names</h3>
<ul>
<li><code>10X</code>: “filtered_contig_annotations.csv”</li>
<li><code>AIRR</code>: “airr_rearrangement.tsv”</li>
<li><code>BD</code>: “Contigs_AIRR.tsv”</li>
<li><code>Dandelion</code>: “all_contig_dandelion.tsv”</li>
<li><code>Immcantation</code>: “_data.tsv” (or similar)</li>
<li><code>JSON</code>: “.json”</li>
<li><code>MiXCR</code>: “clones.tsv”</li>
<li><code>ParseBio</code>: “barcode_report.tsv”</li>
<li><code>TRUST4</code>: “barcode_report.tsv”</li>
<li><code>WAT3R</code>: “barcode_results.csv”</li>
</ul>
<p>Key Parameter(s) for <code>loadContigs()</code></p>
<ul>
<li><code>input</code>: A directory path containing your contig files
(the function will recursively search) or a list/data frame of
pre-loaded contig data.</li>
<li><code>format</code>: A string specifying the data format (e.g.,
<code>10X</code>, <code>TRUST4</code>, <code>WAT3R</code>). If set to
“auto”, the function will attempt to automatically detect the format
based on file names or data structure.</li>
</ul>
<p>You can provide <code>loadContigs()</code> with a directory where
your sequencing experiments are located, and it will recursively load
and process the contig data based on the file names:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Directory example</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>contig.output <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;~/Documents/MyExperiment&quot;</span>)</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>contig.list <span class="ot">&lt;-</span> <span class="fu">loadContigs</span>(<span class="at">input =</span> contig.output, </span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>                           <span class="at">format =</span> <span class="st">&quot;TRUST4&quot;</span>)</span></code></pre></div>
<p>Alternatively, <code>loadContigs()</code> can be given a list of
pre-loaded data frames and process the contig data based on the
specified format:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># List of data frames example</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>S1 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;~/Documents/MyExperiment/Sample1/outs/barcode_results.csv&quot;</span>)</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>S2 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;~/Documents/MyExperiment/Sample2/outs/barcode_results.csv&quot;</span>)</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>S3 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;~/Documents/MyExperiment/Sample3/outs/barcode_results.csv&quot;</span>)</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>S4 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;~/Documents/MyExperiment/Sample4/outs/barcode_results.csv&quot;</span>)</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>contig.list <span class="ot">&lt;-</span> <span class="fu">list</span>(S1, S2, S3, S4)</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>contig.list <span class="ot">&lt;-</span> <span class="fu">loadContigs</span>(<span class="at">input =</span> contig.list, </span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>                           <span class="at">format =</span> <span class="st">&quot;WAT3R&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="multiplexed-experiment" class="section level2">
<h2>Multiplexed Experiment</h2>
<p>It is now easy to create the contig list from a multiplexed
experiment by first generating a single-cell RNA object (either Seurat
or Single Cell Experiment), loading the filtered contig file, and then
using <code>createHTOContigList()</code>. This function will return a
list separated by the <code>group.by</code> variable(s).</p>
<div id="important-considerations-for-createhtocontiglist"
class="section level3">
<h3>Important Considerations for <code>createHTOContigList()</code></h3>
<ul>
<li>This function depends on the match of barcodes between the
single-cell object and contigs. If there is a prefix or different suffix
added to the barcode, this will result in no contigs being
recovered.</li>
<li>It is currently recommended to perform this step before integration
workflows, as integration commonly alters the barcodes.</li>
<li>There is a multi.run variable that can be used on an integrated
object. However, it assumes you have modified the barcodes with the
Seurat pipeline (automatic addition of _# to the end), and your contig
list is in the same order.</li>
</ul>
<p>To create a contig list separated by HTO (Hash Tag Oligo) IDs from a
single-cell object:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>contigs <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;.../outs/filtered_contig_annotations.csv&quot;</span>)</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>contig.list <span class="ot">&lt;-</span> <span class="fu">createHTOContigList</span>(contigs, </span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>                                   Seurat.Obj, </span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>                                   <span class="at">group.by =</span> <span class="st">&quot;HTO_maxID&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="example-data-in-screpertoire" class="section level2">
<h2>Example Data in <code>scRepertoire</code></h2>
<p><code>scRepertoire</code> includes a built-in example dataset to
demonstrate the functionality of the R package. This dataset consists of
T cells derived from four patients with acute respiratory distress with
paired peripheral-blood (B) and bronchoalveolar lavage (L), effectively
creating 8 distinct runs for T cell receptor (TCR) enrichment. More
information on the data set can be found in the corresponding <a
href="https://pubmed.ncbi.nlm.nih.gov/33622974/">manuscript</a>.</p>
<p>The built-in example data is derived from the 10x Cell Ranger
pipeline, so it is ready to go for downstream processing and
analysis.</p>
<p>To load and preview the example data built into
<code>scRepertoire</code>:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;contig_list&quot;</span>) <span class="co">#the data built into scRepertoire</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="fu">head</span>(contig_list[[<span class="dv">1</span>]])</span></code></pre></div>
<pre><code>##              barcode is_cell                   contig_id high_confidence length
## 1 AAACCTGAGTACGACG-1    True AAACCTGAGTACGACG-1_contig_1            True    500
## 2 AAACCTGAGTACGACG-1    True AAACCTGAGTACGACG-1_contig_2            True    478
## 4 AAACCTGCAACACGCC-1    True AAACCTGCAACACGCC-1_contig_1            True    506
## 5 AAACCTGCAACACGCC-1    True AAACCTGCAACACGCC-1_contig_2            True    470
## 6 AAACCTGCAGGCGATA-1    True AAACCTGCAGGCGATA-1_contig_1            True    558
## 7 AAACCTGCAGGCGATA-1    True AAACCTGCAGGCGATA-1_contig_2            True    505
##   chain       v_gene d_gene  j_gene c_gene full_length productive
## 1   TRA       TRAV25   None  TRAJ20   TRAC        True       True
## 2   TRB      TRBV5-1   None TRBJ2-7  TRBC2        True       True
## 4   TRA TRAV38-2/DV8   None  TRAJ52   TRAC        True       True
## 5   TRB     TRBV10-3   None TRBJ2-2  TRBC2        True       True
## 6   TRA     TRAV12-1   None   TRAJ9   TRAC        True       True
## 7   TRB        TRBV9   None TRBJ2-2  TRBC2        True       True
##                 cdr3                                                cdr3_nt
## 1        CGCSNDYKLSF                      TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT
## 2     CASSLTDRTYEQYF             TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC
## 4 CAYRSAQAGGTSYGKLTF TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT
## 5      CAISEQGKGELFF                TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT
## 6     CVVSDNTGGFKTIF             TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT
## 7  CASSVRRERANTGELFF    TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
##   reads umis raw_clonotype_id         raw_consensus_id
## 1  8344    4     clonotype123 clonotype123_consensus_2
## 2 65390   38     clonotype123 clonotype123_consensus_1
## 4 18372    8     clonotype124 clonotype124_consensus_1
## 5 34054    9     clonotype124 clonotype124_consensus_2
## 6  5018    2       clonotype1   clonotype1_consensus_2
## 7 25110   11       clonotype1   clonotype1_consensus_1</code></pre>
<hr />
</div>
</div>
<div id="combining-contigs-into-clones" class="section level1">
<h1>Combining Contigs into Clones</h1>
<p>The <code>combineTCR()</code> function processes a list of TCR
sequencing results, consolidating them to the level of individual cell
barcodes. It handles potential issues with repeated barcodes by adding
prefixes from <code>samples</code> and <code>ID</code> parameters. The
output includes combined reads into clone calls by nucleotide sequence
(<code>CTnt</code>), amino acid sequence (<code>CTaa</code>), VDJC gene
sequence (<code>CTgene</code>), or a combination of nucleotide and gene
sequence (<code>CTstrict</code>).</p>
<p>Key Parameter(s) for <code>combineTCR()</code></p>
<ul>
<li><code>input.data</code>: A list of filtered contig annotations
(e.g., filtered_contig_annotations.csv from 10x Cell Ranger) or outputs
from <code>loadContigs()</code>.</li>
<li><code>samples</code>: Labels for your samples (recommended).</li>
<li><code>ID</code>: Additional sample labels (optional).</li>
<li><code>removeNA</code>: If <code>TRUE</code>, removes any cell
barcode with an NA value in at least one chain (default is
<code>FALSE</code>).</li>
<li><code>removeMulti</code>: If <code>TRUE</code>, removes any cell
barcode with more than two immune receptor chains (default is
<code>FALSE</code>).</li>
<li><code>filterMulti</code>: If <code>TRUE</code>, isolates the top two
expressed chains in cell barcodes with multiple chains (default is
<code>FALSE</code>).</li>
<li><code>filterNonproductive</code>: If <code>TRUE</code>, removes
non-productive chains if the variable exists in the contig data (default
is <code>TRUE</code>).</li>
</ul>
<p>To combine TCR contigs from <code>contig_list</code> and apply sample
prefixes:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>combined.TCR <span class="ot">&lt;-</span> <span class="fu">combineTCR</span>(contig_list, </span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>                           <span class="at">samples =</span> <span class="fu">c</span>(<span class="st">&quot;P17B&quot;</span>, <span class="st">&quot;P17L&quot;</span>, <span class="st">&quot;P18B&quot;</span>, <span class="st">&quot;P18L&quot;</span>, </span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>                                            <span class="st">&quot;P19B&quot;</span>,<span class="st">&quot;P19L&quot;</span>, <span class="st">&quot;P20B&quot;</span>, <span class="st">&quot;P20L&quot;</span>),</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>                           <span class="at">removeNA =</span> <span class="cn">FALSE</span>, </span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>                           <span class="at">removeMulti =</span> <span class="cn">FALSE</span>, </span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>                           <span class="at">filterMulti =</span> <span class="cn">FALSE</span>)</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="fu">head</span>(combined.TCR[[<span class="dv">1</span>]])</span></code></pre></div>
<pre><code>##                    barcode sample                     TCR1           cdr3_aa1
## 1  P17B_AAACCTGAGTACGACG-1   P17B       TRAV25.TRAJ20.TRAC        CGCSNDYKLSF
## 3  P17B_AAACCTGCAACACGCC-1   P17B TRAV38-2/DV8.TRAJ52.TRAC CAYRSAQAGGTSYGKLTF
## 5  P17B_AAACCTGCAGGCGATA-1   P17B      TRAV12-1.TRAJ9.TRAC     CVVSDNTGGFKTIF
## 7  P17B_AAACCTGCATGAGCGA-1   P17B      TRAV12-1.TRAJ9.TRAC     CVVSDNTGGFKTIF
## 9  P17B_AAACGGGAGAGCCCAA-1   P17B        TRAV20.TRAJ8.TRAC      CAVRGEGFQKLVF
## 10 P17B_AAACGGGAGCGTTTAC-1   P17B      TRAV12-1.TRAJ9.TRAC     CVVSDNTGGFKTIF
##                                                  cdr3_nt1
## 1                       TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT
## 3  TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT
## 5              TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT
## 7              TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT
## 9                 TGTGCTGTGCGAGGAGAAGGCTTTCAGAAACTTGTATTT
## 10             TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT
##                           TCR2          cdr3_aa2
## 1   TRBV5-1.None.TRBJ2-7.TRBC2    CASSLTDRTYEQYF
## 3  TRBV10-3.None.TRBJ2-2.TRBC2     CAISEQGKGELFF
## 5     TRBV9.None.TRBJ2-2.TRBC2 CASSVRRERANTGELFF
## 7     TRBV9.None.TRBJ2-2.TRBC2 CASSVRRERANTGELFF
## 9                         &lt;NA&gt;              &lt;NA&gt;
## 10    TRBV9.None.TRBJ2-2.TRBC2 CASSVRRERANTGELFF
##                                               cdr3_nt2
## 1           TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC
## 3              TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT
## 5  TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 7  TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 9                                                 &lt;NA&gt;
## 10 TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
##                                                  CTgene
## 1         TRAV25.TRAJ20.TRAC_TRBV5-1.None.TRBJ2-7.TRBC2
## 3  TRAV38-2/DV8.TRAJ52.TRAC_TRBV10-3.None.TRBJ2-2.TRBC2
## 5          TRAV12-1.TRAJ9.TRAC_TRBV9.None.TRBJ2-2.TRBC2
## 7          TRAV12-1.TRAJ9.TRAC_TRBV9.None.TRBJ2-2.TRBC2
## 9                                  TRAV20.TRAJ8.TRAC_NA
## 10         TRAV12-1.TRAJ9.TRAC_TRBV9.None.TRBJ2-2.TRBC2
##                                                                                              CTnt
## 1                    TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT_TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC
## 3  TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT_TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT
## 5  TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 7  TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 9                                                      TGTGCTGTGCGAGGAGAAGGCTTTCAGAAACTTGTATTT_NA
## 10 TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
##                                CTaa
## 1        CGCSNDYKLSF_CASSLTDRTYEQYF
## 3  CAYRSAQAGGTSYGKLTF_CAISEQGKGELFF
## 5  CVVSDNTGGFKTIF_CASSVRRERANTGELFF
## 7  CVVSDNTGGFKTIF_CASSVRRERANTGELFF
## 9                  CAVRGEGFQKLVF_NA
## 10 CVVSDNTGGFKTIF_CASSVRRERANTGELFF
##                                                                                                                                               CTstrict
## 1                           TRAV25.TRAJ20.TRAC;TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT_TRBV5-1.None.TRBJ2-7.TRBC2;TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC
## 3  TRAV38-2/DV8.TRAJ52.TRAC;TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT_TRBV10-3.None.TRBJ2-2.TRBC2;TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT
## 5          TRAV12-1.TRAJ9.TRAC;TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TRBV9.None.TRBJ2-2.TRBC2;TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 7          TRAV12-1.TRAJ9.TRAC;TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TRBV9.None.TRBJ2-2.TRBC2;TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 9                                                                                      TRAV20.TRAJ8.TRAC;TGTGCTGTGCGAGGAGAAGGCTTTCAGAAACTTGTATTT_NA;NA
## 10         TRAV12-1.TRAJ9.TRAC;TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TRBV9.None.TRBJ2-2.TRBC2;TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT</code></pre>
<p><code>combineTCR()</code> is the essential first step for organizing
raw TCR sequencing data into a structured format for
<code>scRepertoire</code> analyses. It allows for flexible handling of
single and paired chains, barcode disambiguation, and initial filtering,
producing a list of data frames where each row represents a single cell
and its associated TCR clonotypes.</p>
<div id="combinebcr" class="section level2">
<h2>combineBCR</h2>
<p>The <code>combineBCR()</code> function is the primary tool for
processing raw B cell receptor contig data into a format ready for
analysis. It is analogous to <code>combineTCR()</code> but includes
specialized logic for handling the complexities of BCRs, such as somatic
hypermutation. The function consolidates contigs into a single data
frame per sample, with each row representing a unique cell.</p>
<p>By default <code>(call.related.clones = TRUE)</code>,
<code>combineBCR()</code> groups B cells into clones based on the
similarity of their CDR3 sequences.</p>
<div id="how-combinebcr-groups-related-clones" class="section level3">
<h3>How <code>combineBCR()</code> Groups Related Clones</h3>
<ul>
<li>Internally calling <code>clonalCluster()</code> to build a network
of related sequences.</li>
<li>Using the <code>threshold</code> parameter to define connections.
The <code>threshold</code> is a normalized Levenshtein distance, where a
value closer to 1.0 requires higher sequence similarity. The default of
0.85 is a good starting point.</li>
<li>Assigning a cluster-based ID to the CTstrict column.</li>
</ul>
<p>Additionally, the <code>group.by</code> argument allows you to
constrain the clustering analysis to only occur within distinct
categories in your metadata. For example, using
<code>group.by = "sample"</code> ensures that sequences from different
samples are never compared or clustered together, even if they are
identical.</p>
<p>Key Parameter(s) for <code>combineBCR()</code></p>
<ul>
<li><code>call.related.clones</code>: If <code>TRUE</code> (default),
uses <code>clonalCluster()</code> to identify related clones based on
sequence similarity.</li>
<li><code>group.by</code>: The column header used to group clones for
clustering (if <code>NULL</code>, clusters will be calculated across all
samples).</li>
<li><code>threshold</code>: The similarity threshold for
<code>clonalCluster()</code> (default: 0.85).</li>
<li><code>chain</code>: The chain to use for clustering when
call.related.clones = TRUE (default: <code>both</code>).</li>
<li><code>sequence</code>: The sequence type (<code>nt</code> or
<code>aa</code>) for clustering (default: <code>nt</code>).</li>
<li>use.V, use.J: If <code>TRUE</code>, sequences must share the same
V/J gene to be clustered (default: <code>TRUE</code>)</li>
<li><code>cluster.method</code>: The clustering algorithm to apply to
the edit-distanc network (default: <code>components</code>).</li>
</ul>
<p>First, load the example BCR contig data:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co"># Load example BCR contig data</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>BCR.contigs <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;https://www.borch.dev/uploads/contigs/b_contigs.csv&quot;</span>)</span></code></pre></div>
<p>Then, combine BCR contigs using the default similarity
clustering:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="co"># Combine using the default similarity clustering</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>combined.BCR.clustered <span class="ot">&lt;-</span> <span class="fu">combineBCR</span>(BCR.contigs, </span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>                                     <span class="at">samples =</span> <span class="st">&quot;Patient1&quot;</span>, </span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>                                     <span class="at">threshold =</span> <span class="fl">0.85</span>)</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a><span class="co"># The CTstrict column contains cluster IDs (e.g., &quot;cluster.1&quot;)</span></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a><span class="fu">head</span>(combined.BCR.clustered[[<span class="dv">1</span>]][, <span class="fu">c</span>(<span class="st">&quot;barcode&quot;</span>, <span class="st">&quot;CTstrict&quot;</span>, <span class="st">&quot;IGH&quot;</span>, <span class="st">&quot;cdr3_aa1&quot;</span>)])</span></code></pre></div>
<pre><code>##                       barcode CTstrict                            IGH
## 1 Patient1_AAACCTGAGGGCACTA-1     &lt;NA&gt;                           &lt;NA&gt;
## 2 Patient1_AAACCTGAGTACGTTC-1     &lt;NA&gt;   IGHV3-64.IGHD6-6.IGHJ4.IGHG1
## 3 Patient1_AAACCTGAGTCCGGTC-1     &lt;NA&gt;                           &lt;NA&gt;
## 4 Patient1_AAACCTGCACCAGGTC-1     &lt;NA&gt; IGHV1-69-2.IGHD3-10.IGHJ6.IGHM
## 5 Patient1_AAACCTGGTAAATGAC-1     &lt;NA&gt;  IGHV1-46.IGHD6-13.IGHJ5.IGHA2
## 6 Patient1_AAACGGGTCACCTCGT-1     &lt;NA&gt;                           &lt;NA&gt;
##                cdr3_aa1
## 1                  &lt;NA&gt;
## 2      CAKSYSRDLPRYFGSW
## 3                  &lt;NA&gt;
## 4 CVRDGAGHYGSGIGYYGMDVW
## 5    CSRGRLPMPAAGSGLVSW
## 6                  &lt;NA&gt;</code></pre>
</div>
<div id="filtering-and-cleaning-data" class="section level3">
<h3>Filtering and Cleaning Data</h3>
<p><code>combineBCR()</code> includes several arguments to filter and
clean the contig data during processing. *
<code>filterNonproductive = TRUE</code> (Default): Removes any contigs
that are not classified as productive, ensuring that only functional
receptor chains are included in the analysis. *
<code>filterMulti = TRUE</code> (Default): For cells with more than one
heavy or light chain detected, this automatically selects the chain with
the highest UMI count (read abundance) and discards the others. This
helps resolve cellular multiplets or technical artifacts.</p>
<p>Here is an example applying these filters (though they are on by
default):</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>cleaned.BCR <span class="ot">&lt;-</span> <span class="fu">combineBCR</span>(BCR.contigs,</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>                          <span class="at">samples =</span> <span class="st">&quot;Patient1&quot;</span>,</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>                          <span class="at">filterNonproductive =</span> <span class="cn">TRUE</span>,</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>                          <span class="at">filterMulti =</span> <span class="cn">TRUE</span>)</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a></span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a><span class="fu">head</span>(cleaned.BCR[[<span class="dv">1</span>]])</span></code></pre></div>
<pre><code>##                       barcode   sample                            IGH
## 1 Patient1_AAACCTGAGGGCACTA-1 Patient1                           &lt;NA&gt;
## 2 Patient1_AAACCTGAGTACGTTC-1 Patient1   IGHV3-64.IGHD6-6.IGHJ4.IGHG1
## 3 Patient1_AAACCTGAGTCCGGTC-1 Patient1                           &lt;NA&gt;
## 4 Patient1_AAACCTGCACCAGGTC-1 Patient1 IGHV1-69-2.IGHD3-10.IGHJ6.IGHM
## 5 Patient1_AAACCTGGTAAATGAC-1 Patient1  IGHV1-46.IGHD6-13.IGHJ5.IGHA2
## 6 Patient1_AAACGGGTCACCTCGT-1 Patient1                           &lt;NA&gt;
##                cdr3_aa1
## 1                  &lt;NA&gt;
## 2      CAKSYSRDLPRYFGSW
## 3                  &lt;NA&gt;
## 4 CVRDGAGHYGSGIGYYGMDVW
## 5    CSRGRLPMPAAGSGLVSW
## 6                  &lt;NA&gt;
##                                                          cdr3_nt1
## 1                                                            &lt;NA&gt;
## 2                TGTGCGAAATCGTATAGCAGAGACCTGCCGCGGTACTTTGGCTCCTGG
## 3                                                            &lt;NA&gt;
## 4 TGTGTGAGAGATGGGGCGGGTCACTATGGTTCGGGGATAGGCTACTACGGTATGGACGTCTGG
## 5          TGTTCGAGGGGTCGTCTCCCGATGCCAGCAGCTGGCAGTGGTTTGGTCTCCTGG
## 6                                                            &lt;NA&gt;
##                   IGLC       cdr3_aa2
## 1 IGLV1-51.IGLJ3.IGLC2  CGTWDSSLSAGVF
## 2  IGKV3-15.IGKJ4.IGKC    CQQYSNWPLTF
## 3  IGKV3-20.IGKJ4.IGKC     CQQYGDSLPF
## 4 IGLV2-14.IGLJ2.IGLC2 CSSYISTSTLEVLF
## 5  IGKV3-15.IGKJ2.IGKC   CQHYHNWPPYTF
## 6  IGKV3-11.IGKJ1.IGKC    CQQRSNWPLTF
##                                     cdr3_nt2
## 1    TGCGGAACATGGGATAGCAGCCTGAGTGCTGGCGTGTTC
## 2          TGTCAGCAGTATAGTAACTGGCCGCTCACTTTC
## 3             TGTCAACAGTATGGTGACTCTCTCCCTTTC
## 4 TGCAGTTCATATATAAGTACCAGCACTCTCGAGGTCCTATTC
## 5       TGTCAGCACTATCATAACTGGCCTCCGTACACTTTT
## 6          TGTCAGCAGCGTAGCAACTGGCCTCTGACGTTC
##                                                CTgene
## 1                             NA_IGLV1-51.IGLJ3.IGLC2
## 2    IGHV3-64.IGHD6-6.IGHJ4.IGHG1_IGKV3-15.IGKJ4.IGKC
## 3                              NA_IGKV3-20.IGKJ4.IGKC
## 4 IGHV1-69-2.IGHD3-10.IGHJ6.IGHM_IGLV2-14.IGLJ2.IGLC2
## 5   IGHV1-46.IGHD6-13.IGHJ5.IGHA2_IGKV3-15.IGKJ2.IGKC
## 6                              NA_IGKV3-11.IGKJ1.IGKC
##                                                                                                         CTnt
## 1                                                                 NA_TGCGGAACATGGGATAGCAGCCTGAGTGCTGGCGTGTTC
## 2                         TGTGCGAAATCGTATAGCAGAGACCTGCCGCGGTACTTTGGCTCCTGG_TGTCAGCAGTATAGTAACTGGCCGCTCACTTTC
## 3                                                                          NA_TGTCAACAGTATGGTGACTCTCTCCCTTTC
## 4 TGTGTGAGAGATGGGGCGGGTCACTATGGTTCGGGGATAGGCTACTACGGTATGGACGTCTGG_TGCAGTTCATATATAAGTACCAGCACTCTCGAGGTCCTATTC
## 5                TGTTCGAGGGGTCGTCTCCCGATGCCAGCAGCTGGCAGTGGTTTGGTCTCCTGG_TGTCAGCACTATCATAACTGGCCTCCGTACACTTTT
## 6                                                                       NA_TGTCAGCAGCGTAGCAACTGGCCTCTGACGTTC
##                                   CTaa CTstrict
## 1                     NA_CGTWDSSLSAGVF     &lt;NA&gt;
## 2         CAKSYSRDLPRYFGSW_CQQYSNWPLTF     &lt;NA&gt;
## 3                        NA_CQQYGDSLPF     &lt;NA&gt;
## 4 CVRDGAGHYGSGIGYYGMDVW_CSSYISTSTLEVLF     &lt;NA&gt;
## 5      CSRGRLPMPAAGSGLVSW_CQHYHNWPPYTF     &lt;NA&gt;
## 6                       NA_CQQRSNWPLTF     &lt;NA&gt;</code></pre>
<p><code>combineBCR()</code> is designed for processing B cell
repertoire data, going beyond simple contig aggregation to incorporate
advanced clustering based on CDR3 sequence similarity. This enables the
identification of clonally related B cells, crucial for studying B cell
development, affinity maturation, and humoral immune responses. Its
filtering options further ensure the quality and interpretability of the
processed data.</p>
<hr />
</div>
</div>
</div>
<div id="additional-processing" class="section level1">
<h1>Additional Processing</h1>
<div id="addvariable-adding-variables-for-plotting"
class="section level2">
<h2>addVariable: Adding Variables for Plotting</h2>
<p>What if there are more variables to add than just sample and ID? We
can add them by using the <code>addVariable()</code> function. For each
element, the function will add a column (labeled by
<code>variable.name</code>) with the <code>variable</code>. The length
of the <code>variables</code> parameter needs to match the length of the
combined object.</p>
<p>Key Parameter(s) for <code>addVariable()</code></p>
<ul>
<li><code>variable.name</code>: A character string that defines the new
variable to add (e.g., “Type”, “Treatment”).</li>
<li><code>variables</code>: A character vector defining the desired
column value for each list element. Its length must match the number of
elements in the input.data list.</li>
</ul>
<p>As an example, here we add the Type in which the samples were
processed and sequenced to the <code>combined.TCR</code> object:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>combined.TCR <span class="ot">&lt;-</span> <span class="fu">addVariable</span>(combined.TCR, </span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>                            <span class="at">variable.name =</span> <span class="st">&quot;Type&quot;</span>, </span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>                            <span class="at">variables =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">&quot;B&quot;</span>, <span class="st">&quot;L&quot;</span>), <span class="dv">4</span>))</span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a></span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a><span class="fu">head</span>(combined.TCR[[<span class="dv">1</span>]])</span></code></pre></div>
<pre><code>##                    barcode sample                     TCR1           cdr3_aa1
## 1  P17B_AAACCTGAGTACGACG-1   P17B       TRAV25.TRAJ20.TRAC        CGCSNDYKLSF
## 3  P17B_AAACCTGCAACACGCC-1   P17B TRAV38-2/DV8.TRAJ52.TRAC CAYRSAQAGGTSYGKLTF
## 5  P17B_AAACCTGCAGGCGATA-1   P17B      TRAV12-1.TRAJ9.TRAC     CVVSDNTGGFKTIF
## 7  P17B_AAACCTGCATGAGCGA-1   P17B      TRAV12-1.TRAJ9.TRAC     CVVSDNTGGFKTIF
## 9  P17B_AAACGGGAGAGCCCAA-1   P17B        TRAV20.TRAJ8.TRAC      CAVRGEGFQKLVF
## 10 P17B_AAACGGGAGCGTTTAC-1   P17B      TRAV12-1.TRAJ9.TRAC     CVVSDNTGGFKTIF
##                                                  cdr3_nt1
## 1                       TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT
## 3  TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT
## 5              TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT
## 7              TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT
## 9                 TGTGCTGTGCGAGGAGAAGGCTTTCAGAAACTTGTATTT
## 10             TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT
##                           TCR2          cdr3_aa2
## 1   TRBV5-1.None.TRBJ2-7.TRBC2    CASSLTDRTYEQYF
## 3  TRBV10-3.None.TRBJ2-2.TRBC2     CAISEQGKGELFF
## 5     TRBV9.None.TRBJ2-2.TRBC2 CASSVRRERANTGELFF
## 7     TRBV9.None.TRBJ2-2.TRBC2 CASSVRRERANTGELFF
## 9                         &lt;NA&gt;              &lt;NA&gt;
## 10    TRBV9.None.TRBJ2-2.TRBC2 CASSVRRERANTGELFF
##                                               cdr3_nt2
## 1           TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC
## 3              TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT
## 5  TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 7  TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 9                                                 &lt;NA&gt;
## 10 TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
##                                                  CTgene
## 1         TRAV25.TRAJ20.TRAC_TRBV5-1.None.TRBJ2-7.TRBC2
## 3  TRAV38-2/DV8.TRAJ52.TRAC_TRBV10-3.None.TRBJ2-2.TRBC2
## 5          TRAV12-1.TRAJ9.TRAC_TRBV9.None.TRBJ2-2.TRBC2
## 7          TRAV12-1.TRAJ9.TRAC_TRBV9.None.TRBJ2-2.TRBC2
## 9                                  TRAV20.TRAJ8.TRAC_NA
## 10         TRAV12-1.TRAJ9.TRAC_TRBV9.None.TRBJ2-2.TRBC2
##                                                                                              CTnt
## 1                    TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT_TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC
## 3  TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT_TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT
## 5  TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 7  TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 9                                                      TGTGCTGTGCGAGGAGAAGGCTTTCAGAAACTTGTATTT_NA
## 10 TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
##                                CTaa
## 1        CGCSNDYKLSF_CASSLTDRTYEQYF
## 3  CAYRSAQAGGTSYGKLTF_CAISEQGKGELFF
## 5  CVVSDNTGGFKTIF_CASSVRRERANTGELFF
## 7  CVVSDNTGGFKTIF_CASSVRRERANTGELFF
## 9                  CAVRGEGFQKLVF_NA
## 10 CVVSDNTGGFKTIF_CASSVRRERANTGELFF
##                                                                                                                                               CTstrict
## 1                           TRAV25.TRAJ20.TRAC;TGTGGGTGTTCTAACGACTACAAGCTCAGCTTT_TRBV5-1.None.TRBJ2-7.TRBC2;TGCGCCAGCAGCTTGACCGACAGGACCTACGAGCAGTACTTC
## 3  TRAV38-2/DV8.TRAJ52.TRAC;TGTGCTTATAGGAGCGCGCAGGCTGGTGGTACTAGCTATGGAAAGCTGACATTT_TRBV10-3.None.TRBJ2-2.TRBC2;TGTGCCATCAGTGAACAGGGGAAAGGGGAGCTGTTTTTT
## 5          TRAV12-1.TRAJ9.TRAC;TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TRBV9.None.TRBJ2-2.TRBC2;TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 7          TRAV12-1.TRAJ9.TRAC;TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TRBV9.None.TRBJ2-2.TRBC2;TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
## 9                                                                                      TRAV20.TRAJ8.TRAC;TGTGCTGTGCGAGGAGAAGGCTTTCAGAAACTTGTATTT_NA;NA
## 10         TRAV12-1.TRAJ9.TRAC;TGTGTGGTCTCCGATAATACTGGAGGCTTCAAAACTATCTTT_TRBV9.None.TRBJ2-2.TRBC2;TGTGCCAGCAGCGTAAGGAGGGAAAGGGCGAACACCGGGGAGCTGTTTTTT
##    Type
## 1     B
## 3     B
## 5     B
## 7     B
## 9     B
## 10    B</code></pre>
</div>
<div id="subsetclones-filter-out-clonal-information"
class="section level2">
<h2>subsetClones: Filter Out Clonal Information</h2>
<p>Likewise, we can remove specific list elements after
<code>combineTCR()</code> or <code>combineBCR()</code> using the
<code>subsetClones()</code> function. In order to subset, we need to
identify the column header we would like to use for subsetting
(<code>name</code>) and the specific values to include
(<code>variables</code>).</p>
<p>Key Parameter(s) for <code>subsetClones()</code></p>
<ul>
<li><code>name</code>: The column header/name in the metadata of
input.data to use for subsetting (e.g., “sample”, “Type”).</li>
<li><code>variables</code>: A character vector of the specific values
within the chosen name column to retain in the subsetted data.</li>
</ul>
<p>Below, we isolate just the two sequencing results from “P18L” and
“P18B” samples:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a>subset1 <span class="ot">&lt;-</span> <span class="fu">subsetClones</span>(combined.TCR, </span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a>                        <span class="at">name =</span> <span class="st">&quot;sample&quot;</span>, </span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a>                        <span class="at">variables =</span> <span class="fu">c</span>(<span class="st">&quot;P18L&quot;</span>, <span class="st">&quot;P18B&quot;</span>))</span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a></span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a><span class="fu">head</span>(subset1[[<span class="dv">1</span>]][,<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>])</span></code></pre></div>
<pre><code>##                    barcode sample                 TCR1         cdr3_aa1
## 1  P18B_AAACCTGAGGCTCAGA-1   P18B TRAV26-1.TRAJ37.TRAC  CIVRGGSSNTGKLIF
## 3  P18B_AAACCTGCATGACATC-1   P18B    TRAV3.TRAJ20.TRAC    CAVQRSNDYKLSF
## 5  P18B_AAACCTGGTATGCTTG-1   P18B TRAV26-1.TRAJ53.TRAC   CIGSSGGSNYKLTF
## 8  P18B_AAACGGGCAGATGGGT-1   P18B                 &lt;NA&gt;             &lt;NA&gt;
## 9  P18B_AAACGGGTCTTACCGC-1   P18B    TRAV20.TRAJ9.TRAC CAVQAKRYTGGFKTIF
## 12 P18B_AAAGATGAGTTACGGG-1   P18B   TRAV8-3.TRAJ8.TRAC   CAVGGDTGFQKLVF</code></pre>
<p>Alternatively, we can also just select the list elements after
<code>combineTCR()</code> or <code>combineBCR()</code>.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a>subset2 <span class="ot">&lt;-</span> combined.TCR[<span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">4</span>)]</span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a><span class="fu">head</span>(subset2[[<span class="dv">1</span>]][,<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>])</span></code></pre></div>
<pre><code>##                    barcode sample                 TCR1         cdr3_aa1
## 1  P18B_AAACCTGAGGCTCAGA-1   P18B TRAV26-1.TRAJ37.TRAC  CIVRGGSSNTGKLIF
## 3  P18B_AAACCTGCATGACATC-1   P18B    TRAV3.TRAJ20.TRAC    CAVQRSNDYKLSF
## 5  P18B_AAACCTGGTATGCTTG-1   P18B TRAV26-1.TRAJ53.TRAC   CIGSSGGSNYKLTF
## 8  P18B_AAACGGGCAGATGGGT-1   P18B                 &lt;NA&gt;             &lt;NA&gt;
## 9  P18B_AAACGGGTCTTACCGC-1   P18B    TRAV20.TRAJ9.TRAC CAVQAKRYTGGFKTIF
## 12 P18B_AAAGATGAGTTACGGG-1   P18B   TRAV8-3.TRAJ8.TRAC   CAVGGDTGFQKLVF</code></pre>
</div>
<div id="exportclones-save-clonal-data" class="section level2">
<h2>exportClones: Save Clonal Data</h2>
<p>After assigning the clone by barcode, we can export the clonal
information using <code>exportClones()</code> to save for later use or
to integrate with other bioinformatics pipelines. This function supports
various output formats tailored for different analytical needs.</p>
<p>Key Parameter(s) for <code>exportClones()</code> *
<code>format</code>: The desired output format for the clonal data. *
<code>airr</code>: Exports data in an Adaptive Immune Receptor
Repertoire (AIRR) Community-compliant format, with each row representing
a single receptor chain. * <code>immunarch</code>: Exports a list
containing a data frame and metadata formatted for use with the
<code>immunarch</code> package. * <code>paired</code>: Exports a data
frame with paired chain information (amino acid, nucleotide, genes) per
barcode. This is the default. * <code>TCRMatch</code>: Exports a data
frame specifically for the TCRMatch algorithm, containing TRB chain
amino acid sequence and clonal frequency. * <code>tcrpheno</code>:
Exports a data frame compatible with the <code>tcrpheno</code> pipeline,
with TRA and TRB chains in separate columns. * <code>write.file</code>:
If <code>TRUE</code> (default), saves the output to a CSV file. If
<code>FALSE</code>, returns the data frame or list to the R environment.
* <code>dir</code>: The directory where the output file will be saved.
Defaults to the current working directory. * <code>file.name</code>: The
name of the CSV file to be saved.</p>
<p>To export the combined clonotypes as a <code>paired</code> data frame
and save it to a specified directory:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="fu">exportClones</span>(combined, </span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a>             <span class="at">write.file =</span> <span class="cn">TRUE</span>,</span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a>             <span class="at">dir =</span> <span class="st">&quot;~/Documents/MyExperiment/Sample1/&quot;</span></span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a>             <span class="at">file.name =</span> <span class="st">&quot;clones.csv&quot;</span>)</span></code></pre></div>
<p>To return an <code>immunarch</code>-formatted data frame directly to
your R environment without saving a file:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a>immunarch <span class="ot">&lt;-</span> <span class="fu">exportClones</span>(combined.TCR, </span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>                          <span class="at">format =</span> <span class="st">&quot;immunarch&quot;</span>, </span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a>                          <span class="at">write.file =</span> <span class="cn">FALSE</span>)</span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a><span class="fu">head</span>(immunarch[[<span class="dv">1</span>]][[<span class="dv">1</span>]])</span></code></pre></div>
<pre><code>##   Clones   Proportion                                             CDR3.nt
## 1      1 0.0003565062    NA;TGCGCCAGCAGTCGGGGACTAGCGGGATACAATGAGCAGTTCTTC
## 2      1 0.0003565062       NA;TGTGCCATCAGCGCGGACCCCCGCTACAATGAGCAGTTCTTC
## 3      1 0.0003565062 NA;TGTGCCAGCAGCTTGAGGGACAGCTATCGGTACTATGGCTACACCTTC
## 4      2 0.0007130125          NA;TGTGCCAGCAGCCGGCAGGGCGCAGATACGCAGTATTTT
## 5      1 0.0003565062       NA;TGTGCCAGCAGTCCCTTTACAGGGTTCTATGGCTACACCTTC
## 6      1 0.0003565062          NA;TGTGCCAGCTCATCCGGGATCAATCAGCCCCAGCATTTT
##               CDR3.aa      V.name  D.name     J.name   C.name
## 1  NA;CASSRGLAGYNEQFF NA;TRBV10-2 NA;None NA;TRBJ2-1 NA;TRBC2
## 2   NA;CAISADPRYNEQFF NA;TRBV10-3 NA;None NA;TRBJ2-1 NA;TRBC2
## 3 NA;CASSLRDSYRYYGYTF NA;TRBV11-3 NA;None NA;TRBJ1-2 NA;TRBC1
## 4    NA;CASSRQGADTQYF NA;TRBV11-3 NA;None NA;TRBJ2-3 NA;TRBC2
## 5   NA;CASSPFTGFYGYTF NA;TRBV12-4 NA;None NA;TRBJ1-2 NA;TRBC1
## 6    NA;CASSSGINQPQHF   NA;TRBV18 NA;None NA;TRBJ1-5 NA;TRBC1
##                                           Barcode
## 1                         P17B_AGCGGTCCAAAGGAAG-1
## 2                         P17B_GGCTCGAGTCGCGGTT-1
## 3                         P17B_CGCGTTTTCGGCTACG-1
## 4 P17B_CACCAGGGTTCCTCCA-1;P17B_TCTATTGCAGGTGCCT-1
## 5                         P17B_GCTGGGTGTACGAAAT-1
## 6                         P17B_AGGGTGACATTGGTAC-1</code></pre>
</div>
<div id="annotateinvariant" class="section level2">
<h2>annotateInvariant</h2>
<p>The <code>annotateInvariant()</code> function enables the
identification of mucosal-associated invariant T (<code>MAIT</code>)
cells and invariant natural killer T (<code>iNKT</code>) cells in
single-cell sequencing datasets. These specialized T-cell subsets are
defined by their characteristic TCR usage, making them distinguishable
within single-cell immune profiling data. The function extracts TCR
chain information from the provided single-cell dataset and evaluates it
against known invariant TCR criteria for either MAIT or iNKT cells. Each
cell is assigned a score indicating the presence (1) or absence (0) of
the specified invariant T-cell population.</p>
<p>Key Parameter(s) for <code>annotateInavriant()</code></p>
<ul>
<li><code>type</code>: Character string specifying the type of invariant
T cell to annotate (<code>MAIT</code> or <code>iNKT</code>).</li>
<li><code>species</code>: Character string specifying the species
(<code>mouse</code> or <code>human</code>).</li>
</ul>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a>combined <span class="ot">&lt;-</span> <span class="fu">annotateInvariant</span>(combined, </span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a>                              <span class="at">type =</span> <span class="st">&quot;MAIT&quot;</span>, </span>
<span id="cb24-3"><a href="#cb24-3" tabindex="-1"></a>                              <span class="at">species =</span> <span class="st">&quot;human&quot;</span>)</span>
<span id="cb24-4"><a href="#cb24-4" tabindex="-1"></a></span>
<span id="cb24-5"><a href="#cb24-5" tabindex="-1"></a>combined <span class="ot">&lt;-</span> <span class="fu">annotateInvariant</span>(combined, </span>
<span id="cb24-6"><a href="#cb24-6" tabindex="-1"></a>                              <span class="at">type =</span> <span class="st">&quot;iNKT&quot;</span>, </span>
<span id="cb24-7"><a href="#cb24-7" tabindex="-1"></a>                              <span class="at">species =</span> <span class="st">&quot;human&quot;</span>)</span></code></pre></div>
<hr />
</div>
</div>
<div id="basic-clonal-visualizations" class="section level1">
<h1>Basic Clonal Visualizations</h1>
<div id="clonecall-how-to-call-clones." class="section level2">
<h2>cloneCall: How to call clones.</h2>
<ul>
<li><code>gene</code> - use the VDJC genes comprising the TCR/Ig<br />
</li>
<li><code>nt</code> - use the nucleotide sequence of the CDR3
region<br />
</li>
<li><code>aa</code> - use the amino acid sequence of the CDR3
region<br />
</li>
<li><code>strict</code> - use the VDJC genes comprising the TCR + the
nucleotide sequence of the CDR3 region. This is the <a
href="https://www.ncbi.nlm.nih.gov/pubmed/19568742">proper definition of
clonotype</a>. For <code>combineBCR()</code> strict refers to the edit
distance clusters + Vgene of the Ig.</li>
</ul>
<p>It is important to note that the clonotype is called using
essentially the combination of genes or nt/aa CDR3 sequences for both
loci. As of this implementation of scRepertoire, clonotype calling is
not incorporating small variations within the CDR3 sequences. As such
the <code>gene</code> approach will be the most sensitive, while the use
of <code>nt</code> or <code>aa</code> is moderately so, and the most
specific for clonotypes being <code>strict</code>. Additionally, the
clonotype call is trying to incorporate both loci, <em>i.e.</em>, both
<strong>TCRA</strong> and <strong>TCRB</strong> chains and if a single
cell barcode has multiple sequences identified (<em>i.e.</em>, 2 TCRA
chains expressed in one cell). Using the 10x approach, there is a subset
of barcodes that only return one of the immune receptor chains. The
unreturned chain is assigned an <strong>NA</strong> value.</p>
</div>
<div id="clonalquant-quantifying-unique-clones" class="section level2">
<h2>clonalQuant: Quantifying Unique Clones</h2>
<p>The <code>clonalQuant()</code> function is used to explore the clones
by returning the total or relative numbers of unique clones.</p>
<p>Key Parameter(s) for <code>clonalQuant()</code></p>
<ul>
<li><code>scale</code>: If <code>TRUE</code>, converts the output to the
relative percentage of unique clones scaled by the total repertoire
size; if <code>FALSE</code> (default), reports the total number of
unique clones.</li>
</ul>
<p>To visualize the relative percent of unique clones across all chains
(<code>"both"</code>) using the <code>strict</code> clone
definition:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="fu">clonalQuant</span>(combined.TCR, </span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a>            <span class="at">cloneCall=</span><span class="st">&quot;strict&quot;</span>, </span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a>            <span class="at">chain =</span> <span class="st">&quot;both&quot;</span>, </span>
<span id="cb25-4"><a href="#cb25-4" tabindex="-1"></a>            <span class="at">scale =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-18-1.png" width="700" /></p>
<p>Another option is to define the visualization by data classes using
the <code>group.by</code> parameter. Here, we’ll use the
<code>"Type"</code> variable, which was previously added to the
combined.TCR list.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="fu">clonalQuant</span>(combined.TCR, </span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a>            <span class="at">cloneCall=</span><span class="st">&quot;gene&quot;</span>, </span>
<span id="cb26-3"><a href="#cb26-3" tabindex="-1"></a>            <span class="at">group.by =</span> <span class="st">&quot;Type&quot;</span>, </span>
<span id="cb26-4"><a href="#cb26-4" tabindex="-1"></a>            <span class="at">scale =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-19-1.png" width="700" /></p>
</div>
<div id="clonalabundance-distribution-of-clones-by-size"
class="section level2">
<h2>clonalAbundance: Distribution of Clones by Size</h2>
<p><code>clonalAbundance()</code> allows for the examination of the
relative distribution of clones by abundance. It produces a line graph
showing the total number of clones at specific frequencies within a
sample or group.</p>
<p>Key Parameter(s) for <code>clonalAbundance()</code></p>
<ul>
<li><code>scale</code>: If <code>TRUE</code>, converts the graphs into
density plots to show relative distributions; if <code>FALSE</code>
(default), displays raw counts.</li>
</ul>
<p>To visualize the raw clonal abundance using the <code>gene</code>
clone definition:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="fu">clonalAbundance</span>(combined.TCR, </span>
<span id="cb27-2"><a href="#cb27-2" tabindex="-1"></a>                <span class="at">cloneCall =</span> <span class="st">&quot;gene&quot;</span>, </span>
<span id="cb27-3"><a href="#cb27-3" tabindex="-1"></a>                <span class="at">scale =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-20-1.png" width="700" /></p>
<p><code>clonalAbundance()</code> output can also be converted into a
density plot, which may allow for better comparisons between different
repertoire sizes, by setting <code>scale = TRUE</code>.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a><span class="fu">clonalAbundance</span>(combined.TCR, </span>
<span id="cb28-2"><a href="#cb28-2" tabindex="-1"></a>                <span class="at">cloneCall =</span> <span class="st">&quot;gene&quot;</span>, </span>
<span id="cb28-3"><a href="#cb28-3" tabindex="-1"></a>                <span class="at">scale =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-21-1.png" width="700" /></p>
</div>
<div id="clonallength-distribution-of-sequence-lengths"
class="section level2">
<h2>clonalLength: Distribution of Sequence Lengths</h2>
<p><code>clonalLength()</code> allows you to look at the length
distribution of the CDR3 sequences. Importantly, unlike the other basic
visualizations, the <code>cloneCall</code> can only be <code>nt</code>
(nucleotide) or <code>aa</code> (amino acid). Due to the method of
calling clones as outlined previously (e.g., using NA for unreturned
chain sequences or multiple chains within a single barcode), the length
distribution may reveal a multimodal curve.</p>
<p>To visualize the amino acid length distribution for both chains
(“both”):</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a><span class="fu">clonalLength</span>(combined.TCR, </span>
<span id="cb29-2"><a href="#cb29-2" tabindex="-1"></a>             <span class="at">cloneCall=</span><span class="st">&quot;aa&quot;</span>, </span>
<span id="cb29-3"><a href="#cb29-3" tabindex="-1"></a>             <span class="at">chain =</span> <span class="st">&quot;both&quot;</span>) </span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-22-1.png" width="700" /></p>
<p>To visualize the amino acid length distribution for the
<code>TRA</code> chain, scaled as a density plot:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="fu">clonalLength</span>(combined.TCR, </span>
<span id="cb30-2"><a href="#cb30-2" tabindex="-1"></a>             <span class="at">cloneCall=</span><span class="st">&quot;aa&quot;</span>, </span>
<span id="cb30-3"><a href="#cb30-3" tabindex="-1"></a>             <span class="at">chain =</span> <span class="st">&quot;TRA&quot;</span>, </span>
<span id="cb30-4"><a href="#cb30-4" tabindex="-1"></a>             <span class="at">scale =</span> <span class="cn">TRUE</span>) </span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-23-1.png" width="700" /></p>
</div>
<div id="clonalcompare-clonal-dynamics-between-categorical-variables"
class="section level2">
<h2>clonalCompare: Clonal Dynamics Between Categorical Variables</h2>
<p><code>clonalCompare()</code> allows you to look at clones between
samples and changes in dynamics. It is useful for tracking how the
proportions of top clones change between conditions.</p>
<p>Key Parameters for <code>clonalCompare()</code></p>
<ul>
<li><code>samples</code>: A character vector to isolate specific samples
by their list element name.</li>
<li><code>clones</code>: A character vector of specific clonal sequences
to visualize. If used, <code>top.clones</code> will be ignored.
*<code>top.clones</code>: The top n number of clones to graph,
calculated based on the frequency within individual samples.</li>
<li><code>highlight.clones</code>: A character vector of specific clonal
sequences to color; all other clones will be greyed out.</li>
<li><code>relabel.clones</code>: If <code>TRUE</code>, simplifies the
legend by labeling isolated clones numerically (e.g., “Clone: 1”).</li>
<li><code>graph</code>: The type of plot to generate;
<code>alluvial</code> (default) or <code>area</code>.</li>
<li><code>proportion</code>: If <code>TRUE</code> (default), the y-axis
represents proportional abundance; if <code>FALSE</code>, it represents
raw clone counts.</li>
</ul>
<p>To compare the top 10 clones between samples “P17B” and “P17L” using
amino acid sequences as an alluvial plot:</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a><span class="fu">clonalCompare</span>(combined.TCR, </span>
<span id="cb31-2"><a href="#cb31-2" tabindex="-1"></a>                  <span class="at">top.clones =</span> <span class="dv">10</span>, </span>
<span id="cb31-3"><a href="#cb31-3" tabindex="-1"></a>                  <span class="at">samples =</span> <span class="fu">c</span>(<span class="st">&quot;P17B&quot;</span>, <span class="st">&quot;P17L&quot;</span>), </span>
<span id="cb31-4"><a href="#cb31-4" tabindex="-1"></a>                  <span class="at">cloneCall=</span><span class="st">&quot;aa&quot;</span>, </span>
<span id="cb31-5"><a href="#cb31-5" tabindex="-1"></a>                  <span class="at">graph =</span> <span class="st">&quot;alluvial&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-24-1.png" width="700" /></p>
<p>We can also choose to highlight specific clones, such as in the case
of <em>“CVVSDNTGGFKTIF_CASSVRRERANTGELFF”</em> and
<em>“NA_CASSVRRERANTGELFF”</em> using the <code>highlight.clones</code>
parameter. In addition, we can simplify the plot to label the clones as
“Clone: 1”, “Clone: 2”, etc., by setting
<code>relabel.clones = TRUE</code>.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="fu">clonalCompare</span>(combined.TCR, </span>
<span id="cb32-2"><a href="#cb32-2" tabindex="-1"></a>              <span class="at">top.clones =</span> <span class="dv">10</span>,</span>
<span id="cb32-3"><a href="#cb32-3" tabindex="-1"></a>              <span class="at">highlight.clones =</span> <span class="fu">c</span>(<span class="st">&quot;CVVSDNTGGFKTIF_CASSVRRERANTGELFF&quot;</span>, <span class="st">&quot;NA_CASSVRRERANTGELFF&quot;</span>),</span>
<span id="cb32-4"><a href="#cb32-4" tabindex="-1"></a>              <span class="at">relabel.clones =</span> <span class="cn">TRUE</span>,</span>
<span id="cb32-5"><a href="#cb32-5" tabindex="-1"></a>              <span class="at">samples =</span> <span class="fu">c</span>(<span class="st">&quot;P17B&quot;</span>, <span class="st">&quot;P17L&quot;</span>), </span>
<span id="cb32-6"><a href="#cb32-6" tabindex="-1"></a>              <span class="at">cloneCall=</span><span class="st">&quot;aa&quot;</span>, </span>
<span id="cb32-7"><a href="#cb32-7" tabindex="-1"></a>              <span class="at">graph =</span> <span class="st">&quot;alluvial&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-25-1.png" width="700" /></p>
<p>Alternatively, if we only want to show specific clones, we can use
the <code>clones</code> parameter.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" tabindex="-1"></a><span class="fu">clonalCompare</span>(combined.TCR, </span>
<span id="cb33-2"><a href="#cb33-2" tabindex="-1"></a>              <span class="at">clones =</span> <span class="fu">c</span>(<span class="st">&quot;CVVSDNTGGFKTIF_CASSVRRERANTGELFF&quot;</span>, <span class="st">&quot;NA_CASSVRRERANTGELFF&quot;</span>),</span>
<span id="cb33-3"><a href="#cb33-3" tabindex="-1"></a>              <span class="at">relabel.clones =</span> <span class="cn">TRUE</span>,</span>
<span id="cb33-4"><a href="#cb33-4" tabindex="-1"></a>              <span class="at">samples =</span> <span class="fu">c</span>(<span class="st">&quot;P17B&quot;</span>, <span class="st">&quot;P17L&quot;</span>), </span>
<span id="cb33-5"><a href="#cb33-5" tabindex="-1"></a>              <span class="at">cloneCall=</span><span class="st">&quot;aa&quot;</span>, </span>
<span id="cb33-6"><a href="#cb33-6" tabindex="-1"></a>              <span class="at">graph =</span> <span class="st">&quot;alluvial&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-26-1.png" width="700" /></p>
</div>
<div id="clonalscatter-scatterplot-of-two-variables"
class="section level2">
<h2>clonalScatter: Scatterplot of Two Variables</h2>
<p>clonalScatter() organizes two repertoires, quantifies their relative
clone sizes, and produces a scatter plot comparing the two samples.
Clones are categorized by counts into singlets or expanded, either
exclusively present or shared between the selected samples.</p>
<p>Key Parameter(s) for <code>clonalScatter()</code></p>
<ul>
<li><code>x.axis</code>, <code>y.axis</code>: Names of the list elements
or meta data variable to place on the x-axis and y-axis.</li>
<li><code>dot.size</code>: Specifies how dot size is determined;
<code>total</code> (default) displays the total number of clones between
the x- and y-axis, or a specific list element name for size
calculation.</li>
<li><code>graph</code>: The type of graph to display;
<code>proportion</code> for the relative proportion of clones (default)
or <code>count</code> for the total count of clones by sample.</li>
</ul>
<p>To compare samples “P18B” and “P18L” based on <code>gene</code> clone
calls, with dot size representing the total number of clones, and
plotting clone proportions:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" tabindex="-1"></a><span class="fu">clonalScatter</span>(combined.TCR, </span>
<span id="cb34-2"><a href="#cb34-2" tabindex="-1"></a>              <span class="at">cloneCall =</span><span class="st">&quot;gene&quot;</span>, </span>
<span id="cb34-3"><a href="#cb34-3" tabindex="-1"></a>              <span class="at">x.axis =</span> <span class="st">&quot;P18B&quot;</span>, </span>
<span id="cb34-4"><a href="#cb34-4" tabindex="-1"></a>              <span class="at">y.axis =</span> <span class="st">&quot;P18L&quot;</span>,</span>
<span id="cb34-5"><a href="#cb34-5" tabindex="-1"></a>              <span class="at">dot.size =</span> <span class="st">&quot;total&quot;</span>,</span>
<span id="cb34-6"><a href="#cb34-6" tabindex="-1"></a>              <span class="at">graph =</span> <span class="st">&quot;proportion&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-27-1.png" width="700" /></p>
<hr />
</div>
</div>
<div id="visualizing-clonal-dynamics" class="section level1">
<h1>Visualizing Clonal Dynamics</h1>
<div id="clonalhomeostasis-examining-clonal-space"
class="section level2">
<h2>clonalHomeostasis: Examining Clonal Space</h2>
<p>By examining the clonal space, we effectively look at the relative
space occupied by clones at specific proportions. Another way to think
about this would be to consider the total immune receptor sequencing run
as a measuring cup. In this cup, we will fill liquids of different
viscosity—or different numbers of clonal proportions. Clonal space
homeostasis asks what percentage of the cup is filled by clones in
distinct proportions (or liquids of different viscosity, to extend the
analogy). The proportional cut points are set under the
<code>cloneSize</code> variable in the function and can be adjusted.</p>
<p>Default <code>cloneSize</code> Bins</p>
<ul>
<li><strong>Rare</strong>: 0 to 0.0001</li>
<li><strong>Small</strong>: 0.0001 to 0.001</li>
<li><strong>Medium</strong>: 0.001 to 0.01</li>
<li><strong>Large</strong>: 0.01 to 0.1</li>
<li><strong>Hyperexpanded</strong>: 0.1 to 1</li>
</ul>
<p>To visualize clonal homeostasis using gene clone calls with default
<code>cloneSize</code> bins:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" tabindex="-1"></a><span class="fu">clonalHomeostasis</span>(combined.TCR, </span>
<span id="cb35-2"><a href="#cb35-2" tabindex="-1"></a>                  <span class="at">cloneCall =</span> <span class="st">&quot;gene&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-28-1.png" width="700" /></p>
<p>We can reassign the proportional cut points for
<code>cloneSize</code>, which can drastically alter the visualization
and analysis</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" tabindex="-1"></a><span class="fu">clonalHomeostasis</span>(combined.TCR, </span>
<span id="cb36-2"><a href="#cb36-2" tabindex="-1"></a>                  <span class="at">cloneCall =</span> <span class="st">&quot;gene&quot;</span>,</span>
<span id="cb36-3"><a href="#cb36-3" tabindex="-1"></a>                  <span class="at">cloneSize =</span> <span class="fu">c</span>(<span class="at">Rare =</span> <span class="fl">0.001</span>, <span class="at">Small =</span> <span class="fl">0.01</span>, <span class="at">Medium =</span> <span class="fl">0.1</span>, </span>
<span id="cb36-4"><a href="#cb36-4" tabindex="-1"></a>                                <span class="at">Large =</span> <span class="fl">0.3</span>, <span class="at">Hyperexpanded =</span> <span class="dv">1</span>))</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-29-1.png" width="700" /></p>
<p>In addition, we can use the <code>group.by</code> parameter to look
at the relative proportion of clones between groups, such as by tissue
type. First, ensure the “Type” variable is added to
<code>combined.TCR</code>:</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a>combined.TCR <span class="ot">&lt;-</span> <span class="fu">addVariable</span>(combined.TCR, </span>
<span id="cb37-2"><a href="#cb37-2" tabindex="-1"></a>                            <span class="at">variable.name =</span> <span class="st">&quot;Type&quot;</span>, </span>
<span id="cb37-3"><a href="#cb37-3" tabindex="-1"></a>                            <span class="at">variables =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">&quot;B&quot;</span>, <span class="st">&quot;L&quot;</span>), <span class="dv">4</span>))</span></code></pre></div>
<p>Now, visualize clonal homeostasis grouped by “Type”:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" tabindex="-1"></a><span class="fu">clonalHomeostasis</span>(combined.TCR, </span>
<span id="cb38-2"><a href="#cb38-2" tabindex="-1"></a>                  <span class="at">group.by =</span> <span class="st">&quot;Type&quot;</span>,</span>
<span id="cb38-3"><a href="#cb38-3" tabindex="-1"></a>                  <span class="at">cloneCall =</span> <span class="st">&quot;gene&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-31-1.png" width="700" /></p>
<p><code>clonalHomeostasis()</code> provides an assessment of how
different “sizes” of clones (based on their proportional abundance)
contribute to the overall repertoire. This visualization helps to
identify shifts in repertoire structure, such as expansion of large
clones in response to infection or contraction in chronic conditions,
offering insights into immune system activity and health.</p>
</div>
<div id="clonalproportion-examining-space-occupied-by-ranks-of-clones"
class="section level2">
<h2>clonalProportion: Examining Space Occupied by Ranks of Clones</h2>
<p>Like clonal space homeostasis, <code>clonalProportion()</code> also
categorizes clones into separate bins. The key difference is that
instead of looking at the relative proportion of the clone to the total,
<code>clonalProportion()</code> ranks the clones by their total count or
frequency of occurrence and then places them into predefined bins.</p>
<p>The <code>clonalSplit</code> parameter represents the ranking of
clonotypes. For example, 1:10 refers to the top 10 clonotypes in each
sample. The default bins are set under the <code>clonalSplit</code>
variable and can be adjusted.</p>
<p>Default <code>clonalSplit</code> Bins</p>
<ul>
<li><strong>10</strong> (top 1-10 clones)</li>
<li><strong>100</strong>: (top 11-100 clones)</li>
<li><strong>1000</strong>: (top 101-1000 clones)</li>
<li><strong>10000</strong>: (top 1001-10000 clones)</li>
<li><strong>30000</strong>: (top 10001-30000 clones)</li>
<li><strong>100000</strong>: (top 30001-100000 clones)</li>
</ul>
<p>To visualize the clonal proportion using <code>gene</code> clone
calls with default <code>clonalSplit</code> bins:</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" tabindex="-1"></a><span class="fu">clonalProportion</span>(combined.TCR, </span>
<span id="cb39-2"><a href="#cb39-2" tabindex="-1"></a>                 <span class="at">cloneCall =</span> <span class="st">&quot;gene&quot;</span>) </span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-32-1.png" width="700" /></p>
<p>To visualize clonal proportion using <code>nt</code> (nucleotide)
clone calls with custom <code>clonalSplit</code> bins:</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" tabindex="-1"></a><span class="fu">clonalProportion</span>(combined.TCR, </span>
<span id="cb40-2"><a href="#cb40-2" tabindex="-1"></a>                 <span class="at">cloneCall =</span> <span class="st">&quot;nt&quot;</span>,</span>
<span id="cb40-3"><a href="#cb40-3" tabindex="-1"></a>                 <span class="at">clonalSplit =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">10</span>, <span class="dv">100</span>, <span class="dv">1000</span>, <span class="dv">10000</span>)) </span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-33-1.png" width="700" /></p>
<p><code>clonalProportion()</code> complements
<code>clonalHomeostasis()</code> by providing a perspective on how the
“richest” (most abundant) clones contribute to the overall repertoire
space. By segmenting clones into rank-based bins, it helps identify
whether a few highly expanded clones or a larger number of moderately
expanded clones dominate the immune response, offering distinct insights
into repertoire structure and dynamics.</p>
<hr />
</div>
</div>
<div id="summarizing-repertoires" class="section level1">
<h1>Summarizing Repertoires</h1>
<div id="percentgeneusage" class="section level2">
<h2>percentGeneUsage</h2>
<p>Gene quantification and visualization has been redesigned to offer a
more robust and translatable function under
<code>percentGeneUsage()</code>. We have maintained the functionality of
the previous functions for gene-level quantification under the aliases
<code>vizGenes()</code>, <code>percentGenes()</code>, and
<code>percentVJ()</code>.</p>
<div id="vizgenes-flexible-gene-usage-visualization"
class="section level3">
<h3>vizGenes: Flexible Gene Usage Visualization</h3>
<p>The <code>vizGenes()</code> function offers a highly adaptable
approach to visualizing the relative usage of TCR or BCR genes. It acts
as a versatile alias for <code>percentGeneUsage()</code>, allowing for
comparisons across different chains, scaling of values, and selection
between bar charts and heatmaps.</p>
<ul>
<li><code>x.axis</code>: Specifies the gene segment to display along the
x-axis (e.g., “TRBV”, “TRBD”, “IGKJ”).</li>
<li><code>y.axis</code>
<ul>
<li>Another gene segment (e.g., “TRAV”, “TRBJ”) for paired gene
analysis. When x.axis and y.axis are both gene segments, vizGenes()
internally calls percentGeneUsage() with genes = c(x.axis, y.axis),
resulting in a heatmap.</li>
<li>A categorical variable (e.g., “sample”, “orig.ident”) to visualize
gene usage across different groups. When y.axis is a categorical
variable, vizGenes() maps it to the group.by parameter of
percentGeneUsage(), creating facets for each category.</li>
</ul></li>
<li><code>plot</code>: Determines the visualization type:
<ul>
<li>“barplot”: Ideal for visualizing the distribution of a single gene
segment.</li>
<li>“heatmap”: Suitable for single gene usage (with a group.by or y.axis
categorical variable) or for paired gene analysis.</li>
</ul></li>
<li><code>summary.fun</code>: (Inherited from
<code>percentGeneUsage</code>) Defines the statistic to display:
“percent” (default), “proportion”, or “count”. This implicitly handles
the scaling of values.</li>
</ul>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" tabindex="-1"></a><span class="fu">vizGenes</span>(combined.TCR,</span>
<span id="cb41-2"><a href="#cb41-2" tabindex="-1"></a>         <span class="at">x.axis =</span> <span class="st">&quot;TRBV&quot;</span>,</span>
<span id="cb41-3"><a href="#cb41-3" tabindex="-1"></a>         <span class="at">y.axis =</span> <span class="cn">NULL</span>, <span class="co"># No specific y-axis variable, will group all samples</span></span>
<span id="cb41-4"><a href="#cb41-4" tabindex="-1"></a>         <span class="at">plot =</span> <span class="st">&quot;barplot&quot;</span>,</span>
<span id="cb41-5"><a href="#cb41-5" tabindex="-1"></a>         <span class="at">summary.fun =</span> <span class="st">&quot;proportion&quot;</span>) </span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-34-1.png" width="700" /></p>
<p>This plot shows the proportion of each <em>TRBV</em> gene segment
observed across the entire combined.TCR dataset. Since
<code>y.axis</code> is NULL, samples are grouped by the list element of
the combined.TCR data.</p>
<p><code>vizGenes()</code> is particularly useful for examining gene
pairings. Let’s look at the differences in <em>TRBV</em> and
<em>TRBJ</em> usage between peripheral blood and lung samples from your
dataset. We’ll subset combined.TCR for this analysis.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" tabindex="-1"></a><span class="co"># Peripheral Blood Samples</span></span>
<span id="cb42-2"><a href="#cb42-2" tabindex="-1"></a><span class="fu">vizGenes</span>(combined.TCR[<span class="fu">c</span>(<span class="st">&quot;P17B&quot;</span>, <span class="st">&quot;P18B&quot;</span>, <span class="st">&quot;P19B&quot;</span>, <span class="st">&quot;P20B&quot;</span>)],</span>
<span id="cb42-3"><a href="#cb42-3" tabindex="-1"></a>         <span class="at">x.axis =</span> <span class="st">&quot;TRBV&quot;</span>,</span>
<span id="cb42-4"><a href="#cb42-4" tabindex="-1"></a>         <span class="at">y.axis =</span> <span class="st">&quot;TRBJ&quot;</span>,</span>
<span id="cb42-5"><a href="#cb42-5" tabindex="-1"></a>         <span class="at">plot =</span> <span class="st">&quot;heatmap&quot;</span>,</span>
<span id="cb42-6"><a href="#cb42-6" tabindex="-1"></a>         <span class="at">summary.fun =</span> <span class="st">&quot;percent&quot;</span>) <span class="co"># Display percentages</span></span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-35-1.png" width="700" /></p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" tabindex="-1"></a><span class="co"># Lung Samples</span></span>
<span id="cb43-2"><a href="#cb43-2" tabindex="-1"></a><span class="fu">vizGenes</span>(combined.TCR[<span class="fu">c</span>(<span class="st">&quot;P17L&quot;</span>, <span class="st">&quot;P18L&quot;</span>, <span class="st">&quot;P19L&quot;</span>, <span class="st">&quot;P20L&quot;</span>)],</span>
<span id="cb43-3"><a href="#cb43-3" tabindex="-1"></a>         <span class="at">x.axis =</span> <span class="st">&quot;TRBV&quot;</span>,</span>
<span id="cb43-4"><a href="#cb43-4" tabindex="-1"></a>         <span class="at">y.axis =</span> <span class="st">&quot;TRBJ&quot;</span>,</span>
<span id="cb43-5"><a href="#cb43-5" tabindex="-1"></a>         <span class="at">plot =</span> <span class="st">&quot;heatmap&quot;</span>,</span>
<span id="cb43-6"><a href="#cb43-6" tabindex="-1"></a>         <span class="at">summary.fun =</span> <span class="st">&quot;percent&quot;</span>) <span class="co"># Display percentages</span></span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-35-2.png" width="700" /></p>
<p>In these examples, by providing both <code>x.axis</code> and
<code>y.axis</code> as gene segments (“TRBV” and “TRBJ”),
<code>vizGenes()</code> automatically performs a paired gene analysis,
generating a heatmap where the intensity reflects the percentage of each
V-J pairing.</p>
<p>Beyond V-J pairings within a single chain, <code>vizGenes()</code>
can also visualize gene usage across different chains. For instance, to
examine <em>TRBV</em> and <em>TRAV</em> pairings for patient P17’s
samples:</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" tabindex="-1"></a><span class="fu">vizGenes</span>(combined.TCR[<span class="fu">c</span>(<span class="st">&quot;P17B&quot;</span>, <span class="st">&quot;P17L&quot;</span>)],</span>
<span id="cb44-2"><a href="#cb44-2" tabindex="-1"></a>         <span class="at">x.axis =</span> <span class="st">&quot;TRBV&quot;</span>,</span>
<span id="cb44-3"><a href="#cb44-3" tabindex="-1"></a>         <span class="at">y.axis =</span> <span class="st">&quot;TRAV&quot;</span>,</span>
<span id="cb44-4"><a href="#cb44-4" tabindex="-1"></a>         <span class="at">plot =</span> <span class="st">&quot;heatmap&quot;</span>,</span>
<span id="cb44-5"><a href="#cb44-5" tabindex="-1"></a>         <span class="at">summary.fun =</span> <span class="st">&quot;count&quot;</span>) </span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-36-1.png" width="700" /></p>
</div>
<div id="percentgenes-quantifying-single-gene-usage"
class="section level3">
<h3>percentGenes: Quantifying Single Gene Usage</h3>
<p>The <code>percentGenes()</code> function is a specialized alias for
<code>percentGeneUsage()</code> designed to quantify the usage of a
single V, D, or J gene locus for a specified immune receptor chain. By
default, it returns a heatmap visualization.</p>
<p>Key Parameters for <code>percentGenes()</code>:</p>
<ul>
<li><code>chain</code>: Specifies the immune receptor chain (e.g.,
“TRB”, “TRA”, “IGH”, “IGL”).</li>
<li><code>gene</code>: Indicates the gene locus to quantify: “Vgene”,
“Dgene”, or “Jgene”.</li>
<li><code>group.by</code>, <code>order.by</code>,
<code>summary.fun</code>, <code>exportTable</code>,
<code>palette</code>: These parameters are directly passed to
<code>percentGeneUsage()</code>.</li>
</ul>
<p>To quantify and visualize the percentage of TRBV gene usage across
your samples:</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" tabindex="-1"></a><span class="fu">percentGenes</span>(combined.TCR,</span>
<span id="cb45-2"><a href="#cb45-2" tabindex="-1"></a>             <span class="at">chain =</span> <span class="st">&quot;TRB&quot;</span>,</span>
<span id="cb45-3"><a href="#cb45-3" tabindex="-1"></a>             <span class="at">gene =</span> <span class="st">&quot;Vgene&quot;</span>,</span>
<span id="cb45-4"><a href="#cb45-4" tabindex="-1"></a>             <span class="at">summary.fun =</span> <span class="st">&quot;percent&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-37-1.png" width="700" /></p>
<p>This generates a heatmap showing the percentage of each TRBV gene
segment within each sample, allowing for easy visual comparison of gene
usage profiles across your samples.</p>
<p>The raw data returned by <code>percentGenes()</code> (when
<code>exportTable</code> = TRUE) can be a powerful input for further
downstream analysis, such as dimensionality reduction techniques like
Principal Component Analysis. This allows you to summarize the complex
gene usage patterns and identify samples with similar or distinct
repertoires.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" tabindex="-1"></a>df.genes <span class="ot">&lt;-</span> <span class="fu">percentGenes</span>(combined.TCR,</span>
<span id="cb46-2"><a href="#cb46-2" tabindex="-1"></a>                         <span class="at">chain =</span> <span class="st">&quot;TRB&quot;</span>,</span>
<span id="cb46-3"><a href="#cb46-3" tabindex="-1"></a>                         <span class="at">gene =</span> <span class="st">&quot;Vgene&quot;</span>,</span>
<span id="cb46-4"><a href="#cb46-4" tabindex="-1"></a>                         <span class="at">exportTable =</span> <span class="cn">TRUE</span>,</span>
<span id="cb46-5"><a href="#cb46-5" tabindex="-1"></a>                         <span class="at">summary.fun =</span> <span class="st">&quot;proportion&quot;</span>) </span>
<span id="cb46-6"><a href="#cb46-6" tabindex="-1"></a></span>
<span id="cb46-7"><a href="#cb46-7" tabindex="-1"></a><span class="co"># Performing PCA on the gene usage matrix</span></span>
<span id="cb46-8"><a href="#cb46-8" tabindex="-1"></a>pc <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(<span class="fu">t</span>(df.genes) )</span>
<span id="cb46-9"><a href="#cb46-9" tabindex="-1"></a></span>
<span id="cb46-10"><a href="#cb46-10" tabindex="-1"></a><span class="co"># Getting data frame to plot from</span></span>
<span id="cb46-11"><a href="#cb46-11" tabindex="-1"></a>df_plot <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">cbind</span>(pc<span class="sc">$</span>x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="fu">colnames</span>(df.genes)))</span>
<span id="cb46-12"><a href="#cb46-12" tabindex="-1"></a><span class="fu">colnames</span>(df_plot) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;PC1&quot;</span>, <span class="st">&quot;PC2&quot;</span>, <span class="st">&quot;Sample&quot;</span>)</span>
<span id="cb46-13"><a href="#cb46-13" tabindex="-1"></a>df_plot<span class="sc">$</span>PC1 <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(df_plot<span class="sc">$</span>PC1)</span>
<span id="cb46-14"><a href="#cb46-14" tabindex="-1"></a>df_plot<span class="sc">$</span>PC2 <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(df_plot<span class="sc">$</span>PC2)</span>
<span id="cb46-15"><a href="#cb46-15" tabindex="-1"></a></span>
<span id="cb46-16"><a href="#cb46-16" tabindex="-1"></a><span class="fu">ggplot</span>(df_plot, <span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2)) <span class="sc">+</span></span>
<span id="cb46-17"><a href="#cb46-17" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">fill =</span> Sample), <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">size =</span> <span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb46-18"><a href="#cb46-18" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">fill=</span><span class="fu">guide_legend</span>(<span class="at">title=</span><span class="st">&quot;Samples&quot;</span>)) <span class="sc">+</span></span>
<span id="cb46-19"><a href="#cb46-19" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">hcl.colors</span>(<span class="fu">nrow</span>(df_plot), <span class="st">&quot;inferno&quot;</span>)) <span class="sc">+</span></span>
<span id="cb46-20"><a href="#cb46-20" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb46-21"><a href="#cb46-21" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;PCA of TRBV Gene Usage&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-38-1.png" width="700" /></p>
<p>This PCA plot visually clusters samples based on their <em>TRBV</em>
gene usage profiles, helping to identify underlying patterns or
relationships between samples.</p>
</div>
<div id="percentvj-quantifying-v-j-gene-pairings"
class="section level3">
<h3>percentVJ: Quantifying V-J Gene Pairings</h3>
<p>The <code>percentVJ()</code> function is another specialized alias
for <code>percentGeneUsage()</code>, specifically designed to quantify
the proportion or percentage of V and J gene segments paired together
within individual repertoires. It always produces a heatmap
visualization.</p>
<p>Key Parameters for <code>percenVJ()</code>:</p>
<ul>
<li><code>chain</code>: Specifies the immune receptor chain (e.g.,
<strong>“TRB”</strong>, <strong>“TRA”</strong>, <strong>“IGH”</strong>,
<strong>“IGL”</strong>). This dictates which V and J gene segments are
analyzed (e.g., <strong>TRBV</strong> and <strong>TRBJ</strong> for
<code>chain</code> = “TRB”)</li>
</ul>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" tabindex="-1"></a><span class="fu">percentVJ</span>(combined.TCR,</span>
<span id="cb47-2"><a href="#cb47-2" tabindex="-1"></a>          <span class="at">chain =</span> <span class="st">&quot;TRB&quot;</span>,</span>
<span id="cb47-3"><a href="#cb47-3" tabindex="-1"></a>          <span class="at">summary.fun =</span> <span class="st">&quot;percent&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-39-1.png" width="700" /></p>
<p>This heatmap displays the percentage of each <em>TRBV-TRBJ</em>
pairing for each sample, providing a detailed view of the V-J
recombination landscape.</p>
<p>Similar to <code>percentGenes()</code>, the quantitative output of
<code>percentVJ()</code> can be used for dimensionality reduction to
summarize V-J pairing patterns across samples.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" tabindex="-1"></a>df.vj <span class="ot">&lt;-</span> <span class="fu">percentVJ</span>(combined.TCR,</span>
<span id="cb48-2"><a href="#cb48-2" tabindex="-1"></a>                   <span class="at">chain =</span> <span class="st">&quot;TRB&quot;</span>,</span>
<span id="cb48-3"><a href="#cb48-3" tabindex="-1"></a>                   <span class="at">exportTable =</span> <span class="cn">TRUE</span>,</span>
<span id="cb48-4"><a href="#cb48-4" tabindex="-1"></a>                   <span class="at">summary.fun =</span> <span class="st">&quot;proportion&quot;</span>) <span class="co"># Export proportions for PCA</span></span>
<span id="cb48-5"><a href="#cb48-5" tabindex="-1"></a></span>
<span id="cb48-6"><a href="#cb48-6" tabindex="-1"></a><span class="co"># Performing PCA on the V-J pairing matrix</span></span>
<span id="cb48-7"><a href="#cb48-7" tabindex="-1"></a>pc.vj <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(<span class="fu">t</span>(df.vj))</span>
<span id="cb48-8"><a href="#cb48-8" tabindex="-1"></a></span>
<span id="cb48-9"><a href="#cb48-9" tabindex="-1"></a><span class="co"># Getting data frame to plot from</span></span>
<span id="cb48-10"><a href="#cb48-10" tabindex="-1"></a>df_plot_vj <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">cbind</span>(pc.vj<span class="sc">$</span>x[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="fu">colnames</span>(df.vj)))</span>
<span id="cb48-11"><a href="#cb48-11" tabindex="-1"></a><span class="fu">colnames</span>(df_plot_vj) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;PC1&quot;</span>, <span class="st">&quot;PC2&quot;</span>, <span class="st">&quot;Sample&quot;</span>)</span>
<span id="cb48-12"><a href="#cb48-12" tabindex="-1"></a>df_plot_vj<span class="sc">$</span>PC1 <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(df_plot_vj<span class="sc">$</span>PC1)</span>
<span id="cb48-13"><a href="#cb48-13" tabindex="-1"></a>df_plot_vj<span class="sc">$</span>PC2 <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(df_plot_vj<span class="sc">$</span>PC2)</span>
<span id="cb48-14"><a href="#cb48-14" tabindex="-1"></a></span>
<span id="cb48-15"><a href="#cb48-15" tabindex="-1"></a><span class="co"># Plotting the PCA results</span></span>
<span id="cb48-16"><a href="#cb48-16" tabindex="-1"></a><span class="fu">ggplot</span>(df_plot_vj, <span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2)) <span class="sc">+</span></span>
<span id="cb48-17"><a href="#cb48-17" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">fill =</span> Sample), <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">size =</span> <span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb48-18"><a href="#cb48-18" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">fill=</span><span class="fu">guide_legend</span>(<span class="at">title=</span><span class="st">&quot;Samples&quot;</span>)) <span class="sc">+</span></span>
<span id="cb48-19"><a href="#cb48-19" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">hcl.colors</span>(<span class="fu">nrow</span>(df_plot_vj), <span class="st">&quot;inferno&quot;</span>)) <span class="sc">+</span></span>
<span id="cb48-20"><a href="#cb48-20" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb48-21"><a href="#cb48-21" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;PCA of TRBV-TRBJ Gene Pairings&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-40-1.png" width="700" /></p>
</div>
</div>
<div id="percentaa-amino-acid-composition-by-residue"
class="section level2">
<h2>percentAA: Amino Acid Composition by Residue</h2>
<p>Quantify the proportion of amino acids along the CDR3 sequence with
<code>percentAA()</code>. By default, the function will pad the
sequences with NAs up to the maximum of <code>aa.length</code>.
Sequences longer than aa.length will be removed before visualization
(default <code>aa.length</code> = 20).</p>
<p>Key Parameter(s) for <code>percentAA()</code></p>
<ul>
<li><code>aa.length</code>: The maximum length of the CDR3 amino acid
sequence to consider.</li>
</ul>
<p>To visualize the relative percentage of amino acids at each position
of the CDR3 sequences for the TRB chain, up to a length of 20 amino
acids:</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" tabindex="-1"></a><span class="fu">percentAA</span>(combined.TCR, </span>
<span id="cb49-2"><a href="#cb49-2" tabindex="-1"></a>          <span class="at">chain =</span> <span class="st">&quot;TRB&quot;</span>, </span>
<span id="cb49-3"><a href="#cb49-3" tabindex="-1"></a>          <span class="at">aa.length =</span> <span class="dv">20</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-41-1.png" width="700" /></p>
<p>This plot displays the relative proportion of each amino acid at
different positions across the CDR3 sequence for the specified chain. It
provides insights into the amino acid composition and variability at
each residue, which can be indicative of functional constraints or
selection pressures on the CDR3</p>
</div>
<div id="positionalentropy-entropy-across-cdr3-sequences"
class="section level2">
<h2>positionalEntropy: Entropy across CDR3 Sequences</h2>
<p>We can also quantify the level of entropy/diversity across amino acid
residues along the CDR3 sequence. <code>positionalEntropy()</code>
combines the quantification by residue of <code>percentAA()</code> with
diversity calculations. Positions without variance will have a value
reported as 0 for the purposes of comparison.</p>
<p>Key Parameter(s) for <code>positionalEntropy()</code></p>
<ul>
<li><code>method</code>
<ul>
<li><code>shannon</code> - Shannon Index</li>
<li><code>inv.simpson</code> - Inverse Simpson Index</li>
<li><code>gini.simpson</code> - Gini-Simpson Index</li>
<li><code>norm.entropy</code> - Normalized Entropy</li>
<li><code>pielou</code> - Pielou’s Evenness</li>
<li><code>hill1</code>, <code>hill2</code>, <code>hill3</code> - Hill
Numbers</li>
</ul></li>
</ul>
<p>To visualize the normalized entropy across amino acid residues for
the TRB chain, up to a length of 20 amino acids:</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" tabindex="-1"></a><span class="fu">positionalEntropy</span>(combined.TCR, </span>
<span id="cb50-2"><a href="#cb50-2" tabindex="-1"></a>                  <span class="at">chain =</span> <span class="st">&quot;TRB&quot;</span>, </span>
<span id="cb50-3"><a href="#cb50-3" tabindex="-1"></a>                  <span class="at">aa.length =</span> <span class="dv">20</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-42-1.png" width="700" /></p>
<p>The plot generated by <code>positionalEntropy()</code> illustrates
the diversity or entropy at each amino acid position within the CDR3
sequence. Higher entropy values indicate greater variability in amino
acid usage at that position, suggesting less selective pressure or more
promiscuous binding, while lower values suggest conserved positions
critical for structural integrity or antigen recognition.</p>
</div>
<div id="positionalproperty-amino-acid-properties-across-cdr3-sequence"
class="section level2">
<h2>positionalProperty: Amino Acid Properties across CDR3 Sequence</h2>
<p>Like <code>positionalEntropy()</code>, we can also examine a series
of amino acid properties along the CDR3 sequences using
<code>positionalProperty()</code>. Important differences for
<code>positionalProperty()</code> is dropping NA values as they would
void the mean calculation and displaying a ribbon with the 95%
confidence interval surrounding the mean value for the selected
properties.</p>
<p>Key Parameter(s) for <code>positionalEntropy()</code></p>
<ul>
<li><code>method</code>
<ul>
<li><code>atchleyFactors</code>: <a
href="https://pubmed.ncbi.nlm.nih.gov/15851683/">citation</a></li>
<li><code>crucianiProperties</code>: <a
href="https://analyticalsciencejournals.onlinelibrary.wiley.com/doi/abs/10.1002/cem.856">citation</a></li>
<li><code>FASGAI</code>: <a
href="https://pubmed.ncbi.nlm.nih.gov/18318694/">citation</a></li>
<li><code>kideraFactors</code>: <a
href="https://link.springer.com/article/10.1007/BF01025492">citation</a></li>
<li><code>MSWHIM</code>: <a
href="https://pubs.acs.org/doi/10.1021/ci980211b">citation</a></li>
<li><code>ProtFP</code>: <a
href="https://pubmed.ncbi.nlm.nih.gov/24059694/">citation</a></li>
<li><code>stScales</code>: <a
href="https://pubmed.ncbi.nlm.nih.gov/19373543/">citation</a></li>
<li><code>tScales</code>: <a
href="https://www.sciencedirect.com/science/article/abs/pii/S0022286006006314">citation</a></li>
<li><code>VHSE</code>: <a
href="https://pubmed.ncbi.nlm.nih.gov/15895431/">citation</a></li>
<li><code>zScales</code>: <a
href="https://pubmed.ncbi.nlm.nih.gov/9651153/">citation</a></li>
</ul></li>
</ul>
<p>To examine the Atchley factors of amino acids across the CDR3
sequence for the first two samples:</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" tabindex="-1"></a><span class="fu">positionalProperty</span>(combined.TCR[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)], </span>
<span id="cb51-2"><a href="#cb51-2" tabindex="-1"></a>                  <span class="at">chain =</span> <span class="st">&quot;TRB&quot;</span>, </span>
<span id="cb51-3"><a href="#cb51-3" tabindex="-1"></a>                  <span class="at">aa.length =</span> <span class="dv">20</span>, </span>
<span id="cb51-4"><a href="#cb51-4" tabindex="-1"></a>                  <span class="at">method =</span> <span class="st">&quot;atchleyFactors&quot;</span>) <span class="sc">+</span> </span>
<span id="cb51-5"><a href="#cb51-5" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">hcl.colors</span>(<span class="dv">5</span>, <span class="st">&quot;inferno&quot;</span>)[<span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">4</span>)])</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-43-1.png" width="700" /></p>
<p>This function provides a detailed view of how physicochemical
properties of amino acids change along the CDR3 sequence. By visualizing
the mean property value and its confidence interval, one can identify
positions with distinct characteristics or significant variations
between groups, offering insights into structural and functional aspects
of the CDR3.</p>
</div>
<div id="percentkmer-motif-quantification" class="section level2">
<h2>percentKmer: Motif Quantification</h2>
<p>Another quantification of the composition of the CDR3 sequence is to
define motifs by sliding across the amino acid or nucleotide sequences
at set intervals, resulting in substrings or kmers.</p>
<p>Key Parameter(s) for <code>percentKmer()</code>:</p>
<ul>
<li><code>cloneCall</code>: Defines the clonal sequence grouping; only
accepts <code>aa</code> (amino acids) or <code>nt</code>
(nucleotides)</li>
<li><code>motif.length</code>: The length of the kmer to analyze.</li>
<li><code>top.motifs</code>: Displays the n most variable motifs as a
function of median absolute deviation.</li>
</ul>
<p>To visualize the percentage of the top 25 most variable 3-mer amino
acid motifs for the TRB chain:</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" tabindex="-1"></a><span class="fu">percentKmer</span>(combined.TCR, </span>
<span id="cb52-2"><a href="#cb52-2" tabindex="-1"></a>            <span class="at">cloneCall =</span> <span class="st">&quot;aa&quot;</span>,</span>
<span id="cb52-3"><a href="#cb52-3" tabindex="-1"></a>            <span class="at">chain =</span> <span class="st">&quot;TRB&quot;</span>, </span>
<span id="cb52-4"><a href="#cb52-4" tabindex="-1"></a>            <span class="at">motif.length =</span> <span class="dv">3</span>, </span>
<span id="cb52-5"><a href="#cb52-5" tabindex="-1"></a>            <span class="at">top.motifs =</span> <span class="dv">25</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-44-1.png" width="700" /></p>
<p>To perform the same analysis but for nucleotide motifs:</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" tabindex="-1"></a><span class="fu">percentKmer</span>(combined.TCR, </span>
<span id="cb53-2"><a href="#cb53-2" tabindex="-1"></a>            <span class="at">cloneCall =</span> <span class="st">&quot;nt&quot;</span>,</span>
<span id="cb53-3"><a href="#cb53-3" tabindex="-1"></a>            <span class="at">chain =</span> <span class="st">&quot;TRB&quot;</span>, </span>
<span id="cb53-4"><a href="#cb53-4" tabindex="-1"></a>            <span class="at">motif.length =</span> <span class="dv">3</span>, </span>
<span id="cb53-5"><a href="#cb53-5" tabindex="-1"></a>            <span class="at">top.motifs =</span> <span class="dv">25</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-45-1.png" width="700" /></p>
<p>The heatmaps generated by <code>percentKmer()</code> illustrate the
relative composition of frequently occurring k-mer motifs across samples
or groups. This analysis can highlight recurrent sequence patterns in
the CDR3, which may be associated with specific antigen recognition,
disease states, or processing mechanisms. By examining the most variable
motifs, researchers can identify key sequence features that
differentiate immune repertoires.</p>
<hr />
</div>
</div>
<div id="comparing-clonal-diversity-and-overlap" class="section level1">
<h1>Comparing Clonal Diversity and Overlap</h1>
<div id="clonaldiversity-clonal-diversity-quantification"
class="section level2">
<h2>clonalDiversity: Clonal Diversity Quantification</h2>
<p>Diversity can be measured for samples or by other variables.
<code>clonalDiversity()</code> calculates a specified diversity metric
for groups within a dataset. To control for variations in library size,
the function can perform bootstrapping with downsampling. It resamples
each group to the size of the smallest group and calculates the
diversity metric across multiple iterations, returning the mean
values.</p>
<div id="how-clonaldiversity-handles-downsampling-and-bootstrapping"
class="section level3">
<h3>How <code>clonalDiversity()</code> Handles Downsampling and
Bootstrapping</h3>
<ul>
<li>It determines the number of clones in the smallest group.</li>
<li>For each group, it performs <code>n.boots</code> iterations (default
= 100).</li>
<li>In each iteration, it randomly samples the clones (with replacement)
down to the size of the smallest group.</li>
<li>It calculates the selected diversity metric on this downsampled
set.</li>
<li>The final reported diversity value is the mean of the results from
all bootstrap iterations.</li>
<li>This process can be disabled by setting
<code>skip.boots = TRUE</code>.</li>
</ul>
</div>
<div id="available-diversity-metrics-metric" class="section level3">
<h3>Available Diversity Metrics (<code>metric</code>)</h3>
<table>
<colgroup>
<col width="6%" />
<col width="50%" />
<col width="7%" />
<col width="36%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"><strong>Metric</strong></th>
<th align="left"><strong>What it measures</strong></th>
<th align="left"><strong>Typical scale</strong></th>
<th align="left"><strong>Higher value indicates…</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left"><code>shannon</code></td>
<td align="left">Joint <em>richness + evenness</em> of clonotypes;
moderately sensitive to rare clones (entropy of the frequency
distribution).</td>
<td align="left">0 → ∞</td>
<td align="left"><strong>More distinct clonotypes</strong> <em>and</em>
a more even clone-size distribution.</td>
</tr>
<tr class="even">
<td align="left"><code>inv.simpson</code></td>
<td align="left">Evenness-weighted diversity (inverse Simpson’s D);
heavily penalises dominance by a few large clones.</td>
<td align="left">1 → ∞</td>
<td align="left"><strong>Few highly dominant clones</strong>; diversity
spread across many mid-abundance clones.</td>
</tr>
<tr class="odd">
<td align="left"><code>norm.entropy</code></td>
<td align="left">Shannon entropy divided by ln <em>S</em> (where
<em>S</em> = no. clonotypes); removes library-size dependence.</td>
<td align="left">0 → 1</td>
<td align="left">Same interpretation as <code>shannon</code>, but
<strong>comparable across samples of different depths</strong>.</td>
</tr>
<tr class="even">
<td align="left"><code>gini.simpson</code></td>
<td align="left">1 – Simpson’s D; probability that two reads drawn at
random come from <strong>different</strong> clonotypes.</td>
<td align="left">0 → 1</td>
<td align="left"><strong>Greater heterogeneity</strong> (higher chance
the two reads represent distinct clones).</td>
</tr>
<tr class="odd">
<td align="left"><code>chao1</code></td>
<td align="left">Estimated <em>richness only</em> (number of
clonotypes), correcting for unseen singletons/doubletons.</td>
<td align="left">0 → ∞</td>
<td align="left"><strong>More unique clonotypes</strong>, regardless of
their relative abundances.</td>
</tr>
<tr class="even">
<td align="left"><code>ACE</code></td>
<td align="left">Abundance-based Coverage Estimator—alternative richness
estimator robust when many clones are rare.</td>
<td align="left">0 → ∞</td>
<td align="left"><strong>More unique clonotypes</strong>, especially
when the repertoire contains many low-frequency clones.</td>
</tr>
<tr class="odd">
<td align="left"><code>gini</code></td>
<td align="left">Inequality of clone size distribution (Gini
Coefficient); measures how unevenly clone sizes are distributed.</td>
<td align="left">0 → 1</td>
<td align="left"><strong>Greater inequality</strong>; a few highly
abundant clones dominate the repertoire.</td>
</tr>
<tr class="even">
<td align="left"><code>d50</code></td>
<td align="left">Repertoire dominance; how many of the top clones are
needed to constitute 50% of the total library.</td>
<td align="left">1 → ∞</td>
<td align="left"><strong>Less dominance</strong> by top clones; the
repertoire is more evenly distributed.</td>
</tr>
<tr class="odd">
<td align="left"><code>hill0</code></td>
<td align="left">Richness (Hill number q=0); the total number of unique
clonotypes observed in the sample.</td>
<td align="left">0 → ∞</td>
<td align="left"><strong>More unique clonotypes</strong> (greater raw
richness).</td>
</tr>
<tr class="even">
<td align="left"><code>hill1</code></td>
<td align="left">Effective number of abundant clonotypes (Hill number
q=1); the exponential of Shannon entropy.</td>
<td align="left">1 → ∞</td>
<td align="left"><strong>Higher diversity</strong> of common clonotypes,
balancing richness and evenness.</td>
</tr>
<tr class="odd">
<td align="left"><code>hill2</code></td>
<td align="left">Effective number of dominant clonotypes (Hill number
q=2); equivalent to the inverse Simpson index.</td>
<td align="left">1 → ∞</td>
<td align="left"><strong>Higher diversity</strong> among the most
dominant clonotypes in the repertoire.</td>
</tr>
</tbody>
</table>
<p>Key Parameters for <code>clonalDiversity()</code></p>
<ul>
<li><code>metric</code>: The diversity metric to calculate.</li>
<li><code>x.axis</code>: An additional metadata variable to group
samples along the x-axis.</li>
<li><code>return.boots</code>: If <code>TRUE</code>, returns all
bootstrap values instead of the mean.</li>
<li><code>skip.boots</code>: If <code>TRUE</code>, disables downsampling
and bootstrapping.</li>
<li><code>n.boots</code>: The number of bootstrap iterations to perform
(default is 100).</li>
</ul>
<p>To calculate Shannon diversity, grouped by sample:</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" tabindex="-1"></a><span class="fu">clonalDiversity</span>(combined.TCR, </span>
<span id="cb54-2"><a href="#cb54-2" tabindex="-1"></a>                <span class="at">cloneCall =</span> <span class="st">&quot;gene&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-46-1.png" width="700" /></p>
<p>There are two options for grouping in <code>clonalDiversity()</code>:
<code>group.by</code> and <code>x.axis</code>.</p>
<ul>
<li><code>group.by</code>: Reorganizes the clone information into new
groups that the calculation will be based on.</li>
<li><code>x.axis</code>: Keeps the organization of the clone information
the same, but plots along the x-axis for improved visibility or
grouping.</li>
</ul>
<p>First, add a “Patient” variable to combined.TCR</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" tabindex="-1"></a>combined.TCR <span class="ot">&lt;-</span> <span class="fu">addVariable</span>(combined.TCR, </span>
<span id="cb55-2"><a href="#cb55-2" tabindex="-1"></a>                            <span class="at">variable.name =</span> <span class="st">&quot;Patient&quot;</span>, </span>
<span id="cb55-3"><a href="#cb55-3" tabindex="-1"></a>                             <span class="at">variables =</span> <span class="fu">c</span>(<span class="st">&quot;P17&quot;</span>, <span class="st">&quot;P17&quot;</span>, <span class="st">&quot;P18&quot;</span>, <span class="st">&quot;P18&quot;</span>, </span>
<span id="cb55-4"><a href="#cb55-4" tabindex="-1"></a><span class="st">&quot;P19&quot;</span>,<span class="st">&quot;P19&quot;</span>, <span class="st">&quot;P20&quot;</span>, <span class="st">&quot;P20&quot;</span>))</span></code></pre></div>
<p>Now, calculate Inverse Simpson diversity, grouped by “Patient”:</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" tabindex="-1"></a><span class="fu">clonalDiversity</span>(combined.TCR, </span>
<span id="cb56-2"><a href="#cb56-2" tabindex="-1"></a>                <span class="at">cloneCall =</span> <span class="st">&quot;gene&quot;</span>, </span>
<span id="cb56-3"><a href="#cb56-3" tabindex="-1"></a>                <span class="at">group.by =</span> <span class="st">&quot;Patient&quot;</span>, </span>
<span id="cb56-4"><a href="#cb56-4" tabindex="-1"></a>                <span class="at">metric =</span> <span class="st">&quot;inv.simpson&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-48-1.png" width="700" /></p>
<p>Calculate Inverse Simpson diversity with “Patient” on the x-axis,
keeping the original grouping:</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" tabindex="-1"></a><span class="fu">clonalDiversity</span>(combined.TCR, </span>
<span id="cb57-2"><a href="#cb57-2" tabindex="-1"></a>                <span class="at">cloneCall =</span> <span class="st">&quot;gene&quot;</span>, </span>
<span id="cb57-3"><a href="#cb57-3" tabindex="-1"></a>                <span class="at">x.axis =</span> <span class="st">&quot;Patient&quot;</span>, </span>
<span id="cb57-4"><a href="#cb57-4" tabindex="-1"></a>                <span class="at">metric =</span> <span class="st">&quot;inv.simpson&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-49-1.png" width="700" /></p>
<p><code>clonalDiversity()</code> functions in quantifying and comparing
the diversity of immune repertoires across different samples or
experimental conditions. By implementing bootstrapping and offering a
wide range of diversity metrics, it provides robust and comparable
measures of clonal richness and evenness, which are fundamental for
understanding immune responses and disease states</p>
</div>
</div>
<div id="clonalrarefaction-sampling-based-extrapolation"
class="section level2">
<h2>clonalRarefaction: Sampling-based Extrapolation</h2>
<p><code>clonalRarefaction()</code> uses Hill numbers to estimate
rarefaction, or estimating species richness, based on the abundance of
clones across groupings. The underlying rarefaction calculation uses the
observed receptor abundance to compute diversity. This function relies
on the <code>iNEXT</code> R package.</p>
<div id="hill-numbers-and-their-interpretation" class="section level3">
<h3>Hill Numbers and Their Interpretation</h3>
<ul>
<li><code>0</code>: Species richness</li>
<li><code>1</code>: Shannon Diversity</li>
<li><code>2</code>: Simpson Diversity</li>
</ul>
</div>
<div id="plot.type-options" class="section level3">
<h3>plot.type Options</h3>
<ul>
<li><code>1</code>: Sample-size-based rarefaction/extrapolation
curve</li>
<li><code>2</code>: Sample completeness curve</li>
<li><code>3</code>: Coverage-based rarefaction/extrapolation curve</li>
</ul>
<p>Key Parameter(s) for <code>clonalRarefaction()</code></p>
<ul>
<li><code>plot.type</code>: The type of plot to generate.</li>
<li><code>hill.numbers</code>: The Hill numbers to be plotted (0, 1, or
2).</li>
<li><code>n.boots</code>: The number of bootstrap replicates used to
derive confidence intervals (default is 20).</li>
</ul>
<p>This function relies on the <a
href="https://cran.r-project.org/web/packages/iNEXT/index.html">iNEXT</a>
with the accompanying <a
href="http://chao.stat.nthu.edu.tw/wordpress/paper/120_pdf_appendix.pdf">manuscript</a>.
Like the other wrapping functions in scRepertoire, please cite the
original work. The sample completeness curve (<strong>plot.type</strong>
= 2), may not show full sample coverage due to the size/diversity of the
input data.</p>
</div>
<div id="rarefaction-using-species-richness-q-0" class="section level3">
<h3>Rarefaction using Species Richness (q = 0)</h3>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" tabindex="-1"></a><span class="fu">clonalRarefaction</span>(combined.TCR,</span>
<span id="cb58-2"><a href="#cb58-2" tabindex="-1"></a>                  <span class="at">plot.type =</span> <span class="dv">1</span>,</span>
<span id="cb58-3"><a href="#cb58-3" tabindex="-1"></a>                  <span class="at">hill.numbers =</span> <span class="dv">0</span>,</span>
<span id="cb58-4"><a href="#cb58-4" tabindex="-1"></a>                  <span class="at">n.boots =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-50-1.png" width="700" /></p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" tabindex="-1"></a><span class="fu">clonalRarefaction</span>(combined.TCR,</span>
<span id="cb59-2"><a href="#cb59-2" tabindex="-1"></a>                  <span class="at">plot.type =</span> <span class="dv">2</span>,</span>
<span id="cb59-3"><a href="#cb59-3" tabindex="-1"></a>                  <span class="at">hill.numbers =</span> <span class="dv">0</span>,</span>
<span id="cb59-4"><a href="#cb59-4" tabindex="-1"></a>                  <span class="at">n.boots =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-50-2.png" width="700" /></p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" tabindex="-1"></a><span class="fu">clonalRarefaction</span>(combined.TCR,</span>
<span id="cb60-2"><a href="#cb60-2" tabindex="-1"></a>                  <span class="at">plot.type =</span> <span class="dv">3</span>,</span>
<span id="cb60-3"><a href="#cb60-3" tabindex="-1"></a>                  <span class="at">hill.numbers =</span> <span class="dv">0</span>,</span>
<span id="cb60-4"><a href="#cb60-4" tabindex="-1"></a>                  <span class="at">n.boots =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-50-3.png" width="700" /></p>
</div>
<div id="rarefaction-using-shannon-diversity-q-1"
class="section level3">
<h3>Rarefaction using Shannon Diversity (q = 1)</h3>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" tabindex="-1"></a><span class="fu">clonalRarefaction</span>(combined.TCR,</span>
<span id="cb61-2"><a href="#cb61-2" tabindex="-1"></a>                  <span class="at">plot.type =</span> <span class="dv">1</span>,</span>
<span id="cb61-3"><a href="#cb61-3" tabindex="-1"></a>                  <span class="at">hill.numbers =</span> <span class="dv">1</span>,</span>
<span id="cb61-4"><a href="#cb61-4" tabindex="-1"></a>                  <span class="at">n.boots =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-51-1.png" width="700" /></p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" tabindex="-1"></a><span class="fu">clonalRarefaction</span>(combined.TCR,</span>
<span id="cb62-2"><a href="#cb62-2" tabindex="-1"></a>                  <span class="at">plot.type =</span> <span class="dv">2</span>,</span>
<span id="cb62-3"><a href="#cb62-3" tabindex="-1"></a>                  <span class="at">hill.numbers =</span> <span class="dv">1</span>,</span>
<span id="cb62-4"><a href="#cb62-4" tabindex="-1"></a>                  <span class="at">n.boots =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-51-2.png" width="700" /></p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" tabindex="-1"></a><span class="fu">clonalRarefaction</span>(combined.TCR,</span>
<span id="cb63-2"><a href="#cb63-2" tabindex="-1"></a>                  <span class="at">plot.type =</span> <span class="dv">3</span>,</span>
<span id="cb63-3"><a href="#cb63-3" tabindex="-1"></a>                  <span class="at">hill.numbers =</span> <span class="dv">1</span>,</span>
<span id="cb63-4"><a href="#cb63-4" tabindex="-1"></a>                  <span class="at">n.boots =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-51-3.png" width="700" /></p>
<p><code>clonalRarefaction()</code> provides robust estimates of immune
repertoire diversity, allowing for comparison of richness and effective
diversity across samples, even with varying sequencing depths. By
visualizing rarefaction and extrapolation curves, researchers can assess
the completeness of their sampling and make fair comparisons of
diversity, which is essential for understanding immune complexity.</p>
</div>
</div>
<div id="clonalsizedistribution-modeling-clonal-composition"
class="section level2">
<h2>clonalSizeDistribution: Modeling Clonal Composition</h2>
<p>Another method for modeling the repertoire distribution is a discrete
gamma-GPD spliced threshold model, proposed by <a
href="https://pubmed.ncbi.nlm.nih.gov/30485278/">Koch et al.</a> The
spliced model models the repertoire and allows for the application of a
power law distribution for larger clonal-expanded sequences and a
Poisson distribution for smaller clones. After fitting the models,
repertoires can be compared using Euclidean distance. If using this
function, please read/cite <a
href="https://pubmed.ncbi.nlm.nih.gov/30485278/">Koch et al.</a> and
check out the <a
href="https://bioconductor.org/packages/release/bioc/html/powerTCR.html">powerTCR</a>
R package.</p>
<p>Key Parameter(s) for <code>clonalSizeDistribution()</code></p>
<ul>
<li><code>method</code>: The agglomeration method for hierarchical
clustering (e.g., “ward.D2”).</li>
</ul>
<p>To model the clonal size distribution using amino acid clone calls
and hierarchical clustering with the “ward.D2” method:</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" tabindex="-1"></a><span class="fu">clonalSizeDistribution</span>(combined.TCR, </span>
<span id="cb64-2"><a href="#cb64-2" tabindex="-1"></a>                       <span class="at">cloneCall =</span> <span class="st">&quot;aa&quot;</span>, </span>
<span id="cb64-3"><a href="#cb64-3" tabindex="-1"></a>                       <span class="at">method=</span> <span class="st">&quot;ward.D2&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-52-1.png" width="700" /></p>
<p><code>clonalSizeDistribution()</code> offers a sophisticated approach
to modeling immune repertoire composition, distinguishing between the
distribution of rare and expanded clones. By applying a spliced
statistical model, it provides a more accurate representation of the
repertoire’s underlying clonal architecture, enabling robust comparisons
and a deeper understanding of immune system dynamics.</p>
</div>
<div id="clonaloverlap-exploring-sequence-overlap"
class="section level2">
<h2>clonalOverlap: Exploring Sequence Overlap</h2>
<p>If you are interested in measures of similarity between the samples
loaded into scRepertoire, using <code>clonalOverlap()</code> can assist
in the visualization. The underlying <code>clonalOverlap()</code>
calculation varies by the <code>method</code> parameter.</p>
<div id="method-parameters-for-clonaloverlap" class="section level3">
<h3><code>method</code> Parameters for <code>clonalOverlap()</code></h3>
<ul>
<li><code>overlap</code> - Overlap coefficient</li>
<li><code>morisita</code> - Morisita’s overlap index</li>
<li><code>jaccard</code> - Jaccard index</li>
<li><code>cosine</code> - Cosine similarity</li>
<li><code>raw</code> - Exact number of overlapping clones</li>
</ul>
<p>To calculate and visualize the Morisita overlap index using strict
clone calls:</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" tabindex="-1"></a><span class="fu">clonalOverlap</span>(combined.TCR, </span>
<span id="cb65-2"><a href="#cb65-2" tabindex="-1"></a>              <span class="at">cloneCall =</span> <span class="st">&quot;strict&quot;</span>, </span>
<span id="cb65-3"><a href="#cb65-3" tabindex="-1"></a>              <span class="at">method =</span> <span class="st">&quot;morisita&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-53-1.png" width="700" /></p>
<p>To calculate and visualize the raw number of overlapping clones using
strict clone calls:</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" tabindex="-1"></a><span class="fu">clonalOverlap</span>(combined.TCR, </span>
<span id="cb66-2"><a href="#cb66-2" tabindex="-1"></a>              <span class="at">cloneCall =</span> <span class="st">&quot;strict&quot;</span>, </span>
<span id="cb66-3"><a href="#cb66-3" tabindex="-1"></a>              <span class="at">method =</span> <span class="st">&quot;raw&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-54-1.png" width="700" /></p>
<p><code>clonalOverlap()</code> is a tool for assessing the similarity
and shared clonotypes between different immune repertoires. By offering
various quantitative methods and a clear heatmap visualization, it
allows researchers to identify the degree of overlap, providing insights
into shared immune responses, cross-reactivity, or the impact of
different treatments on clonal composition.</p>
<hr />
</div>
</div>
</div>
<div id="combining-clones-and-single-cell-objects"
class="section level1">
<h1>Combining Clones and Single-Cell Objects</h1>
<p>The data in the scRepertoire package is derived from a <a
href="https://pubmed.ncbi.nlm.nih.gov/33622974/">study</a> of acute
respiratory stress disorder in the context of bacterial and COVID-19
infections. The internal single cell data (<code>scRep_example()</code>)
built in to scRepertoire is randomly sampled 500 cells from the fully
integrated Seurat object to minimize the package size. We will use both
Seurat and Single-Cell Experiment (SCE) with scater to perform further
visualizations in tandem.</p>
<div id="preprocessed-single-cell-object" class="section level2">
<h2>Preprocessed Single-Cell Object</h2>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" tabindex="-1"></a>scRep_example <span class="ot">&lt;-</span> <span class="fu">get</span>(<span class="fu">data</span>(<span class="st">&quot;scRep_example&quot;</span>))</span>
<span id="cb67-2"><a href="#cb67-2" tabindex="-1"></a></span>
<span id="cb67-3"><a href="#cb67-3" tabindex="-1"></a><span class="co">#Making a Single-Cell Experiment object</span></span>
<span id="cb67-4"><a href="#cb67-4" tabindex="-1"></a>sce <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">as.SingleCellExperiment</span>(scRep_example)</span>
<span id="cb67-5"><a href="#cb67-5" tabindex="-1"></a></span>
<span id="cb67-6"><a href="#cb67-6" tabindex="-1"></a><span class="co">#Adding patient information</span></span>
<span id="cb67-7"><a href="#cb67-7" tabindex="-1"></a>scRep_example<span class="sc">$</span>Patient <span class="ot">&lt;-</span> <span class="fu">substr</span>(scRep_example<span class="sc">$</span>orig.ident, <span class="dv">1</span>,<span class="dv">3</span>)</span>
<span id="cb67-8"><a href="#cb67-8" tabindex="-1"></a></span>
<span id="cb67-9"><a href="#cb67-9" tabindex="-1"></a><span class="co">#Adding type information</span></span>
<span id="cb67-10"><a href="#cb67-10" tabindex="-1"></a>scRep_example<span class="sc">$</span>Type <span class="ot">&lt;-</span> <span class="fu">substr</span>(scRep_example<span class="sc">$</span>orig.ident, <span class="dv">4</span>,<span class="dv">4</span>)</span></code></pre></div>
</div>
<div id="note-on-dimensional-reduction" class="section level2">
<h2>Note on Dimensional Reduction</h2>
<p>In single-cell RNA sequencing workflows, dimensional reduction is
typically performed by first identifying highly variable features. These
features are then used directly for UMAP/tSNE projection or as inputs
for principal component analysis. The same approach is commonly applied
to clustering as well.</p>
<p>However, in immune-focused datasets, VDJ genes from TCR and BCR are
often among the most variable genes. This variability arises naturally
due to clonal expansion and diversity within lymphocytes. As a result,
UMAP projections and clustering outcomes may be influenced by clonal
information rather than broader transcriptional differences across cell
types.</p>
<p>To mitigate this issue, a common strategy is to exclude VDJ genes
from the set of highly variable features before proceeding with
clustering and dimensional reduction. We introduce a set of functions
that facilitate this process by removing VDJ-related genes from either a
Seurat Object or a vector of gene names (useful for SCE-based
workflows).</p>
<div id="functions-to-exclude-vdj-genes" class="section level3">
<h3>Functions to Exclude VDJ Genes</h3>
<ul>
<li><code>quietVDJgenes()</code> – Removes both TCR and BCR VDJ
genes.</li>
<li><code>quietTCRgenes()</code> – Removes only TCR VDJ genes.</li>
<li><code>quietBCRgenes()</code> – Removes only BCR VDJ genes, but
retains BCR VDJ pseudogenes in the variable features.</li>
</ul>
<p>Let’s first check the top 10 variable features in the scRep_example
Seurat object before any removal:</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" tabindex="-1"></a><span class="co"># Check the first 10 variable features before removal</span></span>
<span id="cb68-2"><a href="#cb68-2" tabindex="-1"></a><span class="fu">VariableFeatures</span>(scRep_example)[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>##  [1] &quot;TRBV7-2&quot;  &quot;HSPA1B&quot;   &quot;HSPA1A&quot;   &quot;TRBV4-1&quot;  &quot;CCL4&quot;     &quot;TRBV5-1&quot; 
##  [7] &quot;TRBV10-3&quot; &quot;TRBV3-1&quot;  &quot;TRBV6-6&quot;  &quot;CD79A&quot;</code></pre>
<p>Now, we’ll remove TCR VDJ genes from the scRep_example object:</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" tabindex="-1"></a><span class="co"># Remove TCR VDJ genes</span></span>
<span id="cb70-2"><a href="#cb70-2" tabindex="-1"></a>scRep_example <span class="ot">&lt;-</span> <span class="fu">quietTCRgenes</span>(scRep_example)</span>
<span id="cb70-3"><a href="#cb70-3" tabindex="-1"></a></span>
<span id="cb70-4"><a href="#cb70-4" tabindex="-1"></a><span class="co"># Check the first 10 variable features after removal</span></span>
<span id="cb70-5"><a href="#cb70-5" tabindex="-1"></a><span class="fu">VariableFeatures</span>(scRep_example)[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>##  [1] &quot;HSPA1B&quot;  &quot;HSPA1A&quot;  &quot;CCL4&quot;    &quot;CD79A&quot;   &quot;HLA-DRA&quot; &quot;CCL20&quot;   &quot;TUBA1B&quot; 
##  [8] &quot;HSPA6&quot;   &quot;TNF&quot;     &quot;MS4A1&quot;</code></pre>
<p>By applying these functions, you can ensure that clustering and
dimensional reduction are driven by broader transcriptomic differences
across cell types rather than being skewed by the inherent variability
due to clonal expansion. This provides a more accurate representation of
cellular heterogeneity independent of clonal lineage.</p>
</div>
</div>
<div id="preprocessed-single-cell-object-1" class="section level2">
<h2>Preprocessed Single-Cell Object</h2>
<p>The data in the <code>scRepertoire</code> package is derived from a
<a href="https://pubmed.ncbi.nlm.nih.gov/33622974/">study</a> of acute
respiratory stress disorder in the context of bacterial and COVID-19
infections. The internal single cell data (<code>scRep_example()</code>)
built in to scRepertoire is randomly sampled 500 cells from the fully
integrated Seurat object to minimize the package size. However, for the
purpose of the vignette we will use the full single-cell object with
30,000 cells. We will use both Seurat and Single-Cell Experiment (SCE)
with <code>scater</code> to perform further visualizations in
tandem.</p>
</div>
<div id="combineexpression" class="section level2">
<h2>combineExpression</h2>
<p>After processing the contig data into clones via
<code>combineBCR()</code> or <code>combineTCR()</code>, we can add the
clonal information to the single-cell object using
<code>combineExpression()</code>.</p>
<p><strong>Importantly</strong>, the major requirement for the
attachment is matching contig cell barcodes and barcodes in the row
names of the metadata of the Seurat or Single-Cell Experiment object. If
these do not match, the attachment will fail. We suggest making changes
to the single-cell object barcodes for ease of use.</p>
<div id="calculating-clonesize" class="section level3">
<h3>Calculating <code>cloneSize</code></h3>
<p>Part of <code>combineExpression()</code> is calculating the clonal
frequency and proportion, placing each clone into groups called
<code>cloneSize</code>. The default <code>cloneSize</code> argument uses
the following bins:
<code>c(Rare = 1e-4, Small = 0.001, Medium = 0.01, Large = 0.1, Hyperexpanded = 1)</code>,
which can be modified to include more/less bins or different names.</p>
<p>Clonal frequency and proportion are dependent on the repertoires
being compared. You can modify the calculation using the
<code>group.by</code> parameter, such as grouping by a Patient variable.
If <code>group.by</code> is not set, <code>combineExpression()</code>
will calculate clonal frequency, proportion, and <code>cloneSize</code>
as a function of individual sequencing runs. Additionally,
<code>cloneSize</code> can use the frequency of clones when
<code>proportion = FALSE</code>.</p>
<p>Key Parameter(s) for <code>combineExpression()</code></p>
<ul>
<li><code>input.data</code>: The product of <code>combineTCR()</code>,
<code>combineBCR()</code>, or a list containing both.</li>
<li><code>sc.data</code>: The Seurat or Single-Cell Experiment (SCE)
object to attach the clonal data to.</li>
<li><code>proportion</code>: If <code>TRUE</code> (default), calculates
the proportion of the clone; if <code>FALSE</code>, calculates total
frequency.</li>
<li><code>cloneSize</code>: Bins for grouping based on proportion or
frequency. If proportion is <code>FALSE</code> and cloneSize bins are
not set high enough, the upper limit will automatically adjust.</li>
<li><code>filterNA</code>: Method to subset the Seurat/SCE object of
barcodes without clone information.</li>
<li><code>addLabel</code>: Adds a label to the frequency header, useful
for trying multiple group.by variables or recalculating frequencies
after subsetting.</li>
</ul>
<p>We can look at the default cloneSize groupings using the Single-Cell
Experiment object we just created, with group.by set to the sample
variable used in combineTCR():</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">combineExpression</span>(combined.TCR, </span>
<span id="cb72-2"><a href="#cb72-2" tabindex="-1"></a>                         sce, </span>
<span id="cb72-3"><a href="#cb72-3" tabindex="-1"></a>                         <span class="at">cloneCall=</span><span class="st">&quot;gene&quot;</span>, </span>
<span id="cb72-4"><a href="#cb72-4" tabindex="-1"></a>                         <span class="at">group.by =</span> <span class="st">&quot;sample&quot;</span>, </span>
<span id="cb72-5"><a href="#cb72-5" tabindex="-1"></a>                         <span class="at">proportion =</span> <span class="cn">TRUE</span>)</span>
<span id="cb72-6"><a href="#cb72-6" tabindex="-1"></a></span>
<span id="cb72-7"><a href="#cb72-7" tabindex="-1"></a><span class="co">#Define color palette </span></span>
<span id="cb72-8"><a href="#cb72-8" tabindex="-1"></a>colorblind_vector <span class="ot">&lt;-</span> <span class="fu">hcl.colors</span>(<span class="at">n=</span><span class="dv">7</span>, <span class="at">palette =</span> <span class="st">&quot;inferno&quot;</span>, <span class="at">fixup =</span> <span class="cn">TRUE</span>)</span>
<span id="cb72-9"><a href="#cb72-9" tabindex="-1"></a></span>
<span id="cb72-10"><a href="#cb72-10" tabindex="-1"></a><span class="fu">plotUMAP</span>(sce, <span class="at">colour_by =</span> <span class="st">&quot;cloneSize&quot;</span>) <span class="sc">+</span></span>
<span id="cb72-11"><a href="#cb72-11" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">rev</span>(colorblind_vector[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">5</span>,<span class="dv">7</span>)]))</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-58-1.png" width="700" /></p>
<p>Alternatively, if we want <code>cloneSize</code> to be based on the
frequency of the clone, we can set proportion = FALSE and will need to
change the <code>cloneSize</code> bins to integers. If we haven’t
inspected our clone data, setting the upper limit of the clonal
frequency might be difficult; <code>combineExpression()</code> will
automatically adjust the upper limit to fit the distribution of the
frequencies.</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" tabindex="-1"></a>scRep_example <span class="ot">&lt;-</span> <span class="fu">combineExpression</span>(combined.TCR, </span>
<span id="cb73-2"><a href="#cb73-2" tabindex="-1"></a>                                   scRep_example, </span>
<span id="cb73-3"><a href="#cb73-3" tabindex="-1"></a>                                   <span class="at">cloneCall=</span><span class="st">&quot;gene&quot;</span>, </span>
<span id="cb73-4"><a href="#cb73-4" tabindex="-1"></a>                                   <span class="at">group.by =</span> <span class="st">&quot;sample&quot;</span>, </span>
<span id="cb73-5"><a href="#cb73-5" tabindex="-1"></a>                                   <span class="at">proportion =</span> <span class="cn">FALSE</span>, </span>
<span id="cb73-6"><a href="#cb73-6" tabindex="-1"></a>                                   <span class="at">cloneSize=</span><span class="fu">c</span>(<span class="at">Single=</span><span class="dv">1</span>, <span class="at">Small=</span><span class="dv">5</span>, <span class="at">Medium=</span><span class="dv">20</span>, <span class="at">Large=</span><span class="dv">100</span>, <span class="at">Hyperexpanded=</span><span class="dv">500</span>))</span>
<span id="cb73-7"><a href="#cb73-7" tabindex="-1"></a></span>
<span id="cb73-8"><a href="#cb73-8" tabindex="-1"></a>Seurat<span class="sc">::</span><span class="fu">DimPlot</span>(scRep_example, <span class="at">group.by =</span> <span class="st">&quot;cloneSize&quot;</span>) <span class="sc">+</span></span>
<span id="cb73-9"><a href="#cb73-9" tabindex="-1"></a>    <span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">rev</span>(colorblind_vector[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">7</span>)]))</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-59-1.png" width="700" /></p>
</div>
<div id="combining-both-tcr-and-bcr" class="section level3">
<h3>Combining both TCR and BCR</h3>
<p>If you have both TCR and BCR enrichment data, or wish to include
information for both gamma-delta and alpha-beta T cells, you can create
a single list containing the outputs of <code>combineTCR()</code> and
<code>combineBCR()</code> and then use
<code>combineExpression()</code>.</p>
<p><strong>Major Note</strong>: If there are duplicate barcodes (e.g.,
if a cell has both Ig and TCR information), the immune receptor
information will not be added for those cells. It might be worth
checking cluster identities and removing incongruent barcodes in the
products of <code>combineTCR()</code> and <code>combineBCR()</code>. As
an anecdote, the <a
href="https://support.10xgenomics.com/single-cell-vdj/datasets/6.0.1/SC5v2_Melanoma_5Kcells_Connect_single_channel_SC5v2_Melanoma_5Kcells_Connect_single_channel">testing
data</a> used to improve this function showed 5-6% barcode overlap.</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" tabindex="-1"></a><span class="co">#This is an example of the process, which will not be evaluated during knit</span></span>
<span id="cb74-2"><a href="#cb74-2" tabindex="-1"></a>TCR <span class="ot">&lt;-</span> <span class="fu">combineTCR</span>(...)</span>
<span id="cb74-3"><a href="#cb74-3" tabindex="-1"></a>BCR <span class="ot">&lt;-</span> <span class="fu">combineBCR</span>(...)</span>
<span id="cb74-4"><a href="#cb74-4" tabindex="-1"></a>list.receptors <span class="ot">&lt;-</span> <span class="fu">c</span>(TCR, BCR)</span>
<span id="cb74-5"><a href="#cb74-5" tabindex="-1"></a></span>
<span id="cb74-6"><a href="#cb74-6" tabindex="-1"></a></span>
<span id="cb74-7"><a href="#cb74-7" tabindex="-1"></a>seurat <span class="ot">&lt;-</span> <span class="fu">combineExpression</span>(list.receptors, </span>
<span id="cb74-8"><a href="#cb74-8" tabindex="-1"></a>                            seurat, </span>
<span id="cb74-9"><a href="#cb74-9" tabindex="-1"></a>                            <span class="at">cloneCall=</span><span class="st">&quot;gene&quot;</span>, </span>
<span id="cb74-10"><a href="#cb74-10" tabindex="-1"></a>                            <span class="at">proportion =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><code>combineExpression()</code> is a core function in
<code>scRepertoire</code> that bridges the immune repertoire data with
single-cell gene expression data. It enriches your Seurat or SCE object
with crucial clonal information, including calculated frequencies and
proportions, allowing for integrated analysis of cellular identity and
clonal dynamics. Its flexibility in defining <code>cloneSize</code> and
handling various grouping scenarios makes it adaptable to diverse
experimental designs, even accommodating the integration of both TCR and
BCR data.</p>
<hr />
</div>
</div>
</div>
<div id="visualizations-for-single-cell-objects" class="section level1">
<h1>Visualizations for Single-Cell Objects</h1>
<div id="clonaloverlay" class="section level2">
<h2>clonalOverlay</h2>
<p>Using the dimensional reduction graphs as a reference,
<code>clonalOverlay()</code> generates an overlay of the positions of
clonally-expanded cells. It highlights areas of high clonal frequency or
proportion on your UMAP or tSNE plots.</p>
<p>Key Parameters for clonalOverlay()</p>
<ul>
<li><code>sc.data</code>: The single-cell object
after<code>combineExpression()</code>.</li>
<li><code>reduction</code>: The dimensional reduction to visualize
(e.g., “umap”, “pca”). Default is “pca”.</li>
<li><code>cut.category</code>: The metadata variable to use for
filtering the overlay (e.g., “clonalFrequency” or
“clonalProportion”).</li>
<li><code>cutpoint</code>: The lowest clonal frequency or proportion to
include in the contour plot.</li>
<li><code>bins</code>: The number of contours to draw.</li>
</ul>
<p><code>clonalOverlay()</code> can be used to look across all cells or
faceted by a metadata variable using <code>facet.by</code>. The overall
dimensional reduction will be maintained as we facet, while the contour
plots will adjust based on the <code>facet.by</code> variable. The
coloring of the dot plot is based on the active identity of the
single-cell object.</p>
<p>This visualization was authored by Dr. Francesco Mazziotta and
inspired by Drs. Carmona and Andreatta and their work with <a
href="https://github.com/carmonalab/ProjecTILs">ProjectTIL</a>, a
pipeline for annotating T cell subtypes.</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" tabindex="-1"></a><span class="fu">clonalOverlay</span>(scRep_example, </span>
<span id="cb75-2"><a href="#cb75-2" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, </span>
<span id="cb75-3"><a href="#cb75-3" tabindex="-1"></a>              <span class="at">cutpoint =</span> <span class="dv">1</span>, </span>
<span id="cb75-4"><a href="#cb75-4" tabindex="-1"></a>              <span class="at">bins =</span> <span class="dv">10</span>, </span>
<span id="cb75-5"><a href="#cb75-5" tabindex="-1"></a>              <span class="at">facet.by =</span> <span class="st">&quot;Patient&quot;</span>) <span class="sc">+</span> </span>
<span id="cb75-6"><a href="#cb75-6" tabindex="-1"></a>              <span class="fu">guides</span>(<span class="at">color =</span> <span class="st">&quot;none&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-61-1.png" width="700" /></p>
</div>
<div id="clonalnetwork" class="section level2">
<h2>clonalNetwork</h2>
<p>Similar to <code>clonalOverlay()</code>, <code>clonalNetwork()</code>
visualizes the network interaction of clones shared between clusters
along the single-cell dimensional reduction. This function shows the
relative proportion of clones flowing from a starting node, with the
ending node indicated by an arrow</p>
<div id="filtering-options-for-clonalnetwork" class="section level3">
<h3>Filtering Options for clonalNetwork()</h3>
<ul>
<li><code>filter.clones</code>:
<ul>
<li>Select a number to isolate the clones comprising the top n number of
cells (e.g., <code>filter.clones = 2000</code>).</li>
<li>Select <code>min</code> to scale all groups to the size of the
minimum group.</li>
</ul></li>
<li><code>filter.identity</code>: For the identity chosen to visualize,
show the “to” and “from” network connections for a single group.</li>
<li><code>filter.proportion</code>: Remove clones from the network that
comprise less than a certain proportion of clones in groups.</li>
<li><code>filter.graph</code>: Remove reciprocal edges from one half of
the graph, allowing for cleaner visualization.</li>
</ul>
<p>Now, visualize the clonal network with no specific identity
filter:</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" tabindex="-1"></a><span class="co">#ggraph needs to be loaded due to issues with ggplot</span></span>
<span id="cb76-2"><a href="#cb76-2" tabindex="-1"></a><span class="fu">library</span>(ggraph)</span>
<span id="cb76-3"><a href="#cb76-3" tabindex="-1"></a></span>
<span id="cb76-4"><a href="#cb76-4" tabindex="-1"></a><span class="co">#No Identity filter</span></span>
<span id="cb76-5"><a href="#cb76-5" tabindex="-1"></a><span class="fu">clonalNetwork</span>(scRep_example, </span>
<span id="cb76-6"><a href="#cb76-6" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, </span>
<span id="cb76-7"><a href="#cb76-7" tabindex="-1"></a>              <span class="at">group.by =</span> <span class="st">&quot;seurat_clusters&quot;</span>,</span>
<span id="cb76-8"><a href="#cb76-8" tabindex="-1"></a>              <span class="at">filter.clones =</span> <span class="cn">NULL</span>,</span>
<span id="cb76-9"><a href="#cb76-9" tabindex="-1"></a>              <span class="at">filter.identity =</span> <span class="cn">NULL</span>,</span>
<span id="cb76-10"><a href="#cb76-10" tabindex="-1"></a>              <span class="at">cloneCall =</span> <span class="st">&quot;aa&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-62-1.png" width="700" /></p>
<p>We can look at the clonal relationships relative to a single cluster
using the <code>filter.identity</code> parameter. For example, focusing
on Cluster 3:</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" tabindex="-1"></a><span class="co">#Examining Cluster 3 only</span></span>
<span id="cb77-2"><a href="#cb77-2" tabindex="-1"></a><span class="fu">clonalNetwork</span>(scRep_example, </span>
<span id="cb77-3"><a href="#cb77-3" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, </span>
<span id="cb77-4"><a href="#cb77-4" tabindex="-1"></a>              <span class="at">group.by =</span> <span class="st">&quot;seurat_clusters&quot;</span>,</span>
<span id="cb77-5"><a href="#cb77-5" tabindex="-1"></a>              <span class="at">filter.identity =</span> <span class="dv">3</span>,</span>
<span id="cb77-6"><a href="#cb77-6" tabindex="-1"></a>              <span class="at">cloneCall =</span> <span class="st">&quot;aa&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-63-1.png" width="700" /></p>
<p>You can also use the <code>exportClones</code> parameter to quickly
get clones that are shared across multiple identity groups, along with
the total number of clone copies in the dataset.</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" tabindex="-1"></a>shared.clones <span class="ot">&lt;-</span> <span class="fu">clonalNetwork</span>(scRep_example, </span>
<span id="cb78-2"><a href="#cb78-2" tabindex="-1"></a>                               <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, </span>
<span id="cb78-3"><a href="#cb78-3" tabindex="-1"></a>                               <span class="at">group.by =</span> <span class="st">&quot;seurat_clusters&quot;</span>,</span>
<span id="cb78-4"><a href="#cb78-4" tabindex="-1"></a>                               <span class="at">cloneCall =</span> <span class="st">&quot;aa&quot;</span>, </span>
<span id="cb78-5"><a href="#cb78-5" tabindex="-1"></a>                               <span class="at">exportClones =</span> <span class="cn">TRUE</span>)</span>
<span id="cb78-6"><a href="#cb78-6" tabindex="-1"></a><span class="fu">head</span>(shared.clones)</span></code></pre></div>
<pre><code>## [38;5;246m# A tibble: 6 × 2[39m
##   clone                                sum
##   [3m[38;5;246m&lt;fct&gt;[39m[23m                              [3m[38;5;246m&lt;int&gt;[39m[23m
## [38;5;250m1[39m CVVSDNTGGFKTIF_CASSVRRERANTGELFF      11
## [38;5;250m2[39m CAELNQAGTALIF_CASSQAPFSTSGELFF         3
## [38;5;250m3[39m CASLSGSARQLTF_CASSSTVAGEQYF            3
## [38;5;250m4[39m NA_CASSVRRERANTGELFF                   3
## [38;5;250m5[39m CAERGSGGSYIPTF_CASSDPSGRQGPRWDTQYF     2
## [38;5;250m6[39m CAGAAETSGSRLTF_CASSRLRTGALYEQYF        2</code></pre>
</div>
</div>
<div id="highlightclones" class="section level2">
<h2>highlightClones</h2>
<p>The <code>highlightClones()</code> function allows you to
specifically visualize the distribution of particular clonal sequences
on your single-cell dimensional reduction plots. This helps in tracking
the location and expansion of clones of interest.</p>
<p>Key Parameters for <code>highlightClones()</code></p>
<ul>
<li><code>cloneCall</code>: The type of sequence to use for highlighting
(e.g., “aa”, “nt”, “strict”).</li>
<li><code>sequence</code>: A character vector of the specific clonal
sequences to highlight.</li>
</ul>
<p>To highlight the most prominent amino acid sequences:
<em>CAERGSGGSYIPTF_CASSDPSGRQGPRWDTQYF</em> and
<em>CARKVRDSSYKLIF_CASSDSGYNEQFF</em>:</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" tabindex="-1"></a>scRep_example <span class="ot">&lt;-</span> <span class="fu">highlightClones</span>(scRep_example, </span>
<span id="cb80-2"><a href="#cb80-2" tabindex="-1"></a>                    <span class="at">cloneCall=</span> <span class="st">&quot;aa&quot;</span>, </span>
<span id="cb80-3"><a href="#cb80-3" tabindex="-1"></a>                    <span class="at">sequence =</span> <span class="fu">c</span>(<span class="st">&quot;CAERGSGGSYIPTF_CASSDPSGRQGPRWDTQYF&quot;</span>, </span>
<span id="cb80-4"><a href="#cb80-4" tabindex="-1"></a>                                 <span class="st">&quot;CARKVRDSSYKLIF_CASSDSGYNEQFF&quot;</span>))</span>
<span id="cb80-5"><a href="#cb80-5" tabindex="-1"></a></span>
<span id="cb80-6"><a href="#cb80-6" tabindex="-1"></a>Seurat<span class="sc">::</span><span class="fu">DimPlot</span>(scRep_example, <span class="at">group.by =</span> <span class="st">&quot;highlight&quot;</span>) <span class="sc">+</span> </span>
<span id="cb80-7"><a href="#cb80-7" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color=</span><span class="fu">guide_legend</span>(<span class="at">nrow=</span><span class="dv">3</span>,<span class="at">byrow=</span><span class="cn">TRUE</span>)) <span class="sc">+</span> </span>
<span id="cb80-8"><a href="#cb80-8" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_blank</span>(), </span>
<span id="cb80-9"><a href="#cb80-9" tabindex="-1"></a>                 <span class="at">legend.position =</span> <span class="st">&quot;bottom&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-65-1.png" width="700" /></p>
</div>
<div id="clonaloccupy" class="section level2">
<h2>clonalOccupy</h2>
<p><code>clonalOccupy()</code> visualizes the count of cells by cluster,
categorized into specific clonal frequency ranges. It uses the cloneSize
metadata variable (generated by <code>combineExpression()</code>) to
plot the number of cells within each clonal expansion designation, using
a secondary variable like cluster. Credit for the idea goes to Drs.
Carmona and Andreatta.</p>
<p>Key Parameters for <code>clonalOccupy()</code> * <code>x.axis</code>:
The variable in the metadata to graph along the x-axis (e.g.,
“seurat_clusters”, “ident”). * <code>label</code>: If TRUE, includes the
number of clones in each category by x.axis variable. *
<code>proportion</code>: If TRUE, converts the stacked bars into
relative proportions. * <code>na.include</code>: If TRUE, visualizes NA
values.</p>
<p>To visualize the count of cells by <code>seurat_clusters</code> based
on <code>cloneSize</code> groupings:</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" tabindex="-1"></a><span class="fu">clonalOccupy</span>(scRep_example, </span>
<span id="cb81-2"><a href="#cb81-2" tabindex="-1"></a>              <span class="at">x.axis =</span> <span class="st">&quot;seurat_clusters&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-66-1.png" width="700" /></p>
<p>To visualize the proportion of cells by <code>ident</code> (active
identity), without labels:</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" tabindex="-1"></a><span class="fu">clonalOccupy</span>(scRep_example, </span>
<span id="cb82-2"><a href="#cb82-2" tabindex="-1"></a>             <span class="at">x.axis =</span> <span class="st">&quot;ident&quot;</span>, </span>
<span id="cb82-3"><a href="#cb82-3" tabindex="-1"></a>             <span class="at">proportion =</span> <span class="cn">TRUE</span>, </span>
<span id="cb82-4"><a href="#cb82-4" tabindex="-1"></a>             <span class="at">label =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-67-1.png" width="700" /></p>
</div>
<div id="alluvialclones" class="section level2">
<h2>alluvialClones</h2>
<p>After the metadata has been modified with clonal information,
<code>alluvialClones()</code> allows you to look at clones across
multiple categorical variables, enabling the examination of the
interchange between these variables. Because this function produces a
graph with each clone arranged by called stratification, it may take
some time depending on the size of the repertoire.</p>
<p>Key Parameters for <code>alluvialClones()</code></p>
<ul>
<li><code>y.axes</code>: The columns that will separate the proportional
visualizations.</li>
<li><code>color</code>: The column header or clone(s) to be
highlighted.</li>
<li><code>facet</code>: The column label to separate facets.</li>
<li><code>alpha</code>: The column header to have gradated opacity.</li>
</ul>
<p>To visualize clonal flow across “Patient”, “ident”, and “Type”,
highlighting specific amino acid clones:</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" tabindex="-1"></a><span class="fu">alluvialClones</span>(scRep_example, </span>
<span id="cb83-2"><a href="#cb83-2" tabindex="-1"></a>               <span class="at">cloneCall =</span> <span class="st">&quot;aa&quot;</span>, </span>
<span id="cb83-3"><a href="#cb83-3" tabindex="-1"></a>               <span class="at">y.axes =</span> <span class="fu">c</span>(<span class="st">&quot;Patient&quot;</span>, <span class="st">&quot;ident&quot;</span>, <span class="st">&quot;Type&quot;</span>), </span>
<span id="cb83-4"><a href="#cb83-4" tabindex="-1"></a>               <span class="at">color =</span> <span class="fu">c</span>(<span class="st">&quot;CVVSDNTGGFKTIF_CASSVRRERANTGELFF&quot;</span>, <span class="st">&quot;NA_CASSVRRERANTGELFF&quot;</span>)) <span class="sc">+</span> </span>
<span id="cb83-5"><a href="#cb83-5" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;grey&quot;</span>, colorblind_vector[<span class="dv">3</span>]))</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-68-1.png" width="700" /></p>
<p>To visualize clonal flow across “Patient”, “ident”, and “Type”,
coloring by “ident”:</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" tabindex="-1"></a><span class="fu">alluvialClones</span>(scRep_example, </span>
<span id="cb84-2"><a href="#cb84-2" tabindex="-1"></a>                   <span class="at">cloneCall =</span> <span class="st">&quot;gene&quot;</span>, </span>
<span id="cb84-3"><a href="#cb84-3" tabindex="-1"></a>                   <span class="at">y.axes =</span> <span class="fu">c</span>(<span class="st">&quot;Patient&quot;</span>, <span class="st">&quot;ident&quot;</span>, <span class="st">&quot;Type&quot;</span>), </span>
<span id="cb84-4"><a href="#cb84-4" tabindex="-1"></a>                   <span class="at">color =</span> <span class="st">&quot;ident&quot;</span>) </span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-69-1.png" width="700" /></p>
<p><code>alluvialClones()</code> provides a visual representation of
clonal distribution and movement across multiple categorical
annotations. It is particularly effective for tracking how specific
clones or clonal groups transition between different states, tissues, or
cell types, offering a dynamic perspective on immune repertoire
evolution and function.</p>
</div>
<div id="getcirclize" class="section level2">
<h2>getCirclize</h2>
<p>Like alluvial graphs, we can also visualize the interconnection of
clusters using chord diagrams from the <a
href="https://jokergoo.github.io/circlize_book/book/">circlize R
package</a>. The first step is to get the data frame output to feed into
the<code>chordDiagram()</code> function in <code>circlize</code>, which
can be done using <code>getCirclize()</code>.</p>
<p>This function calculates the relative number of unique and shared
clones based on the <code>group.by</code> variable using the product of
<code>combineExpression()</code>. It creates a matrix the size of the
<code>group.by</code> variable and then simplifies it into instructions
readable by the <code>circlize</code> R package. The output represents
the total number of unique and shared clones by the group.by variable.
If using the downstream <code>circlize</code> R package, please read and
cite the following <a
href="https://pubmed.ncbi.nlm.nih.gov/24930139/">manuscript</a>.</p>
<p>Key Parameters for getCirclize()</p>
<ul>
<li><code>proportion</code>: If <code>TRUE</code>, normalizes the
relationship by proportion; if <code>FALSE</code> (default), uses unique
clone counts.</li>
<li><code>include.self</code>: If <code>TRUE</code>, includes counting
clones within a single group.by comparison.</li>
</ul>
<p>To get data for a chord diagram showing shared clones between
<code>seurat_clusters</code>:</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" tabindex="-1"></a><span class="fu">library</span>(circlize)</span>
<span id="cb85-2"><a href="#cb85-2" tabindex="-1"></a><span class="fu">library</span>(scales)</span>
<span id="cb85-3"><a href="#cb85-3" tabindex="-1"></a></span>
<span id="cb85-4"><a href="#cb85-4" tabindex="-1"></a>circles <span class="ot">&lt;-</span> <span class="fu">getCirclize</span>(scRep_example, </span>
<span id="cb85-5"><a href="#cb85-5" tabindex="-1"></a>                       <span class="at">group.by =</span> <span class="st">&quot;seurat_clusters&quot;</span>)</span>
<span id="cb85-6"><a href="#cb85-6" tabindex="-1"></a></span>
<span id="cb85-7"><a href="#cb85-7" tabindex="-1"></a><span class="co">#Just assigning the normal colors to each cluster</span></span>
<span id="cb85-8"><a href="#cb85-8" tabindex="-1"></a>grid.cols <span class="ot">&lt;-</span> <span class="fu">hue_pal</span>()(<span class="fu">length</span>(<span class="fu">unique</span>(scRep_example<span class="sc">$</span>seurat_clusters)))</span>
<span id="cb85-9"><a href="#cb85-9" tabindex="-1"></a><span class="fu">names</span>(grid.cols) <span class="ot">&lt;-</span> <span class="fu">unique</span>(scRep_example<span class="sc">$</span>seurat_clusters)</span>
<span id="cb85-10"><a href="#cb85-10" tabindex="-1"></a></span>
<span id="cb85-11"><a href="#cb85-11" tabindex="-1"></a><span class="co">#Graphing the chord diagram</span></span>
<span id="cb85-12"><a href="#cb85-12" tabindex="-1"></a><span class="fu">chordDiagram</span>(circles, <span class="at">self.link =</span> <span class="dv">1</span>, <span class="at">grid.col =</span> grid.cols)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-70-1.png" width="700" /></p>
<p>This can also be used if we want to explore just the lung-specific T
cells by subsetting the single-cell object. For the sake of this
vignette, we can also look at setting <code>proportion = TRUE</code> to
get a scaled output.</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" tabindex="-1"></a>subset <span class="ot">&lt;-</span> <span class="fu">subset</span>(scRep_example, Type <span class="sc">==</span> <span class="st">&quot;L&quot;</span>)</span>
<span id="cb86-2"><a href="#cb86-2" tabindex="-1"></a></span>
<span id="cb86-3"><a href="#cb86-3" tabindex="-1"></a>circles <span class="ot">&lt;-</span> <span class="fu">getCirclize</span>(subset, <span class="at">group.by =</span> <span class="st">&quot;ident&quot;</span>, <span class="at">proportion =</span> <span class="cn">TRUE</span>)</span>
<span id="cb86-4"><a href="#cb86-4" tabindex="-1"></a></span>
<span id="cb86-5"><a href="#cb86-5" tabindex="-1"></a>grid.cols <span class="ot">&lt;-</span> scales<span class="sc">::</span><span class="fu">hue_pal</span>()(<span class="fu">length</span>(<span class="fu">unique</span>(subset<span class="sc">@</span>active.ident)))</span>
<span id="cb86-6"><a href="#cb86-6" tabindex="-1"></a><span class="fu">names</span>(grid.cols) <span class="ot">&lt;-</span> <span class="fu">levels</span>(subset<span class="sc">@</span>active.ident)</span>
<span id="cb86-7"><a href="#cb86-7" tabindex="-1"></a></span>
<span id="cb86-8"><a href="#cb86-8" tabindex="-1"></a><span class="fu">chordDiagram</span>(circles, </span>
<span id="cb86-9"><a href="#cb86-9" tabindex="-1"></a>             <span class="at">self.link =</span> <span class="dv">1</span>, </span>
<span id="cb86-10"><a href="#cb86-10" tabindex="-1"></a>             <span class="at">grid.col =</span> grid.cols, </span>
<span id="cb86-11"><a href="#cb86-11" tabindex="-1"></a>             <span class="at">directional =</span> <span class="dv">1</span>, </span>
<span id="cb86-12"><a href="#cb86-12" tabindex="-1"></a>             <span class="at">direction.type =</span>  <span class="st">&quot;arrows&quot;</span>,</span>
<span id="cb86-13"><a href="#cb86-13" tabindex="-1"></a>             <span class="at">link.arr.type =</span> <span class="st">&quot;big.arrow&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-71-1.png" width="700" /></p>
<p><code>getCirclize()</code> facilitates the creation of visually
striking and informative chord diagrams to represent shared clonal
relationships between distinct groups within your single-cell data. By
providing a flexible way to quantify and format clonal overlap, it
enables researchers to effectively illustrate complex clonal
connectivity patterns, which are crucial for understanding immune
communication and migration.</p>
<hr />
</div>
</div>
<div id="quantifying-clonal-bias" class="section level1">
<h1>Quantifying Clonal Bias</h1>
<div id="startracdiversity" class="section level2">
<h2>StartracDiversity</h2>
<p>From the excellent work by <a
href="https://www.nature.com/articles/s41586-018-0694-x">Zhang et
al. (2018, Nature)</a>, the authors introduced new methods for looking
at clones by cellular origins and cluster identification. We strongly
recommend you read and cite their publication when using this
function.</p>
<p>To use the <code>StartracDiversity()</code> function, you need the
output of the <code>combineExpression()</code> function and a column in
your metadata that specifies the tissue of origin.</p>
<div id="indices-output-from-startracdiversity" class="section level3">
<h3>Indices Output from <code>StartracDiversity()</code></h3>
<ul>
<li><code>expa</code> - Clonal Expansion. Measures the degree of clonal
proliferation within a given cell cluster. It is calculated as
<code>1 - normalized Shannon entropy</code>, where a higher value
indicates that a few clones dominate the cluster.<br />
</li>
<li><code>migr</code> - Cross-tissue Migration. Quantifies the movement
of clonal T cells between different tissues, as defined by the
<code>type</code> parameter. This index is based on the entropy of a
single clonotype’s distribution across the specified tissues.<br />
</li>
<li><code>tran</code> - State Transition. Measures the developmental
transition of T cell clones between different functional clusters. This
index is calculated from the entropy of a single clonotype’s
distribution across the cell clusters identified in the data.</li>
</ul>
<p>Key Parameters for <code>StartracDiversity()</code></p>
<ul>
<li><code>type</code>: The variable in the metadata that provides tissue
type.</li>
<li><code>group.by</code>: A column header in the metadata to group the
analysis by (e.g., “sample”, “treatment”).</li>
</ul>
<p>By default, <code>StartracDiversity()</code> will calculate all three
indices (<code>expa</code>, <code>migr</code>, and <code>tran</code>)
for each cluster and group you define. This provides a comprehensive
overview of the clonal dynamics in your dataset. The output is a
three-paneled plot, with each panel representing one index.</p>
<p>In the example data, <code>type</code> corresponds to the “Type”
column, which includes “P” (peripheral blood) and “L” (lung)
classifiers. The analysis is grouped by the “Patient” column.</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" tabindex="-1"></a><span class="co"># Calculate and plot all three STARTRAC indices</span></span>
<span id="cb87-2"><a href="#cb87-2" tabindex="-1"></a><span class="fu">StartracDiversity</span>(scRep_example, </span>
<span id="cb87-3"><a href="#cb87-3" tabindex="-1"></a>                  <span class="at">type =</span> <span class="st">&quot;Type&quot;</span>, </span>
<span id="cb87-4"><a href="#cb87-4" tabindex="-1"></a>                  <span class="at">group.by =</span> <span class="st">&quot;Patient&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-72-1.png" width="700" /></p>
</div>
<div id="calculating-a-single-index" class="section level3">
<h3>Calculating a Single Index</h3>
<p>If you’re only interested in one aspect of clonal dynamics, you can
specify it using the <code>index</code> parameter. For example, to only
calculate and plot clonal expansion:</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" tabindex="-1"></a><span class="co"># Calculate and plot only the clonal expansion index</span></span>
<span id="cb88-2"><a href="#cb88-2" tabindex="-1"></a><span class="fu">StartracDiversity</span>(scRep_example, </span>
<span id="cb88-3"><a href="#cb88-3" tabindex="-1"></a>                  <span class="at">type =</span> <span class="st">&quot;Type&quot;</span>, </span>
<span id="cb88-4"><a href="#cb88-4" tabindex="-1"></a>                  <span class="at">group.by =</span> <span class="st">&quot;Patient&quot;</span>,</span>
<span id="cb88-5"><a href="#cb88-5" tabindex="-1"></a>                  <span class="at">index =</span> <span class="st">&quot;expa&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-73-1.png" width="700" /></p>
</div>
<div id="pairwise-migration-analysis" class="section level3">
<h3>Pairwise Migration Analysis</h3>
<p>Another feature of <code>StartracDiversity()</code> is the ability to
perform pairwise comparisons. To specifically quantify the migration
between two tissues (e.g., Lung vs. Periphery), set
<code>index = "migr"</code> and tell the function which column to use
for the comparison with <code>pairwise = "Type"</code>.</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" tabindex="-1"></a><span class="co"># # Calculate pairwise migration between tissues</span></span>
<span id="cb89-2"><a href="#cb89-2" tabindex="-1"></a><span class="fu">StartracDiversity</span>(scRep_example, </span>
<span id="cb89-3"><a href="#cb89-3" tabindex="-1"></a>                  <span class="at">type =</span> <span class="st">&quot;Type&quot;</span>, </span>
<span id="cb89-4"><a href="#cb89-4" tabindex="-1"></a>                  <span class="at">group.by =</span> <span class="st">&quot;Patient&quot;</span>,</span>
<span id="cb89-5"><a href="#cb89-5" tabindex="-1"></a>                  <span class="at">index =</span> <span class="st">&quot;migr&quot;</span>,</span>
<span id="cb89-6"><a href="#cb89-6" tabindex="-1"></a>                  <span class="at">pairwise =</span> <span class="st">&quot;Type&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-74-1.png" width="700" /></p>
</div>
</div>
<div id="clonalbias" class="section level2">
<h2>clonalBias</h2>
<p>A new metric proposed by <a
href="https://pubmed.ncbi.nlm.nih.gov/35829695/">Massimo et al</a>,
<code>clonalBias()</code>, like STARTRAC, is a clonal metric that seeks
to quantify how individual clones are skewed towards a specific cellular
compartment or cluster. A clone bias of <code>1</code> indicates that a
clone is composed of cells from a single compartment or cluster, while a
clone bias of <code>0</code> matches the background subtype
distribution. Please read and cite the linked manuscript if using
<code>clonalBias()</code></p>
<p>Key Parameter(s) for <code>clonalBias()</code></p>
<ul>
<li><code>group.by</code>: A column header in the metadata that bias
will be based on.</li>
<li><code>split.by</code>: The variable to use for calculating the
baseline frequencies (e.g., “Type” for lung vs peripheral blood
comparison)</li>
<li><code>n.boots</code>: Number of bootstraps to downsample.</li>
<li><code>min.expand</code>: Clone frequency cut-off for the purpose of
comparison (default = 10).</li>
</ul>
<p>Here we calculate and plot clonal bias using <code>aa</code> clone
calls, splitting by “Patient” and grouping by “seurat_clusters”, with a
minimum expansion of 5 and 10 bootstraps:</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" tabindex="-1"></a><span class="fu">clonalBias</span>(scRep_example, </span>
<span id="cb90-2"><a href="#cb90-2" tabindex="-1"></a>           <span class="at">cloneCall =</span> <span class="st">&quot;aa&quot;</span>, </span>
<span id="cb90-3"><a href="#cb90-3" tabindex="-1"></a>           <span class="at">split.by =</span> <span class="st">&quot;Patient&quot;</span>, </span>
<span id="cb90-4"><a href="#cb90-4" tabindex="-1"></a>           <span class="at">group.by =</span> <span class="st">&quot;seurat_clusters&quot;</span>,</span>
<span id="cb90-5"><a href="#cb90-5" tabindex="-1"></a>           <span class="at">n.boots =</span> <span class="dv">10</span>, </span>
<span id="cb90-6"><a href="#cb90-6" tabindex="-1"></a>           <span class="at">min.expand =</span><span class="dv">5</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-75-1.png" width="700" /></p>
<hr />
</div>
</div>
<div id="clustering-by-edit-distance" class="section level1">
<h1>Clustering by Edit Distance</h1>
<div id="clonalcluster-cluster-by-sequence-similarity"
class="section level2">
<h2>clonalCluster: Cluster by Sequence Similarity</h2>
<p>The <code>clonalCluster()</code> function provides a powerful method
to group clonotypes based on sequence similarity. It calculates the edit
distance between CDR3 sequences and uses this information to build a
network, identifying closely related clusters of T or B cell receptors.
This functionality allows for a more nuanced definition of a “clone”
that extends beyond identical sequence matches.</p>
<div id="core-concepts" class="section level3">
<h3>Core Concepts</h3>
<p>The clustering process follows these key steps:</p>
<ul>
<li><strong>Sequence Selection</strong>: The function selects either the
nucleotide (<code>sequence = "nt"</code>) or amino acid
(<code>sequence = "aa"</code>) CDR3 sequences for a specified
chain.</li>
<li><strong>Distance Calculation</strong>: It calculates the edit
distance between all pairs of sequences. By default, it also requires
sequences to share the same V gene (<code>use.v = TRUE</code>).</li>
<li><strong>Network Construction</strong>: An edge is created between
any two sequences that meet the similarity threshold, forming a network
graph.</li>
<li><strong>Clustering</strong>: A graph-based clustering algorithm is
run to identify connected components or communities within the network.
By default, it identifies all directly or indirectly connected sequences
as a single cluster (<code>cluster.method = "components"</code>).</li>
<li><strong>Output</strong>: The function can either add the resulting
cluster IDs to the input object, return an <code>igraph</code> object
for network analysis, or export a sparse adjacency matrix.</li>
</ul>
</div>
<div id="understanding-the-threshold-parameter" class="section level3">
<h3>Understanding the <code>threshold</code> Parameter</h3>
<p>The behavior of the threshold parameter is critical for controlling
cluster granularity:</p>
<ul>
<li><strong>Normalized Similarity (threshold &lt; 1)</strong>: When the
threshold is a value between 0 and 1 (e.g., <code>0.85</code>), it
represents the normalized edit distance (Levenshtein distance / mean
sequence length). A higher value corresponds to a stricter similarity
requirement. This is useful for comparing sequences of varying
lengths.</li>
<li><strong>Raw Edit Distance (threshold &gt;= 1)</strong>: When the
threshold is a whole number (e.g., 2), it represents the maximum raw
edit distance allowed. A lower value corresponds to a stricter
similarity requirement. This is useful when you want to allow a specific
number of mutations.</li>
</ul>
<p>Key Parameter(s) for <code>clonalCluster()</code></p>
<ul>
<li><code>sequence</code>: Specifies whether to cluster based on
<code>aa</code> (amino acid) or <code>nt</code> (nucleotide)
sequences.</li>
<li><code>threshold</code>: The similarity threshold for clustering.
Values <code>&lt; 1</code> are normalized similarity, while values
<code>&gt;= 1</code> are raw edit distance.</li>
<li><code>group.by</code>: A column header in the metadata or lists to
group the analysis by (e.g., “sample”, “treatment”). If
<code>NULL</code>, clusters are calculated across all sequences.</li>
<li><code>use.V</code>: If <code>TRUE</code>, sequences must share the
same V gene to be clustered together.</li>
<li><code>use.J</code>: If <code>TRUE</code>, sequences must share the
same J gene to be clustered together.</li>
<li><code>cluster.method</code>: The clustering algorithm to use.
Defaults to <code>components</code>, which finds connected
subgraphs.</li>
<li><code>cluster.prefix</code>: A character prefix to add to the
cluster names (e.g., “cluster.”).</li>
<li><code>exportGraph</code>: If <code>TRUE</code>, returns an igraph
object of the sequence network.</li>
<li><code>exportAdjMatrix</code>: If <code>TRUE</code>, returns a sparse
adjacency matrix (dgCMatrix) of the network.</li>
</ul>
</div>
<div id="demonstrating-basic-use" class="section level3">
<h3>Demonstrating Basic Use</h3>
<p>To run clustering on the first two samples for the TRA chain, using
amino acid sequences with a normalized similarity threshold of 0.85:</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" tabindex="-1"></a><span class="co"># Run clustering on the first two samples for the TRA chain</span></span>
<span id="cb91-2"><a href="#cb91-2" tabindex="-1"></a>sub_combined <span class="ot">&lt;-</span> <span class="fu">clonalCluster</span>(combined.TCR[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)], </span>
<span id="cb91-3"><a href="#cb91-3" tabindex="-1"></a>                              <span class="at">chain =</span> <span class="st">&quot;TRA&quot;</span>, </span>
<span id="cb91-4"><a href="#cb91-4" tabindex="-1"></a>                              <span class="at">sequence =</span> <span class="st">&quot;aa&quot;</span>, </span>
<span id="cb91-5"><a href="#cb91-5" tabindex="-1"></a>                              <span class="at">threshold =</span> <span class="fl">0.85</span>)</span>
<span id="cb91-6"><a href="#cb91-6" tabindex="-1"></a></span>
<span id="cb91-7"><a href="#cb91-7" tabindex="-1"></a><span class="co"># View the new cluster column</span></span>
<span id="cb91-8"><a href="#cb91-8" tabindex="-1"></a><span class="fu">head</span>(sub_combined[[<span class="dv">1</span>]][, <span class="fu">c</span>(<span class="st">&quot;barcode&quot;</span>, <span class="st">&quot;TCR1&quot;</span>, <span class="st">&quot;TRA.Cluster&quot;</span>)])</span></code></pre></div>
<pre><code>##                   barcode                     TCR1 TRA.Cluster
## 1 P17B_AAACCTGAGTACGACG-1       TRAV25.TRAJ20.TRAC        &lt;NA&gt;
## 2 P17B_AAACCTGCAACACGCC-1 TRAV38-2/DV8.TRAJ52.TRAC        &lt;NA&gt;
## 3 P17B_AAACCTGCAGGCGATA-1      TRAV12-1.TRAJ9.TRAC  cluster.24
## 4 P17B_AAACCTGCATGAGCGA-1      TRAV12-1.TRAJ9.TRAC  cluster.24
## 5 P17B_AAACGGGAGAGCCCAA-1        TRAV20.TRAJ8.TRAC cluster.148
## 6 P17B_AAACGGGAGCGTTTAC-1      TRAV12-1.TRAJ9.TRAC  cluster.24</code></pre>
</div>
<div id="demonstrating-clustering-with-a-single-cell-object"
class="section level3">
<h3>Demonstrating Clustering with a Single-Cell Object</h3>
<p>You can calculate clusters based on specific metadata variables
within a single-cell object by using the <code>group.by</code>
parameter. This is useful for analyzing clusters on a per-sample or
per-patient basis without subsetting the data first.</p>
<p>First, add patient and type information to the
<code>scRep_example</code> Seurat object:</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" tabindex="-1"></a><span class="co">#Adding patient information</span></span>
<span id="cb93-2"><a href="#cb93-2" tabindex="-1"></a>scRep_example<span class="sc">$</span>Patient <span class="ot">&lt;-</span> <span class="fu">substr</span>(scRep_example<span class="sc">$</span>orig.ident, <span class="dv">1</span>,<span class="dv">3</span>)</span>
<span id="cb93-3"><a href="#cb93-3" tabindex="-1"></a></span>
<span id="cb93-4"><a href="#cb93-4" tabindex="-1"></a><span class="co">#Adding type information</span></span>
<span id="cb93-5"><a href="#cb93-5" tabindex="-1"></a>scRep_example<span class="sc">$</span>Type <span class="ot">&lt;-</span> <span class="fu">substr</span>(scRep_example<span class="sc">$</span>orig.ident, <span class="dv">4</span>,<span class="dv">4</span>)</span></code></pre></div>
<p>Now, run clustering on the <code>scRep_example</code> Seurat object,
grouping calculations by “Patient”:</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" tabindex="-1"></a><span class="co"># Run clustering, but group calculations by &quot;Patient&quot;</span></span>
<span id="cb94-2"><a href="#cb94-2" tabindex="-1"></a>scRep_example <span class="ot">&lt;-</span> <span class="fu">clonalCluster</span>(scRep_example, </span>
<span id="cb94-3"><a href="#cb94-3" tabindex="-1"></a>                               <span class="at">chain =</span> <span class="st">&quot;TRA&quot;</span>, </span>
<span id="cb94-4"><a href="#cb94-4" tabindex="-1"></a>                               <span class="at">sequence =</span> <span class="st">&quot;aa&quot;</span>, </span>
<span id="cb94-5"><a href="#cb94-5" tabindex="-1"></a>                               <span class="at">threshold =</span> <span class="fl">0.85</span>, </span>
<span id="cb94-6"><a href="#cb94-6" tabindex="-1"></a>                               <span class="at">group.by =</span> <span class="st">&quot;Patient&quot;</span>)</span>
<span id="cb94-7"><a href="#cb94-7" tabindex="-1"></a></span>
<span id="cb94-8"><a href="#cb94-8" tabindex="-1"></a><span class="co">#Define color palette </span></span>
<span id="cb94-9"><a href="#cb94-9" tabindex="-1"></a>num_clusters <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">unique</span>(<span class="fu">na.omit</span>(scRep_example<span class="sc">$</span>TRA.Cluster)))</span>
<span id="cb94-10"><a href="#cb94-10" tabindex="-1"></a>cluster_colors <span class="ot">&lt;-</span> <span class="fu">hcl.colors</span>(<span class="at">n =</span> num_clusters, <span class="at">palette =</span> <span class="st">&quot;inferno&quot;</span>)</span>
<span id="cb94-11"><a href="#cb94-11" tabindex="-1"></a></span>
<span id="cb94-12"><a href="#cb94-12" tabindex="-1"></a><span class="fu">DimPlot</span>(scRep_example, <span class="at">group.by =</span> <span class="st">&quot;TRA.Cluster&quot;</span>) <span class="sc">+</span></span>
<span id="cb94-13"><a href="#cb94-13" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> cluster_colors, <span class="at">na.value =</span> <span class="st">&quot;grey&quot;</span>) <span class="sc">+</span> </span>
<span id="cb94-14"><a href="#cb94-14" tabindex="-1"></a>  <span class="fu">NoLegend</span>()</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-78-1.png" width="700" /></p>
</div>
<div id="returning-an-igraph-object" class="section level3">
<h3>Returning an igraph Object:</h3>
<p>Instead of modifying the input object, <code>clonalCluster()</code>
can export the underlying network structure for advanced analysis. Set
<code>exportGraph = TRUE</code> to get an igraph object consisting of
the networks of barcodes by the indicated clustering scheme.</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb95-2"><a href="#cb95-2" tabindex="-1"></a><span class="co">#Clustering Patient 19 samples</span></span>
<span id="cb95-3"><a href="#cb95-3" tabindex="-1"></a>igraph.object <span class="ot">&lt;-</span> <span class="fu">clonalCluster</span>(combined.TCR[<span class="fu">c</span>(<span class="dv">5</span>,<span class="dv">6</span>)],</span>
<span id="cb95-4"><a href="#cb95-4" tabindex="-1"></a>                               <span class="at">chain =</span> <span class="st">&quot;TRB&quot;</span>,</span>
<span id="cb95-5"><a href="#cb95-5" tabindex="-1"></a>                               <span class="at">sequence =</span> <span class="st">&quot;aa&quot;</span>,</span>
<span id="cb95-6"><a href="#cb95-6" tabindex="-1"></a>                               <span class="at">group.by =</span> <span class="st">&quot;sample&quot;</span>,</span>
<span id="cb95-7"><a href="#cb95-7" tabindex="-1"></a>                               <span class="at">threshold =</span> <span class="fl">0.85</span>, </span>
<span id="cb95-8"><a href="#cb95-8" tabindex="-1"></a>                               <span class="at">exportGraph =</span> <span class="cn">TRUE</span>)</span>
<span id="cb95-9"><a href="#cb95-9" tabindex="-1"></a></span>
<span id="cb95-10"><a href="#cb95-10" tabindex="-1"></a><span class="co">#Setting color scheme</span></span>
<span id="cb95-11"><a href="#cb95-11" tabindex="-1"></a>col_legend <span class="ot">&lt;-</span> <span class="fu">factor</span>(igraph<span class="sc">::</span><span class="fu">V</span>(igraph.object)<span class="sc">$</span>group)</span>
<span id="cb95-12"><a href="#cb95-12" tabindex="-1"></a>col_samples <span class="ot">&lt;-</span> <span class="fu">hcl.colors</span>(<span class="dv">2</span>,<span class="st">&quot;inferno&quot;</span>)[<span class="fu">as.numeric</span>(col_legend)]</span>
<span id="cb95-13"><a href="#cb95-13" tabindex="-1"></a>color.legend <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">unique</span>(igraph<span class="sc">::</span><span class="fu">V</span>(igraph.object)<span class="sc">$</span>group))</span>
<span id="cb95-14"><a href="#cb95-14" tabindex="-1"></a>sample.vertices <span class="ot">&lt;-</span> <span class="fu">V</span>(igraph.object)[<span class="fu">sample</span>(<span class="fu">length</span>(igraph.object), <span class="dv">500</span>)]</span>
<span id="cb95-15"><a href="#cb95-15" tabindex="-1"></a></span>
<span id="cb95-16"><a href="#cb95-16" tabindex="-1"></a>subgraph.object <span class="ot">&lt;-</span> <span class="fu">induced_subgraph</span>(igraph.object, <span class="at">vids =</span> sample.vertices)</span>
<span id="cb95-17"><a href="#cb95-17" tabindex="-1"></a><span class="fu">V</span>(subgraph.object)<span class="sc">$</span>degrees <span class="ot">&lt;-</span> igraph<span class="sc">::</span><span class="fu">degree</span>(subgraph.object)</span>
<span id="cb95-18"><a href="#cb95-18" tabindex="-1"></a>edge_alpha_color <span class="ot">&lt;-</span> <span class="fu">adjustcolor</span>(<span class="st">&quot;gray&quot;</span>, <span class="at">alpha.f =</span> <span class="fl">0.3</span>)</span>
<span id="cb95-19"><a href="#cb95-19" tabindex="-1"></a></span>
<span id="cb95-20"><a href="#cb95-20" tabindex="-1"></a><span class="co">#Plotting</span></span>
<span id="cb95-21"><a href="#cb95-21" tabindex="-1"></a><span class="fu">plot</span>(subgraph.object,</span>
<span id="cb95-22"><a href="#cb95-22" tabindex="-1"></a>     <span class="at">layout =</span> <span class="fu">layout_nicely</span>(subgraph.object),</span>
<span id="cb95-23"><a href="#cb95-23" tabindex="-1"></a>     <span class="at">vertex.label =</span> <span class="cn">NA</span>,</span>
<span id="cb95-24"><a href="#cb95-24" tabindex="-1"></a>     <span class="at">vertex.size =</span> <span class="fu">sqrt</span>(igraph<span class="sc">::</span><span class="fu">V</span>(subgraph.object)<span class="sc">$</span>degrees), </span>
<span id="cb95-25"><a href="#cb95-25" tabindex="-1"></a>     <span class="at">vertex.color =</span> col_samples[sample.vertices],</span>
<span id="cb95-26"><a href="#cb95-26" tabindex="-1"></a>     <span class="at">vertex.frame.color =</span> <span class="st">&quot;white&quot;</span>, </span>
<span id="cb95-27"><a href="#cb95-27" tabindex="-1"></a>     <span class="at">edge.color =</span> edge_alpha_color,</span>
<span id="cb95-28"><a href="#cb95-28" tabindex="-1"></a>     <span class="at">edge.arrow.size =</span> <span class="fl">0.05</span>,</span>
<span id="cb95-29"><a href="#cb95-29" tabindex="-1"></a>     <span class="at">edge.curved =</span> <span class="fl">0.05</span>, </span>
<span id="cb95-30"><a href="#cb95-30" tabindex="-1"></a>     <span class="at">margin =</span> <span class="sc">-</span><span class="fl">0.1</span>)</span>
<span id="cb95-31"><a href="#cb95-31" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;topleft&quot;</span>, </span>
<span id="cb95-32"><a href="#cb95-32" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">levels</span>(color.legend), </span>
<span id="cb95-33"><a href="#cb95-33" tabindex="-1"></a>       <span class="at">pch =</span> <span class="dv">16</span>, </span>
<span id="cb95-34"><a href="#cb95-34" tabindex="-1"></a>       <span class="at">col =</span> <span class="fu">unique</span>(col_samples), </span>
<span id="cb95-35"><a href="#cb95-35" tabindex="-1"></a>       <span class="at">bty =</span> <span class="st">&quot;n&quot;</span>)</span></code></pre></div>
<p><img src="/Users/borcherding.n/Documents/GitHub/scRepertoire/docs/articles/vignette_files/figure-html/unnamed-chunk-79-1.png" width="700" /></p>
</div>
<div id="returning-a-sparse-adjacency-matrix" class="section level3">
<h3>Returning a Sparse Adjacency Matrix</h3>
<p>For computational applications, you can export a sparse adjacency
matrix using <code>exportAdjMatrix = TRUE</code>. This matrix represents
the connections between all barcodes in the input data, with the edit
distance that meet the threshold in places of connection.</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" tabindex="-1"></a><span class="co"># Generate the sparse matrix</span></span>
<span id="cb96-2"><a href="#cb96-2" tabindex="-1"></a>adj.matrix <span class="ot">&lt;-</span> <span class="fu">clonalCluster</span>(combined.TCR[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)],</span>
<span id="cb96-3"><a href="#cb96-3" tabindex="-1"></a>                            <span class="at">chain =</span> <span class="st">&quot;TRB&quot;</span>,</span>
<span id="cb96-4"><a href="#cb96-4" tabindex="-1"></a>                            <span class="at">exportAdjMatrix =</span> <span class="cn">TRUE</span>)</span>
<span id="cb96-5"><a href="#cb96-5" tabindex="-1"></a></span>
<span id="cb96-6"><a href="#cb96-6" tabindex="-1"></a><span class="co"># View the dimensions and a snippet of the matrix</span></span>
<span id="cb96-7"><a href="#cb96-7" tabindex="-1"></a><span class="fu">dim</span>(adj.matrix)</span></code></pre></div>
<pre><code>## [1] 5698 5698</code></pre>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" tabindex="-1"></a><span class="fu">print</span>(adj.matrix[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span></code></pre></div>
<pre><code>## 10 x 10 sparse Matrix of class &quot;dgCMatrix&quot;</code></pre>
<pre><code>##                                                                          
## P17B_AAACCTGAGTACGACG-1 . . .      .      . .      .      .      . .     
## P17B_AAACCTGCAACACGCC-1 . . .      .      . .      .      .      . .     
## P17B_AAACCTGCAGGCGATA-1 . . .       1e-06 .  1e-06  1e-06  1e-06 .  1e-06
## P17B_AAACCTGCATGAGCGA-1 . .  1e-06 .      .  1e-06  1e-06  1e-06 .  1e-06
## P17B_AAACGGGAGAGCCCAA-1 . . .      .      . .      .      .      . .     
## P17B_AAACGGGAGCGTTTAC-1 . .  1e-06  1e-06 . .       1e-06  1e-06 .  1e-06
## P17B_AAACGGGAGGGCACTA-1 . .  1e-06  1e-06 .  1e-06 .       1e-06 .  1e-06
## P17B_AAACGGGAGTGGTCCC-1 . .  1e-06  1e-06 .  1e-06  1e-06 .      .  1e-06
## P17B_AAACGGGCAGTTAACC-1 . . .      .      . .      .      .      . .     
## P17B_AAACGGGGTCGCCATG-1 . .  1e-06  1e-06 .  1e-06  1e-06  1e-06 . .</code></pre>
</div>
<div id="using-both-chains" class="section level3">
<h3>Using Both Chains</h3>
<p>You can analyze the combined network of both TRA/TRB or IGH/IGL
chains by setting <code>chain = "both"</code>. This will create a single
cluster column named <code>Multi.Cluster</code>.</p>
<div class="sourceCode" id="cb101"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" tabindex="-1"></a><span class="co"># Cluster using both TRB and TRA chains simultaneously</span></span>
<span id="cb101-2"><a href="#cb101-2" tabindex="-1"></a>clustered_both <span class="ot">&lt;-</span> <span class="fu">clonalCluster</span>(combined.TCR[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)], </span>
<span id="cb101-3"><a href="#cb101-3" tabindex="-1"></a>                                <span class="at">chain =</span> <span class="st">&quot;both&quot;</span>)</span>
<span id="cb101-4"><a href="#cb101-4" tabindex="-1"></a></span>
<span id="cb101-5"><a href="#cb101-5" tabindex="-1"></a><span class="co"># View the new &quot;Multi.Cluster&quot; column</span></span>
<span id="cb101-6"><a href="#cb101-6" tabindex="-1"></a><span class="fu">head</span>(clustered_both[[<span class="dv">1</span>]][, <span class="fu">c</span>(<span class="st">&quot;barcode&quot;</span>, <span class="st">&quot;TCR1&quot;</span>, <span class="st">&quot;TCR2&quot;</span>, <span class="st">&quot;Multi.Cluster&quot;</span>)])</span></code></pre></div>
<pre><code>##                   barcode                     TCR1                        TCR2
## 1 P17B_AAACCTGAGTACGACG-1       TRAV25.TRAJ20.TRAC  TRBV5-1.None.TRBJ2-7.TRBC2
## 2 P17B_AAACCTGCAACACGCC-1 TRAV38-2/DV8.TRAJ52.TRAC TRBV10-3.None.TRBJ2-2.TRBC2
## 3 P17B_AAACCTGCAGGCGATA-1      TRAV12-1.TRAJ9.TRAC    TRBV9.None.TRBJ2-2.TRBC2
## 4 P17B_AAACCTGCATGAGCGA-1      TRAV12-1.TRAJ9.TRAC    TRBV9.None.TRBJ2-2.TRBC2
## 5 P17B_AAACGGGAGAGCCCAA-1        TRAV20.TRAJ8.TRAC                        &lt;NA&gt;
## 6 P17B_AAACGGGAGCGTTTAC-1      TRAV12-1.TRAJ9.TRAC    TRBV9.None.TRBJ2-2.TRBC2
##   Multi.Cluster
## 1          &lt;NA&gt;
## 2          &lt;NA&gt;
## 3    cluster.23
## 4    cluster.23
## 5   cluster.145
## 6    cluster.23</code></pre>
</div>
<div id="using-different-clustering-algorithms" class="section level3">
<h3>Using Different Clustering Algorithms</h3>
<p>While the default <code>cluster.method = "components"</code> is
robust, you can use other algorithms from the igraph package, such as
<code>walktrap</code> or <code>louvain</code>, to potentially uncover
different community structures.</p>
<div class="sourceCode" id="cb103"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" tabindex="-1"></a><span class="co"># Cluster using the walktrap algorithm</span></span>
<span id="cb103-2"><a href="#cb103-2" tabindex="-1"></a>graph_walktrap <span class="ot">&lt;-</span> <span class="fu">clonalCluster</span>(combined.TCR[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)],</span>
<span id="cb103-3"><a href="#cb103-3" tabindex="-1"></a>                                <span class="at">cluster.method =</span> <span class="st">&quot;walktrap&quot;</span>,</span>
<span id="cb103-4"><a href="#cb103-4" tabindex="-1"></a>                                <span class="at">exportGraph =</span> <span class="cn">TRUE</span>)</span>
<span id="cb103-5"><a href="#cb103-5" tabindex="-1"></a></span>
<span id="cb103-6"><a href="#cb103-6" tabindex="-1"></a><span class="co"># Compare the number of clusters found</span></span>
<span id="cb103-7"><a href="#cb103-7" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(<span class="fu">V</span>(graph_walktrap)<span class="sc">$</span>cluster))</span></code></pre></div>
<pre><code>## [1] 391</code></pre>
<p>Overall, <code>clonalCluster()</code> is a versatile function for
defining and analyzing clonal relationships based on sequence
similarity. It allows researchers to move beyond exact sequence matches,
providing a more comprehensive understanding of clonal families. The
ability to customize parameters like <code>threshold</code>,
<code>chain</code> selection, and <code>group.by</code> ensures
adaptability to diverse research questions. Furthermore, the option to
export <code>igraph</code> objects or sparse adjacency matrices provides
advanced users with the tools for in-depth network analysis.</p>
<hr />
</div>
</div>
</div>
<div id="conclusion" class="section level1">
<h1>Conclusion</h1>
<p>This has been a general overview of the capabilities of scRepertoire
from the initial processing and visualization to attach to the mRNA
expression values in a single-cell object. If you have any questions,
comments, or suggestions, please visit the <a
href="https://github.com/BorchLab/scRepertoire">GitHub
repository</a>.</p>
<div id="session-info" class="section level3">
<h3>Session Info</h3>
<div class="sourceCode" id="cb105"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.5.1 (2025-06-13)
## Platform: aarch64-apple-darwin20
## Running under: macOS Sonoma 14.6
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.0.dylib 
## LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## time zone: America/Chicago
## tzcode source: internal
## 
## attached base packages:
## [1] stats4    stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] scales_1.4.0                circlize_0.4.16            
##  [3] ggraph_2.2.2                scRepertoire_2.6.0         
##  [5] igraph_2.2.0                Seurat_5.3.1.1000          
##  [7] SeuratObject_5.2.0          sp_2.2-0                   
##  [9] scater_1.35.0               ggplot2_4.0.0              
## [11] scuttle_1.18.0              SingleCellExperiment_1.30.1
## [13] SummarizedExperiment_1.38.1 Biobase_2.68.0             
## [15] GenomicRanges_1.60.0        GenomeInfoDb_1.44.3        
## [17] IRanges_2.42.0              S4Vectors_0.48.0           
## [19] BiocGenerics_0.54.1         generics_0.1.4             
## [21] MatrixGenerics_1.20.0       matrixStats_1.5.0          
## [23] BiocStyle_2.36.0           
## 
## loaded via a namespace (and not attached):
##   [1] RcppAnnoy_0.0.22        splines_4.5.1           later_1.4.4            
##   [4] tibble_3.3.0            polyclip_1.10-7         fastDummies_1.7.5      
##   [7] lifecycle_1.0.4         globals_0.18.0          lattice_0.22-7         
##  [10] MASS_7.3-65             magrittr_2.0.4          plotly_4.11.0          
##  [13] sass_0.4.10             rmarkdown_2.30          jquerylib_0.1.4        
##  [16] yaml_2.3.10             httpuv_1.6.16           sctransform_0.4.2      
##  [19] spam_2.11-1             spatstat.sparse_3.1-0   reticulate_1.43.0      
##  [22] cowplot_1.2.0           pbapply_1.7-4           RColorBrewer_1.1-3     
##  [25] abind_1.4-8             rvest_1.0.5             Rtsne_0.17             
##  [28] purrr_1.1.0             hash_2.2.6.3            tweenr_2.0.3           
##  [31] evmix_2.12              GenomeInfoDbData_1.2.14 ggrepel_0.9.6          
##  [34] irlba_2.3.5.1           listenv_0.9.1           spatstat.utils_3.2-0   
##  [37] iNEXT_3.0.2             MatrixModels_0.5-4      goftest_1.2-3          
##  [40] RSpectra_0.16-2         spatstat.random_3.4-2   fitdistrplus_1.2-4     
##  [43] parallelly_1.45.1       pkgdown_2.1.3           codetools_0.2-20       
##  [46] DelayedArray_0.34.1     xml2_1.4.0              ggforce_0.5.0          
##  [49] shape_1.4.6.1           tidyselect_1.2.1        UCSC.utils_1.4.0       
##  [52] farver_2.1.2            ScaledMatrix_1.16.0     viridis_0.6.5          
##  [55] spatstat.explore_3.5-3  jsonlite_2.0.0          BiocNeighbors_2.2.0    
##  [58] tidygraph_1.3.1         progressr_0.16.0        ggalluvial_0.12.5      
##  [61] ggridges_0.5.7          survival_3.8-3          systemfonts_1.3.1      
##  [64] tools_4.5.1             ragg_1.5.0              ica_1.0-3              
##  [67] Rcpp_1.1.0              glue_1.8.0              gridExtra_2.3          
##  [70] SparseArray_1.8.1       xfun_0.53               dplyr_1.1.4            
##  [73] withr_3.0.2             BiocManager_1.30.26     fastmap_1.2.0          
##  [76] SparseM_1.84-2          digest_0.6.37           rsvd_1.0.5             
##  [79] R6_2.6.1                mime_0.13               colorspace_2.1-2       
##  [82] textshaping_1.0.4       scattermore_1.2         tensor_1.5.1           
##  [85] spatstat.data_3.1-8     utf8_1.2.6              tidyr_1.3.1            
##  [88] data.table_1.17.8       graphlayouts_1.2.2      httr_1.4.7             
##  [91] htmlwidgets_1.6.4       S4Arrays_1.8.1          uwot_0.2.3             
##  [94] pkgconfig_2.0.3         gtable_0.3.6            lmtest_0.9-40          
##  [97] S7_0.2.0                XVector_0.48.0          htmltools_0.5.8.1      
## [100] dotCall64_1.2           bookdown_0.45           png_0.1-8              
## [103] spatstat.univar_3.1-4   ggdendro_0.2.0          knitr_1.50             
## [106] rstudioapi_0.17.1       rjson_0.2.23            reshape2_1.4.4         
## [109] nlme_3.1-168            GlobalOptions_0.1.2     cachem_1.1.0           
## [112] zoo_1.8-14              stringr_1.5.2           KernSmooth_2.23-26     
## [115] parallel_4.5.1          miniUI_0.1.2            vipor_0.4.7            
## [118] desc_1.4.3              pillar_1.11.1           grid_4.5.1             
## [121] vctrs_0.6.5             RANN_2.6.2              promises_1.3.3         
## [124] BiocSingular_1.24.0     beachmat_2.24.0         xtable_1.8-4           
## [127] cluster_2.1.8.1         beeswarm_0.4.0          evaluate_1.0.5         
## [130] isoband_0.2.7           cli_3.6.5               compiler_4.5.1         
## [133] rlang_1.1.6             crayon_1.5.3            future.apply_1.20.0    
## [136] labeling_0.4.3          immApex_1.4.0           plyr_1.8.9             
## [139] fs_1.6.6                ggbeeswarm_0.7.2        stringi_1.8.7          
## [142] viridisLite_0.4.2       deldir_2.0-4            BiocParallel_1.42.2    
## [145] gsl_2.1-8               lazyeval_0.2.2          spatstat.geom_3.6-0    
## [148] quantreg_6.1            Matrix_1.7-4            RcppHNSW_0.6.0         
## [151] patchwork_1.3.2         future_1.67.0           shiny_1.11.1           
## [154] ROCR_1.0-11             memoise_2.0.1           bslib_0.9.0</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p><p>Developed by Nick Borcherding, Qile Yang, Ksenia Safina.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>
