<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Getting Started with deepMatchR • deepMatchR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting Started with deepMatchR">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">deepMatchR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/deepMatchR.html">Get Started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/index.html">Overview</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Core Workflows</h6></li>
    <li><a class="dropdown-item" href="../articles/sequence-analysis.html">Sequence &amp; Mismatch Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/eplet-analysis.html">Eplet Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/peptide-binding.html">Peptide Binding Prediction</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Visualization</h6></li>
    <li><a class="dropdown-item" href="../articles/antibody-visualization.html">Antibody Visualization</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/BorchLab/deepMatchR" aria-label="GitHub"><span class="fa fab fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Getting Started with deepMatchR</h1>
                        <h4 data-toc-skip class="author">Nick
Borcherding</h4>
            <address class="author_afil">
      Washington University in St. Louis, School of Medicine, St. Louis,
MO, USA<br><h4 data-toc-skip class="date">Compiled: January 25,
2026</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BorchLab/deepMatchR/blob/main/vignettes/deepMatchR.Rmd" class="external-link"><code>vignettes/deepMatchR.Rmd</code></a></small>
      <div class="d-none name"><code>deepMatchR.Rmd</code></div>
    </address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Human leukocyte antigen (HLA) molecules are central to both innate
and adaptive immune responses. Characterizing HLA antibodies and typing
plays a pivotal role in bone marrow and solid organ transplantation.</p>
<p><strong>deepMatchR</strong> provides a collection of tools that use
advanced statistical approaches and deep learning to better characterize
HLA immunogenicity for clinical and research applications.</p>
<p><strong>Key capabilities include:</strong></p>
<ul>
<li>Sequence-level mismatch quantification (amino acid and eplet)</li>
<li>Peptide-MHC binding prediction</li>
<li>Visualization of antibody reactivity data</li>
<li>Integration with standard HLA nomenclature</li>
</ul>
<p>For more information, visit the <a href="https://github.com/BorchLab/deepMatchR" class="external-link">deepMatchR GitHub
repository</a>.</p>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install from Bioconductor (when available)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"deepMatchR"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Or install the development version from GitHub</span></span>
<span><span class="co"># remotes::install_github("BorchLab/deepMatchR")</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="quick-start">Quick Start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BorchLab/deepMatchR" class="external-link">deepMatchR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="working-with-hla-data">Working with HLA Data<a class="anchor" aria-label="anchor" href="#working-with-hla-data"></a>
</h2>
<div class="section level3">
<h3 id="built-in-example-data">Built-in Example Data<a class="anchor" aria-label="anchor" href="#built-in-example-data"></a>
</h3>
<p><strong>deepMatchR</strong> includes simulated assay data for
learning and testing:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Class I SAB</strong> - Single Antigen Bead assay
results</li>
<li>
<strong>Class II SAB</strong> - Single Antigen Bead assay
results</li>
<li>
<strong>PRA</strong> - Panel Reactive Antibody assay results</li>
</ol>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"deepMatchR_example"</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="data-format-requirements">Data Format Requirements<a class="anchor" aria-label="anchor" href="#data-format-requirements"></a>
</h4>
<p>For downstream analysis, your data must include these columns:</p>
<table class="table">
<thead><tr class="header">
<th>Column</th>
<th>Description</th>
<th>Example</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>BeadID</code></td>
<td>Unique bead identifier</td>
<td>1, 2, 3…</td>
</tr>
<tr class="even">
<td><code>SpecAbbr</code></td>
<td>Abbreviated specificity</td>
<td>A01, B35</td>
</tr>
<tr class="odd">
<td><code>Specificity</code></td>
<td>Full specificity string</td>
<td>“A<em>01:01, B</em>07:01”</td>
</tr>
<tr class="even">
<td><code>NormalValue</code></td>
<td>Measured MFI value</td>
<td>15234</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Class I SAB example</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">deepMatchR_example</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt;   BeadID         SpecAbbr       Specificity NormalValue  RawData CountValue</span></span>
<span><span class="co">#&gt; 1      1  -,-,-,-,-,-,-,-       -,-,-,-,-,-          NA    48.04         77</span></span>
<span><span class="co">#&gt; 2      2  -,-,-,-,-,-,-,-       -,-,-,-,-,-          NA 14068.42         78</span></span>
<span><span class="co">#&gt; 3      3 A1,-,-,-,-,-,-,- A*01:01,-,-,-,-,-           0    37.41         73</span></span>
<span></span>
<span><span class="co"># Class II SAB example</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">deepMatchR_example</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt;   BeadID          SpecAbbr                      Specificity  RawData</span></span>
<span><span class="co">#&gt; 1      1   -,-,-,-,-,-,-,-          -,-,-,-,-,-,-,-,-,-,-,-    15.51</span></span>
<span><span class="co">#&gt; 2      2   -,-,-,-,-,-,-,-          -,-,-,-,-,-,-,-,-,-,-,- 13759.52</span></span>
<span><span class="co">#&gt; 3      3 DR1,-,-,-,-,-,-,- DRB1*01:01,-,-,-,-,-,-,-,-,-,-,-   131.86</span></span>
<span><span class="co">#&gt;   NormalValue CountValue</span></span>
<span><span class="co">#&gt; 1          NA        128</span></span>
<span><span class="co">#&gt; 2          NA         56</span></span>
<span><span class="co">#&gt; 3       75.22        117</span></span>
<span></span>
<span><span class="co"># PRA example</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">deepMatchR_example</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>, <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt;     BeadID               SpecAbbr                         Specificity</span></span>
<span><span class="co">#&gt; 98       1                                                           </span></span>
<span><span class="co">#&gt; 99       2                                                           </span></span>
<span><span class="co">#&gt; 100      3 A2,-,B46,-,Bw6,-,Cw1,- A*02:01,A*02:07,B*46:01,-,C*01:02,-</span></span>
<span><span class="co">#&gt;     NormalValue RawData CountValue</span></span>
<span><span class="co">#&gt; 98           NA   12.13        272</span></span>
<span><span class="co">#&gt; 99           NA 9209.35        277</span></span>
<span><span class="co">#&gt; 100       551.9  602.17        152</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="creating-hla-genotypes">Creating HLA Genotypes<a class="anchor" aria-label="anchor" href="#creating-hla-genotypes"></a>
</h3>
<p>The <code><a href="../reference/hlaGeno.html">hlaGeno()</a></code> function creates standardized genotype
objects used throughout the package.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define recipient and donor genotypes</span></span>
<span><span class="va">recipient</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  A_1 <span class="op">=</span> <span class="st">"A*01:01"</span>, A_2 <span class="op">=</span> <span class="st">"A*02:01"</span>,</span>
<span>  B_1 <span class="op">=</span> <span class="st">"B*07:02"</span>, B_2 <span class="op">=</span> <span class="st">"B*08:01"</span>,</span>
<span>  DQA1_1 <span class="op">=</span> <span class="st">"DQA1*02:01"</span>, DQA1_2 <span class="op">=</span> <span class="st">"DQA1*05:05"</span>,</span>
<span>  DQB1_1 <span class="op">=</span> <span class="st">"DQB1*02:02"</span>, DQB1_2 <span class="op">=</span> <span class="st">"DQB1*03:01"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">donor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  A_1 <span class="op">=</span> <span class="st">"A*03:01"</span>, A_2 <span class="op">=</span> <span class="st">"A*24:02"</span>,</span>
<span>  B_1 <span class="op">=</span> <span class="st">"B*44:02"</span>, B_2 <span class="op">=</span> <span class="st">"B*51:01"</span>,</span>
<span>  DQA1_1 <span class="op">=</span> <span class="st">"DQA1*05:05"</span>, DQA1_2 <span class="op">=</span> <span class="st">"DQA1*01:02"</span>,</span>
<span>  DQB1_1 <span class="op">=</span> <span class="st">"DQB1*06:02"</span>, DQB1_2 <span class="op">=</span> <span class="st">"DQB1*03:01"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create genotype objects</span></span>
<span><span class="va">rgeno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hlaGeno.html">hlaGeno</a></span><span class="op">(</span><span class="va">recipient</span><span class="op">)</span></span>
<span><span class="va">dgeno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hlaGeno.html">hlaGeno</a></span><span class="op">(</span><span class="va">donor</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="sequence-level-analysis">Sequence-Level Analysis<a class="anchor" aria-label="anchor" href="#sequence-level-analysis"></a>
</h2>
<div class="section level3">
<h3 id="retrieving-allele-sequences">Retrieving Allele Sequences<a class="anchor" aria-label="anchor" href="#retrieving-allele-sequences"></a>
</h3>
<p>The <code><a href="../reference/getAlleleSequence.html">getAlleleSequence()</a></code> function retrieves protein
sequences from the IMGT/HLA database.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the amino acid sequence for A*01:01</span></span>
<span><span class="va">a0101_seq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getAlleleSequence.html">getAlleleSequence</a></span><span class="op">(</span><span class="st">"A*01:01"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View first 50 amino acids</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="va">a0101_seq</span>, <span class="fl">1</span>, <span class="fl">50</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRFIAVG"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="quantifying-amino-acid-mismatches">Quantifying Amino Acid Mismatches<a class="anchor" aria-label="anchor" href="#quantifying-amino-acid-mismatches"></a>
</h3>
<p>The <code><a href="../reference/quantifyMismatch.html">quantifyMismatch()</a></code> function compares two protein
sequences and counts differences. You can optionally filter by
biophysical properties.</p>
<div class="section level4">
<h4 id="basic-usage">Basic Usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h4>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seq1</span> <span class="op">&lt;-</span> <span class="st">"YFAMYGEKVAHTHVDTLYVRYHY"</span></span>
<span><span class="va">seq2</span> <span class="op">&lt;-</span> <span class="st">"YFDMYGEKVAHTHVDTLYVRFHY"</span></span>
<span></span>
<span><span class="co"># Count all mismatches</span></span>
<span><span class="fu"><a href="../reference/quantifyMismatch.html">quantifyMismatch</a></span><span class="op">(</span><span class="va">seq1</span>, <span class="va">seq2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 2</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="filtering-by-biophysical-properties">Filtering by Biophysical Properties<a class="anchor" aria-label="anchor" href="#filtering-by-biophysical-properties"></a>
</h4>
<p><strong>deepMatchR</strong> uses these classification models:</p>
<ul>
<li>
<strong>Charge:</strong> K/R/H = positive; D/E = negative; others =
neutral</li>
<li>
<strong>Polarity:</strong> Nonpolar {A,V,L,I,P,M,F,W,G}; Polar
{S,T,N,Q,Y,C,H,K,R,D,E}</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Only charge-changing mismatches</span></span>
<span><span class="fu"><a href="../reference/quantifyMismatch.html">quantifyMismatch</a></span><span class="op">(</span><span class="va">seq1</span>, </span>
<span>                 <span class="va">seq2</span>, </span>
<span>                 filter_charge <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1</span></span>
<span></span>
<span><span class="co"># Only polarity-changing mismatches</span></span>
<span><span class="fu"><a href="../reference/quantifyMismatch.html">quantifyMismatch</a></span><span class="op">(</span><span class="va">seq1</span>, </span>
<span>                 <span class="va">seq2</span>, </span>
<span>                 filter_polarity <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 2</span></span>
<span></span>
<span><span class="co"># Both charge AND polarity changing</span></span>
<span><span class="fu"><a href="../reference/quantifyMismatch.html">quantifyMismatch</a></span><span class="op">(</span><span class="va">seq1</span>, </span>
<span>                 <span class="va">seq2</span>, </span>
<span>                 filter_charge <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                 filter_polarity <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="detailed-position-by-position-analysis">Detailed Position-by-Position Analysis<a class="anchor" aria-label="anchor" href="#detailed-position-by-position-analysis"></a>
</h4>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">details</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantifyMismatch.html">quantifyMismatch</a></span><span class="op">(</span><span class="va">seq1</span>, </span>
<span>                            <span class="va">seq2</span>, </span>
<span>                            return <span class="op">=</span> <span class="st">"detail"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">details</span><span class="op">)</span></span>
<span><span class="co">#&gt;   alignment_position ref alt is_gap_ref is_gap_alt is_mismatch charge_ref</span></span>
<span><span class="co">#&gt; 1                  1   Y   Y      FALSE      FALSE       FALSE        neu</span></span>
<span><span class="co">#&gt; 2                  2   F   F      FALSE      FALSE       FALSE        neu</span></span>
<span><span class="co">#&gt; 3                  3   A   D      FALSE      FALSE        TRUE        neu</span></span>
<span><span class="co">#&gt; 4                  4   M   M      FALSE      FALSE       FALSE        neu</span></span>
<span><span class="co">#&gt; 5                  5   Y   Y      FALSE      FALSE       FALSE        neu</span></span>
<span><span class="co">#&gt; 6                  6   G   G      FALSE      FALSE       FALSE        neu</span></span>
<span><span class="co">#&gt;   charge_alt charge_change polarity_ref polarity_alt polarity_change counted</span></span>
<span><span class="co">#&gt; 1        neu         FALSE        polar        polar           FALSE   FALSE</span></span>
<span><span class="co">#&gt; 2        neu         FALSE     nonpolar     nonpolar           FALSE   FALSE</span></span>
<span><span class="co">#&gt; 3        neg          TRUE     nonpolar        polar            TRUE    TRUE</span></span>
<span><span class="co">#&gt; 4        neu         FALSE     nonpolar     nonpolar           FALSE   FALSE</span></span>
<span><span class="co">#&gt; 5        neu         FALSE        polar        polar           FALSE   FALSE</span></span>
<span><span class="co">#&gt; 6        neu         FALSE     nonpolar     nonpolar           FALSE   FALSE</span></span></code></pre></div>
<p>The detail output includes:</p>
<ul>
<li>
<code>is_mismatch</code>: Whether amino acids differ at this
position</li>
<li>
<code>charge_change</code> / <code>polarity_change</code>: Whether
that property changed</li>
<li>
<code>counted</code>: Whether this position counts toward the final
tally</li>
</ul>
</div>
<div class="section level4">
<h4 id="function-reference">Function Reference<a class="anchor" aria-label="anchor" href="#function-reference"></a>
</h4>
<table class="table">
<colgroup>
<col width="31%">
<col width="28%">
<col width="40%">
</colgroup>
<thead><tr class="header">
<th>Argument</th>
<th>Default</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>
<code>sequence1</code>, <code>sequence2</code>
</td>
<td>–</td>
<td>Two protein sequences (same length)</td>
</tr>
<tr class="even">
<td><code>filter_charge</code></td>
<td><code>NULL</code></td>
<td>Filter by charge change
(<code>TRUE</code>/<code>FALSE</code>/<code>NULL</code>)</td>
</tr>
<tr class="odd">
<td><code>filter_polarity</code></td>
<td><code>NULL</code></td>
<td>Filter by polarity change
(<code>TRUE</code>/<code>FALSE</code>/<code>NULL</code>)</td>
</tr>
<tr class="even">
<td><code>return</code></td>
<td><code>"count"</code></td>
<td>Output type: <code>"count"</code> or <code>"detail"</code>
</td>
</tr>
<tr class="odd">
<td><code>na_action</code></td>
<td><code>"exclude"</code></td>
<td>Handle unknown residues: <code>"exclude"</code>,
<code>"error"</code>, <code>"count"</code>
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="section level2">
<h2 id="donor-recipient-matching">Donor-Recipient Matching<a class="anchor" aria-label="anchor" href="#donor-recipient-matching"></a>
</h2>
<div class="section level3">
<h3 id="calculating-mismatch-load">Calculating Mismatch Load<a class="anchor" aria-label="anchor" href="#calculating-mismatch-load"></a>
</h3>
<p>The <code><a href="../reference/calculateMismatchLoad.html">calculateMismatchLoad()</a></code> function aggregates amino
acid mismatches across HLA loci between a donor and recipient.</p>
<div class="section level4">
<h4 id="total-mismatch-load">Total Mismatch Load<a class="anchor" aria-label="anchor" href="#total-mismatch-load"></a>
</h4>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/calculateMismatchLoad.html">calculateMismatchLoad</a></span><span class="op">(</span><span class="va">rgeno</span>, </span>
<span>                      <span class="va">dgeno</span>, </span>
<span>                      parallel <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 378</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="per-locus-breakdown">Per-Locus Breakdown<a class="anchor" aria-label="anchor" href="#per-locus-breakdown"></a>
</h4>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">per_locus</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateMismatchLoad.html">calculateMismatchLoad</a></span><span class="op">(</span></span>
<span>  <span class="va">rgeno</span>, <span class="va">dgeno</span>,</span>
<span>  return <span class="op">=</span> <span class="st">"per_locus"</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">per_locus</span></span>
<span><span class="co">#&gt;   locus mismatch_load</span></span>
<span><span class="co">#&gt; 1     A            96</span></span>
<span><span class="co">#&gt; 2     B           128</span></span>
<span><span class="co">#&gt; 3  DQA1            71</span></span>
<span><span class="co">#&gt; 4  DQB1            83</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="pairwise-allele-matrix">Pairwise Allele Matrix<a class="anchor" aria-label="anchor" href="#pairwise-allele-matrix"></a>
</h4>
<p>View mismatches between each recipient-donor allele pair at a
specific locus:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mB</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateMismatchLoad.html">calculateMismatchLoad</a></span><span class="op">(</span></span>
<span>  <span class="va">rgeno</span>, <span class="va">dgeno</span>,</span>
<span>  return <span class="op">=</span> <span class="st">"pairwise"</span>,</span>
<span>  pairwise_locus <span class="op">=</span> <span class="st">"B"</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">mB</span></span>
<span><span class="co">#&gt;          donor</span></span>
<span><span class="co">#&gt; recipient B*44:02 B*51:01</span></span>
<span><span class="co">#&gt;   B*07:02      37      33</span></span>
<span><span class="co">#&gt;   B*08:01      31      27</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="applying-filters">Applying Filters<a class="anchor" aria-label="anchor" href="#applying-filters"></a>
</h4>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Charge-changing mismatches only</span></span>
<span><span class="fu"><a href="../reference/calculateMismatchLoad.html">calculateMismatchLoad</a></span><span class="op">(</span><span class="va">rgeno</span>, </span>
<span>                      <span class="va">dgeno</span>, </span>
<span>                      filter_charge <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                      parallel <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 140</span></span>
<span></span>
<span><span class="co"># Polarity-changing mismatches only</span></span>
<span><span class="fu"><a href="../reference/calculateMismatchLoad.html">calculateMismatchLoad</a></span><span class="op">(</span><span class="va">rgeno</span>, </span>
<span>                      <span class="va">dgeno</span>, </span>
<span>                      filter_polarity <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                      parallel <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 138</span></span>
<span></span>
<span><span class="co"># Restrict to specific loci</span></span>
<span><span class="fu"><a href="../reference/calculateMismatchLoad.html">calculateMismatchLoad</a></span><span class="op">(</span><span class="va">rgeno</span>, </span>
<span>                      <span class="va">dgeno</span>, </span>
<span>                      loci <span class="op">=</span> <span class="st">"A"</span>, </span>
<span>                      parallel <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 96</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="function-reference-1">Function Reference<a class="anchor" aria-label="anchor" href="#function-reference-1"></a>
</h4>
<table class="table">
<thead><tr class="header">
<th>Argument</th>
<th>Default</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>
<code>recipient_geno</code>, <code>donor_geno</code>
</td>
<td>–</td>
<td>HLA genotypes from <code><a href="../reference/hlaGeno.html">hlaGeno()</a></code>
</td>
</tr>
<tr class="even">
<td><code>loci</code></td>
<td><code>NULL</code></td>
<td>Restrict to specific loci (e.g., <code>c("A","B")</code>)</td>
</tr>
<tr class="odd">
<td>
<code>filter_charge</code>, <code>filter_polarity</code>
</td>
<td><code>NULL</code></td>
<td>Biophysical filters</td>
</tr>
<tr class="even">
<td><code>return</code></td>
<td><code>"total"</code></td>
<td>
<code>"total"</code>, <code>"per_locus"</code>, or
<code>"pairwise"</code>
</td>
</tr>
<tr class="odd">
<td><code>pairwise_locus</code></td>
<td><code>NULL</code></td>
<td>Locus for pairwise matrix</td>
</tr>
<tr class="even">
<td><code>parallel</code></td>
<td><code>TRUE</code></td>
<td>Use parallel processing</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="section level2">
<h2 id="eplet-analysis">Eplet Analysis<a class="anchor" aria-label="anchor" href="#eplet-analysis"></a>
</h2>
<p>Eplets are short amino acid sequences that form antibody-binding
epitopes on HLA molecules. They are key determinants of HLA
immunogenicity.</p>
<div class="section level3">
<h3 id="quantifying-eplet-mismatches">Quantifying Eplet Mismatches<a class="anchor" aria-label="anchor" href="#quantifying-eplet-mismatches"></a>
</h3>
<p>The <code><a href="../reference/quantifyEpletMismatch.html">quantifyEpletMismatch()</a></code> function calculates the
number of non-shared eplets between two alleles.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Basic eplet comparison (A1/A2 evidence level)</span></span>
<span><span class="fu"><a href="../reference/quantifyEpletMismatch.html">quantifyEpletMismatch</a></span><span class="op">(</span><span class="st">"A*01:01"</span>, </span>
<span>                      <span class="st">"A*02:01"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 13</span></span>
<span></span>
<span><span class="co"># Filter by exposition level</span></span>
<span><span class="fu"><a href="../reference/quantifyEpletMismatch.html">quantifyEpletMismatch</a></span><span class="op">(</span><span class="st">"B*07:02"</span>, </span>
<span>                      <span class="st">"B*44:02"</span>, </span>
<span>                      exposition_filter <span class="op">=</span> <span class="st">"High"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 6</span></span>
<span></span>
<span><span class="co"># Filter by reactivity confirmation</span></span>
<span><span class="fu"><a href="../reference/quantifyEpletMismatch.html">quantifyEpletMismatch</a></span><span class="op">(</span><span class="st">"C*07:01"</span>, </span>
<span>                      <span class="st">"C*06:02"</span>,</span>
<span>                      evidence_level <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                      reactivity_filter <span class="op">=</span> <span class="st">"Confirmed"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 9</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="calculating-eplet-load">Calculating Eplet Load<a class="anchor" aria-label="anchor" href="#calculating-eplet-load"></a>
</h3>
<p>The <code><a href="../reference/calculateEpletLoad.html">calculateEpletLoad()</a></code> function calculates total
donor-specific eplets across all loci.</p>
<div class="section level4">
<h4 id="total-eplet-load">Total Eplet Load<a class="anchor" aria-label="anchor" href="#total-eplet-load"></a>
</h4>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/calculateEpletLoad.html">calculateEpletLoad</a></span><span class="op">(</span><span class="va">rgeno</span>, </span>
<span>                   <span class="va">dgeno</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 19</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="per-locus-summary">Per-Locus Summary<a class="anchor" aria-label="anchor" href="#per-locus-summary"></a>
</h4>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">per</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateEpletLoad.html">calculateEpletLoad</a></span><span class="op">(</span><span class="va">rgeno</span>, </span>
<span>                          <span class="va">dgeno</span>, </span>
<span>                          return <span class="op">=</span> <span class="st">"per_locus"</span><span class="op">)</span></span>
<span><span class="va">per</span></span>
<span><span class="co">#&gt;   locus eplet_load</span></span>
<span><span class="co">#&gt; 1     A          4</span></span>
<span><span class="co">#&gt; 2     B          6</span></span>
<span><span class="co">#&gt; 3  DQA1          1</span></span>
<span><span class="co">#&gt; 4  DQB1          8</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="pairwise-allele-matrix-1">Pairwise Allele Matrix<a class="anchor" aria-label="anchor" href="#pairwise-allele-matrix-1"></a>
</h4>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mB</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateEpletLoad.html">calculateEpletLoad</a></span><span class="op">(</span></span>
<span>  <span class="va">rgeno</span>, <span class="va">dgeno</span>,</span>
<span>  return <span class="op">=</span> <span class="st">"pairwise"</span>,</span>
<span>  pairwise_locus <span class="op">=</span> <span class="st">"B"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">mB</span></span>
<span><span class="co">#&gt;          donor</span></span>
<span><span class="co">#&gt; recipient B*44:02 B*51:01</span></span>
<span><span class="co">#&gt;   B*07:02       5       4</span></span>
<span><span class="co">#&gt;   B*08:01       4       3</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="applying-evidence-and-exposition-filters">Applying Evidence and Exposition Filters<a class="anchor" aria-label="anchor" href="#applying-evidence-and-exposition-filters"></a>
</h4>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># High-exposition eplets only</span></span>
<span><span class="fu"><a href="../reference/calculateEpletLoad.html">calculateEpletLoad</a></span><span class="op">(</span><span class="va">rgeno</span>, </span>
<span>                   <span class="va">dgeno</span>, </span>
<span>                   exposition_filter <span class="op">=</span> <span class="st">"High"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 15</span></span>
<span></span>
<span><span class="co"># Antibody-confirmed eplets only</span></span>
<span><span class="fu"><a href="../reference/calculateEpletLoad.html">calculateEpletLoad</a></span><span class="op">(</span><span class="va">rgeno</span>, </span>
<span>                   <span class="va">dgeno</span>, </span>
<span>                   reactivity_filter <span class="op">=</span> <span class="st">"Confirmed"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 19</span></span>
<span></span>
<span><span class="co"># All evidence levels (no filter)</span></span>
<span><span class="fu"><a href="../reference/calculateEpletLoad.html">calculateEpletLoad</a></span><span class="op">(</span><span class="va">rgeno</span>, </span>
<span>                   <span class="va">dgeno</span>, </span>
<span>                   evidence_level <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 79</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="peptide-binding-prediction">Peptide Binding Prediction<a class="anchor" aria-label="anchor" href="#peptide-binding-prediction"></a>
</h2>
<div class="section level3">
<h3 id="mhcnuggets-integration">MHCnuggets Integration<a class="anchor" aria-label="anchor" href="#mhcnuggets-integration"></a>
</h3>
<p>The <code><a href="../reference/predictMHCnuggets.html">predictMHCnuggets()</a></code> function provides an R interface
to the MHCnuggets deep learning model for peptide-MHC binding
prediction.</p>
<blockquote>
<p><strong>Note:</strong> This section requires Python dependencies and
is skipped on Windows due to path length limitations with the
TensorFlow/MHCnuggets installation.</p>
</blockquote>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define peptides and allele</span></span>
<span><span class="va">peptides</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SIINFEKL"</span>, <span class="st">"LLFGYPVYV"</span><span class="op">)</span></span>
<span><span class="va">allele</span> <span class="op">&lt;-</span> <span class="st">"A*02:01"</span></span>
<span></span>
<span><span class="co"># Predict binding affinity (IC50)</span></span>
<span><span class="va">binding_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/predictMHCnuggets.html">predictMHCnuggets</a></span><span class="op">(</span></span>
<span>  peptides <span class="op">=</span> <span class="va">peptides</span>,</span>
<span>  allele <span class="op">=</span> <span class="va">allele</span>,</span>
<span>  mhc_class <span class="op">=</span> <span class="st">"I"</span>,</span>
<span>  rank_output <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Predicting for 2 peptides</span></span>
<span><span class="co">#&gt; Number of peptides skipped/total due to length 0 / 0</span></span>
<span><span class="co">#&gt; Building model</span></span>
<span><span class="co">#&gt; Closest allele found HLA-A02:01</span></span>
<span><span class="co">#&gt; BA_to_HLAp model found, predicting with BA_to_HLAp model...</span></span>
<span><span class="co">#&gt; Rank output selected, computing peptide IC50 ranks against human proteome peptides...</span></span>
<span><span class="co">#&gt; Writing output files...</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">binding_results</span><span class="op">)</span></span>
<span><span class="co">#&gt;     peptide    ic50</span></span>
<span><span class="co">#&gt; 1  SIINFEKL 5600.06</span></span>
<span><span class="co">#&gt; 2 LLFGYPVYV  535.92</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="calculating-peptide-binding-load">Calculating Peptide Binding Load<a class="anchor" aria-label="anchor" href="#calculating-peptide-binding-load"></a>
</h3>
<p>The <code><a href="../reference/calculatePeptideBindingLoad.html">calculatePeptideBindingLoad()</a></code> function predicts
transplant risk by analyzing peptide-HLA binding between recipient
molecules and donor-derived peptides.</p>
<p><strong>Workflow:</strong></p>
<ol style="list-style-type: decimal">
<li>
<strong>Input processing</strong> - Standardizes recipient/donor
allele formats</li>
<li>
<strong>Peptide derivation</strong> - Generates mismatched peptides
from sequence differences</li>
<li>
<strong>Binding prediction</strong> - Predicts IC50 using PWM,
NetMHCpan, or MHCflurry</li>
<li>
<strong>Risk calculation</strong> - Aggregates predictions into risk
scores</li>
</ol>
<div class="section level4">
<h4 id="total-risk-score">Total Risk Score<a class="anchor" aria-label="anchor" href="#total-risk-score"></a>
</h4>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">total_risk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculatePeptideBindingLoad.html">calculatePeptideBindingLoad</a></span><span class="op">(</span></span>
<span>  recipient <span class="op">=</span> <span class="va">rgeno</span>,</span>
<span>  donor <span class="op">=</span> <span class="va">dgeno</span>,</span>
<span>  return <span class="op">=</span> <span class="st">"total"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">total_risk</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="per-allele-summary">Per-Allele Summary<a class="anchor" aria-label="anchor" href="#per-allele-summary"></a>
</h4>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">summary_load</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculatePeptideBindingLoad.html">calculatePeptideBindingLoad</a></span><span class="op">(</span></span>
<span>  recipient <span class="op">=</span> <span class="va">rgeno</span>,</span>
<span>  donor <span class="op">=</span> <span class="va">dgeno</span>,</span>
<span>  return <span class="op">=</span> <span class="st">"summary"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">summary_load</span><span class="op">)</span></span>
<span><span class="co">#&gt;   hla_allele n_peptides n_strong n_weak risk_contribution</span></span>
<span><span class="co">#&gt; 1    A*01:01          0        0      0                 0</span></span>
<span><span class="co">#&gt; 2    A*02:01          0        0      0                 0</span></span>
<span><span class="co">#&gt; 3    B*07:02          0        0      0                 0</span></span>
<span><span class="co">#&gt; 4    B*08:01          0        0      0                 0</span></span>
<span><span class="co">#&gt; 5 DQA1*02:01          0        0      0                 0</span></span>
<span><span class="co">#&gt; 6 DQA1*05:05          0        0      0                 0</span></span>
<span><span class="co">#&gt; 7 DQB1*02:02          0        0      0                 0</span></span>
<span><span class="co">#&gt; 8 DQB1*03:01          0        0      0                 0</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="detailed-peptide-level-results">Detailed Peptide-Level Results<a class="anchor" aria-label="anchor" href="#detailed-peptide-level-results"></a>
</h4>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">detailed_load</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculatePeptideBindingLoad.html">calculatePeptideBindingLoad</a></span><span class="op">(</span></span>
<span>  recipient <span class="op">=</span> <span class="va">rgeno</span>,</span>
<span>  donor <span class="op">=</span> <span class="va">dgeno</span>,</span>
<span>  return <span class="op">=</span> <span class="st">"detail"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">detailed_load</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] peptide        hla_allele     predicted_ic50 binding_level  contribution  </span></span>
<span><span class="co">#&gt; &lt;0 rows&gt; (or 0-length row.names)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="visualizing-antibody-data">Visualizing Antibody Data<a class="anchor" aria-label="anchor" href="#visualizing-antibody-data"></a>
</h2>
<div class="section level3">
<h3 id="antibody-plots-with-plotantibodies">Antibody Plots with <code>plotAntibodies()</code><a class="anchor" aria-label="anchor" href="#antibody-plots-with-plotantibodies"></a>
</h3>
<p>The <code><a href="../reference/plotAntibodies.html">plotAntibodies()</a></code> function creates
publication-quality visualizations of SAB and PRA assay data.</p>
<div class="section level4">
<h4 id="sab-bar-plot">SAB Bar Plot<a class="anchor" aria-label="anchor" href="#sab-bar-plot"></a>
</h4>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotAntibodies.html">plotAntibodies</a></span><span class="op">(</span></span>
<span>  result_file <span class="op">=</span> <span class="va">deepMatchR_example</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"SAB"</span>,</span>
<span>  bead_cutoffs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2000</span>, <span class="fl">1000</span>, <span class="fl">500</span>, <span class="fl">250</span><span class="op">)</span>,</span>
<span>  add_table <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  palette <span class="op">=</span> <span class="st">"spectral"</span>,</span>
<span>  highlight_antigen <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Bw4"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="deepMatchR_files/figure-html/plot-sab-1.png" class="r-plt" alt="" width="960"></p>
</div>
<div class="section level4">
<h4 id="pra-bar-plot">PRA Bar Plot<a class="anchor" aria-label="anchor" href="#pra-bar-plot"></a>
</h4>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotAntibodies.html">plotAntibodies</a></span><span class="op">(</span></span>
<span>  result_file <span class="op">=</span> <span class="va">deepMatchR_example</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"PRA"</span>,</span>
<span>  class <span class="op">=</span> <span class="st">"I"</span>,</span>
<span>  add_table <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  palette <span class="op">=</span> <span class="st">"spectral"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="deepMatchR_files/figure-html/plot-pra-1.png" class="r-plt" alt="" width="960"></p>
</div>
<div class="section level4">
<h4 id="time-series-trend-plot">Time-Series Trend Plot<a class="anchor" aria-label="anchor" href="#time-series-trend-plot"></a>
</h4>
<p>Track antibody levels over multiple time points:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Simulate longitudinal data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">sab_data_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">"01/01/2023"</span> <span class="op">=</span> <span class="va">deepMatchR_example</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  <span class="st">"02/01/2023"</span> <span class="op">=</span> <span class="va">deepMatchR_example</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>NormalValue <span class="op">=</span> <span class="va">NormalValue</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>, <span class="fl">0.5</span>, <span class="fl">0.8</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="st">"03/01/2023"</span> <span class="op">=</span> <span class="va">deepMatchR_example</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>NormalValue <span class="op">=</span> <span class="va">NormalValue</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>, <span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotAntibodies.html">plotAntibodies</a></span><span class="op">(</span></span>
<span>  result_file <span class="op">=</span> <span class="va">sab_data_list</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"SAB"</span>,</span>
<span>  plot_trend <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  highlight_threshold <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  vline_dates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"2023-02-15"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="deepMatchR_files/figure-html/plot-trend-1.png" class="r-plt" alt="" width="960"></p>
</div>
</div>
<div class="section level3">
<h3 id="eplet-reactivity-analysis-with-epletauc">Eplet Reactivity Analysis with <code>epletAUC()</code><a class="anchor" aria-label="anchor" href="#eplet-reactivity-analysis-with-epletauc"></a>
</h3>
<p>The <code><a href="../reference/calculateAUC.html">epletAUC()</a></code> function calculates the Area Under the
Curve (AUC) for eplet reactivity across multiple MFI thresholds. This
provides a more robust measure than a single cutoff.</p>
<div class="section level4">
<h4 id="visualize-reactivity-curves">Visualize Reactivity Curves<a class="anchor" aria-label="anchor" href="#visualize-reactivity-curves"></a>
</h4>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/calculateAUC.html">epletAUC</a></span><span class="op">(</span></span>
<span>  result_file <span class="op">=</span> <span class="va">deepMatchR_example</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  group_by <span class="op">=</span> <span class="st">"evidence"</span>,</span>
<span>  evidence_level <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A1"</span>, <span class="st">"A2"</span>, <span class="st">"B"</span>, <span class="st">"D"</span><span class="op">)</span>,</span>
<span>  plot_results <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  cut_min <span class="op">=</span> <span class="fl">250</span>,</span>
<span>  cut_max <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  cut_step <span class="op">=</span> <span class="fl">250</span>,</span>
<span>  palette <span class="op">=</span> <span class="st">"inferno"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="deepMatchR_files/figure-html/eplet-auc-plot-1.png" class="r-plt" alt="" width="768"></p>
</div>
<div class="section level4">
<h4 id="get-auc-values-as-data">Get AUC Values as Data<a class="anchor" aria-label="anchor" href="#get-auc-values-as-data"></a>
</h4>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">auc_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateAUC.html">epletAUC</a></span><span class="op">(</span></span>
<span>  result_file <span class="op">=</span> <span class="va">deepMatchR_example</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  plot_results <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  cut_min <span class="op">=</span> <span class="fl">250</span>,</span>
<span>  cut_max <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  cut_step <span class="op">=</span> <span class="fl">250</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">auc_result</span><span class="op">)</span></span>
<span><span class="co">#&gt;     eplet      AUC total_count   loci  norm_AUC</span></span>
<span><span class="co">#&gt;    &lt;char&gt;    &lt;num&gt;       &lt;int&gt; &lt;char&gt;     &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1:   82LR 7817.708          24   A; B 0.7817708</span></span>
<span><span class="co">#&gt; 2:    80I 7022.059          17   A; B 0.7022059</span></span>
<span><span class="co">#&gt; 3:   69AA 5641.667          15      B 0.5641667</span></span>
<span><span class="co">#&gt; 4:  163EW 6691.667          15   A; B 0.6691667</span></span>
<span><span class="co">#&gt; 5:    41T 9583.333          12      B 0.9583333</span></span>
<span><span class="co">#&gt; 6:  65QIA 4761.364          11      B 0.4761364</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="advanced-filtering">Advanced Filtering<a class="anchor" aria-label="anchor" href="#advanced-filtering"></a>
</h4>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/calculateAUC.html">epletAUC</a></span><span class="op">(</span></span>
<span>  result_file <span class="op">=</span> <span class="va">deepMatchR_example</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  label <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plot_results <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  eplet_filter <span class="op">=</span> <span class="fl">10</span>,      <span class="co"># Minimum bead count</span></span>
<span>  percPos_filter <span class="op">=</span> <span class="fl">1.0</span>    <span class="co"># Minimum positivity threshold</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="deepMatchR_files/figure-html/eplet-auc-filter-1.png" class="r-plt" alt="" width="768"></p>
</div>
</div>
<div class="section level3">
<h3 id="eplet-visualization-with-ploteplets">Eplet Visualization with <code>plotEplets()</code><a class="anchor" aria-label="anchor" href="#eplet-visualization-with-ploteplets"></a>
</h3>
<p>The <code><a href="../reference/plotEplets.html">plotEplets()</a></code> function offers three visualization
types for eplet reactivity data.</p>
<div class="section level4">
<h4 id="treemap-view">Treemap View<a class="anchor" aria-label="anchor" href="#treemap-view"></a>
</h4>
<p>Ideal for an overview of eplet prominence. Tile area represents
importance (positive beads x positivity rate):</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotEplets.html">plotEplets</a></span><span class="op">(</span></span>
<span>  result_file <span class="op">=</span> <span class="va">deepMatchR_example</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  plot_type <span class="op">=</span> <span class="st">"treemap"</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  evidence_level <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A1"</span>, <span class="st">"A2"</span>, <span class="st">"B"</span><span class="op">)</span>,</span>
<span>  percPos_filter <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>  palette <span class="op">=</span> <span class="st">"spectral"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="deepMatchR_files/figure-html/plot-treemap-1.png" class="r-plt" alt="" width="960"></p>
</div>
<div class="section level4">
<h4 id="bar-plot">Bar Plot<a class="anchor" aria-label="anchor" href="#bar-plot"></a>
</h4>
<p>Ranks eplets by positivity rate at a specific threshold:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotEplets.html">plotEplets</a></span><span class="op">(</span></span>
<span>  result_file <span class="op">=</span> <span class="va">deepMatchR_example</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  plot_type <span class="op">=</span> <span class="st">"bar"</span>,</span>
<span>  group_by <span class="op">=</span> <span class="st">"loci"</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  evidence_level <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A1"</span>, <span class="st">"A2"</span>, <span class="st">"B"</span><span class="op">)</span>,</span>
<span>  percPos_filter <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>  top_eplets <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  palette <span class="op">=</span> <span class="st">"spectral"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="deepMatchR_files/figure-html/plot-bar-1.png" class="r-plt" alt="" width="768"></p>
</div>
<div class="section level4">
<h4 id="auc-bar-plot">AUC Bar Plot<a class="anchor" aria-label="anchor" href="#auc-bar-plot"></a>
</h4>
<p>Ranks eplets by performance across multiple thresholds:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotEplets.html">plotEplets</a></span><span class="op">(</span></span>
<span> result_file <span class="op">=</span> <span class="va">deepMatchR_example</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  plot_type <span class="op">=</span> <span class="st">"AUC"</span>,</span>
<span>  percPos_filter <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>  group_by <span class="op">=</span> <span class="st">"evidence_level"</span>,</span>
<span>  cut_min <span class="op">=</span> <span class="fl">250</span>,</span>
<span>  cut_max <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  cut_step <span class="op">=</span> <span class="fl">250</span>,</span>
<span>  top_eplets <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  palette <span class="op">=</span> <span class="st">"spectral"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="deepMatchR_files/figure-html/plot-auc-bar-1.png" class="r-plt" alt="" width="768"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.2 (2025-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ggplot2_4.0.1     dplyr_1.1.4       deepMatchR_0.99.0 BiocStyle_2.38.0 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] ggfittext_0.10.3       gtable_0.3.6           dir.expiry_1.18.0     </span></span>
<span><span class="co">#&gt;  [4] xfun_0.56              bslib_0.9.0            lattice_0.22-7        </span></span>
<span><span class="co">#&gt;  [7] quadprog_1.5-8         vctrs_0.7.1            tools_4.5.2           </span></span>
<span><span class="co">#&gt; [10] generics_0.1.4         stats4_4.5.2           parallel_4.5.2        </span></span>
<span><span class="co">#&gt; [13] tibble_3.3.1           pkgconfig_2.0.3        Matrix_1.7-4          </span></span>
<span><span class="co">#&gt; [16] data.table_1.18.0      RColorBrewer_1.1-3     S7_0.2.1              </span></span>
<span><span class="co">#&gt; [19] desc_1.4.3             S4Vectors_0.48.0       readxl_1.4.5          </span></span>
<span><span class="co">#&gt; [22] lifecycle_1.0.5        compiler_4.5.2         farver_2.1.2          </span></span>
<span><span class="co">#&gt; [25] textshaping_1.0.4      Biostrings_2.78.0      Seqinfo_1.0.0         </span></span>
<span><span class="co">#&gt; [28] htmltools_0.5.9        sass_0.4.10            yaml_2.3.12           </span></span>
<span><span class="co">#&gt; [31] pillar_1.11.1          pkgdown_2.2.0          crayon_1.5.3          </span></span>
<span><span class="co">#&gt; [34] jquerylib_0.1.4        cachem_1.1.0           basilisk_1.22.0       </span></span>
<span><span class="co">#&gt; [37] tidyselect_1.2.1       rvest_1.0.5            digest_0.6.39         </span></span>
<span><span class="co">#&gt; [40] stringi_1.8.7          bookdown_0.46          labeling_0.4.3        </span></span>
<span><span class="co">#&gt; [43] fastmap_1.2.0          grid_4.5.2             treemapify_2.6.0      </span></span>
<span><span class="co">#&gt; [46] cli_3.6.5              magrittr_2.0.4         patchwork_1.3.2       </span></span>
<span><span class="co">#&gt; [49] withr_3.0.2            filelock_1.0.3         scales_1.4.0          </span></span>
<span><span class="co">#&gt; [52] rmarkdown_2.30         pwalign_1.6.0          XVector_0.50.0        </span></span>
<span><span class="co">#&gt; [55] httr_1.4.7             reticulate_1.44.1      cellranger_1.1.0      </span></span>
<span><span class="co">#&gt; [58] ragg_1.5.0             png_0.1-8              memoise_2.0.1         </span></span>
<span><span class="co">#&gt; [61] evaluate_1.0.5         knitr_1.51             IRanges_2.44.0        </span></span>
<span><span class="co">#&gt; [64] immReferent_0.99.6     rlang_1.1.7            Rcpp_1.1.1            </span></span>
<span><span class="co">#&gt; [67] glue_1.8.0             BiocManager_1.30.27    xml2_1.5.2            </span></span>
<span><span class="co">#&gt; [70] directlabels_2025.6.24 BiocGenerics_0.56.0    svglite_2.2.2         </span></span>
<span><span class="co">#&gt; [73] jsonlite_2.0.0         R6_2.6.1               systemfonts_1.3.1     </span></span>
<span><span class="co">#&gt; [76] fs_1.6.6</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="getting-help">Getting Help<a class="anchor" aria-label="anchor" href="#getting-help"></a>
</h2>
<p>If you have questions, suggestions, or encounter issues:</p>
<ul>
<li>
<strong>GitHub Issues:</strong> <a href="https://github.com/ncborcherding/deepMatchR/issues" class="external-link">github.com/ncborcherding/deepMatchR/issues</a>
</li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/ncborcherding" class="external-link">Nick Borcherding</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
