<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Peptide Binding Prediction • deepMatchR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Peptide Binding Prediction">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">deepMatchR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/deepMatchR.html">Get Started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/index.html">Overview</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Core Workflows</h6></li>
    <li><a class="dropdown-item" href="../articles/sequence-analysis.html">Sequence &amp; Mismatch Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/eplet-analysis.html">Eplet Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/peptide-binding.html">Peptide Binding Prediction</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Visualization</h6></li>
    <li><a class="dropdown-item" href="../articles/antibody-visualization.html">Antibody Visualization</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/BorchLab/deepMatchR" aria-label="GitHub"><span class="fa fab fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Peptide Binding Prediction</h1>
                        <h4 data-toc-skip class="author">Nick
Borcherding</h4>
            <address class="author_afil">
      Washington University in St. Louis, School of Medicine, St. Louis,
MO, USA<br><h4 data-toc-skip class="date">Compiled: January 25,
2026</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BorchLab/deepMatchR/blob/HEAD/vignettes/peptide-binding.Rmd" class="external-link"><code>vignettes/peptide-binding.Rmd</code></a></small>
      <div class="d-none name"><code>peptide-binding.Rmd</code></div>
    </address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Peptide-MHC binding prediction is central to understanding T
cell-mediated immunity. In the context of transplantation, predicting
how recipient HLA molecules might present donor-derived peptides helps
assess immunological risk.</p>
<p>This article covers:</p>
<ul>
<li>Using MHCnuggets for peptide binding prediction</li>
<li>Calculating peptide binding load for donor-recipient pairs</li>
<li>Interpreting binding predictions</li>
</ul>
</div>
<div class="section level2">
<h2 id="loading-the-package">Loading the Package<a class="anchor" aria-label="anchor" href="#loading-the-package"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BorchLab/deepMatchR" class="external-link">deepMatchR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<div class="section level3">
<h3 id="peptide-mhc-binding">Peptide-MHC Binding<a class="anchor" aria-label="anchor" href="#peptide-mhc-binding"></a>
</h3>
<p>MHC molecules present peptides to T cells. The binding affinity
between a peptide and an MHC molecule determines:</p>
<ul>
<li>
<strong>Presentation likelihood:</strong> Strong binders are more
likely to be presented</li>
<li>
<strong>T cell activation potential:</strong> Better presentation
leads to stronger T cell responses</li>
<li>
<strong>Immunogenicity:</strong> Peptides that bind well and differ
from self are immunogenic</li>
</ul>
</div>
<div class="section level3">
<h3 id="binding-thresholds">Binding Thresholds<a class="anchor" aria-label="anchor" href="#binding-thresholds"></a>
</h3>
<p>Binding affinity is typically measured as IC50 (half-maximal
inhibitory concentration):</p>
<table class="table">
<thead><tr class="header">
<th>IC50 (nM)</th>
<th>Classification</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>&lt; 50</td>
<td>Strong binder</td>
</tr>
<tr class="even">
<td>50-500</td>
<td>Intermediate binder</td>
</tr>
<tr class="odd">
<td>500-5000</td>
<td>Weak binder</td>
</tr>
<tr class="even">
<td>&gt; 5000</td>
<td>Non-binder</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="mhcnuggets-prediction">MHCnuggets Prediction<a class="anchor" aria-label="anchor" href="#mhcnuggets-prediction"></a>
</h2>
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p><code><a href="../reference/predictMHCnuggets.html">predictMHCnuggets()</a></code> provides an R interface to
MHCnuggets, a deep learning model for peptide-MHC binding
prediction.</p>
<blockquote>
<p><strong>Note:</strong> MHCnuggets requires Python dependencies via
basilisk. This functionality is not available on Windows due to path
length limitations.</p>
</blockquote>
</div>
<div class="section level3">
<h3 id="basic-usage">Basic Usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define peptides and allele</span></span>
<span><span class="va">peptides</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SIINFEKL"</span>, <span class="st">"LLFGYPVYV"</span>, <span class="st">"GILGFVFTL"</span><span class="op">)</span></span>
<span><span class="va">allele</span> <span class="op">&lt;-</span> <span class="st">"A*02:01"</span></span>
<span></span>
<span><span class="co"># Predict binding affinity</span></span>
<span><span class="va">binding_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/predictMHCnuggets.html">predictMHCnuggets</a></span><span class="op">(</span></span>
<span>  peptides <span class="op">=</span> <span class="va">peptides</span>,</span>
<span>  allele <span class="op">=</span> <span class="va">allele</span>,</span>
<span>  mhc_class <span class="op">=</span> <span class="st">"I"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Predicting for 3 peptides</span></span>
<span><span class="co">#&gt; Number of peptides skipped/total due to length 0 / 0</span></span>
<span><span class="co">#&gt; Building model</span></span>
<span><span class="co">#&gt; Closest allele found HLA-A02:01</span></span>
<span><span class="co">#&gt; BA_to_HLAp model found, predicting with BA_to_HLAp model...</span></span>
<span><span class="co">#&gt; Writing output files...</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">binding_results</span><span class="op">)</span></span>
<span><span class="co">#&gt;     peptide    ic50</span></span>
<span><span class="co">#&gt; 1  SIINFEKL 5600.06</span></span>
<span><span class="co">#&gt; 2 LLFGYPVYV  535.92</span></span>
<span><span class="co">#&gt; 3 GILGFVFTL  589.25</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="interpretation">Interpretation<a class="anchor" aria-label="anchor" href="#interpretation"></a>
</h3>
<p>The output includes:</p>
<ul>
<li>
<strong>peptide:</strong> Input peptide sequence</li>
<li>
<strong>allele:</strong> HLA allele tested</li>
<li>
<strong>ic50:</strong> Predicted binding affinity in nM</li>
<li>
<strong>rank:</strong> Percentile rank (if
<code>rank_output = TRUE</code>)</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="peptide-binding-load">Peptide Binding Load<a class="anchor" aria-label="anchor" href="#peptide-binding-load"></a>
</h2>
<div class="section level3">
<h3 id="concept">Concept<a class="anchor" aria-label="anchor" href="#concept"></a>
</h3>
<p>The <code><a href="../reference/calculatePeptideBindingLoad.html">calculatePeptideBindingLoad()</a></code> function estimates
transplant risk by:</p>
<ol style="list-style-type: decimal">
<li>Identifying mismatched peptides from donor sequences</li>
<li>Predicting binding to recipient HLA molecules</li>
<li>Aggregating predictions into a risk score</li>
</ol>
</div>
<div class="section level3">
<h3 id="creating-genotypes">Creating Genotypes<a class="anchor" aria-label="anchor" href="#creating-genotypes"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">recipient</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  A_1 <span class="op">=</span> <span class="st">"A*02:01"</span>, A_2 <span class="op">=</span> <span class="st">"A*03:01"</span>,</span>
<span>  B_1 <span class="op">=</span> <span class="st">"B*07:02"</span>, B_2 <span class="op">=</span> <span class="st">"B*08:01"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">donor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  A_1 <span class="op">=</span> <span class="st">"A*01:01"</span>, A_2 <span class="op">=</span> <span class="st">"A*24:02"</span>,</span>
<span>  B_1 <span class="op">=</span> <span class="st">"B*44:02"</span>, B_2 <span class="op">=</span> <span class="st">"B*51:01"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">rgeno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hlaGeno.html">hlaGeno</a></span><span class="op">(</span><span class="va">recipient</span><span class="op">)</span></span>
<span><span class="va">dgeno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hlaGeno.html">hlaGeno</a></span><span class="op">(</span><span class="va">donor</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="total-risk-score">Total Risk Score<a class="anchor" aria-label="anchor" href="#total-risk-score"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">total_risk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculatePeptideBindingLoad.html">calculatePeptideBindingLoad</a></span><span class="op">(</span></span>
<span>  recipient <span class="op">=</span> <span class="va">rgeno</span>,</span>
<span>  donor <span class="op">=</span> <span class="va">dgeno</span>,</span>
<span>  return <span class="op">=</span> <span class="st">"total"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">total_risk</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="per-allele-summary">Per-Allele Summary<a class="anchor" aria-label="anchor" href="#per-allele-summary"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">summary_load</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculatePeptideBindingLoad.html">calculatePeptideBindingLoad</a></span><span class="op">(</span></span>
<span>  recipient <span class="op">=</span> <span class="va">rgeno</span>,</span>
<span>  donor <span class="op">=</span> <span class="va">dgeno</span>,</span>
<span>  return <span class="op">=</span> <span class="st">"summary"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">summary_load</span><span class="op">)</span></span>
<span><span class="co">#&gt;   hla_allele n_peptides n_strong n_weak risk_contribution</span></span>
<span><span class="co">#&gt; 1    A*02:01          0        0      0                 0</span></span>
<span><span class="co">#&gt; 2    A*03:01          0        0      0                 0</span></span>
<span><span class="co">#&gt; 3    B*07:02          0        0      0                 0</span></span>
<span><span class="co">#&gt; 4    B*08:01          0        0      0                 0</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="detailed-peptide-results">Detailed Peptide Results<a class="anchor" aria-label="anchor" href="#detailed-peptide-results"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">detailed_load</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculatePeptideBindingLoad.html">calculatePeptideBindingLoad</a></span><span class="op">(</span></span>
<span>  recipient <span class="op">=</span> <span class="va">rgeno</span>,</span>
<span>  donor <span class="op">=</span> <span class="va">dgeno</span>,</span>
<span>  return <span class="op">=</span> <span class="st">"detail"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">detailed_load</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] peptide        hla_allele     predicted_ic50 binding_level  contribution  </span></span>
<span><span class="co">#&gt; &lt;0 rows&gt; (or 0-length row.names)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="backends">Backends<a class="anchor" aria-label="anchor" href="#backends"></a>
</h2>
<div class="section level3">
<h3 id="available-backends">Available Backends<a class="anchor" aria-label="anchor" href="#available-backends"></a>
</h3>
<p><code><a href="../reference/calculatePeptideBindingLoad.html">calculatePeptideBindingLoad()</a></code> supports multiple
prediction backends:</p>
<table class="table">
<thead><tr class="header">
<th>Backend</th>
<th>Description</th>
<th>Requirements</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>pwm</code></td>
<td>Position Weight Matrix (default)</td>
<td>R only</td>
</tr>
<tr class="even">
<td><code>mhcnuggets</code></td>
<td>Deep learning model</td>
<td>Python</td>
</tr>
<tr class="odd">
<td><code>netmhcpan</code></td>
<td>External tool</td>
<td>NetMHCpan installed</td>
</tr>
<tr class="even">
<td><code>mhcflurry</code></td>
<td>Deep learning model</td>
<td>Python</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="using-pwm-default">Using PWM (Default)<a class="anchor" aria-label="anchor" href="#using-pwm-default"></a>
</h3>
<p>The PWM backend uses pre-computed position-specific scoring matrices
based on HLA supertypes:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result_pwm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculatePeptideBindingLoad.html">calculatePeptideBindingLoad</a></span><span class="op">(</span></span>
<span>  recipient <span class="op">=</span> <span class="va">rgeno</span>,</span>
<span>  donor <span class="op">=</span> <span class="va">dgeno</span>,</span>
<span>  backend <span class="op">=</span> <span class="st">"pwm"</span>,</span>
<span>  return <span class="op">=</span> <span class="st">"summary"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">result_pwm</span><span class="op">)</span></span>
<span><span class="co">#&gt;   hla_allele n_peptides n_strong n_weak risk_contribution</span></span>
<span><span class="co">#&gt; 1    A*02:01          0        0      0                 0</span></span>
<span><span class="co">#&gt; 2    A*03:01          0        0      0                 0</span></span>
<span><span class="co">#&gt; 3    B*07:02          0        0      0                 0</span></span>
<span><span class="co">#&gt; 4    B*08:01          0        0      0                 0</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="customizing-thresholds">Customizing Thresholds<a class="anchor" aria-label="anchor" href="#customizing-thresholds"></a>
</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Adjust binding thresholds</span></span>
<span><span class="va">result_custom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculatePeptideBindingLoad.html">calculatePeptideBindingLoad</a></span><span class="op">(</span></span>
<span>  recipient <span class="op">=</span> <span class="va">rgeno</span>,</span>
<span>  donor <span class="op">=</span> <span class="va">dgeno</span>,</span>
<span>  binding_threshold <span class="op">=</span> <span class="fl">100</span>,   <span class="co"># Strong binder threshold</span></span>
<span>  weak_threshold <span class="op">=</span> <span class="fl">1000</span>,     <span class="co"># Weak binder threshold</span></span>
<span>  return <span class="op">=</span> <span class="st">"summary"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">result_custom</span><span class="op">)</span></span>
<span><span class="co">#&gt;   hla_allele n_peptides n_strong n_weak risk_contribution</span></span>
<span><span class="co">#&gt; 1    A*02:01          0        0      0                 0</span></span>
<span><span class="co">#&gt; 2    A*03:01          0        0      0                 0</span></span>
<span><span class="co">#&gt; 3    B*07:02          0        0      0                 0</span></span>
<span><span class="co">#&gt; 4    B*08:01          0        0      0                 0</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="clinical-applications">Clinical Applications<a class="anchor" aria-label="anchor" href="#clinical-applications"></a>
</h2>
<div class="section level3">
<h3 id="transplant-risk-assessment">Transplant Risk Assessment<a class="anchor" aria-label="anchor" href="#transplant-risk-assessment"></a>
</h3>
<p>Peptide binding load can complement eplet-based analysis:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Eplet load:</strong> Predicts antibody-mediated rejection
risk</li>
<li>
<strong>Peptide binding load:</strong> Predicts T cell-mediated
rejection risk</li>
</ol>
</div>
<div class="section level3">
<h3 id="integration-with-other-metrics">Integration with Other Metrics<a class="anchor" aria-label="anchor" href="#integration-with-other-metrics"></a>
</h3>
<p>Consider combining peptide binding predictions with:</p>
<ul>
<li>HLA mismatch count</li>
<li>Eplet mismatch load</li>
<li>Pre-existing antibody profiles</li>
<li>Clinical risk factors</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="function-reference">Function Reference<a class="anchor" aria-label="anchor" href="#function-reference"></a>
</h2>
<div class="section level3">
<h3 id="predictmhcnuggets">
<code>predictMHCnuggets()</code><a class="anchor" aria-label="anchor" href="#predictmhcnuggets"></a>
</h3>
<table class="table">
<thead><tr class="header">
<th>Argument</th>
<th>Default</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>peptides</code></td>
<td>–</td>
<td>Character vector of peptide sequences</td>
</tr>
<tr class="even">
<td><code>allele</code></td>
<td>–</td>
<td>HLA allele (e.g., “A*02:01”)</td>
</tr>
<tr class="odd">
<td><code>mhc_class</code></td>
<td><code>"I"</code></td>
<td>MHC class: “I” or “II”</td>
</tr>
<tr class="even">
<td><code>rank_output</code></td>
<td><code>FALSE</code></td>
<td>Include percentile rank</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="calculatepeptidebindingload">
<code>calculatePeptideBindingLoad()</code><a class="anchor" aria-label="anchor" href="#calculatepeptidebindingload"></a>
</h3>
<table class="table">
<thead><tr class="header">
<th>Argument</th>
<th>Default</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>recipient</code></td>
<td>–</td>
<td>Recipient genotype or allele vector</td>
</tr>
<tr class="even">
<td><code>donor</code></td>
<td>–</td>
<td>Donor peptides or genotype</td>
</tr>
<tr class="odd">
<td><code>backend</code></td>
<td><code>"pwm"</code></td>
<td>Prediction backend</td>
</tr>
<tr class="even">
<td><code>peptide_length</code></td>
<td><code>9L</code></td>
<td>Peptide length for analysis</td>
</tr>
<tr class="odd">
<td><code>binding_threshold</code></td>
<td><code>500</code></td>
<td>IC50 threshold for strong binding</td>
</tr>
<tr class="even">
<td><code>weak_threshold</code></td>
<td><code>5000</code></td>
<td>IC50 threshold for weak binding</td>
</tr>
<tr class="odd">
<td><code>return</code></td>
<td><code>"total"</code></td>
<td>Output type: “total”, “summary”, “detail”</td>
</tr>
<tr class="even">
<td><code>aggregate_method</code></td>
<td><code>"sum"</code></td>
<td>Aggregation: “sum”, “max”, “mean”</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.0 (2025-04-11)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">#&gt; Running under: macOS Sonoma 14.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/Chicago</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] dplyr_1.1.4       deepMatchR_0.99.0 BiocStyle_2.36.0 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] gtable_0.3.6            dir.expiry_1.16.0       xfun_0.55              </span></span>
<span><span class="co">#&gt;  [4] bslib_0.9.0             ggplot2_4.0.1           htmlwidgets_1.6.4      </span></span>
<span><span class="co">#&gt;  [7] lattice_0.22-7          vctrs_0.6.5             tools_4.5.0            </span></span>
<span><span class="co">#&gt; [10] generics_0.1.4          stats4_4.5.0            parallel_4.5.0         </span></span>
<span><span class="co">#&gt; [13] tibble_3.3.0            pkgconfig_2.0.3         Matrix_1.7-4           </span></span>
<span><span class="co">#&gt; [16] data.table_1.18.0       RColorBrewer_1.1-3      S7_0.2.1               </span></span>
<span><span class="co">#&gt; [19] desc_1.4.3              S4Vectors_0.46.0        readxl_1.4.5           </span></span>
<span><span class="co">#&gt; [22] lifecycle_1.0.4         GenomeInfoDbData_1.2.14 compiler_4.5.0         </span></span>
<span><span class="co">#&gt; [25] farver_2.1.2            textshaping_1.0.4       Biostrings_2.76.0      </span></span>
<span><span class="co">#&gt; [28] GenomeInfoDb_1.44.3     htmltools_0.5.9         sass_0.4.10            </span></span>
<span><span class="co">#&gt; [31] yaml_2.3.12             pillar_1.11.1           pkgdown_2.2.0          </span></span>
<span><span class="co">#&gt; [34] crayon_1.5.3            jquerylib_0.1.4         cachem_1.1.0           </span></span>
<span><span class="co">#&gt; [37] basilisk_1.20.0         tidyselect_1.2.1        rvest_1.0.5            </span></span>
<span><span class="co">#&gt; [40] digest_0.6.39           bookdown_0.46           fastmap_1.2.0          </span></span>
<span><span class="co">#&gt; [43] grid_4.5.0              cli_3.6.5               magrittr_2.0.4         </span></span>
<span><span class="co">#&gt; [46] patchwork_1.3.2         withr_3.0.2             filelock_1.0.3         </span></span>
<span><span class="co">#&gt; [49] scales_1.4.0            UCSC.utils_1.4.0        pwalign_1.4.0          </span></span>
<span><span class="co">#&gt; [52] rmarkdown_2.30          XVector_0.48.0          httr_1.4.7             </span></span>
<span><span class="co">#&gt; [55] otel_0.2.0              reticulate_1.44.1       cellranger_1.1.0       </span></span>
<span><span class="co">#&gt; [58] ragg_1.5.0              png_0.1-8               memoise_2.0.1          </span></span>
<span><span class="co">#&gt; [61] evaluate_1.0.5          knitr_1.51              IRanges_2.42.0         </span></span>
<span><span class="co">#&gt; [64] basilisk.utils_1.20.0   immReferent_0.99.0      rlang_1.1.6            </span></span>
<span><span class="co">#&gt; [67] Rcpp_1.1.0              glue_1.8.0              BiocManager_1.30.27    </span></span>
<span><span class="co">#&gt; [70] xml2_1.5.1              BiocGenerics_0.54.1     rstudioapi_0.17.1      </span></span>
<span><span class="co">#&gt; [73] jsonlite_2.0.0          R6_2.6.1                systemfonts_1.3.1      </span></span>
<span><span class="co">#&gt; [76] fs_1.6.6</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/ncborcherding" class="external-link">Nick Borcherding</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
