<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Sequence &amp; Mismatch Analysis • deepMatchR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Sequence &amp; Mismatch Analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">deepMatchR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/deepMatchR.html">Get Started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/index.html">Overview</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Core Workflows</h6></li>
    <li><a class="dropdown-item" href="../articles/sequence-analysis.html">Sequence &amp; Mismatch Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/eplet-analysis.html">Eplet Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/peptide-binding.html">Peptide Binding Prediction</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Visualization</h6></li>
    <li><a class="dropdown-item" href="../articles/antibody-visualization.html">Antibody Visualization</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/BorchLab/deepMatchR" aria-label="GitHub"><span class="fa fab fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Sequence &amp; Mismatch Analysis</h1>
                        <h4 data-toc-skip class="author">Nick
Borcherding</h4>
            <address class="author_afil">
      Washington University in St. Louis, School of Medicine, St. Louis,
MO, USA<br><h4 data-toc-skip class="date">Compiled: January 27,
2026</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BorchLab/deepMatchR/blob/main/vignettes/sequence-analysis.Rmd" class="external-link"><code>vignettes/sequence-analysis.Rmd</code></a></small>
      <div class="d-none name"><code>sequence-analysis.Rmd</code></div>
    </address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This article covers sequence-level HLA analysis in
<strong>deepMatchR</strong>, including:</p>
<ul>
<li>Retrieving allele sequences from IMGT/HLA</li>
<li>Quantifying amino acid mismatches between sequences</li>
<li>Calculating mismatch load between donor-recipient pairs</li>
</ul>
<p>These tools provide the foundation for understanding immunogenicity
at the molecular level.</p>
</div>
<div class="section level2">
<h2 id="loading-the-package">Loading the Package<a class="anchor" aria-label="anchor" href="#loading-the-package"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.borch.dev/uploads/deepMatchR">deepMatchR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="retrieving-hla-sequences">Retrieving HLA Sequences<a class="anchor" aria-label="anchor" href="#retrieving-hla-sequences"></a>
</h2>
<div class="section level3">
<h3 id="basic-usage">Basic Usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h3>
<p>The <code><a href="../reference/getAlleleSequence.html">getAlleleSequence()</a></code> function retrieves protein
sequences from the IMGT/HLA database.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the amino acid sequence for A*01:01</span></span>
<span><span class="va">a0101_seq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getAlleleSequence.html">getAlleleSequence</a></span><span class="op">(</span><span class="st">"A*01:01"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View first 60 amino acids</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="va">a0101_seq</span>, <span class="fl">1</span>, <span class="fl">60</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRFIAVGYVDDTQFVRF"</span></span>
<span></span>
<span><span class="co"># Full sequence length</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">a0101_seq</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 365</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="batch-retrieval">Batch Retrieval<a class="anchor" aria-label="anchor" href="#batch-retrieval"></a>
</h3>
<p>For multiple alleles, use <code><a href="../reference/batchGetSequences.html">batchGetSequences()</a></code> for
efficient parallel retrieval:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alleles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A*01:01"</span>, <span class="st">"A*02:01"</span>, <span class="st">"B*07:02"</span><span class="op">)</span></span>
<span><span class="va">sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/batchGetSequences.html">batchGetSequences</a></span><span class="op">(</span><span class="va">alleles</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View sequence lengths</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">sequences</span>, <span class="va">nchar</span><span class="op">)</span></span>
<span><span class="co">#&gt; A*01:01 A*02:01 B*07:02 </span></span>
<span><span class="co">#&gt;     365     365     362</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="quantifying-amino-acid-mismatches">Quantifying Amino Acid Mismatches<a class="anchor" aria-label="anchor" href="#quantifying-amino-acid-mismatches"></a>
</h2>
<div class="section level3">
<h3 id="basic-mismatch-counting">Basic Mismatch Counting<a class="anchor" aria-label="anchor" href="#basic-mismatch-counting"></a>
</h3>
<p>The <code><a href="../reference/quantifyMismatch.html">quantifyMismatch()</a></code> function compares two protein
sequences and counts differences:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seq1</span> <span class="op">&lt;-</span> <span class="st">"YFAMYGEKVAHTHVDTLYVRYHY"</span></span>
<span><span class="va">seq2</span> <span class="op">&lt;-</span> <span class="st">"YFDMYGEKVAHTHVDTLYVRFHY"</span></span>
<span></span>
<span><span class="co"># Count all mismatches</span></span>
<span><span class="fu"><a href="../reference/quantifyMismatch.html">quantifyMismatch</a></span><span class="op">(</span><span class="va">seq1</span>, <span class="va">seq2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 2</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="biophysical-property-filters">Biophysical Property Filters<a class="anchor" aria-label="anchor" href="#biophysical-property-filters"></a>
</h3>
<p>You can filter mismatches by biophysical properties to focus on
functionally significant changes:</p>
<div class="section level4">
<h4 id="charge-classification">Charge Classification<a class="anchor" aria-label="anchor" href="#charge-classification"></a>
</h4>
<ul>
<li>
<strong>Positive:</strong> K, R, H</li>
<li>
<strong>Negative:</strong> D, E</li>
<li>
<strong>Neutral:</strong> All others</li>
</ul>
</div>
<div class="section level4">
<h4 id="polarity-classification">Polarity Classification<a class="anchor" aria-label="anchor" href="#polarity-classification"></a>
</h4>
<ul>
<li>
<strong>Nonpolar:</strong> A, V, L, I, P, M, F, W, G</li>
<li>
<strong>Polar:</strong> S, T, N, Q, Y, C, H, K, R, D, E</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Only charge-changing mismatches</span></span>
<span><span class="fu"><a href="../reference/quantifyMismatch.html">quantifyMismatch</a></span><span class="op">(</span><span class="va">seq1</span>, <span class="va">seq2</span>, filter_charge <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1</span></span>
<span></span>
<span><span class="co"># Only polarity-changing mismatches</span></span>
<span><span class="fu"><a href="../reference/quantifyMismatch.html">quantifyMismatch</a></span><span class="op">(</span><span class="va">seq1</span>, <span class="va">seq2</span>, filter_polarity <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 2</span></span>
<span></span>
<span><span class="co"># Both charge AND polarity changing</span></span>
<span><span class="fu"><a href="../reference/quantifyMismatch.html">quantifyMismatch</a></span><span class="op">(</span><span class="va">seq1</span>, <span class="va">seq2</span>, filter_charge <span class="op">=</span> <span class="cn">TRUE</span>, filter_polarity <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="detailed-position-analysis">Detailed Position Analysis<a class="anchor" aria-label="anchor" href="#detailed-position-analysis"></a>
</h3>
<p>Get position-by-position breakdown:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">details</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantifyMismatch.html">quantifyMismatch</a></span><span class="op">(</span><span class="va">seq1</span>, <span class="va">seq2</span>, return <span class="op">=</span> <span class="st">"detail"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">details</span><span class="op">)</span></span>
<span><span class="co">#&gt;   alignment_position ref alt is_gap_ref is_gap_alt is_mismatch charge_ref</span></span>
<span><span class="co">#&gt; 1                  1   Y   Y      FALSE      FALSE       FALSE        neu</span></span>
<span><span class="co">#&gt; 2                  2   F   F      FALSE      FALSE       FALSE        neu</span></span>
<span><span class="co">#&gt; 3                  3   A   D      FALSE      FALSE        TRUE        neu</span></span>
<span><span class="co">#&gt; 4                  4   M   M      FALSE      FALSE       FALSE        neu</span></span>
<span><span class="co">#&gt; 5                  5   Y   Y      FALSE      FALSE       FALSE        neu</span></span>
<span><span class="co">#&gt; 6                  6   G   G      FALSE      FALSE       FALSE        neu</span></span>
<span><span class="co">#&gt;   charge_alt charge_change polarity_ref polarity_alt polarity_change counted</span></span>
<span><span class="co">#&gt; 1        neu         FALSE        polar        polar           FALSE   FALSE</span></span>
<span><span class="co">#&gt; 2        neu         FALSE     nonpolar     nonpolar           FALSE   FALSE</span></span>
<span><span class="co">#&gt; 3        neg          TRUE     nonpolar        polar            TRUE    TRUE</span></span>
<span><span class="co">#&gt; 4        neu         FALSE     nonpolar     nonpolar           FALSE   FALSE</span></span>
<span><span class="co">#&gt; 5        neu         FALSE        polar        polar           FALSE   FALSE</span></span>
<span><span class="co">#&gt; 6        neu         FALSE     nonpolar     nonpolar           FALSE   FALSE</span></span></code></pre></div>
<p>The detail output includes:</p>
<table class="table">
<thead><tr class="header">
<th>Column</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>position</code></td>
<td>Position in the sequence</td>
</tr>
<tr class="even">
<td>
<code>aa1</code> / <code>aa2</code>
</td>
<td>Amino acid in each sequence</td>
</tr>
<tr class="odd">
<td><code>is_mismatch</code></td>
<td>Whether amino acids differ</td>
</tr>
<tr class="even">
<td><code>charge_change</code></td>
<td>Whether charge class changed</td>
</tr>
<tr class="odd">
<td><code>polarity_change</code></td>
<td>Whether polarity changed</td>
</tr>
<tr class="even">
<td><code>counted</code></td>
<td>Whether position counts toward total</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="handling-unknown-residues">Handling Unknown Residues<a class="anchor" aria-label="anchor" href="#handling-unknown-residues"></a>
</h3>
<p>The <code>na_action</code> parameter controls how unknown residues
(X, -, etc.) are handled:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seq_with_unknown</span> <span class="op">&lt;-</span> <span class="st">"YFAMYGEKVAHTHVDTLYVRXHY"</span></span>
<span></span>
<span><span class="co"># Exclude unknown positions (default)</span></span>
<span><span class="fu"><a href="../reference/quantifyMismatch.html">quantifyMismatch</a></span><span class="op">(</span><span class="va">seq1</span>, <span class="va">seq_with_unknown</span>, na_action <span class="op">=</span> <span class="st">"exclude"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1</span></span>
<span></span>
<span><span class="co"># Count unknowns as mismatches</span></span>
<span><span class="fu"><a href="../reference/quantifyMismatch.html">quantifyMismatch</a></span><span class="op">(</span><span class="va">seq1</span>, <span class="va">seq_with_unknown</span>, na_action <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="donor-recipient-mismatch-load">Donor-Recipient Mismatch Load<a class="anchor" aria-label="anchor" href="#donor-recipient-mismatch-load"></a>
</h2>
<div class="section level3">
<h3 id="creating-hla-genotypes">Creating HLA Genotypes<a class="anchor" aria-label="anchor" href="#creating-hla-genotypes"></a>
</h3>
<p>First, create genotype objects using <code><a href="../reference/hlaGeno.html">hlaGeno()</a></code>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">recipient</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  A_1 <span class="op">=</span> <span class="st">"A*01:01"</span>, A_2 <span class="op">=</span> <span class="st">"A*02:01"</span>,</span>
<span>  B_1 <span class="op">=</span> <span class="st">"B*07:02"</span>, B_2 <span class="op">=</span> <span class="st">"B*08:01"</span>,</span>
<span>  DRB1_1 <span class="op">=</span> <span class="st">"DRB1*03:01"</span>, DRB1_2 <span class="op">=</span> <span class="st">"DRB1*04:01"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">donor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  A_1 <span class="op">=</span> <span class="st">"A*03:01"</span>, A_2 <span class="op">=</span> <span class="st">"A*24:02"</span>,</span>
<span>  B_1 <span class="op">=</span> <span class="st">"B*44:02"</span>, B_2 <span class="op">=</span> <span class="st">"B*51:01"</span>,</span>
<span>  DRB1_1 <span class="op">=</span> <span class="st">"DRB1*07:01"</span>, DRB1_2 <span class="op">=</span> <span class="st">"DRB1*11:01"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">rgeno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hlaGeno.html">hlaGeno</a></span><span class="op">(</span><span class="va">recipient</span><span class="op">)</span></span>
<span><span class="va">dgeno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hlaGeno.html">hlaGeno</a></span><span class="op">(</span><span class="va">donor</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="calculating-total-mismatch-load">Calculating Total Mismatch Load<a class="anchor" aria-label="anchor" href="#calculating-total-mismatch-load"></a>
</h3>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/calculateMismatchLoad.html">calculateMismatchLoad</a></span><span class="op">(</span><span class="va">rgeno</span>, <span class="va">dgeno</span>, parallel <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 310</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="per-locus-breakdown">Per-Locus Breakdown<a class="anchor" aria-label="anchor" href="#per-locus-breakdown"></a>
</h3>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">per_locus</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateMismatchLoad.html">calculateMismatchLoad</a></span><span class="op">(</span></span>
<span>  <span class="va">rgeno</span>, <span class="va">dgeno</span>,</span>
<span>  return <span class="op">=</span> <span class="st">"per_locus"</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">per_locus</span></span>
<span><span class="co">#&gt;   locus mismatch_load</span></span>
<span><span class="co">#&gt; 1     A            96</span></span>
<span><span class="co">#&gt; 2     B           128</span></span>
<span><span class="co">#&gt; 3  DRB1            86</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="pairwise-allele-matrix">Pairwise Allele Matrix<a class="anchor" aria-label="anchor" href="#pairwise-allele-matrix"></a>
</h3>
<p>View mismatches between each recipient-donor allele pair:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mB</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateMismatchLoad.html">calculateMismatchLoad</a></span><span class="op">(</span></span>
<span>  <span class="va">rgeno</span>, <span class="va">dgeno</span>,</span>
<span>  return <span class="op">=</span> <span class="st">"pairwise"</span>,</span>
<span>  pairwise_locus <span class="op">=</span> <span class="st">"B"</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">mB</span></span>
<span><span class="co">#&gt;          donor</span></span>
<span><span class="co">#&gt; recipient B*44:02 B*51:01</span></span>
<span><span class="co">#&gt;   B*07:02      37      33</span></span>
<span><span class="co">#&gt;   B*08:01      31      27</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="applying-filters">Applying Filters<a class="anchor" aria-label="anchor" href="#applying-filters"></a>
</h3>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Charge-changing mismatches only</span></span>
<span><span class="fu"><a href="../reference/calculateMismatchLoad.html">calculateMismatchLoad</a></span><span class="op">(</span><span class="va">rgeno</span>, <span class="va">dgeno</span>, filter_charge <span class="op">=</span> <span class="cn">TRUE</span>, parallel <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 123</span></span>
<span></span>
<span><span class="co"># Polarity-changing mismatches only</span></span>
<span><span class="fu"><a href="../reference/calculateMismatchLoad.html">calculateMismatchLoad</a></span><span class="op">(</span><span class="va">rgeno</span>, <span class="va">dgeno</span>, filter_polarity <span class="op">=</span> <span class="cn">TRUE</span>, parallel <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 108</span></span>
<span></span>
<span><span class="co"># Restrict to specific loci</span></span>
<span><span class="fu"><a href="../reference/calculateMismatchLoad.html">calculateMismatchLoad</a></span><span class="op">(</span><span class="va">rgeno</span>, <span class="va">dgeno</span>, loci <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, parallel <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 224</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="function-reference">Function Reference<a class="anchor" aria-label="anchor" href="#function-reference"></a>
</h2>
<div class="section level3">
<h3 id="quantifymismatch">
<code>quantifyMismatch()</code><a class="anchor" aria-label="anchor" href="#quantifymismatch"></a>
</h3>
<table class="table">
<colgroup>
<col width="31%">
<col width="28%">
<col width="40%">
</colgroup>
<thead><tr class="header">
<th>Argument</th>
<th>Default</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>
<code>sequence1</code>, <code>sequence2</code>
</td>
<td>–</td>
<td>Two protein sequences (same length)</td>
</tr>
<tr class="even">
<td><code>filter_charge</code></td>
<td><code>NULL</code></td>
<td>Filter by charge change</td>
</tr>
<tr class="odd">
<td><code>filter_polarity</code></td>
<td><code>NULL</code></td>
<td>Filter by polarity change</td>
</tr>
<tr class="even">
<td><code>return</code></td>
<td><code>"count"</code></td>
<td>Output type: <code>"count"</code> or <code>"detail"</code>
</td>
</tr>
<tr class="odd">
<td><code>na_action</code></td>
<td><code>"exclude"</code></td>
<td>Handle unknown residues: <code>"exclude"</code>,
<code>"error"</code>, <code>"count"</code>
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="calculatemismatchload">
<code>calculateMismatchLoad()</code><a class="anchor" aria-label="anchor" href="#calculatemismatchload"></a>
</h3>
<table class="table">
<thead><tr class="header">
<th>Argument</th>
<th>Default</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>
<code>recipient_geno</code>, <code>donor_geno</code>
</td>
<td>–</td>
<td>HLA genotypes from <code><a href="../reference/hlaGeno.html">hlaGeno()</a></code>
</td>
</tr>
<tr class="even">
<td><code>loci</code></td>
<td><code>NULL</code></td>
<td>Restrict to specific loci</td>
</tr>
<tr class="odd">
<td>
<code>filter_charge</code>, <code>filter_polarity</code>
</td>
<td><code>NULL</code></td>
<td>Biophysical filters</td>
</tr>
<tr class="even">
<td><code>return</code></td>
<td><code>"total"</code></td>
<td>
<code>"total"</code>, <code>"per_locus"</code>, or
<code>"pairwise"</code>
</td>
</tr>
<tr class="odd">
<td><code>pairwise_locus</code></td>
<td><code>NULL</code></td>
<td>Locus for pairwise matrix</td>
</tr>
<tr class="even">
<td><code>parallel</code></td>
<td><code>TRUE</code></td>
<td>Use parallel processing</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.2 (2025-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] dplyr_1.1.4       deepMatchR_0.99.0 BiocStyle_2.38.0 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] sass_0.4.10         generics_0.1.4      xml2_1.5.2         </span></span>
<span><span class="co">#&gt;  [4] lattice_0.22-7      immReferent_0.99.6  digest_0.6.39      </span></span>
<span><span class="co">#&gt;  [7] magrittr_2.0.4      evaluate_1.0.5      grid_4.5.2         </span></span>
<span><span class="co">#&gt; [10] RColorBrewer_1.1-3  bookdown_0.46       fastmap_1.2.0      </span></span>
<span><span class="co">#&gt; [13] cellranger_1.1.0    jsonlite_2.0.0      Matrix_1.7-4       </span></span>
<span><span class="co">#&gt; [16] BiocManager_1.30.27 rvest_1.0.5         httr_1.4.7         </span></span>
<span><span class="co">#&gt; [19] scales_1.4.0        Biostrings_2.78.0   textshaping_1.0.4  </span></span>
<span><span class="co">#&gt; [22] jquerylib_0.1.4     cli_3.6.5           rlang_1.1.7        </span></span>
<span><span class="co">#&gt; [25] crayon_1.5.3        XVector_0.50.0      cachem_1.1.0       </span></span>
<span><span class="co">#&gt; [28] yaml_2.3.12         tools_4.5.2         dir.expiry_1.18.0  </span></span>
<span><span class="co">#&gt; [31] parallel_4.5.2      memoise_2.0.1       ggplot2_4.0.1      </span></span>
<span><span class="co">#&gt; [34] filelock_1.0.3      basilisk_1.22.0     BiocGenerics_0.56.0</span></span>
<span><span class="co">#&gt; [37] reticulate_1.44.1   vctrs_0.7.1         R6_2.6.1           </span></span>
<span><span class="co">#&gt; [40] png_0.1-8           stats4_4.5.2        lifecycle_1.0.5    </span></span>
<span><span class="co">#&gt; [43] pwalign_1.6.0       Seqinfo_1.0.0       IRanges_2.44.0     </span></span>
<span><span class="co">#&gt; [46] S4Vectors_0.48.0    fs_1.6.6            ragg_1.5.0         </span></span>
<span><span class="co">#&gt; [49] pkgconfig_2.0.3     desc_1.4.3          pkgdown_2.2.0      </span></span>
<span><span class="co">#&gt; [52] pillar_1.11.1       bslib_0.10.0        gtable_0.3.6       </span></span>
<span><span class="co">#&gt; [55] data.table_1.18.0   glue_1.8.0          Rcpp_1.1.1         </span></span>
<span><span class="co">#&gt; [58] systemfonts_1.3.1   xfun_0.56           tibble_3.3.1       </span></span>
<span><span class="co">#&gt; [61] tidyselect_1.2.1    knitr_1.51          farver_2.1.2       </span></span>
<span><span class="co">#&gt; [64] patchwork_1.3.2     htmltools_0.5.9     rmarkdown_2.30     </span></span>
<span><span class="co">#&gt; [67] compiler_4.5.2      S7_0.2.1            readxl_1.4.5</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/ncborcherding" class="external-link">Nick Borcherding</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
