<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Changelog • scRepertoire</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-news">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scRepertoire</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Loading.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/Combining_Contigs.html">Combining Contigs into Clones</a>
    </li>
    <li>
      <a href="../articles/Processing.html">Additional Processing</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/Clonal_Visualizations.html">Basic Clonal Analysis</a>
    </li>
    <li>
      <a href="../articles/Clonal_Dynamics.html">Visualizing Clonal Dynamics</a>
    </li>
    <li>
      <a href="../articles/Repertoire_Summary.html">Summarizing Repertoires</a>
    </li>
    <li>
      <a href="../articles/Clonal_Diversity.html">Clonal Diversity, Rarefaction, and Overlap</a>
    </li>
    <li>
      <a href="../articles/Clonal_Cluster.html">Clustering Sequences by Edit Distance</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/Attaching_SC.html">Combining Clones and Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/SC_Visualizations.html">Visualizations for Single-Cell Objects</a>
    </li>
    <li>
      <a href="../articles/Clonal_Bias.html">Clonal Bias</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/Trex.html">Combining Deep Learning and TCRs with Trex</a>
    </li>
    <li>
      <a href="../articles/Ibex.html">Combining Deep Learning and BCRs with Ibex</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small></small></h1>
      
    </div>

    <div class="section level2">
<h2 class="page-header" data-toc-text="2.0.0" id="screpertoire-version-200-2023-11-4">scRepertoire VERSION 2.0.0 (2023-11-4)<a class="anchor" aria-label="anchor" href="#screpertoire-version-200-2023-11-4"></a></h2>
<div class="section level3">
<h3 id="new-features-2-0-0">NEW FEATURES<a class="anchor" aria-label="anchor" href="#new-features-2-0-0"></a></h3>
<ul><li>Added <code><a href="../reference/percentAA.html">percentAA()</a></code>
</li>
<li>Added <code><a href="../reference/percentGenes.html">percentGenes()</a></code>
</li>
<li>Added <code><a href="../reference/percentVJ.html">percentVJ()</a></code>
</li>
<li>Added <code><a href="../reference/percentKmer.html">percentKmer()</a></code>
</li>
<li>Added <code><a href="../reference/exportClones.html">exportClones()</a></code>
</li>
<li>Changed compareClonotypes to <code><a href="../reference/clonalCompare.html">clonalCompare()</a></code>
</li>
<li>Changed clonotypeSizeDistribution to <code><a href="../reference/clonalSizeDistribution.html">clonalSizeDistribution()</a></code>
</li>
<li>Changed scatterClonotypes to <code><a href="../reference/clonalScatter.html">clonalScatter()</a></code>
</li>
<li>Changed quantContig to <code><a href="../reference/clonalQuant.html">clonalQuant()</a></code>
</li>
<li>Changed highlightClonotypes to <code><a href="../reference/highlightClones.html">highlightClones()</a></code>
</li>
<li>Changed lengthContigs to <code><a href="../reference/clonalLength.html">clonalLength()</a></code>
</li>
<li>Changed occupiedscRepertoire to <code><a href="../reference/clonalOccupy.html">clonalOccupy()</a></code>
</li>
<li>Changed abundanceContig to <code><a href="../reference/clonalAbundance.html">clonalAbundance()</a></code>
</li>
<li>Changed alluvialClonotypes to <code>clonalClones()</code>
</li>
<li>Added features to <code><a href="../reference/clonalCompare.html">clonalCompare()</a></code> to allow for highlighting sequences, relabeling clonotypes.</li>
</ul></div>
<div class="section level3">
<h3 id="underlying-changes-2-0-0">UNDERLYING CHANGES<a class="anchor" aria-label="anchor" href="#underlying-changes-2-0-0"></a></h3>
<ul><li>Removed internal <strong>.quiet()</strong> function.</li>
<li>
<strong>.theCall()</strong> now allows for a custom header/variable and checks the colnames.</li>
<li>Replaced data arguments to be more descriptive: <em>df</em> is now <em>input.data</em>, <em>dir</em> is now <em>input</em>, and <em>sc</em> is now <em>sc.data</em>
</li>
<li>Deep clean on the documentation for each function for increased consistency and explainability</li>
<li>
<code>Startracdiversity()</code> metric re-implemented to remove startrac-class object intermediary</li>
<li>Implemented powerTCR locally to reduce dependencies and continue support</li>
<li>Universalized underlying function language and intermediate variables</li>
<li>License change to MIT</li>
<li>
<strong>group.by</strong> and <strong>split.by</strong> have been consolidated into single <strong>group.by</strong> parameter</li>
<li>Added support for .json, Omniscope, and MiXCR formats for <code><a href="../reference/loadContigs.html">loadContigs()</a></code>
</li>
<li>Made GitHub.io website for support/vignettes/FAQ</li>
<li>Restructured NEWS Tracking</li>
<li>Added testhat for all exported and internal functions</li>
</ul></div>
<div class="section level3">
<h3 id="deprecated-and-defunct-2-0-0">DEPRECATED AND DEFUNCT<a class="anchor" aria-label="anchor" href="#deprecated-and-defunct-2-0-0"></a></h3>
<ul><li>Deprecate stripBarcodes()</li>
<li>Deprecate expression2List() (now only an internal function).</li>
<li>Deprecate checkContigs()</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.11.0" id="screpertoire-version-1110">scRepertoire VERSION 1.11.0<a class="anchor" aria-label="anchor" href="#screpertoire-version-1110"></a></h2>
<ul><li>Rebasing for the purposes of bioconductor version</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.7.5" id="screpertoire-version-175">scRepertoire VERSION 1.7.5<a class="anchor" aria-label="anchor" href="#screpertoire-version-175"></a></h2>
<ul><li>Fixed combineBCR() to allow for non-related sequences</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.7.4" id="screpertoire-version-174">scRepertoire VERSION 1.7.4<a class="anchor" aria-label="anchor" href="#screpertoire-version-174"></a></h2>
<div class="section level3">
<h3 id="new-features-1-7-4">NEW FEATURES<a class="anchor" aria-label="anchor" href="#new-features-1-7-4"></a></h3>
<ul><li>checkContigs() function to quantify the percentages of NA values by genes or sequences</li>
<li>exportClones to clonalNetwork() to isolate clones shared across identities.</li>
</ul></div>
<div class="section level3">
<h3 id="underlying-changes-1-7-4">UNDERLYING CHANGES<a class="anchor" aria-label="anchor" href="#underlying-changes-1-7-4"></a></h3>
<ul><li>Fix issue with clonalDiversity() and skipping boots</li>
<li>Fixing underlying assumptions with clonalBias()</li>
<li>Adding reads variable to parseAIRR</li>
<li>Fixing handling of samples parameter in combine contain functions</li>
<li>removed need to relevel the cloneType factor after combineExpression()</li>
<li>set up lapply() for combineBCR() and clusterTCR() - no more pairwise distance matrix calculation</li>
<li>loadContigs() support for data.frames or lists of contigs</li>
<li>Added examples for loadContigs() to test function</li>
<li>Removed requirement for T cell type designation - will combine A/B and G/D simultaneously</li>
<li>Updated combineBCR() to chunk nucleotide edit distance calculations by V gene and give option to skip edit distance calculation with call.related.clones = FALSE</li>
<li>Updated clusterTCR() to use lvcompare() function and base edit distances of V gene usage.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.7.3" id="screpertoire-version-173">scRepertoire VERSION 1.7.3<a class="anchor" aria-label="anchor" href="#screpertoire-version-173"></a></h2>
<ul><li>Fix misspellings for parse contains functions</li>
<li>Optimize WAT3R and 10x loadContigs()</li>
<li>Remove combineTRUST4 - superseded by loadContigs()</li>
<li>Added support of TRUST4 for combineBCR()</li>
<li>Added support for BD in loadContigs()</li>
<li>loadContigs() TRUST4 parsing allows for all NA values in a chain</li>
<li>combineExpression() group.by = NULL will now collapse the whole list.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.7.2" id="screpertoire-version-172">scRepertoire VERSION 1.7.2<a class="anchor" aria-label="anchor" href="#screpertoire-version-172"></a></h2>
<ul><li>ClonalDiversity() now has skip.boots to stop bootstrapping and downsampling</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.7.0" id="screpertoire-version-170">scRepertoire VERSION 1.7.0<a class="anchor" aria-label="anchor" href="#screpertoire-version-170"></a></h2>
<ul><li>Rebumping the version change with new release of Bioconductor</li>
<li>Added mean call to the heatmap of vizGenes()</li>
<li>To combineTCR, filteringMulti now checks to remove list elements with 0 cells.</li>
<li>Removed top_n() call as it is now deprecated, using slice_max() without ties.</li>
<li>Add arrange() call during parseTCR() to organize the chains</li>
<li>Correct the gd flip in the combineContig and subsequent functions</li>
<li>Removed viridis call in the clonalNetwork() function that was leading to errors</li>
<li>Matched syntax for strict clonotype in combineBCR()</li>
<li>Added group.by variable to all applicable visualizations</li>
<li>Added return.boots to clonalDiversity(), allow for export of all bootstrapped values</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.5.4" id="screpertoire-version-154">scRepertoire VERSION 1.5.4<a class="anchor" aria-label="anchor" href="#screpertoire-version-154"></a></h2>
<ul><li>modified grabMeta() internal function to no longer assume the active Identity is clusters.</li>
<li>checkBlanks() now checks if a blank was detected before trying to remove it</li>
<li>clonalNetwork() automatically resulted in default error message, bug now removed.</li>
<li>clonotypeBias now adds z-score of bias when matrix is exported. exportTable parameter is now fixed.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.5.3" id="screpertoire-version-153">scRepertoire VERSION 1.5.3<a class="anchor" aria-label="anchor" href="#screpertoire-version-153"></a></h2>
<ul><li>Added loadContigs for non-10X formatted single-cell data</li>
<li>removed combineTRUST4, superseded by loadContigs</li>
<li>combineTCR() now allows for &gt; 3 recovered TCRs per barcode</li>
<li>Readded the filtering steps to combineTCR(), will detect if data is from 10X and automatically remove nonproductive or multi chains.</li>
<li>Updated parseTCR() to include evaluation for gamma/delta chains.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.5.2" id="screpertoire-version-152">scRepertoire VERSION 1.5.2<a class="anchor" aria-label="anchor" href="#screpertoire-version-152"></a></h2>
<ul><li>Arbitrarily numbering system to match new bio conductor dev version</li>
<li>highlightClonotypes() now returns the specific clones instead of clonotype 1, …</li>
<li>compareClonotypes numbers parameter now for group-wide numbers and not overall top X numbers</li>
<li>Fixed issue with clonalDiversity that cause errors when group.by parameter was used.</li>
<li>modified parseBCR() to reduce complexity and assume lambda &gt;&gt; kappa</li>
<li>fixed clusterTCR() function broken with Seurat Objects</li>
<li>checkContigs no ensures data frames and that “” are converted into NAs</li>
<li>modified makeGenes() internal function changing na.omit to str_replace_na() and separating the BCR calls by chain to prevent combination errors.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.5" id="screpertoire-version-135">scRepertoire VERSION 1.3.5<a class="anchor" aria-label="anchor" href="#screpertoire-version-135"></a></h2>
<ul><li>Modified parseBCR() to check for contents of the chains. Resolve issue with placing light chain into heavy chain slots when 2 contigs are present.</li>
<li>Updated checkBlanks to include NA evaluation and placed the check in all .viz functions</li>
<li>Added clonalNetwork() function</li>
<li>Modified diversity visualization to remove outliers and place graphs on a single line</li>
<li>Modified clonalOverlay() to use new internal getCoord() function like clonalNetwork()</li>
<li>Added threshold parameter to clonesizeDistribution()</li>
<li>Added support for single-cell objects to clusterTCR()</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.4" id="screpertoire-version-134">scRepertoire VERSION 1.3.4<a class="anchor" aria-label="anchor" href="#screpertoire-version-134"></a></h2>
<ul><li>Modification in clusterTCR() and combineBCR() to speed up the comparison and use less memory</li>
<li>FilteringMulti, now isolates the top contig by chain, then for barcodes with chains &gt; 2, isolates the top expressing chains. This substantially increases the speed of the filtering step.</li>
<li>Modified makeGenes() internal function to use strings str_c()</li>
<li>Added threshold parameter to combineTRUST4 for B cell manipulation</li>
<li>Changed combineTCR function to prevent cell type mix up and clarified in function documentation.</li>
<li>vizGenes can now be used to look at other component genes of the receptor and “separate” parameter was replaced by “y.axis” parameter.</li>
<li>Added clonotypeBias() function for inter-cluster comparison.</li>
<li>Fixed clusterTCR() and combineBCR() assumption that you will have unrelated clones.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.3" id="screpertoire-version-133">scRepertoire VERSION 1.3.3<a class="anchor" aria-label="anchor" href="#screpertoire-version-133"></a></h2>
<ul><li>CombineBCR() auto naming function updated to actually name the list elements.</li>
<li>Added createHTOContigList() function to create contain list of multiplexed experiments. Fixed issue with groupBy variable</li>
<li>Added Inv.Pielou matric to diversity call - this is essentially 1-shannon/ln(length). Due to the bootstrapping the length with be constant.</li>
<li>Added include.na and split.by to occupiedscRepertoire and changed labeling depending on frequency vs proportion</li>
<li>Added support for single-cell objects for most visualizations, list organizing by single-cell object can be called using split.by variable</li>
<li>All group and groupBy parameters are now group.by.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.2" id="screpertoire-version-132">scRepertoire VERSION 1.3.2<a class="anchor" aria-label="anchor" href="#screpertoire-version-132"></a></h2>
<ul><li>This is the new numbering scheme apologies - we are all up-to-date now and now cell ranger &gt;= 5 will # work on bioconductor, so let’s all just take that as a win.</li>
<li>added dot.size parameter to scatterClonotype</li>
<li>filteringMulti now subsets clonotypes with contains &gt;=2, to prevent 2 of the same chains</li>
<li>changed how coldata is added to SCE objects using merge instead of union</li>
<li>Can now add BCR and TCR simultaneously by making large list</li>
<li>scatter plotting code is not so ugly and allows for user to select dot.size as a variable on the x or y axis</li>
<li>Removed regressClonotype function - too many dependencies required, adding an additional vignette to go through the process</li>
<li>Added chain option to visualizations and combineExpression to allow users to facilitate single chains - removed chain option from combineTCR/BCR/TRUST4 (the combined object will have both chains no matter what)</li>
<li>Added NA filter to combineTCR/BCR/TRUST4 for cell barcodes with only NA values</li>
<li>Added NA filter to expression2List() for cells with NA clonotypes.</li>
<li>Updated VizGene to order the genes automatically by highest to lowest variance</li>
<li>Updated VizGene to pull the correct genes based on selection</li>
<li>Updated parse method - old version had issue with place V–&gt;J–&gt;D in the TRB/Heavy chains</li>
<li>Simplified the clonalDiversity() to allow for more options in organizing plot and box plots.</li>
<li>CombineExpression() adds the groupBy variable to Frequency, allowing for multiple calculations to be saved in the meta data.</li>
<li>Default color scheme now uses viridis plasma, because it I am on transfusion medicine.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.2.2" id="screpertoire-version-122">scRepertoire VERSION 1.2.2<a class="anchor" aria-label="anchor" href="#screpertoire-version-122"></a></h2>
<ul><li>added the combineTRUST4 function to parse contigs from TUST4 pipeline</li>
<li>added the filter of contigs by chain in the combineTCR, combineBCR, and combineTRUST4 functions</li>
<li>no longer require the ID in the combineTCR/BCR/TRUST4 functions</li>
<li>added jaccard index for overlap analysis</li>
<li>replaced vizVgene with vizGene - allowing users to look at any gene in the combinedContig object</li>
<li>Fixed coloring scale on the overlap analysis</li>
<li>Added regressClonotype function using harmony to remove the clonotype effect on feature space</li>
<li>allowed occupiedRepertoire to use proportion.</li>
<li>added scatterClonotype function to Viz.R</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.2.1" id="screpertoire-version-121">scRepertoire VERSION 1.2.1<a class="anchor" aria-label="anchor" href="#screpertoire-version-121"></a></h2>
<ul><li>number of changes to the parseTCR/BCR functions to limit assumptions</li>
<li>Changed grabMeta to include assessment of colnames</li>
<li>fixed lengthDF handling of single chains with multi chains stored - ;</li>
<li>Added labels to alluvialClonotype and occupiedClonotype plotting</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.4" id="screpertoire-version-114">scRepertoire VERSION 1.1.4<a class="anchor" aria-label="anchor" href="#screpertoire-version-114"></a></h2>
<ul><li>replaced hammingCompare with lvCompare to enable superior clonotype calling in combineBCR function.</li>
<li>added proportion to combineExpression() function so users no longer need to know absolute frequencies when combining the contiguous information.</li>
<li>added clusterTCR() and clonalOverlay() functions.</li>
<li>added downsampling to the diversity calculations</li>
<li>replaced hammingCompare with lvCompare to enable superior clonotype calling in combineBCR function.</li>
<li>added proportion to combineExpression() function so users no longer need to know absolute frequencies when combining the contiguous information.</li>
<li>added clusterTCR() and clonalOverlay() functions.</li>
<li>added downsampling to the diversity calculations</li>
<li>Clonal Overlap Coefficient issue fixed, was comparing unique barcodes and not clonotypes</li>
<li>Added function checkBlanks to remove list elements without clonotypes, this prevents errors for visualizations</li>
<li>Re-added Startrac metrics by stripping down the package and adding it piecemeal</li>
<li>Heavily modified dependencies to reduce total number</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.0.0" id="screpertoire-version-100">scRepertoire VERSION 1.0.0<a class="anchor" aria-label="anchor" href="#screpertoire-version-100"></a></h2>
<ul><li>removed dependencies ggfittext and ggdendrogram</li>
<li>clonesizeDistribution now returns a plot() function</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.18" id="screpertoire-version-09918">scRepertoire VERSION 0.99.18<a class="anchor" aria-label="anchor" href="#screpertoire-version-09918"></a></h2>
<ul><li>Updated author information in the vignette</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.17" id="screpertoire-version-09917">scRepertoire VERSION 0.99.17<a class="anchor" aria-label="anchor" href="#screpertoire-version-09917"></a></h2>
<ul><li>Updated NEWS formatting</li>
<li>Edited DESCRIPTION to Single Cell Experiment R package</li>
<li>Updated information in the vignette</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.16" id="screpertoire-version-09916">scRepertoire VERSION 0.99.16<a class="anchor" aria-label="anchor" href="#screpertoire-version-09916"></a></h2>
<ul><li>Added getCirclize()</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.15" id="screpertoire-version-09915">scRepertoire VERSION 0.99.15<a class="anchor" aria-label="anchor" href="#screpertoire-version-09915"></a></h2>
<ul><li>Modified numerator for index function</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.14" id="screpertoire-version-09914">scRepertoire VERSION 0.99.14<a class="anchor" aria-label="anchor" href="#screpertoire-version-09914"></a></h2>
<ul><li>Removed bracket from indexing function</li>
</ul><p>#scRepertoire VERSION 0.99.13</p>
<ul><li>Added exportTable to remaining viz functions</li>
<li>Modified morisita index to correct error</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.12" id="screpertoire-version-09912">scRepertoire VERSION 0.99.12<a class="anchor" aria-label="anchor" href="#screpertoire-version-09912"></a></h2>
<ul><li>Reducing the size of the screp_example to fulfill &lt; 5 mB requirement. Randomly samples 100 cells and removed RNA counts from Seurat object</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.11" id="screpertoire-version-09911">scRepertoire VERSION 0.99.11<a class="anchor" aria-label="anchor" href="#screpertoire-version-09911"></a></h2>
<ul><li>Updated compareClonotype to allow for clonotype comparisons</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.10" id="screpertoire-version-09910">scRepertoire VERSION 0.99.10<a class="anchor" aria-label="anchor" href="#screpertoire-version-09910"></a></h2>
<ul><li>Bioconductor did not detect the version update.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.9" id="screpertoire-version-0999">scRepertoire VERSION 0.99.9<a class="anchor" aria-label="anchor" href="#screpertoire-version-0999"></a></h2>
<ul><li>Bioconductor had no love - changed the Seurat package to imports instead of required, see if that will address the compiling issue that results in a killed: 9 error.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.8" id="screpertoire-version-0998">scRepertoire VERSION 0.99.8<a class="anchor" aria-label="anchor" href="#screpertoire-version-0998"></a></h2>
<ul><li>Passed checks on system, let’s see how much bioconductor hates it</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.7" id="screpertoire-version-0997">scRepertoire VERSION 0.99.7<a class="anchor" aria-label="anchor" href="#screpertoire-version-0997"></a></h2>
<ul><li>But really this time, changed the colData import</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.6" id="screpertoire-version-0996">scRepertoire VERSION 0.99.6<a class="anchor" aria-label="anchor" href="#screpertoire-version-0996"></a></h2>
<ul><li>Changed colData import</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.5" id="screpertoire-version-0995">scRepertoire VERSION 0.99.5<a class="anchor" aria-label="anchor" href="#screpertoire-version-0995"></a></h2>
<ul><li>Added screp_example data to package</li>
<li>Added visVgene function for visualizing the distribution of V genes in TCR</li>
<li>Added support for monocle to combineExpression function</li>
<li>Updated documentation for combineTCR() and combineBCR()</li>
<li>Updated documentation to utilize SingleCellExperiment formats</li>
<li>Updated Vignette to utilize SingleCellExperiment formats</li>
<li>Added Author information to vignette</li>
<li>Add intro and conclusion to vignette</li>
<li>Removed html knitted vignette</li>
<li>Removed descriptive code snippets</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.4" id="screpertoire-version-0994">scRepertoire VERSION 0.99.4<a class="anchor" aria-label="anchor" href="#screpertoire-version-0994"></a></h2>
<ul><li>Modified expression2List() to allow for variables across meta data</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.1" id="screpertoire-version-0991">scRepertoire VERSION 0.99.1<a class="anchor" aria-label="anchor" href="#screpertoire-version-0991"></a></h2>
<ul><li>Changed R (&gt;= 3.6) to R (&gt;= 4.0)</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.0" id="screpertoire-version-0990">scRepertoire VERSION 0.99.0<a class="anchor" aria-label="anchor" href="#screpertoire-version-0990"></a></h2>
<ul><li>Changed DESCRIPTION version to 0.99.0</li>
<li>Removed file seurat_example.rda, accidentally committed</li>
<li>Deleted git attributes</li>
<li>reduced Seurat object size for alluvialClonotype in vignette</li>
<li>Changed the alluvialClonotype assessment to account for only 1 condition</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.2.3" id="screpertoire-version-123">scRepertoire VERSION 1.2.3<a class="anchor" aria-label="anchor" href="#screpertoire-version-123"></a></h2>
<ul><li>Changed the access of the sample data to github.io repo: readRDS(url(“<a href="https://ncborcherding.github.io/vignettes/scRepertoire_example.rds" class="external-link uri">https://ncborcherding.github.io/vignettes/scRepertoire_example.rds</a>”))</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.2.2" id="screpertoire-version-122-1">scRepertoire VERSION 1.2.2<a class="anchor" aria-label="anchor" href="#screpertoire-version-122-1"></a></h2>
<ul><li>Removed Startrac-based functions in order to pass build on Bioconductor.</li>
</ul><p>DEPRECATED AND DEFUNCT</p>
<ul><li>Deprecate StartracDiversity()</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.2.0" id="screpertoire-version-120">scRepertoire VERSION 1.2.0<a class="anchor" aria-label="anchor" href="#screpertoire-version-120"></a></h2>
<p>SIGNIFICANT USER-VISIBLE CHANGES</p>
<ul><li>Added support for SingleCellExperiement format.</li>
</ul><p>DEPRECATED AND DEFUNCT</p>
<ul><li>Deprecate combineSeurat in favor or combineExpression().</li>
<li>Deprecate seurat2List in favor of expression2List().</li>
</ul></div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Nick Borcherding, Qile Yang.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

